// bxuip-angular.js-ui/src/app/components/button.js
/**
 * @class bxButton
 *
 * @desc bxButton Directive
 *
 */

/**
 * bxButton Directive
 *
 * @name directive
 *
 * @private
 * @memberOf bxButton
 *
 */

/**
 * @alias directive
 *
 * @memberOf bxButton
 *
 */

const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxButton';

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function compile($scope, $element, iAttrs) {

            $element = $($element);

            const bxButton = UTIL.createCustomElement('bx-button');
            const $bxButton = $(bxButton);
            const { attrs, events } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);

            const nativeAttrs = _getNativeAttr(attrs);
            const wrapAttrs = _getWrapAttr(attrs);

            UTIL.applyAttrs($bxButton, wrapAttrs);

            $element.append($bxButton);

            const instance = $element.bxButton().initialize();
            const elem = instance.elems[0];
            const $elem = $(elem);
            const type = iAttrs.type || 'default';
            const opts = instance.opts[type];

            _.forEach($elem.find('input, button').toArray(), elem => {
                UTIL.applyAttrs($(elem), nativeAttrs);
            });

            $element.after($compile(elem)($scope));
            $element.remove();

            UTIL.addEvent($elem, events, opts);

            UTIL.callBxRender(elem, events);

        }
    };
}

/**
 * 컴포넌트 내부의 native element (input, button)에 할당할 속성을 반환하는 함수
 *
 * @param attrs
 * @returns {Object}
 * @private
 */
function _getNativeAttr(attrs = {}){

    const { type } = attrs;

    if (type === 'toggle') return _.pickBy(attrs, (v,k) => UTIL.isNgAttr(k));
    else                   return _.pickBy(attrs, (v,k) => UTIL.isNgAttr(k) || k === 'value');

}

/**
 * 컴포넌트에 할당할 속성을 반환하는 함수
 *
 * @param attrs
 * @returns {Object}
 * @private
 */
function _getWrapAttr(attrs = {}){

    const { type } = attrs;

    if (type === 'toggle') return _.pickBy(attrs, (v,k) => !UTIL.isNgAttr(k));
    else                   return _.pickBy(attrs, (v,k) => !UTIL.isNgAttr(k) && k !== 'value');

}

module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
