// bxuip-angular.js-ui/src/app/components/datepicker.js
/**
 * @class bxDatepicker
 *
 * @desc bxDatepicker Directive
 *
 */

/**
 * bxDatepicker Directive
 *
 * @name directive
 *
 * @private
 * @memberOf bxDatepicker
 *
 */

/**
 * @alias directive
 *
 * @memberOf bxDatepicker
 *
 */


const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxDatepicker';

const ATTR_NAMES = _getComponentPropertyNames();
const EVENT_NAMES = _getComponentEventNames();


function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function compile($scope, $element, iAttrs) {

            const $bxDatepicker = $('<div>');
            const { attrs, events } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);

            const componentAttr = _.pickBy(attrs, (v, k) => !UTIL.isNgAttr(k));
            const ngAttr = _.pickBy(attrs, (v, k) => UTIL.isNgAttr(k));

            _initComponent.call($bxDatepicker, {componentAttr, events, $scope, iAttrs});

            UTIL.applyAttrs($bxDatepicker.children(), ngAttr);

            $element.append($bxDatepicker);

            $element.after($compile($bxDatepicker.get(0))($scope));
            $element.remove();

        }
    };
}

/**
 * jquery version 컴포넌트 초기화 함수
 *
 * @param {Object} config
 * @returns {Object} 컴포넌트 인스턴스 객체
 * @private
 */
function _initComponent(config){

    const opts = _getComponentOptions(config);

    return this.bxDatePicker(opts).render();

}

/**
 *
 * directive attribute를 컴포넌트 초기화에 필요한 옵션 객체로 반환하는 함수
 *
 * @param componentAttr
 * @param events
 * @param $scope
 * @param iAttrs
 * @returns {Object} 컴포넌트 옵션 객체
 * @private
 */
function _getComponentOptions({componentAttr, events, $scope, iAttrs}){

    const opts = _.mapKeys(componentAttr, (v, k) => ATTR_NAMES[k.toLowerCase()]);
    const onSelect = events['bx-select'];

    _.forEach(events, (v, k) => {

        const eventName = `on${_.startCase(k.split('-').pop())}`;

        if (_.indexOf(EVENT_NAMES, eventName) !== -1){
            opts[eventName] = v;
        }

    });

    opts.onSelect = (date, ui) => {
        _.set($scope, iAttrs.ngModel, date);
        $scope.$apply();

        onSelect && onSelect.call(ui, date, ui);
    };

    return opts;
}

/**
 *
 * directive attribute 이름 객체를 반환하는 함수 (http 스펙에 의한 소문자/대문자 처리를 위함)
 *
 * @returns {Object} 컴포넌트 프로퍼티 이름 객체
 * @private
 */
function _getComponentPropertyNames(){

    const ATTR_NAMES = {};

    if ($().bxDatePicker){

        const instance = $().bxDatePicker();
        const symbolOPT = Object.getOwnPropertySymbols(instance);

        _.forEach(Object.getOwnPropertyNames(instance[symbolOPT.pop()]), (v) => {
            ATTR_NAMES[v.toLowerCase()] = v;
        });
    }

    return ATTR_NAMES;

}

/**
 *
 * directive attribute에서 bx-* 함수명을 컴포넌트 이벤트명으로 반환하는 함수
 *
 * @returns {Array} 컴포넌트 이벤트 이름 배열
 * @private
 */
function _getComponentEventNames(){

    return _.filter(ATTR_NAMES, (v, k) => _.words(v).shift() === 'on');

}


module.exports = {
    directive: {
        name: 'bxDatepicker',
        $inject: ['$compile'],
        constructor: _directive
    }
};
