// bxuip-angular.js-ui/src/app/components/fileUpload.js

const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxFileUpload';

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function ($scope, $element, iAttrs){

            $element = $($element);

            const bxFileUpload = UTIL.createCustomElement('bx-fileUpload');
            const $bxFileUpload = $(bxFileUpload);

            $element.append($bxFileUpload);

            const { events, attrs } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            UTIL.applyAttrs($bxFileUpload, attrs);

            const instance = $element.bxFileUpload().initialize();
            const elem = instance.elems[0];
            const $elem = $(elem);
            const $fileUploadWrap = $elem.parent();
            const $wrap = $elem.find('.bx-fileupload-wrap');

            $element.after($compile($fileUploadWrap.get(0))($scope));
            $element.remove();

            UTIL.addEvent($wrap, events, instance.opts);

            UTIL.callBxRender(elem, events);
        }
    };
}

module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
