// bxuip-angular.js-ui/src/app/components/img.js
const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxImg';

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function ($scope, $element, iAttrs){

            $element = $($element);

            const bxImg = UTIL.createCustomElement('bx-img');
            const $bxImg = $(bxImg);

            $element.append($bxImg);

            const { events, attrs } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);

            UTIL.applyAttrs($bxImg, attrs);

            const instance = $element.BxImg().initialize();
            const elem = instance.elems[0];
            const $elem = $(elem);
            const $wrap = $elem.parents('.-bx-ui-img');
            const wrap = $wrap.get(0);

            $element.after($compile(wrap)($scope));
            $element.remove();

            UTIL.addEvent($elem, events, instance.opts);

            UTIL.callBxRender(elem, events);
        }
    };
}

module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
