// bxuip-angular.js-ui/src/app/components/input.js
/**
 * @class bxInput
 *
 * @desc bxInput Directive
 *
 */

/**
 * bxInput Directive
 *
 * @name directive
 *
 * @private
 * @memberOf bxInput
 *
 */

/**
 * @alias directive
 *
 * @memberOf bxInput
 *
 */

const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxInput';

const INPUT_TYPES = {
    SINGLE_TYPES: ['text', 'password', 'checkbox', 'radio', 'number', 'timepicker', 'file'],
    MULTI_TYPES: ['ip', 'email', 'tel', 'pid', 'bzno']
};

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function compile($scope, $element, iAttrs) {

            $element = $($element);

            const bxInput = UTIL.createCustomElement('bx-input');
            const $bxInput = $(bxInput);
            const { attrs, events } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);
            UTIL.applyAttrs($bxInput, attrs);

            $element.append($bxInput);

            const instance = $($element).bxInput().initialize();
            const elem = instance.elems[0];
            const $elem = $(elem);
            const $inputs = $elem.find('input');
            const type = iAttrs.type || 'text';
            const opts = instance[type].opts;

            $element.after($compile(elem)($scope));
            $element.remove();

            if (_isMultipleType(type)) {
                _bindModel(iAttrs.ngModel, $scope, $elem);
                UTIL.addEvent($elem, events, opts);
            }
            else {
                UTIL.addEvent($inputs, events, opts);
            }

            UTIL.callBxRender(elem, events);

        }
    };
}

/**
 * multi type 여부 반환 함수
 * @param {string} type - type 명
 * @returns {boolean}
 * @private
 */
function _isMultipleType(type = 'text') {
    return _.indexOf(INPUT_TYPES.MULTI_TYPES, type) > -1;
}

/**
 *
 * multi type (ip, email, tel....)에 angular의 twoway binding을 설정하는 함수
 *
 * @param {string} modelPath - $scope에 할당된 model의 path
 * @param {object} $scope - $scope 객체
 * @param {jQuery|HTMLElement} $elem - component jquery 객체
 * @private
 */
function _bindModel(
    modelPath = '',
    $scope = null,
    $elem = null
){

    const targetElem = $elem.children().get(0);
    const $inputs = $elem.find('bx-input');
    const $bxSelect = $elem.find('.bx-select-wrap');

    if (!modelPath) return;

    targetElem.value = _.get($scope, modelPath);

    $scope.$watch(modelPath, (newValue, oldValue) => {
        if (newValue === oldValue) return;
        targetElem.value = newValue;
    }, true);

    $inputs.on('change', () => {
        _.set($scope, modelPath, targetElem.value);
        $scope.$apply();
    });

    if ($bxSelect.length){
        _.forEach($bxSelect.toArray(), (select) => {
            select.bxhide = () => {
                _.set($scope, modelPath, targetElem.value);
                $scope.$apply();
            };
        });
    }

}

module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
