// bxuip-angular.js-ui/src/app/components/progress.js
/**
 * @class bxProgress
 *
 * @desc bxProgress Directive
 *
 */

/**
 * bxProgress Directive
 *
 * @name directive
 *
 * @private
 * @memberOf bxProgress
 *
 */

/**
 * @alias directive
 *
 * @memberOf bxProgress
 *
 */

const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxProgress';

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function compile($scope, $element, iAttrs) {

            $element = $($element);

            const bxProgress = UTIL.createCustomElement('bx-progress');
            const $bxProgress = $(bxProgress);
            const modelPath = iAttrs.ngModel;

            const { attrs, events } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            if (_.get($scope, modelPath)) attrs.defaultValue = _.get($scope, modelPath);

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);
            UTIL.applyAttrs($bxProgress, attrs);

            $element.append($bxProgress);

            const instance = $element.bxProgress().initialize();
            const elem = instance.elems[0];

            $element.after($compile(elem)($scope));
            $element.remove();

            UTIL.addEvent($(elem), events, instance.opts.common);

            _bindModel(modelPath, $scope, elem);

            UTIL.callBxRender(elem, events);
        }
    };
}

/**
 *
 * angular의 twoway binding을 설정하는 함수
 *
 * @param {string} modelPath - $scope에 할당된 model의 path
 * @param {object} $scope - $scope 객체
 * @param {jQuery|HTMLElement} $elem - component jquery 객체
 * @private
 */
function _bindModel(
    modelPath = '',
    $scope = null,
    elem = null
){

    if (!modelPath) return;

    const changedEvent = elem.bxchange;

    $scope.$watch(modelPath, (newValue, oldValue) => {
        if (newValue === oldValue) return;
        elem.defaultValue = newValue;
    }, true);

    elem.bxchange = v => {
        _.set($scope, modelPath, v);

        // slider 타입의 경우, 사용자 interaction change 이벤트가 다수 발생하여
        // $digest 루프가 종료되기 전에 $apply가 호출되는 경우가 발생함.
        // 따라서 비동기 처리함
        setTimeout(() => $scope.$apply());
        changedEvent && changedEvent.call(elem);
    };

}


module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
