// bxuip-angular.js-ui/src/app/components/select.js


const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxSelect';

const ELEMENT_NAME = {
    "select": 'bx-select',
    "group": 'bx-optGroup',
    "option": 'bx-option'
};

function _directive() {

    return {
        restrict: 'E',
        require: ['?ngModel'],
        link: function ($scope, $element, iAttrs, ngModel) {

            $element = $($element);

            const key = iAttrs.$attr.items;
            const name = iAttrs[key];
            let $prevElement = null;
            
            $scope.$watch(name, (newValue, oldValue) => {

                const items = newValue;

                if (!_.isArray(items)) {
                    BXUI.Console.prop.throw(_directive, 'items', 'type');
                }

                const { events, attrs } = UTIL.separateEventsAndAttrs($scope, iAttrs);

                const select = UTIL.createCustomElement(ELEMENT_NAME.select);
                const $select = $(select);

                UTIL.applyAttrs($select, attrs);

                _createSelectElement({
                    children: items
                }, $select);

                $element.append($select);

                const instance = $element.bxSelect().initialize();
                const elem = instance.elems[0];
                const $elem = $(elem);
                const $selectWrap = $elem.parent();

                if (_.isNull($prevElement)) {
                    $element.after($selectWrap);
                    $element.remove();
                }
                else {
                    $prevElement.after($selectWrap);
                    $prevElement.remove();
                }

                $prevElement = $selectWrap;

                if (events.hasOwnProperty('bx-selected')) {
                    _wrapBxSelected(elem, ngModel, events);
                }

                UTIL.addEvent($elem, events, instance.opts);

                UTIL.callBxRender(elem, events);
            }, true);
        }
    };
}

/**
 * element 생성함수
 * @param items
 * @param $elem
 * @private
 */
function _createSelectElement(items = {}, $elem = null) {

    let $child;

    _addAttribute(items, $elem);

    _.forEach(items.children, (v) => {
        if (v.isGroup) {
            const child = UTIL.createCustomElement(ELEMENT_NAME.group);
            $child = $(child);
        }
        else {
            const child = UTIL.createCustomElement(ELEMENT_NAME.option);
            $child = $(child);
        }
        $elem.append($child);
        _createSelectElement(v, $child);
    });
}

/**
 *
 * @param items
 * @param $elem
 * @private
 */
function _addAttribute(items, $elem) {
    _.forEach(items, (v, k) => {
        if (k === 'children' || k === 'isGroup') return;

        $elem.get(0).setAttribute(k, v);
    });
}

/**
 * bxSelected wraping 함수
 * @param ngModel
 * @param events
 * @private
 */
function _wrapBxSelected(elem = null, ngModel = null, events = {}){

    if (!ngModel) return;

    const prevSelect = events['bx-selected'];

    events['bx-selected'] = function(type, state){
        const value  = this.value;
        ngModel[0].$setViewValue(value);
        prevSelect && prevSelect.call(this, type, state);
    };
}

module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: [],
        constructor: _directive
    }
};
