// bxuip-angular.js-ui/src/app/components/tab.js
/**
 * 탭 선택 시, 호출될 callback 함수
 * @callback onActivate
 *
 * @param {object} data - 생성된 엘리먼트 집합
 * @param {object} data.$elem - 선택된 탭의 인스턴스
 * @param {number} data.idx - 선택된 탭의 인덱스
 * @param {string} data.fromState - 이전 state 이름
 * @param {string} data.toSate - 현재 state 이름
 * @param {object} data.fromParams - parameter 객체
 * @memberOf bxAngularTab
 */

/**
 * 랜더링 완료 시, 호출될 callback 함수
 *
 * @callback onRender
 *
 * @memberOf bxAngularTab
 */

/**
 * @class bxAngularTab
 *
 * @desc bxAngularTab Directive<br>
 * Ui-Router Guide page : {@link https://github.com/angular-ui/ui-router/wiki/URL-Routing Ui-Router}
 *
 */

/**
 * bxAngularTab Directive
 *
 * @name directive
 * @param {object} $compile 컴파일 서비스
 * @param {object} $state   ui-route state 서비스
 *
 * @private
 * @memberOf bxAngularTab
 *
 */

/**
 * @alias directive
 *
 * @memberOf bxAngularTab
 *
 * @property {Object} attribute
 * @property {object[]} attribute.items 아이템 집합
 * @property {string} attribute.items.title - 추가 탭 영역
 * @property {object} attribute.items.$state - <s>$stateProvider 에 등록한 객체</s> <span class='important'> ( Added in: v3.0.0, Deprecated since: v3.2.0 )</span>
 * @property {string} attribute.items.route - 아이템 route 이름($state 에 등록된, route 이름)
 * @property {number} attribute.height panel 높이
 * @property {string} [attribute.defaultRoute = ''] 초기 라우트 값
 * @property {function} [attribute.onActivate = function(){}] 탭 선택 시, 호출될 callback 함수
 * @property {function} [attribute.onRender = function(){}] 랜더링 완료 시, 호출될 callback 함수
 *
 */

const BXUI = require('../../lib/bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

function _directive($compile, $state) {

    return {
        replace: true,
        scope: {
            // array
            items: '=?items',
            // primitive
            height: '=?height',
            defaultRoute: '=?defaultRoute',
            // primitive(차후 append 및 remove 추가 시, 활용할 프로퍼티값)
            //maxTabCounter: '@',
            // callback function
            onActivate: '&?onActivate',
            onRender: '&?onRender'
        },
        transclude: true,
        restrict: 'E',
        templateUrl: 'bxAngularTab.tpl.cache.html',
        controller: 'bxAngularTabController',
        compile: function (iElement, iAttrs, transclude) {

            return {
                pre: function ($scope, iElement, iAttrs, controller) {

                    let $ul = iElement.find('ul');

                    $scope.setHeight(iElement, height);

                    // render
                    $scope.render($scope.items, $ul);

                    // onRender 이벤트 핸들러 함수 호출
                    $scope.onRender();

                    let stateChangeSuccess = (event, toState, toParams, fromState, fromParams) => {

                        let toStateIdx = _.findIndex($scope.items, function(item){
                            return toState.name === item.route;
                        });

                        if (toStateIdx === -1) {
                            let defaultRouteIdx = _getDefaultRouteIdx.call($scope, $scope.items, toState.name);
                            let defaultRoute = $scope.items[defaultRouteIdx].route;

                            $state.go(defaultRoute);
                        }

                        let route = toState.name;
                        let id = $scope.getIdByRoute(route);

                        let $li = $(iElement).find('#' + id);
                        let name = $li.attr('name');

                        $scope.activateTab($li);
                        $scope.onActivate({data: {
                            $elem: $li,
                            idx: parseInt(name, 10),
                            fromState: fromState.name,
                            toState: toState.name,
                            fromParams: fromParams
                        }});
                    };

                    $scope.$on('$stateChangeSuccess', stateChangeSuccess);

                },
                post: function ($scope, iElement, iAttrs, controller) {
                }
            };
        }
    };
}
/**
 * bxAngularTab 디렉티브 컨트롤러
 *
 * @param {object} $scope 스코프
 * @param {object} $state ui-route state 서비스
 * @private
 * @memberOf bxAngularTab
 */
function _controller($scope, $state){

    // type 이 안맞는 경우 err
    if (!_.isArray($scope.items)) {
        BXUI.Console.prop.throw(_directive, 'items', 'type');
    }
    // tab 은 item 에 대한 dynamic 추가가 되지 않으므로 [] 대한 체크실행
    if (_.isEmpty($scope.items)) {
        BXUI.Console.prop.throw(_directive, 'items', 'value');
    }
    // height 가 정확한 정수형일 경우에만 참조하도록
    if (_.isNaN(parseInt($scope.height))) {
        BXUI.Console.prop.throw(_directive, 'height', 'value');
    }

    // onActivate 는 function
    if (!_.isFunction($scope.onActivate)) {
        BXUI.Console.prop.throw(_directive, 'onActivate', 'type');
    }
    // onRender 는 function
    if (!_.isFunction($scope.onRender)) {
        BXUI.Console.prop.throw(_directive, 'onRender', 'type');
    }

    $scope.render = (items = [], $ul = null) => {

        if (_.isNull($ul)) return;

        let length = items.length;
        // default route 에 따른 설정
        let defaultRouteIdx = _getDefaultRouteIdx.call($scope, items, $state.current.name);
        let defaultRoute = items[defaultRouteIdx];

        console.log('_getDefaultRouteIdx', defaultRouteIdx, defaultRoute);

        _.forEach(items, (item, i) => {

            let $li = _createLiElem.call($scope, item, i);

            // 0 번째 $li 엘리먼트를 활성화 시킨다.
            if (i === 0) $li.addClass('first-tab');

            if (defaultRouteIdx === i) $li.addClass('on-tab');

            if (i === (length - 1)) $li.addClass('last-tab');

            $ul.append($li);

        });

        $state.go(defaultRoute.route);

    };

    // 이벤트 함수
    $scope.click = (e = null, route = '') => {

        if (!_.isString(route) || route === ''){
            console.error('route 값이 존재하지 않습니다.');
            return;
        }

        $state.go(route);
    };

    $scope.moveTab = (e = null, route = '') => {

        if (!_.isString(route) || route === ''){
            console.error('route 값이 존재하지 않습니다.');
            return;
        }

        $state.go(route);

    };

    // tab 업데이트 함수
    $scope.activateTab = ($elem = null) => {

        if (_.isNull($elem)) return;

        // class 추가 / 삭제
        let $ul = $elem.parent('ul');
        $ul.find('li').removeClass('on-tab');
        $elem.addClass('on-tab');
    };

    $scope.getIdByRoute = (route = '') => {

        if (!_.isString(route) || route === ''){
            console.error('route 값이 존재하지 않습니다.');
            return;
        }

        let ids = route.split('.');
        let id = ids[ids.length - 1];

        return id;
    };

    $scope.setHeight = ($elem = null, height = 0) => {

        let tabContainer = $elem.find('.tab-content');

        tabContainer.height(height);
    };
}
/**
 * bxAngularTab 디렉티브 템플릿
 *
 * @param {object} $templateCache 템플릿캐시 서비스
 * @private
 * @memberOf bxAngularTab
 */
function _template($templateCache) {

    let tpl = [];
    let className = BXUI.Base.getComponentClassName('angular-tab');

    tpl.push('<div class="' + className + '">');
    tpl.push('<section class="tab-title add-bd2-b add-mg-t">');
    tpl.push('<ul></ul>');
    tpl.push('</section>');

    tpl.push('<section class="tab-content">');
    // ui-view area
    tpl.push('<ng-transclude></ng-transclude>');
    tpl.push('</section>');
    tpl.push('</div>');

    $templateCache.put('bxAngularTab.tpl.cache.html', tpl.join(''));
}

/**
 * defaultRoute 인덱스 반환 함수
 * @desc URL, defaultRouter 순의 우선순위를 통해 결정된다.
 * @param items
 * @param defaultRoute
 * @returns {number}
 * @private
 */
function _getDefaultRouteIdx(items, currentRoute) {

    let self = this;
    let i = 0;
    let defaultIdx = -1;

    // currentRoute 와 defaultRoute 가 다른 경우, currentRoute 가 우선시 된다
    if (currentRoute !== this.defaultRoute) {
        defaultIdx = _.findIndex(items, (item) => {
            return currentRoute === item.route;
        });
    }
    // currentRoute 값이 아이템과 일치하지 않는 경우, defaultRoute 로 설정
    // 또는 currentRoute 와 default route 가 같은 경우, defaultRoute 로 설정
    if (_.isEqual(defaultIdx, -1) && !_.isUndefined(self.defaultRoute)) {
        defaultIdx = _.findIndex(items, (item) => {
            return self.defaultRoute === item.route;
        });
    }
    // item 에 존재하지도 않고, defaultRoute 도 없는 경우, 0번째 index 로 설정
    if (_.isEqual(defaultIdx, -1)) defaultIdx = 0;

    return defaultIdx;
}

/**
 * 아이템을 엘리먼트로 반환하는 함수
 *
 * @param {string} title - 메뉴 타이틀
 * @param {object} state - 메뉴 state
 *
 * @returns {jQuery|HTMLElement} li - li 엘리먼트
 * @private
 * @memberOf bxAngularTab
 */
function _createLiElem({title = '', route = {}} = {}, i = 0){

    // let {route, params, options} = state;

    let id = this.getIdByRoute(route);

    // $li 엘리먼트 생성
    let $li = $(`<li id="${id}" name="${i}"></li>`);

    let _h = [];
    _h.push(`<a href="#"> ${title}`);
    _h.push('<span class="tab-bg-before"></span>');
    _h.push('<span class="tab-bg"></span>');
    _h.push('</a>');

    // a 엘리먼트 생성
    let $a = $(_h.join(''));

    // li 엘리먼트의 자식으로 a 엘리먼트를 할당한다.
    $li.append($a);

    // 이벤트 바인딩
    $li.find('a').click((e) => {
        // controller method call
        this.click(e, route);
        // 기본 이벤트 해제
        e.preventDefault();
    });

    Mousetrap($li.get(0)).bind('left', (e) => {

        _moveTab.apply(this, [e, $(e.target).parents('ul'), 'left']);

    });

    Mousetrap($li.get(0)).bind('right', (e) => {

        _moveTab.apply(this, [e, $(e.target).parents('ul'), 'right']);

    });

    Mousetrap($li.get(0)).bind('space', (e) => {
        // controller method call
        this.click(e, route);
        // 기본 이벤트 해제
        e.preventDefault();

    });

    return $li;

}

/**
 * 키보드 동작에 따라 탭을 활성화 하는 함수
 *
 * @param {object} e - 키보드 이벤트 객체
 * @param {jQuery|HTMLElement} $tab - ui jquery 객체
 * @param {string} side - 현재 탭을 기준으로 활성화 할 탭의 방향
 *
 * @private
 * @memberOf bxAngularTab
 */
function _moveTab(e, $tab, side){

    let $activeTab = $tab.find('.on-tab');
    let idx = parseInt($activeTab.attr('name'));
    let items = this.items;

    if (side === 'left'){
        --idx;
        if (idx < 0) idx = items.length - 1;
    }
    else{
        ++idx;
        if (idx > items.length - 1) idx = 0;
    }

    $tab.find('li').removeClass('on-tab');
    $tab.find(`[name=${idx}]`).addClass('on-tab');

    // controller method call
    this.moveTab(e, items[idx].route);
}


module.exports = {
    directive: {
        name: 'bxAngularTab',
        $inject: ['$compile', '$state'],
        constructor: _directive
    },
    controller: {
        name: 'bxAngularTabController',
        $inject: ['$scope', '$state'],
        constructor: _controller
    },
    template: {
        $inject: ['$templateCache'],
        constructor: _template
    }
};
