// bxuip-angular.js-ui/src/app/components/textarea.js
const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxTextarea';

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function ($scope, $element, iAttrs){

            $element = $($element);

            const bxTextArea = UTIL.createCustomElement('bx-textarea');
            const $bxTextArea = $(bxTextArea);

            $element.append($bxTextArea);

            const { events, attrs } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);

            UTIL.applyAttrs($bxTextArea, attrs);

            const instance = $element.bxTextarea().initialize();
            const elem = instance.elems[0];
            const $textArea = $(elem).find('textarea');

            $element.after($compile(elem)($scope));
            $element.remove();

            UTIL.addEvent($textArea, events, instance.opts);

            UTIL.callBxRender(elem, events);
        }
    };
}

module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
