// bxuip-angular.js-ui/src/app/components/tooltip.js
/**
 * @class bxTooltip
 *
 * @desc bxTooltip Directive
 *
 */

/**
 * bxTooltip Directive
 *
 * @name directive
 *
 * @private
 * @memberOf bxTooltip
 *
 */

/**
 * @alias directive
 *
 * @memberOf bxTooltip
 *
 */

const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../../lib/bx-ui-core/entry');

const UTIL = require('../util');
const COMPONENT_NAME = 'bxTooltip';

function _directive($compile) {

    return {
        scope: false,
        restrict: 'E',
        link: function compile($scope, $element, iAttrs) {

            $element = $($element);

            const bxTooltip = UTIL.createCustomElement('bx-tooltip');
            const $bxTooltip = $(bxTooltip);

            const { attrs, events } = UTIL.separateEventsAndAttrs($scope, iAttrs);

            attrs.content = attrs.content || $element.text();
            $element.text('');

            UTIL.hasNgRepeat(attrs) && UTIL.deleteNgRepeat(attrs);
            UTIL.applyAttrs($bxTooltip, attrs);

            $element.append($bxTooltip);

            const instance = $element.bxTooltip().initialize();
            const elem = instance.elems[0];

            $element.after($compile(elem)($scope));
            $element.remove();

            UTIL.addEvent($(elem), events, instance.opts);

            UTIL.callBxRender(elem, events);
        }
    };
}


module.exports = {
    directive: {
        name: COMPONENT_NAME,
        $inject: ['$compile'],
        constructor: _directive
    }
};
