// bxuip-angular.js-ui/src/lib/bx-ui-core/app/base.js
/**
 * Created by sgjeon on 16. 8. 19..
 */

/**
 *
 * @namespace BXUI.Base
 *
 */

const $ = require('jquery');
const _ = require('lodash');

const CONSOLE = require('./console');
const UTIL = require('./util');

const PREFIX = '-bx-ui-';

const BASE = {
    getComponentClassName: _getComponentClassName,
    init: _init
};

window.preFocusElement = {
    base: null,
    hoverer: null,
    // "key" 값은 base, hoverer 중, 택 1 로 설정해야하며, default: 'base'
    key: 'base'
};

/**
 * document focus event 등록 함수.
 */
(function _appendBlurEventListener(){

    document.addEventListener('blur', e => {

        let $pre = window.preFocusElement;
        let key = $pre.key;
        let $target = $(e.target);

        $pre[key] = $target;

        UTIL.setPreFocusInstanceKey('base');

    }, true);

})();

/**
 * 최종 컴포넌트 클래스 이름을 반환한다.
 *
 * @param {String} className - 클래스 이름
 *
 * @returns {string} - 생성된 컴포넌트 클래스 이름
 *
 * @private
 *
 * @memberof BXUI.Base
 */

function _getComponentClassName(className = ''){

    if (!_.isString(className) || className === ''){
        throw new Error('className 값이 존재하지 않습니다.');
    }

    return PREFIX + className;
}

/**
 * 문서상에 존재하는 모든 커스텀 엘리먼트를 초기화 시킨다.
 *
 * @param {Object} $rootElem - root 엘리먼트
 * @param {Array} tagNames - 커스텀 엘리먼트의 태그명 집합
 *
 * @alias init
 *
 * @memberOf BXUI.Base
 */
function _init($rootElem = null, tagNames = null){

    if (_.isNull($rootElem) || (!_.isFunction($rootElem.get) || !_.isElement($rootElem.get(0)))){
        $rootElem = $(document.body);
    }

    if (!_.isEmpty(tagNames) && !_.isArray(tagNames)){
        CONSOLE.prop.throw('BXUI.Base.init', 'tagNames', 'type');
    }

    let filterJQConstructor = null;
    let jqConstructors = _getJQConstructorAll();

    if (tagNames){

        _.forEach(tagNames, (v) => {

            let constructor = jqConstructors[v];

            if (constructor){
                filterJQConstructor[v] = constructor;
            }
        });
    }
    else{
        filterJQConstructor = jqConstructors;
    }

    _.map(filterJQConstructor, (constructor, k) => {

        let $elem = $rootElem.find(k);

        if ($elem.length){

            // `커스텀 태그`의 생성자 함수를 반환한다.
            let customElementConstructor = constructor.call($rootElem);

            // `커스텀 태그`를 초기화 한다.
            customElementConstructor.initialize();
        }
    });

    return this;
}

/**
 * 등록된 모든 커스텀 엘리먼트 집합을 반환한다.
 *
 *
 * @private
 *
 * @memberOf BXUI.Base
 */
function _getJQConstructorAll(){

    let ptn = /^(bx|Bx)(.+)/;

    let _proto = Object.getPrototypeOf($());
    let ret = {};

    for (let k in _proto){

        let jqConstructor = _proto[k];

        if (ptn.test(k) && jqConstructor.isCustomElement){

            let customTagName = k.replace(ptn, function(){

                let args = arguments;

                let prefixName = args[1].toLowerCase();
                let tagName = args[2];

                tagName = tagName[0].toLowerCase() + tagName.substr(1);

                return `${prefixName}-${tagName}`;
            });

            ret[customTagName] = jqConstructor;
        }
    }

    return ret;
}

//function _mixin(superClass){
//
//    class Mix {
//
//        constructor(superClass){
//            this.superclass = superClass;
//        }
//
//        with(...mixins){
//
//            console.log(mixins);
//
//            return mixins.reduce(function(c, mixin){
//
//                console.log(c);
//                console.log(mixin);
//
//                return mixin(c);
//
//            }, this.superclass);
//        }
//    }
//
//    return new Mix(superClass);
//}


module.exports = BASE;
