// bxuip-angular.js-ui/src/lib/bx-ui-core/app/file.js
/**
 * Created by sgjeon on 17. 1. 4..
 */

const $ = require('jquery');
const _ = require('lodash');

const CONSOLE = require('./console');
const UTIL = require('./util');

const MESSAGES = {
    "noSupportIE9": "IE9 에서는 지원하지 않는 API 입니다",
    "noUseInnerForm": "bx-fileUpload 컴포넌트는 form 내부에서 사용할 수 없습니다.",
    "notAllowExtensions": "허용하지 않는 확장자 입니다.",
    "overTotalSize": "최대 허용 용량을 초과했습니다.",
    "overFileSize": "개 파일이 개별 허용 용량을 초과했습니다.",
    "overFileCount": "파일 최대 개수를 초과했습니다."
};

const FILE = _.assign({}, {
    /**
     * @memberof BXUI.File
     */
    MESSAGES: MESSAGES,
    /**
     * @memberof BXUI.File
     */
    "readAsDataURL": (...args) => {
        return _readFileData.apply(FILE, args.concat(['dataUrl']));
    },
    /**
     * @memberof BXUI.File
     */
    "readAsArrayBuffer": (...args) => {
        return _readFileData.apply(FILE, args.concat(['arrayBuffer']));
    },
    /**
     * @memberof BXUI.File
     */
    "readAsText": (...args) => {
        return _readFileData.apply(FILE, args.concat(['text']));
    },
    /**
     * @memberof BXUI.File
     */
    "getFilesInfo": (...args) => {
        return _getFilesInfo.apply(FILE, args);
    },
    /**
     * @memberof BXUI.File
     */
    "getFiles": (...args) => {
        return _getFiles.apply(FILE, args);
    },
    /**
     * @memberof BXUI.File
     */
    "getImageFileInfo": (...args) => {
        return _getImageFileInfo.apply(FILE, args);
    },
    /**
     * @memberof BXUI.File
     */
    "getFileNameInfo": (...args) => {
        return _getFileNameInfo.apply(FILE, args);
    }
});

/**
 *
 * 여러가지 형태로 파일 정보를 반환한다.
 *
 * @param {Element} $elem - 파일 엘리먼트 객체
 * @param {Function|Object} arg - callback 함수 또는 Optional 객체
 *
 * @return {object} BXUI.File 객체
 *
 */

function _readFileData(...args){

    let deferred = $.Deferred();
    let promise = deferred.promise();

    if (!window.File || !window.FileList || !window.FileReader) return promise;

    let firstArg = args[0] || null;
    let secondArg = args[1];
    let readType = args[2];

    if (_.isNull(firstArg)) return promise;

    let files = _getFileObjects.call(this, firstArg);

    let opts = {
        "onloadstart": function (){},
        "onprogress": function (){},
        "onloadend": function (){},
        "onabort": function (){},
        "oncompleted": deferred.resolve,
        "onerror": function (){}
    };

    if (_.isString(secondArg)){
        readType = secondArg;
    }
    else if (_.isPlainObject(secondArg)){

        opts.onloadstart = _.isFunction(secondArg.onloadstart) ? secondArg.onloadstart : opts.onloadstart;
        opts.onprogress = _.isFunction(secondArg.onprogress) ? secondArg.onprogress : opts.onprogress;
        opts.onloadend = _.isFunction(secondArg.onloadend) ? secondArg.onloadend : opts.onloadend;
        opts.onabort = _.isFunction(secondArg.onabort) ? secondArg.onabort : opts.onabort;
        opts.onerror = _.isFunction(secondArg.onerror) ? secondArg.onerror : opts.onerror;
    }

    let length = files.length;

    let loadedIndex = 0;
    let ret = [];

    _eachFiles(files, (file, info) => {

        let reader = new FileReader();

        _addReaderEvent.call(reader, readType, info, opts, length, ++loadedIndex, ret);

        switch (readType){

            case 'dataUrl':{

                let ptn = /image.*/g;

                if (info.type.match(ptn)){
                    reader.readAsDataURL(file);
                }
                else{
                    CONSOLE.prop.warn(readType, info.name + ' 파일 타입이 이미지가 아닙니다.');
                    return promise;
                }

                break;
            }
            case 'arrayBuffer':{
                reader.readAsArrayBuffer(file);
                break;
            }
            case 'text':{
                reader.readAsText(file);
                break;
            }
        }
    });

    return promise;
}

/**
 *
 * 파일 정보를 반환한다.
 *
 * @param {Element|Blob} arg - 파일 엘리먼트 객체
 *
 * @return {Array} 파일 정보 객체
 *
 */
function _getFilesInfo(arg = null){

    if (_.isNull(arg)) return this;

    let ret = [];

    let files = _getFileObjects.call(this, arg);
    let length = files.length;

    for (let i = 0; i < length; i++){
        ret.push(_convertFileInfo(files[i]));
    }

    return ret;
}

/**
 *
 * 파일 객체를 반환한다.
 *
 * @param {Element|Blob} arg - 파일 엘리먼트 객체
 *
 * @return {Array} 파일 객체
 *
 */
function _getFiles(arg = null){

    if (_.isNull(arg)) return this;

    let ret = [];

    let files = _getFileObjects.call(this, arg);

    let length = files.length;

    for (let i = 0; i < length; i++){
        ret.push(files[i]);
    }

    return ret;
}

/**
 *
 * 파일 객체를 반환한다.
 *
 * @param {Element|Blob} arg - 파일 엘리먼트 객체
 *
 * @return {Array} 파일 객체
 *
 */
function _getFileObjects(arg = null){

    if (_.isNull(arg)) return this;

    // 인자값이 `파일 엘리먼트` 또는 `Blob 데이터`가 아닐경우
    if (
        !(!arg.get || arg.prop('type') !== 'file') &&
        !(arg.constructor !== Array || arg.constructor !== Blob)){

        return this;
    }

    let ret = [];

    // 파일 엘리먼트인 경우
    if (arg.get && arg.prop('type') === 'file'){

        // `ie9` 브라우저일 경우에는, 해당 `파일 엘리먼트`를 반환한다.
        ret = UTIL.isIE9() ? [arg.get(0)] : arg.prop('files');
    }
    else{

        // blob 데이터인 경우
        let blobs = arg.constructor === Array ? arg : [arg];

        _.map(blobs, (v, k) => {
            ret.push(new File([v], v.name, {"type": v.type, "lastModified": v.lastModified}));
        });
    }

    return ret;
}

/**
 *
 * fileReader 이벤트를 등록한다.
 *
 * @param {String} readType - reader 타입
 * @param {Object} info - 파일 정보
 * @param {Object} opts - optional 객체
 * @param {Int} length - 전체 파일 길이
 * @param {Int} loadedIndex - 로드가 완료된 파일 수
 * @param {Array} ret - `completed callback` 함수에 전달될 반환 데이터
 *
 *
 */
function _addReaderEvent(readType, info, opts, length, loadedIndex, ret){

    let readyStateText = {
        "0": 'EMPTY',
        "1": 'LOADING',
        "2": 'DONE'
    };

    let progressStateText = {
        "0": 'LOADSTART',
        "1": 'PROGRESS',
        "2": 'LOADEND',
        "3": 'COMPLETED',
        "4": 'ABORT',
        "5": 'ERROR'
    };

    let target = null;

    this.onloadstart = (e) => {

        target = e.target;

        info[readType] = target.result;

        opts.onloadstart.apply(this, [info, readyStateText[target.readyState], progressStateText[0]]);
    };

    this.onprogress = (e) => {

        target = e.target;
        let loaded = e.loaded;
        let total = e.total;

        info[readType] = target.result;

        opts.onprogress.apply(this, [info, loaded, total, readyStateText[target.readyState], progressStateText[1]]);
    };

    this.onloadend = (e) => {

        target = e.target;

        info[readType] = target.result;

        ret.push(info);

        opts.onloadend.apply(this, [info, readyStateText[target.readyState], progressStateText[2]]);

        if (loadedIndex === length){
            opts.oncompleted.apply(this, [ret, readyStateText[target.readyState], progressStateText[3]]);
        }
    };

    this.onabort = (e) => {

        target = e.target;

        info[readType] = target.result;

        opts.onabort.apply(this, [info, readyStateText[target.readyState], progressStateText[4]]);
    };


    this.onerror = (e) => {

        target = e.target;

        info[readType] = target.result;

        opts.onerror.apply(this, [info, readyStateText[target.readyState], progressStateText[5]]);
    };
}

/**
 *
 * 파일 정보를 반환한다.
 *
 * @param {Object} file - 파일 객체
 *
 * @return {object} 파일 정보
 *
 */
function _convertFileInfo(file = {}){

    if (!_.isObject(file)) return this;

    let ret = {
        "originFile": file,
        "type": '',
        "name": '',
        "size": '',
        "lastModified": '',
        "path": ''
    };

    let name = '';

    if (UTIL.isIE9()){

        let path = '';

        file.select();

        try {
            path = document.selection.createRangeCollection()[0].text.toString();
        }
        catch (e){
            path = file.value;
        }

        name = path.substr(path.lastIndexOf('\\') + 1, path.length);

        ret.name = name;
        ret.path = path;

    }
    else{

        ret.type = file.type;

        name = file.name;
        ret.name = name;

        ret.size = file.size;
        ret.lastModified = file.lastModified;
    }

    ret.ext = _getFileNameInfo(name).ext;

    return ret;
}

/**
 *
 * 파일 객체를 전달된 callback 함수에 전달한다.
 *
 * @param {Object} files - 파일 객체 집합
 *
 *
 */
function _eachFiles(files = {}, callback = function (){}){

    if (!_.isObject(files) || !_.isFunction(callback)) return this;

    let length = files.length;

    for (let i = 0; i < length; i++){

        let file = files[i];

        callback(file, _convertFileInfo(file));
    }
}

/**
 *
 * 이미지 파일 객체 정보를 반환한다.
 *
 * @param {String} url - 로컬 이미지 파일 URL
 *
 * @return {object} 이미지 파일 정보
 *
 */
function _getImageFileInfo(url = ''){

    let deferred = $.Deferred();
    let promise = deferred.promise();


    if (!_.isString(url) || !_.trim(url).length) return promise;

    // 반환값
    let ret = {};
    let self = this;

    let $img = $(new Image());

    // 전달받은 URL 값의 이미지를 로드 시킨다.
    $img.prop('src', url);

    $img.on('load', function(){

        let elem = this;

        // 파일명
        let fileInfo = _getFileNameInfo(url);

        let name = fileInfo.name;
        let ext = fileInfo.ext;

        // 화면상에 보이는 이미지 사이즈
        let width = elem.width;
        let height = elem.height;

        // 원본 이미지 사이즈
        let naturalWidth = elem.naturalWidth;
        let naturalHeight = elem.naturalHeight;

        ret.name = name;
        ret.ext = ext;

        // 화면상에 보이는 이미지 사이즈
        ret.width = width;
        ret.height = height;

        // 원본 이미지 사이즈
        ret.naturalWidth = naturalWidth;
        ret.naturalHeight = naturalHeight;

        // origin 이 포함된 이미지 주소
        let src = $img.prop('src');
        let origin = UTIL.getLocationOrigin();
        let imgOrigin = UTIL.getLocationOriginWithUrl(src);

        // 해당 이미지가 로컬 이미지일 경우에만 ajax 를 통해, 더 많은 이미지 정보를 알아낸다.
        if (origin === imgOrigin){

            // detail xhr 설정 객체
            let xhrFields = {};

            // 기본 dataType
            let dataType = 'text';

            if (!UTIL.isIE9()){
                xhrFields = {"responseType": 'blob'};
                dataType = 'binary';
            }

            $.ajax({
                url: url,
                type: 'get',
                xhrFields: xhrFields,
                dataType: dataType,
                beforeSend: xhr => {
                },
                error: xhr => {
                },
                success: (blobData, status, xhr) => {

                    let size = 0;
                    let type = '';
                    let lastModified = new Date(xhr.getResponseHeader('Last-Modified'));

                    if (UTIL.isIE9()){

                        size = xhr.getResponseHeader('Content-Length');
                        type = xhr.getResponseHeader('Content-Type');

                        ret.type = type;
                        ret.size = size;
                        ret.lastModified = lastModified.getTime();

                        deferred.resolve(ret);
                    }
                    else{

                        blobData.name = name;
                        blobData.lastModified = lastModified;

                        let files = self.getFilesInfo(blobData);

                        let ret = files[0];

                        // 화면상에 보이는 이미지 사이즈
                        ret.width = width;
                        ret.height = height;

                        // 원본 이미지 사이즈
                        ret.naturalWidth = naturalWidth;
                        ret.naturalHeight = naturalHeight;

                        deferred.resolve(ret);
                    }
                }
            });
        }
        else{
            deferred.resolve(ret);
        }
    });

    return promise;
}

/**
 *
 * 이미지 파일 객체 정보를 반환한다.
 *
 * @param {String} 파일 경로 또는 파일명
 *
 * @return {object} 파일 정보 객체
 *
 */
function _getFileNameInfo(v = ''){

    v = v.replace('\\', '/');

    let name = v.substr(v.lastIndexOf('/') + 1).split('?')[0];
    let ext = name.substr(name.lastIndexOf('.') + 1).split('?')[0].toLowerCase();

    return {
        name,
        ext
    };
}

module.exports = FILE;
