// bxuip-angular.js-ui/src/lib/bx-ui-core/app/form.js
/**
 * Created by sgjeon on 17. 1. 4..
 */

const $ = require('jquery');
const _ = require('lodash');

const CONSOLE = require('./console');
const UTIL = require('./util');
const XHR = require('./xhr');
const FILE = require('./file');

const ENCTYPES = {
    "default": 'application/x-www-form-urlencoded',
    "multipart": 'multipart/form-data',
    "text": 'text/plain'
};

const TITLES = {
    "selectFile": '파일 선택하기',
    "addFile": '파일 추가하기'
};

const FORM = _.assign({}, {
    /**
     * @memberof BXUI.FORM
     */
    "createFileUploadElements": (...args) => {
        return _createFileUploadElements.apply(FORM, args);
    },
    /**
     * @memberof BXUI.FORM
     */
    "submit": (...args) => {
        return _submit.apply(FORM, args);
    },
    /**
     * @memberof BXUI.FORM
     */
    "clear": (...args) => {
        return _clear.apply(FORM, args);
    }
});

/**
 *
 * 파일 업로드 관련 엘리먼트들을 반환한다.
 *
 * @param {object} [opts]
 * @param {string} [opts.type] - 엘리먼트 타입(input, button)
 * @param {string} [opts.prefixId] - prefixId
 * @param {string} [opts.action] - form action
 * @param {function} [opts.onFileChange] - file change event handler
 *
 * @return {object} fileUpload 엘리먼트 집합
 *
 */
function _createFileUploadElements({
    type = 'button',
    prefixId = '',
    action = '',
    onFileChange = function(){}
} = {}){

    if (!_.isString(type)){
        CONSOLE.prop.throw('BXUI.Form.createFileUploadFormElement', 'type', 'type');
    }

    if (!_.isString(prefixId)){
        CONSOLE.prop.throw('BXUI.Form.createFileUploadFormElement', 'prefixId', 'type');
    }

    if (!_.isString(action) || !_.trim(action).length){
        CONSOLE.prop.throw('BXUI.Form.createFileUploadFormElement', 'action', 'all');
    }

    if (!_.isFunction(onFileChange)){
        CONSOLE.prop.throw('BXUI.Form.createFileUploadFormElement', 'onFileChange', 'type');
    }

    // iframe 아이디
    let iframeId = `bxIframe_${_.uniqueId()}`;
    // input file 타입 아이디
    let inputFileId = `bxInputFile_${_.uniqueId()}`;

    if (prefixId){
        iframeId = `${prefixId}_${iframeId}`;
        inputFileId = `${prefixId}_${inputFileId}`;
    }

    // form 엘리먼트
    let $form = $(`<form method="post" action="${action}" enctype="${ENCTYPES.multipart}" target="${iframeId}" class="blind">`);

    if (type === 'input'){
        $form.append(`<input type="text" title="${TITLES.selectFile}" disabled />`);
    }

    $form.append(`<input type="file" id="${inputFileId}" name="${inputFileId}" aria-hidden="true" class="blind" />`);

    // label 엘리먼트
    let $label = $(`<label for="${inputFileId}" title="${TITLES.addFile}"></label>`);

    // iframe 엘리먼트
    let $iframe = $(`<iframe id="${iframeId}" name="${iframeId}" style="display:none;visibility:hidden" class="blind"></iframe>`);

    // 이벤트 바인딩
    let $inputFile = $form.find('input[type="file"]');
    $inputFile.on('change', onFileChange);

    $inputFile.on('focusin focusout', e => {

        if (e.type === 'focusin') $label.addClass('focus');
        else $label.removeClass('focus');
    });

    return {
        $form: $form,
        $label: $label,
        $iframe: $iframe
    };
}

/**
 *
 * form submit 함수
 *
 * @param {object} [opts]
 * @param {object} [opts.$formElem=null] - 폼 엘리먼트
 * @param {object} [opts.$iframeElem=null] - 아이프레임 엘리먼트
 * @param {string} [opts.url=''] - url
 * @param {string} [opts.method=''] - method
 * @param {object} [opts.headers={}] - headers
 * @param {object} [opts.params={}] - params
 * @param {string} [opts.dataType=''] - dataType
 * @param {object} [opts.file=null] - file 객체
 * @param {function} [opts.before=function(){}] - before event handler
 * @param {function} [opts.loadStart=function(){}] - loadStart event handler
 * @param {function} [opts.progress=function(){}] - progress event handler
 * @param {function} [opts.success=function(){}] - success event handler
 * @param {function} [opts.error=function(){}] - error event handler
 *
 * @return {object} BXUI.Form 객체
 *
 */
function _submit({
    $formElem = null,
    $iframeElem = null,
    url = '',
    method = 'POST',
    headers = {},
    params = {},
    dataType = 'json',
    file = {},
    before = function(){},
    loadStart = function(){},
    progress = function(){},
    success = function(){},
    error = function(){}
} = {}){

    if (window.FormData){
        _xhrSubmit.call(this, {url, method, headers, params, dataType, file, before, loadStart, progress, success, error});
    }
    else{
        _formSubmit.call(this, {$formElem, $iframeElem, params, before, success});
    }

    return this;
}

/**
 *
 * xhr 파일 업로드 함수
 *
 * @param {object} [opts]
 * @param {string} [opts.url=''] - url
 * @param {string} [opts.method=''] - method
 * @param {object} [opts.headers={}] - headers
 * @param {object} [opts.params={}] - params
 * @param {string} [opts.dataType=''] - dataType
 * @param {object} [opts.file=null] - file 객체
 * @param {function} [opts.before=function(){}] - before event handler
 * @param {function} [opts.loadStart=function(){}] - loadStart event handler
 * @param {function} [opts.progress=function(){}] - progress event handler
 * @param {function} [opts.success=function(){}] - success event handler
 * @param {function} [opts.error=function(){}] - error event handler
 *
 * @return {object} BXUI.Form 객체
 *
 */
function _xhrSubmit({
    url = '',
    method = '',
    headers = {},
    params = {},
    dataType = '',
    file = null,
    before = function(){},
    loadStart = function(){},
    progress = function(){},
    success = function(){},
    error = function(){}
} = {}){

    if (!_.isString(url) || !_.trim(url).length){
        CONSOLE.prop.throw('BXUI.Form.submit', 'url', 'all');
    }

    if (!_.isString(method)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'method', 'type');
    }

    if (!_.isPlainObject(headers)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'headers', 'type');
    }

    if (!_.isPlainObject(params)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'params', 'type');
    }

    if (!_.isString(dataType)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'dataType', 'type');
    }

    if (!_.isObject(file)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'file', 'type');
    }

    if (!_.isFunction(before)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'before', 'type');
    }

    if (!_.isFunction(loadStart)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'loadStart', 'type');
    }

    if (!_.isFunction(progress)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'progress', 'type');
    }

    if (!_.isFunction(success)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'success', 'type');
    }

    if (!_.isFunction(error)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'error', 'type');
    }

    let formData = new FormData();

    let filesInfo = FILE.getFilesInfo(file);

    _.forEach(filesInfo, v => {
        formData.append(v.name, v.originFile);
    });

    let ajaxOption = {
        url: url,
        transferType: 'upload',
        processData: false,
        contentType: false,
        dataType: dataType,
        type: method,
        data: formData,
        beforeSend: (xhr, settings) =>{

            // `IE9 브라우저`에서, 캐시 사용(json 등)을 무력화 시키기위한 header 설정
            xhr.setRequestHeader('If-Modified-Since', 'Mon, 26 Jul 1997 05:00:00 GMT');
            xhr.setRequestHeader('Cache-Control', 'no-cache');
            xhr.setRequestHeader('Pragma', 'no-cache');

            // 사용자가 입력한 Header Setting
            _.forEach(headers, (v, k) => {
                xhr.setRequestHeader(k, v);
            });

            before(xhr);

            _.forEach(params, (v, k) => {
                settings.data.append(k,v);
            });
        },
        loadStart: (xhr) => {
            loadStart(xhr);
        },
        progress: (xhr, loaded, total, percent) => {
            progress(xhr, loaded, total, percent);
        },
        success: data => {
            success(data);
        },
        error: xhr => {
            error(xhr);
        }
    };

    XHR.ajax(ajaxOption);

    return this;
}

/**
 *
 * 폼 파일 업로드 함수
 *
 * @param {object} [opts]
 * @param {string} [opts.$formElem=null] - 폼 엘리먼트
 * @param {string} [opts.$iframeElem=null] - 아이프레임 엘리먼트
 * @param {object} [opts.params={}] - params
 * @param {function} [opts.before=function(){}] - before event handler
 * @param {function} [opts.success=function(){}] - success event handler
 *
 * @return {object} BXUI.Form 객체
 *
 */
function _formSubmit({
    $formElem = null,
    $iframeElem = null,
    params = {},
    before = function(){},
    success = function(){}
    } = {}){

    if (_.isNull($formElem)){
        CONSOLE.prop.throw('BXUI.Form.submit', '$formElem', 'type');
    }

    if (!_.isFunction($formElem.get) || !_.isElement($formElem.get(0))){
        CONSOLE.prop.throw('BXUI.Form.submit', '$formElem', 'all');
    }

    if (_.isNull($iframeElem)){
        CONSOLE.prop.throw('BXUI.Form.submit', '$iframeElem', 'type');
    }

    if (!_.isFunction($iframeElem.get) || !_.isElement($iframeElem.get(0))){
        CONSOLE.prop.throw('BXUI.Form.submit', '$iframeElem', 'all');
    }

    if (!_.isPlainObject(params)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'params', 'type');
    }

    if (!_.isFunction(before)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'before', 'type');
    }

    if (!_.isFunction(success)){
        CONSOLE.prop.throw('BXUI.Form.submit', 'success', 'type');
    }

    _.map(params, (v, k) => {

        let $input = $formElem.find(`input[id="${k}"]`);

        $input.remove();
    });


    _.map(params, (v, k) => {

        let $input = $('<input type="hidden" />');

        $input.attr('id', k);
        $input.attr('name', k);
        $input.attr('value', v);

        $formElem.append($input);
    });

    before();

    $iframeElem.get(0).onload = () => {

        let data = $($iframeElem.get(0).contentDocument.body).text();

        try{
            success(JSON.parse(data));
        }
        catch (e){
            success(_.toString(data));
        }
    };

    $formElem.submit();

    return this;
}

/**
 *
 * 폼 데이터 삭제 함수
 *
 * @param {object=null} $elem
 *
 * @return {object} BXUI.Form 객체
 *
 */
function _clear($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Form.clear', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Form.clear', '$elem', 'all');
    }

    let type = $elem.prop('type');
    let tagName = $elem.prop('tagName').toLowerCase();

    if (tagName === 'form'){
        $elem.find('input, textarea, select').each((k, v) => {
            _clear($(v));
        });
    }

    if (tagName === 'input'){

        if (type === 'text' || type === 'password' || type === 'hidden'){
            $elem.val('');
        }
        else if (type === 'file'){
            $elem.prop('type', '');
            $elem.prop('type', 'file');
        }
        else if (type === 'checkbox' || type === 'radio'){
            $elem.prop('checked', false);
        }
    }

    if (tagName === 'textarea'){
        $elem.val('');
    }

    if (tagName === 'select'){
        $elem.prop('selectedIndex', -1);
    }

    return this;
}



module.exports = FORM;
