// bxuip-angular.js-ui/src/lib/bx-ui-core/app/util.js
/**
 * Created by sgjeon on 16. 8. 19..
 */

const $ = require('jquery');
const _ = require('lodash');

const CONSOLE = require('./console');
const TYPE = require('./type');

const JQ_BROWSER = require('jquery.browser');


const UTIL = _.assign({}, {
    /**
     * @memberOf BXUI.Util
     */
    "getScrollSize": _getScrollSize,
    /**
     * @memberOf BXUI.Util
     */
    "attr": _attr,
    /**
     * @memberOf BXUI.Util
     */
    "addStyle": _addStyle,
    /**
     * @memberOf BXUI.Util
     */
    "removeWhiteSpace": _removeWhiteSpace,
    /**
     * @memberOf BXUI.Util
     */
    "addAllEventListenerWithProperty": _addAllEventListenerWithProperty,
    //"removeAllEventListenerWithProperty": _removeAllEventListenerWithProperty,
    /**
     * @memberOf BXUI.Util
     */
    "setHSL": _setHSL,
    /**
     * @memberOf BXUI.Util
     */
    "colorToHSL": _colorToHSL,
    /**
     * @memberOf BXUI.Util
     */
    "RGBAToHSLA": _RGBAToHSLA,
    /**
     * @memberOf BXUI.Util
     */
    "toJSON": _toJSON,
    /**
     * @memberOf BXUI.Util
     */
    "JSParser": _JSParser,
    /**
     * @memberOf BXUI.Util
     */
    "hexToRGB": _hexToRGB,
    /**
     * @memberOf BXUI.Util
     */
    "colorLevelToHSL": _colorLevelToHSL,
    /**
     * @memberOf BXUI.Util
     */
    "objectToQueryString": _objectToQueryString,
    /**
     * @memberOf BXUI.Util
     */
    "queryStringToObject": _queryStringToObject,
    /**
     * @memberOf BXUI.Util
     */
    "getAttrValueByCustomElement": _getAttrValueByCustomElement,
    /**
     * @memberOf BXUI.Util
     */
    "isAttrNameByCustomElement": _isAttrNameByCustomElement,
    /**
     * @memberOf BXUI.Util
     */
    "objectToStringAttr": _objectToStringAttr,
    /**
     * @memberOf BXUI.Util
     */
    "stringAttrToObject": _stringAttrToObject,
    /**
     * @memberOf BXUI.Util
     */
    "attrsToGenerator": _attrsToGenerator,
    /**
     * @memberOf BXUI.Util
     */
    "attrsToObject": _attrsToObject,
    /**
     * @memberOf BXUI.Util
     */
    "setProperty": _setProperty,
    /**
     * @memberOf BXUI.Util
     */
    "setReadOnlyProperty": _setReadOnlyProperty,
    /**
     * @memberOf BXUI.Util
     */
    "setNonEnumerableProperty": _setNonEnumerableProperty,
    /**
     * @memberOf BXUI.Util
     */
    "setPropertyWithDescriptor": _setPropertyWithDescriptor,
    /**
     * @memberOf BXUI.Util
     */
    "convertInternalPropertyName": _convertInternalPropertyName,
    /**
     * @memberOf BXUI.Util
     */
    "styleToObject": _styleToObject,
    /**
     * @memberOf BXUI.Util
     */
    "objectToStyle": _objectToStyle,
    /**
     * @memberOf BXUI.Util
     */
    "isIE": _isIE,
    /**
     * @memberOf BXUI.Util
     */
    "isIE9": _isIE9,
    /**
     * @memberOf BXUI.Util
     */
    "getFormatString": _getFormatString,
    /**
     * @memberOf BXUI.Util
     */
    "getUnFormatString": _getUnFormatString,
    /**
     * @memberOf BXUI.Util
     */
    "formatLength": _formatLength,
    /**
     * @memberOf BXUI.Util
     */
     "checkFormatString": _checkFormatString,
    /**
     * @memberOf BXUI.Util
     */
    "extractDateFormat": _extractDateFormat,
    /**
     * @memberOf BXUI.Util
     */
    "convertDateFormat": _convertDateFormat,
    /**
     * @memberOf BXUI.Util
     */
    "isDate": _isDate,
    /**
     * @memberOf BXUI.Util
     */
    "getElementScrollBarSize": _getElementScrollBarSize,
    /**
     * @memberOf BXUI.Util
     */
    "checkVisibleArea": _checkVisibleArea,
    /**
     * @memberOf BXUI.Util
     */
    "createComponentOptions": _createComponentOptions,
    /**
     * @memberOf BXUI.Util
     */
    "parseObject": _parseObject,
    /**
     * @memberOf BXUI.Util
     */
    "isAppliedCSSCalc": _isAppliedCSSCalc,
    /**
     * @memberOf BXUI.Util
     */
    "setPosition": _setPosition,
    /**
     * @memberOf BXUI.Util
     */
    "getSelectorInstance": _getSelectorInstance,
    /**
     * @memberOf BXUI.Util
     */
    "getLocationOrigin": _getLocationOrigin,
    /**
     * @memberOf BXUI.Util
     */
    "getLocationOriginWithUrl": _getLocationOriginWithUrl,
    /**
     * @memberOf BXUI.Util
     */
    "getLocationWithUrl": _getLocationWithUrl,
    /**
     * @memberOf BXUI.Util
     */
    "convertStringToByte": _convertStringToByte,
    /**
     * @memberOf BXUI.Util
     */
    "isInViewport": _isInViewport,
    /**
     * @memberOf BXUI.Util
     */
    "isBelowSafariVersion": _isBelowSafariVersion,
    /**
     * @memberOf BXUI.Util
     */
    "JSONToCSV": _JSONToCSV,
    /**
     * @memberOf BXUI.Util
     */
    "CSVToJSON": _CSVToJSON,
    /**
     * @memberOf BXUI.Util
     */
    "encodeBase64": _encodeBase64,
    /**
     * @memberOf BXUI.Util
     */
    "decodeBase64": _decodeBase64,
    /**
     * @memberOf BXUI.Util
     */
    "encodeUTF8": _encodeUTF8,
    /**
     * @memberOf BXUI.Util
     */
    "decodeUTF8": _decodeUTF8,
    /**
     * @memberOf BXUI.Util
     */
    "isSupportedDataURI": _isSupportedDataURI,
    /**
     * @memberOf BXUI.Util
     */
    "getBrowserVersion": _getBrowserVersion,
    /**
     * @memberOf BXUI.Util
     */
    "isAboveSafariVersion": _isAboveSafariVersion,
    /**
     * @memberOf BXUI.Util
     */
    "getPreFocusInstance": _getPreFocusInstance,
    /**
     * @memberOf BXUI.Util
     */
    "setPreFocusInstanceKey": _setPreFocusInstanceKey,
    /**
     * @memberOf BXUI.Util
     */
    "isAngular": _isAngular,
    /**
     * @memberOf BXUI.Util
     */
    "isReact": _isReact,
    /**
     * @memberOf BXUI.Util
     */
    "setIE9PlaceHolder": _setIE9PlaceHolder,
    /**
     * @memberOf BXUI.Util
     */
    "removePropertys": _removePropertys,
    /**
     * @memberOf BXUI.Util
     */
    "removeItems": _removeItems,
    /**
     * @memberOf BXUI.Util
     */
    "uniqueId": _uniqueId
});

/**
 *
 * @param {Element} $elem - 엘리먼트 객체
 *
 * @return {object} 엘리먼트 Scroll 사이즈
 *
 */
function _getScrollSize($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.getScrollSize', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.getScrollSize', '$elem', 'type');
    }

    let ret = 0;

    if ($elem.prop('nodeName') === 'BODY'){

        let $html = $(document.documentElement);

        ret = {
            width: Math.max(
                $elem.prop('scrollWidth'), $html.prop('scrollWidth'),
                $elem.prop('offsetWidth'), $html.prop('offsetWidth'),
                $elem.prop('clientWidth'), $html.prop('clientWidth')
            ),
            height: Math.max(
                $elem.prop('scrollHeight'), $html.prop('scrollHeight'),
                $elem.prop('offsetHeight'), $html.prop('offsetHeight'),
                $elem.prop('clientHeight'), $html.prop('clientHeight')
            )
        };
    }
    else{
        ret = {
            width: $elem.prop('scrollWidth'),
            height: $elem.prop('scrollHeight')
        };
    }

    return ret;
}

/**
 *
 * 전달된 $elem 에 attr 속성을 할당하는 함수
 *
 * @param {Element} $elem - 엘리먼트 객체
 * @param {object} attr - attr 속성 객체
 *
 * @return {object} BXUI.Util 객체
 *
 */
function _attr($elem = null, attr = {}){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.attr', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.attr', '$elem', 'type');
    }

    if (!_.isPlainObject(attr)){
        CONSOLE.prop.throw('BXUI.Util.attr', 'attr', 'type');
    }

    _.forEach(attr, function (v, k){

        if (k === 'class'){
            $elem.addClass(v);
        }
        else if (k === 'style'){
            _addStyle($elem, v);
        }
        else{
            $elem.attr(k, v);
        }
    });

    return this;
}

/**
 *
 * 전달된 $elem 에 style 을 할당하는 함수
 *
 * @param {Element} $elem - 엘리먼트 객체
 * @param {string} style - style
 *
 * @return {object} BXUI.Util 객체
 *
 */
function _addStyle($elem = null, style = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.addStyle', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.addStyle', '$elem', 'type');
    }

    if (!_.isString(style)){
        CONSOLE.prop.throw('BXUI.Util.addStyle', 'style', 'type');
    }

    if (style){

        let arr = [];
        let _style = $elem.attr('style');

        if (_style) arr.push(_style);

        arr.push(style);

        $elem.attr('style', arr.join(';'));
    }

    return this;
}

/**
 *
 * 전달된 문자열의 공백을 제거하는 함수
 *
 *
 * @param {string} str - 공백을 제거할려는 문자열
 *
 * @return {string} 공백이 제거된 문자열
 *
 */

function _removeWhiteSpace(str = ''){

    if (!_.isString(str)){
        CONSOLE.prop.throw('BXUI.Util.removeWhiteSpace', 'str', 'all');
    }

    return str.replace(/\s+/g, '');
}

/**
 *
 * 이벤트 추가 함수
 *
 *
 * @param {object} $elem - 엘리먼트 객체
 * @param {object} property - 바인딩시킬 프로퍼티 객체
 * @param {object} datas - 바인딩된 이벤트로 전달될 데이터 객체
 *
 * @return {object} BXUI.Util 객체
 *
 */
function _addAllEventListenerWithProperty($elem = null, property = {}, ...datas){ //jshint ignore:line

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.addAllEventListenerWithProperty', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.addAllEventListenerWithProperty', '$elem', 'type');
    }

    if (!_.isPlainObject(property)){
        CONSOLE.prop.throw('BXUI.Util.addAllEventListenerWithProperty', 'property', 'type');
    }

    _.map(property, (v, k) =>{

        let mt = k.match(/^bx-[a-z]+/);

        if (mt && mt[0]){

            let evtName = k.replace('bx-', '').toLowerCase();

            _.isFunction(v) && $elem.on(evtName, function(e){
                v.bind(this, ...[e].concat(datas))();
            });
        }
    });

    return this;
}

/**
 *
 * 전달된 $elem 에 HSL color를 할당하는 함수
 *
 * @param {Element} $elem - 엘리먼트 객체
 * @param {object} attr - attr 속성 객체
 *
 * @return {object} BXUI.Util 객체
 *
 */
function _setHSL($elem = null, attributes = {}){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.setHSL', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.setHSL', '$elem', 'type');
    }

    if (!_.isPlainObject(attributes)){
        CONSOLE.prop.throw('BXUI.Util.setHSL', 'attributes', 'type');
    }

    _.map(attributes, function (v, k){

        let h = v.h ? v.h : 0;
        let s = v.s ? v.s : 0;
        let l = v.l ? v.l : 0;
        let a = v.a ? v.a : 0;
        let _v = v.v ? v.v : '';

        let hslInfo = [h, s + '%', l + '%', a];

        let hsl = k + ':' + _v + ' hsla(' + hslInfo.join(',') + ')';

        _addStyle($elem, hsl);
    });

    return this;
}

/**
 *
 * RGB color를  HSL color로 변환하는 함수
 *
 * @param {number} r - red value
 * @param {number} g - green value
 * @param {number} b - blur value
 * @param {number | string} a - alpha value
 *
 * @return {object} HSL 객체
 *
 */
function _RGBAToHSLA(r = 0, g = 0, b = 0, a = 0){

    if (!_.isNumber(r)){
        CONSOLE.prop.throw('BXUI.Util.RGBAToHSLA', 'r', 'type');
    }

    if (!_.isNumber(g)){
        CONSOLE.prop.throw('BXUI.Util.RGBAToHSLA', 'g', 'type');
    }

    if (!_.isNumber(b)){
        CONSOLE.prop.throw('BXUI.Util.RGBAToHSLA', 'b', 'type');
    }

    if (!_.isNumber(a) && !_.isString(a)){
        CONSOLE.prop.throw('BXUI.Util.RGBAToHSLA', 'a', 'type');
    }


    r = parseInt(r) / 255;
    g = parseInt(g) / 255;
    b = parseInt(b) / 255;

    let max = Math.max(r, g, b), min = Math.min(r, g, b);
    let h, s, l = (max + min) / 2;

    if (max === min){
        h = 0;
        s = 0;
    }
    else{

        let d = max - min;
        s = l > 0.5 ? d / (2 - max - min) : d / (max + min);

        switch (max){

            case r:
                h = (g - b) / (d + (g < b ? 6 : 0));
                break;
            case g:
                h = (b - r) / d + 2;
                break;
            case b:
                h = (r - g) / d + 4;
                break;
        }

        h /= 6;
    }

    h = h > 0 ? Math.round(h * 360) : 0;
    s = s > 0 ? Math.round(s * 100) : 0;
    l = l > 0 ? Math.round(l * 100) : 0;

    return {
        "h": h,
        "s": s,
        "l": l,
        "a": a
    };
}

/**
 *
 * color 를  HSL color 로 변환하는 함수
 *
 * http://vision0814.tistory.com/86(계산식)
 *
 * @param {string} color
 *
 * @return {object} HSL 객체
 */
function _colorToHSL(color = ''){

    if (!_.isString(color)){
        CONSOLE.prop.throw('BXUI.Util.colorToHSL', 'color', 'type');
    }

    let r = parseInt(color.substr(1, 2), 16) / 255;
    let g = parseInt(color.substr(3, 2), 16) / 255;
    let b = parseInt(color.substr(5, 2), 16) / 255;

    let max = Math.max(r, g, b), min = Math.min(r, g, b);
    let h, s, l = (max + min) / 2;

    if (max === min){
        h = 0;
        s = 0;
    }
    else{

        let d = max - min;
        s = l > 0.5 ? d / (2 - max - min) : d / (max + min);

        switch (max){

            case r:
                h = (g - b) / d + (g < b ? 6 : 0);
                break;
            case g:
                h = (b - r) / d + 2;
                break;
            case b:
                h = (r - g) / d + 4;
                break;
        }

        h /= 6;
    }

    h = h > 0 ? Math.round(h * 360) : 0;
    s = s > 0 ? Math.round(s * 100) : 0;
    l = l > 0 ? Math.round(l * 100) : 0;

    return {
        "h": h,
        "s": s,
        "l": l
    };
}

/**
 *
 * HEX code color를  RGB color로 변환하는 함수
 *
 * @param {string} color - hex code
 *
 * @return {object} RGB 객체
 *
 */
function _hexToRGB(hex = ''){

    if (!_.isString(hex)){
        CONSOLE.prop.throw('BXUI.Util.hexToRGB', 'hex', 'type');
    }

    let ret = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);

    return ret ? {

        "r": parseInt(ret[1], 16),
        "g": parseInt(ret[2], 16),
        "b": parseInt(ret[3], 16)

    } : null;
}

/**
 * colorLevel로 전달된 HSL color를 계산하여 반환하는 함수
 *
 * @param {number} colorLevel - color level (1~5)
 * @param {object} hsl - color level을 적용하고자 하는 hsl 객체
 *
 * @returns {object} HSL 객체
 *
 */
function _colorLevelToHSL(colorLevel = 3, hsl = {}){

    if (!_.isNumber(colorLevel)){
        CONSOLE.prop.throw('BXUI.Util.colorToHSL', 'colorLevel', 'type');
    }

    if (!_.isPlainObject(hsl)){
        CONSOLE.prop.throw('BXUI.Util.colorToHSL', 'hsl', 'type');
    }

    if (colorLevel < 3){

        let level = (3 - colorLevel);
        let h = level * 8;
        let l = level * 5;

        hsl.h -= h;
        hsl.l -= l;

    }else{

        let level = (colorLevel - 3);
        let h = level * 8;
        let l = level * 5;

        hsl.h += h;
        hsl.l += l;

    }

    return hsl;

}

function _toJSON(value = ''){

    if (!_.isString(value)){
        CONSOLE.prop.throw('BXUI.Util.toJSON', 'value', 'type');
    }

    let ret = null;

    try{
        ret = JSON.parse(value);
    }
    catch (e){
        // eslint-disable-next-line no-eval
        ret = eval('(' + value + ')');
    }

    return ret;
}

function _JSParser(value = ''){

    if (!_.isString(value)){
        CONSOLE.prop.throw('BXUI.Util.JSParser', 'value', 'type');
    }

    // eslint-disable-next-line no-eval
    return eval('(' + value + ')');
}

/**
 * Object 데이터를 QueryString 파라미터 형식으로 변환하는 함수
 *
 * @param {object} params - JSON 데이터
 * @param {boolean} isEnc - URI 인코딩 여부
 *
 * @returns {string} 변환된 uri 파라미터
 */
function _objectToQueryString(params = {}, isEnc = false){

    if (!_.isPlainObject(params)){
        CONSOLE.prop.throw('BXUI.Util.objectToQueryString', 'params', 'type');
    }

    if (!_.isBoolean(isEnc)){
        CONSOLE.prop.throw('BXUI.Util.objectToQueryString', 'isEnc', 'type');
    }

    let results = [];

    _.forEach(params, (value, key) =>{

        if (_.isNull(value) || _.isUndefined(value)) return;

        if (_.isArray(value)){
            let tempParams = [];

            _.forEach(value, (item, i)=> tempParams.push(`${isEnc ? encodeURIComponent(item) : item}`));
            results.push(`${key}=` + tempParams.join());
        }else{
            results.push(`${key}=${isEnc ? encodeURIComponent(value) : value}`);
        }
    });

    return results.join('&');
}

/**
 * QueryString 파라미터를 Object 형식으로 변환하는 함수
 *
 * @param {string} url - 변환할 url
 *
 * @returns {object} 변환된 object
 */
function _queryStringToObject(url = ''){

    if (!_.isString(url)){
        CONSOLE.prop.throw('BXUI.Util.queryStringToObject', 'url', 'type');
    }

    let result = {};

    if (url.length === 0) return result;

    if (url.indexOf('?') > -1){

        let sliceData = url.slice(url.indexOf('?') + 1, url.length);
        let params = sliceData.split('&');

        _.forEach(params, value =>{

            if (value.indexOf('=') === -1) return;

            let objKey = decodeURIComponent(value.split('=')[0]);
            let objValue = decodeURIComponent(value.split('=')[1]);

            result[objKey] = objValue;

        });
    }

    return result;
}

/**
 * custom 엘리먼트의 attr 값을 반환한다.(ie(9) 대응)
 *
 * @param {object} target - target object
 * @param {string} attrName - attr 이름
 *
 * @returns {object} attr 값
 */
function _getAttrValueByCustomElement(target = {}, attrName = ''){

    if (!_.isPlainObject(target)){
        CONSOLE.prop.throw('BXUI.Util.getAttrValueByCustomElement', 'target', 'type');
    }

    if (!_.isString(attrName)){
        CONSOLE.prop.throw('BXUI.Util.getAttrValueByCustomElement', 'attrName', 'type');
    }

    // 보통 브라우저에서는 attr 이름은 전부 소문자로 처리한다.
    let lowerAttrName = _.toLower(attrName);

    return target[lowerAttrName] || target[attrName];
}

/**
 * custom 엘리먼트의 attr 이름 유/무 (ie(9) 대응)
 *
 * @param {string} targetName - target 이름
 * @param {string} attrName - attr 이름
 *
 * @returns {object} attr 값
 */
function _isAttrNameByCustomElement(targetName = '', attrName = ''){

    if (!_.isString(targetName)){
        CONSOLE.prop.throw('BXUI.Util.isAttrNameByCustomElement', 'targetName', 'type');
    }

    if (!_.isString(attrName)){
        CONSOLE.prop.throw('BXUI.Util.isAttrNameByCustomElement', 'attrName', 'type');
    }

    // 보통 브라우저에서는 attr 이름은 전부 소문자로 처리한다.
    let lowerAttrName = _.toLower(attrName);

    return (targetName === attrName) || (targetName === lowerAttrName);
}

/**
 * object 를 String Attribute 문자열로 변환하는 함수
 *
 * @param {object} obj - attr object
 *
 * @returns {string} string attr 값
 */
function _objectToStringAttr(obj = {}){

    if (!_.isPlainObject(obj)){
        CONSOLE.prop.throw('BXUI.Util.objectToStringAttr', 'obj', 'type');
    }

    let ret = [];

    _.map(obj, (v, k) =>{
        ret.push(`${k}="${v}"`);
    });

    return ret.join(' ');
}

/**
 * string Attribute 문자열을 object 로 변환하는 함수
 *
 * @param {string} attr - string attr 문자열
 *
 * @returns {object} attr object 값
 */
function _stringAttrToObject(attr = ''){

    if (!_.isString(attr)){
        CONSOLE.prop.throw('BXUI.Util.stringAttrToObject', 'attr', 'type');
    }

    let ret = {};

    let _attr = attr.split(' ');

    _.map(_attr, (v, k) =>{

        let attr = v.split('=');

        let _k = attr[0];
        let _v = attr[1];

        if (_k) ret[_k] = _v.replace(/(^"|"$)/g, '');
    });

    return ret;
}

/**
 * Attribute 배열을 generator 로 변환하는 함수
 *
 * @param {object} attrs - DOM object
 *
 * @returns {object} attr object
 *
 */
/* jshint ignore:start */
function* _attrsToGenerator(attrs = {}){

    // attrs `프로토타입 체인`에 NamedNodeMap.prototype 객체가 포함되지는 여부
    if (!(attrs instanceof NamedNodeMap)){
        CONSOLE.prop.throw('BXUI.Util.attrsToGenerator', 'attrs', 'type');
    }

    for (let i in attrs){

        let attr = attrs[i];

        if (attr instanceof Attr){

            let k = attr.name,
                v = attr.value;

            yield {k: k, v: v};
        }
    }

    return this;
}
/* jshint ignore:end */

/**
 * Attribute 배열을 object 로 변환하는 함수
 *
 * @param {object} attrs - DOM object
 * @param {object} callback - callback 함수 false 반환시 property 제외, true 반환시 property 설정
 *
 * @returns {object} attr object
 *
 */

function _attrsToObject(attrs = {}, callback = function (){
    return true;
}){

    // attrs `프로토타입 체인`에 NamedNodeMap.prototype 객체가 포함되지는 여부
    if (!(attrs instanceof NamedNodeMap)){
        CONSOLE.prop.throw('BXUI.Util.attrsToObject', 'attrs', 'type');
    }

    if (!_.isFunction(callback)){
        CONSOLE.prop.throw('BXUI.Util.attrsToObject', 'callback', 'type');
    }

    let ret = {};

    for (let i in attrs){

        let attr = attrs[i];

        if (attr instanceof Attr){

            let k = attr.name,
                v = attr.value;

            if (!callback(k, v)) continue;

            ret[k] = v;
        }
    }

    return ret;
}

/**
 * 객체 프로퍼티를 생성한다.
 *
 * @param {object} target - target 객체
 * @param {string} propertyName - property 이름
 * @param {value} [value=undefined] - 값
 *
 * @return {object} BXUI.Util 객체
 *
 */

function _setProperty(target = {}, propertyName = '', /* jshint ignore:start */ value = undefined /* jshint ignore:end */){

    if (!_.isPlainObject(target)){
        CONSOLE.prop.throw('BXUI.Util.setProperty', 'target', 'type');
    }

    if (!_.isString(propertyName)){
        CONSOLE.prop.throw('BXUI.Util.setProperty', 'propertyName', 'type');
    }


    let descriptor = Object.getOwnPropertyDescriptor(target, propertyName);

    if (descriptor && !descriptor.configurable){

        CONSOLE.prop.warn('BXUI.Util.setProperty', `객체 프로퍼티(${propertyName})를 재설정할 수 없으며, 전달받은 값을 엘리먼트 프로퍼티로 설정합니다.`);

        if (_.isElement(target) && value){
            $(target).prop(propertyName, value);
        }

        return this;
    }

    Object.defineProperty(target, propertyName, {
        enumerable: true,
        configurable: true,
        writable: true,
        value: value
    });

    return this;
}

/**
 * 객체에 읽기전용 프로퍼티를 생성한다.
 *
 * @param {object} target - target 객체
 * @param {string} propertyName - property 이름
 * @param {value} [value=undefined] - 값
 *
 * @return {object} BXUI.Util 객체
 *
 */

function _setReadOnlyProperty(target = {}, propertyName = '', /* jshint ignore:start */ value = undefined /* jshint ignore:end */){

    if (!_.isObject(target)){
        CONSOLE.prop.throw('BXUI.Util.setReadOnlyProperty', 'target', 'type');
    }

    if (!_.isString(propertyName)){
        CONSOLE.prop.throw('BXUI.Util.setReadOnlyProperty', 'propertyName', 'type');
    }

    let descriptor = Object.getOwnPropertyDescriptor(target, propertyName);

    if (descriptor && !descriptor.configurable){

        CONSOLE.prop.warn('BXUI.Util.setReadOnlyProperty', `객체 프로퍼티(${propertyName})를 재설정할 수 없으며, 전달받은 값을 엘리먼트 프로퍼티로 설정합니다.`);

        if (_.isElement(target) && value){
            $(target).prop(propertyName, value);
        }

        return this;
    }

    Object.defineProperty(target, propertyName, {
        enumerable: false,
        configurable: false,
        writable: false,
        value: value
    });

    return this;
}

/**
 * 객체에 비열거자 프로퍼티를 생성한다.
 *
 * @param {object} target - target 객체
 * @param {string} propertyName - property 이름
 * @param {object} [value=undefined] - 값
 *
 * @return {object} BXUI.Util 객체
 *
 */
function _setNonEnumerableProperty(target = {}, propertyName = '', /* jshint ignore:start */ value = undefined /* jshint ignore:end */){

    if (!_.isObject(target)){
        CONSOLE.prop.throw('BXUI.Util.setNonEnumerableProperty', 'target', 'type');
    }

    if (!_.isString(propertyName)){
        CONSOLE.prop.throw('BXUI.Util.setNonEnumerableProperty', 'propertyName', 'type');
    }


    let descriptor = Object.getOwnPropertyDescriptor(target, propertyName);

    if (descriptor && !descriptor.configurable){

        CONSOLE.prop.warn('BXUI.Util.setNonEnumerableProperty', `객체 프로퍼티(${propertyName})를 재설정할 수 없으며, 전달받은 값을 엘리먼트 프로퍼티로 설정합니다.`);

        if (_.isElement(target) && value){
            $(target).prop(propertyName, value);
        }

        return this;
    }

    Object.defineProperty(target, propertyName, {
        enumerable: false,
        configurable: true,
        writable: true,
        value: value
    });

    return this;
}

/**
 * 서술자(descriptor)를 통해, 객체 프로퍼티를 생성한다.
 *
 * @param {object} target - target 객체
 * @param {string} propertyName - property 이름
 * @param {object} descriptor - 서술자 객체
 * @param {object} [propertyValue=null] - property 값
 *
 * @return {object} BXUI.Util 객체
 *
 */
function _setPropertyWithDescriptor(target = {}, propertyName = '', descriptor = {}, propertyValue = null){

    if (!_.isObject(target)){
        CONSOLE.prop.throw('BXUI.Util.setPropertyWithDescriptor', 'target', 'type');
    }

    if (!_.isString(propertyName)){
        CONSOLE.prop.throw('BXUI.Util.setPropertyWithDescriptor', 'propertyName', 'type');
    }

    if (!_.isPlainObject(descriptor)){
        CONSOLE.prop.throw('BXUI.Util.setPropertyWithDescriptor', 'descriptor', 'type');
    }

    let _descriptor = Object.getOwnPropertyDescriptor(target, propertyName);

    if (_descriptor && !_descriptor.configurable){

        CONSOLE.prop.warn('BXUI.Util.setPropertyWithDescriptor', `객체 프로퍼티(${propertyName})를 재설정할 수 없으며, 전달받은 값을 엘리먼트 프로퍼티로 설정합니다.`);

        if (_.isElement(target) && propertyValue){
            $(target).prop(propertyName, propertyValue);
        }

        return this;
    }

    Object.defineProperty(target, propertyName, descriptor);

    return this;
}

/**
 * 접근자 프로퍼티를 위한, 내부 프로퍼티 이름을 생성해준다.
 *
 * @param {string} propertyName - property 이름
 *
 * @return {string} 내부 속성 이름
 *
 */
function _convertInternalPropertyName(propertyName = ''){

    if (!_.isString(propertyName)){
        CONSOLE.prop.throw('BXUI.Util.convertInternalPropertyName', 'propertyName', 'type');
    }

    return `__${propertyName}__`;
}

/**
 * style 문자열을 객체로 반환해준다.
 *
 * @param {string} [style=''] style 문자열
 *
 * @returns {object} o - style 정보를 갖고 있는 객체
 */
function _styleToObject(style = ''){

    if (!_.isString(style)){
        CONSOLE.prop.throw('BXUI.Util.styleToObject', 'style', 'type');
    }

    if (_.isEmpty(style)) return {};

    return JSON.parse(`{"${style.replace(/;$/, '').replace(/\s+/g, '').replace(/;/g, '","').replace(/:/g, '":"')}"}`);
}

/**
 * style 정보를 갖고 있는 객체를 문자열로 변환해준다.
 *
 * @param o - style 정보를 갖고 있는 객체
 *
 * @returns {string} style - style 문자열
 */
function _objectToStyle(o = {}){

    if (!_.isPlainObject(o)){
        CONSOLE.prop.throw('BXUI.Util.objectToStyle', 'o', 'type');
    }

    if (_.isEmpty(o)) return '';

    return JSON.stringify(o).replace(/("|{|})/g, '').replace(/,/g, ';') + ';';
}

/**
 * browser 가 ie 계열인지 확인 하는 함수
 *
 * @returns {boolean} isIE - ie 계열 브라우저인지 확인
 *
 */
function _isIE(){
    return JQ_BROWSER.msie || JQ_BROWSER.msedge;
}

/**
 * IE9 브라우저 인지 체크하는 함수
 *
 * @returns {boolean} isIE9 - ie9 브라우저인지 유무
 * @private
 */
function _isIE9(){
    return JQ_BROWSER.msie && JQ_BROWSER.versionNumber === 9;
}

/**
 * 전달받은 사파리 버전 보다, 하위 버전인지 체크하는 함수
 *
 * @param {int} majorVer - majorVer
 * @param {int} minorVer - minorVer
 *
 * @returns {Boolean} - 버전 여부
 *
 * @private
 */
function _isBelowSafariVersion(majorVer = null, minorVer = null){

    if (_.isNull(majorVer)){
        CONSOLE.prop.throw('BXUI.Util.isBelowSafariVersion', 'majorVer', 'type');
    }

    let ret = false;

    let browserVer = this.getBrowserVersion();
    let browserMajorVer = browserVer[0];
    let browserMinorVer = browserVer[1];

    if (JQ_BROWSER.safari){

        if (!_.isNull(majorVer) && _.isNull(minorVer)){
            ret = browserMajorVer < majorVer;
        }
        else if (!_.isNull(majorVer) && !_.isNull(minorVer)){
            ret = browserMajorVer === majorVer && browserMinorVer < minorVer;
        }
    }

    return ret;
}

/**
 * 전달받은 사파리 버전 보다, 상위 버전인지 체크하는 함수
 *
 * @param {int} majorVer - majorVer
 * @param {int} minorVer - minorVer
 *
 * @returns {Boolean} - 버전 여부
 *
 * @private
 */
function _isAboveSafariVersion(majorVer = null, minorVer = null){

    if (_.isNull(majorVer)){
        CONSOLE.prop.throw('BXUI.Util.isBelowSafariVersion', 'majorVer', 'type');
    }

    let ret = false;

    let browserVer = this.getBrowserVersion();
    let browserMajorVer = browserVer[0];
    let browserMinorVer = browserVer[1];

    if (JQ_BROWSER.safari){

        if (!_.isNull(majorVer) && _.isNull(minorVer)){
            ret = browserMajorVer > majorVer;
        }
        else if (!_.isNull(majorVer) && !_.isNull(minorVer)){
            ret = browserMajorVer === majorVer && browserMinorVer > minorVer;
        }
    }

    return ret;
}

/**
 * 브라우저 버전 정보를 반환한다.
 *
 * @returns {Array} - 브라우저 버전 정보
 *
 * @private
 */
function _getBrowserVersion(){

    let ret = [];
    let buffer = [];

    if (JQ_BROWSER.safari){

        let ua = window.navigator.userAgent;

        let mt = ua.match(/Version\/([\d.]+)/);

        if (mt) buffer = mt[1];

    }
    else{
        buffer = JQ_BROWSER.version;
    }

    _.forEach(buffer.split('.'), v => {
        ret.push(_.toNumber(v));
    });

    return ret;
}


/**
 * 특정 type별 value formatting 함수
 *
 * @param {string} formatType - formatting 할 type
 * @param {string} value - formatting 할 값
 * @param {object} options - formatting에 필요한 기타 옵션 (separator, date관련 format)
 *
 * @returns {string} result - type별 formatting된 값
 *
 * @private
 */
function _getFormatString(formatType = '', value = '', options = {}){

    if (!_.isString(formatType) || formatType.length === 0){
        CONSOLE.prop.throw('BXUI.Util.getFormatString', 'type', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw('BXUI.Util.getFormatString', 'value', 'type');
    }

    if (!_.isPlainObject(options)){
        CONSOLE.prop.throw('BXUI.Util.getFormatString', 'options', 'type');
    }

    let reg = null;
    let resObj = {};
    let pattern = '';
    let cardName = '';
    let defaultSeparator = '';
    let type = formatType.toLowerCase();

    switch (type){

        case 'exp':{
            defaultSeparator = '/';
            break;
        }
        case 'currency': {
            defaultSeparator = ',';
            break;
        }
        case 'pct': {
            defaultSeparator = '%';
            break;
        }
        default : {
            defaultSeparator = '-';
        }
    }

    let	separator = !_.isUndefined(options.separator) && options.separator !== '' ? options.separator : defaultSeparator;

    value = _.isNumber(value) ? '' + value : value;

    if (type !== 'date' && type !== 'pct' && type !== 'currency'){
        value = _formatLength(type, value);
    }

    switch (type){

        case 'currency':

            let prefix = options.prefix || '';
            let suffix = options.suffix || '';
            let decimalPlaces = TYPE.isNumber(options.decimalPlaces) ? options.decimalPlaces : 0;

            let tempData = value.split('.');

            // 정수부분 콤마 변경
            tempData[0] = tempData[0].replace(/\B(?=(\d{3})+(?!\d))/g, separator);

            // 소수 부분 자릿수 제한
            if (!_.isUndefined(tempData[1])) tempData[1] = tempData[1].substr(0, decimalPlaces);

            // 소수점 영역 결합
            tempData = tempData.join('.');

            // prefix와 suffix 결합
            value = prefix + tempData + suffix;

            break;

        case 'date':

            let dataFormat = options.dataFormat;
            let viewFormat = options.viewFormat;
            let dateObj = this.extractDateFormat(dataFormat, value);

            // safari에서 ISO8601 type만 date 변환이 되어 모두 적용되도록 수정
            let dateString = `${dateObj.year}-${dateObj.month}-${dateObj.day}T${dateObj.hour}:${dateObj.min}:${dateObj.sec}+09:00`;

            if (this.isDate(new Date(dateString))){
                value = this.convertDateFormat(dateString, viewFormat);
            }

            break;
        case 'tel':

            if (value.length < 8){
                reg = /(^02.{0}|^01\d{1}|\d{3})(\d+)/;
                pattern = `$1${separator}$2`;
            }
            else{
                reg = /(^02.{0}|^01\d{1}|\d{3})(\d+)(\d{4})/;
                pattern = `$1${separator}$2${separator}$3`;
            }

            break;
        case 'pid':

            reg = /(\d{6})(\d{1,7})/;
            pattern = `$1${separator}$2`;

            break;
        case 'bzno':

            if (value.length < 6){
                reg = /(\d{3})(\d{1,2})/;
                pattern = `$1${separator}$2`;
            }
            else{
                reg = /(\d{3})(\d{0,2})(\d{1,5})/;
                pattern = `$1${separator}$2${separator}$3`;
            }

            break;
        case 'card':

            let amex = /^3[47]\d{1,13}/,
                dinersClub = /^3(0[0-5]|[68]\d)\d{1,11}/,
                masterCard = /^5[1-5]\d{14}]/,
                visa = /^4\d{12}(\d{3})?/,
                discover = /(6(011|5\d{2})\d{12})/,
                jcb = /(^2131|^1800|^35\d{3})\d{11}/,
                bc = /^94\d{12}/;

            value = value.replace(/\D/g,'');

            //american express : start with 34/37
            if (amex.test(value)){

                cardName = 'amex';

                if (value.length < 11){
                    reg = /(\d{4})(\d{1,6})/;
                    pattern = `$1${separator}$2`;
                }
                else{
                    reg = /(\d{4})(\d{6})(\d{1,5})/;
                    pattern = `$1${separator}$2${separator}$3`;
                }
            }
            //diners club : start with 30
            else if (dinersClub.test(value)){
                cardName = 'diners Club';

                if (value.length < 11){
                    reg = /(\d{4})(\d{1,6})/;
                    pattern = `$1${separator}$2`;
                }
                else{
                    reg = /(\d{4})(\d{6})(\d{1,4})/;
                    pattern = `$1${separator}$2${separator}$3`;
                }

            }
            else{
                if (masterCard.test(value)){
                    cardName = 'masterCard';
                }
                else if (visa.test(value)){
                    cardName = 'visa';
                }
                else if (discover.test(value)){
                    cardName = 'discover';
                }
                else if (jcb.test(value)){
                    cardName = 'jcb';
                }
                else if (bc.test(value)){
                    cardName = 'bc';
                }
                else {
                    cardName = 'etc';
                }

                if (value.length < 9){
                    reg = /(\d{4})(\d{1,4})/;
                    pattern = `$1${separator}$2`;
                }
                else if (value.length < 13){
                    reg = /(\d{4})(\d{4})(\d{1,4})/;
                    pattern = `$1${separator}$2${separator}$3`;
                }
                else{
                    reg = /(\d{4})(\d{4})(\d{4})(\d{1,4})/;
                    pattern = `$1${separator}$2${separator}$3${separator}$4`;
                }

            }

            resObj.cardName = cardName;

            break;
        case 'exp':
            reg = /(\d{2})(\d{1,2})/;
            pattern = `$1${separator}$2`;

            break;
        case 'zip':
            reg = /[^0-9]/g;
            pattern = ``;

            break;
        case 'pct':
            reg = /([0-9\.]+)/;
            pattern = `$1${separator}`;

            break;
        case 'byte':

            let tempValue = parseInt(value);

            if (!isNaN(tempValue)) {

                let step = options.step || 0;
                const unit = ['B', 'KB', 'MB', 'GB', 'TB'];
                const kb = 1024;

                if (step >= unit.length){
                    value = value + unit[unit.length - 1];
                }
                else if (tempValue >= kb){

                    let calcValue = parseInt(tempValue / kb);

                    value = _getFormatString('byte', calcValue.toString(), {step: ++step});
                }
                else {
                    value = value + unit[step];
                }
            }

            break;
    }

    resObj.value = _.isNull(reg) ? value : value.replace(reg, pattern);

    return type === 'card' ? resObj : resObj.value;
}

/**
 * 특정 type별 value unformatting 함수
 *
 * @param {string} formatType - unformatting 할 type
 * @param {string} value - unformatting 할 값
 * @param {object} options - unformatting에 필요한 기타 옵션
 *
 * @returns {string} result - type별 unformatting된 값
 *
 * @private
 */
function _getUnFormatString(formatType = '', value = '', options = {}){

    if (!_.isString(formatType) || formatType.length === 0){
        CONSOLE.prop.throw('BXUI.Util.getUnFormatString', 'type', 'type');
    }

    if (!_.isPlainObject(options)){
        CONSOLE.prop.throw('BXUI.Util.getUnFormatString', 'options', 'type');
    }

    let result = '';

    if (formatType === 'date'){

        let dataFormat = options.dataFormat;
        let viewFormat = options.viewFormat;
        let dateObj = this.extractDateFormat(viewFormat, value);

        // safari에서 ISO8601 type만 date 변환이 되어 모두 적용되도록 수정
        let dateString = `${dateObj.year}-${dateObj.month}-${dateObj.day}T${dateObj.hour}:${dateObj.min}:${dateObj.sec}+09:00`;

        if (this.isDate(new Date(dateString))){
            result = this.convertDateFormat(dateString, dataFormat);
            result = result === '--' ? value : result;
        }
        else{
            result = value;
        }
    }
    else{
        // 숫자가 아닌 모든 문자를 삭제
        result = value.replace(/\D/g, '');
    }

    return result;
}
/**
 * 특정 type별 format 길이 check 함수
 *
 * @param {string} type - 길이 체크 할 type
 * @param {string} value - type별 길이 반영 전 value
 *
 * @returns {string} value - type별 길이 반영 된 value
 *
 * @private
 */
function _formatLength(type = '', value = ''){

    if (!_.isString(type) || type.length === 0){
        CONSOLE.prop.throw('BXUI.Util.formatLength', 'type', 'type');
    }

    value = value.replace(/\D/g, '');

    switch(type){
        case 'pid': value = value.substr(0, 13); break;

        case 'bzno': value = value.substr(0, 10); break;

        case 'tel':
            let seoulLocal = /^02\d{1,8}/;

            if (seoulLocal.test(value)) value = value.substr(0, 10);
            else value = value.substr(0, 11);

            break;

        case 'card':

            let amex = /^3[47]\d{1,13}/,
                dinersClub = /^3(0[0-5]|[68]\d)\d{1,11}/;

            if (amex.test(value)) value = value.substr(0, 15);
            else if (dinersClub.test(value)) value = value.substr(0, 14);
            else value = value.substr(0, 16);

            break;

        case 'date': value = value.substr(0, 8); break;
        case 'exp': value = value.substr(0, 4); break;
        case 'zip': value = value.substr(0, 5); break;

    }

    return value;
}

/**
 * 특정 type별 value formatting check 함수
 *
 * @param {string} type - formatting 체크 할 type
 * @param {string} value - formatting 체크 할 value
 *
 * @returns {string} value - type별 체크완료 된 value
 *
 * @private
 */
function _checkFormatString(type = '', value = ''){

    if (!_.isString(type) || type.length === 0){
        CONSOLE.prop.throw('BXUI.Util.checkFormatString', 'type', 'type');
    }

    let reg = null;

    value = _.isNumber(value) ? '' + value : value;

    switch(type.toLowerCase()){

        //email, ip는 현재 사용하지 않음 (추후 예정)
        case 'email': reg = /^([\w-]+(?:\.[\w-]+)*)@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$/;break;
        case 'ip': reg = /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/;break;
    }

    if (reg.test(value) === false) {
        value = '';
    }

    return value;
}

/**
 * format 형식으로 값을 추출
 *
 * @param {string} format - 추출 할 format 형식
 * @param {string} value - 추출 할 값
 *
 * @returns {object} result - date값 별로 추출된 날짜들
 *
 * @private
 */
function _extractDateFormat(format = '', value = ''){

    if (!_.isString(format) || format.length === 0){
        CONSOLE.prop.throw('BXUI.Util.extractDateFormat', 'format', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw('BXUI.Util.extractDateFormat', 'value', 'type');
    }

    let _year = '',
        _month = '',
        _day = '',
        _hour = '',
        _min = '',
        _sec = '';

    value = String(value);

    _.forEach(format, (formatString, index)=>{
        let v = value[index];

        switch (formatString){
            case 'y':
                _year += String(v);
                break;
            case 'M':
                _month += String(v);
                break;
            case 'd':
                _day += String(v);
                break;
            case 'h':
                _hour += _.isUndefined(v) ? '0' : String(v);
                break;
            case 'm':
                _min += _.isUndefined(v) ? '0' : String(v);
                break;
            case 's':
                _sec += _.isUndefined(v) ? '0' : String(v);
                break;
        }
    });

    _hour = _hour === '' ? '00' : _hour;
    _min = _min === '' ? '00' : _min;
    _sec = _sec === '' ? '00' : _sec;

    return {
        year: _year,
        month: _month,
        day: _day,
        hour: _hour,
        min: _min,
        sec: _sec
    };
}

/**
 * dateString을 format형태로 변경
 *
 * @param {string} dateString - date 형식 날짜 string 값
 * @param {string} format - 변경 할 format
 * @param {boolean} is24Hours - 24시간 표시 설정 유무
 *
 * @returns {string} result - format 형태로 변환된 문자열
 *
 * @private
 */
function _convertDateFormat(dateString = '', format = '', is24Hours = false){

    if (!_.isString(dateString)){
        CONSOLE.prop.throw('BXUI.Util.convertDateFormat', 'dateString', 'type');
    }

    if (!_.isString(format)){
        CONSOLE.prop.throw('BXUI.Util.convertDateFormat', 'format', 'type');
    }


    let date = new Date(dateString);
    let ret = '';

    if (!this.isDate(date)){
        return dateString;
    }

    ret = format.replace(/(yyyy|yy|MM|dd|hh|mm|ss|a\/p)/gi, (v)=>{
        switch (v){
            case 'yyyy':
                return String(date.getFullYear());
            case 'yy':
                return String(date.getFullYear()).substr(2, 2);
            case 'MM':
                return date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : String(date.getMonth() + 1);
            case 'dd':
                return date.getDate() < 10 ? `0${date.getDate()}` : String(date.getDate());
            case 'hh':

                let hours = date.getHours();

                if (is24Hours){
                    hours = hours;
                }
                else if (hours % 12 === 0){
                    hours = 12;
                }
                else {
                    hours = hours % 12;
                }

                return hours < 10 ? `0${hours}` : String(hours);

            case 'mm':
                return date.getMinutes() < 10 ? `0${date.getMinutes()}` : date.getMinutes();
            case 'ss':
                return date.getSeconds() < 10 ? `0${date.getSeconds()}` : date.getSeconds();
            case 'a/p':
                return date.getHours() < 12 ? 'AM' : 'PM';
            default:
                return v;
        }
    });

    return ret;
}

/**
 *
 * date 객체에 대한, 유효성 검사 체크 함수
 *
 * @param {object} [date=null] date 객체
 *
 * @returns {boolean} date 체크 값
 *
 * @private
 */
function _isDate(date = null){

    if (!_.isObject(date)){
        CONSOLE.prop.throw('BXUI.Util.isDate', 'date', 'type');
    }

    // date 객체 여부
    let ret = _.isDate(date);

    // new Date 메서드로 전달된 스트링이 유효한 경우에는 `Invalid Date` 가 반환되지 않는다.
    ret = ret && _.toString(date) !== 'Invalid Date' ? true : false;

    return ret;
}

/**
 * $elem 의 scrollbar size 반환 함수
 * @param {{}} $elem 엘리먼트 객체
 * @returns {number} scrollBarSize
 * @private
 */
function _getElementScrollBarSize($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.getBrowserScrollBarSize', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.getBrowserScrollBarSize', '$elem', 'type');
    }

    let {clientWidth, offsetWidth} = $elem.get(0);

    let scrollBarSize = offsetWidth - clientWidth;

    return scrollBarSize;
}

/**
 *
 * element에 대한 보이는 넓이 산출 후, 관련 정보 리턴 함수
 *
 * @param {Element} $elem - 엘리먼트 객체
 *
 * @return {object} element 의 원래 넓이, 보이는 넓이, 가려지는지 유무
 *
 * @private
 */
function _checkVisibleArea($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.checkVisibleArea', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.checkVisibleArea', '$elem', 'type');
    }

    let $window = $(window);

    let {top, left} = $elem.offset(),

        innerWidth = $window.innerWidth(),

        innerHeight = $window.innerHeight(),

        width = $elem.width(),

        height = $elem.height();

    let scrollLeft = $window.scrollLeft(),
        scrollTop = $window.scrollTop();

    // 4면의 안보이는 길이 체크 후, 저장
    let leftMinus = (left - scrollLeft < 0) ? Math.abs(left - scrollLeft) : 0,

        rightMinus = ((left - scrollLeft) + width > innerWidth) ? Math.abs(((left - scrollLeft) + width) - innerWidth) : 0,

        topMinus = (top - scrollTop < 0) ? Math.abs(top - scrollTop) : 0,

        bottomMinus = ((top - scrollTop) + height > innerHeight) ? Math.abs(((top - scrollTop) + height) - innerHeight) : 0;

    //originArea : element의 전체 넓이, visibleArea : 현재 보이고 있는 넓이, isFullvisible : 전체 영역이 보이는지에 유무
    let originArea = width * height,

        visibleWidth = width - (leftMinus + rightMinus),

        visibleHeight = height - (topMinus + bottomMinus),

        visibleArea = Math.round(visibleWidth * visibleHeight),

        isFullVisible = (originArea === visibleArea) ? true : false;

    return {originArea, visibleArea, isFullVisible};
}

/**
 *
 * 각 `custom` 엘리먼트에 정의된 attr 를 opt 객체로 반환한다.
 *
 * @param {Object} targetObject - 대상 객체
 * @param {Object} defaultOpts - 기본 옵션 객체
 * @param {Object} attrNames - IE9 대응을 위한, attr 객체
 * @param {Object} withOptionKeyCallBack - 전달받은 옵션 객체가, 해당 키를 가진경우 호출될 callback 함수
 * @param {Object} withOptionKeyCallBack - 전달받은 옵션 객체가, 해당 키를 갖지못한경우 호출될 callback 함수
 *
 * @returns {object} 생성된 옵션 객체
 *
 * @private
 */
function _createComponentOptions(targetObject = null, defaultOpts = null, attrNames = {}, withOptionKeyCallBack = function (){}, withoutOptionKeyCallBack = function (){}){

    if (!_.isFunction(targetObject.get) && !_.isPlainObject(targetObject)){
        CONSOLE.prop.throw('BXUI.Util.createComponentOptions', 'targetObject', 'type');
    }

    if (!_.isPlainObject(defaultOpts)){
        CONSOLE.prop.throw('BXUI.Util.createComponentOptions', 'defaultOpts', 'type');
    }

    if (!_.isPlainObject(attrNames)){
        CONSOLE.prop.throw('BXUI.Util.createComponentOptions', 'attrNames', 'type');
    }

    if (!_.isFunction(withOptionKeyCallBack)){
        CONSOLE.prop.throw('BXUI.Util.createComponentOptions', 'withOptionKeyCallBack', 'type');
    }

    if (!_.isFunction(withoutOptionKeyCallBack)){
        CONSOLE.prop.throw('BXUI.Util.createComponentOptions', 'withoutOptionKeyCallBack', 'type');
    }

    let attrs = null;

    let opts = _.clone(defaultOpts);

    opts.attr = {};
    opts.events = {};

    let bxPtn = /^bx-/;

    // 인자값이 엘리먼트일 경우
    if (_.isFunction(targetObject.get) && _.isElement(targetObject.get(0))){
        attrs = _attrsToObject(targetObject.get(0).attributes);
    }
    else{
        attrs = targetObject;
    }

    _.map(attrs, (v, k) =>{

        k = k.toLowerCase();

        k = attrNames[k] ? attrNames[k] : k;

        v = _.trim(v);

        // native event 를 사용하지 못하도록 함. (bx-<eventName> 사용)
        if (k.match(/^on[a-z]+/)){
            CONSOLE.prop.warn('BXUI.Util.createComponentOptions', k, 'type');
            return;
        }

        // bx 이벤트일 경우
        if (k.match(bxPtn)){

            v = this.parseObject(k, v);

            if (!_.isFunction(v)){
                CONSOLE.prop.throw('BXUI.Util.createComponentOptions', k, 'type');
            }
        }

        if (v === 'true') v = true;
        else if (v === 'false') v = false;

        if (opts.hasOwnProperty(k)){

            let _v = withOptionKeyCallBack.apply(opts, [k, v]);

            if (_.isUndefined(_v)) _v = v;

            opts[k] = _v;
        }
        else{

            let _v = withoutOptionKeyCallBack.apply(opts, [k, v]);

            if (_.isUndefined(_v)) _v = v;

            if (k.match(bxPtn)){
                opts.events[k] = _v;
            }
            else{

                opts.attr[k] = _v;
            }
        }
    });


    return opts;
}

/**
 * object parsing 함수
 * @param {string} attribute name
 * @param {string} attribute value
 * @private
 */
function _parseObject(name = '', string = ''){

    if (!_.isString(name)){
        CONSOLE.prop.throw('BXUI.Util.parseObject', 'name', 'type');
    }

    if (!_.isString(string)){
        CONSOLE.prop.throw('BXUI.Util.parseObject', 'string', 'type');
    }

    try{
        return this.JSParser(string);
    }
    catch (e){
        CONSOLE.prop.throw('BXUI.Util.parseObject', name, 'type');
    }
}

/**
 * window Safari 인지 유무 반환 함수
 * @returns {boolean}
 * @private
 */
function _isAppliedCSSCalc(){

    // calc() 지원 브라우저
    // https://www.w3schools.com/cssref/func_calc.asp
    let supportBrowserVersions = {
        chrome: 26,
        msie: 9,
        mozilla: 16,
        opera: 15,
        safari: 7
    };

    let {name, versionNumber} = JQ_BROWSER;

    return supportBrowserVersions[name] >= versionNumber;
}
/**
 * 포지션 지정 함수
 * @param {{}} [$elem=null] - jquery element instance
 * @param {{}} [position={}] - 포지션 객체
 * @param {int} [position.top=0]- top pixel
 * @param {int} [position.left=0] - left pixel
 * @param {int} [position.right=0] - right pixel
 * @param {int} [position.bottom=0] - bottom pixel
 * @param {boolean} [isRound=false] - 반올림 유무
 * @private
 */
function _setPosition($elem = null, {top = 0 , left = 0, right = 0, bottom = 0} = {}, isRound = false){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.setPosition', '$elem', 'value');
    }

    let position = {};

    Object.assign(position, {
        top, left, right, bottom
    });

    for (let k in position){

        let v = isRound ? Math.round(position[k]) : position[k];

        if (_.isEqual(v, 0)) continue;

        $elem.css(k, `${v}px`);
    }
}

/**
 * $elem 의 하위 selector 에 해당하는 객체 반환 함수
 * @param $elem - jquery instance
 * @param selector - selector class 객체
 * @returns {{}}
 * @private
 */
function _getSelectorInstance($elem = null, selector = {}){

    let o = {};

    _.map(selector, (v, k) =>{
        o[k] = $elem.find(`.${v}`);
    });

    return o;
}

/**
 * IE9에서 placeholder를 지원하는 함수
 *
 * @param $elem - jquery instance
 * @param placeHolderText - placeholder text
 * @param separateValue - separateValue text
 * @private
 */
function _setIE9PlaceHolder($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.setPlaceHolder', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.setPlaceHolder', '$elem', 'type');
    }

    if (!this.isIE9()) return;

    let className = 'placeholder';

    let value = _.trim($elem.val());
    let placeholder = _.trim($elem.attr('placeholder'));

    let isPlaceHolder = $elem.hasClass(className);


    if (!value || isPlaceHolder){
        $elem.val(placeholder);
        $elem.addClass(className);
    }

    $elem.on('focusin', e => {

        let isPlaceHolder = $elem.hasClass(className);

        if (isPlaceHolder){
            $elem.val('');
            $elem.removeClass(className);
        }
    });

    $elem.on('focusout', e => {

        let value = _.trim($elem.val());
        let placeholder = _.trim($elem.attr('placeholder'));

        if (!value){
            $elem.val(placeholder);
            $elem.addClass(className);
        }
    });

    return this;
}

/**
 * 브라워저 별 `window.location.origin` 값을 반환한다.
 *
 * @returns {string} ret - locationOrigin 값
 *
 * @private
 */
function _getLocationOrigin(){

    let ret = '';

    let location = window.location;

    if (this.isIE9()){
        ret = location.protocol + '//' + location.host;
    }
    else{
        ret = location.origin;
    }

    return ret;
}

/**
 * 전달받은 URL 값을통해, location 객체를 반환한다.
 *
 * @param url - url
 *
 * @returns {object} ret - location 객체
 *
 * @private
 */
function _getLocationWithUrl(url = ''){

    if (!_.isString(url) || !_.trim(url).length){
        CONSOLE.prop.throw('BXUI.Util._getLocationWithUrl', 'url', 'all');
    }

    let ptn = /^(http:|https:)+\/\/([^\/]+)/;

    let mt = ptn.exec(url);

    let origin = '';
    let protocol = '';
    let host = '';
    let href = '';
    let port = '';
    let hostName = '';

    if (mt){

        origin = mt[0];
        protocol = mt[1];
        host = mt[2];
        href = url;
        port = '';
        hostName = host;

        // port 값 추출
        if (host.indexOf(':') > -1){

            let hostArr = host.split(':');

            if (hostArr.length){
                port = hostArr[1];
            }

            hostName = hostArr[0];
        }
    }

    return {
        origin,
        protocol,
        host,
        href,
        port,
        hostName
    };
}

/**
 * 전달받은 URL 값을통해, location origin 값을 반환한다.
 *
 * @param url - url
 *
 * @returns {string} location.origin - location origin 값
 *
 * @private
 */
function _getLocationOriginWithUrl(url = ''){

    if (!_.isString(url) || !_.trim(url).length){
        CONSOLE.prop.throw('BXUI.Util.getLocationOriginWithUrl', 'url', 'all');
    }

    let location = _getLocationWithUrl(url);

    return location.origin;
}

/**
 *
 * 문자나 숫자로 된 데이터 크기를 Byte 값으로 계산한다
 *
 * @param {String} value - 데이터 사이즈가 표시가 된 문자열
 *
 * @returns {Number} result - 계산된 Byte 크기
 *
 * @private
 *
 */
function _convertStringToByte(value = '') {

    let number = _.parseInt(value);
    let dataType = _.isNumber(value) ? -1 : value.replace(/\d/gi,'');
    let unit = ['B', 'KB', 'MB', 'GB', 'TB'];
    let kb = 1024;
    let result = value;

    if (!_.isNaN(number)){

        let pow = _.indexOf(unit, dataType);

        pow = pow === -1 ? _.indexOf(unit, 'MB') : pow;

        result = Math.pow(kb, pow) * number;
    }

    return result;
}

/**
 * 받은 element의 위치가 viewport 안에 있는지 체크한다
 *
 * @param {object} $elem - 체크 할 element
 *
 * @returns {boolean}
 *
 * @private
 */
function _isInViewport($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw('BXUI.Util.isInViewport', '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw('BXUI.Util.isInViewport', '$elem', 'type');
    }

    let elem = $elem.get(0);
    // 엘리먼트 크기와 위치값을 반환한다.
    let rect = elem.getBoundingClientRect();

    if (!rect.width || !rect.height) return false;

    let docElem = document.documentElement;

    // 가시화면(실제 눈으로 보는 크기) 사이즈
    let viewportWidth = docElem.clientWidth;
    let viewportHeight = docElem.clientHeight;

    // 아래 모든 값들은, 가시화면의 top 값의 상대적 위치값을 반환한다.

    // 즉 반환되는 top 값은, (요소의 원래 top 위치 - scrollTop) 값이 된다.
    // 즉 가장 상위 요소의 top 값이 0px 이 었을 경우, 이후 40px 만큼 아래로 스크롤하면, 이 요소의 top 값은 -40px 로 변경된다.(가시화면의 상대적 위치값)
    // 또한 bottom 값은 아래에서 위로 스크롤할 경우를 위해 쓰인다.(원리는 동일하다)
    let top = rect.top >= 0 && rect.top < viewportHeight;
    let bottom = rect.bottom >= 0 && rect.bottom < viewportHeight;
    let left = rect.left >= 0 && rect.left < viewportWidth;
    let right = rect.right >= 0 && rect.right < viewportWidth;

    return (top || bottom) && (left || right);
}


/**
 * JSON 데이터를 CSV 데이터로 변환한다.
 *
 * @param {Array|Object} v - json 데이터
 * @param {Boolean} enableHeader - header 데이터 적용 여부
 *
 * @returns {String} records - CSV 데이터
 *
 * @private
 *
 */
function _JSONToCSV(v = [], enableHeader = true){

    if (!_.isArray(v) && !_.isPlainObject(v)){
        CONSOLE.prop.throw('BXUI.Util.JSONToCSV', 'v', 'all');
    }

    // CSV 설명: https://ko.wikipedia.org/wiki/CSV_(%ED%8C%8C%EC%9D%BC_%ED%98%95%EC%8B%9D)
    v = _.isArray(v) ? v : [v];

    let records = '';
    let delimiter = ',';
    let newLine = '\n';

    if (enableHeader){

        let header = v[0];

        if (!_.isPlainObject(header)){
            CONSOLE.prop.throw('BXUI.Util.JSONToCSV', 'header', 'type');
        }

        _.map(header, (v, k) => {
            records += _.trim(k) + `${delimiter}`;
        });

        // 두번째 argument 음수는, 배열의 마지막 요소부터 indexing 한다.(-1(배열의 마지막 첫번째 요소를 나타낸다))
        records = records.slice(0, -1);
        records += newLine;
    }

    _.forEach(v, (vv) => {

        _.map(vv, (vvv) => {
            records += _.trim(vvv) + `${delimiter}`;
        });

        records = records.slice(0, -1);
        records += newLine;
    });

    records = records.replace(/([\n])?$/g, '');

    return records;
}


/**
 * CSV 데이터를 JSON 데이터로 변환한다.
 *
 * @param {Array|Object} v - CSV 데이터
 * @param {Boolean} isIncludeHeader - header 데이터 포함 여부
 *
 * @returns {Object} records - JSON 데이터
 *
 * @private
 */
function _CSVToJSON(v = '', isIncludeHeader = true){

    if (!_.isString(v) || !_.trim(v).length){
        CONSOLE.prop.throw('BXUI.Util.CSVToJSON', 'v', 'all');
    }

    // CSV 설명: https://ko.wikipedia.org/wiki/CSV_(%ED%8C%8C%EC%9D%BC_%ED%98%95%EC%8B%9D)
    v = v.split('\n');

    let records = [];

    let header = '';

    if (isIncludeHeader){

        header = v[0];
        v.shift();
    }
    else{

        _.forEach(v[0].split(','), (v, k) => {
            header += k + ',';
        });

        header = header.slice(0, -1);
    }

    let keys = header.split(',');

    _.forEach(v, (vv) => {

        let t = {};

        _.forEach(vv.split(','), (vvv, kkk) => {
            t[keys[kkk]] = vvv;
        });

        records.push(t);
    });


    return records;
}

/**
 * BASE64 데이터로 인코딩한다.
 *
 * @param {String} v - 인코딩될 문자
 *
 * @returns {String} ret - BASE64 데이터
 *
 * @private
 */
function _encodeBase64(v){

    if (!_.isString(v) || !_.trim(v).length){
        CONSOLE.prop.throw('BXUI.Util.encodeBase64', 'v', 'all');
    }

    // base64 변환 방법
    // http://dakuo.tistory.com/entry/base64-%EC%9D%B8%EC%BD%94%EB%94%A9-%EC%95%8C%EA%B3%A0%EB%A6%AC%EC%A6%98

    // encodeURIComponent: 영문 대/소, 숫자가 아니라면, 모두 유니코드 문자(https://ko.wikipedia.org/wiki/%EC%9C%A0%EB%8B%88%EC%BD%94%EB%93%9C)로 인코딩한다.
    // 단 일부 특수문자(!, . 등)는 제외된다.
    // 유니코드표: http://www.tamasoft.co.jp/en/general-info/unicode.html
    if (window.btoa) return window.btoa(encodeURIComponent(v));

    /* jshint ignore:start */
    let symbols = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';

    let ret = '';

    let chr1 = null;
    let chr2 = null;
    let chr3 = null;

    let enc1 = null;
    let enc2 = null;
    let enc3 = null;
    let enc4 = null;

    let i = 0;

    v = _encodeUTF8(v);

    while (i < v.length) {

        chr1 = v.charCodeAt(i++);
        chr2 = v.charCodeAt(i++);
        chr3 = v.charCodeAt(i++);

        enc1 = chr1 >> 2;
        enc2 = ((chr1 & 3) << 4) | (chr2 >> 4);
        enc3 = ((chr2 & 15) << 2) | (chr3 >> 6);
        enc4 = chr3 & 63;

        if (isNaN(chr2)) {
            enc3 = enc4 = 64;
        }
        else if (isNaN(chr3)) {
            enc4 = 64;
        }


        ret = ret + symbols.charAt(enc1) + symbols.charAt(enc2) + symbols.charAt(enc3) + symbols.charAt(enc4);
    }

    /* jshint ignore:end */

    return ret;
}

/**
 * BASE64 문자를 디코딩한다.
 *
 * @param {String} v - BASE64 문자
 *
 * @returns {String} ret - 디코딩된 문자
 *
 * @private
 */
function _decodeBase64(v){

    if (!_.isString(v) || !_.trim(v).length){
        CONSOLE.prop.throw('BXUI.Util.decodeBase64', 'v', 'all');
    }

    // encodeURIComponent: 영문 대/소, 숫자가 아니라면, 모두 유니코드 문자(https://ko.wikipedia.org/wiki/%EC%9C%A0%EB%8B%88%EC%BD%94%EB%93%9C)로 인코딩한다.
    // 단 일부 특수문자(!, . 등)는 제외된다.
    // 유니코드표: http://www.tamasoft.co.jp/en/general-info/unicode.html
    if (window.atob) return decodeURIComponent(window.atob(v));

    /* jshint ignore:start */
    let symbols = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';

    let ret = '';

    let chr1 = null;
    let chr2 = null;
    let chr3 = null;

    let enc1 = null;
    let enc2 = null;
    let enc3 = null;
    let enc4 = null;

    let i = 0;

    v = v.replace(/[^A-Za-z0-9\+\/\=]/g, '');

    while (i < v.length) {

        enc1 = symbols.indexOf(v.charAt(i++));
        enc2 = symbols.indexOf(v.charAt(i++));
        enc3 = symbols.indexOf(v.charAt(i++));
        enc4 = symbols.indexOf(v.charAt(i++));

        chr1 = (enc1 << 2) | (enc2 >> 4);
        chr2 = ((enc2 & 15) << 4) | (enc3 >> 2);
        chr3 = ((enc3 & 3) << 6) | enc4;

        ret = ret + String.fromCharCode(chr1);

        if (enc3 !== 64) {
            ret = ret + String.fromCharCode(chr2);
        }


        if (enc4 !== 64) {
            ret = ret + String.fromCharCode(chr3);
        }
    }

    ret = _decodeUTF8(ret);

    /* jshint ignore:end */

    return ret;
}

/**
 * UTF8 데이터로 인코딩한다.
 *
 * @param {String} v - 인코딩될 문자
 *
 * @returns {String} ret - UTF8 데이터
 *
 * @private
 */
function _encodeUTF8(v = ''){

    if (!_.isString(v) || !_.trim(v).length){
        CONSOLE.prop.throw('BXUI.Util.encodeUTF8', 'v', 'all');
    }

    v = v.replace(/\r\n/g, '\n');

    /* jshint ignore:start */

    let ret = '';

    for (let i = 0; i < v.length; i++){

        let c = v.charCodeAt(i);

        if (c < 128) {
            ret += String.fromCharCode(c);
        }
        else if ((c > 127) && (c < 2048)){
            ret += String.fromCharCode((c >> 6) | 192);
            ret += String.fromCharCode((c & 63) | 128);
        }


        else {
            ret += String.fromCharCode((c >> 12) | 224);
            ret += String.fromCharCode(((c >> 6) & 63) | 128);
            ret += String.fromCharCode((c & 63) | 128);

        }
    }

    /* jshint ignore:end */

    return ret;
}

/**
 * UTF8 문자를 디코딩한다.
 *
 * @param {String} v - UTF8 문자
 *
 * @returns {String} ret - 디코딩된 문자
 *
 * @private
 */
function _decodeUTF8(v = ''){

    if (!_.isString(v) || !_.trim(v).length){
        CONSOLE.prop.throw('BXUI.Util.decodeUTF8', 'v', 'all');
    }

    /* jshint ignore:start */

    let ret = '';

    let i = 0;
    let c = 0;
    let c1 = 0;
    let c2 = 0;

    while (i < v.length) {

        c = v.charCodeAt(i);

        if (c < 128) {
            ret += String.fromCharCode(c);
            i++;
        }
        else if ((c > 191) && (c < 224)) {
            c1 = v.charCodeAt(i + 1);
            ret += String.fromCharCode(((c & 31) << 6) | (c1 & 63));
            i += 2;
        }
        else{

            c1 = v.charCodeAt(i + 1);
            c2 = v.charCodeAt(i + 2);

            ret += String.fromCharCode(((c & 15) << 12) | ((c1 & 63) << 6) | (c2 & 63));
            i += 3;
        }
    }

    /* jshint ignore:end */

    return ret;
}

/**
 * DATA URI Scheme 지원 여부를 반환한다.
 *
 * DATA URI Scheme 이란? https://hyeonseok.com/soojung/webstandards/2011/02/17/641.html
 *
 * @returns {object} promise - promise 객체
 *
 * @private
 */
function _isSupportedDataURI(){

    let deferred = $.Deferred();
    let promise = deferred.promise();

    let ret = false;

    let $img = $(new Image());

    $img.on('load', callback);
    $img.on('error', callback);


    function callback(){

        ret = this.width === 1 ? true : false;
        deferred.resolve(ret);
    }

    // 가로 / 세로 1px 짜리 이미지 데이터
    // 주의: IE8 이상 data uri scheme 을 지원한다.
    $img.prop('src', 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==');

    return promise;
}

/**
 * 이전 포커스 된 jquery instance 객체 반환 함수
 *
 * @returns {object} o - 이전 포커스 객체의 jquery instance
 *
 * @private
 *
 */
function _getPreFocusInstance(){

    let $pre = window.preFocusElement;
    let key = $pre.key;

    return $pre[key];
}

/**
 * BXUI.Base.preFocusElement 의 key 값 setting 함수.
 *
 * @param {string} k - BXUI.Base.preFocusElement 의 key 값
 *
 * @private
 *
 */
function _setPreFocusInstanceKey(k = ''){

    let $pre = window.preFocusElement;

    if (k !== 'base' && k !== 'hoverer'){
        CONSOLE.prop.throw('BXUI.Util.setPreFocusInstanceKey', `${k} 값을 preFocusInstance 의 key 값으로 사용할 수 없습니다.`);
    }

    $pre.key = k;
}


/**
 * AngularJS 사용 유무
 *
 * @returns {boolean}
 *
 * @private
 *
 */
function _isAngular(){
    return window.angular;
}

/**
 * ReactJS 사용 유무

 * @returns {boolean}
 *
 * @private
 *
 */
function _isReact(){
    return window.React && window.ReactDOM;
}

/**
 * 오브젝트 내부 프로퍼티를 삭제시킨다.

 * @returns {Object} 정제된 객체
 *
 * @private
 *
 */
function _removePropertys(v = {}, propertys = null){

    if (!_.isPlainObject(v)){
        CONSOLE.prop.throw('BXUI.Util.removeProperty', 'v', 'type');
    }

    if (!_.isString(propertys) && !_.isArray(propertys)){
        CONSOLE.prop.throw('BXUI.Util.removeProperty', 'propertys', 'type');
    }

    propertys = _.isArray(propertys) ? propertys : [propertys];

    v = _.cloneDeep(v);

    _.map(v, (vv, propertyName) => {

        _.forEach(propertys, _propertyName => {

            if (propertyName === _propertyName){
                delete v[propertyName];
            }
        });
    });

    return v;
}

/**
 * 오브젝트 내부 프로퍼티를 삭제시킨다.

 * @returns {Object} 정제된 객체
 *
 * @private
 *
 */
function _removeItems(v = [], items = null){

    if (!_.isArray(v)){
        CONSOLE.prop.throw('BXUI.Util.removeItems', 'v', 'type');
    }

    if (!_.isString(items) && !_.isArray(items)){
        CONSOLE.prop.throw('BXUI.Util.removeItems', 'items', 'type');
    }

    v = _.cloneDeep(v);

    items = _.isArray(items) ? items : [items];

    _.forEach(items, item => {

        let idx = v.indexOf(item);

        if (idx > -1) v.splice(idx, 1);
    });

    return v;
}

function _uniqueId(prefix = ''){

    if (!_.isString(prefix)){
        CONSOLE.prop.throw('BXUI.Util.uniqueId', 'prefix', 'type');
    }

    let ret = _.uniqueId();

    if (prefix){
        ret = `${prefix}_${ret}`;
    }

    return ret;
}

module.exports = UTIL;
