// bxuip-angular.js-ui/src/lib/bx-ui/app/components/fileUpload/_base.js

/**
 * @class Base
 * @classDesc bxFileUpload Base 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 */
const FILEUPLOAD_TYPE = {
    "dnd": "dnd",
    "list": "list"
};

const EXTENSIONS_GROUP = {
    // 압축 파일
    "7z": "zip",
    "zip": "zip",
    "alz": "zip",
    "rar": "zip",
    "tar": "zip",
    // 이미지 파일
    "png": "img",
    "bmp": "img",
    "jpg": "img",
    "jpeg": "img",
    "gif": "img",
    "psd": "img",
    "ai": "img",
    // 텍스트 파일
    "doc": "doc",
    "docx": "doc",
    "txt": "txt",
    "md": "md",
    "hwp": "hwp",
    "ppt": "ppt",
    "pptx": "ppt",
    // pdf 파일
    "pdf": "pdf",
    // 엑셀 파일
    "xls": "xls",
    "xlsm": "xls",
    "xlsx": "xls",
    // 영상 파일
    "avi": "media",
    "wma": "media",
    "mpg": "media",
    "mpeg": "media",
    "asf": "media",
    "mp4": "media",
    // 음악 파일
    "mp3": "music",
    "wav": "music",
    "mid": "music",
    // 웹 소스 파일
    "java": "java",
    "js": "js",
    "html": "html",
    "css": "css"
};

const CLASS_NAME = {
    "bxFileUploadWrap": "bx-fileupload-wrap",
    "bxInputWrap": "bx-input-wrap",
    "fileType": "file-type",
    "fileList": "file-list",
    "fileName": "file-name",
    "fileProgress": "file-progress",
    "progressBar": "progress-bar",
    "progressInfo": "progress-info",
    "errorMessage": "error-message"
};

const OPTS = {
    // component 제약 조건
    "accept": '*/*',                                // readonly, ie9 x 네이티브 옵션으로 제공
    "allowFileType": '',                               // readonly, validator 체크용

    "maxTotalSize": 0,                               // readonly, ie9 x, 사용하려면 activeX 를 써야한다
    "maxFileSize": 0,                                // readonly, ie9 x, 사용하려면 activeX 를 써야한다
    "maxFileCount": 0,                               // readonly

    // component 디자인
    "type": 'list',                                     // readonly         DnD, list
    "width": null,                                       // get, set
    "height": null,                                      // get, set
    "multiple": false,
    "enableSubmitButton": true,
    // "enableDragAndDrop": false,                         // 이번에 안들어감

    // server info
    "url": null,                                          // readonly
    "timeout": 15000,

    "bx-render": function(){},                          // get, set, 컴포넌트 렌더링 후
    "bx-selected": function(){},                        // get, set, 파일 선택시
    "bx-error": function(){},                           // get, set, 에러 발생시 호출
    "bx-prepare": function(){},                         // get, set, 파일 업로드 전
    "bx-loadstart": function(){},                       // get, set, ie9는 제공 안함
    "bx-progress": function(){},                        // get, set, ie9는 제공 안함
    "bx-complete": function(){},                        // get, set, 개별 파일 업로드 후
    "bx-done": function(){}                             // get, set, 모든 파일 업로드 후
};

const DIM_STYLE = {
    enableIcon: true,
    iconType: 'default',
    color: 'black'
};

const COMPONENT_DATA = {};

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const UTIL = BXUI.Util;
const FILE = BXUI.File;

class Base {
    
    constructor(){

        Object.assign(this, {
            msg: FILE.MESSAGES,
            dimStyle: DIM_STYLE,
            opts: OPTS,
            className: CLASS_NAME,
            extensionsGroup: EXTENSIONS_GROUP,
            type: FILEUPLOAD_TYPE
        });
    }

    /**
     * file list를 추가한다
     *
     * @param {object|array} files - 추가할 file object
     * @param {object} $contentsContainer - component container
     *
     * @private
     * @memberOf BxFileUpload
     */
    addFileList(files = null, $contentsContainer = {}){
        _.forEach(files, file => $contentsContainer.append(_createContentsList.call(this, file)));
    }

    /**
     * 에러 container를 보여준다
     *
     * @param {object} $componentContainer - component container
     * @param {object} errorObj - error 오브젝트
     *
     * @private
     * @memberOf BxFileUpload
     */
    animateErrorContainer($componentContainer = {}, errorObj = {}){

        const duration = 1000;
        const delay = 3000;
        let $errorContainer = $componentContainer.find(`.${CLASS_NAME.fileList} > .error`);
        let changeText = $componentContainer.prop('bxerror')(errorObj);

        if (!_.isString(changeText) || changeText.length === 0) changeText = errorObj.msg;

        $errorContainer.children(`.${CLASS_NAME.errorMessage}`).text(changeText);

        if ($errorContainer.css('display') === 'block') return;

        // slide down & up
        $errorContainer.show()
            .velocity("slideDown", {duration: duration})
            .velocity("slideUp", {
                delay: delay,
                duration: duration
            }).hide();
    }

    /**
     * 문자로 된 데이터 크기를 Byte값으로 계산한다
     *
     * @param {String} value - 데이터 사이즈가 표시가 된 문자열
     *
     * @returns {Number} result - 계산된 Byte 크기
     *
     * @private
     * @memberOf BxFileUpload
     */
    convertStringToByte(value = ''){

        let number = _.parseInt(value);
        let dataType = _.isNumber(value) ? -1 : value.replace(/\d/gi,'');
        let unit = ['B', 'KB', 'MB', 'GB', 'TB'];
        let kb = 1024;
        let result = value;

        if (!_.isNaN(number)){

            let pow = _.indexOf(unit, dataType);

            pow = pow === -1 ? _.indexOf(unit, 'MB') : pow;

            result = Math.pow(kb, pow) * number;
        }

        return result;
    }

    /**
     * input을 초기화 한다
     * 
     *
     * @param {object} $target - input element
     *
     * @private
     * @memberOf BxFileUpload
     */
    resetInputFile($target = {}){

        if (UTIL.isIE()) {
            // 전달 받은 원본 event target은 하단 함수를 통해 reset이 된다
            $target.replaceWith($target.clone(true));
        }
        else {
            $target.val('');
        }
    }

    /**
     * 비어있는 값인지 체크한다
     *
     * @param {string} value - 체크 할 값 
     *
     * @private
     * @memberOf BxFileUpload
     */
    isNull(value){
        return _.isUndefined(value) || _.isNull(value) ||  !_.isString(value) || value.length === 0 ? true : false;
    }
}

/**
 * file list item을 반환한다
 *
 * @param {object} fileInfo={} - 추가할 file info
 *
 * @return 생성한 item container
 *
 * @private
 * @memberOf BxFileUpload
 */
function _createContentsList(fileInfo = {}){
    
    let itemHtml = [];

    let fileName = fileInfo.name;
    let fileIcon = _getFileIcon(_.last(fileName.split('.')));
    let fileSize = UTIL.getFormatString('byte', fileInfo.size.toString());

    itemHtml.push('<li>');
    itemHtml.push(`<span class="file-icon ${fileIcon}"></span>`);
    itemHtml.push(`<span class="file-name">${fileName}</span>`);
    itemHtml.push(`<span class="file-size">${fileSize}</span>`);
    itemHtml.push(`<button type="button"><i class="file-icon del i-15"></i></button>`);
    itemHtml.push('</li>');

    return $(itemHtml.join(''));
}

/**
 * file icon을 가져온다
 *
 * @param {string} extensions='' - 가져올 icon 확장자
 *
 * @return icon class
 *
 * @private
 * @memberOf BxFileUpload
 */
function _getFileIcon(extensions = ''){
    return _.isUndefined(EXTENSIONS_GROUP[extensions]) ? 'file' : EXTENSIONS_GROUP[extensions];
}

module.exports = {
    Base,
    COMPONENT_DATA
};
