// bxuip-angular.js-ui/src/lib/bx-ui/app/components/fileUpload/_ie9.js

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const FILE = BXUI.File;

const {Base: BASE_CLASS, COMPONENT_DATA} = require('./_base');

class IE9 extends BASE_CLASS {

    constructor(){
        super();
    }

    /**
     * form 내부 input의 유효성 검사를 체크한다
     *
     * @param {object} $componentContainer - component container
     *
     * @private
     * 
     * @memberOf BxFileUpload
     */
    validateForm($componentContainer = {}){

        let targetId = $componentContainer.attr('id');
        let iframeId = COMPONENT_DATA[targetId].iframeId;
        let $componentAllButton = $componentContainer.find('button');

        // 모든 버튼 disabled 설정
        $componentAllButton.prop('disabled', true);

        // submit 전 validate 체크
        let validForm = _validateSubmitBefore.call(this, $componentContainer);
        
        if (validForm){

            // dim 그리기
            _showDim.call(this, $componentContainer, targetId);

            // form의 target을 내부 iframe Id로 설정
            $componentContainer.parents('form').attr('target', iframeId);
            
            // 보낼때 onload 이벤트를 넣은 이유는 삽입 후 화면 렌더링시에도 onload가 실행되기 때문에 보낼때 onload 이벤트를 추가한다
            document.getElementById(iframeId).onload = _onLoadedIframe.bind(this);

            // bx-prepare 호출
            $componentContainer.prop('bxprepare')();

            // form submit
            $(`#${targetId}`).parent().submit();
        }
        else {
            // 모든 버튼 disabled 해제
            $componentAllButton.prop('disabled', false);
        }
    }

    /**
     * input 파일 선택 후 이벤트 처리 함수
     *
     * @param {object} e - 클릭 이벤트 객체
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    onChangeFile(e){
    
        let $target = $(e.target);
        let $componentContainer = $target.parents(`.${this.className.bxFileUploadWrap}`);
        let $contentsContainer = $componentContainer.find('ul');

        let targetId = $componentContainer.attr('id');
        let files = FILE.getFilesInfo($target);

        let data = COMPONENT_DATA[targetId];
        let maxFileCount = data.opts.maxFileCount;
        let currentFileCount = $contentsContainer.children('li').length;

        let resultObj = null;

        // 이벤트 전파 방지
        e.returnValue = false;
        e.cancelBubble = true;

        // selected callback 호출
        $componentContainer.prop('bxselected')(files);

        // 처음 파일 선택시에 이미 파일 최대수가 넘어갔다면 아무것도 하지 않는다
        if (maxFileCount !== 0 && currentFileCount >= maxFileCount){
            this.animateErrorContainer($componentContainer, {msg: this.msg.overFileCount});
            return false;
        }
        
        // validation 체크
        resultObj = _validateFiles.call(this, $componentContainer, files[0]);

        // 에러 컨테이너 출력
        if (resultObj.errorObj.length){
            this.animateErrorContainer($componentContainer, resultObj.errorObj[0]);
        }
        
        // 파일 리스트 추가
        this.addFileList.call(this, files, $contentsContainer);
    
        return true;    
    }

    /**
     * 파일을 리스트에서 삭제한다
     *
     * @param {object} $componentContainer - component container
     * @param {number} index - 삭제할 파일 index
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    removeFileList($componentContainer, index){
        $componentContainer.children('.file-type').eq(index).remove();
    }

    getUploadList(targetId = '', $componentContainer = {}){

        let filesList = $componentContainer.find('input[type="file"]:not(:last)');
        let resData = [];

        _.forEach(filesList, item => {

            let $item = $(item);

            resData.push(FILE.getFilesInfo($item)[0]);
        });

        return resData;
    }
}

/**
 * submit 전에 validation 체크 함수
 *
 * @param {object} $componentContainer - component Container
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _validateSubmitBefore($componentContainer = {}){

    let files = $componentContainer.find('input[type="file"]:not(:last)');
    let resultObj = [];
    let result = true;

    if (files.length !== 0){
        _.forEach(files, file => {

            let fileInfo = FILE.getFilesInfo($(file));

            resultObj.push(_validateFiles.call(this, $componentContainer, fileInfo[0]));
        });
        
        // submit 전에 validation 체크시 에러가 있는 경우 submit 중지
        _.forEach(resultObj, item => {

            if (item.errorObj.length !== 0){

                this.animateErrorContainer($componentContainer, item.errorObj[0]);
                
                result = false;

                return false;
            }
        });
    }
    else {
        result = false;
    }
    
    return result;
}

/**
 * file validation 체크
 *
 * @param {object} $componentContainer - component container
 * @param {object} fileInfo - file
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _validateFiles($componentContainer = {}, fileInfo = null){

    let targetId = $componentContainer.attr('id');
    let filesCount = $componentContainer.find('ul > li').length;
    
    let componentData = COMPONENT_DATA[targetId];
    let opts = componentData.opts;

    let allowedList = opts.allowFileType.replace(/\s/g, '').split(',');
    let maxFileCount = parseInt(opts.maxFileCount);
    
    let extensions = fileInfo.ext;
    let fileName = fileInfo.name;
    
    let errorObj = [];

    // 확장자 체크
    if (opts.allowFileType.length && _.intersection(allowedList, [extensions]).length === 0){
        
        errorObj.push({
            type: 'notAllowExtensions',
            fileName: fileName,
            file: fileInfo,
            msg: this.msg.notAllowExtensions
        });
    }
    
    // 파일 최대 수 체크
    if (maxFileCount !== 0 && filesCount > maxFileCount){

        errorObj.push({
            type: 'overFileCount',
            msg: this.msg.overFileCount
        });
    }

    return {
        errorObj: errorObj
    };
}

/**
 * input 파일 선택 후 이벤트 처리 함수
 *
 * @param {object} $componentContainer - component container
 * @param {string} targetId - component ID
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _showDim($componentContainer = {}, targetId = ''){

    let $dim = $componentContainer.bxDim(this.dimStyle).show();

    // ie9에서 사용할 $dim 데이터 저장
    COMPONENT_DATA[targetId].$dim = $dim;
}

/**
 * iframe onload 함수
 *
 * @param {object} e - onload callback 객
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _onLoadedIframe(e){

    let $componentContainer = $(e.target).parents(`.${this.className.bxFileUploadWrap}`);
    let $fileList = $componentContainer.find('ul > li');
    let targetId = $componentContainer.attr('id');

    // 아이템 success class 추가
    _.map($fileList, $item => $item.className = 'success');

    // dim 해제
    COMPONENT_DATA[targetId].$dim.hide();

    // 버튼 disabled 해제
    $componentContainer.find('button').prop('disabled', false);

    $componentContainer.prop('bxcomplete')();

    $componentContainer.prop('bxdone')();
}

module.exports = IE9;
