// bxuip-angular.js-ui/src/lib/bx-ui/app/components/fileUpload/_modern.js

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const {Base: BASE_CLASS, COMPONENT_DATA} = require('./_base');
const UTIL = BXUI.Util;
const FILE = BXUI.File;
const XHR = BXUI.Xhr;

class ETC extends BASE_CLASS {

    constructor(){
        super();
    }

    /**
     * 남은 용량 표시 컨테이너 생성
     *
     * @param {object} $opts - component options
     *
     * @return {object} - 생성된 html
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    createFileSizeContainer(opts = {}){

        let html = [];
        let maxTotalSize = opts.maxTotalSize;

        maxTotalSize = UTIL.getFormatString('byte', _.toString(maxTotalSize), {step:2});

        html.push('<div class="file-progress">');
        html.push('<div class="progress-bar" style="width:0"></div>');
        html.push(`<div class="progress-info"><span class="user-capa">0</span>/<span>${maxTotalSize}</span></div>`);
        html.push('</div>');

        return $(html.join(''));
    }

    /**
     * 남은 용량을 변경하는 외부 함수
     *
     * @param {object} $componentContainer - component container
     * @param {string} targetId - component ID
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    changeFileSize($componentContainer = {}, targetId = ''){
        _changeFileSize.call(this, $componentContainer, targetId);
    }

    /**
     * form 내부 input의 유효성 검사를 체크한다
     *
     * @param {object} $componentContainer - component container
     *
     * @private
     * 
     * @memberOf BxFileUpload
     */
    validateForm($componentContainer = {}){

        let $fileList = $componentContainer.find('ul > li');
        let targetId = $componentContainer.attr('id');
        let $componentAllButton = $componentContainer.find('button');

        // submit 전 validate 체크
        let validForm = _validateSubmitBefore.call(this, $componentContainer, targetId);
        
        if (validForm){
            
            // 모든 버튼 disabled 설정
            $componentAllButton.prop('disabled', true);

            _sendForm.call(this, $componentContainer, targetId);
        }
    }

    /**
     * input 파일 선택 후 이벤트 처리 함수
     *
     * @param {object} e - 클릭 이벤트 객체
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    onChangeFile(e = {}){
    
        let $target = $(e.target);
        let $componentContainer = $target.parents(`.${this.className.bxFileUploadWrap}`);
        let $inputContainer = $target.parents(`.${this.className.fileType}`);
        let $contentsContainer = $componentContainer.find('ul');

        let targetId = $target.parents(`.${this.className.bxFileUploadWrap}`).attr('id');
        let files = FILE.getFilesInfo($target);

        let data = COMPONENT_DATA[targetId];
        let maxFileCount = data.opts.maxFileCount;
        let maxTotalSize = UTIL.convertStringToByte(data.opts.maxTotalSize.toString());
        let currentFileCount = data.fileStack.length;
        let currentFileSize = 0;

        _.forEach(data.fileStack, item => currentFileSize += item.size);

        let resultObj = null;

        // selected callback 호출
        $componentContainer.prop('bxselected')(files);

        // 처음 파일 선택시에 이미 파일 최대수가 넘어가면 오류를 출력한다
        if (maxFileCount !== 0 && currentFileCount >= maxFileCount){
            this.animateErrorContainer($componentContainer, {
                type: 'overFileCount',
                msg: this.msg.overFileCount
            });
            return false;
        }

        // 처음 파일 선택시에 이미 총 파일 허용 용량이 넘어가면 오류를 출력한다
        if (maxTotalSize !== 0 && currentFileSize > maxTotalSize){
            this.animateErrorContainer($componentContainer, {
                type: 'overTotalSize',
                msg: this.msg.overTotalSize
            });

            return false;
        }

        // validation 체크
        resultObj = _validateFiles.call(this, $componentContainer, files);
        
        // fileStack에 validate된 파일 붙이기
        COMPONENT_DATA[targetId].fileStack = _.concat(COMPONENT_DATA[targetId].fileStack, resultObj.validFiles);
        
        // 파일 리스트 추가
        this.addFileList.call(this, resultObj.validFiles, $contentsContainer);

        // 파일 사이즈 변경
        _changeFileSize.call(this, $componentContainer, targetId);

        // input 초기화
        this.resetInputFile($target);

        // 에러 컨테이너 출력
        if (resultObj.errorObj.length){
            this.animateErrorContainer($componentContainer, resultObj.errorObj[0]);
        }
    }

    /**
     * 파일을 리스트에서 삭제한다
     *
     * @param {object} $componentContainer - component container
     * @param {number} index - 삭제할 파일 index
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    removeFileList($componentContainer, index){

        let targetId = $componentContainer.attr('id');

        COMPONENT_DATA[targetId].fileStack.splice(index, 1);
    }

    /**
     * 컴포넌트에 등록된 파일 리스트를 반환한다
     *
     * @param {string} targetId - component id
     *
     * @private
     *
     * @memberOf BxFileUpload
     */
    getUploadList(targetId = ''){
        return COMPONENT_DATA[targetId].fileStack;
    }
}

/**
 * 남은 용량 변경 함수
 *
 * @param {object} $componentContainer - component container
 * @param {string} targetId - component ID
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _changeFileSize($componentContainer = {}, targetId = ''){
    
    let $fileContainer = $componentContainer.children(`.${this.className.fileProgress}`);

    if ($fileContainer.length === 0) return;

    let $fileProgressbar = $fileContainer.children(`.${this.className.progressBar}`);
    let $fileSizeContainer = $fileContainer.children(`.${this.className.progressInfo}`);

    let totalFileSize = 0;
    let maxPercent = 100;
    let maxTotalSize = $fileSizeContainer.children(':last').text();
    let fileList = COMPONENT_DATA[targetId].fileStack;

    maxTotalSize = UTIL.convertStringToByte(maxTotalSize);

    // 적재된 파일 총 용량 계
    _.forEach(fileList, file => totalFileSize += file.size);

    // 최대 용량 문자를 Byte로 변환
    let convertFileMaxSize = UTIL.getFormatString('byte', totalFileSize.toString());
    // 남은 용량 퍼센트로 계산
    let remainFilePercent = parseInt(totalFileSize / maxTotalSize * maxPercent);

    // 남은 용량의 텍스트 출력
    $fileSizeContainer.children(':first').text(convertFileMaxSize);

    // 100%가 넘어가면 100%로 고정 및 에러 표시
    if (remainFilePercent >= maxPercent){
        $fileSizeContainer.addClass('exceed').children(':first').addClass('error');
        remainFilePercent = maxPercent;
    }
    else {
        $fileSizeContainer.removeClass('exceed').children(':first').removeClass('error');
    }

    $fileProgressbar.css('width', remainFilePercent + '%');
}

/**
 * submit 전에 validation 체크 함수
 *
 * @param {object} $componentContainer - component Container
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _validateSubmitBefore($componentContainer = {}, targetId = ''){

    let resultObj = [];
    let result = true;
    let filesInfo = COMPONENT_DATA[targetId].fileStack;

    if (filesInfo.length !== 0){
        resultObj = _validateFiles.call(this, $componentContainer);
    
        // submit 전에 validation 체크시 에러가 있는 경우 submit 중지
        if (resultObj.errorObj.length !== 0){
            this.animateErrorContainer($componentContainer, resultObj.errorObj[0]);
            result = false;
        }
    }
    else {
        result = false;
    }

    return result;
}

/**
 * file validation 체크
 *
 * @param {object} $componentContainer - component container
 * @param {object} fileInfo - file
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _validateFiles($componentContainer = {}, filesInfo = []){
    
    let targetId = $componentContainer.attr('id');
    let $fileSizeContainer = $componentContainer.children(`.${this.className.fileProgress}`);
    
    let opts = COMPONENT_DATA[targetId].opts;
    let fileStack = COMPONENT_DATA[targetId].fileStack;

    let totalFileCount = fileStack.length + filesInfo.length;
    let maxTotalSize = UTIL.convertStringToByte(opts.maxTotalSize.toString());
    let maxFileSize = UTIL.convertStringToByte(opts.maxFileSize.toString());
    let maxFileCount = parseInt(opts.maxFileCount);

    let allowedList = opts.allowFileType.replace(/\s/g, '').split(',');

    let res = [];
    let errorObj = [];
    let overFileSizeCount = 0;
    let totalValidFileSize = 0;
    
    filesInfo = filesInfo.length === 0 ? fileStack : filesInfo;

    _.forEach(filesInfo, file => {
        
        let extensions = file.ext;
        let fileName = file.name;
        
        // 확장자 체크
        if (opts.allowFileType.length && _.intersection(allowedList, [extensions]).length === 0){
            
            errorObj.push({
                type: 'notAllowExtensions',
                fileName: fileName,
                file: file,
                msg: this.msg.notAllowExtensions
            });

            return;
        }
        
        // 파일 크기 체크
        if (maxFileSize !== 0 && file.size >= maxFileSize){

            overFileSizeCount++;
            errorObj.push({
                type: 'overFileSize',
                fileName: fileName,
                file: file
            });

            return;
        }

        // 전부 통과하면 임시 스택에 입력
        res.push(file);

        // valid 통과한 파일의 size 합 계산
        totalValidFileSize += file.size;

    });

    // 파일 최대 개수 체크
    if (maxFileCount !== 0 && totalFileCount > maxFileCount){
        errorObj.push({
            type: 'overFileCount',
            msg: this.msg.overFileCount
        });
    }

    // 개별 파일 크기가 틀린 경우 개수를 체크하여 메세지를 출력한다
    if (overFileSizeCount !== 0){
        _.map(errorObj, item => {
            if (item.type === 'overFileSize'){
                item.msg = `${overFileSizeCount}${this.msg.overFileSize}`;
            }
        });
    }

    // 최대 파일 허용 용량 체크
    if (maxTotalSize !== 0 && totalValidFileSize > maxTotalSize){
        errorObj.push({
            type: 'overTotalSize',
            msg: this.msg.overTotalSize
        });
    }

    return {
        validFiles: res,
        errorObj: errorObj
    };
}

/**
 * 파일 업로드 함수
 *
 * @param {object} $componentContainer - component container
 * @param {string} targetId - component ID
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _sendForm($componentContainer = {}, targetId = ''){

    let componentData = COMPONENT_DATA[targetId];
    let opts = componentData.opts;
    let files = componentData.fileStack;
    let url = opts.url;
    let timeout = opts.timeout;
    let enableSubmitButton = opts.enableSubmitButton;
    let completeCount = 0;
    
    _.forEach(files, (file, i) => {

        let formData = new FormData();
        let $itemContainer = $componentContainer.find('ul > li').eq(i);
        $itemContainer.addClass('add-dim');

        formData.append(file.name, file.originFile);

        XHR.post({
            url: url,
            transferType: 'upload',
            processData: false,
            contentType: false,
            timeout: timeout,
            data: formData,
            beforeSend: xhr => {
                $componentContainer.prop('bxprepare').call(xhr, file);
            },
            loadStart: (xhr) => {
                $componentContainer.prop('bxloadstart').call(xhr);
            },
            progress: (xhr, loaded, total, percent) => {
                $componentContainer.prop('bxprogress').apply(xhr, [file, loaded, total, percent]);
            },
            error: xhr => {
                $itemContainer.addClass('failed').removeClass('add-dim');
                $componentContainer.prop('bxerror').call(xhr, {type: 'transactionError', file: file});
                $componentContainer.find('button').prop('disabled', false);
            },
            success: data => {
                console.log('success', data);
                $itemContainer.addClass('success').removeClass('add-dim');
            }
        }).done((data) => {

            $componentContainer.prop('bxcomplete').call(file, data);

            if (++completeCount === files.length){
                $componentContainer.prop('bxdone').call(files);
                $componentContainer.find('button').prop('disabled', false);

                if (!enableSubmitButton) {$componentContainer.find('.submit').parent().prop('disabled', true);}
            }
        });
    });
}

module.exports = ETC;
