// bxuip-angular.js-ui/src/lib/bx-ui/app/components/fileUpload/fileUpload.js
/**
 * Created by hyunjang on 2017. 5. 30..
 */
/**
 * @class BxFileUpload
 * @classDesc BxFileUpload 클래스 생성자
 * @desc Mime Type List : [MIME 타입의 전체 목록]{@link https://developer.mozilla.org/ko/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Complete_list_of_MIME_types}
 * @example
 *
 *  <bx-fileUpload
 *      id="fileUpload1"
 *      url="/api/upload"
 *      type="list"
 *      maxFileCount="5"
 *      maxFileSize="20"
 *      maxTotalSize="10"
 *      accept="image/png"
 *      allowFileType="jpg,png,txt"
 *      multiple="false"
 *      width="500"
 *      height="500"
 *      timeout="3000"
 *      bx-render="function(){console.log('render');}"
 *      bx-selected="function(){console.log('selected')}"
 *      bx-error="function(){console.log('error');}"
 *      bx-prepare="function(){console.log('prepare');}"
 *      bx-loadStart="function(){console.log('loadStart')}"
 *      bx-progress="function(){console.log('progress');}"
 *      bx-complete="function(){console.log('complete');}"
 *      bx-done="function(){console.log('done');}">
 *  </bx-fileUpload>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxFileUpload().initialize();
 *
 *      var fileUpload = document.getElementById('fileUpload1');
 *
 *      fileUpload.bxselected = function(files){
 *          console.log('files', files);
 *      }
 *
 *      fileUpload.bxerror = function(errorObj){
 *          console.log('errorObj', errorObj);
 *      }
 *
 *      fileUpload.bxprepare = function(file){
 *          console.log('xhr', this);
 *          console.log('file', file);
 *
 *          this.setRequestHeader('Cache-Control', 'no-cache');
 *      }
 *
 *      fileUpload.bxloadstart = function(xhrEvent){
 *          console.log('this', this);
 *          console.log('xhrEvent', xhrEvent);
 *      }
 *
 *      fileUpload.bxprogress = function(file, loaded, total, percent){
 *          console.log('xhr', this);
 *          console.log('file', file);
 *          console.log('loaded', loaded);
 *          console.log('total', total);
 *          console.log('percent', percent);
 *      }
 *
 *      fileUpload.bxcomplete = function(file){
 *          console.log('file', file);
 *      }
 *
 *      fileUpload.bxdone = function(files){
 *          console.log('files', files);
 *      }
 *
 *      fileUpload.width = 300;
 *      fileUpload.height = 600;
 *      console.log('width', fileUpload.width);
 *      console.log('height', fileUpload.height);
 *
 *  });
 *
 * </script>
 *
 */

/**
 * @property {function} [bx-render = function(){}] - 컴포넌트 render 시, 호출될 callback 함수
 * @callback bxrender
 *
 * @memberOf BxFileUpload
 */

/**
 * @property {function} [bx-selected = function(){}] - 파일 선택 시, 호출될 callback 함수
 * @callback bxselected
 *
 * @param {object} files - 선택된 파일들
 * @memberOf BxFileUpload
 */

/**
 * @property {function} [bx-error = function(){}] - 컴포넌트 render 시, 호출될 callback 함수
 * @callback bxerror
 *
 * @param {object} errorObj - 에러 오브젝트
 * @return {string} text - 사용자 설정 에러 문구
 * @memberOf BxFileUpload
 */

/**
 * @property {function} [bx-prepare = function(){}] - upload 전에 호출될 callback 함수, <span class='important'>ie9는 parameter가 없음</span>
 * @callback bxprepare
 *
 * @this xhrObject
 * @param {file} file - upload 될 파일
 *
 * @memberOf BxFileUpload
 */

/**
 * @property {function} [bx-loadStart = function(){}] - upload 시작할 때 호출될 callback 함수, <span class='important'>ie9는 지원되지 않음</span>
 * @callback bxloadstart
 *
 * @this xhrObject
 * @param {object} xhrEvent - xhr event
 * @memberOf BxFileUpload
 */

 /**
 * @property {function} [bx-progress = function(){}] - upload 중에 호출될 callback 함수, <span class='important'>ie9는 지원되지 않음</span>
 * @callback bxprogress
 *
 * @this xhrObject
 * @param {object} file - 진행중인 file Object
 * @param {int} loaded - 업로드 된 용량
 * @param {int} total - 업로드 총 용량
 * @param {int} percent - 진행된 퍼센트
 *
 * @memberOf BxFileUpload
 */

 /**
 * @property {function} [bx-complete = function(){}] - 개별 upload 완료 시, 호출될 callback 함수
 * @callback bxcomplete
 *
 * @this fileObject
 * @param {object} data - 서버 Response Data
 *
 * @memberOf BxFileUpload
 */

/**
 * @property {function} [bx-done = function(){}] - 전체 upload 완료 시, 호출될 callback 함수
 * @callback bxdone
 *
 * @this fileObject
 *
 * @memberOf BxFileUpload
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;

const IE9_CLASS = require('./_ie9');
const MODERN_CLASS = require('./_modern');
const {Base: BASE_CLASS, COMPONENT_DATA} = require('./_base');


const READONLY_PROPERTY = {
    /**
     * @property {object} bx-fileUpload
     * @property {string} [bx-fileUpload.type='list'] - <span class='important'>readonly</span> - fileUpload 타입, 현재는 list Type만 지원
     * @property {string} bx-fileUpload.url='' - <span class='important'>readonly</span> - 업로드 할 url
     * @property {int} [bx-fileUpload.maxFileCount=0] - <span class='important'>readonly</span> - 업로드 할 최대 파일 개수
     * @property {int} [bx-fileUpload.maxFileSize=0] - <span class='important'>readonly</span> - 개당 파일 최대 용량 설정, 기본은 MB,<br/> <span class='important'>ie9는 지원되지 않음</span>
     * @property {int} [bx-fileUpload.maxTotalSize=0] - <span class='important'>readonly</span> - 업로드 파일 총 용량 설정, 기본은 MB,<br/> <span class='important'>ie9는 지원되지 않음</span>
     * @property {string} [bx-fileUpload.accept=''] - <span class='important'>readonly</span> - 파일 선택 시 제한 설정,<br/> <span class='important'>ie9는 지원되지 않음</span>
     * @property {string} [bx-fileUpload.allowFileType=''] - <span class='important'>readonly</span> - 허용할 파일 확장자 설정
     * @property {boolean} [bx-fileUpload.multiple=false] - <span class='important'>readonly</span> - 파일 다중선택 유무 설정,<br/> <span class='important'>ie9, safari 6 이하는 지원되지 않음</span>
     * @property {boolean} [bx-fileUpload.enableSubmitButton=true] - <span class='important'>readonly</span> - submit 버튼 활성화 여부
     * @property {int} [bx-fileUpload.width=300] - 컴포넌트 width 설정
     * @property {int} [bx-fileUpload.height=200] - 컴포넌트 height 설정
     * @property {int} [bx-fileUpload.timeout=3000] - <span class='important'>readonly</span> - submit의 timeout 설정, 기본은 ms, <span class='important'>ie9는 지원되지 않음</span>
     * @property {function} [bx-fileUpload.bx-render=function(){}] - 컴포넌트 render 시, 호출될 callback 함수
     * @property {function} [bx-fileUpload.bx-selected=function(){}] - 파일 선택 시, 호출될 callback 함수
     * @property {function} [bx-fileUpload.bx-error=function(){}] - 컴포넌트에서 에러가 발생 시, 호출될 callback 함수, <span class='important'>ie9는 통신 callback에러가 지원되지 않음</span>
     * @property {function} [bx-fileUpload.bx-prepare=function(){}] - upload 전에 호출될 callback 함수, <span class='important'>ie9는 지원되지 않음</span>
     * @property {function} [bx-fileUpload.bx-loadStart=function(){}] - upload 시작할 때 호출될 callback 함수, <span class='important'>ie9는 지원되지 않음</span>
     * @property {function} [bx-fileUpload.bx-progress=function(){}] - upload 중에 호출될 callback 함수, <span class='important'>ie9는 지원되지 않음</span>
     * @property {function} [bx-fileUpload.bx-complete=function(){}] - 개별 upload 완료 시, 호출될 callback 함수
     * @property {function} [bx-fileUpload.bx-done=function(){}] - 전체 upload 완료 시, 호출될 callback 함수
     * @property {function} [bx-fileUpload.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
     *
     * @alias Attributes
     * @memberOf BxFileUpload
     */
    "accept": "accept",
    "multiple": "multiple",
    "allowFileType": "allowFileType",
    "maxTotalSize": "maxTotalSize",
    "maxFileSize": "maxFileSize",
    "maxFileCount": "maxFileCount",
    "type": "type",
    "enableSubmitButton": "enableSubmitButton",
    "url": "url",
    "timeout": "timeout"
};

const ATTRNAMES = {
    "accept": 'accept',
    "allowfiletype": "allowFileType",
    "maxtotalsize": "maxTotalSize",
    "maxfilesize": "maxFileSize",
    "maxfilecount": "maxFileCount",
    "enablesubmitbutton": "enableSubmitButton",
    "enabledraganddrop": "enableDragAndDrop"
};

const APIS = {
    "submit": _submit,
    "getUploadList": _getUploadList
};

const SELFS = {};

class BxFileUpload extends BASE_CLASS{

    constructor(){

        super();

        SELFS.self = this;
        SELFS.extendInstance = UTIL.isIE9() ? new IE9_CLASS() : new MODERN_CLASS();
    }

    /**
     * 초기화 함수
     *
     * @returns {object} 생성된 인스턴스
     *
     * @memberOf BxFileUpload
     */
    initialize(){

        this.$fileUpload = this.$root.find('bx-fileUpload');

        this.elems = [];

        _.forEach(this.$fileUpload, (elem, i) => {

            let $customElem = $(elem);

            // 사용자 커스텀 엘리먼트를 파싱한다.
            let $component = _parseCustomElement.call(this, $customElem);

            $component.children().prop('bxrender')();

            this.elems.push($component.get(0));

        });

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxFileUpload.isCustomElement = true;

/**
 * 파일을 업로드 한다
 *
 * @returns {element} 파일 엘리먼트
 *
 * @alias submit
 * @memberOf BxFileUpload
 */
function _submit(){

    let $this = $(this);

    _validateForm.call(SELFS.self, $this);

    return this;
}

/**
 * 선택된 파일 리스트를 가져온다
 *
 * @returns {object} 선택된 file object
 *
 * @alias getUploadList
 * @memberOf BxFileUpload
 */
function _getUploadList(){

    let $this = $(this);
    let targetId = $this.attr('id');

    return SELFS.extendInstance.getUploadList.call(SELFS.self, targetId, $this);
}

/**
 * 사용자 `custom` 태그를 `native` 태그로 대치시킨다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - 사용자 custom 엘리먼트
 *
 * @returns {object} 생성한 $container 엘리먼트
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _parseCustomElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxFileUpload, '$customElem', 'type');
    }

    // $container 엘리먼트를 생성한다.
    let $container = BASE.appendContainerElement(BxFileUpload.$root, BxFileUpload.className);

    // 컴포넌트 엘리먼트를 생성한다.
    let $component = _getComponentElement.call(this, $customElem);

    // 반환받은 `컴포넌트 엘리먼트`를 `컨테이너 엘리먼트`의 자식으로 추가한다.
    $container.append($component);

    // $customElem 엘리먼트를 생성된 `$container` 엘리먼트로 대치시킨다.
    $customElem.replaceWith($container);

    return $component;
}

/**
 * component 엘리먼트를 반환한다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - 사용자 custom 엘리먼트
 *
 * @returns {object} 생성한 component 엘리먼트
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _getComponentElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxFileUpload, '$customElem', 'type');
    }

    const fileType = $customElem.attr('type');
    const opts = _createComponentOptions.call(this, $customElem, this.opts);
    const uniqueId = opts.attr.id || 'bxForm' + _.uniqueId();

    // this에 form ID별로 데이터 적제할 공간 생성
    Object.assign(COMPONENT_DATA, {
        [uniqueId]: {
            opts: opts,
            fileStack: []
        }
    });

    // BxFileUpload 컴포넌트가 form 태그에 감싸여 있다면 에러를 호출한다
    if ($customElem.parents('form').length){
        CONSOLE.prop.throw(BxFileUpload, this.msg.noUseInnerForm);
    }

    let $formContainer = _createFormContainer.call(this, opts, uniqueId);
    let $componentContainer = $formContainer.children(`.${this.className.bxFileUploadWrap}`);

    const $inputContainer = _createInputContainer.call(this, opts);
    const $contentsContainer = _createContentsContainer.call(this, opts, fileType, uniqueId);

    $componentContainer.append($inputContainer);
    $componentContainer.append($contentsContainer);
    // ie9 거나 maxTotalSize 값이 없을 때는 progress Container를 그리지 않는다
    if (!UTIL.isIE9() && opts.maxTotalSize !== this.opts.maxTotalSize){

        const $fileSizeContainer = SELFS.extendInstance.createFileSizeContainer(opts);

        $componentContainer.append($fileSizeContainer);
    }

    // native event 바인드
    UTIL.addAllEventListenerWithProperty($componentContainer, opts.events);

    // 리스트 외의 영역 클릭시 선택 해제
    $componentContainer.on('click', e => {

        let target = e.target;

        if (target.tagName === 'UL' ||
            target.className !== this.className.fileName ||
            (target.tagName === 'INPUT' && target.className === this.className.fileName)){
            $(e.currentTarget).find('ul > li').removeClass('on');
        }
    });

    return $formContainer;
}

/**
 * form 컨테이너를 반환한다
 *
 * @param {object} [opts={}] - component options
 * @param {string} [uniqueId] - component unique ID
 *
 * @return 생성한 form container
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _createFormContainer(opts = {}, uniqueId = ''){

    const url = opts.url;

    let $formContainer = $(`<form method="post" action="${url}" enctype="multipart/form-data"></form>`);
    let $componentContainer = $(`<div id="${uniqueId}" class="${this.className.bxFileUploadWrap}"></div>`);

    if (!_.isNull(opts.width)){
        $componentContainer.css('width', opts.width);
    }

    if (!_.isNull(opts.height)){
        $componentContainer.css('height', opts.height);
    }

    $formContainer.append($componentContainer);

    _setElementProperties.call(this, opts, $componentContainer.get(0));

    return $formContainer;
}

/**
 * File List container를 반환한다
 *
 * @param {object} opts={} - component options
 * @param {object} fileType='' - component file type
 * @param {string} uniqueId='' - component unique Id
 *
 * @return 생성한 file list container
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _createContentsContainer(opts = {}, fileType = '', uniqueId = ''){

    let wrapHTML = [];
    let inputHeight = 40;

    let height = opts.height > 200 ? opts.height - inputHeight : opts.height;

    wrapHTML.push(`<div class="${this.className.fileList}">`);
    wrapHTML.push('<ul></ul>');
    wrapHTML.push('<div class="error">');
    wrapHTML.push(`<span class="error-message">ERROR</span>`);
    wrapHTML.push('</div>');
    wrapHTML.push('</div>');

    let $wrapHTML = $(wrapHTML.join(''));

    $wrapHTML.find('.error').hide();

    switch (fileType){
        case this.type.dnd:

            break;
        case this.type.list:

            if (UTIL.isIE9()){

                let iframe = document.createElement('iframe');
                let uniqueIFrameId = 'bxIframe' + _.uniqueId();

                iframe.style.display = 'none';
                iframe.style.visibility = 'hidden';

                iframe.id = uniqueIFrameId;
                iframe.name = uniqueIFrameId;

                Object.assign(COMPONENT_DATA[uniqueId], {
                    iframeId: uniqueIFrameId
                });

                $wrapHTML.append($(iframe));
            }

            break;
    }

    return $wrapHTML;
}

/**
 * 각 `custom` 엘리먼트에 정의된 attr 를 opt 객체로 반환한다.
 *
 * @param {object} [targetObject=null]
 * @param {object} [defaultOpts={}]
 *
 * @returns {object} 생성된 opt 객체
 *
 * @private
 * @memberOf BxFileUpload
 */
function _createComponentOptions(targetObject = null, defaultOpts = {}){

    if (_.isNull(targetObject)){
        CONSOLE.prop.throw(BxFileUpload, 'targetObject', 'type');
    }

    if (!_.isFunction(targetObject.get) && !_.isPlainObject(targetObject)){
        CONSOLE.prop.throw(BxFileUpload, 'targetObject', 'type');
    }

    if (!_.isPlainObject(defaultOpts)){
        CONSOLE.prop.throw(BxFileUpload, 'defaultOpts', 'type');
    }

    let opts = UTIL.createComponentOptions(targetObject, defaultOpts, ATTRNAMES, function(k, v){

        if (k === 'maxTotalSize' || k === 'maxFileSize' || k === 'maxFileCount'){

            v = parseInt(v, 10);

            if (isNaN(v)){
                CONSOLE.prop.throw(BxFileUpload, k, 'type');
            }
        }

        if (k === 'timeout' && !UTIL.isIE9()){

            v = parseInt(v, 10);

            if (isNaN(v)){
                CONSOLE.prop.throw(BxFileUpload, k, 'type');
            }
        }

        if (k === 'width' || k === 'height'){

            let componentMinWidth = 300;
            let componentMinHeight = 200;

            v = parseInt(v, 10);

            if (!_.isNull(v) && isNaN(v)){
                CONSOLE.prop.throw(BxFileUpload, k, 'type');
            }

            if (k === 'width' && v < componentMinWidth){
                v = componentMinWidth;
            }
            else if (k === 'height' && v < componentMinHeight){
                v = componentMinHeight;
            }
        }

        if (k === 'multiple' || k === 'enableDragAndDrop' || k === 'enableSubmitButton'){
            if (!(v === true || v === false || v === 'true' || v === 'false')){
                CONSOLE.prop.throw(BxFileUpload, k, 'type');
            }

            v = v === 'true' || v === true ? true : false;
        }

        if (k === 'accept' && !_.isString(v)){
            CONSOLE.prop.throw(BxFileUpload, k, 'type');
        }

        if (k === 'type' && v !== 'list'){
            CONSOLE.prop.throw(BxFileUpload, k, 'type');
        }

        return v;

    }, function(k, v){

        return v;
    });

    if (_.isNull(opts.url)){
        CONSOLE.prop.throw(BxFileUpload, 'url', 'all');
    }

    return opts;
}

/**
 * file 엘리먼트의 프로퍼티를 설정한다.
 *
 * @param {object} [opts={}] - 옵션 객체
 * @param {object} [elem=null] - 프로퍼티가 추가될 엘리먼트
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxFileUpload
 */
function _setElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxFileUpload, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxFileUpload, 'elem', 'type');
    }

    let $elem = $(elem);
    let bxPtn = /^bx/;
    let defaultWidth = this.opts.width || 300;
    let defaultHeight = this.opts.height || 200;

    _.map(opts, (v, k) => {

        // attr 속성은 무시한다.
        if (k === 'attr') return;

        // 읽기 전용 속성
        if ( READONLY_PROPERTY[k] ){
            UTIL.setReadOnlyProperty(elem, k, v);
        }
        else{

            if (k.match(bxPtn)){

                let _k = k.replace('-', '');
                delete opts[_k];

                k = _k;
            }

            let _k = UTIL.convertInternalPropertyName(k);

            elem[_k] = v;

            UTIL.setPropertyWithDescriptor(elem, k, {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: v => {

                    if (k === 'width' || k === 'height'){

                        if (_.isString(v)){
                            CONSOLE.prop.throw(BxFileUpload, k, 'type');
                        }

                        if (k === 'width' && v < defaultWidth){
                            v = defaultWidth;
                        }

                        if (k === 'height' && v < defaultHeight){
                            v = defaultHeight;
                        }

                        if (isNaN(parseInt(v, 10))){
                            CONSOLE.prop.throw(BxFileUpload, k, 'type');
                        }
                    }

                    elem[_k] = v;

                    if (k === 'width' || k === 'height'){
                        _setComponentSize(elem, k, v);
                    }
                }
            });
        }
    });

    // API 에 대한, 프로퍼티 할당
    _.map(APIS, (v, k) => {
        UTIL.setReadOnlyProperty(elem, k, v);
    });
}

/**
 * 컴포넌트 size를 설정한다
 *
 * @param {object} elem=null - 프로퍼티가 추가될 엘리먼트
 * @param {string} type='' - 변경될 사이즈 타입
 * @param {string} value=0 - 변경될 사이즈 크기
 *
 * @private
 * @memberOf BxFileUpload
 */
function _setComponentSize(elem = {}, type = '', value = 0){

    let $component = $(elem);

    $component.css(type, value);
}

/**
 * input 컨테이너를 반환한다
 *
 * @param {object} opts={} - component options
 *
 * @return 생성한 input container
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _createInputContainer(opts = {}){

    let $inputWrapper = _createInputElement.call(this, opts);

    let $buttonTrash = $(`<button type="button"><i class="file-icon i-25 trash" title="삭제하기"></i></button>`);
    let $buttonSubmit = $(`<button type="button"><i class="file-icon i-25 submit" title="전송하기"></i></button>`);

    $inputWrapper.children('.upload-hidden').on('change', _onChangeFile.bind(this));

    $buttonTrash.on('click', _onClickTrash.bind(this)).hover(_onHoverIcon);
    $buttonSubmit.on('click', _onClickSubmit.bind(this)).hover(_onHoverIcon);

    if (!opts.enableSubmitButton){
        $buttonSubmit.prop('disabled', true);
    }

    $inputWrapper.append($buttonTrash);
    $inputWrapper.append($buttonSubmit);

    return $inputWrapper;
}

/**
 * input file Type의 엘리먼트를 반환한다
 *
 * @param {object} [opts={}] - component options
 *
 * @return 생성한 input file type element
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _createInputElement(opts = {}){

    let uniqueId = 'bxFile' + _.uniqueId();
    // let attr = UTIL.objectToStringAttr(opts.attr);
    let accept = opts.accept;

    let $html = $(`<div class="${this.className.bxInputWrap} ${this.className.fileType}"></div>`);

    $html.append(`<input type="text" class="file-name" title="파일 선택하기" aria-hidden="true" disabled/>`);
    $html.append(`<label for="${uniqueId}" title="파일 추가하기">파일 추가하기</label>`);
    $html.append(`<input type="file" id="${uniqueId}" name="${uniqueId}" class="upload-hidden" accept="${accept}"/>`);

    let $file = $html.find('input[type="file"]');
    let $label = $html.find('label');

    if (!UTIL.isIE9() && !UTIL.isBelowSafariVersion(6) && opts.multiple){
        $file.prop('multiple', opts.multiple);
    }

    $file.on('focusin focusout', e => {

        if (e.type === 'focusin'){
            $label.addClass('focus');
        }
        else {
            $label.removeClass('focus');
        }
    });

    return $html;
}

/**
 * 컴포넌트 리스트에서 해당 항목을 삭제한다.
 *
 * @param {object} $componentContainer={} - component container
 * @param {object} $contentsContainer={} - file list container
 * @param {array} index=[] - 삭제 할 리스트 index
 *
 * @private
 * @memberOf BxFileUpload
 */
function _removeFileList($componentContainer = {}, $contentsContainer = {}, index = []){

    // 뒤에서부터 지우도록 정렬을 한다
    let removeIndex = _.reverse(_.sortBy(index));

    _.forEach(removeIndex, _index => {

        SELFS.extendInstance.removeFileList($componentContainer, _index);

        // 리스트에서 삭제
        $contentsContainer.children('li').eq(_index).remove();
    });

    // COMPONENT_DATA = this;
}

/**
 * 파일 선택시 이벤트를 처리한다
 *
 * @param {object} e={} - file 선택 시 이벤트 객체
 *
 * @private
 * @memberOf BxFileUpload
 */
function _onChangeFile(e = {}){

    let $target = $(e.target);
    let $componentContainer = $target.parents(`.${this.className.bxFileUploadWrap}`);

    let isValid = SELFS.extendInstance.onChangeFile.call(this, e);

    let $contentsList = $componentContainer.find('ul > li');

    // ie9용 파일 input 컨테이너 추가
    if (UTIL.isIE9() && isValid){

        let $inputContainer = $target.parents(`.${this.className.fileType}`);
        let targetId = $componentContainer.attr('id');

        _changeInputContainer.call(this, $inputContainer, targetId);
    }

    // // file list 클릭 이벤트 등록
    $contentsList.off('click').on('click', _onClickFileList.bind(this));

    // // x 버튼 클릭 이벤트 등록
    $contentsList.children('button').off('click').on('click', _onClickDeleteButton.bind(this));
}

function _changeInputContainer($inputContainer = {}, targetId = ''){

    let opts = COMPONENT_DATA[targetId].opts;
    let addInputContainer = _createInputContainer.call(this, opts);

    $inputContainer.after(addInputContainer).hide();
}

/**
 * 개별 파일 삭제 버튼 이벤트를 처리한다
 *
 * @param {object} fileInfo={} - 추가할 file info
 *
 * @private
 *
 * @memberOf BxFileUpload
 */
function _onClickDeleteButton(e = {}){

    let $target = $(e.target);
    let targetId = $target.parents(`.${this.className.bxFileUploadWrap}`).attr('id');
    let $componentContainer = $target.parents(`.${this.className.bxFileUploadWrap}`);
    let $contentsContainer = $componentContainer.find('ul');
    let index = $contentsContainer.children('li').index($target.parent());

    _removeFileList.call(this, $componentContainer, $contentsContainer, [index]);

    if (!UTIL.isIE9()){
        SELFS.extendInstance.changeFileSize.call(this, $componentContainer, targetId);
    }
    else {
        // ie9중 파일이 삭제되었을 때 제일 마지막 input을 초기화한다 (보이는 input)
        // 초기화 하지 않으면 마지막 선택한 파일을 재 선택시 파일이 선택되지 않는다
        let $inputFileElement = $componentContainer.find('input[type="file"]:last');

        this.resetInputFile($inputFileElement);
    }
}

/**
 * file list를 선택한다
 *
 * @param {object} e={} - 클릭 이벤트 객체
 *
 * @private
 * @memberOf BxFileUpload
 */
function _onClickFileList(e = {}){

    let $target = $(e.currentTarget);
    let $componentContainer = $target.parents(`.${this.className.bxFileUploadWrap}`);
    let $contentsContainer = $componentContainer.find('ul');

    if (!(e.metaKey || e.ctrlKey)) $contentsContainer.children('li').removeClass('on');

    $target.addClass('on');
}


/**
 * 삭제 버튼 이벤트를 처리한다
 *
 * @param {object} e={} - 클릭 이벤트 객체
 *
 * @private
 * @memberOf BxFileUpload
 */
function _onClickTrash(e = {}){

    let $componentContainer = $(e.currentTarget).parents(`.${this.className.bxFileUploadWrap}`);
    let $contentsContainer = $componentContainer.find('ul');
    let $selectedList = $contentsContainer.children('.on');

    let targetId = $componentContainer.attr('id');
    let selectedListIndex = [];

    // 아무것도 선택되지 않은 경우 전체 선택
    if ($selectedList.length === 0){
        $selectedList = $contentsContainer.children();
    }

    _.forEach($selectedList, item => selectedListIndex.push($contentsContainer.children('li').index(item)));

    _removeFileList.call(this, $componentContainer, $contentsContainer, selectedListIndex);

    if (!UTIL.isIE9()){
        SELFS.extendInstance.changeFileSize.call(this, $componentContainer, targetId);
    }
    else {

        let $inputFileElement = $componentContainer.find(`.${this.className.uploadHidden}`);

        this.resetInputFile($inputFileElement);
    }
}

/**
 * 아이콘 hover시 `on` Class를 추가시킨다
 * (firefox만 hover가 button에 적용 안되고 i에 적용되서 이벤트로 처리)
 *
 * @private
 * @memberOf BxFileUpload
 */
function _onHoverIcon(){
    $(this).children().toggleClass('on');
}

/**
 * submit 버튼 이벤트를 처리한다
 *
 * @param {object} e={} - 클릭 이벤트 객체
 *
 * @private
 * @memberOf BxFileUpload
 */
function _onClickSubmit(e = {}){

    let $componentContainer = $(e.currentTarget).parents(`.${this.className.bxFileUploadWrap}`);

    _validateForm.call(this, $componentContainer);
}

/**
 * submit 전에 유효성 검사를 진행한다
 *
 * @param {object} $componentContainer={} - component container
 *
 * @private
 * @memberOf BxFileUpload
 */
function _validateForm($componentContainer = {}){

    // 파일 업로드가 한번 진행되었던 경우에 success, fail class 제거
    $componentContainer.find(`.${this.className.fileList} li`).removeClass();

    // submit 전 validate 체크
    SELFS.extendInstance.validateForm.call(this, $componentContainer);

}

module.exports = BxFileUpload;
