// bxuip-angular.js-ui/src/lib/bx-ui/app/components/accordion.js
/**
 * 메뉴가 펼쳐질 경우, 호출될 callback 함수
 * @callback onActivate
 *
 * @param {event} e - 이벤트 객체
 * @param {object} ui - 생성된 엘리먼트 집합
 *
 * @memberOf BxAccordion
 */

/**
 * 메뉴 선택 시, 호출될 callback 함수
 * @callback onSelected
 *
 * @param {event} e - 이벤트 객체
 * @param {object} ui - 생성된 엘리먼트 집합
 *
 * @memberOf BxAccordion
 */

/**
 * 랜더링 완료 시, 호출될 callback 함수
 *
 * @callback onRender
 *
 * @param {event} e - 이벤트 객체
 * @param {object} ui - 생성된 엘리먼트 집합
 *
 * @memberOf BxAccordion
 */

/**
 * @class BxAccordion
 * @classDesc BxAccordion 클래스 생성자
 *
 * @parma {object} [OPTS]
 * @param {object} OPTS
 * @param {Object[]} OPTS.items
 * @param {string} OPTS.items.title - 헤더 타이틀 설정
 * @param {string} OPTS.items.icon - 헤더 아이콘 설정
 * @param {string[]} OPTS.items.items - 자식 아이템 집합
 * @param {int} [OPTS.active = ''] - 활성화될 아이템 인덱스
 * @param {string} [OPTS.event = 'click'] - 바인딩될 이벤트 타입('mouseover', 'click')
 * @param {int} [OPTS.animate = 200] - animation duration
 * @param {string} [OPTS.heightStyle = content] <s>아이템 컨테이너의 높이 스타일 설정 ('auto', 'fill', 'content')</s><span class='important'> ( Added in: v3.0.0, Deprecated since: v3.8.0 )</span>
 * @param {boolean} [OPTS.collapsible = false] 모든 section 에 대한, closed 유/무(단 active 프로퍼티에 값이 존재하지 않아야 한다)
 * @param {boolean} [OPTS.enableToggle = false] 토글 기능 사용 유/무
 * @param {object} [OPTS.stateOptions = {}] - storage 에 accrodion 의 열기/닫기 상태를 설정
 * @param {string} OPTS.stateOptions.key - storage 에 저장 할 고유의 state key <br/><b class="dec">(필수 값)</b>
 * @param {boolean} [OPTS.stateOptions.expire = true] - 브라우저 종료 시 state 초기화
 * @param {boolean} [OPTS.stateOptions.fold = false] - 열림 / 닫힘 상태 저장 유무
 * @param {function} [OPTS.onActivate = function(){}] -  메뉴가 펼쳐질 경우, 호출될 callback 함수
 * @param {function} [OPTS.onSelected = function(){}] -  메뉴 선택 시, 호출될 callback 함수
 * @param {function} [OPTS.onRender = function(){}] - 랜더링 완료 시, 호출될 callback 함수
 * @returns {object} 생성된 인스턴스
 *
 * @example
 // render 사용 예제
 let items = [
 {
     id: 'menu1',
     title: '메뉴1',
     items: [
         {
             id: 'menu1-1',
             title: '메뉴1-1',
             items: [
                 '<span>메뉴1-1-1</span>',
                 '<span>메뉴1-1-2</span>',
                 {
                     id: 'menu1-1-3',
                     title: '메뉴1-1-3',
                     items: [
                         {
                             title: '메뉴1-1-3-1',
                             items: [
                                 {
                                     id: 'menu1-1-3-1-1',
                                     title: '메뉴1-1-3-1-1',
                                     items: [
                                         {
                                             id: 'menu1-1-3-1-1-2',
                                             title: '메뉴1-1-3-1-1-2',
                                             icon: 'bw-icon i-30 i-mng-online',
                                             items: [
                                                 'test',
                                                 '<span>메뉴1-1-3-1-1-2-1</span>',
                                                 '<span>메뉴1-1-3-1-1-2-2</span>'
                                             ]
                                         },
                                         {
                                             title: '메뉴1-1-3-1-1-3',
                                             icon: 'bw-icon i-30 i-mng-online',
                                             items: [
                                                 '<span>메뉴1-1-3-1-1-2-1</span>',
                                                 '<span>메뉴1-1-3-1-1-2-2</span>'
                                             ]
                                         }
                                     ]
                                 }
                             ]
                         }
                     ]
                 },
                 {
                     id: 'menu1-1-4',
                     title: '메뉴1-1-4'
                 }
             ]
         },
         '<span>메뉴1-2</span>'
     ]
 },
 '<span>메뉴2</span>',
 {
     id: 'menu3',
     title: '메뉴3',
     icon: 'bw-icon i-30 i-mng-online',
     items: [
         {
             id: 'menu3-1',
             title: '메뉴3-1',
             items: [
                 '<span>메뉴3-1-1</span>',
                 '<span>메뉴3-1-2</span>',
                 '<span>메뉴3-1-3</span>'
             ]
         }
     ]
 }
 ];

 $('#accordion').BxAccordion({
    items: items,
    active : 0,
    animate : 200,
    event : "click",
    collapsible: false,
    onActivate: function(e, ui){
        console.log('active');
    },
    onRender: function(e, ui){
        console.log('render Complate');
    }
 }).render();
 //---------------------------------------------------------------

 // bind 사용 예제, angularJS
 <div id="accordion">
     <div class="bx-accordion-wrap">
         <ul class="bx-accordion">
            <li class="has-children" id="menu1">
                <a class="label" href="#" onclick="return false;">메뉴1</a>
                <ul>
                    <li class="has-children" id="menu1-1">
                        <a class="label" href="#" onclick="return false;">메뉴1-1</a>
                        <ul>
                            <li id="accordion-item-4">
                                <a href="#"><span>메뉴1-1-1</span></a>
                            </li>
                            <li id="accordion-item-5">
                                <a href="#"><span>메뉴1-1-2</span></a>
                            </li>
                            <li class="has-children" id="accordion-item-6">
                                <a href="#"><span>메뉴1-1-3</span></a>
                            </li>
                        </ul>
                    </li>
                </ul>
            </li>
            <li id="menu2">
                <a href="#"><span>메뉴2</span></a>
            </li>
            <li id="menu3">
                <a href="#"><span>메뉴3</span></a>
            </li>
        </ul>
     </div>
 </div>

 $('#accordion').BxAccordion({
    active : 1,
    animate : 200,
    event : "click",
    onActivate: function(e, ui){
        console.log('active');
    },
    onRender: function(e, ui){
        console.log('render Complate');
    }
 }).bind();
 //---------------------------------------------------------------

 // bind 사용 예제, react
 <div id="accordion">
     <div className="bx-accordion-wrap">
         <ul className="bx-accordion">
            <li className="has-children" id="menu1">
                <a className="label" href="#" onClick={function(){return false;}}>메뉴1</a>
                <ul>
                    <li className="has-children" id="menu1-1">
                        <a className="label" href="#" onClick={function(){return false;}}>메뉴1-1</a>
                        <ul>
                            <li id="accordion-item-4">
                                <a href="#"><span>메뉴1-1-1</span></a>
                            </li>
                            <li id="accordion-item-5">
                                <a href="#"><span>메뉴1-1-2</span></a>
                            </li>
                            <li className="has-children" id="accordion-item-6">
                                <a href="#"><span>메뉴1-1-3</span></a>
                            </li>
                        </ul>
                    </li>
                </ul>
            </li>
            <li id="menu2">
                <a href="#"><span>메뉴2</span></a>
            </li>
            <li id="menu3">
                <a href="#"><span>메뉴3</span></a>
            </li>
         </ul>
     </div>
 </div>

 $('#accordion').BxAccordion({
    active : 2,
    animate : 200,
    event : "click",
    onActivate: function(e, ui){
        console.log('active');
    },
    onRender: function(e, ui){
        console.log('render Complate');
    }
 }).bind();
 */
// * @param {string} [OPT.header ='h3'] - 아코디언 메뉴의 타이틀 부분의 태그를 지정하는 프로퍼티

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');

const CONSOLE = BXUI.Console;
const STORAGE = BXUI.Storage;
const Util = BXUI.Util;
const OPTS = Symbol();

const SELECTOR_NAME = {
    wrapper: 'bx-accordion-wrap',
    superParent: 'bx-accordion',
    parent: 'has-children',
    label: 'label',
    checked: 'checked',
    opened: 'opened',
    expander: 'accordion-expander',
    active: 'active-accordion',
    item: 'accordion-item'
};

const STYLE_VALUE = {
    paddingLeft: 20,
    space: '&nbsp;'
};

const STATE = {
    open : 'openState',
    select : 'selectState'
};

const MESSAGE = {
    noStringTypeAdd: 'string 타입의 item에는 추가할 수 없습니다.'
};

const DEFAULT_STATE_OPTIONS = {
    fold : false,
    expire : true,
    key : ''
};

const ROOT_ID = 'root';

class BxAccordion {

    constructor({
        items = [],
        active = null,
        // mouseover, click
        event = 'click',
        // 애니메이션 시간 설정.
        animate = 200,
        // @deprecated 
        heightStyle = 'content',
        enableToggle = false,
        collapsible = false,
        stateOptions,
        onActivate = function (){},
        onSelected = function (){},
        onRender = function (){},
    } = {}){

        if (!_.isArray(items)){
            CONSOLE.prop.throw(BxAccordion, 'items', 'type');
        }

        if (!_.isNull(active) && !_.isNumber(active)){
            CONSOLE.prop.throw(BxAccordion, 'active', 'type');
        }

        if (!_.isString(event)){
            CONSOLE.prop.throw(BxAccordion, 'event', 'type');
        }

        if (!_.isNumber(animate)){
            CONSOLE.prop.throw(BxAccordion, 'animate', 'type');
        }

        if (!_.isString(heightStyle)){
            CONSOLE.prop.throw(BxAccordion, 'heightStyle', 'type');
        }

        if (!_.isBoolean(enableToggle)){
            CONSOLE.prop.throw(BxAccordion, 'enableToggle', 'type');
        }

        if (!_.isBoolean(collapsible)){
            CONSOLE.prop.throw(BxAccordion, 'collapsible', 'type');
        }

        if (!_.isFunction(onActivate)){
            CONSOLE.prop.throw(BxAccordion, 'onActivate', 'type');
        }

        if (!_.isFunction(onSelected)){
            CONSOLE.prop.throw(BxAccordion, 'onSelected', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxAccordion, 'onRender', 'type');
        }

        if (!_.isUndefined(stateOptions)) {

            if (_.isPlainObject(stateOptions)){
                initStateOptions(stateOptions);
            }
            else {
                CONSOLE.prop.throw(BxAccordion, 'stateOptions', 'type');
            }
        }

        function initStateOptions(stateOptions = {}){

            let {fold, expire, key} = stateOptions;

            if (_.isUndefined(fold)) stateOptions.fold = DEFAULT_STATE_OPTIONS.fold;

            if (_.isUndefined(expire)) stateOptions.expire = DEFAULT_STATE_OPTIONS.expire;

            if (_.isUndefined(key)) stateOptions.key = DEFAULT_STATE_OPTIONS.key;

            if (!_.isBoolean(stateOptions.expire)){
                CONSOLE.prop.throw(BxAccordion, 'stateOptions.expire', 'type');
            }

            if (_.isString(stateOptions.key)) {

                if (_.isEmpty(stateOptions.key.trim())) {
                    CONSOLE.prop.throw(BxAccordion, 'stateOptions.key', 'value');
                }
            }
            else {
                CONSOLE.prop.throw(BxAccordion, 'stateOptions.key', 'type');
            }

            if (!_.isBoolean(stateOptions.expire)){
                CONSOLE.prop.throw(BxAccordion, 'stateOptions.expire', 'type');
            }

            if (!_.isBoolean(stateOptions.fold)){
                CONSOLE.prop.throw(BxAccordion, 'stateOptions.fold', 'type');
            }
        }

        this[OPTS] = {active, event, animate, collapsible, enableToggle, stateOptions};

        let $root = BxAccordion.$root;

        this.$root = $root;

        this.items = _parseItems(items);
        this.model = _parseTreeModel(this.items);

        this.onActivate = onActivate;
        this.onSelected = onSelected;
        this.onRender = onRender;

        this.depth = 0;
    }

    /**
     * 렌더 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    render(){

        let $container = BASE.appendContainerElement(BxAccordion.$root, BxAccordion.className);

        this.$container = $container;

        let html = _createDOMElement(this.items);

        $container.append(html);

        _initState.call(this);
        _addEvent.call(this);

        // active 설정값으로 해당 아이템을 활성화 시킨다.
        if (_.isNumber(this[OPTS].active)){

            $container.find(`.${SELECTOR_NAME.superParent}>li`)
                .eq(this[OPTS].active)
                .children('a')
                .click();
        }

        this.onRender.call(this, this, this.$container);

        return this;
    }

    /**
     * 템플릿 바인딩 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    bind(){

        /*
         내부 바인드 함수
         */
        let _bind = () => {

            let $children = this.$root.children();

            this.$container = BASE.appendContainerElement(this.$root, BxAccordion.className);

            /*
             * $children DOM Tree 이동
             * Jq.append 함수 이용시, 존재하는 DOM 을 인자로 사용하는 경우, 해당 돔의 위치만 이동하게 된다.
             * $children 은 Jq Dom 인스턴스이며, 내부에 context 를 통해 해당 돔을 삭제 및 지정 한 DOM 에 추가 한다.
             * http://api.jquery.com/append/ 참고.
             */
            this.$container.append($children);

            _initState.call(this);
            _addEvent.call(this);

            this.onRender.call(this, this, this.$container);
        };

        window.setTimeout(() => _bind(), 1);

        return this;
    }

    /**
     *
     * @deprecated Deprecated in: v3.0.0, Deprecated since: v3.8.0
     * 모든 section 에 대한, closed 유/무를 설정하는 함수.
     *
     * @param {boolean} [collapsibleType = true] collapsible 유/무
     * @param {number} [activeIndex = ''] active 될 section 번호
     *
     * @returns {object} 생성된 인스턴스
     */
    collapsible(collapsibleType = true, activeIndex = 0){
        CONSOLE.prop.throw(BxAccordion, 'collapsible', 'deprecatedMethod');
    }

    /**
     *
     * item 열기/닫기 함수. (collapsible 이 false일 경우 열기만 가능)
     *
     * @param {string} id - item id
     *
     * @returns {object} 생성된 인스턴스
     */
    toggle(id = ''){

        if (!_.isString(id) || _.isEqual(id.trim(), '')) {
            CONSOLE.prop.throw(BxAccordion, 'id', 'type');
        }

        _toggle.call(this, id);
        _selectItem.call(this, id);

        return this;
    }

    /**
     *
     * item 열기 함수.
     *
     * @param {string} id - item id
     *
     * @returns {object} 생성된 인스턴스
     */
    open(id = '') {

        if (!_.isString(id) || _.isEqual(id.trim(), '')) {
            CONSOLE.prop.throw(BxAccordion, 'id', 'type');
        }

        let $target = this.$container.find(`#${id}`);
        let $parents = $target.parents('li').not(`.${SELECTOR_NAME.opened}`);

        // 부모가 닫혀 있거나 또는 자기 자신이 닫혀 있을 경우
        // 열기를 시도 한다.
        if ($parents.length || !$target.hasClass(SELECTOR_NAME.opened)) _toggle.call(this, id);

        _selectItem.call(this, id);

        return this;
    }

    /**
     *
     * item 닫기 함수. (collapsible 이 false일 경우 닫기 불가능)
     *
     * @param {string} id - item id
     *
     * @returns {object} 생성된 인스턴스
     */
    close(id = '') {

        if (!_.isString(id) || _.isEqual(id.trim(), '')) {
            CONSOLE.prop.throw(BxAccordion, 'id', 'type');
        }

        let $target = this.$container.find(`#${id}`);

        if ($target.hasClass(SELECTOR_NAME.opened)) _toggle.call(this, id);

        _selectItem.call(this, id);

        return this;
    }

    /**
     *
     * 아이템 추가 함수 (마지막 순서로 아이템을 추가한다.)
     *
     * @param {string} [parentId = ''] - 부모 id
     * @param {object[]|object|string} item - item 객체
     * @param {function}[callback = function(){}] - callback
     *
     * @returns {object} 생성된 아이템
     */
    addItems(...args){
        return  _insertItems.call(this, 'append', [...args]);
    }

    /**
     *
     * 아이템 추가 함수 (처음 순서로 아이템을 추가한다.)
     *
     * @param {string} [parentId = ''] - 부모 id
     * @param {object[]|object|string} item - item 객체
     * @param {function}[callback = function(){}] - callback
     *
     * @returns {object} 생성된 아이템
     */
    prependItems(...args){
        return  _insertItems.call(this, 'prepend', [...args]);
    }

    /**
     *
     * 아이템 삭제 함수
     *
     * @param {array|string} id - item id
     * @param {function} [callback=function(){}] - callback
     *
     * @returns {object} 생성된 아이템
     */
    deleteItems(ids = '', callback = function (){}){

        if (_.isEmpty(ids) || (!_.isString(ids) && !_.isArray(ids))){
            CONSOLE.prop.throw(bxAccordion, 'ids', 'type');
        }

        if (!_.isFunction(callback)){
            CONSOLE.prop.throw(BxAccordion, 'callback', 'type');
        }

        const $container = this.$container;
        const item = {
            ids: [].concat(ids)
        };

        _.map(item.ids, (id)=>{
            _deleteItem.call(this, this.model, id);
            let $target = $container.find(`#${id}`);
            let $parent = $target.parent().parent();

            $target.detach();

            if ($parent.hasClass(SELECTOR_NAME.parent) &&
                $parent.find('>ul').children().length === 0){

                $parent.removeClass(SELECTOR_NAME.parent);
                $parent.find(`>.${SELECTOR_NAME.label}`)
                    .removeClass(`${SELECTOR_NAME.label} ${SELECTOR_NAME.checked}`);
            }
        });

        callback.call(item);

        return item;
    }

    /**
     *
     * 아이콘 변경 함수
     *
     * @param {string} id - item id
     * @param {string} iconClass - icon class 선택자
     *
     * @returns {object} 생성된 인스턴스
     */
    setIcon(id = '', iconClass = ''){

        if (!_.isString(id) || _.isEqual(id.trim(), '')) {
            CONSOLE.prop.throw(BxAccordion, 'id', 'type');
        }

        if (!_.isString(iconClass)){
            CONSOLE.prop.throw(BxAccordion, 'iconClass', 'type');
        }

        let item = _getItem(this.model, id);

        if (!item) return this;

        if (item.icon){
            this.$container.find(`#${id}>a>i`)
                .attr('class', iconClass);
        }
        else{

            if (item.items){
                this.$container.find(`#${id}>a`)
                    .prepend(`<i class="${iconClass}"></i>`);
            }
            else{
                this.$container.find(`#${id}>a`)
                    .prepend(`<i class="${iconClass}"></i>`);
            }

        }

        item.icon = iconClass;

        return this;
    }

    /**
     *
     * item 객체 반환 함수
     *
     * @param {string} id - item id
     *
     * @returns {object} item 객체
     */
    getItem(id = ''){

        if (!_.isString(id) || _.isEqual(id.trim(), '')) {
            CONSOLE.prop.throw(BxAccordion, 'id', 'type');
        }

        return _getItem(this.model, id);
    }

    /**
     *
     * 선택된 item id 반환 함수
     *
     * @returns {string} 선택된 item id
     */
    getSelectedId(){

        return _getSelectedId.call(this);
    }
}

/**
 * items parse 함수
 * @param items - 입력받은 items
 * @param depth - 깊이
 * @returns {Array} - item 객체
 * @private
 */
function _parseItems(items = [], depth = 1){

    return items.map((item) =>{

        if (_.isString(item)){
            item = {title: item};
        }

        item.id = item.id || Util.uniqueId(`${SELECTOR_NAME.item}`);
        item.depth = depth;
        item.title = item.title || STYLE_VALUE.space;

        if (item.items) item.items = _parseItems(item.items, depth + 1);

        return item;

    });
}

/**
 * items를 검색/추가/삭제 용이하도록 tree model 형태로 변환하는 함수
 *
 * @param {array} items - items
 * @param {string} parentId - 부모 id (재귀 호출 시 사용)
 * @param {object} parentId - 변환된 객체 (재귀 호출 시 사용)
 *
 * @private
 * @memberOf BxAccordion
 */
function _parseTreeModel(items = [], parentId = ROOT_ID, treeModel = {}){

    _.map(items, (item) =>{

        let id = item.id;

        treeModel[id] = {
            id: id,
            parentId: parentId,
            title: item.title,
            icon: item.icon,
            depth: item.depth
        };

        if (item.items){
            treeModel[id].items = item.items;
            _parseTreeModel(item.items, id, treeModel);
        }

    });

    return treeModel;
}

/**
 * accordion html을 반환하는 함수
 *
 * @param {array} items - item parameter 배열
 *
 * @private
 * @memberOf BxAccordion
 */
function _createDOMElement(items = []){

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxAccordion, 'type', 'items');
    }

    let html = [];

    html.push(`<div class="${SELECTOR_NAME.wrapper}">`);
    html.push(`<ul class="${SELECTOR_NAME.superParent}">`);

    html.push(_createItemsElements(items));

    html.push('</ul>');
    html.push('</div>');

    return html.join('');
}

/**
 * item list html을 반환하는 함수
 *
 * @param {array} items - item parameter 배열
 *
 * @private
 * @memberOf BxAccordion
 */
function _createItemsElements(items = []){

    return items.map((item) =>{
        return _createItemElement(item);

    }).join('');
}


/**
 * accordion item html을 반환하는 함수
 *
 * @param {array} items - item parameter 배열
 * @param {number} [depth = 1] - group의 depth
 *
 * @private
 * @memberOf BxAccordion
 */
function _createItemElement(item = {}){

    if (!_.isPlainObject(item) && !_.isString(item)){
        CONSOLE.prop.throw(BxAccordion, 'item', 'type');
    }

    let html = [];

    let id = item.id || Util.uniqueId(`${SELECTOR_NAME.item}`);
    let depth = item.depth;
    let padding = depth * STYLE_VALUE.paddingLeft;

    html.push(`<li ${item.items ? 'class="' + SELECTOR_NAME.parent + '"' : ''} id="${id}">`);

    if (item.items){

        let id = Util.uniqueId(`${SELECTOR_NAME.expander}`);

        html.push(`<a id="${id}" href="#" onclick="return false;" class="label" style="padding-left:${padding}px">`);

        if (item.icon) html.push(`<i class="${item.icon}"></i>`);

        html.push(`${item.title}</a>`);

        html.push(_createChildItemElement(item.items, ++depth));
    }
    else{

        let title = item.title;
        let $item = $(title);

        if (/^</.test(item.title) && $item[0] && $item[0].tagName === 'A'){

            if (item.icon) $item.prepend(`<i class="${item.icon}"></i>`);

            $item.css('paddingLeft', padding);

            html.push($item.get(0).outerHTML);
        }
        else{

            html.push(`<a href="#" onclick="return false;" style="padding-left:${padding}px">`);

            if (item.icon) html.push(`<i class="${item.icon}"></i>`);

            html.push(`${title}</a>`);
        }
    }

    html.push('</li>');

    return html.join('');
}

/**
 * accordion group html을 반환하는 함수
 *
 * @param {array} items - item parameter 배열
 *
 * @private
 * @memberOf BxAccordion
 */
function _createChildItemElement(items = []){

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxAccordion, 'type', 'items');
    }

    let html = [];

    html.push(`<ul>`);

    html.push(_createItemsElements(items));

    html.push('</ul>');

    return html.join('');
}

/**
 * 초기 state 설정
 *
 * @private
 * @memberOf BxAccordion
 */
function _initState() {

    let enableState = _.isPlainObject(this[OPTS].stateOptions);

    if (!enableState) return;

    _setSession.call(this);
    _setDefaultState.call(this);
}

/**
 * 브라우져 종료시 state 초기화 유무를 설정
 *
 * @private
 * @memberOf BxAccordion
 */
function _setSession(){

    let stateOptions = this[OPTS].stateOptions;
    let expire = stateOptions.expire;

    if (expire){

        let key = stateOptions.key;
        let session = STORAGE.sessionStorage.get(key);

        if (!session){

            STORAGE.localStorage.remove(key);
            STORAGE.localStorage.remove(key + '_state');
            STORAGE.sessionStorage.set(stateOptions.key, 'sessionOn');
        }
    }
}

/**
 * state 정보를 통해 초기 accroidon 열기/닫기 또는 포커스 설정
 *
 * @private
 * @memberOf BxAccordion
 */
function _setDefaultState() {

    let originState = _getOriginState.call(this);
    let $container = this.$container;

    // state 기능을 사용시 active 를 사용 할 수 없다.
    this[OPTS].active = null;

    if (!originState) return;

    _.forEach(originState[STATE.open], (value) => {

        let $target =  $container.find('#' + value);

        $target.addClass(SELECTOR_NAME.opened);

        $target.find('>a').addClass(SELECTOR_NAME.checked);
        $target.find('ul:first').slideDown(0);

        this.onActivate.apply(this, [null, $target]);
    });

    _.forEach(originState[STATE.select], (id) => {

        const $target = _selectItem.call(this, id);

        this.onSelected.apply(this, [null, $target]);
    });
}

/**
 * local 저장소에 저장되어 있는 state 를 반환
 *
 * @private
 * @memberOf BxAccordion
 */
function _getOriginState(){

    let originState = false;
    let key = this[OPTS].stateOptions.key;

    if (STORAGE.localStorage.get(key)){
        originState = JSON.parse(STORAGE.localStorage.get(key));
    }

    return originState;
}

/**
 * 이벤트 추가 함수
 *
 * @private
 * @memberOf BxAccordion
 */
function _addEvent(){

    let $container = this.$container;
    let opts = this[OPTS];

    if (opts.collapsible || opts.enableToggle){

        $container.on('click', `a.${SELECTOR_NAME.label}`, (e) =>{

            _collapsible.call(this, $(e.currentTarget).parent(), opts.animate, e);

            //this.onSelected.call(this, e, $(e.target));

        });

    }
    else{

        $container.on('click', `a.${SELECTOR_NAME.label}`, (e) =>{

            _nonCollapsible.call(this, $(e.currentTarget).parent(), opts.animate, e);

            //this.onSelected.call(this, e, $(e.target));
        });
    }

    if (opts.event === 'mouseover'){

        $container.on('mouseover', `a.${SELECTOR_NAME.label}`, (e)=>{
            $(e.target).click();
        });
    }

    $container.on('click', (e) => {

        const $target = $(e.target);
        const $li = $target.parent();
        const id = $li.attr('id');

        if (_.isEmpty(id) || _.isUndefined(id)) return;

        _selectItem.call(this,  id);
        _updateSelectItemState.call(this, $li, true);

        this.onSelected.call(this, e, $li);
    });
}

/**
 * Collapsible 가 true 일 때 열기/닫기 함수
 *
 * @param {object} $target = null - 열기/닫기 할 타겟 jquery 객체
 * @param {number} [animate = 0] - 애니메이션 duration
 * @param {object} [e = null] - target event
 *
 * @private
 * @memberOf BxAccordion
 */
function _collapsible($target = null, animate = 0, e = null){

    let {enableToggle} = this[OPTS];

    if ($target.length && !_.isElement($target.get(0))){
        CONSOLE.prop.throw(BxAccordion, '$target', 'type');
    }

    if ($target.hasClass(SELECTOR_NAME.opened)){
        _closeItem.call(this, $target, animate);
    }
    else {

        !enableToggle && _closeItem.call(this, $target.parent().find(`>.${SELECTOR_NAME.opened}`), animate);
        _openItem.call(this, $target, animate);

        this.onActivate.call(this, e, $target);
    }
}

/**
 * Collapsible 가 false 일 때 열기/닫기 함수
 *
 * @param {object} $target = null - 열기/닫기 할 타겟 jquery 객체
 * @param {number} [animate = 0] - 애니메이션 duration
 * @param {object} [e = null] - target event
 *
 * @private
 * @memberOf BxAccordion
 */
function _nonCollapsible($target = null, animate = 0, e = null){

    let { enableToggle } = this[OPTS];

    if ($target.length && !_.isElement($target.get(0))){
        CONSOLE.prop.throw(BxAccordion, '$target', 'type');
    }

    if (!_.isNumber(animate)){
        CONSOLE.prop.throw(BxAccordion, 'animate', 'type');
    }

    if (!$target.hasClass(SELECTOR_NAME.opened)){
        !enableToggle && _closeItem.call(this, $target.parent().find(`>.${SELECTOR_NAME.opened}`), animate);
        _openItem.call(this, $target, animate);
        this.onActivate.call(this, e, $target);
    }
}

/**
 * id 아이템을 선택
 *
 * @param {string} id - 선택할 아이디
 *
 * @private
 * @memberOf BxAccordion
 */
function _selectItem(id = '') {

    const $target =  this.$container.find(`#${id}`);
    const $anchor =  $target.find('>a');

    if (!$anchor.hasClass('selected')) {

        _clearSelectedItem.call(this);

        $anchor.addClass('selected');
    }

    return $target;
}

/**
 * 아이템 선택 해제
 *
 * @private
 * @memberOf BxAccordion
 */
function _clearSelectedItem() {

    let $target = this.$container.find('.selected');

    if ($target.length < 1) return;

    if ($target.attr('class')  === 'selected') {
        $target.removeAttr('class');
    }
    else {
        $target.removeClass('selected');
    }
}

/**
 * 선택 되어 있는 아이템 id를 storage에 저장
 *
 * @param {object} $elem - 열려 있는 element의 jquery 객체
 * @param {boolean} isSelect - 선택 상태
 *
 * @private
 * @memberOf BxAccordion
 */
function _updateSelectItemState($elem = null, isSelect = true) {

    if (!_.isPlainObject(this[OPTS].stateOptions)) return;

    const id = $elem.attr('id');
    const key = this[OPTS].stateOptions.key;
    const originState = _getOriginState.call(this) ? _getOriginState.call(this): {};

    let changedState = {
        selectState : [],
        openState : originState[STATE.open]
    };

    if (isSelect) {
        changedState[STATE.select].push(id);
    }
    else {
        changedState[STATE.select] = [];
    }

    originState[STATE.select] = _.union(originState[STATE.select], changedState[STATE.select]);

    STORAGE.localStorage.set(key, JSON.stringify(changedState));
}

/**
 * item 펼침 함수
 *
 * @param {object} [$item = null] - 타겟 li jquery 객체
 * @param {number} animate - 애니메이션 duration
 *
 * @private
 * @memberOf BxAccordion
 */
function _openItem($target = null, animate = 0){

    if ($target.length && !_.isElement($target.get(0))){
        CONSOLE.prop.throw(BxAccordion, '$item', 'type');
    }

    if (!_.isNumber(animate)){
        CONSOLE.prop.throw(BxAccordion, 'animate', 'type');
    }

    _updateOpenItemsState.call(this, $target, true);

    $target.addClass(SELECTOR_NAME.opened);
    $target.find('>a').addClass(SELECTOR_NAME.checked);
    $target.children('ul').stop(true, true).slideDown(animate);
}

/**
 * item 닫힘 함수
 *
 * @param {object} [$item = null] - 타겟 li jquery 객체
 * @param {number} animate - 애니메이션 duration
 *
 * @private
 * @memberOf BxAccordion
 */
function _closeItem($target = null, animate = 0){

    if ($target.length && !_.isElement($target.get(0))){
        CONSOLE.prop.throw(BxAccordion, '$item', 'type');
    }

    if (!_.isNumber(animate)){
        CONSOLE.prop.throw(BxAccordion, 'animate', 'type');
    }

    _updateOpenItemsState.call(this, $target, false);

    $target.removeClass(SELECTOR_NAME.opened);
    $target.find('>a').removeClass(SELECTOR_NAME.checked);
    $target.children('ul').stop(true, true).slideUp(animate);
}

/**
 * open 되어 있는 아이템 id들을 storage에 저장
 *
 * @param {object} $elem - 열려 있는 element의 jquery 객체
 * @param {boolean} isOpen - 열기/닫기 상태 정보
 *
 * @private
 * @memberOf BxAccordion
 */
function _updateOpenItemsState($elem = null, isOpen = true) {

    if (!_.isPlainObject(this[OPTS].stateOptions)) return;

    let id =  $elem.attr('id');
    let fold = this[OPTS].stateOptions.fold;
    let key = this[OPTS].stateOptions.key;

    if (_.isUndefined(id) || !fold) return;

    let originState = _getOriginState.call(this) ? _getOriginState.call(this): {};

    let changedState = {
        openState : [],
        selectState : originState[STATE.select]
    };

    if (isOpen){
        changedState[STATE.open].push(id);
    }
    else {

        originState[STATE.open] = _.remove(originState[STATE.open], function(n) {
            return n !== id;
        });
    }

    originState[STATE.open] = _.union(originState[STATE.open], changedState[STATE.open]);

    STORAGE.localStorage.set(key, JSON.stringify(originState));
}

/**
 * item 추가 함수
 *
 * @param {string} [type = 'append'] - 추가 순서 타입 (append / prepend)
 * @param {object} [args = []] - arguments
 *
 * @private
 * @memberOf BxAccordion
 */
function _insertItems(type = 'append', args = []){

    let pId = '', items, callback;

    if (args.length === 0){

        CONSOLE.prop.throw(BxAccordion, 'items', 'value');
    }
    else if (args.length === 1){

        items = args[0];
    }
    else if (args.length === 2){

        if (_.isString(args[0]) && !_.isFunction(args[1])){
            pId = args[0];
            items = args[1];
        }
        else{
            items = args[0];
            callback = args[1];
        }
    }
    else if (args.length === 3){

        pId = args[0];
        items = args[1];
        callback = args[2];
    }

    if (pId && !_.isString(pId)){
        CONSOLE.prop.throw(BxAccordion, 'parentId', 'type');
    }
    if (!_.isString(items) && !_.isPlainObject(items) && !_.isArray(items)){
        CONSOLE.prop.throw(BxAccordion, 'items', 'type');
    }
    if (callback && !_.isFunction(callback)){
        CONSOLE.prop.throw(BxAccordion, 'callback', 'type');
    }

    items = _.isArray(items) ? items : [items];

    if (_.isEmpty(pId) || _.isEqual(pId, ROOT_ID)){
        items = _parseItems(items);
        let itemHtmls = _createItemsElements(items);

        this.$container.find(`.${SELECTOR_NAME.superParent}`)[type](itemHtmls);
    }
    else{

        let item = _getItem(this.model, pId);
        if (!item) return this;

        let depth = item.depth + 1;

        if (!item.items || !item.items.length){

            item.items = _parseItems(_.isString(items) ? [items] : items, depth);

            let $target = this.$container.find(`#${pId}`);
            $target.addClass(SELECTOR_NAME.parent);
            $target.html(_createParentItem(item).join(''));
        }
        else{

            items = _parseItems(items, depth);
            let itemHtmls = _createItemsElements(items);

            this.$container.find(`#${pId}>ul`)[type](itemHtmls);
        }

        if (type === 'append') this.model[pId].items.push(...items);
        else                   this.model[pId].items.unshift(...items);

    }

    _parseTreeModel(items, pId, this.model);

    items = _.assign(items, {parentId:pId});

    callback && callback.call(items);

    return items;
}

/**
 * 부모 item html을 반환하는 함수
 *
 * @param {array} items - item 객체
 *
 * @private
 * @memberOf BxAccordion
 */
function _createParentItem(item = {}){

    let html = [];

    let id = Util.uniqueId(`${SELECTOR_NAME.expander}`);
    let depth = item.depth;
    let padding = depth * STYLE_VALUE.paddingLeft;

    html.push(`<a id="${id}" class="label" style="padding-left:${padding}px">`);

    if (item.icon) html.push(`<i class="${item.icon}"></i>`);

    html.push(`${item.title}</a>`);

    html.push(_createChildItemElement(item.items));

    return html;
}

/**
 * items 를 id로 탐색하여 일치하는 item을 삭제하는 함수
 *
 * @param {object} model - model 객체
 * @param {string} id - 검색하고자 하는 id
 *
 * @private
 * @memberOf BxAccordion
 */
function _deleteItem(model = {}, id = ''){

    if (!_.isPlainObject(model)){
        CONSOLE.prop.throw(BxAccordion, 'model', 'type');
    }

    if (!_.isString(id)){
        CONSOLE.prop.throw(BxAccordion, 'id', 'type');
    }

    const isExist = model.hasOwnProperty(id);
    if (!isExist) return;

    const target = model[id];
    const children = _.clone(target.items);

    _.forEach(children, (v, k) =>{
        _deleteItem.call(this, model, v.id);
    });

    if (target && model[target.parentId]){
        _.remove(model[target.parentId].items, item => item.id === id);
    }
    _.unset(model, id);
}

/**
 * items 를 id로 탐색하여 일치하는 item을 반환하는 함수
 *
 * @param {object} model - model 객체
 * @param {string} id - 검색하고자 하는 id
 *
 * @private
 * @memberOf BxAccordion
 */
function _getItem(model = {}, id = ''){

    return model[id];
}


/**
 *
 * item 열기/닫기 함수. (collapsible 이 false일 경우 열기만 가능)
 *
 * @param {string} id - item id
 *
 * @returns {object} 생성된 인스턴스
 */
function _toggle(id = '') {

    let $target = this.$container.find(`#${id}`);
    let $parents = $target.parents('li').not(`.${SELECTOR_NAME.opened}`);
    let _this = this;

    let $superParent = null;

    // 부모 여부를 확인하여 닫혀있는 부모를 열기 상태로 변경한다.
    // 부모 중 하나라도 닫혀있는 상태일 경우, 자신 또한 닫혀있는 상태로 간주한다.

    if ($parents.length) _closeItem.call(this, $target, 0);

    _.forEach($parents, (parent, i) =>{
        let $parent = $(parent);
        let length = $parents.length - 1;

        if (length !== 1 && i === length) $superParent = $parent;
        else _activeParent($parent, 0);
    });

    // 애니메이션 효과를 위해 부모가 있는 경우 닫힌 부모 중 최상위 부모를 마지막에 open 한다.
    let animate = $superParent ? 0 : this[OPTS].animate;

    if (this[OPTS].collapsible)  _collapsible.call(this, $target, animate);
    else                         _nonCollapsible.call(this, $target, animate);

    if ($superParent) _activeParent($superParent, this[OPTS].animate);

    function _activeParent($parent, animate){
        _closeItem.call(_this, $parent.parent().find(`>.${SELECTOR_NAME.opened}`), animate);
        _openItem.call(_this, $parent, animate);
    }

    return this;
}

/**
 * 선택 되었을때 이벤트 등록
 *
 * @private
 * @memberOf BxAccordion
 */
//function _addSelectEventListener() {
//
//    let $container = this.$container;
//
//    $container.on('click', (e) => {
//
//        let $target = $(e.target);
//        let id = $target.parent().attr('id');
//
//        if (_.isEmpty(id) || _.isUndefined(id)) return;
//
//        _selectItem.call(this,  id);
//        _updateSelectItemState.call(this, $target, true);
//    });
//}

/**
 * 선택된 item id 반환 함수
 *
 * @returns {string} id
 * @private
 */
function _getSelectedId(){

    let $item = this.$container.find('.selected').parent();
    let id = $item.attr('id');

    if (_.isEmpty(id)) id = '';

    return id;
}


module.exports = BxAccordion;
