// bxuip-angular.js-ui/src/lib/bx-ui/app/components/button.js
/**
 * render 메서드 호출 시, 호출될 callback 함수
 *
 * @callback bx-render
 *
 * @memberOf BxButton
 */

/**
 * eventName에 할당한 이벤트 발생 시, 호출될 callback 함수
 * (ex. bx-click="function(e){}")
 *
 * @callback bx-eventName
 * @param {object} event - event 객체
 *
 * @memberOf BxButton
 */

/**
 * 버튼 toggle 시, 호출될 callback 함수
 *
 * @callback bx-toggle
 * @param {string} value - toggle된 버튼의 value
 *
 * @memberOf BxButton
 */

/**
 * @class BxButton
 * @classDesc BxButton 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 * @example
 *
 * <script>
 *
 *     $(document).ready(function(){
 *         // DOM이 ready된 시점에서 초기화 함수를 호출한다.
 *         $().BxButton().initialize();
 *     });
 *
 *     window.formMethod = {
 *       setButtonRender: function(e){
 *         console.log('Render')
 *       },
 *       setButtonLabel: function(e){
 *         this.setLabel('change button');
 *       },
 *       getSliderChecked: function(e){
 *         var target = document.getElementById('slider1');
 *         alert('체크된 버튼 값' + target.getChecked());
 *      }
 *     };
 *
 *    function toggle(value){
 *        alert(value);
 *     }
 *
 * </script>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // default type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *
 *  <bx-button id="button1"
 *             label="button"
 *             fill="false"
 *             icon="bw-icon i-20 i-trash"
 *             bx-click="formMethod.setButtonLabel"
 *             bx-render="formMethod.setButtonRender"
 *             colorLevel="1">
 *  </bx-button>
 *
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // toggle type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *  <bx-button id="toggle1"
 *             type="toggle"
 *             bx-toggle="toggle"
 *             label="{left: 'Week', right: 'Month'}"
 *             value="{left: 'week', right: 'month'}"
 *             defaultValue="month"
 *             colorLevel="1"
 *             bx-render="formMethod.setButtonRender">
 *  </bx-button>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // switch type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *  <bx-button id="switch"
 *             type="switch"
 *             bx-toggle="toggle"
 *             defaultValue="on"
 *             bx-mouseover="function(){console.log('mouseover')}"
 *             bx-click="function(){console.log('click')}"
 *             bx-render="formMethod.setButtonRender"
 *             label="ON">
 *  </bx-button>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // slider type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *  <bx-button id="slider"
 *             type="slider"
 *             defaultValue="on"
 *             bx-toggle="toggle"
 *             bx-render="formMethod.setButtonRender">
 *  </bx-button>
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;

// 사용자에게 제공되는 API 목록
const APIS = {

    setLabel: _setLabel,
    setIcon: _setIcon,
    setChecked: _setChecked,
    getChecked: _getChecked

};

// 전역 클래스 객체
const CLASS_NAME = {
    "bxUiButton": 'bx-button-'
};

const SELECTOR_NAME = {
    default: 'bx-button',
    switch: 'bx-switch',
    switchType1: 'bx-switch3',
    switchType2: 'bx-switch4',
    switchInput: 'switch-input',
    switchLabel: 'switch-label',
    switchLabelOn: 'switch-label-on',
    switchLabelOff: 'switch-label-off',
    switchSelection: 'switch-selection',
    toggle: 'bx-switch2',
    sliderId: 'slider-',
    switchId: 'switch-',
    toggleId: 'toggle-',
    leftSuffix: '-left',
    rightSuffix: '-right',
    basic: 'basic',
    color: 'color',
    blind: 'blind'
};

const OPTS = {
    /**
     * @property {object} bx-button - 모든 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-button.type='default'] - <span class='important'>readonly</span> - button 타입(default, toggle, switch, slider)
     * @property {function} [bx-button.bx-render=function(){}] - render 시, 호출될 callback 함수
     * @property {function} [bx-button.bx-[eventName]=function(){}] - 생성된 button 엘리먼트에 해당 이벤트 핸들러를 바인딩시킨다.(ex. bx-click -> elem.onclick = function(){})
     *
     * @alias Attributes - All Type
     *
     * @memberof BxButton
     */
    common: {},
    /**
     * @property {object} bx-button - default 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-button.action='button'] - <span class='important'>readonly</span> - 버튼 action(button, submit, reset)
     * @property {number} [bx-button.colorLevel=3] - <span class='important'>readonly</span> - 버튼 색상 단계 설정 ( 1 ~ 5)
     * @property {boolean} [bx-button.fill=true] - <span class='important'>readonly</span> - 버튼의 배경 색상 유/무
     * @property {string} [bx-button.label=''] - label의 text 값
     * @property {string} [bx-button.icon=''] - 버튼 icon의 class 설정, 또는 로딩 아이콘 설정 (default, spin, dolphin)
     *
     * @alias Attributes - Default Type
     *
     * @memberof BxButton
     */
    default: {
        "attr": {},
        "label": '',
        "action": 'button',
        "colorLevel": 3,
        "fill": true,
        "icon": '',
        "bx-render": function (){
        }
    },
    /**
     * @property {object} bx-button - toggle 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {number} [bx-button.colorLevel=3] - <span class='important'>readonly</span> - 버튼 색상 단계 설정 ( 1 ~ 5)
     * @property {string} [bx-button.defaultValue=value.left] - 버튼에 기본으로 active 될 value
     * @property {object} bx-button.value - <span class='important'>readonly</span> - toggle 버튼의 value 설정 객체 ({left: '', right: ''})
     * @property {object} bx-button.label - toggle 버튼의 label 설정 객체 ({left: '', right: ''})
     * @property {function} [bx-button.bx-toggle=function(){}] - 버튼 toggle 시, 호출될 callback 함수
     *
     * @alias Attributes - Toggle Type
     *
     * @memberof BxButton
     */
    toggle: {
        "attr": {},
        "label": {},
        "value": {},
        "defaultValue": '',
        "colorLevel": 3,
        "bx-toggle": function (){
        },
        "bx-render": function (){
        }
    },
    /**
     * @property {object} bx-button - switch 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-button.label=''] - label의 text 값
     * @property {string} [bx-button.icon=''] - icon의 class
     * @property {string} [bx-button.defaultValue='off'] - 버튼에 기본으로 active 될 value ( 'on' 또는 'off' 만 사용 가능)
     * @property {function} [bx-button.bx-toggle=function(){}] - 버튼 toggle 시, 호출될 callback 함수
     *
     * @alias Attributes - Switch Type
     *
     * @memberof BxButton
     */
    switch: {
        "attr": {},
        "label": '',
        "icon": '',
        "defaultValue": '',
        "bx-toggle": function (){
        },
        "bx-render": function (){
        }
    },
    /**
     * @property {object} bx-button - slider 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-button.defaultValue='off'] - 버튼에 기본으로 active 될 value ( 'on' 또는 'off' 만 사용 가능)
     * @property {function} [bx-button.bx-toggle=function(){}] - 버튼 toggle 시, 호출될 callback 함수
     *
     * @alias Attributes - Slider Type
     *
     * @memberof BxButton
     */
    slider: {
        "defaultValue": 'off',
        "bx-toggle": function (){
        },
        "bx-render": function (){
        }
    }
};

const ATTRNAMES = {
    "defaultvalue": 'defaultValue',
    "colorlevel": 'colorLevel'
};

class BxButton {

    constructor(){
        this.$root = BxButton.$root;
        this.opts = OPTS;
    }

    /**
     * 초기화 함수
     *
     * <p>
     *     DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
     * </p>
     *
     * @returns {object} 생성된 인스턴스
     */
    initialize(){

        this.elems = [];

        $.each(this.$root.find('bx-button'), (i, el) =>{

            let $el = $(el);

            let options = _createComponentOptions($el);

            let $container = _parseCustomElement($el, options);

            _setNativeAttribute.call($container, options);

            _setThemeColor($container.children(), options);

            if (options.type !== 'default'){

                UTIL.addAllEventListenerWithProperty(
                options.type !== 'slider' ? $container.find('label') : $container, options.events);
                _setToggleEvent.call($container, options);
            }
            else{

                UTIL.addAllEventListenerWithProperty($container, options.events);
            }

            $container.prop('bxrender')();

            this.elems.push($container.get(0));

        });

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxButton.isCustomElement = true;

/**
 * 버튼 아이콘 설정 함수
 * @param {string} iconClass|loadingType - 아이콘 클래스 | loading icon type (default, spin, dolphin)
 * @returns {object} 생성된 인스턴스
 * @alias setIcon
 * @memberOf BxButton
 */
function _setIcon(icon = ''){

    if (!icon || !_.isString(icon)){
        CONSOLE.prop.throw(BxButton, 'icon', 'all');
    }

    let $buttonElem = $(this);

    let buttonType = _getButtonType($buttonElem);

    if (buttonType !== 'default' && buttonType !== 'switch'){
        return this;
    }
    $buttonElem.find('i').detach();

    let isLoadingIcon = icon === 'dolphin' || icon === 'spin' || icon === 'default';

    if (isLoadingIcon){

        let loadingIcon = 'bw-icon i-20 bx-pulse ';

        switch (icon){

            case 'dolphin' :
                loadingIcon += 'i-spinner3 ';
                break;
            case 'spin' :
                loadingIcon += 'i-spinner2';
                break;
            case 'default' :
                loadingIcon += 'i-spinner';
                break;
        }

        if (this.fill) loadingIcon += '-white';

        icon = loadingIcon;

    }

    let iconElem = '<i class="' + icon + '"></i>';

    if (buttonType === 'default'){
        $buttonElem.find('button').prepend(iconElem);
    }
    else{
        $buttonElem.find('.switch-label').html(iconElem);
    }

    return this;

}

/**
 * 버튼 text 설정 함수
 * @param {string|object} [label=''] - 버튼 text
 * @returns {object} 생성된 인스턴스
 * @alias setLabel
 * @memberOf BxButton
 */
function _setLabel(label = ''){

    let $buttonElem = $(this);

    let type = _getButtonType($buttonElem);

    if (type === 'slider') return this;

    if (type === 'toggle'){

        if (!_.isPlainObject(label)){
            CONSOLE.prop.throw(BxButton, 'label', 'type');
        }

        let left = label.left, right = label.right;

        $buttonElem.find('label').first().text(left);
        $buttonElem.find('label').last().text(right);

    }
    else if (type === 'switch'){

        if (!_.isString(label)){
            CONSOLE.prop.throw(BxButton, 'label', 'type');
        }

        let $div = $buttonElem.find('input').parent('div');

        if (label !== ''){
            $div.removeClass(SELECTOR_NAME.switchType1);
            $div.addClass(SELECTOR_NAME.switchType2);
        }
        else{
            $div.removeClass(SELECTOR_NAME.switchType2);
            $div.addClass(SELECTOR_NAME.switchType1);
        }

        $buttonElem.find('label').html(label);
    }
    else{

        if (!_.isString(label)){
            CONSOLE.prop.throw(BxButton, 'label', 'type');
        }

        $buttonElem.find('span').html(label);
    }

    return this;
}

/**
 * 버튼 활성화 함수 (switch / slider / toggle)
 * @param {string} value - 활성화 하고자 하는 버튼 value
 * @returns {object} 생성된 인스턴스
 * @alias setChecked
 * @memberOf BxButton
 */
function _setChecked(value = ''){

    if (!value || !_.isString(value)){
        CONSOLE.prop.throw(BxButton, 'value', 'all');
    }

    let $buttonElem = $(this);

    let buttonElem = $buttonElem.get(0);

    let type = _getButtonType($buttonElem);

    if (type === 'default') return this;

    let prevValue = null;

    if (type === 'toggle'){

        prevValue = $buttonElem.find('input:checked').val();

        let $checkElem = $buttonElem.find('[value="' + value + '"]');

        if ($checkElem.length < 1){
            CONSOLE.prop.throw(BxButton, 'value', 'value');
        }

        $buttonElem.find('input:checked').prop('checked', false);

        $checkElem.prop('checked', true);

    }
    else{

        if ((type === 'slider' || type === 'switch') && (value !== 'on' && value !== 'off')){

            CONSOLE.prop.throw(BxButton, 'value', 'value');
        }

        let $checkElem = $buttonElem.find('input');

        prevValue = $checkElem.prop('checked') ? 'on' : 'off';

        let isChecked = value === 'off' ? false : true;

        $checkElem.prop('checked', isChecked);

    }

    if (value !== prevValue){
        buttonElem['bxtoggle'].call(buttonElem, value);
    }

    return this;
}

/**
 * 활성화된 버튼의 value를 반환하는 함수 (switch / slider / toggle)
 *
 * @returns {object} 생성된 인스턴스
 * @alias getChecked
 * @memberOf BxButton
 */
function _getChecked(){

    let $buttonElem = $(this);

    let type = _getButtonType($buttonElem);

    if (type === 'default') return this;

    if (type === 'toggle'){
        return $buttonElem.find(':checked').val();
    }
    else{
        return $buttonElem.find('input').prop('checked') ? 'on' : 'off';
    }

}

/**
 * custom 태그를 native 태그로 치환한다.
 *
 * @param {object} custom jquery element
 * @param {object} component option
 *
 * @private
 * @memberOf BxButton
 */
function _parseCustomElement($customElem = $(document.documentElement), options = {}){

    let $container = BASE.appendContainerElement(BxButton.$root, BxButton.className);

    if (!options.attr.id) options.attr.id = _.uniqueId(CLASS_NAME.bxUiButton);

    let $component = $(_getButtonHTML($customElem, options.type, options));

    $container.append($component);

    let buttonElem = $container.get(0);

    _setElementProperties(options, buttonElem);

    $customElem.replaceWith($container);

    return $container;

}

/**
 * native attribute 설정 함수
 *
 * @param {object} component options
 *
 * @private
 * @memberOf BxButton
 */
function _setNativeAttribute(options = {}){

    let $target = $(this);

    $.each(options.attr, function (attrName, attrValue){

        if (attrName === 'class'){
            $target.addClass(attrValue);
        }
        else{
            $target.attr(attrName, attrValue);
        }

    }.bind(this));
}

/**
 * 버튼 타입 반환
 * @param {elem} 버튼 elem
 * @returns {string}
 * @private
 * @memberOf BxButton
 */
function _getButtonType(elem){

    let type = elem.attr('type');

    if (type === 'slider' || type === 'switch' || type === 'toggle'){

        return type;
    }
    else{

        return 'default';
    }
}

/**
 * button 설정 함수
 *
 * @param {object} container jquery element
 * @param {string} button type
 * @param {object} component options
 *
 * @private
 * @memberOf BxButton
 */
function _getButtonHTML($container = $(document.documentElement), type = 'default', options = {}){

    switch (type){
        case 'slider' :
            return _setSliderButton($container, options);
            break;
        case 'switch' :
            return _setSwitchButton($container, options);
            break;
        case 'toggle' :
            return _setToggleButton($container, options);
            break;
        default :
            return _setDefaultButton($container, options);
            break;
    }
}


/**
 * toggle event 설정 함수
 *
 * @private
 * @memberOf BxButton
 */
function _setToggleEvent(options){

    let type = options.type;

    switch (type){
        case 'toggle' :
            _setToggleButtonHandler.call(this);
            return;
        case 'slider' :
            _setSliderButtonHandler.call(this);
            return;
        case 'switch' :
            _setSwitchButtonHandler.call(this);
            return;
    }

}

/**
 * switch 버튼의 toggle event 설정 함수
 *
 * @private
 * @memberOf BxButton
 */
function _setSwitchButtonHandler(){

    let $buttonElem = $(this);

    let buttonElem = $buttonElem.get(0);

    $buttonElem.on('change', 'input', (e)=>{

        let $target = $(e.currentTarget);

        let status = $target.is(':checked') ? 'on' : 'off';

        let bxToggle = buttonElem['bxtoggle'] || function (){};

        if (!_.isFunction(bxToggle)){
            CONSOLE.prop.throw(BxButton, 'bxToggle', 'type');
        }

        bxToggle.call(buttonElem, status);

        buttonElem.defaultValue = status;
    });
}

/**
 * toggle 버튼의 toggle event 설정 함수
 *
 * @private
 * @memberOf BxButton
 */
function _setToggleButtonHandler(){

    let $buttonElem = $(this);

    let buttonElem = $buttonElem.get(0);

    $buttonElem.on('change', 'input', (e)=>{

        let $target = $(e.currentTarget);

        let value = $target.val();

        // if (value === buttonElem.defaultValue) return;

        let bxToggle = buttonElem['bxtoggle'] || function (){};

        if (!_.isFunction(bxToggle)){
            CONSOLE.prop.throw(BxButton, 'bxToggle', 'type');
        }

        bxToggle.call(buttonElem, value);

        buttonElem.defaultValue = value;

    });
}

/**
 * slider 버튼의 toggle event 설정 함수
 *
 * @private
 * @memberOf BxButton
 */
function _setSliderButtonHandler(){

    let $buttonElem = $(this);

    let buttonElem = $buttonElem.get(0);

    $buttonElem.on('change', 'input', (e)=>{
        let $target = $(e.currentTarget);

        let status = $target.is(':checked') ? 'on' : 'off';

        let bxToggle = buttonElem['bxtoggle'] || function (){};

        if (!_.isFunction(bxToggle)){
            CONSOLE.prop.throw(BxButton, 'bxToggle', 'type');
        }

        bxToggle.call(buttonElem, status);

        buttonElem.defaultValue = status;
    });
}

/**
 * attribute 정보를 통해 selector 반환
 *
 * @param {object} component options
 * @param {string} button type
 *
 * @return {object}
 *
 * @private
 * @memberOf BxButton
 */
function _setSelector(options = {}, type = ''){

    let id = options.id;
    let elClass = options.class;
    let fill = options.fill;
    let attrs = 'class="' + type;

    if (elClass) attrs += ' ' + elClass;

    if (!fill) attrs += ' ' + SELECTOR_NAME.basic + ' ' + SELECTOR_NAME.color;

    attrs += '"';

    if (id) attrs += ' id="' + id + '"';

    return attrs;

}

/**
 * defulat type button 치환 함수
 *
 * @param {object} container jquery element
 * @param {object} component options
 *
 * @return {string}
 *
 * @private
 * @memberOf BxButton
 */
function _setDefaultButton($container = $(document.documentElement), options = {}){

    let html = [];

    let action = options.action;

    html.push(`<button type="${action}" ${_setSelector(options, SELECTOR_NAME.default)}
                       ${_getTitleAttrText(options.title)}>`);

    if (options.icon) html.push(`<i class="${options.icon}"></i>`);

    if (options.label) html.push(_getAltText(options.label, options.title));

    html.push('</button>');

    return html.join('');

}

/**
 *
 * 라벨과 타이틀 유무에 따라 대체 택스트를 반환하는 함수
 *
 * @param {string} label
 * @param {string} title
 * @returns {string}
 * @private
 */
function _getAltText(label = '', title = ''){

    let altText = label;

    if (_.isEmpty(label)){

        if (!_.isEmpty(title)){
            altText =  `<span class="${SELECTOR_NAME.blind}">${title}</span>`;
        }

    }
    else {
        altText = `<span>${label}</span>`;
    }

    return altText;

}

/**
 * 타이틀 값으로 태그의 속성 텍스트를 반환하는 함수
 *
 * @param {string} title
 * @returns {string}
 * @private
 */
function _getTitleAttrText(title = ''){

    let titleTxt = title;

    if (!_.isEmpty(title)) titleTxt = `title="${title}"`;

    return titleTxt;

}

/**
 * slider type button 치환 함수
 *
 * @param {object} container jquery element
 * @param {object} component options
 *
 * @return {string}
 *
 * @private
 * @memberOf BxButton
 */
function _setSliderButton($container = $(document.documentElement), options = {}){
    const html = [];

    const defaultValue = (options.defaultValue || 'off') === 'off' ? false : true;
    const id = _.uniqueId(SELECTOR_NAME.sliderId);
    const {title, label} = options;

    html.push('<div class="bx-switch">');

    html.push(`<input type="checkbox" id="${id}" ${_getTitleAttrText(title)}`);

    if (defaultValue) html.push('checked="checked"');

    html.push('>');

    html.push(`<label for="${id}" ${_getTitleAttrText()}>`);

    html.push(`<i></i>`);

    if (!_.isEmpty(title)) html.push(_getAltText(label, title));

    html.push(`</label>`);

    html.push('<span></span>');

    html.push('</div>');

    return html.join('');
}

/**
 * toggle type button 치환 함수
 *
 * @param {object} container jquery element
 * @param {object} component options
 *
 * @return {string}
 *
 * @private
 * @memberOf BxButton
 */
function _setToggleButton($container = $(document.documentElement), options = {}){

    let label = _parseObject('label', options.label);
    let value = _parseObject('value', options.value);

    if (!label || !_.isPlainObject(label)){
        CONSOLE.prop.throw(BxButton, 'label', 'all');
    }

    if (!value || !_.isPlainObject(value)){
        CONSOLE.prop.throw(BxButton, 'value', 'all');
    }

    let identifier = _.uniqueId('toggle-');

    let leftText = label.left, rightText = label.right;
    let leftValue = value.left, rightValue = value.right;
    let leftId = identifier + SELECTOR_NAME.leftSuffix,
        rightId = identifier + SELECTOR_NAME.rightSuffix;

    let defaultValue = (options.defaultValue || leftValue) === leftValue ? false : true;

    let switchInputCls = SELECTOR_NAME.switchInput,
        switchLabelCls = SELECTOR_NAME.switchLabel;

    let html = [];

    html.push(`<div class="${SELECTOR_NAME.toggle}">`);

    html.push(`<input type="radio" class="${switchInputCls}" name="${identifier}" id="${leftId}" value="${leftValue}"`);

    if (!defaultValue) html.push(' checked="checked"');

    html.push('>');

    html.push(`<label for="${leftId}" class="${switchLabelCls} ${SELECTOR_NAME.switchLabelOff}" 
                      ${_getTitleAttrText(leftText)}>${leftText}</label>`);

    html.push(`<input type="radio" class="${switchInputCls}" name="${identifier}" id="${rightId}" value="${rightValue}"`);

    if (defaultValue) html.push(' checked="checked"');

    html.push('>');

    html.push(`<label for="${rightId}" class="${switchLabelCls} ${SELECTOR_NAME.switchLabelOn}"
                      ${_getTitleAttrText(rightText)}>${rightText}</label>`);

    html.push(`<span class="${SELECTOR_NAME.switchSelection}"></span>`);

    html.push('</div>');

    return html.join('');
}

/**
 * switch type button 치환 함수
 *
 * @param {object} custom element
 * @param {object} component options
 *
 * @return {string}
 *
 * @private
 * @memberOf BxButton
 */
function _setSwitchButton($container = $(document.documentElement), options = {}){

    let identifier = _.uniqueId(SELECTOR_NAME.switchId);

    let elClass = '',
        {icon, label, title} = options,
        defaultValue = (options.defaultValue || 'off') === 'off' ? false : true;


    if (icon || label){
        elClass = SELECTOR_NAME.switchType2;
    }
    else{
        elClass = SELECTOR_NAME.switchType1;
    }

    let html = [];

    html.push(`<div class="${elClass}">`);

    html.push(`<input type="checkbox" id="${identifier}" class="${SELECTOR_NAME.switchInput}"`);

    if (defaultValue) html.push(' checked="checked"');

    html.push('>');

    html.push(`<label 
                    class="${SELECTOR_NAME.switchLabel}" 
                    for="${identifier}" 
                    data-label-on="on" 
                    data-label-off="off" 
                    ${_getTitleAttrText(title)}>`);

    if (icon){
        html.push(`<i class="${icon}"></i>`);
        html.push(_getAltText('', label || title));
    }
    else {
        html.push(_getAltText(label, title));
    }

    html.push('</label>');

    html.push('</div>');

    return html.join('');

}


/**
 * colorLevel을 통해 HSL 반환
 * @param {object} custom element
 * @param {string} colorLevel
 * @return {object}
 * @private
 * @memberOf BxButton
 */
function _getHSLToColorLevel(themeColor = '', colorLevel = 3){

    if (colorLevel > 5) colorLevel = 5;
    if (colorLevel < 1) colorLevel = 1;

    let reg = new RegExp(/\((.*?)\)/g);

    let themeColorCss = reg.exec(themeColor);

    if (!themeColorCss) return;

    let rgbs = themeColorCss[1].split(',').map((item)=>{
        return parseInt(item);
    });

    let r = rgbs[0], g = rgbs[1], b = rgbs[2], a = rgbs[3];

    if (!_.isNumber(a)) a = 1;

    let hsla = UTIL.RGBAToHSLA(r, g, b, a);

    return UTIL.colorLevelToHSL(colorLevel, hsla);
}

/**
 * default type button 색상 설정 함수
 * @param {object} custom element
 * @param {object} attributes
 * @private
 * @memberOf BxButton
 */
function _setDefaultButtonColor($container = $(document.documentElement), options = {}){

    let isFill = options.fill;

    let colorLevel = parseInt(options.colorLevel) || 3;

    let cssName = isFill ? 'backgroundColor' : 'color';

    let themeColor = $container.filter('.bx-button').css(cssName);

    let hsla = _getHSLToColorLevel(themeColor, colorLevel);

    if (!hsla) return;

    let resultColor = {};

    if (isFill){
        resultColor['background-color'] = hsla;
    }
    else{
        resultColor['color'] = hsla;
        resultColor['box-shadow'] = $.extend({v: 'inset 0 0 0 1px'}, hsla);
    }

    UTIL.setHSL($container, resultColor);

}

/**
 * toggle type button 색상 설정 함수
 * @param {object} custom element
 * @param {object} attributes
 * @private
 * @memberOf BxButton
 */
function _setToggleButtonColor($container = $(document.documentElement), options = {}){

    let colorLevel = parseInt(options.colorLevel) || 3;

    let $switch = $container.find('.' + SELECTOR_NAME.switchSelection);

    let themeColor = $switch.css('backgroundColor');

    let hsla = _getHSLToColorLevel(themeColor, colorLevel);

    if (!hsla) return;

    let resultColor = {'background-color': hsla};

    UTIL.setHSL($switch, resultColor);

}

/**
 * theme 색상 설정 함수
 * @param {object} custom element
 * @param {object} attributes
 * @private
 * @memberOf BxButton
 */
function _setThemeColor($container = $(document.documentElement), options = {}){

    let buttonType = options.type;

    // default type
    if (!buttonType || buttonType === 'default'){
        _setDefaultButtonColor($container, options);
    }

    // toggle Type
    if (buttonType === 'toggle'){
        _setToggleButtonColor($container, options);
    }

}

/**
 * component option 반환 함수
 *
 * @param {object} custom jquery element
 *
 * @return {object} component opts
 *
 * @private
 * @memberOf BxButton
 */
function _createComponentOptions($targetObject = null){

    if (!_.isFunction($targetObject.get) && !_.isPlainObject($targetObject)){
        CONSOLE.prop.throw(BxButton, 'targetObject', 'type');
    }

    let defaultOpts = {};
    let type = $targetObject.attr('type') === undefined ? 'default' : $targetObject.attr('type');

    switch (type){

        case 'toggle':
            $.extend(defaultOpts, OPTS.toggle);
            break;

        case 'switch':
            $.extend(defaultOpts, OPTS.switch);
            break;

        case 'slider':
            $.extend(defaultOpts, OPTS.slider);
            break;

        case 'default':
            $.extend(defaultOpts, OPTS.default);
            break;

        default :
            $.extend(defaultOpts, OPTS.default);
    }

    let opts = UTIL.createComponentOptions($targetObject, defaultOpts, ATTRNAMES, function (k, v){

        return v;

    }, function (k, v){

        let _opts = this;

        //if (!_.isPlainObject(_opts.events)) _opts.events = {};

        if (k === 'class'){
            v = ' ' + v;
        }

        if (k === 'type'){
            _opts[k] = v;
        }

        if (k === 'title'){
            _opts[k] = v;
        }

        return v;
    });

    if (!opts.type) opts.type = 'default';

    // 접근성 처리를 위해 wrapper 에는 적용하지 않도록 attr 에서 삭제한다.
    if (opts.title) delete opts.attr.title;

    return opts;
}

/**
 * 프로퍼티 세팅 함수
 *
 * @param {object} 옵션 객체
 * @param {object} 프로퍼티를 세팅할 엘리먼트
 *
 * @private
 * @memberOf BxButton
 */
function _setElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxButton, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxButton, 'elem', 'type');
    }
    _.map(opts, (v, k) =>{

        // attr 속성은 무시한다.
        if (k === 'attr') return;
        // events 속성은 무시한다.
        if (k === 'events') return;

        // if (k === 'type') return;

        if (k.match(/^bx-/)){
            let _k = k.replace('-', '');
            delete opts[k];
            k = _k;
        }

        v = _checkType(opts, k, v);

        if (
            k === 'colorLevel' ||
            k === 'fill' ||
            k === 'value' ||
            k === 'type' ||
            k === 'action'
        ){
            UTIL.setReadOnlyProperty(elem, k, v);
        }
        else{

            let _k = UTIL.convertInternalPropertyName(k);

            elem[_k] = v;

            UTIL.setPropertyWithDescriptor(elem, k, {
                configurable: true,
                enumerable: true,
                get: () =>{

                    return elem[_k];
                },
                set: (v) =>{

                    switch (k){
                        case 'label' :
                            _setLabel.call(elem, v);
                            elem[_k] = v;
                            break;

                        case 'icon' :
                            _setIcon.call(elem, v);
                            elem[_k] = v;
                            break;

                        case 'defaultValue' :
                            _setChecked.call(elem, v);
                            elem[_k] = v;
                            break;

                        case 'bxrender':
                            if (!_.isFunction(v)){
                                CONSOLE.prop.throw(BxButton, 'bxrender', 'type');
                            }
                            else{
                                elem[_k] = v;
                            }
                            break;
                        case 'bxtoggle':
                            if (!_.isFunction(v)){
                                CONSOLE.prop.throw(BxButton, 'bxToggle', 'type');
                            }
                            else{
                                elem[_k] = v;
                            }
                            break;
                        default :
                            elem[_k] = v;
                    }
                }
            });
        }
    });

    _.map(APIS, (v, k) =>{

        UTIL.setReadOnlyProperty(elem, k, v);
    });

    return this;
}

/**
 * 속성 타입을 재 지정 해준다
 * @param {string} opts - 옵션 객체
 * @param {string} k - 속성명
 * @param {string} v - 속성값
 * @private
 * @memberOf BxButton
 */
function _checkType(opts, k = '', v = ''){

    let type = opts.type;

    if (k === 'colorLevel'){
        v = parseInt(v);
    }
    else if (k === 'fill'){
        k === 'false' ? v = false : v = true;
    }
    else if (type === 'toggle' && (k === 'value' || k === 'label')){
        v = _parseObject(k, v);
    }

    return v;
}


/**
 * object parsing 함수
 * @param {string} attribute name
 * @param {string} attribute value
 * @private
 * @memberOf BxButton
 */
function _parseObject(name = '', string = ''){

    if (!string) return;

    try{
        return UTIL.JSParser(string);
    }
    catch (e){
        CONSOLE.prop.throw(BxButton, name, 'type');
    }
}

module.exports = BxButton;
