// bxuip-angular.js-ui/src/lib/bx-ui/app/components/datepicker.js
/**
 * 날짜 선택 시, 호출될 callback 함수
 *
 * @callback onSelect
 * @param {string} date - 선택된 날짜
 * @param {object} ui - ui 엘리먼트 집합
 *
 * @memberOf BxDatePicker
 */

/**
 * piker close 시, 호출될 callback 함수
 *
 * @callback onClose
 * @param {string} date - 선택된 날짜
 * @param {object} ui - ui 엘리먼트 집합
 *
 * @memberOf BxDatePicker
 */

/**
 * @class BxDatePicker
 * @classDesc BxDatePicker 클래스 생성자
 * @param {object} OPTS
 * @param {string} [OPTS.dateFormat = 'yy/mm/dd']- 날짜 포맷 형식<br/>'y': 두자리 년도, 'm': 1자리 월, 'd': 1자리 일<br/> 'yy': 네자리 년도, 'mm': 2자리 월, 'dd': 2자리 일
 * @param {boolean} [OPTS.changeMonth = true] - "달" select 엘리먼트 사용유무
 * @param {boolean} [OPTS.changeYear = true] - "년" select 엘리먼트 사용유무
 * @param {string} [OPTS.yearSuffix = ' 년']- "년" select 엘리먼트 오른쪽의 문자열을 지정한다.
 * @param {number} [OPTS.yearRange = 5] - "년" select 엘리먼트 범위 값
 * @param {string} [OPTS.minDate = ''] - 선택 가능한 최소 날짜를 설정한다.
 * @param {string} [OPTS.maxDate = ''] - 선택 가능한 최대 날짜를 설정한다.
 * @param {string} [OPTS.defaultDate = ''] - 최초 선택될 날짜를 설정한다.
 * @param {boolean} [OPTS.showOtherMonths = false] - datepicker 실행시 현재 달력에서 다른 달의 출력여부를 설정한다
 * @param {object} [OPTS.attr = {}] - input 엘리먼트의 attr 속성
 * @param {object} [OPTS.disabled = false] - datepicker disabled 설정 유무
 * @param {function} [OPTS.onSelect = function(){}] - 날짜 선택 시, 호출될 callback 함수
 * @param {function} [OPTS.onClose = function(){}] - piker close 시, 호출될 callback 함수
 * @returns {object} 생성된 인스턴스
 *
 * @example
 $('#datepicker').BxDatePicker({
    dateFormat: 'yy/mm/dd',
    changeMonth: true,
    changeYear: true,
    yearSuffix: '년',
    yearRange: 5,
    minDate: '2015/10/01',
    maxDate: '2017/01/05',
    defaultDate: '2016/12/25',
    showOtherMonths: true,
    attr: {
        style: 'width: 100px; height: 50px;'
    },
    disabled: true,
    onSelect : function(date, ui){
        console.log(ui);
    },
    onClose : function(date, ui){
        if (!date){
            alert("날짜를 선택하세요");
        }
    }
}).render();
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

// * @param {object} OPT - bxDatePicker 메뉴를 생성하기 위해 필요한 속성들을 담은 객체
const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

//      kr.* --> OPTS 설정 값들 예시.
// const _lang = {
//     kr : {
//         // 달 문자 설정. 값 없을 시, 영문
//         monthNames : ['1월', '2월', '3월', '4월', '5월', '6월', '7월', '8월', '9월', '10월', '11월', '12월'],
//         monthNamesShort : ['1월', '2월', '3월', '4월', '5월', '6월', '7월', '8월', '9월', '10월', '11월', '12월'],
//         // 일 문자 설정. 값 없을 시, 영문
//         dayNames : ['일', '월', '화', '수', '목', '금', '토'],
//         dayNamesShort : ['일', '월', '화', '수', '목', '금', '토'],
//         dayNamesMin : ['일', '월', '화', '수', '목', '금', '토']
//     }
// };

const ALT_TEXT = {
    'datePickerInput': '날짜 입력',
    'caption': '달력',
    'summary': '달력',
    'selectYear': '년 선택',
    'selectMonth': '월 선택',
    'prev': '이전',
    'next': '다음'
};

const ERR_MSG = {
    'lengthError': '의 length가 올바르지 않습니다.'
};

class BxDatePicker {

    constructor({
        //@private
        showOn = 'focus',
        buttonText = '',
        constrainInput = true,
        dayNames,
        monthNames,
        dateFormat = 'yy/mm/dd',
        changeMonth = true,
        changeYear = true,
        yearSuffix = ' 년',
        yearRange = 5,
        minDate = '',
        maxDate = '',
        defaultDate = '',
        showOtherMonths = false,
        attr = {},
        disabled = false,
        onSelect = function (){},
        onClose = function (){}
    } = {}){

        if (!_.isString(dateFormat)){
            CONSOLE.prop.throw(BxDatePicker, 'dateFormat', 'type');
        }

        if (!_.isBoolean(changeMonth)){
            CONSOLE.prop.throw(BxDatePicker, 'changeMonth', 'type');
        }

        if (!_.isBoolean(changeYear)){
            CONSOLE.prop.throw(BxDatePicker, 'changeYear', 'type');
        }

        if (!_.isString(yearSuffix)){
            CONSOLE.prop.throw(BxDatePicker, 'yearSuffix', 'type');
        }

        if (!_.isNumber(yearRange)){
            CONSOLE.prop.throw(BxDatePicker, 'yearRange', 'type');
        }

        if (!_.isString(minDate)){
            CONSOLE.prop.throw(BxDatePicker, 'minDate', 'type');
        }

        if (!_.isString(maxDate)){
            CONSOLE.prop.throw(BxDatePicker, 'maxDate', 'type');
        }

        if (!_.isString(defaultDate)){
            CONSOLE.prop.throw(BxDatePicker, 'defaultDate', 'type');
        }

        if (!_.isBoolean(showOtherMonths)){
            CONSOLE.prop.throw(BxDatePicker, 'showOtherMonths', 'type');
        }

        if (!_.isPlainObject(attr)){
            CONSOLE.prop.throw(BxDatePicker, 'attr', 'type');
        }

        if (!_.isBoolean(disabled)){
            CONSOLE.prop.throw(BxDatePicker, 'disabled', 'type');
        }

        if (!_.isFunction(onSelect)){
            CONSOLE.prop.throw(BxDatePicker, 'onSelect', 'type');
        }

        if (!_.isFunction(onClose)){
            CONSOLE.prop.throw(BxDatePicker, 'onClose', 'type');
        }

        if (!_.isUndefined(dayNames) && !_.isArray(dayNames)){
            CONSOLE.prop.throw(BxDatePicker, 'dayNames', 'type');
        }

        if (!_.isUndefined(dayNames) && dayNames.length !== 7){
            CONSOLE.prop.throw(BxDatePicker, `dayNames${ERR_MSG.lengthError}`);
        }

        if (!_.isUndefined(monthNames) && !_.isArray(monthNames)){
            CONSOLE.prop.throw(BxDatePicker, 'monthNames', 'type');
        }

        if (!_.isUndefined(monthNames) && monthNames.length !== 12){
            CONSOLE.prop.throw(BxDatePicker, `monthNames${ERR_MSG.lengthError}`);
        }


        this[OPTS] = {showOn, buttonText, constrainInput, dateFormat, changeMonth, changeYear, yearSuffix, defaultDate};

        this.defaultDateString = '';

        if (showOtherMonths){
            this[OPTS].showOtherMonths = true;
            this[OPTS].selectOtherMonths = true;
        }

        if (_.trim(minDate).length){
            this[OPTS].minDate = minDate;
            this[OPTS].hideIfNoPrevNext = true;
        }

        if (_.trim(maxDate).length){
            this[OPTS].maxDate = maxDate;
            this[OPTS].hideIfNoPrevNext = true;
        }

        if (_.trim(defaultDate).length){
            this.defaultDateString = defaultDate;
        }

        if (!_.isEmpty(dayNames)){
            this[OPTS].dayNamesMin = dayNames;
            this[OPTS].dayNamesShort = dayNames;
            this[OPTS].dayNames = dayNames;
        }

        if (!_.isEmpty(monthNames)){
            this[OPTS].monthNames = monthNames;
            this[OPTS].monthNamesShort = monthNames;
        }

        this.$root = BxDatePicker.$root;
        this.attr = attr;
        this.disabled = disabled;

        this.yearRange = yearRange;

        this.onSelect = onSelect;
        this.onClose = onClose;


    }

    /**
     * 렌더 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    render(){

        let opts = this[OPTS];

        // Render 시점에서 Container 생성 후 Root에 Append
        let inputTitle = ALT_TEXT.datePickerInput;

        this.$container = BASE.appendContainerElement(this.$root, BxDatePicker.className, $(`<input type="text" class="bw-input ipt-cal" title="` + inputTitle + `"/>`));

        // 연도 범위 값 설정.
        opts.yearRange = _calcYearRange(this.yearRange);

        // add 이벤트
        opts.onSelect = (e, ui) =>{
            this.onSelect.apply(this, [e, ui]);
        };

        opts.onClose = (e, ui) =>{
            this.onClose.apply(this, [e, ui]);
        };

        this.$container.datepicker(opts);

        // 기본 날짜를 설정한다.
        this.setDate(this.defaultDateString);

        // 엘리먼트에 전달받은 attr 객체 속성을 추가한다.
        UTIL.attr(this.$container, this.attr);

        if (this.disabled){
            this.setDisable();
        }

        // 접근성 요소 추가
        let $inputElem = this.$root.find('.-bx-ui-date-picker');

        $inputElem.off("focus");

        $inputElem.click(() =>{

            this.show();
            _setAccessibility.call(this);

        });

        //달력박스에 focus 가능하도록 jquery keyboard이벤트 일부 삭제
        this.$container.off('keydown');

        return this;
    }

    /**
     * 날짜 반환 함수
     * @returns {object} date 정보를 담은 객체 <br/>
     *
     * @example
     * {
     *      value: '',      // [string] input text   , default: ''
     *      year: 0,        // [number] year {yyyy}  , default: 0
     *      month: 0        // [number] month        , default: 0
     *      date: 0,        // [number] date         , default: 0
     *      ms: 0           // [number] millisecond  , default: 0
     * }
     */
    getDate(){

        let $container = this.$container;

        let result = {
            value: $container.val(),
            year: 0,
            month: 0,
            date: 0,
            ms: 0
        };

        if (!_.isEqual(result.value, "")){

            let date = new Date($container.datepicker('getDate'));

            result.year = date.getFullYear();
            result.month = date.getMonth() + 1;
            result.date = date.getDate();
            result.ms = date.getTime();

        }

        return result;
    }

    /**
     * 날짜 지정 함수
     * @param {string} [date = ''] - 날짜 문자열
     * @returns {object} 생성된 인스턴스
     */
    setDate(date = ''){

        if (!_.isString(date)){
            CONSOLE.prop.throw(BxDatePicker, 'date', 'type');
        }

        date = date.trim() === '' ? null : date;

        this.$container.datepicker('setDate', date);

        _setAccessibility.call(this);

        return this;
    }

    /**
     * hide 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    hide(){

        this.$container.datepicker('hide');

        return this;
    }

    /**
     * show 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    show(){

        this.$container.datepicker('show');

        _setAccessibility.call(this);

        return this;
    }

    /**
     * input 엘리먼트를 `disabled` 시킨다.
     *
     * @returns {object} 생성된 인스턴스
     */
    setDisable(){

        this.$container.attr('disabled', true);

        return this;
    }

    /**
     * input 엘리먼트를 `enabled` 시킨다.
     *
     * @returns {object} 생성된 인스턴스
     */
    setEnable(){

        this.$container.attr('disabled', false);

        return this;
    }
}

/**
 * yearRange 계산 함수
 *
 * @param {number} yearRange - 년도 범위 값
 * @returns {string}
 *
 * @private
 * @memberOf BxDatePicker
 */
function _calcYearRange(yearRange){

    let currentYear = new Date().getFullYear();

    return (currentYear - yearRange) + ':' + (currentYear + yearRange);

}


/**
 * 접근성 관련 함수를 호출
 * @returns {object} 생성된 인스턴스
 * @private
 * @memberOf BxDatePicker
 */
function _setAccessibility(){

    let $datepicker = $('#ui-datepicker-div');

    _setAccessibilityTable.call(this, $datepicker);

    _setAccessibilityTableHeader.call(this, $datepicker);

    return this;
}

/**
 * datepicker의 table에 접근성 필요 요소를 추가
 * @param {object} [$datepicker=null] - datepicker 달력 박스 객체
 * @returns {object} 생성된 인스턴스
 * @private
 * @memberOf BxDatePicker
 */
function _setAccessibilityTable($datepicker = null){

    if (!_.isElement($datepicker.get(0))){
        CONSOLE.prop.throw(BxDatePicker, '$datepicker', 'type');
    }

    let $table = $datepicker.find('.ui-datepicker-calendar');

    $table.find('caption').remove();
    $table.attr({summary: ALT_TEXT.summary});
    $table.prepend('<caption>' + ALT_TEXT.caption + '</caption>');

    return this;
}

/**
 * datepicker의 table title에 접근성 필요 요소를 추가
 * @param {object} [$datepicker=null] - datepicker 달력 박스 객체
 * @returns {object} 생성된 인스턴스
 * @private
 * @memberOf BxDatePicker
 */
function _setAccessibilityTableHeader($datepicker = null){

    if (!_.isElement($datepicker.get(0))){
        CONSOLE.prop.throw(BxDatePicker, '$datepicker', 'type');
    }

    let $tableHeader = $datepicker.find('.ui-datepicker-header');

    $tableHeader.find('.ui-datepicker-year').attr({'title': ALT_TEXT.selectYear});
    $tableHeader.find('.ui-datepicker-month').attr({'title': ALT_TEXT.selectMonth});
    $tableHeader.find('.ui-datepicker-prev').attr({'title': ALT_TEXT.prev, 'href': '#', 'onclick': 'return false'});
    $tableHeader.find('.ui-datepicker-next').attr({'title': ALT_TEXT.next, 'href': '#', 'onclick': 'return false'});
    $tableHeader.find('.ui-datepicker-prev span').html(ALT_TEXT.prev);
    $tableHeader.find('.ui-datepicker-next span').html(ALT_TEXT.next);

    return this;
}

module.exports = BxDatePicker;
