// bxuip-angular.js-ui/src/lib/bx-ui/app/components/dialog.js
/**
 * @class BxDialog
 * @classDesc BxDialog 클래스 생성자
 *
 * @param {object} OPTS
 * @param {string} [OPTS.type = 'default'] - dialog 모양 ( default, stick, edge, folder, pure, square)
 * @param {int} [OPTS.width = 400] - dialog 가로 길이
 * @param {string} [OPTS.confirmText = '확인'] - 확인 버튼 문구
 * @param {string} [OPTS.closeText = '취소'] - 취소 버튼 문구
 * @param {boolean} [OPTS.invert = false] - 취소, 확인 버튼 위치 변경 유무
 * @param {object} OPTS.dimStyle
 * @param {string} [OPTS.dimStyle.color = 'black'] - dim 색깔 ( black, white, clear, theme )
 * @param {boolean} [OPTS.dimStyle.enableBlur = false] - dim 흐림 처리 유무 ( true, false )
 * @param {boolean} [OPTS.isDim] - <s>dim 출력 유무</s><span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @returns {object} 생성된 인스턴스
 *
 * @example
 // alert
 $().BxDialog({
    type : 'stick',
    width : 0,
    confirmText : '확인',
    closeText : '취소',
    dimStyle : {
        color : 'black',
        enableBlur : false
    }
 }).alert({
    title: 'title',
    content: 'content',
    onConfirm: function(e){
        alert('확인');
        this.hide();
    }
});

 //confirm
 $().BxDialog({
    type : 'stick',
    width : 0,
    confirmText : '확인',
    closeText : '취소',
    dimStyle : {
        color : 'black',
        enableBlur : false
    }
 }).confirm({
    title: 'title',
    content: 'content',
    onConfirm: function(e){
        alert('확인');
        this.hide();
    },
    onClose: function(e){
        console.log("닫기");
    }
});

 //say
 $().BxDialog({
    type : 'stick',
    width : 0,
    dimStyle : {
        color : 'black',
        enableBlur : false
    }
 }).say({
    title: 'title',
    content: 'content',
});

 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

const ALT_TEXT = {
    'close': '닫기'
};

const DIALOG_TYPE = {
    default: 'default',
    stick: 'stick',
    edge: 'edge',
    folder: 'folder',
    pure: 'pure',
    square: 'square'
};

const SELECTOR = {
    wrap: 'dialog-wrap',
    close: 'bx-btn-close',
    confirm: 'confirm',
    cancel: 'cancel',
    action: 'dialog-action'
};

class BxDialog {

    constructor({
        width = 0,
        type = '',
        invert = false,
        confirmText = '',
        closeText = '',
        isDim = null, // jshint ignore:line
        dimStyle = {}
    } = {}){

        let $root = BxDialog.$root;

        // $root 엘리먼트의 nodeName 이 BODY 가 아닐경우, body 엘리먼트를 $root 변수에 할당시킨다.
        if ($root.prop('nodeName') !== 'BODY') {
            console.warn('$root 엘리먼트가 body 엘리먼트가 아닙니다.');
            $root = $(document.body);
        }

        if (!_.isObject(dimStyle)){
            dimStyle = {};
        }

        if (!_.isNull(isDim)){
            CONSOLE.prop.throw(BxDialog, 'isDim', 'deprecatedProperty');
        }

        if (!_.isString(type)){
            CONSOLE.prop.throw(BxDialog, 'type', 'all');
        }
        else if (_.trim(type).length === 0){
            type = '';
        }

        if (!_.isString(confirmText)){
            CONSOLE.prop.throw(BxDialog, 'confirmText', 'type');
        }
        else if (_.trim(confirmText).length === 0){
            confirmText = '확인';
        }

        if (!_.isString(closeText)){
            CONSOLE.prop.throw(BxDialog, 'closeText', 'type');
        }
        else if (_.trim(closeText).length === 0){
            closeText = '취소';
        }

        if (!_.isNumber(width) || isNaN(width)){
            CONSOLE.prop.throw(BxDialog, 'width', 'all');
        }
        else if (width < 400){
            width = 400;
        }

        if (!_.isBoolean(invert)){
            CONSOLE.prop.warn(BxDialog, 'invert', 'type');
            invert = false;
        }

        Object.assign(this, {
            $root,
            $container: null,
            $dialogWrap: null
        });

        this[OPTS] = Object.assign({}, {
            width, type, invert, confirmText, closeText, dimStyle
        });

        return this;
    }

    /**
     * 버튼이 한개인 dialog를 출력한다
     *
     * @param {string} [title = ''] - 제목
     * @param {string} [content = ''] - 내용
     * @param {function}[onConfirm=function(){}] - 확인 버튼 클릭 시, 호출될 callback 함수
     * @returns {object} 생성된 인스턴스
     */
    alert({
        title = '',
        content = '',
        onConfirm = function(){}
    } = {}){

        if (!_.isString(title)){
            CONSOLE.prop.throw(BxDialog, 'title', 'type');
        }

        if (!_.isString(content) || _.trim(content).length === 0) {
            CONSOLE.prop.throw(BxDialog, 'content','value');
        }

        _renderDOM.call(this);

        let { $root, $container } = this;

        _.trim(title).length === 0 ? $container.find('.dialog-header').remove() : $container.find('.bx-tt').html(title);
        $container.find('.bx-desc p').html(content);

        let {confirm: $confirm, cancel: $cancel, close: $close, wrap: $wrap} = UTIL.getSelectorInstance($container, SELECTOR);

        _setWrapVerticalPosition({$root, $wrap});

        if (DIALOG_TYPE[this[OPTS].type] !== DIALOG_TYPE.pure && DIALOG_TYPE[this[OPTS].type] !== DIALOG_TYPE.default) {
            $confirm.parent().addClass('btn-single');
        }

        $confirm.addClass('on');

        $cancel.hide();

        _setButtonEvent.call(this, {$confirm, $close}, {onConfirm});

        return this;
    }

    /**
     * 버튼이 두개인 dialog를 출력한다
     *
     * @param {string} [title = ''] - 제목
     * @param {string} [content = '']- 내용
     * @param {function} [onConfirm = function(){}] - 확인 버튼 클릭 시, 호출될 callback 함수
     * @param {function} [onClose = function(){}] - 닫기 버튼 클릭 시, 호출될 callback 함수
     * @returns {object} 생성된 인스턴스
     */
    confirm({
        title = '',
        content = '',
        onConfirm = function(){},
        onClose = function(){}
    } = {}){

        if (!_.isString(title)){
            CONSOLE.prop.throw(BxDialog, 'title', 'type');
        }

        if (!_.isString(content) || _.trim(content).length === 0) {
            CONSOLE.prop.throw(BxDialog, 'content','value');
        }

        _renderDOM.call(this);

        let { $root, $container } = this;

        _.trim(title).length === 0 ? $container.find('.dialog-header').remove() : $container.find('.bx-tt').html(title);
        $container.find('.bx-desc p').html(content);

        let {confirm: $confirm, cancel: $cancel, close: $close, wrap: $wrap} = UTIL.getSelectorInstance($container, SELECTOR);

        _setWrapVerticalPosition({$root, $wrap});

        _setButtonEvent.call(this, {$confirm, $cancel, $close}, {onConfirm, onClose});

        return this;
    }

    /**
     * 버튼이 없는 dialog를 출력한다
     *
     * @param {string} [title = ''] - 제목
     * @param {string} [content = ''] - 내용
     * @returns {object} 생성된 인스턴스
     */
    say({
        title = '',
        content = ''
    } = {}){

        if (!_.isString(title)){
            CONSOLE.prop.throw(BxDialog, 'title', 'type');
        }

        if (!_.isString(content) || _.trim(content).length === 0) {
            CONSOLE.prop.throw(BxDialog, 'content','value');
        }

        _renderDOM.call(this);

        let { $root, $container } = this;

        _.trim(title).length === 0 ? $container.find('.dialog-header').remove() : $container.find('.bx-tt').html(title);

        $container.find('.bx-desc p').html(content);

        let {action: $action, close: $close, wrap: $wrap}= UTIL.getSelectorInstance($container, SELECTOR);

        $action.remove();

        _setWrapVerticalPosition({$root, $wrap});

        _setButtonEvent.call(this, {$close});

        return this;
    }

    /**
     * BxDialog 숨김 함수
     * @returns {object} 생성된 인스턴스
     */
    hide(){

        this.$container.remove();
        this.dim && this.dim.hide();

        //dialog hide 할 때, focus 가져올 곳을 base로 설정.
        UTIL.setPreFocusInstanceKey('base');
        let $target = UTIL.getPreFocusInstance();

        if ($target) {
            window.setTimeout(() => {
                $target.focus();
            }, 0);
        }

        return this;
    }
}

/**
 *
 * DOM render 함수
 *
 * @private
 * @memberOf BxDialog
 */
function _renderDOM(){

    if (this.$container) this.hide();

    this.dim = $().bxDim(this[OPTS].dimStyle).show();

    this.$container = BASE.appendContainerElement(this.$root, BxDialog.className);

    let html = _setDialogDesign(this[OPTS]);

    this.$container.append(html);

    this.$dialogWrap = this.$container.find('.dialog-wrap');
    this.$dialogWrap.find('.dialog-focus').focus();

    this.$dialogWrap.on('focusin', function(e){
        UTIL.setPreFocusInstanceKey('hoverer');
    });

    window.setTimeout(()=>{
        UTIL.setPreFocusInstanceKey('hoverer');
    }, 0);
}

/**
 *
 * BxDialog 타입별 디자인 설정 함수
 * @param {object} options
 * @param {int} [options.width = 400] - dialog 가로 길이
 * @param {string} [options.type = 'default'] - dialog 디자인 타입
 * @param {boolean} [options.invert = false] - dialog button 위치 변경
 * @param {string} [options.confirmText = '확인'] - dialog 확인 버튼 문구
 * @param {string} [options.closeText = '취소'] - dialog 취소 버튼 문구
 *
 * @returns {string} 생성된 다이얼로그 디자인 html
 *
 * @private
 * @memberOf BxDialog
 */
function _setDialogDesign(options){
    let type = options.type;
    let width = options.width;
    let confirmText = options.confirmText;
    let closeText = options.closeText;
    let isInvert = options.invert;
    let html = [];
    let themeColor = "";

    switch (DIALOG_TYPE[options.type]) {
        case DIALOG_TYPE.stick : type = 'd-type2'; break;
        case DIALOG_TYPE.edge : type = 'd-type3'; break;
        case DIALOG_TYPE.folder : type = 'd-type5'; break;
        case DIALOG_TYPE.pure : type = 'd-type7'; break;
        case DIALOG_TYPE.square : type = 'd-type8'; break;
        default : type = 'd-type6';
    }

    if (DIALOG_TYPE[options.type] === DIALOG_TYPE.pure && options.dimStyle.color !== 'black') {
        themeColor = "theme-color";
    }

    html.push(`<div class="dialog-wrap ${type} ${themeColor}" style="width:${width}px;margin-left:-${width/2}px;">`);
    html.push(`<a href="#" class="dialog-focus" onclick="return false;"><span class="blind">focus</span></a>`);
    html.push('<div class="dialog-header">');
    html.push('<h2 class="bx-tt"></h2>');
    html.push('</div>');
    html.push('<div class="dialog-content">');
    html.push('<div class="bx-desc"><p></p></div>');
    html.push('</div>');
    html.push('<div class="dialog-action">');
    html.push('<div class="btn-wrap">');

    if (isInvert === true) {
        html.push(`<button type="button" class="bx-btn-txt cancel" title="${closeText}">${closeText}</button>`);
        html.push(`<button type="button" class="bx-btn-txt confirm on" title="${confirmText}">${confirmText}</button>`);
    } else {
        html.push(`<button type="button" class="bx-btn-txt confirm" title="${confirmText}">${confirmText}</button>`);
        html.push(`<button type="button" class="bx-btn-txt cancel on" title="${closeText}">${closeText}</button>`);
    }

    html.push('</div>');
    html.push('</div>');

    if (DIALOG_TYPE[options.type] === DIALOG_TYPE.pure) {
        html.push('</div><button type="button" class="bx-btn-close" title="'+ALT_TEXT.close+'"><i class="bw-icon i-70 i-close"><span class="blind">'+ALT_TEXT.close+'</span></i></button>');
    } else {
        html.push('<button type="button" class="bx-btn-close" title="'+ALT_TEXT.close+'"><i class="bw-icon i-30 i-close"><span class="blind">'+ALT_TEXT.close+'</span></i></button></div>');
    }

    return html.join('');
}

/**
 * @param {{}}[instance={}] instance 객체
 * @param {{}}[$root=null] BxDialog root jquery instance
 * @param {{}}[$wrap=null] BxDialog wrap jquery instance
 * @private
 */
function _setWrapVerticalPosition({$root = null, $wrap = null} = {}){

    let rootHalfHeight = $root.height() / 2,
        wrapHalfHeight = $wrap.outerHeight() / 2,
        top = rootHalfHeight - wrapHalfHeight;

    UTIL.setPosition($wrap, {top: top});
}

/**
 * 버튼 이벤트 바인딩 함수
 * @param {{}}[instance] instance 객체
 * @param {{}}[instance.$confirm=null] 확인 버튼 jquery instance
 * @param {{}}[instance.$cancel=null] 취소 버튼 jquery instance
 * @param {{}}[instance.$close=null] x 버튼 jquery instance
 * @param {{}}[callback] callback 객체
 * @param {function}[callback.onConfirm=function(){}] 확인 버튼 click callback 함수
 * @param {function}[callback.onClose=function(){}] 취소 버튼 click callback 함수
 * @private
 */
function _setButtonEvent({$confirm = null, $cancel = null, $close = null} = {}, {onConfirm = function(){}, onClose = function(){}} = {}){

    _setConfirmButtonEvent.call(this, {$confirm, onConfirm});

    _setCancelButtonEvent.call(this, {$cancel, onClose});

    _setCloseButtonEvent.call(this, {$close});
}

/**
 * 확인 버튼 click event 바인딩 함수
 * @param $confirm
 * @param onConfirm
 * @private
 */
function _setConfirmButtonEvent({$confirm = null, onConfirm = function(){}} = {}){

    $confirm && $confirm.click((e) => {
        onConfirm.call(this, e);
        this.hide();
    });
}
/**
 * 취소 버튼 click event 바인딩 함수
 * @param $cancel
 * @param onClose
 * @private
 */
function _setCancelButtonEvent({$cancel= null, onClose = function(){}} = {}){

    $cancel && $cancel.click((e) => {
        onClose.call(this, e);
        this.hide();
    });
}

/**
 * 닫기 버튼 click event 바인딩 함수
 * @param $close
 * @private
 */
function _setCloseButtonEvent({$close = null} = {}){

    $close && $close.click((e) => {
        this.hide();
    });
}


module.exports = BxDialog;
