// bxuip-angular.js-ui/src/lib/bx-ui/app/components/dim.js
/**
 * @class BxDim
 * @classDesc BxDim 클래스 생성자
 *
 * @param {object} OPTS
 * @param {string} [OPTS.color = 'theme'] - 바탕 색상 <br/> <span class='dec'>['black', 'theme', 'clear']</span>
 * @param {boolean} [OPTS.enableBlur = false] - 블러 효과 사용 유무 <br/> <span class='dec'>(Edge를 제외한 IE 계열 브라우저 미지원)</span>
 * @param {boolean} [OPTS.isIcon = false] - <s>로딩 아이콘 사용 유무</s><span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @param {boolean} [OPTS.enableIcon = false] - 로딩 아이콘 사용 유무
 * @param {string} [OPTS.iconType = 'default'] - 로딩 아이콘 타입 <br/> <span class='dec'>['default', 'spin', 'juggling', 'square', 'cube', 'dribble', 'think', 'equalizer']</span>
 * @param {string} [OPTS.text = ''] - 로딩 텍스트
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 * // show
 * var dim = $('#id').BxDim({
 *      color: 'black',
 *      enableBlur: true,
 *      enableIcon: true,
 *      iconType: 'spin',
 *      text: '로딩중입니다. 잠시만 기다려주세요.'
 * }).show();
 *
 * // hide
 * dim.hide();
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

const CLASS_NAME = {
    wrap: "bx-dim",
    innerWrap: "bx-dim-inner",
    blur: "-bx-ui-dim-style2",
    text: "bx-message",
    dimContainer: '-bx-ui-dim-container',
    noneRoot: 'none-root'
};

const ICON_TYPE = {
    default: "circle",
    spin: "spin",
    juggling: "juggling",
    square: "square",
    cube: "cube",
    dribble: "dribble",
    equalizer: "equalizer",
    think: "think",
    ie9: "ie9"
};

const ICON_CLASS_REPLACE_TOKEN = '${count}';
const ICON_CLASS_NAME = {
    circle: {
        className: `bx-circle${ICON_CLASS_REPLACE_TOKEN} bx-circle`,
        count: 12
    },
    spin: {
        className: `bx-circle${ICON_CLASS_REPLACE_TOKEN} bx-child`,
        count: 12
    },
    juggling: {
        className: `dot${ICON_CLASS_REPLACE_TOKEN}`,
        count: 2
    },
    square: {
        className: ``,
        count: 0
    },
    cube: {
        className: `cube${ICON_CLASS_REPLACE_TOKEN}`,
        count: 2
    },
    equalizer: {
        className: `rect${ICON_CLASS_REPLACE_TOKEN}`,
        count: 5
    },
    dribble: {
        className: `double-bounce${ICON_CLASS_REPLACE_TOKEN}`,
        count: 2
    },
    think: {
        className: `bounce${ICON_CLASS_REPLACE_TOKEN}`,
        count: 3
    }
};

const DIM_COLOR = {
    supportColor: ["black", "theme", "clear"],
    black: "dim-color1",
    theme: "dim-color2",
    clear: ""
};

class BxDim {

    constructor({
        color = 'theme',
        enableBlur = false,
        enableIcon = false,
        // @deprecated 
        isIcon = null,
        iconType = '',
        text = ''
    } = {}){

        if (!_.isString(color)){
            CONSOLE.prop.throw(BxDim, 'color', 'type');
        }

        if (!_.isBoolean(enableBlur)){
            CONSOLE.prop.throw(BxDim, 'enableBlur', 'type');
        }

        if (!_.isBoolean(enableIcon)){
            CONSOLE.prop.throw(BxDim, 'enableIcon', 'type');
        }

        // @deprecated 
        if (!_.isNull(isIcon)){
            CONSOLE.prop.throw(BxDim, 'isIcon', 'deprecatedProperty');
        }

        if (!_.isString(iconType)){
            CONSOLE.prop.throw(BxDim, 'iconType', 'type');
        }

        if (!_.isString(text)){
            CONSOLE.prop.throw(BxDim, 'text', 'type');
        }

        this[OPTS] = {color, enableBlur, enableIcon, iconType, text: _.trim(text)};

        let $root = BxDim.$root;

        this.$root = $root;
        this.$container = null;

        this.isBody = _isBody.call(this);
    }

    /**
     * BxDim 출력 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    show(){

        const { $root , $container, isBody } = this;

        if ($container) $container.remove();

        this.$container = BASE.appendContainerElement($root, BxDim.className);

        let options = this[OPTS];

        // enableBlur 옵션은, edge 를 제외한 IE 계열 브라우저에서만 지원한다.
        if (options.enableBlur && !_isUnderEdge()) _enableBlur.call(this);

        _setInnerElement.call(this);

        _setDimInnerElementPosition.call(this);

        $root.addClass(CLASS_NAME.dimContainer);

        if (isBody){
            _setBodyDimPosition.call(this);
        }
        else{
            _setContainerDimPosition.call(this);
            $root.addClass(CLASS_NAME.noneRoot);
        }

        return this;
    }

    /**
     * BxDim 숨김 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    hide(){

        const { $root, $container, isBody } = this;

        if (_.isNull($container)) return this;

        if (this[OPTS].enableBlur) _disableBlur.call(this);

        $root.removeClass(CLASS_NAME.dimContainer);
        if (!isBody){
            $root.removeClass(CLASS_NAME.noneRoot);
        }
        // target 엘리먼트
        $container.remove();

        return this;
    }

    /**
     * @deprecated Deprecated in: v3.1.0, Deprecated since: v3.2.1
     *
     * @returns {object} 생성된 인스턴스
     */
    resize(){}
}

/**
 * root의 Body 여부 반환 함수
 *
 * @private
 * @memberOf BxDim
 */
function _isBody(){
    return this.$root.get(0).tagName === 'BODY' ? true : false;
}

/**
 * 브라우저가 IE edge 이하인지 여부 반환 함수
 * @private
 */
function _isUnderEdge(){
    return $.browser.msie;
}

/**
 * blur 효과를 세팅한다.
 *
 * edge 를 제외한 IE 계열 브라우저에서는 제공하지 않는 함수
 *
 * @private
 * @memberOf BxDim
 */
function _enableBlur(){
    // edge 를 제외한 IE 계열 브라우저에서는 blur filter 가 제대로 지원되지 않는다.
    // http://stackoverflow.com/questions/24736594/filterblur-svg-in-opera-and-ie-window-linux-not-working
    this.$root.addClass(`${CLASS_NAME.blur}`);
}

/**
 * dim size를 세팅한다.
 *
 * @private
 * @memberOf BxDim
 */
function _setInnerElement(){
    let options = this[OPTS];

    let html = _getInnerElement.call(this, options);

    this.$container.append(html);
}

/**
 * icon, text를 HTML tag로 반환한다.
 *
 * @param {object} options
 *
 * @returns {string} - html 문자열
 * @private
 * @memberOf BxDim
 */
function _getInnerElement(options = {
    color: 'theme',
    enableBlur: false,
    enableIcon: false,
    iconType: '',
    text: ''
}){

    let html = [];

    html.push(`<div class="${CLASS_NAME.wrap} ${_getDimColorClass(options.color)}">`);

    if (options.enableIcon || options.text){

        html.push(`<div class="${CLASS_NAME.innerWrap}">`);

        if (options.enableIcon) html.push(_setIcon(options));

        if (options.text !== '') html.push(_getText(options));

        html.push('</div>');

    }

    html.push('</div>');

    return html.join('');

}

/**
 * dim color class name을 반환한다.
 *
 * @param {string} color
 * @return {string} class name
 * @private
 * @memberOf BxDim
 */
function _getDimColorClass(color = 'theme'){

    const { supportColor, theme } = DIM_COLOR;
    const colorIdx = supportColor.indexOf(color);
    let dimColor = theme;

    if (colorIdx !== -1) {
        const color = supportColor[colorIdx];
        dimColor = DIM_COLOR[color];
    }

    return dimColor;
}

/**
 * icon을 세팅한다.
 *
 * @param {object} options
 * @private
 * @memberOf BxDim
 */
function _setIcon(options = {
    color: 'theme',
    enableBlur: false,
    enableIcon: false,
    iconType: '',
    text: ''
}){

    if (UTIL.isIE9()) return _getDefaultTypeIconForIE9(options);

    let html = [];

    html.push(_getFirstElementOfIcon(options));

    let iconType = ICON_TYPE[options.iconType];

    if (!iconType) iconType = ICON_TYPE.default;

    html.push(_getIcon(iconType));

    html.push('</div>');

    return html.join('');
}

/**
 * IE 9 전용 이미지 아이콘 반환
 *
 * @param {object} options
 * @return {string}
 * @private
 * @memberOf BxDim
 */
function _getDefaultTypeIconForIE9(options){

    options.iconType = 'ie9';

    let html = [];

    html.push(_getFirstElementOfIcon(options));

    html.push('<div class="ie-loading-img"/>');

    html.push('</div>');

    return html.join('');

}

/**
 * Icon의 첫번째 HTML 반환
 *
 * @param {object} options
 * @return {string}
 * @private
 * @memberOf BxDim
 */
function _getFirstElementOfIcon({
   color = 'theme',
   enableBlur = false,
   enableIcon = false,
   iconType = '',
   text = ''
} = {}){
    let iconClass = 'bx-loading-icon bx-loading-type' + _getIconTypeNum(iconType);

    if (text) iconClass += ' with-text';

    if (color !== 'black') iconClass += ' theme-color';

    return '<div class="' + iconClass + '">';
}

/**
 * icon type name으로 type number를 반환
 *
 * @param {string} iconType
 * @return {number} icon type number
 * @private
 * @memberOf BxDim
 */
function _getIconTypeNum(iconType = 'default'){
    switch (iconType){
        case ICON_TYPE.spin:
            return 2;
        case ICON_TYPE.juggling:
            return 3;
        case ICON_TYPE.square:
            return 4;
        case ICON_TYPE.cube:
            return 5;
        case ICON_TYPE.equalizer:
            return 6;
        case ICON_TYPE.dribble:
            return 7;
        case ICON_TYPE.think:
            return 8;
        case ICON_TYPE.ie9:
            return 9;
        default:
            return 1;
    }
}

/**
 * 타입별 아이콘 생성 함수
 * @param type
 * @returns {string}
 * @private
 */
function _getIcon(type = 'default') {
    const { className, count } = ICON_CLASS_NAME[type];
    let html = new Array(count);

    return _.map(html, (v, i)=> {
       return `<div class="${className.replace(ICON_CLASS_REPLACE_TOKEN, i + 1)}"></div>`;
    }).join('');
}

/**
 * 로딩 텍스트를 HTML tag로 반환한다.
 *
 * @param {object} options
 * @return {string}
 * @private
 * @memberOf BxDim
 */
function _getText({
                      color = 'theme',
                      enableBlur = false,
                      enableIcon = false,
                      iconType = '',
                      text = ''
                  } = {}){

    let textClass = 'bx-message';

    if (!enableIcon) textClass += ' message-type2';

    if (color !== 'black') textClass += ' theme-color';

    return '<div class="' + textClass + '"><p>' + text + '</p></div>';
}

/**
 * icon, text의 위치를 세팅한다.
 *
 * @private
 * @memberOf BxDim
 */
function _setDimInnerElementPosition(){

    let $inner = this.$container.find(`.${CLASS_NAME.innerWrap}`);

    $inner.css({
        'margin-top': -1 * ($inner.height() / 2)
    });
}

/**
 * root가 Body인 Dim의 position 세팅 함수
 *
 * @private
 * @memberOf BxDim
 */
function _setBodyDimPosition(){
    this.$container.find(`.${CLASS_NAME.wrap}`).css('position', 'fixed');
}

/**
 * root가 Container인 Dim의 position 세팅 함수
 *
 * @private
 * @memberOf BxDim
 */
function _setContainerDimPosition(){
    let $root = this.$root;

    // scroll된 경우를 고려하여, scroll된 만큼의 위치로 조정한다.
    this.$container.find(`.${CLASS_NAME.wrap}`).css({
        top: $root.scrollTop(),
        left: $root.scrollLeft()
    });
}

/**
 * blur 효과를 헤제한다.
 *
 * edge 를 제외한 IE 계열 브라우저에서는 제공하지 않는 함수
 *
 * @private
 * @memberOf BxDim
 */
function _disableBlur(){
    this.$root.removeClass(CLASS_NAME.blur);
}

module.exports = BxDim;
