// bxuip-angular.js-ui/src/lib/bx-ui/app/components/dropDown.js
/**
 * 선택 이벤트 callback 함수
 *
 * @callback onActivate
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf bxPaper
 * */

/*
 * 드래그 이벤트 callback 함수
 *
 * @callback onDrag
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf bxPaper
 * */

/*
 * 드래그 시작 이벤트 callback 함수
 *
 * @callback onDragStart
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf bxPaper
 * */

/*
 * 드래그 종료 이벤트 callback 함수
 *
 * @callback onDragStop
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf bxPaper
 * */

/*
 * 닫기 이벤트 callback 함수
 *
 * @callback onClosed
 * @memberOf bxPaper
 * */

/*
 * 렌더 종료 이벤트 callback 함수
 *
 * @callback onRender
 * @memberOf bxPaper
 * */

/*
 * @class bxPaper
 * @classDesc bxPaper 클래스 생성자
 * @param {object} OPTS
 * @param {int} [OPTS.width=400] bxPaper content 영역 넓이
 * @param {int} [OPTS.height=300] bxPaper content 영역 높이
 * @param {object} [OPTS.position={x:0, y:0}] 초기 렌더링 시, 화면 정 가운데를 기준으로 위치를 이동시킬 x, y 좌표
 * @param {string} [OPTS.color='white'] bxPaper 의 색상 <br/> <span class='dec'>['white', 'yellow', 'theme']</span>
 * @param {string} [OPTS.type='flat'] bxPaper 의 타입 <br/> <span class='dec'>['flat', 'round', 'line', 'stitch']</span>
 * @param {string} [OPTS.shadow='none'] bxPaper 의 그림자 <br/> <span class='dec'>['none', 'left', 'right', 'both', 'bottom']</span>
 * @param {string} [OPTS.title=''] bxPaper 의 타이틀
 * @param {string} [OPTS.titleAlign='left'] bxPaper 의 타이틀 정렬 <br/> <span class='dec'>['left', 'center']</span>
 * @param {string} [OPTS.template=''] bxPaper 의 내용
 * @param {string} [OPTS.templateUrl=''] template 로 사용할 파일의 절대 경로 <br/><span class='dec'> template 와 동시에 쓰이는 경우, templateUrl 의 우선순위가 더 높게 적용. </span>
 * @param {object} [OPTS.dimStyle={}] paper 사용 시, dim 에 대한 설정
 * @param {boolean} [OPTS.enableDrag=true] drag 사용 유무
 * @param {function} [OPTS.onActivate=function(e, $elem){}] 선택 이벤트 callback 함수
 * @param {function} [OPTS.onDrag=function(e, $elem){}] 드래그 이벤트 callback 함수
 * @param {function} [OPTS.onDragStart=function(e, $elem){}] 드래그 시작 이벤트 callback 함수
 * @param {function} [OPTS.onDragStop=function(e, $elem){}] 드래그 종료 이벤트 callback 함수
 * @param {function} [OPTS.onClosed=function(){}] 닫기 이벤트 callback 함수
 * @param {function} [OPTS.onRender=function(){}] 렌더 종료 이벤트 callback 함수
 * @example

 // 1. template 예제

 var template = [];

 template.push('<div>template 예제 입니다.</div>');
 template.push('<div> template 에 대한 내용을 입력해 주세요.</div>');
 template = template.join('');

 $().bxPaper(
 {
 width:500,
 height: 500,
 position: {
 x:0,
 y:0
 },
 color: 'theme',
 type: 'stitch',
 shadow: 'both',
 title:'template 예제 #1',
 titleAlign: 'center',
 template: template,
 dimStyle: {
 color: 'theme' // ref bxDim Document
 },
 enableDrag: true,
 onActivate: function(e, $elem) {
 console.log('onActivate', e, $elem);
 },
 onDrag: function (e, $elem){
 console.log('onDrag', e, $elem);
 },
 onDragStart: function (e, $elem){
 console.log('onDragStart', e, $elem);
 },
 onDragStop: function(e, $elem){
 console.log('onDragStop', e, $elem);
 },
 onClosed: function () {
 console.log('bxPaper - onClose');
 },
 onRender: function () {
 console.log('bxPaper - onRender')
 }
 }
 ).render();

 ----------------------------------------------------------------------

 // 2. templateUrl 예제
 $().bxPaper(
 {
 width: 500,
 height: 500,
 position: {
 x:0,
 y:0
 },
 color: 'theme',
 type: 'stitch',
 shadow: 'both',
 title:'templateUrl 예제 #1',
 titleAlign: 'center',
 templateUrl: './app/views/onlineAdmin/log/test.tpl.html',
 dimStyle: {
 color: 'theme' // ref bxDim Document
 },
 enableDrag: true,
 onActivate: function(e, $elem) {
 console.log('onActivate', e, $elem);
 },
 onDrag: function (e, $elem){
 console.log('onDrag', e, $elem);
 },
 onDragStart: function (e, $elem){
 console.log('onDragStart', e, $elem);
 },
 onDragStop: function(e, $elem){
 console.log('onDragStop', e, $elem);
 },
 onClosed: function () {
 console.log('bxPaper - onClose');
 },
 onRender: function () {
 console.log('bxPaper - onRender')
 }
 }
 ).render();

 ----------------------------------------------------------------------

 // 3. bind 함수 예제

 // html

 <div id="bind_paper" style="visibility: hidden;">
 <div class="bx-paper-wrap">
 <a href="#" class="focus" onclick="return false"></a>
 <div class="paper-drag"></div>
 <div class="paper-header">
 <h2 class="bx-tt"> bind 예제 #1 </h2>
 </div>
 <div class="paper-content paper-curve-btm">
 <div class="bx-desc">
 <div> bind 예제 입니다.</div>
 <div> 사용하고자 하는 template 는 </div>
 <div> 이곳 bx-desc 하위에 입력해주세요. </div>
 </div>
 </div>
 <div class="paper-btn-wrap">
 <button type="button" class="bx-btn-close">
 <i class="bw-icon i-30 i-close"></i>
 </button>
 </div>
 </div>
 </div>

 // js

 $('#bind_paper').bxPaper(
 {
 width: 500,
 height: 500,
 position: {
 x:0,
 y:0
 },
 color: 'theme',
 type: 'stitch',
 shadow: 'both',
 titleAlign: 'center',
 dimStyle: {
 color: 'black' // ref bxDim Document
 },
 enableDrag: true,
 onActivate: function(e, $elem) {
 console.log('onActivate', e, $elem);
 },
 onDrag: function (e, $elem){
 console.log('onDrag', e, $elem);
 },
 onDragStart: function (e, $elem){
 console.log('onDragStart', e, $elem);
 },
 onDragStop: function(e, $elem){
 console.log('onDragStop', e, $elem);
 },
 onClosed: function () {
 console.log('bxPaper - onClose');
 },
 onRender: function () {
 console.log('bxPaper - onRender')
 }
 }
 ).bind();

 ----------------------------------------------------------------------

 // 4. angular 예제

 angularApp.controller('angularAppController', function($scope, $compile, $timeout){

 var paper1 = $().bxPaper({
 // paper options ....
 });

 paper1.render($compile, $scope);

 var paper2 = $().bxPaper({
 // paper options ....
 });

 paper2.bind($timeout);

 });


 */


//const Base = require('../base');
//const OPTS = Symbol();
//
//// 전역 클래스 객체
//const CLASS_NAME = {
//    //"bxUiSelect": '-bx-ui-select',
//    "bxDropDownWrap": 'bx-dropdown-wrap'
//    //"overflow": 'overflow',
//    //"bxSelectHeader": 'bx-select-header',
//    //"bxSelectList": 'bx-select-list',
//    //"bxSelectHeaderClick": 'bx-select-header-click',
//    //"option": 'option',
//    //"optionGroup": 'option-group',
//    //"optionValue": 'option-value',
//    //"icon": 'icon',
//    //"disabled": 'disabled',
//    //"overflowSelector": 'overflow-selector',
//    //"focus": 'focus',
//    //"selected": 'selected',
//    //"bxCheckbox": 'bx-checkbox',
//    //"text": 'text'
//};
//
//class bxDropDown {
//
//
//    constructor({
//        //type = 'default',
//        width = 100,
//        items = {},
//        // 위 아래 포지션이 있어야하나?
//        positions = {},
//        onShow = function (){
//        },
//        onHide = function (){
//        },
//        onSelected = function (){
//        },
//        onRender = function (){
//        }
//        } = {}){
//
//        //if (!_.isString(type)) {
//        //    BXUI.Console.prop.throw(bxDropDown, 'type', 'type');
//        //}
//
//        if (!_.isNumber(width)){
//            BXUI.Console.prop.throw(bxDropDown, 'width', 'type');
//        }
//
//        if (!_.isObject(items)){
//            BXUI.Console.prop.throw(bxDropDown, 'items', 'type');
//        }
//
//        if (!_.isObject(positions)){
//            BXUI.Console.prop.throw(bxDropDown, 'positions', 'type');
//        }
//
//        if (!_.isFunction(onShow)){
//            BXUI.Console.prop.throw(bxDropDown, 'onShow', 'type');
//        }
//
//        if (!_.isFunction(onHide)){
//            BXUI.Console.prop.throw(bxDropDown, 'onHide', 'type');
//        }
//
//        if (!_.isFunction(onSelected)){
//            BXUI.Console.prop.throw(bxDropDown, 'onSelected', 'type');
//        }
//
//        if (!_.isFunction(onRender)){
//            BXUI.Console.prop.throw(bxDropDown, 'onRender', 'type');
//        }
//
//        // 있을 경우
//        positions = {
//            menu: 'top', // bottom
//            subMenu: 'left' // right
//        };
//
//        items = [
//            {
//                id: 'menu1',
//                title: '드롭다운 리스트 영역 1',
//                items: [
//                    {
//                        id: 'menu1-1',
//                        title: '메뉴1-1',
//                        icon: '',
//                        desc: '설명 문구',
//                        items: [
//                            {
//                                id: 'menu1-1-3',
//                                title: '메뉴1-1-3',
//                                items: [
//                                    {
//                                        title: '메뉴1-1-3-1',
//                                        items: [
//                                            {
//                                                id: 'menu1-1-3-1-1',
//                                                title: '메뉴1-1-3-1-1'
//                                            }
//                                        ]
//                                    }]
//                            },
//                            {
//                                id: 'menu1-1-4',
//                                title: '메뉴1-1-4'
//                            }
//                        ]
//                    }
//                ]
//            },
//            {
//                id: 'menu2',
//                title: '드롭다운 리스트 영역 2',
//                enableDivider: true,
//                items: [
//                    {
//                        id: 'menu1-1',
//                        title: '메뉴1-1'
//                    }
//                ]
//            }
//        ];
//
//        // 상태(에러가 무슨 상황이었지?)
//
//        this[OPTS] = {type: 'default', width, items, positions, onShow, onHide, onSelected, onRender};
//
//        //this.items = _parseItems(items);
//    }
//
//    render(){
//
//        let $container = Base.appendContainerElement(bxDropDown.$root, bxDropDown.className);
//
//        this.$container = $container;
//
//        let html = _createDOMElement.call(this);
//
//        $container.append(html);
//
//        this.onRender.call(this, this.$container);
//    }
//}
//
///*
// * accordion html을 반환하는 함수
// *
// * @param {array} items - item parameter 배열
// *
// * @private
// * @memberOf bxAccordion
// */
//function _createDOMElement(){
//
//    let html = [];
//
//    html.push(`<div class="${CLASS_NAME.bxDropDownWrap}">`);
//    html.push(`<ul class="${CLASS_NAME.bxDropDownWrap}">`);
//
//    html.push(_createItemsElements(items));
//
//    html.push('</ul>');
//    html.push('</div>');
//
//    return html.join('');
//
//}
//
////function _parseItems(items = [], depth = 1){
////
////    return items.map((item) => {
////
////        if (_.isString(item)){
////            item = {title: item};
////        }
////
////        item.id = item.id ||  _.uniqueId(`${SELECTOR_NAME.item}-`);
////        item.depth = depth;
////        item.title = item.title || STYLE_VALUE.space;
////
////        if (item.items) item.items = _parseItems(item.items, depth + 1);
////
////        return item;
////
////    });
////}
//
//module.exports = bxDropDown;
