// bxuip-angular.js-ui/src/lib/bx-ui/app/components/grid.js
// OPT 내부 함수를 DOCS로 빼내기 위한 @callback 설정.
/**
 * 수정 및 삭제 버튼 클릭 시 호출될 callback 함수
 *
 * @callback onRowAction
 * @param {string} id - 이벤트가 발생한 행의 아이디
 * @param {object} data -  이벤트가 발생한 행의 데이터
 * @param {string} condition - 'del', 'edit', 'save', 'cancel'
 * @memberOf BxGrid
 */

/**
 * 수정 및 삭제 버튼 클릭 시 호출될 callback 함수 이고 executor()을 호출 해야 inline 기능이 작동
 *
 * @callback onBeforeRowAction
 * @param {string} id - 이벤트가 발생한 행의 아이디
 * @param {object} data -  이벤트가 발생한 행의 데이터
 * @param {string} condition - 'del', 'edit', 'save', 'cancel'
 * @param {function} executor - 호출 시 inline 기능이 작동
 * @memberOf BxGrid
 */

/**
 * inline edit mode 설정시, 저장 버튼 클릭 이벤트의 callback 함수
 * @callback onInlineSave
 * @param {string} id - 이벤트가 발생한 행의 아이디
 * @param {object} data - 이벤트가 발생한 행의 데이터
 * @param {boolean} state - 저장 성공 유무
 * @memberOf BxGrid
 */

/**
 * inline edit mode 설정 시, 수정 버튼 클릭 이벤트의 callback 함수
 * @callback onInlineEdit
 * @param {string} id - 이벤트가 발생한 행의 아이디
 * @param {object} data -  이벤트가 발생한 행의 데이터
 * @memberOf BxGrid
 */

/**
 * inline edit mode 설정 시, 취소 버튼 클릭 이벤트의 callback 함수
 * @callback onInlineCancel
 * @param {string} id - 이벤트가 발생한 행의 아이디
 * @param {object} data -  이벤트가 발생한 행의 데이터
 * @memberOf BxGrid
 */

/**
 * inline edit mode 설정 시, 삭제 버튼 클릭 이벤트의 callback 함수
 * @callback onInlineDelete
 * @param {string} id - 이벤트가 발생한 행의 아이디
 * @param {object} data -  이벤트가 발생한 행의 데이터
 * @memberOf BxGrid
 */

/**
 * 그리드가 업데이트 되었을때 호출될 callback 함수
 *
 * @callback onLoadComplete
 * @param {object} data
 * @param {object} data.page - 현재 페이지
 * @param {object} data.records - 총 record 수
 * @param {object} data.rows - row 배열
 * @param {object} data.total - 총 페이지 수
 * @param {string} data.action - onLoadComplete가 실행 되기 직전의 CRUD 상태
 * @memberOf BxGrid
 */

/**
 * row 선택 시 호출될 callback 함수
 *
 * @callback onSelectRow
 * @param {object} data - 이벤트가 발생한 곳의 데이터
 * @param {object} e - event 객체
 * @memberOf BxGrid
 */

/**
 * cell 선택 시 호출될 callback 함수 (편집 상태에서는 호출되지 않음)
 *
 * @callback onSelectCell
 * @param {int} index - cell 인덱스
 * @param {string} content - cell content
 * @param {object} rowData - row data
 * @param {string} colName -  column 명
 * @param {object} e - 이벤트 객체
 * @memberOf BxGrid
 */

/**
 * row 더블클릭 시 호출될 callback 함수
 *
 * @callback onDblClick
 * @param {object} data - 이벤트가 발생한 곳의 데이터
 * @param {object} e - event 객체
 * @memberOf BxGrid
 */

/**
 * @class BxGrid
 * @classDesc BxGrid 클래스 생성자
 *
 * @param {object} OPTS
 * @param {int} OPTS.height - grid 높이
 * @param {int} [OPTS.width] - grid 넓이.<br/>
 * width가 100 이상 일 경우 테이블 크기에 따라 가로 스크롤이 생긴다.
 * @param {boolean} OPTS.resizable - grid 리사이즈 여부
 * @param {string} [OPTS.title = ''] - <s>grid title을 지정, 접근성을 위해 grid caption에 사용 <b class="dec">예) title+' 목록'</b></s> <br/><span class='important'>( Added in: v3.6.0, Deprecated since: v3.8.0 )</span>
 * @param {object} [OPTS.$pageContainer = null] - 페이징 container 엘리먼트
 * @param {object} [OPTS.rowNumbers = false] - 행 번호 열 보임 유무
 * @param {int} [OPTS.rowNumWidth = 25] - 행 번호 width 사이즈
 * @param {string} [OPTS.sortname = ''] - <s>정렬할 컬럼 이름</s> <span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @param {string} [OPTS.sortName = ''] - 정렬할 컬럼 이름
 * @param {string} [OPTS.sortorder = 'desc'] - <s>정렬 타입 ('asc', 'desc')</s> <span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @param {string} [OPTS.sortOrder = 'desc'] - 정렬 타입 ('asc', 'desc')
 * @param {boolean} [OPTS.multiSort = false] - 다중 정렬 유무
 * @param {boolean} [OPTS.multiSelect = false] - 다중 행 선택 설정 유무
 * @param {object[]} [OPTS.items = []] - 그리드 아이템.<br/>
 * 아이템 내부 속성 명은 colModel.name과 매칭되어야 하며, colModel의 속성과 <b class="dec">반드시 1:1 매칭이 돼야 한다</b> <br/>
 * 아이템 내부 <span class="dec">id 속성이 존재</span>하는 경우, 이는 <b class="dec">반드시 유니크한 값을 사용</b>해야 한다
 * @param {object[]} [OPTS.colModel = []]
 * @param {boolean} [OPTS.colModel.key='false'] - <s><b class="dec">primary key</b>인 경우 사용<br />
 * 단, 하나만 적용 된다. 만약 여러 컬럼에 key값이 있는 경우엔 <b class="dec">먼저 호출된 key가 적용</b>된다. </s>
 * <span class='important'>( Added in: v3.0.0, Deprecated since: v3.9.0 )</span>
 * @param {string} OPTS.colModel.label - 그리드 컬럼 이름
 * @param {string} OPTS.colModel.name - 데이터와 매칭되는 컬럼 이름
 * <span class="dec">[ required ]</span><br/>
 * @param {int} [OPTS.colModel.width=''] - 컬럼 넓이, <b class="dec">값이 지정되지 않을 경우 컬럼 개수에 따라 넓이가 자동으로 지정된다.</b>
 * @param {boolean} [OPTS.colModel.hidden=false] - 컬럼 숨김 유무
 * @param {string} [OPTS.colModel.align='center'] - cell 위치 정렬 타입 ('left', 'center', 'right')
 * @param {string} [OPTS.colModel.sorttype='text'] - 정렬 타입 ('int', 'float', 'number', 'currency', 'date', 'text')
 * @param {boolean} [OPTS.colModel.sortable=true] - 정렬 가능 유무
 * @param {object[]} [OPTS.colModel.editOption = undefined]
 * @param {string} [OPTS.colModel.editOption.type] - inline Edit Mode 시, 사용 엘리먼트 지정 ('text', 'textarea', 'checkbox', 'radio', 'select', 'custom')
 * @param {object[]} [OPTS.colModel.editOption.items] - editOption.type 이 <b>checkbox, radio, select</b> 인 경우,
 * 사용될 객체로 객체는 아래와 같이 구성된다.<br/>
 * {label: '', value: ''}</br>
 * <b class="dec">label</b> 은 <i>해당 컬럼의 데이터 값</i>이며, <b class="dec">value</b> 값은 <i>editOption.targetCol 에 설정한 컬럼의 데이터 값</i>이 된다.
 * @param {string} [OPTS.colModel.editOption.targetCol] - editOption.type 이 <b>checkbox, radio, select</b> 인 경우,
 * editOption.items 에 정의한 value 값이 대응된다.
 * @param {string} [OPTS.colModel.editOption.separator] - editOption.type = 'checkbox' 인 경우 사용이 가능하며, 체크박스 다중 선택시 구분자.
 * @param {object} [OPTS.colModel.editOption.attr] - inline Edit Mode 의 사용 엘리먼트의 속성값을 지정.
 * 'text', 'textarea' 는 해당 태그에 속성값을 지정.
 * 'radio', 'checkbox', 'select' 는 edit 컨테이너에 속성값 지정.
 * 'custom' 은 사용자가 직접입력이 가능하므로 이는 배재.
 * <span class="essential">id, name 속성은 할당이 불가하다.</span>
 *
 * @param {object} [OPTS.colModel.editOption.validateOption]
 * @param {function} [OPTS.colModel.editOption.validateOption.validator] - 유효성 검사 함수. 유효성 검사 성공시 true, 실패시 false 를 반환해야 한다.
 * @param {string} [OPTS.colModel.editOption.validateOption.message] - 유효성 검사 실패시, 다이얼로그에 나오게 할 문구. <span class='important'> ( Added in: v3.0.3, Deprecated since: v3.11.0 )</span>
 *
 * @param {function} [OPTS.colModel.editOption.customElement] - editOption.type = 'custom' 인 경우, 사용될 HTML 을 return 하는 함수를 선언. <br/>
 * 내부에 들어가는 태그에는 반드시 <span class="important">data-id가 attribute 로 포함되어야 한다.</sapn>
 * @param {function} [OPTS.colModel.editOption.customElementValue] - editOption.type = 'custom' 인 경우, customElement DOM 을 인자로 받아 값을 return 하는 함수를 선언.
 *
 * @param {object} [OPTS.colModel.formatOption = undefined]
 * @param {string} OPTS.colModel.formatOption.type - 해당 셀의 formatting 형식을 지정 ('currency', 'date', 'pid', 'bzno', 'tel', 'card', 'custom') <br/>
 * currency - 통화<br/>
 * date - 날짜<br/>
 * pid - 주민등록번호<br/>
 * bzno - 사업자등록번호<br/>
 * tel - 전화번호(핸드폰번호)<br/>
 * card - 카드번호<br/>
 * custom - 사용자 설정<br/>
 * @param {object} [OPTS.colModel.formatOption.options]
 * @param {string} [OPTS.colModel.formatOption.options.prefix = ''] - <b class="dec">formatOption.type</b>이 <b class="dec">'currency'</b> 인 경우 사용이 가능하며, 가장 앞에 붙일 문자를 설정한다.
 * @param {string} [OPTS.colModel.formatOption.options.suffix = ''] - <b class="dec">formatOption.type</b>이 <b class="dec">'currency'</b> 인 경우 사용이 가능하며, 가장 뒤에 붙일 문자를 설정한다.
 * @param {int} [OPTS.colModel.formatOption.options.decimalPlaces = 2] - <b class="dec">formatOption.type</b>이 <b class="dec">'currency'</b> 인 경우 사용이 가능하며, 소수점 출력 자릿수를 설정한다.
 * @param {string} OPTS.colModel.formatOption.options.dataFormat - <b class="dec">formatOption.type</b>이 <b class="dec">'date'</b> 인 경우 사용이 가능하며, item에 들어간 날짜 데이터의 형식을 설정해야한다.
 * @param {string} OPTS.colModel.formatOption.options.viewFormat - <b class="dec">formatOption.type</b>이 <b class="dec">'date'</b> 인 경우 사용이 가능하며, grid에 출력될 format을 설정한다.<br/>
 *     yyyy - 년도 (4자리)<br/>
 *     yy - 년도 (2자리)<br/>
 *     MM - 월<br/>
 *     dd - 일<br/>
 *     hh - 시간<br/>
 *     mm - 분<br/>
 *     ss - 초<br/>
 * @param {object} OPTS.colModel.formatOption.customFormatter - <b class="dec">formatOption.type</b>이 <b class="dec">'custom'</b> 인 경우 사용이 가능하며, grid에 출력될 데이터 format을 설정한다.
 * @param {object} OPTS.colModel.formatOption.customUnFormatter - <b class="dec">formatOption.type</b>이 <b class="dec">'custom'</b> 인 경우 사용이 가능하며, item에 저장될 데이터 format을 설정한다.
 *
 * @param {string} OPTS.colModel.formatter - <s>cell 값에 대한 format 을 정의한다. 'actions' 값을 통해, 수정, 삭제 버튼을 생성할 수 있다.</s> <span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 *
 * @param {function} [OPTS.colModel.attr] - cell(td)의 속성 설정 함수. {속성명: 속성값} 형태의 객체를 반환해야 한다.
 *
 * @param {boolean} [OPTS.isLoadingBar = false] - <s>로딩바 설정 유무</s> <span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 *
 * @param {boolean} [OPTS.isPaging = false] - <s>페이징 처리 유무</s> <span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @param {boolean} [OPTS.enablePaging = false] - 페이징 처리 유무
 * @param {int} [OPTS.rowNum = 0] - 화면에 보여질 행 수
 * @param {int} [OPTS.pageGroupSize = 10] - 페이지 네이션 그룹 사이즈
 *
 * @param {object} [OPTS.ajaxModel] - <b class="dec">AjaxModel을 사용 할 경우에 반드시 primary key에 해당하는 필드가 colModel에 추가되어야 한다.</b></br>
 * request 요청의 query String 은 기본적으로 4가지로 구성된며 아래와 같다.<br/>
 * <b class="dec">pageNo: 현재 페이지, rowNum: 페이지 사이즈, sort: 정렬 컬럼, order: 정렬 순서<br/>
 * 예) [url]/api/page?pageNo=1&rowNum=10&sort=Name,Salary&order=asc,asc<br/>
 * </b>
 * response 는 2가지 데이터를 담고 있어야 하며 아래와 같다.<br/>
 * <b class="dec">data: 데이터 배열이며, totalCount: 전체 데이터의 수.</b>
 * @param {object} [OPTS.ajaxModel.readOption]
 * @param {string} [OPTS.ajaxModel.readOption.url] - Ajax에서 아이템을 가져올 때 사용할 url<br>
 * @param {string} [OPTS.ajaxModel.readOption.method] - Ajax 통신 할 method
 * @param {object} [OPTS.ajaxModel.readOption.headers] - Ajax에 추가 할 헤더 정보
 * @param {object} [OPTS.ajaxModel.readOption.params] - Ajax에 요청할 파라미터
 * @param {function} [OPTS.ajaxModel.readOption.before] - Ajax 통신 실행 전 callback 함수
 * @param {function} [OPTS.ajaxModel.readOption.success] - Ajax 통신 성공 시 실행 할 callback 함수
 * @param {function} [OPTS.ajaxModel.readOption.fail] - Ajax 통신 실패 시 실행 할 callback 함수
 * @param {object} [OPTS.ajaxModel.createOption]
 * @param {string} [OPTS.ajaxModel.createOption.url] - Ajax에서 아이템을 추가할 때 사용할 url
 * @param {string} [OPTS.ajaxModel.createOption.method] - Ajax 통신 할 method
 * @param {object} [OPTS.ajaxModel.createOption.headers] - Ajax에 추가 할 헤더 정보
 * @param {object} [OPTS.ajaxModel.createOption.params] - Ajax에 요청할 파라미터
 * @param {function} [OPTS.ajaxModel.createOption.before] - Ajax 통신 실행 전 callback 함수
 * @param {function} [OPTS.ajaxModel.createOption.success] - Ajax 통신 성공 시 실행 할 callback 함수
 * @param {function} [OPTS.ajaxModel.createOption.fail] - Ajax 통신 실패 시 실행 할 callback 함수
 * @param {object} [OPTS.ajaxModel.deleteOption]
 * @param {string} [OPTS.ajaxModel.deleteOption.url] - Ajax에서 아이템을 삭제할 때 사용할 url
 * @param {string} [OPTS.ajaxModel.deleteOption.method] - Ajax 통신 할 method
 * @param {object} [OPTS.ajaxModel.deleteOption.headers] - Ajax에 추가 할 헤더 정보
 * @param {object} [OPTS.ajaxModel.deleteOption.params] - Ajax에 요청할 파라미터
 * @param {function} [OPTS.ajaxModel.deleteOption.before] - Ajax 통신 실행 전 callback 함수
 * @param {function} [OPTS.ajaxModel.deleteOption.success] - Ajax 통신 성공 시 실행 할 callback 함수
 * @param {function} [OPTS.ajaxModel.deleteOption.fail] - Ajax 통신 실패 시 실행 할 callback 함수
 * @param {object} [OPTS.ajaxModel.updateOption]
 * @param {string} [OPTS.ajaxModel.updateOption.url] - Ajax에서 아이템을 변경할 때 사용 url
 * @param {string} [OPTS.ajaxModel.updateOption.method] - Ajax 통신 할 method
 * @param {object} [OPTS.ajaxModel.updateOption.headers] - Ajax에 추가 할 헤더 정보
 * @param {object} [OPTS.ajaxModel.updateOption.params] - Ajax에 요청할 파라미터
 * @param {function} [OPTS.ajaxModel.updateOption.before] - Ajax 통신 실행 전 callback 함수
 * @param {function} [OPTS.ajaxModel.updateOption.success] - Ajax 통신 성공 시 실행 할 callback 함수
 * @param {function} [OPTS.ajaxModel.updateOption.fail] - Ajax 통신 실패 시 실행 할 callback 함수
 *
 * @param {string} [OPTS.ajaxModel.reqModel] - 기본 parameter(paging, sort)가 설정 될 path
 *
 * @param {object} [OPTS.ajaxModel.resModel]
 * @param {string} [OPTS.ajaxModel.resModel.totalCount='totalCount'] - 서버 반환 값의 전체 데이터 수
 * @param {string} [OPTS.ajaxModel.resModel.items='data'] - 서버 반환 값의 데이터
 *
 * @param {object} [OPTS.exportModel]
 * @param {string} [OPTS.exportModel.type='excel'] - export 형식을 지정 (현재는 excel만 지원)
 * @param {string} [OPTS.exportModel.fileName='unknown'] - export 될 파일 명
 * @param {string} [OPTS.exportModel.sheetName=''] - excel 파일의 sheet 명
 * @param {string} [OPTS.exportModel.dataFormat='formatted'] - export 데이터의 포매팅 형식 ('formatted', 'plain')
 * @param {boolean} [OPTS.exportModel.enableExpButton=false] - export 버튼 유무
 * @param {boolean} [OPTS.exportModel.enableTimeStamp=false] - 파일명 뒤에 timestamp 추가 유무
 * @param {object} [OPTS.exportModel.ajaxOption]
 * @param {string} OPTS.exportModel.ajaxOption.url - export 데이터를 가져올 url
 * @param {string} [OPTS.exportModel.ajaxOption.method] - Ajax 통신 할 method
 * @param {object} [OPTS.exportModel.ajaxOption.headers] - Ajax 통신에 추가 할 헤더 정보
 * @param {object} [OPTS.exportModel.ajaxOption.params] - Ajax에 요청 할 파라미터
 * @param {function} [OPTS.exportModel.ajaxOption.before] - Ajax 통신 실행 전 callback 함수
 * @param {function} [OPTS.exportModel.ajaxOption.success] - Ajax 통신 성공 시 실행 할 callback 함수
 * @param {function} [OPTS.exportModel.ajaxOption.fail] - Ajax 통신 실패 시 실행 할 callback 함수
 * @param {string} [OPTS.exportModel.ajaxOption.resModel] - 응답 데이터 모델
 *
 * @param {object} [OPTS.importModel]
 * @param {string} [OPTS.importModel.type='excel'] - import 형식을 지정 (현재는 excel만 지원)
 * @param {boolean} [OPTS.importModel.enableImpButton=false] - import 버튼 유무
 * @param {boolean} [OPTS.importModel.enableResetData=true] - ajax 미사용시, 기존 데이터 초기화 여부
 * @param {object} OPTS.importModel.ajaxOption
 * @param {string} OPTS.importModel.ajaxOption.url - Ajax excel 파일을 전송 할 타겟 url
 * @param {string} [OPTS.importModel.ajaxOption.method] - Ajax 통신 할 method
 * @param {object} [OPTS.importModel.ajaxOption.headers] - Ajax에 추가 할 헤더 정보
 * @param {object} [OPTS.importModel.ajaxOption.params] - Ajax에 요청할 파라미터
 * @param {function} [OPTS.importModel.ajaxOption.before] - Ajax 통신 실행 전 callback 함수
 * @param {function} [OPTS.importModel.ajaxOption.success] - Ajax 통신 성공 시 실행 할 callback 함수
 * @param {function} [OPTS.importModel.ajaxOption.fail] - Ajax 통신 실패 시 실행 할 callback 함수
 * @param {string} [OPTS.importModel.ajaxOption.resModel] - 응답 데이터 모델
 *
 * @param {boolean} [OPTS.inlineEditCol = false] - 컬럼 넓이가 100인 inline Edit 버튼 컬럼 추가 설정 유무
 * @param {boolean} [OPTS.inlineEditColModel = {}]
 * @param {boolean} [OPTS.inlineEditColModel.cancel = {}]
 * @param {boolean} [OPTS.inlineEditColModel.cancel.show] - 일반모드 시, 취소버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.cancel.showEditMode] - 편집모드 시, 취소버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.delete = {}]
 * @param {boolean} [OPTS.inlineEditColModel.delete.show] - 일반모드 시, 취소버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.delete.showEditMode] - 편집모드 시, 취소버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.save = {}]
 * @param {boolean} [OPTS.inlineEditColModel.save.show] - 일반모드 시, 저장버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.save.showEditMode] - 편집모드 시, 저장버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.edit = {}]
 * @param {boolean} [OPTS.inlineEditColModel.edit.show] - 일반모드 시, 수정버튼 사용 유무
 * @param {boolean} [OPTS.inlineEditColModel.edit.showEditMode] - 편집모드 시, 수정버튼 사용 유무
 *
 * @param {string} [OPTS.inlineEditColLabel = '편집'] - inline Edit 컬럼의 라벨 명 지정
 * @param {boolean} [OPTS.enableStartEditMode = false] - inline Edit 로 시작 유무 [ajax 사용시 사용 불가하다.]
 *
 * @param {object[]} [OPTS.groupHeaders = []] - groupHeader 객체
 * @param {string} [OPTS.groupHeaders.startColName=''] - 시작 컬럼 명 (colModel 아이템 중 name 과 매칭되야 한다)
 * @param {int} [OPTS.groupHeaders.numberOfCol=0] - 시작 컬럼 명 기준으로 colspan 개수를 지정 한다.
 * @param {string} [OPTS.groupHeaders.content=''] - groupHeader HTML
 *
 * @param {array[]} [OPTS.footers = null] - footers 내부에 배열 개수 만큼 footer를 추가 한다.
 * @param {string} [OPTS.footers.colName=''] - 시작 컬럼 명 (colModel 아이템 중 name 과 매칭되야 한다.)
 * @param {string} [OPTS.footers.type='label'] - label, count, sum, avg, min , max 을 지원 한다.
 * @param {string} [OPTS.footers.content=''] - 해당 컬럼 명에 content 정보를 적용하고 {0}을 치환 한다. 예) 합계: {0} $ -> 합계: 100 $
 * @param {int} [OPTS.footers.numberOfCol=1] - 시작 컬럼 명 기준으로 colspan 개수를 지정 한다.
 *
 * @param {function} [OPTS.onBeforeRowAction = function(){}] - inline edit mode 설정 시 수정,삭제,저장,취소 버튼 클릭 이벤트의 callback 함수이고 <b class="dec">exec()</b>를 호출해야 inline 기능이 작동한다
 * @param {function} [OPTS.onRowAction = function(){}] - inline edit mode 설정 시 수정,삭제,저장,취소 버튼 클릭 이벤트의 callback 함수
 * @param {function} [OPTS.onInlineSave = function(){}] - inline edit mode 설정 시 저장 버튼 클릭 이벤트의 callback 함수
 * @param {function} [OPTS.onInlineEdit = function(){}] - inline edit mode 설정 시수정 버튼 클릭 이벤트의 callback 함수
 * @param {function} [OPTS.onInlineCancel = function(){}] - inline edit mode 설정 시 취소 버튼 클릭 이벤트의 callback 함수
 * @param {function} [OPTS.onInlineDelete = function(){}] - inline edit mode 설정 시 삭제 버튼 클릭 이벤트의 callback 함수
 *
 * @param {function} [OPTS.onRender = function(){} ] - 랜더링 완료 시 호출될 callback 함수
 * @param {function} [OPTS.onLoadComplete = function(){}] - 그리드가 업데이트 되었을때 호출될 callback 함수
 * @param {function} [OPTS.onReload = function(){}] - 다시 렌더링 시 호출될 callback 함수
 * @param {function} [OPTS.onSelectRow = function(){}] - row 선택 시 호출될 callback 함수
 * @param {function} [OPTS.onSelectCell = function(){}] - cell 선택 시 호출될 callback 함수 (편집 상태에서는 호출되지 않음)
 * @param {function} [OPTS.onDblClick = function(){}] - row 더블클릭 시 호출될 callback 함수
 *
 * @returns {object} 생성된 인스턴스
 * @example

 var item = [
 {"Name":"AAAAA", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20160101100546", "Salary":"323522.3800", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"HANAR", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20151130094523", "Salary":"564565.8300", "tel":"03199998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"VICTE", "logCode":"CODE002", "logCodeName": "RED", "date": "20161231231105", "Salary":"452341.3400", "tel":"01011112222", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"SUPRD", "logCode":"CODE002", "logCodeName": "RED", "date": "20160301180524", "Salary":"5523451.3000", "tel":"0182223333", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"HANAR", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20160101023415", "Salary":"567568.1700", "tel":"0119996666", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "400", "stateCodeName": "NOT FOUND", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"CHOPS", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20160228133000", "Salary":"212352.9800", "tel":"0439994444", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"RICSU", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20160315203010", "Salary":"145648.3300", "tel":"03199998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"WELLI", "logCode":"CODE002", "logCodeName": "RED", "date": "20160411181741", "Salary":"1345633.9700", "tel":"0189998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"HILAA", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20160508164005", "Salary":"567881.9100", "tel":"0179998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "400", "stateCodeName": "NOT FOUND", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"ERNSH2", "logCode":"CODE002", "logCodeName": "RED", "date": "20160620030303", "Salary":"185640.5100", "tel":"0169998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"CENTC2", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20160717040404", "Salary":"33553.2500", "tel":"0419998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "400", "stateCodeName": "NOT FOUND", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"OLDWO2", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20160820050505", "Salary":"112355.0900", "tel":"0639998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"QUEDE2", "logCode":"CODE002", "logCodeName": "RED", "date": "20160924060606", "Salary":"33123.0500", "tel":"0119998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "300", "stateCodeName": "MODIFY", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"RATTC2", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20161001070707", "Salary":"53448.2900", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"ERNSH2", "logCode":"CODE001", "logCodeName": "BLUE", "date": "201601010080808", "Salary":"143456.0600", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "300", "stateCodeName": "MODIFY", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"FOLKO2", "logCode":"CODE002", "logCodeName": "RED", "date": "20161111090909", "Salary":"352735.6700", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"BLONP2", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20161225101010", "Salary":"579965.2800", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "400", "stateCodeName": "NOT FOUND", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"WARTH2", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20150518111111", "Salary":"263435.7300", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"FRANK2", "logCode":"CODE002", "logCodeName": "RED", "date": "20151025121212", "Salary":"206548.5800", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"ERNSH3", "logCode":"CODE002", "logCodeName": "RED", "date": "20150420131313", "Salary":"1631440.5100", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "400", "stateCodeName": "NOT FOUND", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"CENTC3", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20150130141414", "Salary":"5424573.2500", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"OLDWO3", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20150214151515", "Salary":"585635.0900", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"QUEDE3", "logCode":"CODE002", "logCodeName": "RED", "date": "20150314161616", "Salary":"3543567.0500", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"RATTC3", "logCode":"CODE002", "logCodeName": "RED", "date": "20150515171717", "Salary":"43123568.2900", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"ERNSH3", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20150320181818", "Salary":"112356846.0600", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "400", "stateCodeName": "NOT FOUND", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"FOLKO3", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20150830191919", "Salary":"3123689.6700", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "3", "testCodeName": "ORANGE"},
 {"Name":"BLONP3", "logCode":"CODE201", "logCodeName": "YELLOW", "date": "20151101202020", "Salary":"5125775.2800", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "2", "testCodeName": "BANANA"},
 {"Name":"WARTH3", "logCode":"CODE002", "logCodeName": "RED", "date": "20151212212121", "Salary":"258888.7300", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "200", "stateCodeName": "OK", "testCode": "1", "testCodeName": "APPLE"},
 {"Name":"FRANK3", "logCode":"CODE001", "logCodeName": "BLUE", "date": "20150801222222", "Salary":"20535778.5800", "tel":"029998888", "bzno": "1112233333", "pid": "9010011091296", "card":"1111222233334444", "stateCode": "500", "stateCodeName": "ERROR", "testCode": "2", "testCodeName": "BANANA"}
 ];

 // var grid = $('#grid').BxGrid({

    height: 200,
    resizable: false,
    $pageContainer: $('#pager'),
    colModel: [
        {
        // 해당 object는 서버쪽 primary key 설정을 위한 필드값 입니다
            name: 'no',
            hidden: true
        },
        {
            name: 'logCode',
            hidden: true
        },
        {
            name: 'stateCode',
            hidden: true
        },
        {
            name: 'testCode',
            hidden: true
        },
        {
            label: 'Customer Name',
            name: 'Name',
            width: 80,
            attr: function(rowId, value, colModel, rowData){
                return {
                    id: rowId + '-' + colModel.name,
                    headers: 'header1 header3 header4'
                }
            },
            editOption: {
                type: 'custom',
                customElement: function(gridValue){
                    return '<input type="text" value="' + gridValue + '"/>';
                },
                customElementValue : function($elem){
                    return $elem.get(0).value;
                }
            }
        },
        {
            label: 'Salary',
            name: 'Salary',
            width: 100,
            sorttype: 'float',
            editOption: {
                type: 'text',
                validateOption: {
                    validator: function(value, colName, rowData){
                        var pattern1 = /^\d+.?\d+$/.test(value);
                        var pattern2 = /^-?(\d*([.]\d{0,3})?)?$/.test(value);

                        if (!pattern1) {
                            alert('에러문구를 지정 할 수 있습니다.')
                        }

                        if(!pattern2) {
                            alert('소수점 세째자리 까지 됩니다.')
                        }

                        return pattern1 && pattern2; // false 일 경우 저장 되지 않음
                     }
                }
            },
            formatOption: {
                type: 'currency',
                options: {
                    prefix: '$',
                    suffix: '~',
                    decimalPlaces: 3
                }
            }
        },
        {
            label: 'Date',
            name: 'date',
            width: 150,
            formatOption: {
                type: 'date',
                options: {
                    dataFormat: 'yyyyMMddhhmmss',
                    viewFormat: 'MM-dd-yyyy'
                }
            }
        },
        {
            label: '주민등록번호',
            name: 'pid',
            width: 150,
            formatOption: {
                type: 'pid'
            }
        },
        {
            label: '사업자등록번호',
            name: 'bzno',
            width: 150,
            formatOption: {
                type: 'bzno'
            }
        },
        {
            label: '전화번호',
            name: 'tel',
            width: 150,
            formatOption: {
                type: 'tel'
            }
        },
        {
            label: '카드번호',
            name: 'card',
            width: 150,
            formatOption: {
                type: 'card'
            }
        },
        {
            label: 'log Name',
            name: 'logCodeName',
            width: 150,
            editOption: {
                type: 'radio',
                items: [
                    { label: 'BLUE', value:'CODE001'},
                    { label: 'RED', value:'CODE002'},
                    { label: 'YELLOW', value:'CODE201'}
                ],
                attr: {
                    class: 'test'
                },
                targetCol: 'logCode'
            }
        },
        {
            label: 'State Name',
            name: 'stateCodeName',
            width: 120,
            editOption: {
                type: 'select',
                items: [
                    {label: 'OK', value: '200' },
                    {label: 'MODIFY', value: '300' },
                    {label: 'NOT FOUND', value: '400' },
                    {label: 'ERROR', value: '500' }
                ],
                attr: {
                    class: 'test'
                },
                targetCol: 'stateCode'
            }
        },
        {
            label: 'Test Code',
            name: 'testCodeName',
            editOption: {
                type: 'checkbox',
                items: [
                    {label: 'APPLE', value: '1' },
                    {label: 'BANANA', value: '2' },
                    {label: 'ORANGE', value: '3' }
                ],
                attr: {
                    class: 'test2'
                },
                separator: '/',
                targetCol: 'testCode'
            }
        }
    ],
    items: item,
    rowNumbers: true,
    rowNumWidth : 50,
    sortName: '',
    sortOrder: 'asc',
    multiSort: true,
    multiSelect: true,
    // paging
    enablePaging: true,
    rowNum: 30,
    pageGroupSize: 3,
    // export
    exportModel: {
        type: 'excel',
        fileName: 'excel',
        sheetName: 'sheet001',
        enableExpButton: true,
        enableTimeStamp: false,
        dataFormat: 'formatted',
		ajaxOption: {
		   url: '/api/page/export',
		   method: 'get',
		   headers: {
		       "Content-Type": 'application/json'
		   },
		   params: {},
		   before: function(xhr){
		       console.log('ajax export before', xhr);
		   },
		   success: function(res){
		       console.log("ajax export success", res);
		   },
		   fail: function(jqXHR){
		       console.log("ajax export error", jqXHR);
		   }
		}
    },
    importModel: {
        type: 'excel',           // import 형식을 지정 (현재는 excel만 지원)
        enableImpButton: true,   // import 버튼 유무
        enableResetData: true,
        ajaxOption: {
            url: '/api/import',
            method: 'post',
            headers: {
                Authorization : 'key=AIzaSyB4Vc0Liohg35e7qSlTCZiJ6PbYOuwyzuI',
            },
            params: {
                a:'sd',
            },
            before: function(xhr){
                console.log("file before");
                this.params.time = new Date().getTime();
            },
            success: function(res){
                console.log("file success");
            },
            fail: function(jqXHR){
                console.log("file error");
            }
        }
    },
    ajaxModel: {
        readOption: {
            url: '/api/page',
            method: 'get',
            headers: {
                Authorization : 'key=AIzaSyB4Vc0Liohg35e7qSlTCZiJ6PbYOuwyzuI'
            },
            params: {
            },
            before: function(xhr){
                console.log('ajax get before', xhr);
            },
            success: function(res){
                console.log("ajax get success", res);
            },
            fail: function(jqXHR){
                console.log("ajax get error", jqXHR);
            }
        },
        createOption: {
            url: '/api/page',
            method: 'post',
            headers: {},
            params: {},
            before: function(xhr){
                        console.log('create before', xhr);
            },
            success: function(res){
                console.log("ajax create success", res);
            },
            fail: function(jqXHR){
                console.log("ajax create error", jqXHR);
            }
        },
        deleteOption: {
            url: '/api/page',
            method: 'delete',
            headers: {},
            params: {},
            before: function(xhr){
                console.log('delete before', xhr);
            },
            success: function(res){
                console.log("ajax delete success", res);
            },
            fail: function(jqXHR){
                console.log("ajax delete error", jqXHR);
            }
        },
        updateOption: {
            url: '/api/page',
            method: 'put',
            headers: {},
            params: {},
            before: function(xhr){
                console.log('update before', xhr);
            },
            success: function(res){
                console.log("ajax update success", res);
            },
            fail: function(jqXHR){
                console.log("ajax update error", jqXHR);
            }
        },
        resModel: {
            totalCount: 'totalCount',
            items: 'data'
        }
    },
    // inlineEdit
    inlineEditCol: true,
    inlineEditColModel: {
        edit: {
            show: true ,
            showEditMode: false
        },
        save: {
            show: false ,
            showEditMode: true
        },
        delete: {
            show: true ,
            showEditMode: true
        },
        cancel: {
            show: false,
            showEditMode: true
        }
    },
    inlineEditColLabel: '편집!',
    enableStartEditMode: false,
    groupHeaders: [
        {
            startColName: 'Salary',
            numberOfCol: 2,
            content: [
                '<tr scope="row">',
                    '<th scope="col" id="header1">',
                        'group #1',
                    '</th>',
                    '<th scope="col" id="header2">',
                        'group #2',
                    '</th>',
                '</tr>',
                '<tr scope="col">',
                    '<th colspan="2" id="header3">',
                        'group #3',
                    '</th>',
                '</tr>',
                '<tr scope="col" id="header4">',
                    '<th colspan="2">',
                        'group #4',
                    '</th>',
                '</tr>'
            ].join('')
        },
        {
            startColName: 'stateCodeName',
            numberOfCol: 2,
            content: [
                '<tr>',
                    '<th rowspan="2">',
                        'asdfasgroup #1',
                    '</th>',
                '</tr>'
            ].join('')
        }
    ],
    footers : [
        [
            {
                colName:"Name",
                type: 'label',
                content: "전체 리스트 수:",
                numberOfCol: 6
            },
            {
                colName:'testCodeName',
                type: 'count',
                content: 'row: {0} 개',
                numberOfCol:1
            }

        ],
        [
            {
                colName:'Name',
                type: 'label',
                content: '급여 : ',
                numberOfCol:1
            },
            {
                colName:'Salary',
                type: 'max',
                content: '합계: {0}',
                numberOfCol:1
            }
        ],
        [
            {
                colName:'Name',
                type: 'label',
                content: '금액 합: ',
                numberOfCol:1
            },
            {
                colName:'Salary',
                type: 'sum',
                content: '{0}',
                numberOfCol:4
            }
        ],
        [
            {
                colName:'Name',
                type: 'label',
                content: '금액 합: ',
                numberOfCol:1
            },
            {
                colName:'Salary',
                type: 'sum',
                content: '',
                numberOfCol:1
            }
        ]
    ],
    onRowAction: function (id, data, condition) {
            console.log('user onRowAction', id, data, condition);

    },
    onBeforeRowAction : function (id, data, condition, executor) {

        $().BxDialog({
            type : 'stick',
            width : 0,
            confirmText : '확인',
            closeText : '취소',
            dimStyle : {
                color : 'black',
                enableBlur : false
            }
        }).confirm({
            title: 'title',
            content: 'content',
            onConfirm: function(e){
                executor();
            },
            onClose: function(e){
                console.log("닫기");
            }
        });
    },
    onInlineCancel: function(id, data){
        console.log("OnInLine cancel", id, data);
    },
    onInlineEdit: function(id, data){
        console.log("OnILine Edit", id, data);
    },
    onInlineSave: function(id, data, state){

        if (state) {
            console.log("OnILine Save success", id, data, state);
        }
        else {
            console.log("OnILine Save fail", id, data, state);
        }
    },
    onInlineDelete: function(id, data){
        console.log("OnILine delete", id, data);
    },
    onRender: function () {
        console.log("onRender");
    },
    onReload: function (data){
        console.log("onReload", data);
    },
    onSelectRow: function(rowData, e){
                console.log('user onSelectRow', rowData, e);
    },
    onLoadComplete: function(data) {
            console.log('onLoadComplete',data);
    },
    onDblClick: function (rowData, e) {
        console.log(JSON.stringify(rowData), e);
    },
    onSelectCell: function (index, content, rowData, colName, e) {
        console.log('cell index is ', index);
    }
}).render();
*/

//* @param {string} [OPTS.caption='']- BxGrid의 상단 제목 설정
//* @param {boolean} [OPTS.grouping = false] - 그룹화 여부 설정
// * @param {object} [OPTS.groupingView = {}] - 그룹화할 컬럼 설정. 내부 프로퍼티는 그룹화할 필드 배열을 설정하는 groupField,
// *                                              그룹화할 필드의 정렬 상태를 설정하는 groupOrder가 존재한다


// * @param {object} [OPTS.colModel.editOption.validateOption.type] - inline Edit Mode 의 사용 엘리먼트의 유효성 검사 타입지정 ('number', 'phone', 'email', 'date', 'datetime', 'time', 'custom')
// * @param {object} [OPTS.colModel.editOption.validateOption.pattern] - 유효성 검사 타입이 'custom' 인 경우, 사용될 패턴

/* jshint camelcase: false */
const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');

const {Console: CONSOLE, Util: UTIL, Type: TYPE, File: FILE, Xhr:XHR, Exporter: EXPORTER, Form: FORM} = BXUI;

const OPTS = Symbol();

// POSTMAN Method 기준 RESTful
const RESTFUL_METHOD = {
    GET: 'get',
    POST: 'post',
    PUT: 'put',
    PATCH: 'patch',
    DELETE: 'delete',
    COPY: 'copy',
    HEAD: 'head',
    OPTIONS: 'options',
    LINK: 'link',
    UNLINK: 'unlink',
    PURGE: 'purge',
    LOCK: 'lock',
    UNLOCK: 'unlock',
    PROPFIND: 'propfind',
    VIEW: 'view'
};

const STYLE_VALUE = {
    ROW_HEIGHT: 40
};

const ALT_TEXT = {
    'fa-pencil': '쓰기',
    'fa-trash-o': '삭제',
    'i-save': '저장',
    'i-undo': '취소',
    'next': '다음',
    'backward': '맨뒤',
    'pre': '이전',
    'forward': '맨앞',
    'select': '선택',
    'caption': '이 표는 {columns} 를 나타내고 있다'
};

const JQGRID_QUERYSTRING_KEY = {
    page: null,
    rows: null,
    sort: null,
    order: null,
    search: null,
    nd: null,
    id: null,
    oper: null,
    editoper: null,
    addoper: null,
    deloper: null,
    subgridid: null,
    npage: null,
    totalrows: null
};

// AJAX DEFAULT REQ MODEL
const DEFAULT_REQ_PARAMS_MODEL = {
    paging: {
        currentPage: 'pageNo',
        pageSize: 'rowNum'
    },
    sort: {
        columns: 'sort',
        orderBy: 'order'
    },
    extend: {}
};

// reqModel 기본값
const DEFAULT_REQ_MODEL = '';

// resModel 기본값
const DEFAULT_RES_MODEL = '';

// AJAX DEFAULT RES MODEL
const DEFAULT_AJAX_MODEL_RES_MODEL = {
    totalCount: 'totalCount',
    items: 'data'
};

// AJAX OPTION
const AJAX_OPTION = {
    READ_OPTION: 'readOption',
    CREATE_OPTION: 'createOption',
    UPDATE_OPTION: 'updateOption',
    DELETE_OPTION: 'deleteOption',
    EXPORT_OPTION: 'exportOption',
    IMPORT_OPTION: 'importOption'
};

// FORMATTER TYPE
const FORMAT_TYPE = {
    CURRENCY: 'currency',
    DATE: 'date',
    TEL: 'tel',
    PID: 'pid',
    BZNO: 'bzno',
    CARD: 'card',
    CUSTOM: 'custom'
};

// EXPORT TYPE
const EXPORT_MODEL = {
    "excel": {
        "type": 'excel',
        "name": '엑셀',
        "action": 'export'
    },
    "downloadText": '다운로드',
    "dataFormat": {
        "plain": 'plain',
        "formatted": 'formatted'
    }
};

// EXPORT TYPE
const IMPORT_MODEL = {
    "excel": {
        "type": 'excel',
        "name": '엑셀',
        "action": 'import'
    },
    "uploadText": '업로드'
};

// error MSG
const ERR_MSG = {
    minWidth: '넓이의 최소 값은 100 이상이어야 합니다.',
    emptySelectedId: '선택된 행이 없습니다.',
    pageParam: 'page 는 1이상 값을 인자값으로 주셔야 합니다.',
    notSupportGetChangedDataFunc: 'ajax 를 사용하실 경우, getChangedData 를 지원하지 않습니다.',
    notSupportResetGridDataFunc: 'ajaxModel을 설정한 상태에서는 resetGridData 를 사용할 수 없습니다. loadData() 함수를 사용해주세요',
    notSupportLoadDataFunc: 'loadData() 함수는 ajaxModel 설정시에 사용 가능합니다. resetGridData() 함수를 사용해주세요.',
    notSupportEnableStartEditMode: 'ajaxModel 설정시 enableStartEditMode 를 사용할 수 없습니다.',
    notFoundItemViaResModel: '전달받은 item 이, resModel 구조와 일치하지 않습니다.',
    notFoundTotalCountViaResModel: '전달받은 totalCount 가, resModel.totalCount 구조와 일치하지 않습니다.'
};

class BxGrid {

    constructor({
        // basic option
        height = 0,
        width = null,
        resizable = false,
        colModel = [],
        items = [],
        rowNumbers = false,
        rowNumWidth = 25,
        caption = '',
        sortName = '',
        sortOrder = 'desc',
        multiSort = false,
        multiSelect = false,
        enablePaging = false,
        $pageContainer = null,
        rowNum = 0,
        pageGroupSize = 10,
        ajaxModel = {},
        enableStartEditMode = false,
        // export options
        exportModel = {},
        importModel = {},
        // inlineEdit option
        inlineEditCol = false,
        inlineEditColModel = {},
        inlineEditColLabel = '편집',
        // grouping 기능은 현재 디자인(아이콘 등) 부재로 인해, 차기 버전에서 추가 시킨다.
        grouping = false,
        groupingView = {},
        groupHeaders = [],
        // footer
        footerrow = false,
        userDataOnFooter = false,
        footers = null,
        // @deprecated.
        title = null,
        // @deprecated.
        isPaging = null,
        // @deprecated.
        isLoadingBar = null,
        // @deprecated.
        sortname = null,
        // @deprecated.
        sortorder = null,
        // inline edit function
        onRowAction = function (){},
        onBeforeRowAction = null,
        onInlineSave = function (){},
        onInlineEdit = function (){},
        onInlineCancel = function (){},
        onInlineDelete = function (){},
        onSelectCell = function (){},
        // grid event callback
        onSelectRow = function (){},
        onDblClick = function (){},
        onReload = function (){},
        onLoadComplete = function (){},
        onRender = function (){}} = {}){

        // basic option validation
        if (!_.isNumber(height)){
            CONSOLE.prop.throw(BxGrid, 'height', 'type');
        }

        if (!_.isBoolean(resizable)){
            CONSOLE.prop.throw(BxGrid, 'resizable', 'type');
        }

        if (height <= 0){
            height = rowNum * STYLE_VALUE.ROW_HEIGHT;
        }

        if (!_.isNull(width)){

            if (!_.isNumber(width)){
                CONSOLE.prop.throw(BxGrid, 'width', 'type');
            }

            if (width < 100){
                CONSOLE.prop.throw(BxGrid, ERR_MSG.minWidth);
            }
        }

        if (!_.isArray(colModel)){
            CONSOLE.prop.throw(BxGrid, 'colModel', 'type');
        }

        if (_.isEmpty(colModel[0]) || !_.isPlainObject(colModel[0])){
            CONSOLE.prop.throw(BxGrid, 'colModel', 'value');
        }

        if (!_.isArray(items)){
            CONSOLE.prop.throw(BxGrid, 'items', 'type');
        }

        if (!_.isBoolean(rowNumbers)){
            CONSOLE.prop.throw(BxGrid, 'rowNumbers', 'type');
        }

        if (!_.isNumber(rowNumWidth)){
            CONSOLE.prop.throw(BxGrid, 'rowNumWidth', 'type');
        }

        // @deprecated
        if (!_.isNull(sortname)){
            CONSOLE.prop.throw(BxGrid, 'sortname', 'deprecatedProperty');
        }

        if (!_.isString(sortName)){
            CONSOLE.prop.throw(BxGrid, 'sortName', 'type');
        }

        // @deprecated
        if (!_.isNull(sortorder)){
            CONSOLE.prop.throw(BxGrid, 'sortorder', 'deprecatedProperty');
        }

        if (!_.isString(sortOrder)){
            CONSOLE.prop.throw(BxGrid, 'sortOrder', 'type');
        }

        // @deprecated
        if (!_.isNull(title)){
            CONSOLE.prop.throw(BxGrid, 'title', 'deprecatedProperty');
        }

        if (!_.isBoolean(multiSort)){
            CONSOLE.prop.throw(BxGrid, 'multiSort', 'type');
        }

        if (!_.isBoolean(multiSelect)){
            CONSOLE.prop.throw(BxGrid, 'multiSelect', 'type');
        }

        // @deprecated
        if (!_.isNull(isLoadingBar)){
            CONSOLE.prop.throw(BxGrid, 'isLoadingBar', 'deprecatedProperty');
        }

        // @deprecated
        if (!_.isNull(isPaging)){
            CONSOLE.prop.throw(BxGrid, 'isPaging', 'deprecatedProperty');
        }

        if (!_.isBoolean(enablePaging)){
            CONSOLE.prop.throw(BxGrid, 'enablePaging', 'type');
        }

        if (enablePaging && !_.isObject($pageContainer)){
            CONSOLE.prop.throw(BxGrid, '$pageContainer', 'all');
        }

        if (enablePaging && !_.isNumber(rowNum)){
            CONSOLE.prop.throw(BxGrid, 'rowNum', 'type');
        }

        if (enablePaging && rowNum <= 0){
            CONSOLE.prop.throw(BxGrid, 'rowNum', 'value');
        }

        if (enablePaging && !_.isNumber(pageGroupSize)){
            CONSOLE.prop.throw(BxGrid, 'pageGroupSize', 'type');
        }

        if (enablePaging && pageGroupSize < 0){
            CONSOLE.prop.throw(BxGrid, 'pageGroupSize', 'value');
        }

        if (!enablePaging){
            rowNum = -1;
        }

        if (footers){

            if (!_.isArray(footers)){
                CONSOLE.prop.throw(BxGrid, 'footers', 'type');

            }
            else {
                footerrow = true;
                userDataOnFooter = true;
            }
        }

        if (!_.isPlainObject(ajaxModel)){
            CONSOLE.prop.throw(BxGrid, 'ajaxModel', 'type');
        }

        if (!_.isBoolean(enableStartEditMode)){
            CONSOLE.prop.throw(BxGrid, 'enableStartEditMode', 'type');
        }
        else if (enableStartEditMode && !_.isEmpty(ajaxModel)){
            CONSOLE.prop.throw(BxGrid, ERR_MSG.notSupportEnableStartEditMode);
        }

        if (!_.isPlainObject(importModel)){
            CONSOLE.prop.throw(BxGrid, 'importModel', 'type');
        }

        if (!_.isPlainObject(exportModel)){
            CONSOLE.prop.throw(BxGrid, 'exportModel', 'type');
        }

        // inline Edit option validate
        if (!_.isBoolean(inlineEditCol)){
            CONSOLE.prop.throw(BxGrid, 'inlineEditCol', 'type');
        }

        if (inlineEditCol && !_.isPlainObject(inlineEditColModel)){
            CONSOLE.prop.throw(BxGrid, 'inlineEditColModel', 'type');
        }

        if (inlineEditCol && _.isEmpty(inlineEditColModel)){
            CONSOLE.prop.throw(BxGrid, 'inlineEditColModel', 'value');
        }

        if (inlineEditCol && !_.isString(inlineEditColLabel)){
            CONSOLE.prop.throw(BxGrid, 'inlineEditColLabel', 'type');
        }

        if (inlineEditCol && _.isEqual(inlineEditColLabel.trim(), '')){
            CONSOLE.prop.throw(BxGrid, 'inlineEditColLabel', 'value');
        }

        if (!_.isArray(groupHeaders)){
            CONSOLE.prop.throw(BxGrid, 'groupHeaders', 'type');
        }

        if (!_.isFunction(onRowAction)){
            CONSOLE.prop.throw(BxGrid, 'onRowAction', 'type');
        }

        if (!(_.isFunction(onBeforeRowAction) || _.isNull(onBeforeRowAction))){
            CONSOLE.prop.throw(BxGrid, 'onBeforeRowAction', 'type');
        }

        if (!_.isFunction(onInlineSave)){
            CONSOLE.prop.throw(BxGrid, 'onInlineSave', 'type');
        }

        if (!_.isFunction(onInlineEdit)){
            CONSOLE.prop.throw(BxGrid, 'onInlineEdit', 'type');
        }

        if (!_.isFunction(onInlineCancel)){
            CONSOLE.prop.throw(BxGrid, 'onInlineCancel', 'type');
        }

        if (!_.isFunction(onInlineDelete)){
            CONSOLE.prop.throw(BxGrid, 'onInlineDelete', 'type');
        }

        if (!_.isFunction(onSelectRow)){
            CONSOLE.prop.throw(BxGrid, 'onSelectRow', 'type');
        }

        if (!_.isFunction(onSelectCell)){
            CONSOLE.prop.throw(BxGrid, 'onSelectCell', 'type');
        }

        if (!_.isFunction(onDblClick)){
            CONSOLE.prop.throw(BxGrid, 'onDblClick', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxGrid, 'onRender', 'type');
        }

        if (!_.isFunction(onReload)){
            CONSOLE.prop.throw(BxGrid, 'onReload', 'type');
        }

        if (!_.isFunction(onLoadComplete)){
            CONSOLE.prop.throw(BxGrid, 'onLoadComplete', 'type');
        }

        this[OPTS] = Object.assign({}, _getJqGridInitOption({
            width,
            height,
            colModel,
            sortName,
            sortOrder,
            multiSort,
            multiSelect,
            rowNum,
            caption,
            grouping,
            groupingView,
            rowNumbers,
            rowNumWidth,
            items,
            userDataOnFooter,
            footerrow
        }));

        Object.assign(this, {
            // $root: BxGrid.$root,
            $root: BxGrid.$root,
            movePage: 1,
            $pageContainer,
            enablePaging,
            pageGroupSize,
            ajaxModel,
            title,
            width,
            action: 'read',
            originTblWidth: null,
            enableStartEditMode,
            targetColModelValues: {},
            changedObj: {},
            editFail: false,
            preSelectRow: null,
            exportModel,
            importModel,
            inlineEditCol,
            inlineFunc: null,
            resizable,
            groupHeaders,
            footers,
            // edit event callback
            onRowAction,
            onBeforeRowAction,
            onInlineEdit,
            onInlineDelete,
            onInlineSave,
            onInlineCancel,
            // render evnet callback
            onReloadComplete: onReload,
            onRender: onRender,
            onLoadComplete: onLoadComplete,
            // click evnet callback
            onDblClick,
            onSelectRow,
            onSelectCell
        });

        // importModel 이 존재하는 경우, 초기 셋팅
        if (!_.isEmpty(importModel)){

            if (!_.isBoolean(importModel.enableImpButton)) {
                importModel.enableImpButton = false;
            }

            if (!_.isBoolean(importModel.enableResetData)){
                importModel.enableResetData = true;
            }

            if (!_.isPlainObject(importModel.ajaxOption)) {
                importModel.ajaxOption = {};
            }

            importModel.type = IMPORT_MODEL.excel.type;

            // ajax 옵션 초기화
            _initAjaxOption.call(this, AJAX_OPTION.IMPORT_OPTION);
        }

        // exportModel 이 존재하는 경우, 초기 셋팅
        if (!_.isEmpty(exportModel)){

            if (_.isUndefined(EXPORT_MODEL[exportModel.type])){
                exportModel.type = EXPORT_MODEL.excel.type;
            }

            if (!_.isString(exportModel.fileName)){
                exportModel.fileName = '';
            }

            if (!_.isString(exportModel.sheetName)){
                exportModel.sheetName = '';
            }

            if (!_.isBoolean(exportModel.enableTimeStamp)){
                exportModel.enableTimeStamp = false;
            }

            if (!_.isBoolean(exportModel.enableExpButton)){
                exportModel.enableExpButton = false;
            }

            if (_.isUndefined(EXPORT_MODEL.dataFormat[exportModel.dataFormat])){
                exportModel.dataFormat = EXPORT_MODEL.dataFormat.formatted;
            }

            if (!_.isPlainObject(exportModel.ajaxOption)){
                exportModel.ajaxOption = {};
            }

            // ajax 옵션 초기화
            _initAjaxOption.call(this, AJAX_OPTION.EXPORT_OPTION);
        }

        // ajaxModel 이 존재하는 경우, 초기 셋팅
        if (!_.isEmpty(ajaxModel)){

            ajaxModel.isAjaxTransaction = true;

            ajaxModel.extendParams = {};

            // ajaxSortList 초기화
            _setAjaxSortList.call(this, sortName, sortOrder);

            // ajax 옵션 초기화(CRUD)
            _initAjaxOption.call(this, AJAX_OPTION.CREATE_OPTION);
            _initAjaxOption.call(this, AJAX_OPTION.READ_OPTION);
            _initAjaxOption.call(this, AJAX_OPTION.UPDATE_OPTION);
            _initAjaxOption.call(this, AJAX_OPTION.DELETE_OPTION);

            // reqModel 기본값 설정
            if (!_.isString(ajaxModel.reqModel)){
                ajaxModel.reqModel = DEFAULT_REQ_MODEL;
            }

            _initAjaxResModel.call(this);

            // Ajax 오브젝트가 있는 경우 jqGrid의 기본설정을 진행한다
            this[OPTS].datatype = 'jsonstring';
        }

        _expandColModel.call(this, {
            inlineEditCol,
            inlineEditColModel,
            inlineEditColLabel
        });
    }

    /**
     * 렌더 함수
     *
     * @returns {object} 생성된 인스턴스
     * */
    render(){

        // Render 시점에서 Container 생성 후 Root에 Append
        this.$container = BASE.appendContainerElement(this.$root, BxGrid.className, $('<table></table>'));

        // 생성된 인스턴스를 $container 엘리먼트의 __this__ 내부 프로퍼티로 할당해둔다.
        this.$container.get(0).__this__ = this;

        // _render 함수에 위임한다.
        return _render.call(this);
    }

    /**
     * 단일, 다중 행 데이터 반환 함수<br/>
     * 편집 모드에서 데이터 가져오는 경우, element 의 data-id 와 값으로 매칭된다.
     *
     * @param {string | string[]} [ids=null]- row 아이디
     *
     * @returns {object|object[]} 아이템
     */
    getRowData(ids = null){

        let o = null,
            $container = this.$container;

        if (!_.isString(ids) && !_.isArray(ids)){
            CONSOLE.prop.throw(BxGrid, 'ids', 'type');
        }

        if (_.isArray(ids)){
            if (!_.isString(ids[0])){
                CONSOLE.prop.throw(BxGrid, 'ids', 'value');
            }
            o = [];
            ids.map((id) =>{
                let obj = $container.getRowData(id);
                if (!_.isEmpty(obj)){
                    delete obj.format;
                    if (_isEditable.call(this, id)){
                        _.assign(obj, _getEditData.call(this, id));
                    }
                    //
                    o.push(obj);
                }
            });
        }
        else{
            if (ids.trim() === ''){
                CONSOLE.prop.throw(BxGrid, 'ids', 'value');
            }
            let id = ids;
            o = $container.getRowData(ids);

            delete o.format;
            // 소스 추가
            if (_isEditable.call(this, id)){
                _.assign(o, _getEditData.call(this, id));
            }
        }

        return o;
    }

    /**
     * 행 데이터 재설정 함수
     *
     * @param {} item
     * @param {object} item.data - 아이템 객체
     * @param {object} [item.css={}] - css 객체
     * @param {string} item.id - row 아이디
     *
     * @returns {object} 생성된 인스턴스
     */
    setRowData({data = {}, css = {}, id = ''} = {}){

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxGrid, 'id', 'type');
        }

        if (id.trim() === ''){
            CONSOLE.prop.throw(BxGrid, 'id', 'value');
        }

        if (!_.isPlainObject(data)){
            CONSOLE.prop.throw(BxGrid, 'data', 'type');
        }

        if (_.isEmpty(data)){
            CONSOLE.prop.throw(BxGrid, 'data', 'value');
        }

        if (!_.isEmpty(css) && !_.isPlainObject(css)){
            CONSOLE.prop.throw(BxGrid, 'css', 'type');
        }

        let optionName = AJAX_OPTION.UPDATE_OPTION;

        if (_isAjaxTransaction.call(this)){

            let changedData = this.getRowData(id);

            _.assign(changedData, data);

            let params = _getRequestParams.call(this, optionName, changedData);

            _transactionCRUD.call(this, optionName, params).then(req => {
                this.goPage(this.getGridParam('page'));
            });
        }
        else {
            this.$container.setRowData(id, data, css);
            _changedDataSync.call(this, id, 'update');
        }

        return this;
    }
    /**
     * 행 추가 함수
     *
     * @param {object|array} data - 아이템 객체
     *
     * @returns {object} 생성된 아이디 집합
     * @example
     *
     * //enableStartEditMode: false
     * grid.addRowDatas({
     *  "Name":"newData",
     *  "logCode":"CODE001",
     *  "logCodeName": "BLUE",
     *  "Salary":"0",
     *  "stateCode": "200",
     *  "stateCodeName": "OK",
     *  "testCode": "1",
     *  "testCodeName": "APPLE"
     * });
     *
     * //enableStartEditMode: true
     *  grid.addRowDatas();
     */
    addRowDatas(data = {}){

        let isEditMode = this.enableStartEditMode;

        if (!isEditMode && !arguments.length){
            CONSOLE.prop.throw(BxGrid, 'data', 'value');
        }

        if (!_.isPlainObject(data) && !_.isArray(data)){
            CONSOLE.prop.throw(BxGrid, 'data', 'type');
        }

        if (!isEditMode && _.isEmpty(data)){
            CONSOLE.prop.throw(BxGrid, 'data', 'value');
        }

        let $container = this.$container;
        let optionName = AJAX_OPTION.CREATE_OPTION;
        let ret = [];

        this.action = optionName.replace('Option', '');

        data = _.isPlainObject(data) ? [data] : data;

        /**
        // 반환되는 id 값은, 현재 중요도가 떨어지는 부분이므로, ajax 상황일때는 이후 작업에서 처리하기로 한다.
        */
        if (!isEditMode && _isAjaxTransaction.call(this)){

            let params = {items: data};

            params = _getRequestParams.call(this, optionName, params);

            _transactionCRUD.call(this, optionName, params).then(() => {
                this.goPage(this.getGridParam('page'));
            });
        }
        else if (isEditMode){

            _.forEach(data, (v, k) =>{
                // id 값을 생성한다.
                ret.push(_getLastId.call(this));

                $container.addRowData(ret[k], v);
                $container.editRow(ret[k]);
                $container.setSelection(ret[k]);

            });

            this.reloadGrid();
        }
        else{

            _.forEach(data, (v, k) =>{

                ret.push(_getLastId.call(this));

                $container.addRowData(ret[k], v);

                _changedDataSync.call(this, ret[k], 'insert');
            });

            this.reloadGrid();
        }

        return {ids: ret};
    }

    /**
     * 단일, 다중 행 삭제 함수
     * @param {string|string[]} [ids = null] row 아이디 or row 아이디 배열
     *
     * @returns {Array} 생성된 아이디 집합
     */
    delRowData(ids = null){

        if (!_.isString(ids) && !_.isArray(ids)){
            CONSOLE.prop.throw(BxGrid, 'ids', 'type');
        }

        ids = _.isString(ids) ? [ids] : ids;

        let optionName = AJAX_OPTION.DELETE_OPTION;
        this.action = optionName.replace('Option', '');

        // 아이디 반환 값
        let ret = [];
        // 아이템 배열
        let items = [];

        _.forEach(ids, (v, k) =>{

            let data = this.getRowData(v);

            if (!_.isEmpty(data)){

                items.push(data);

                if (_isAjaxTransaction.call(this)){


                    if (ids.length === (k + 1)){

                        let params = {items: items};

                        params = _getRequestParams.call(this, optionName, params);

                        _transactionCRUD.call(this, optionName, params).then(() =>{
                            this.goPage(this.getGridParam('page'));
                            _disableAllSelected.call(this);
                        });
                    }
                }
                else{

                    _changedDataSync.call(this, v, 'delete');

                    this.$container.delRowData(v);

                }

                ret.push(v);
            }
        });


        if (!_isAjaxTransaction.call(this)){

            this.reloadGrid();

            _disableAllSelected.call(this);
        }

        return {ids: ret};
    }

    /**
     * 그리드 리로드 함수
     *
     * @returns {object} 생성된 인스턴스
     * @private
     */
    reloadGrid(){

        this.$container.trigger('reloadGrid');

        if (this.enableStartEditMode){

            let $container = this.$container;
            let items = this.getPageData($container.getGridParam('page'));

            for (let i in items){

                let id = items[i].id;

                $container.editRow(id);
                $container.setSelection(id);

                _disableAllSelected.call(this);
            }
        }

        return this;
    }

    /**
     * 그리드의 파라메터 설정 함수
     *
     * @param {object} param - 파라메터 객체
     * @returns {object} 생성된 인스턴스
     * @private
     */
    setGridParam(param = {}){

        if (!_.isPlainObject(param)){
            CONSOLE.prop.throw(BxGrid, 'param', 'type');
        }

        if (_.isEmpty(param)){
            CONSOLE.prop.throw(BxGrid, 'param', 'value');
        }

        this.$container.setGridParam(param);
        this.reloadGrid();

        return this;
    }

    /**
     * 그리드 파라메터 반환 함수
     * @param {string} [param=''] 그리드 파마리터
     * @returns {object} 해당 그리드 파라메터 데이터 및 객체
     * @private
     */
    getGridParam(param = ''){

        if (!_.isString(param)){
            CONSOLE.prop.throw(BxGrid, 'param', 'type');
        }

        if (param.trim() === ''){
            CONSOLE.prop.throw(BxGrid, 'param', 'value');
        }

        return this.$container.getGridParam(param);
    }

    /**
     * 페이지 이동 함수
     *
     * @param {int} [pageNumber=1]- page number
     *
     * @returns {object} 생성된 인스턴스
     *
     */
    goPage(pageNumber = 1){

        let params = {};

        if (!_.isNumber(pageNumber)){
            CONSOLE.prop.throw(BxGrid, 'pageNumber', 'type');
        }

        if (pageNumber < 1){
            CONSOLE.prop.throw(BxGrid, 'pageNumber', 'value');
        }

        let totalPage = this.getGridParam('total');

        // 전달받은 페이지값이 전체 페이지보다 클 경우..
        // 전체 페이지를 넘어갈 수 없게만든다
        if (pageNumber > totalPage){
            pageNumber = totalPage;
        }

        let optionName = AJAX_OPTION.READ_OPTION;

        this.action = optionName.replace('Option', '');

        if (_isAjaxTransaction.call(this)){

            let defer = $.Deferred();

            this.movePage = pageNumber;

            params = _getRequestParams.call(this, optionName);

            _transactionCRUD.call(this, optionName, params).then((data) => {

                _setResponseDataToGridParam.call(this, data, false);

                _disableAllSelected.call(this);

                return defer.resolve();
            });

            return defer.promise();
        }
        else {

            params = {page: pageNumber};

            this.setGridParam(params);
            this.movePage = pageNumber;
        }

        return this;
    }

    /**
     * 전체 데이터 반환 함수
     *
     * @return {object[]} 그리드 전체 데이터
     *
     * @example
     * // ajax 사용시 (Promise)
     * grid.getAllRowData().then(function(res){
     *      console.log(res);
     * });
     *
     * // ajax 미 사용시
     * console.log(grid.getAllRowData());
     */
    getAllRowData(){

        let optionName = AJAX_OPTION.READ_OPTION;

        if (_isAjaxTransaction.call(this)){

            this.action = optionName.replace('Option', '');

            let defer = $.Deferred();

            let params = _getRequestSortParams.call(this);

            _transactionCRUD.call(this, optionName, params).then(res =>{
                defer.resolve(_.get(res, this.ajaxModel.resModel.items));
            });

            return defer.promise();
        }
        else{

            // 정렬된 데이터 기준
            // this.$container.jqGrid('getGridParam', 'lastSelectedData');

            // 정렬되지 않은 데이터 기준
            // this.$container.jqGrid('getGridParam', 'data');
            return this.$container.jqGrid('getGridParam', 'lastSelectedData');
        }
    }

    /**
     * 페이지 데이터 반환 함수
     *
     * @param {int} [page = currentPageNumber] 페이지 번호
     * @returns {object[]} 페이지의 데이터
     *
     * @example
     * // ajax 사용시 (Promise)
     * grid.getPageData(page).then(function(res){
     *      console.log(res);
     * });
     *
     * // ajax 미 사용시
     * console.log(grid.getPageData(page));
     */
    getPageData(page = 0){

        if (!_.isNumber(page)){
            CONSOLE.prop.throw(BxGrid, 'page', 'type');
        }

        if (page === 0) {
            page = this.getGridParam('page');
        }

        let optionName = AJAX_OPTION.READ_OPTION;

        this.action = optionName.replace('Option', '');

        if (_isAjaxTransaction.call(this)){

            let defer = $.Deferred();

            let resModel = this.ajaxModel.resModel;
            this.movePage = page;

            let params = _getRequestParams.call(this, optionName);

            _transactionCRUD.call(this, optionName, params).then(res => {
                defer.resolve(_.get(res, resModel.items));
            });

            return defer.promise();
        }
        else {

            let lastPage = this.$container.jqGrid('getGridParam', 'lastpage');

            if (lastPage < page) return [];

            let rowNum = this[OPTS].rowNum;

            let startIdx = (page - 1) * rowNum;
            // enablePaging: false 면서 rowNum -1 인 경우 문제가 발생해 추가 수정
            let endIdx = this[OPTS].enablePaging ? startIdx + rowNum : this.getGridParam('data').length;

            return this.getAllRowData().slice(startIdx, endIdx);
        }
    }

    /**
     * 그리드 넓이 지정 함수
     * @param {int}[width=100] 넓이
     * @returns {object} 생성된 인스턴스
     */
    setGridWidth(width = 100){

        if (!_.isNumber(width)){
            CONSOLE.prop.throw(BxGrid, 'width', 'type');
        }

        if (width < 100){
            CONSOLE.prop.throw(BxGrid, 'width', 'value');
        }

        const HEADER_WRAPPER_CLASS = '.ui-state-default.ui-corner-top.ui-jqgrid-hdiv',
            HEADER_SUB_WRAPPER_CLASS = '.ui-jqgrid-hbox',
            BODY_WRAPPER_CLASS = '.ui-jqgrid-bdiv',
            FOOTER_WRAPPER_CLASS = '.ui-jqgrid-sdiv',
            FOOTER_SUB_WRAPPER_CLASS = '.ui-jqgrid-hbox',
            FOOTER_SUB_WRAPPER_TR_CLASS = 'tr.footrow-ltr',
            FOOTER_REMOVE_BORDER = 'border-right: 0px !important';

        let { $root, $container, $pageContainer, inlineEditCol} = this,
            $headerWrapper = $root.find(HEADER_WRAPPER_CLASS),
            $headerSubWrapper = $headerWrapper.find(HEADER_SUB_WRAPPER_CLASS),
            $bodyWrapper = $root.find(BODY_WRAPPER_CLASS),
            $footerWrapper = $root.find(FOOTER_WRAPPER_CLASS),
            $footerSubWrapper = $footerWrapper.find(FOOTER_SUB_WRAPPER_CLASS),
            // $footerSubWrapperTr = $footerSubWrapper.find(FOOTER_SUB_WRAPPER_TR_CLASS),
            scrollBarSize = UTIL.getElementScrollBarSize($bodyWrapper);

        if (scrollBarSize < 3){
            scrollBarSize = 0;
        }
        else{
            scrollBarSize = scrollBarSize - 2;
        }

        // 탭키로 스크롤을 움직일수 있어야 한다.(웹접근성)
        $bodyWrapper.attr('tabindex', 0);

        if (!this[OPTS].shrinkToFit){
            _setHorizonScroll.call(this, width);
        }
        else{

            const optimizationPixel = 2;

            width -= optimizationPixel;

            $container.setGridWidth(width);
            $headerWrapper.css('width', `${width + optimizationPixel}px`);
            $bodyWrapper.css('width', `${width + optimizationPixel}px`);
            $footerWrapper.css('width', `${width + optimizationPixel}px`);

            $pageContainer && $pageContainer.css('width', width + optimizationPixel);
        }

        $headerSubWrapper.css('padding-right', scrollBarSize);
        $footerSubWrapper.css('padding-right', scrollBarSize);

        _resizeFooter.call(this);

        return this;
    }

    /**
     * 그리드 높이 지정 함수
     * @param {int}[height=100] 높이
     * @returns {object} 생성된 인스턴스
     */
    setGridHeight(height = 100){

        if (!_.isNumber(height)){
            CONSOLE.prop.throw(BxGrid, 'height', 'type');
        }

        if (height < 100){
            CONSOLE.prop.throw(BxGrid, 'height', 'value');
        }

        this.$container.setGridHeight(height);

        return this;
    }

    /**
     * 그리드의 변경 데이터 반환 함수
     * @return {object} 추가, 변경 삭제에 대한 row data 값 반환
     */
    getChangedData(){

        if (_isAjaxTransaction.call(this)){
            CONSOLE.prop.throw(BxGrid, ERR_MSG.notSupportGetChangedDataFunc, 'empty');
        }

        return this.changedObj.data;
    }

    /**
     * inline Edit Mode 실행 함수
     * @param {string} id='' row 아이디
     * @returns {object} 생성된 인스턴스
     */
    editRow(id = ''){

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxGrid, 'id', 'type');
        }

        if (id.trim() === ''){
            CONSOLE.prop.throw(BxGrid, 'id', 'value');
        }

        let $container = this.$container;

        if (!this.enableStartEditMode && !this[OPTS].multiselect){

            // this.setGridParam({selrow: ''});
            this.cancelSelectedRow();
        }

        if (!this.enableStartEditMode && this.getSelectedId().length !== 0){
            this.cancelSelectedRow();
        }

        $container.editRow(id);
        $container.setSelection(id);

        let $firstSelectedRow = $($container.find(".ui-state-highlight:first"));
        $firstSelectedRow.find("input[type='text']:first").focus();

        if (!this[OPTS].multiselect){
            this.preSelectRow = id;
        }

        _disableAllSelected.call(this);

        return this;
    }

    /**
     * inline Edit Mode 을 통해 변경된 데이터 값 저장 함수
     * @param {string} id='' row 아이디
     * @returns {object} 생성된 인스턴스
     */
    saveRow(id = ''){

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxGrid, 'id', 'type');
        }

        if (id.trim() === ''){
            CONSOLE.prop.throw(BxGrid, 'id', 'value');
        }

        let $container = this.$container;
        let defer = $.Deferred();

        let isAjaxTransaction = _isAjaxTransaction.call(this);

        let ret = isAjaxTransaction ? defer.promise() : this;

        this.editFail = false;

        // 현재 보여지는 페이지의 데이터를 가져와 바뀌기 이전값에 대한 정보를 가져온다.
        let pageData = isAjaxTransaction ? this.getGridParam('data') : this.getPageData(this.getGridParam('page')),
            preData = {},
            targetObj = {};

        for (let i in pageData){

            if (pageData[i].id === id){

                targetObj = pageData[i];
                preData = JSON.parse(JSON.stringify(targetObj).replace('null', '""'));
                preData.id = String(preData.id);
                break;
            }
        }

        $container.saveRow(id);
        // 편집 버튼으로 인해 자동으로 동기화 되는 영역을 삭제 해준다.
        // 삭제하지 않을시, getPageData 부분에서 undefined 속성이 추가 되어 나오게 된다.
        // if (this.inlineEditCol) delete targetObj.format;
        let data = this.getRowData(String(id));

        if (_.isEmpty(data)) return ret;

        if (this.editFail){
            return isAjaxTransaction ? defer.resolve(this) : ret;
        }

        let targetColModelValues = this.targetColModelValues;
        let optionName = AJAX_OPTION.UPDATE_OPTION;

        this.action = optionName.replace('Option', '');

        if (isAjaxTransaction){

            if (_.size(targetColModelValues)){
                _.assign(data, targetColModelValues);
            }

            this.$container.setRowData(id, preData);

            let selectedId = this.getSelectedId();
            let editableRow = [];

            for (let i in selectedId){
                if ($(`tr#${selectedId[i]} .editable`).length > 0){
                    editableRow.push(selectedId[i]);
                }
            }

            let params = _getRequestParams.call(this, optionName, data);

            _transactionCRUD.call(this, optionName, params).then(() => {

                if (editableRow.length === 0){

                    this.goPage(this.getGridParam('page')).then(()=>{
                        return defer.resolve(this);
                    });
                }
                else {

                    this.$container.setRowData(id, data);
                }
            });

            _disableAllSelected.call(this);

            return ret;
        }
        else {

            if (_.size(targetColModelValues)){
                // 데이터 변경 처리
                this.setRowData({id: id, data: targetColModelValues});
            }

            _changedDataSync.call(this, id, 'update', preData);
        }

        if (this.getSelectedId().indexOf(id) !== -1){
            $container.setSelection(id, false);
        }

        _disableAllSelected.call(this);

        return ret;
    }

    /**
     * inline Edit Mode 취소 함수
     * @param {string} id='' - row 아이디
     * @returns {object} 생성된 아이디 집합
     */
    cancelRow(id = ''){

        if (!_.isString(id)){
            CONSOLE.prop.warn('id', 'type');
            return;
        }

        if (_.isEqual(id, '')){
            CONSOLE.prop.warn('id', 'value');
            return;
        }

        let $container = this.$container;

        $container.restoreRow(id);

        if (this.getSelectedId().indexOf(id) !== -1){
            $container.setSelection(id);
        }


        _disableAllSelected.call(this);

        return {id: id};
    }

    /**
     * 단일, 다중 행 선택 함수
     *
     * @param {string | array} ids - row 아이디 or row 아이디 배열
     * @returns {object} - 생성된 아이디 집합
     */
    selectRows(ids = ''){

        if (!_.isString(ids) && !_.isArray(ids)){
            CONSOLE.prop.throw(BxGrid, 'ids', 'type');
        }

        if (_.isString(ids) && _.isEmpty(ids.trim())) {
            CONSOLE.prop.throw(BxGrid, 'ids', 'value');
        }

        ids = [].concat(ids);

        _setSelection.call(this, ids, 'select');

        return {ids};
    }

    /**
     * 단일, 다중 행 선택 해제 함수
     *
     * @param {string | array} ids='' - row 아이디 or row 아이디 배열
     * @returns {object} - 생성된 아이디 집합
     */
    clearSelectedRows(ids = ''){

        if (!_.isString(ids) && !_.isArray(ids)){
            CONSOLE.prop.throw(BxGrid, 'ids', 'type');
        }

        if (_.isString(ids) && _.isEmpty(ids.trim())) {
            CONSOLE.prop.throw(BxGrid, 'ids', 'value');
        }

        ids = [].concat(ids);

        _setSelection.call(this, ids, 'clear');

        return {ids};
    }

    /**
     * 선택된 행 inline Edit Mode 변경 함수
     * @returns {object} 생성된 아이디 집합
     */
    editSelectedRow(){

        const $container = this.$container;
        const selectedId = this.getSelectedId().slice();
        const $firstSelectedRow = $($container.find(".ui-state-highlight:first"));

        if (_.isEmpty(selectedId)){
            CONSOLE.prop.throw(BxGrid, ERR_MSG.emptySelectedId, 'empty');
        }

        selectedId.forEach((id)=>{
            $container.editRow(id);
        });

        $firstSelectedRow.find("input[type='text']:first").focus();

        _disableAllSelected.call(this);

        return {ids : selectedId};
    }

    /**
     * 선택된 행 inline Edit Mode 취소 함수
     * @returns {object} 생성된 아이디 집합
     */
    cancelSelectedRow(){

        let $container = this.$container;
        let selectedId = this.getSelectedId().slice();


        selectedId.map((id)=>{
            $container.restoreRow(id);
        });

        _disableAllSelected.call(this);

        $container.resetSelection();

        return {ids : selectedId};
    }

    /**
     * 현재 페이지의 모든 행에 대한 inline Edit Mode 저장 함수
     * @returns {object} 생성된 아이디 집합
     */
    saveSelectedRow(){

        let selectedId = this.getSelectedId().slice();

        if (_.isEmpty(selectedId)){
            CONSOLE.prop.throw(BxGrid, ERR_MSG.emptySelectedId, 'empty');
        }

        this.editFail = false;

        for (let i in selectedId){

            let id = selectedId[i];

            if (this.editFail){
                break;
            }

            this.saveRow(id);

        }

        return {ids : selectedId};
    }

    /**
     * 선택된 행 삭제 함수
     * @returns {object} 생성된 아이디 집합
     */
    deleteSelectedRow(){

        let selectedId = this.getSelectedId().slice();

        this.delRowData(selectedId);

        return {ids : selectedId};
    }

    /**
     * 선택 된 행 아이디 반환 함수
     * @return {string[]} 선택 된 행 아이디
     */
    getSelectedId(){

        let result = [];

        if (this[OPTS].multiselect){

            result = this.getGridParam('selarrrow');

            // 전체 선택했을때 첫번째 인자로 빈값으로 들어옴
            if(result[0] === '') {
                result.splice(0, 1);
            }
        }
        else{

            result.push(this.getGridParam('selrow'));
        }

        if (!this[OPTS].multiselect && this.getGridParam('selrow') === null){
            result = [];
        }

        return result;
    }

    /**
     * 선택 된 행 데이터 반환 함수
     * @return {object[]} 선택 된 데이터
     */
    getSelectedData(){

        let result = [];
        let selectedIDs = this.getSelectedId();

        result = this.getRowData(selectedIDs);

        return result;
    }

    /**
     * 그리드의 데이터 삭제 함수
     * @returns {object} 생성된 인스턴스
     */
    clearGridData(){

        this.clearChangedData();
        this.$container.clearGridData();

        this.action = AJAX_OPTION.UPDATE_OPTION.replace('Option', '');

        if (_isAjaxTransaction.call(this)){

            this.movePage = 1;

            let resModel = this.ajaxModel.resModel;
            let params = {};

            _.set(params, resModel.items, []);
            _.set(params, resModel.totalCount, 0);

            _setResponseDataToGridParam.call(this, params, false);

        }
        else{
            this.reloadGrid();
        }

        return this;
    }

    /**
     * 그리드 데이터 리셋 함수
     * @param {object[]} [data=[]] 새로 지정할 그리드 데이터
     * @returns {object} 생성된 인스턴스
     */
    resetGridData(data = []){

        if (!_.isArray(data)){
            CONSOLE.prop.throw(BxGrid, 'data', 'type');
        }

        let optionName = AJAX_OPTION.UPDATE_OPTION;

        this.action = optionName.replace('Option', '');

        if (_isAjaxTransaction.call(this)){

            CONSOLE.prop.throw(BxGrid, ERR_MSG.notSupportResetGridDataFunc, 'empty');

            this.movePage = 1;
            let resModel = this.ajaxModel.resModel;
            let params = {};

            _.set(params, resModel.items, data);
            _.set(params, resModel.totalCount, data.length);

            _setResponseDataToGridParam.call(this, params, false);
        }
        else {

            this.clearChangedData();
            this.$container.clearGridData();
            this.action = optionName.replace('Option', '');
            this.setGridParam({data: data});
        }

        return this;
    }

    /**
     * 그리드 changedData 초기화 함수
     * @returns {object} 생성된 인스턴스
     */
    clearChangedData(){
        this.changedObj = {
            id: {
                'insert': [],
                'update': [],
                'delete': []
            },
            data: {
                'insert': [],
                'update': [],
                'delete': []
            }
        };
        return this;
    }

    /**
     * 요청한 parameter 를 기본 param 에 대한 값으로 ajax 를 요청해 결과를 BxGrid 에 반영하는 함수
     *
     * @param {object} [params = {}] - readOption 에 추가할 parameter
     * @returns {object} - 생성된 인스턴스
     */
    loadData(params = {}){

        if (!_isAjaxTransaction.call(this)){
            CONSOLE.prop.throw(BxGrid, ERR_MSG.notSupportLoadDataFunc, 'empty');
        }

        if (!_.isPlainObject(params)){
            CONSOLE.prop.throw(BxGrid, 'param', 'type');
        }

        let defer = $.Deferred();
        let optionName = AJAX_OPTION.READ_OPTION;

        this.movePage = 1;
        this.action = optionName.replace('Option', '');
        this.ajaxModel.extendParams = params;

        params = _getRequestParams.call(this, optionName, params);

        _transactionCRUD.call(this, optionName, params).then(resData => {

            _setResponseDataToGridParam.call(this, resData, false);

            return defer.resolve(resData);
        });

        return defer.promise();
    }

    /**
     * 입력된 행이 InlineEdit Mode 상태의 유무 반환 함수
     * @param {string} id='' - rowId
     * @returns {boolean} - 입력된 rowId 에 해당하는 행이 inLineEdit mode 유무
     */
    isInlineEdit(id = ''){

        let isInlineEdit = false;


        if (!_.isString(id)){
            CONSOLE.prop.throw(BxGrid, 'id', 'type');
        }

        if (_.isEmpty(id)){
            CONSOLE.prop.throw(BxGrid, 'id', 'value');
        }

        isInlineEdit = _isEditable.call(this, id);

        return isInlineEdit;
    }

    getOPTS(){
        return this[OPTS];
    }

    /**
     * 해당하는 컬럼을 지정한 타입으로 정렬 한다
     *
     * @param {string} colName - 컬럼 이름
     * @param {string} [orderType = 'default'] - 정렬 타입 (default, asc, desc)
     * @returns {object} - 생성된 인스턴스
     */
    sort(colName = '', orderType = 'default') {

        if (!_.isString(colName)){
            CONSOLE.prop.throw(BxGrid, 'colName', 'type');
        }

        if (!_.isString(orderType)){
            CONSOLE.prop.throw(BxGrid, 'orderType', 'type');
        }

        let type = '';
        let $container = this.$container;
        let colModel = this[OPTS].colModel;
        let multiSort = this[OPTS].multiSort;

        switch (orderType){
            case 'asc'         :
                type = "asc";
                break;
            case 'desc'        :
                type = "asc-desc";
                break;
            case 'default'     :
                type = "";
                break;
            default            :
                return this;
        }

        for (let i = 0; i < colModel.length; i++){

            if (colModel[i].name === colName){

                $container.setColProp(colName, {lso: type});

                if (multiSort){

                    $container.sortGrid(colName, true);
                }
                else{
                    if (orderType === 'default'){
                        // 싱글일때 default을 지원 않음 onSortCol을 커스텀 사용하는 것 보다 아래 방법이 더 좋음
                        $container.setGridParam({multiSort: true})
                            .sortGrid(colName, true)
                            .setGridParam({multiSort: false});
                    }
                    else{
                        $container.sortGrid(colName, true, orderType);
                    }
                }
            }
        }

        return this;
    }
    /**
     * grid 데이터를 excel 파일로 export 한다
     *
     * @returns {object} - 생성된 인스턴스
     */
    exportExcel(){

        let optionName = AJAX_OPTION.EXPORT_OPTION;
        let ajaxOption = _getAjaxOption.call(this, optionName);

        // exportModel url o
        // exportModel url x, ajaxModel o
        if (ajaxOption.url || _isAjaxTransaction.call(this)){

            let params = _getRequestParams.call(this, optionName);

            if (ajaxOption.url){

	            _transactionCRUD.call(this, optionName, params).then(pageDatas => {
	                _createExportFile.call(this, pageDatas);
	            });
	        }
	        else {

	            _extractPageData.call(this, this.getGridParam('page')).then(pageDatas => {
	                _createExportFile.call(this, pageDatas);
	            });
	        }
        }
        else {

            let pageDatas = _extractPageData.call(this);

            _createExportFile.call(this, pageDatas);
        }

        return this;
    }
}

/**
 * jqGrid 생성 시 초기 옵션 객체 반환 함수
 * @param {int}[width=null] - 가로
 * @param {int} [height=0] - 높이
 * @param {Array} [colModel=[]] - column Model 배열
 * @param {string} [sortName='']- 정렬 할 컬럼 명
 * @param {string} [sortOrder=''] - 정렬하고자 하는 컬럼의 정렬 순서
 * @param {boolean} [multiSort=false] - 다중 정렬 유무
 * @param {boolean} [multiSelect=false] - 다중 선택 유무
 * @param {int} [rowNum=0] - 보여질 row
 * @param {string} [caption=''] - jqGrid Caption 값
 * @param {boolean} [grouping=false] - 그룹핑 유무 [추후 들어갈예정 ]
 * @param {{}} [groupingView={}] - 그룹핑 뷰 객체 [추, 들어갈예정 ]
 * @param {boolean} [rowNumbers=false] - rownumber 보임 유무
 * @param {int} [rowNumWidth=25] - rowNumbers가 true일떄 컬럼의 width 값 지정
 * @param {Array} [items=[]] - grid 에 맵핑될 데이터 배열
 * @returns {{}} o - jqGrid 초기 옵션 객체
 * @private
 */
function _getJqGridInitOption({
    width = null,
    height = 0,
    colModel = [],
    sortName = '',
    sortOrder = '',
    multiSort = false,
    multiSelect = false,
    rowNum = 0,
    caption = '',
    grouping = false,
    groupingView = {},
    footerrow = false,
    userDataOnFooter = false,
    rowNumbers = false,
    rowNumWidth = 25,
    items = []}){

    const DATA_TYPE = 'local',
        LOAD_UI = 'disable',
        ICON_SET = 'fontAwesome';

    let o = {};

    Object.assign(o, {
        autowidth: true,
        // column resize 방지 option [ 컬럼 리사이즈시 계산으로 인한 scroll 생기는 것을 방지한다. ]
        // >> IE에서 colHeader 깨지는 문제가 발생해 이부분 삭제
        shrinkToFit: width ? false : true,
        // 기본 option
        datatype: DATA_TYPE,
        // 최소 생성시에만 데이터 요청을 한다.
        loadonce: true,
        // datatype: 'json',
        viewrecords: true,
        height,
        // grid 가 제공하는 기본 paging 을 사용하지 않는다.
        colModel,
        sortname: sortName,
        sortorder: sortOrder,
        multiSort,
        multiselect: multiSelect,
        pager: false,
        // paging 처리 불가시 rowNum 제한시 bug...
        rowNum,
        caption,
        //grouping option
        grouping,
        groupingView,
        footerrow,
        userDataOnFooter,
        rownumbers: rowNumbers,
        rownumWidth: rowNumWidth,
        data: items,
        loadui: LOAD_UI,
        // 현재 css 가 fontAwesome 기준으로 맞춰져 있기 때문에 아래 프로퍼티값을 제거할 수 없다.
        // 차후 제거 시킨 후, css class 명을 변경해줘야한다.
        iconSet: ICON_SET
    });

    return o;
}
//
// /**
//  * BxGrid this 에 할당할 option 생성 함수.
//  * @param {function} [onReload=function(){}] - 리로드 함수
//  * @param {function} [onRender=function(){}] - 렌더 함수
//  * @returns {{}} option - option object
//  * @private
//  */
// function _getBxGridOptions({
//     onReload = function (){},
//     onRender = function (){}
// } = {}){
//
//     let o = {};
//
//     Object.assign(o, {
//         $root: BxGrid.$root,
//         movePage: 1,
//         targetColModelValues: {},
//         changedObj: {},
//         originTblWidth: null,
//         preSelectRow: null,
//         inlineFunc: null,
//         editFail: false,
//         isAjaxTransaction: false,
//         onReloadComplete: onReload,
//         onLoadComplete: onRender
//     });
//
//     return o;
// }

/**
 * BxGrid 페이지네이션 및 BxGrid 를 렌더링 하는 내부 함수.
 *
 * @memberOf BxGrid
 *
 * @returns {BxGrid}
 * @private
 */
function _render(){

    let $container = this.$container,
        options = this[OPTS];

    options.loadComplete = (data) =>{
        // enablePaging: true -> 페이지네이션
        // enablePaging: false -> 무한스크롤

        data.action = this.action;

        this.onLoadComplete.call(this, data);

        if (data.rows.length === 0 && data.page !== 1){
            window.setTimeout(() =>{
                this.goPage(data.page - 1);
            }, 0);
            return;
        }

        // 전체, 현재 페이지, rownum.
        this.enablePaging && _initPage.call(this, data.records, data.page, this[OPTS].rowNum);

        if (data.userdata) delete data.userdata;
        // setting Param about Total Page
        this.setGridParam({total: data.total});
        // 사용자 reloadcomplete call
        this.onReloadComplete.call(this, data);

        _initFooter.call(this, this.footers);

        _initMacScrollBorder.call(this);

    };

    options.ondblClickRow = (id, dataIndex, colIndex, e) =>{

        let isEditable = _isEditable.call(this, id);

        if (isEditable){
            return;
        }

        let o = this.getRowData(id);

        this.onDblClick.call(this, o, e);
    };

    options.onSelectRow = (id, isSelect, e) =>{

        // multiSelect: false 인 경우, 오른쪽 버튼 누를 경우 call 되는 문제때문에 아래와 같이 처리
        // editCol 의 icon 클릭 시, e 가 undefined 가 들어오므로 이에 대한 처리도 같이한다.
        // if (_.isUndefined(e)) return;
        // if (!_.isEqual(e.type, 'click')) return;
        // console.log('selected!!!!!');

        if (options.multiselect && _isSelectedAllRow.call(this)){
            $('#cb_jqg1')[0].checked = true;
        }

        let isEditable = _isEditable.call(this, id);

        if (isEditable){
            return;
        }

        let o = this.getRowData(id);

        this.onSelectRow.call(this, o, e);
    };

    options.onCellSelect = (rowId, index, content, e) =>{

        let modelIdx = index;

        if (options.rownumbers) --modelIdx;
        if (options.multiselect) --modelIdx;

        let rowData = this.getRowData(rowId);
        let colName = null;

        if (modelIdx > 0) colName = options.colModel[modelIdx].name;
        else if ($(e.target).hasClass('jqgrid-rownum')) colName = 'rowNumbers';
        else colName = 'multiSelect';

        this.onSelectCell.apply(this, [index, content, rowData, colName, e]);

    };


    options.onSortCol = (sortName, colIdx, sortOrder) =>{

        let optionName = AJAX_OPTION.READ_OPTION;

        if (_isAjaxTransaction.call(this)){

            // 초기화
            _setAjaxSortList.call(this, sortName, sortOrder);

            let params = _getRequestParams.call(this, optionName);

            _transactionCRUD.call(this, optionName, params).then((data)=>{

                _setResponseDataToGridParam.call(this, data, false);

            });

            return false;
        }
        else {
            // onSortCol 이 불리기 전, page 가 1로 초기화되는 문제로 인해 해당 부분 추가.
            $container.setGridParam({page: this.movePage});
            return true;
        }
    };

    this.clearChangedData();

    // $.fn.fmatter.rowactions 내부에서 사용되기 위한 함수.
    this.inlineFunc = {

        edit(id, condition){

            let o = _getDataById.call(this, id);
            this.editRow(id);

            this.onInlineEdit.call(this, id, o);
            this.onRowAction.call(this, id, o, condition);

        },
        delete(id, condition){

            let o = _getDataById.call(this, id);

            this.delRowData(id);

            this.onInlineDelete.call(this, id, o);
            this.onRowAction.call(this, id, o, condition);
        },
        save(id, condition){


            if (_isAjaxTransaction.call(this)){

                this.saveRow(id).then(() =>{
                    _save.call(this);
                });
            }
            else {

                this.saveRow(id);

                _save.call(this);
            }

            /*
             * ajax 사용 시, inlineSave 및 rowAction callback 의 인자값의 차이로 아래와 같이 내부 함수를 만들어 사용
             * ajax 사용하는 경우 _getDataById 를 통해 해당 데이터를 구할때
             * 해당 값이 변경이전의 값을 갖고 있어 promise 를 통한 지연을 통해 아래 함수를 실행한다.
             * 이때, colModel 의 key 값을 변경해 해당 값을 가져오려 하였으나, _getDataById, _getLastId 등 내부 함수에서
             * id 컬럼을 사용하기 때문에 id 를 통한 값으로 데이터를 구하는 경우 추가적인 side effect 이 발생할 우려가 있어
             * 우선적으로 colModel 에 key 값 할당에 대한 부분을 doc 에서 제외하였으며,
             * 이 후 추가적으로 작업이 들어가야 한다.
             * */
            function _save(){

                let o = _getDataById.call(this, id);

                this.onInlineSave.call(this, id, o, !this.editFail);

                this.onRowAction.call(this, id, o, condition);
            }
        },
        cancel(id, condition){

            this.cancelRow(id);

            let o = _getDataById.call(this, id);

            this.onInlineCancel.call(this, id, o);
            this.onRowAction.call(this, id, o, condition);
        }
    };

    // 초기 랜더링시 각 action button(수정/삭제)에 바인딩된 이벤트 핸들러인 $.fn.fmatter.rowactions 메서드를 override 하여, 해당 이벤트를 가로챈다.

    // ** 각 버튼 엘리먼트에는 아래와 같은 방식으로 inline event 가 바인딩 되어있다.(this === btn 엘리먼트) **

    /*
     <div id="btn" ... onclick="return jQuery.fn.fmatter.rowactions.call(this, event, 'edit');" ....>
     */

    // 하지만 전역 메서드(rowactions) 내부 self 변수가 참조하는 this 값은, 항상 마지막으로 초기화된 this 값을 참조하게된다(동일 페이지에서 여러개의 컴포넌트를 생성했을 경우)
    // 즉 self 변수를 사용해도, 메서드 내부에서 개별 인스턴스 값을 가질 수 없는 상황이며, 이를 해결해기위해, 초기 $container 엘리먼트의
    // 내부 프로퍼티로 할당한 __this__ 프로퍼티를 사용한다.

    // 수정, 삭제 event

    // 함수 객체가 매번 재 할당된다.
    $.fn.fmatter.rowactions = function (e, condition){

        let $elem = $(this);

        let id = $elem.attr('id').replace(/^\D+_/g, '');

        let componentClassName = BXUI.Base.getComponentClassName('grid');

        // $container 엘리먼트
        let $container = $elem.parents(`.${componentClassName}`);

        // $container 엘리먼트에 할당시킨, 현재 this 값
        let self = $container.get(0).__this__;
        let o = _getDataById.call(self, id);

        if (_.isFunction(self.onBeforeRowAction)) {
            self.onBeforeRowAction.call(self, id, o, condition, execute);
        }
        else {
            execute();
        }

        function execute() {

            self.isInlineEdit = true;

            if (OPTS.multiselect) self.$container.setSelection(id);

            self.inlineFunc[condition].call(self, id, condition);
        }

    };

    // jqGrid 확장 소스....
    // errMassage 는 colModel Index를 전송 >> colModel.editrules 를 설정할때 사용가능
    $.extend($.jgrid, {

        info_dialog: (err = '', errMessage = '', c_b = '', modalopt = {}) =>{
            // editing 제외한 곳에서 다이얼로그 안뜨게 만들어줌.
            if (!parseInt(errMessage)) return;

            let idx = errMessage;
            let col = this[OPTS].colModel[idx];

            this.editFail = true;

            if (_.isUndefined(col.message)) return; // message가 지정 되지 않을 경우 사용자가 팝업을 직접 생성

            let message = col.message;

            $().bxDialog({
                style: 'stick',
                width: 400,
                confirmText: '확인',
                closeText: '취소',
                dimStyle: {
                    color: 'black',
                    enableBlur: false
                }
            }).alert({
                title: '유효성 검사 실패',
                content: message,
                onConfirm: function (e){
                    this.hide();
                }
            });
        },
        fixScrollOffsetAndhBoxPadding: () => {
            let self = this, grid = self.grid;

            if (!grid){
                return;
            }

            let p = self.p, bDiv = grid.bDiv,
                fixhBox = function (hDiv){
                    let $hDivhBox = $(hDiv).children("div").first();
                    $hDivhBox.css($hDivhBox.hasClass("ui-jqgrid-hbox-rtl") ? "padding-left" : "padding-right", p.scrollOffset);
                    hDiv.scrollLeft = bDiv.scrollLeft;
                };

            if ($(bDiv).width() > 0){
                p.scrollOffset = (bDiv.offsetWidth - bDiv.clientWidth) - 3; // can be 0 if no scrollbar exist
                // TODO: add detection of the width of vertical scroll bar if the grid is hidden
                // at the moment of executing fixScrollOffsetAndhBoxPadding (for example inside of inactive jQuery UI Tab)
                // one need just create close construction with visible:hidden style, add to body and get its width
                fixhBox(grid.hDiv);
                if (grid.sDiv){
                    fixhBox(grid.sDiv);
                }
            }
        }
    });

    _setCellAttr(this[OPTS].colModel);

    _renderJqGrid.call(this, () => {

        let {onRender, $pageContainer, groupHeaders, width, resizable} = this;

        $container.jqGrid(this[OPTS]);

        onRender.call(this);

        _addIconWrapper.call(this);

        _addImportExcel.call(this);

        _addExportButton.call(this);

        _setGroupHeaders.call(this, groupHeaders);

        _removeScopeAttr.call(this, this.$root);

        _setAccessibility.call(this);

        // set resize about grid table.
        if (resizable)  _addResizeListenerAtWindow.call(this);

        if (!options.shrinkToFit) _setHorizonScroll.call(this, width);

        if (_.isNull(width)) width = this.$container.width();

        this.setGridWidth(width);

    });

    return this;
}
// TODO - private Resize function
/**
 * window resize 시, grid 넓이를 조정하는 eventListner 를 추가하는 함수.
 * @private
 * @memberOf BxGrid
 */
function _addResizeListenerAtWindow(){

    window.addEventListener('resize', () =>{

        let width = this.$root.width();

        this.setGridWidth(width);
        this.$pageContainer && this.$pageContainer.css('width', width);

    });
}

/**
 * 가로 스크롤을 추가 하는 함수
 * @param {object} [width=null] 가로 크기 값
 * @private
 * @memberOf BxGrid
 */
function _setHorizonScroll(width = null){

    let { $root, $container, $pageContainer } = this;
    let tblwidth = $container.get(0).p.tblwidth;

    const BODY_WRAPPER_CLASS = '.ui-jqgrid-bdiv';

    // if (UTIL.isIE()) {
    //
    //     // Jqgrid는 false일 경우, 컬럼 사이즈를 브라우저에게 맡기고 있다. 모던브라우저는 원래 사이즈보다 테이블이 커지면 auto size가 적용 되지만
    //     // IE에서는 auto가 적용 되지 않는 문제가 있다. 반면에 true일 경우 jqgrid에서 보이지 않는 <td>를 만들어 사이즈를 조절 하고 있다.
    //
    //     if ( this.originTblWidth === null ) this.originTblWidth = tblwidth;
    //
    //     $container.setGridWidth(this.originTblWidth, true);
    // }

    $container.setGridWidth(width, width > tblwidth); //true는 autosize

    if (width < tblwidth){
        $root.find(BODY_WRAPPER_CLASS).css('overflow-x', 'auto');
    }
    else{
        $root.find(BODY_WRAPPER_CLASS).css('overflow-x', 'hidden');
    }

    // 브라우저 싱크에 따라 스크롤 사이즈가 잘못 계산 될 수 있어 다시 계산해준다
    $container.get(0).fixScrollOffsetAndhBoxPadding();

    $pageContainer && $pageContainer.css('width', width);

}
/**
 * inline Edit Mode 인지 여부 반환 함수
 * @param {string} [id=null] row id
 * @returns {boolean} editMode 여부
 * @private
 * @memberOf BxGrid
 */
function _isEditable(id = null){

    let isEditable = false;

    isEditable = this.$container.find(`tr${!id ? '' : '#' + id} .editable`).length !== 0;

    return isEditable;
}


/**
 * 모든 행이 선택되었는지에 대한 여부 반환 함수
 * @returns {boolean}
 * @private
 * @memberOf BxGrid
 */
function _isSelectedAllRow(){

    let rowNum = this[OPTS].rowNum,
        isSelectedAllRow = rowNum === this.getSelectedId().length;

    return isSelectedAllRow;
}

/**
 * jqGrid Render 함수.
 * @private
 * @memberOf BxGrid
 */
function _renderJqGrid(callback){

    let optionName = AJAX_OPTION.READ_OPTION;

    if (_isAjaxTransaction.call(this)){

        let params = _getRequestParams.call(this, optionName);

        _transactionCRUD.call(this, optionName, params).then(data =>{

            _setResponseDataToGridParam.call(this, data, true);

            callback();
        });
    }
    else{
        callback();
    }
}


/**
 * BxGrid Ajax 통신을 위한 내부 함수
 *
 * @param {string} optionName - Ajax 통신때 사용할 옵션명
 * @param {object} params - Ajax 통신 request 파라미터
 *
 * @private
 * @memberOf BxGrid
 */
function _transactionCRUD(optionName = '', params = {}){

    let ajaxOption = _.cloneDeep(_getAjaxOption.call(this, optionName));

    _validateCheckForAjax.call(this, optionName, ajaxOption);

    let defer = $.Deferred();
    let url = ajaxOption.url;
    let method = ajaxOption.method.toLowerCase();

    // ajax callback method
    let before = ajaxOption.before;
    let success = ajaxOption.success;
    let fail = ajaxOption.fail;

    // ajax headers
    let headers = ajaxOption.headers;

    // request header 의 `Content-Type`에 `application/json` 이 포함되었는지 여부
    let isJSONContentType = (() => {

        let ret = false;

        _.map(headers, v => {

            if (v.indexOf('application/json') > -1){
                ret = true;
            }
        });

        return ret;

    })();

    $.ajaxSetup({cache: false});

    let xhrOption = {
        contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
        dataType: 'json',
        type: method,
        beforeSend: (xhr, settings) =>{

            // `IE9 브라우저`에서, 캐시 사용(json 등)을 무력화 시키기위한 header 설정
            xhr.setRequestHeader('If-Modified-Since', 'Mon, 26 Jul 1997 05:00:00 GMT');
            xhr.setRequestHeader('Cache-Control', 'no-cache');
            xhr.setRequestHeader('Pragma', 'no-cache');

            // 사용자가 입력한 Header Setting
            _.forEach(headers, (v, k) => {
                xhr.setRequestHeader(k, v);
            });

            // 사용자가 최종 수정한, params 객체를 서버로 전달하기위해, 내부적으로 생성한 `params` 객체와, ajaxModel 을 통해, 생성된 `default params`
            // 객체를 병합한 후, `before callback` 을 통해, 최종 전달하는 방식을 사용한다.
            params = $.extend(true, ajaxOption.params, params);

            before.call(ajaxOption, xhr);

            if (method === 'get'){

                // 전달받은 url 을 파라메터를 통해, 새롭게 가공한다.
                let urlObj = _getUrlObjectToMethod(method, url, params);

                settings.url = urlObj.url;
            }
            else {

                settings.url = url;
                settings.data = isJSONContentType ? JSON.stringify(params) : $.param(params);
            }
        },
        loadStart: (xhr) => {
        },
        progress: (xhr, loaded, total, percent) => {
        },
        error: xhr => {
        },
        success: data => {
        }
    };

    let ajaxPromise = XHR.ajax(xhrOption);

    ajaxPromise.done(res => {

        success.call(ajaxOption, res);

        defer.resolve(res);
    });

    ajaxPromise.fail(res => {

        fail.call(ajaxOption, res);

        defer.reject(res);
    });

    return defer.promise();
}

/**
 * 옵션 이름에 따라 사용자가 설정한 ajaxOption 을 반환
 *
 * @param {string} optionName - Ajax 통신때 사용할 옵션명
 * @returns {object} ajaxOption 반환
 *
 * @private
 * @memberOf BxGrid
 */
function _getAjaxOption(optionName = '') {

    let ret = null;

    switch (optionName) {

        case AJAX_OPTION.IMPORT_OPTION :

            ret = this.importModel.ajaxOption;

            break;

        case AJAX_OPTION.EXPORT_OPTION :

            ret = this.exportModel.ajaxOption;

            break;

        default :

            ret = this.ajaxModel[optionName] || {};
    }

    return ret;
}

/**
 * Ajax를 사용하는 모델의 ReqModel, ResModel을 가져오는 함수
 *
 * @param {string} optionName - Ajax 통신때 사용할 옵션명
 *
 * @returns {object} ret
 *
 * @private
 * @memberOf BxGrid
 */
function _getTransModel(optionName = ''){

	let ret = {};

	let reqModel = '';
	let resModel = '';
	
	if (optionName === AJAX_OPTION.IMPORT_OPTION){

	    resModel = this.importModel.ajaxOption.resModel;
	}
	else if (optionName === AJAX_OPTION.EXPORT_OPTION){

	    resModel = this.exportModel.ajaxOption.resModel;
	}
	else{
        reqModel = this.ajaxModel.reqModel;
	    resModel = this.ajaxModel.resModel;
	}

    if (!_.isEmpty(reqModel)){
        _.set(ret, 'reqModel', reqModel);
    }

    if (!_.isEmpty(resModel)){
        _.set(ret, 'resModel', resModel);
    }

	return ret;
}


/**
 * ajaxModel 을 반환한다.
 *
 * @param {string} optionName - Ajax 통신때 사용할 옵션명
 *
 *
 * @private
 *
 * @memberOf BxGrid
 */
function _getAjaxModel(optionName = ''){

	let ret = null;

	if (optionName === AJAX_OPTION.IMPORT_OPTION){

		ret = this.importModel.ajaxOption;
	}
	else if (optionName === AJAX_OPTION.EXPORT_OPTION){

		ret = this.exportModel.ajaxOption;
	}
	else{

		ret = this.ajaxModel;
	}

	return ret;
}

/**
 * Ajax Response data 초기 설정 함수
 *
 * @private
 * @memberOf BxGrid
 */
function _initAjaxResModel(){

    let ajaxModel = this.ajaxModel;
    let resModel = ajaxModel.resModel;

    if (!_.isPlainObject(resModel)){
        ajaxModel.resModel = DEFAULT_AJAX_MODEL_RES_MODEL;
    }
    else {

        const items = resModel.items;
        const totalCount = resModel.totalCount;

        if (_.isUndefined(items)){
            resModel.items = DEFAULT_AJAX_MODEL_RES_MODEL.items;
        }

        if (_.isUndefined(totalCount)){
            resModel.totalCount = DEFAULT_AJAX_MODEL_RES_MODEL.totalCount;
        }
    }

    return this;
}

/**
 * AjaxModel의 validation 체크 함수
 *
 * @params {string} modelName - ajaxModel 내부 key 값
 * @params {object} ajaxOption - default 속성
 *
 * @private
 * @memberOf BxGrid
 */
function _validateCheckForAjax(modelName = '', ajaxOption = {}){

    let url = ajaxOption.url;
    let method = ajaxOption.method;

    // ajax 를 위한 최소한의 유효성 검사
    if (_.isUndefined(ajaxOption) || _.isNull(ajaxOption)){
        CONSOLE.prop.throw(BxGrid, modelName, 'all');
    }

    if (_.isUndefined(url) || !_.isString(url) || _.isEmpty(url)){
        CONSOLE.prop.throw(BxGrid, 'url', 'all');
    }

    if (_.isUndefined(method) || !_.isString(method) || _.isUndefined(RESTFUL_METHOD[method.toUpperCase()])){
        CONSOLE.prop.throw(BxGrid, 'method', 'all');
    }
}

/**
 * BxGrid 마지막 아이템의 ID값을 가져오는 함수
 *
 * @returns {string} id - 마지막 아이템의 ID
 *
 * @private
 * @memberOf BxGrid
 */
function _getLastId(){

    let data = this[OPTS].data;
    let lastIdx = data.length - 1;

    let id = lastIdx < 0 ? 'jqg1' : `jqg${parseInt(data[lastIdx].id.replace(/\D+/g, '')) + 1}`;

    return id;
}

/**
 * 전체 선택 체크박스에 대한 diabled 속성을 결정하는 함수
 *
 * @param {boolean} isDisabled - 전체 선택 체크박스 사용 유무
 *
 * @private
 * @memberOf BxGrid
 */
function _disableAllSelected(){

    let allSelectCheckboxId = '[id^=cb_jqg]',
        $allCheck = $(this.$root.find(allSelectCheckboxId));


    if (_isEditable.call(this)){
        $allCheck.attr('disabled', true);
    }
    else{
        $allCheck.attr('disabled', false);
    }
}

/**
 * 해당 아이디를 갖는 데이터를 반환한다.
 *
 * @param {string} id - 검색할 id
 *
 * @returns {object} id로 검색된 데이터
 *
 * @private
 * @memberOf BxGrid
 */
function _getDataById(id = ''){

    let o = {},
        items = this[OPTS].data,
        rowNum = items.length,
        prefix = id.replace(/\d+$/, ''),
        num = parseInt(id.replace(/\D+/, ''));

    for (let i in items){
        // ajax 시, reload 를 통해 rowId 달라지는 문제로 인해서 아래 prefix 및 rowNum 다시 구하는 부분을 추가함.
        // 추후 진행시, 다시 작업되야 합니다. [ key 값 달라지는 경우, 파생되는 문제들 전부 해결 ]
        if (_.isEqual(String(items[i].id), id) || _.isEqual(String(items[i].id), `${prefix}${num + rowNum}`)){
            // console.log(items[i]);
            o = items[i];
            delete o.format;
            break;
        }
    }

    return o;
}

/**
 * changedData 동기화를 위한 내부 함수
 *
 * @param {string} id - id 값
 * @param {string} condition - 변경될 상태 값
 * @param {object} preData - 이전 데이터
 * @private
 * @memberOf BxGrid
 */
function _changedDataSync(id = '', condition = '', preData = {}){

    let changedObj = this.changedObj;
    let insertIdx = changedObj.id.insert.indexOf(id);
    let updateIdx = changedObj.id.update.indexOf(id);
    let deleteIdx = changedObj.id.delete.indexOf(id);

    let data = _getDataById.call(this, id);

    switch (condition){
        case 'insert':
            if (deleteIdx !== -1 && _.isEqual(data, changedObj.data.delete[deleteIdx])){
                changedObj.id.delete.splice(deleteIdx, 1);
                changedObj.data.delete.splice(deleteIdx, 1);
            }

            changedObj.id.insert.push(id);
            changedObj.data.insert.push(data);

            break;

        case 'update':

            if (_.isEqual(data, preData)){
                return;
            }

            if (insertIdx === -1){

                if (updateIdx === -1){
                    changedObj.id.update.push(id);
                    changedObj.data.update.push(data);
                }
                else{
                    changedObj.data.update[updateIdx] = data;
                }
            }
            else{
                changedObj.data.insert[insertIdx] = data;
            }

            break;

        case 'delete':

            if (insertIdx !== -1){
                changedObj.id.insert.splice(insertIdx, 1);
                changedObj.data.insert.splice(insertIdx, 1);
            }
            else {

                if (updateIdx !== -1){
                    changedObj.id.update.splice(updateIdx, 1);
                    changedObj.data.update.splice(updateIdx, 1);
                }

                changedObj.id.delete.push(id);
                changedObj.data.delete.push(data);
            }

            break;

        default:
            break;
    }
}

/**
 * colModel 확장에 대한 랩핑 함수
 *
 * @param {object} inlineEditColOption - 인라인 edit컬럼 설정 옵션
 *
 * @private
 * @memberOf BxGrid
 */
function _expandColModel(inlineEditColOption = {}){

    // colModel 검사 및 key 값의 유무를 확인 및 keyColumn 셋팅.
    _validateColModel.call(this);
    // colModel editable 추가 및 확장
    _rowEditCustomizing.call(this);
    // editColModel 있는 경우 추가한다
    _addInlineBtnColumn.call(this, inlineEditColOption);
    // formatOption 추가 및 확장
    _expandFormatter.call(this);

}

/**
 * colModel 유효성 검사 및 key 값 설정 (id field 는 jqGrid 내부 id 를 사용을 원칙으로 한다.)
 *
 * @returns {string} key 로 설정된 컬럼의 name 속성
 *
 * @private
 * @memberOf BxGrid
 */
function _validateColModel(){

    let colModel = this[OPTS].colModel;
    let existKey = false;

    colModel.map((data) =>{
        if (!data.name){
            CONSOLE.prop.throw(BxGrid, 'colModel 내부 name', 'value');
        }

        if (data.name === 'id'){
            data.key = true;
            existKey = true;

            return;
        }
    });


    if (!existKey){
        // colModel 추가.
        let idCol = {
            name: 'id',
            hidden: true,
            key: true
        };
        colModel.splice(0, 0, idCol);
        existKey = true;
    }
}

/**
 * inline Eidt Mode 시, 엘리먼트와 유효성검사 커스터마이징 함수
 *
 * @private
 * @memberOf BxGrid
 */
function _rowEditCustomizing(){

    let colModel = this[OPTS].colModel;

    colModel.map((data, i) =>{

        let { editOption, name: colName, label: colLabel } = data;

        if (editOption){

            if (!_.isObject(editOption)){
                CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption', 'type');
            }

            let editType = editOption.type;
            let offerEditTypeArr = ['select', 'checkbox', 'radio', 'text', 'textarea', 'custom'];

            if (!_.isString(editType)){
                CONSOLE.prop.throw(BxGrid, colName + ' 의 editType', 'type');
            }

            if (_.isEqual(editType.trim(), '')){
                CONSOLE.prop.throw(BxGrid, colName + ' 의 editType', 'value');
            }

            if (offerEditTypeArr.indexOf(editType) === -1){
                CONSOLE.prop.throw(BxGrid, colName + ' 의 editType', 'value');
            }

            let { attr: editAttr, items: editItems, separator: editSeparator }= editOption;

            data.editable = true;
            data.edittype = 'custom';

            if (_.isEqual(editType, 'custom')){

                if (!_.isFunction(editOption.customElement)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 customElement', 'type');
                }

                if (!_.isFunction(editOption.customElementValue)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 customElementValue', 'type');
                }

                if (!_.isUndefined(editAttr)){
                    CONSOLE.prop.warn(BxGrid, colName + ' 의 editAttr 는 사용할 수 없습니다.', 'empty');
                }

                if (!_.isUndefined(editItems)){
                    CONSOLE.prop.warn(BxGrid, colName + ' 의 editItems 는 사용할 수 없습니다.', 'empty');
                }

                if (!_.isUndefined(editSeparator)){
                    CONSOLE.prop.warn(BxGrid, colName + ' 의 editOption.separator 는 사용할 수 없습니다.', 'empty');
                }

                data.editoptions = {
                    custom_element: editOption.customElement,
                    custom_value: editOption.customElementValue
                };
            }
            else{

                let havingItem = ['select', 'checkbox', 'radio'].indexOf(editType) !== -1;
                let editTargetCol = editOption.targetCol;

                // select, checkbox, radio 버튼 일 경우 items에 대한 type, value check..
                if (havingItem && !_.isArray(editItems)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editoption.items', 'type');
                }
                if (havingItem && (!_.isPlainObject(editItems[0]) || _.isEmpty(editItems[0]))){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editoption.items', 'value');
                }

                if (!havingItem && !_.isUndefined(editItems)){
                    CONSOLE.prop.warn(BxGrid, colName + ' 의 editItems 는 사용할 수 없습니다.', 'empty');
                }
                if (!_.isEqual(editType, 'checkbox') && !_.isUndefined(editSeparator)){
                    CONSOLE.prop.warn(BxGrid, colName + ' 의 ' + editType + ' 타입은 editOption.separator 는 사용할 수 없습니다.', 'empty');
                }

                // editAttr 값이 들어온 경우 attr에 대한 type, value check
                if (!_.isUndefined(editAttr) && !_.isPlainObject(editAttr)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption.attr', 'type');
                }
                if (!_.isUndefined(editAttr) && _.isEmpty(editAttr)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption.attr', 'value');
                }

                if (!_.isUndefined(editSeparator) && !_.isString(editSeparator)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption.separator', 'type');
                }

                data.editoptions = _createEditOptions.call(this, colName, colLabel, editType, editAttr, editItems, editSeparator, editTargetCol);

            }

            let validateOption = editOption.validateOption;
            // validation이 있는 경우만 추가 한다.
            if (validateOption){

                let validator = validateOption.validator;
                let message = validateOption.message;

                if (!_.isFunction(validator)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption.validateOption.validator', 'type');
                }

                // @deprecated
                if (!_.isUndefined(message) && !_.isString(message)){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption.validateOption.message', 'type');
                }

                // @deprecated
                if (!_.isUndefined(message) && _.isEqual(message.trim(), '')){
                    CONSOLE.prop.throw(BxGrid, colName + ' 의 editOption.validateOption.message', 'value');
                }

                if (!_.isUndefined(message)){
                    data.message = message;
                }

                data.editrules = {
                    custom: true,
                    custom_func: (data, colName, i) =>{

                        // i 는 colModel.map 의 인자값.
                        const $container = this.$container;
                        const id = $container.getGridParam('selrow');
                        const rowData = $container.getRowData(id);

                        const editOptions = $container.get(0).p.colModel[i].editOption;
                        const type = editOptions.type;
                        const items = editOptions.items;

                        const getFindValue = value => {

                            let findValue = _.find(items, {label: value});

                            return _.isUndefined(findValue) ? '' : findValue.value;
                        };

                        if (type === 'checkbox'){

                            const separator = editOptions.separator;

                            let divValue = data.split(separator);
                            let result = [];

                            _.forEach(divValue, v => result.push(getFindValue(v)));

                            data = result.join(separator);
                        }
                        else if (type === 'select' || type === 'radio'){
                            data = getFindValue(data);
                        }

                        let validateFail = validator.call(this, data, colName, rowData);

                        return [validateFail, i];
                    }
                };

            }

            //delete data.editOption;
        }
    });
}
/**
 * edit mode 변경시, 사용할 Element 지정 함수
 *
 * @param {string} colName - 컬럼 명
 * @param {string} colLabel - 컬럼 라벨
 * @param {string} editType - edit 타입
 * @param {string} editAttr - edit 속성
 * @param {array} editItems - edit 아이템
 * @param {string} editSeparator - edit 분리 기호
 * @param {string} editTargetCol - edit 할 타겟 컬럼 명
 *
 * @private
 * @memberOf BxGrid
 */
function _createEditOptions(colName, colLabel, editType, editAttr, editItems, editSeparator, editTargetCol){

    let self = this;

    let o = {
        //TODO 접근성 custom_element
        custom_element: function (cellValue, colData){

            let elem = [],
                rowId = colData.rowId,
                editAttrStr = editAttr ? _addEditElementAttr(colName, editAttr) : '';

            switch (editType){

                case 'text':

                    elem.push(`<input type="text" data-id="${rowId}" title="${colLabel}" value="${cellValue}" `);
                    elem.push(editAttrStr);
                    elem.push('/>');

                    break;

                case 'textarea':

                    elem.push(`<textarea title="${colLabel}" data-id=${rowId}`);
                    elem.push(editAttrStr);
                    elem.push('>' + cellValue + '</textarea>');

                    break;

                case 'checkbox':

                    editSeparator = editSeparator ? editSeparator : ', ';

                    let valueArr = cellValue.split(editSeparator);
                    elem.push('<div ');
                    elem.push(editAttrStr);
                    elem.push('>');

                    for (let i in editItems){

                        let editItem = editItems[i],
                            label = editItem.label,
                            value = editItem.value;

                        elem.push('<label class="edit_label">');

                        if (_.isUndefined(label)){
                            CONSOLE.prop.warn(BxGrid, colName + ' 의 editItem.label', 'value');
                        }

                        if (editTargetCol){

                            if (!_.isString(editTargetCol)){
                                CONSOLE.prop.warn(BxGrid, colName + ' 의 editOption.targetCol', 'type');
                            }

                            if (_.isUndefined(value)){
                                CONSOLE.prop.warn(BxGrid, colName + ' 의 editItem.value', 'value');
                            }

                            elem.push(`<input title="${label}" name="check_${rowId}" type="checkbox" value="${value}" data-id="${rowId}_${i}" `);
                        }
                        else{
                            elem.push(`<input title="${label}" name="check_${rowId}" type="checkbox" value="${label}" data-id="${rowId}_${i}" `);
                        }

                        elem.push(valueArr.indexOf(label) === -1 ? '' : 'checked');

                        elem.push(' />');

                        elem.push(label + '</label>');
                    }

                    elem.push('</div>');

                    break;

                case 'radio':
                    elem.push('<div ');
                    elem.push(editAttrStr);
                    elem.push('>');

                    for (let i  in editItems){

                        let editItem = editItems[i],
                            label = editItem.label,
                            value = editItem.value;

                        if (_.isUndefined(label)){
                            CONSOLE.prop.warn(BxGrid, colName + ' 의 editItem.label', 'value');
                        }

                        elem.push('<label class="edit_label">');

                        if (editTargetCol){

                            if (!_.isString(editTargetCol)){
                                CONSOLE.prop.warn(BxGrid, colName + ' 의 editOption.targetCol', 'type');
                            }

                            if (_.isUndefined(value)){
                                CONSOLE.prop.warn(BxGrid, colName + ' 의 editItem.value', 'value');
                            }

                            elem.push(`<input title="${label}" name="radio_${rowId}" type="radio" value="${value}" data-id="${rowId}_${i}"`);
                        }
                        else{
                            elem.push(`<input title="${label}" name="radio_${rowId}" type="radio" value="${value}" data-id="${rowId}_${i}"`);
                        }

                        elem.push(_.isEqual(label, cellValue) ? 'checked' : '');

                        elem.push(' />');
                        elem.push(label + '</label>');
                    }

                    elem.push('</div>');
                    break;

                case 'select':
                    elem.push('<div ');
                    elem.push(editAttrStr);
                    elem.push('>');

                    elem.push(`<select title="${colLabel}" data-id="${rowId}">`);
                    for (let i in editItems){

                        let editItem = editItems[i],
                            label = editItem.label,
                            value = editItem.value;

                        if (_.isUndefined(label)){
                            CONSOLE.prop.warn(BxGrid, colName + ' 의 editItem.label', 'value');
                        }

                        // targetCol 프로퍼티가 존재할경우
                        if (editTargetCol){

                            if (!_.isString(editTargetCol)){
                                CONSOLE.prop.warn(BxGrid, colName + ' 의 editOption.targetCol', 'type');
                            }

                            if (_.isUndefined(value)){
                                CONSOLE.prop.warn(BxGrid, colName + ' 의 editItem.value', 'value');
                            }

                            elem.push(`<option value="${value}" ${label === cellValue ? 'selected' : ''}>`);
                        }
                        else{
                            elem.push(`<option value="${value}" ${label === cellValue ? 'selected' : ''}>`);
                        }

                        elem.push(label + '</option>');
                    }

                    elem.push('</select>');
                    elem.push('</div>');

                    break;

                default :
                    break;
            }
            return elem.join('');
        },
        custom_value: function ($elem){

            let text = '';
            let value = '';

            switch (editType){

                case 'text':

                    text = $elem.val();

                    break;

                case 'textarea':

                    text = $elem.val();

                    break;

                case 'checkbox':

                    editSeparator = editSeparator ? editSeparator : ',';

                    let length = $elem.parent().find('input:checked').length;

                    if (length){

                        let i = 0;
                        let $checkbox = null;
                        let $label = null;

                        for (i = 0; i < length - 1; i++){

                            $checkbox = $($elem.parent().find('input:checked').get(i));
                            $label = $checkbox.parent();

                            text += $label.text() + editSeparator;

                            if (editTargetCol) value += $checkbox.val() + editSeparator;
                        }


                        $checkbox = $($elem.parent().find('input:checked').get(i));
                        $label = $checkbox.parent();

                        text += $label.text();

                        if (editTargetCol){
                            value += $checkbox.val();
                            self.targetColModelValues[editTargetCol] = value;
                        }
                    }
                    else {
                        self.targetColModelValues[editTargetCol] = '';
                    }

                    break;

                case 'radio':

                    let $radio = $elem.parent().find('input:checked');
                    let $label = $radio.parent();

                    text = $label.text();

                    if (editTargetCol){
                        value = $radio.val();
                        self.targetColModelValues[editTargetCol] = value;
                    }

                    break;

                case 'select':

                    $elem = $elem.find('select option:selected');

                    text = $elem.text();

                    if (editTargetCol){
                        value = $elem.val();
                        self.targetColModelValues[editTargetCol] = value;
                    }

                    break;
            }

            return text;
        }
    };

    return o;
}

/**
 * inline Edit Mode 의 Element 속성 부여 함수
 * @param {string} colName - 컬럼 명
 * @param {string} editAttr - edit 속성
 *
 * @returns {string} 추가된 element
 *
 * @private
 * @memberOf BxGrid
 */
function _addEditElementAttr(colName = '', editAttr = ''){

    let elem = [];
    let excludeAttr = ['data-id', 'id'];

    for (let name in editAttr){

        if (excludeAttr.indexOf(name) !== -1){
            CONSOLE.prop.warn(BxGrid, colName + ' 의 editoption.attr 의 ' + name + '속성은 사용할 수 없습니다.', 'empty');
            continue;
        }

        elem.push(name + '="' + editAttr[name] + '" ');
    }
    return elem.join('');
}

// /**
//  * 유효성 검사 추가 함수.
//  * @private
//  * @memberOf BxGrid
//  */
// function _createEditPattern(type){
//     let pattern = null;
//
//     switch (type) {
//         case 'number':
//             pattern = /^\d+.?\d+$/;
//             break;
//         case 'email':
//             pattern = /^[^@]+@{1}[^@]+$/;
//             break;
//         // case 'phone':
//         //     pattern = /^\+?\d+-?\d+-?\d+$/;
//         //     break;
//         // case 'date':
//         //     pattern = /^\d+[.-]{1}\d+[.-]{1}\d+$/;
//         //     break;
//         // case 'datetime':
//         //     pattern = /^\d+[.-]{1}\d+[.-]{1}\d+\s*\d+:\d+:\d+$/;
//         //     break;
//         // case 'time':
//         //     pattern = /^\d+:\d+:\d+$/;
//         default:
//             throw new throw(type + '유형은 현재 지원하지 않습니다.');
//             break;
//     }
//
//     return pattern;
// }

/**
 * colModel 에 inlineBtn 컬럼 추가 함수
 * @private
 * @memberOf BxGrid
 */
function _addInlineBtnColumn(inlineEditColOption){

    if (inlineEditColOption.inlineEditCol){

        let inlineEditColModel = inlineEditColOption.inlineEditColModel;
        let editColWidth = Object.keys(inlineEditColModel).length * 25;
        let editable = inlineEditColModel['edit'] === undefined ? false : true;

        const iconSet = {
            edit: 'fa-pencil',
            delete: 'fa-trash-o',
            save: 'i-save',
            cancel: 'i-undo'
        };

        let editBtnCol = {
            // label: inLineEditColLable,
            // name: 'format',
            // width: 100,
            // sortable: false,
            // editable: true,
            // formatter: 'actions',
            // edittype: 'custom',
            // editoptions: {
            //     custom_element: function(dummy, obj){
            //         let test = [];
            //         test.push('<div class="fommater ui-jqgrid-actions">')
            //         test.push('<div class="ui-corner-all ui-pg-div ui-inline-save" id="jEditButton_' + obj.rowId + '" onclick="return $.fn.fmatter.rowactions.call(this,event,\'save\');" onmouseover="$(this).addClass(\'ui-state-hover\');" onmouseout="$(this).removeClass(\'ui-state-hover\');">');
            //         test.push('<i class="bw-icon i-25 i-save"></i>');
            //         test.push('</div>');
            //         test.push('<div class="ui-corner-all ui-pg-div ui-inline-cancel" id="jDeleteButton_' + obj.rowId + '"onclick="return $.fn.fmatter.rowactions.call(this, event, \'cancel\')" onmouseover="$(this).addClass(\'ui-state-hover\');" onmouseout="$(this).removeClass(\'ui-state-hover\');">');
            //         test.push('<i class="bw-icon i-25 i-undo"></i>');
            //         test.push('</div>');
            //         test.push('</div>')
            //         return test.join('');
            //     },
            //     custom_value: function(elem){
            //         return '';
            //     }
            // },
            // align: 'center'
            label: inlineEditColOption.inlineEditColLabel,
            name: 'format',
            width: editColWidth,
            align: "center",
            editable: editable,
            sortable: false,
            formatter: function (cellvalue, options, rowObject){
                let rowId = options.rowId;
                let html = createBtnElement(rowId, 'show');
                return html;
            }
        };

        let createBtnElement = function (rowId, prop){

            let btnTemplate = [];

            btnTemplate.push('<div class="ui-corner-all ui-pg-div ui-inline-{state}"');
            btnTemplate.push(' id="j{state}Button_{rowId}" onclick="return $.fn.fmatter.rowactions.call(this,event, \'{state}\');"');
            btnTemplate.push(' onmouseover="$(this).addClass(\'ui-state-hover\');" onmouseout="$(this).removeClass(\'ui-state-hover\');">');
            btnTemplate.push('<button type="button" class="bw-icon i-25 {iconSet}" title="{iconStr}"><span class="blind">{iconStr}</span></button>');
            btnTemplate.push('</div>');

            btnTemplate = btnTemplate.join('');

            let html = [];

            html.push('<div class="fommater ui-jqgrid-actions">');

            for (let key in inlineEditColModel){

                if (!inlineEditColModel[key][prop]) continue;

                html.push(btnTemplate.replace(/\{state}/g, key)
                    .replace(/\{rowId}/, rowId)
                    .replace(/\{iconSet}/, iconSet[key])
                    .replace(/\{iconStr}/g, ALT_TEXT[iconSet[key]]));
            }

            html.push('</div>');

            return html.join('');
        };

        if (editable){

            editBtnCol.edittype = 'custom';

            editBtnCol.editoptions = {
                custom_element: function (dummy, obj){
                    let test = createBtnElement(obj.rowId, 'showEditMode');
                    return test;
                },
                custom_value: function (elem){
                    return '';
                }
            };
        }

        this[OPTS].colModel.push(editBtnCol);

    }
}

/**
 * formatter 기능 설정 및 커스터마이징 함수
 *
 * @private
 * @memberOf BxGrid
 */
function _expandFormatter(){

    let colModel = this[OPTS].colModel;

    colModel.forEach((data, i) =>{

        if (_.isPlainObject(data.formatOption)){

            let name = data.name;
            let item = data.formatOption;
            let options = item.options || {};
            let type = item.type;


            if (!_.isString(type) || _.isUndefined(FORMAT_TYPE[type.toUpperCase()])){
                CONSOLE.prop.throw(BxGrid, 'formatter type', 'type');
            }

            if (type === FORMAT_TYPE.CUSTOM && (!_.isFunction(item.customFormatter) || !_.isFunction(item.customUnFormatter))){
                CONSOLE.prop.throw(BxGrid, `colModel ${name}의 formatter type이 custom인 경우 customFormatter와 customUnformatter 함수가 반드시 필요합니다.`, 'empty');
            }

            if (type === FORMAT_TYPE.DATE && (
                    !_.isString(options.dataFormat) || !_.isString(options.viewFormat) ||
                    options.dataFormat.length === 0 ||
                    options.viewFormat.length === 0
                )
            ){
                CONSOLE.prop.throw(BxGrid, `colModel ${name}의 formatter type이 date인 경우 dataFormat 값과 viewFormat 값이 반드시 필요합니다.`, 'empty');
            }

            if (type === FORMAT_TYPE.CURRENCY && !_.isUndefined(options.decimalPlaces)){

                if (!_.isNumber(options.decimalPlaces)){
                    CONSOLE.prop.throw(BxGrid, 'decimalPlaces', 'type');
                }

                if (options.decimalPlaces < 0){
                    CONSOLE.prop.throw(BxGrid, `colModel ${name}의 formatOption.options.decimalPlaces 값은 양수만 입력 가능합니다`, 'empty');
                }

                if (_.toString(options.decimalPlaces).indexOf('.') > -1){
                    CONSOLE.prop.throw(BxGrid, `colModel ${name}의 formatOption.options.decimalPlaces는 정수만 입력 가능합니다`, 'empty');
                }
            }

            let maxDecimalPlaces = 30;

            switch (type){

                case FORMAT_TYPE.CURRENCY:
                    let tempOptions = {};

                    if (!_.isUndefined(options.prefix)) tempOptions.prefix = options.prefix;
                    if (!_.isUndefined(options.suffix)) tempOptions.suffix = options.suffix;

                    if (!_.isUndefined(options.decimalPlaces)){
                        tempOptions.decimalPlaces = options.decimalPlaces > maxDecimalPlaces ? maxDecimalPlaces : options.decimalPlaces;
                    }

                    tempOptions.thousandsSeparator = ',';
                    tempOptions.defaultValue = '';

                    data.formatter = type;
                    data.formatoptions = tempOptions;

                    break;
                case FORMAT_TYPE.DATE:
                case FORMAT_TYPE.TEL:
                case FORMAT_TYPE.PID:
                case FORMAT_TYPE.BZNO:
                case FORMAT_TYPE.CARD:
                case FORMAT_TYPE.CUSTOM:
                    data.formatter = _customFormatter.bind(this, item);
                    data.unformat = _customUnFormatter.bind(this, item);
                    break;
            }

        }
    });
}

/**
 * type별 formatter 기능 설정
 *
 * @private
 * @memberOf BxGrid
 */
function _customFormatter(formatOptions, cellvalue, options, rowObject){

    let result = '';
    let type = formatOptions.type;
    let option = formatOptions.options;
    let separator = option && option.separator ? option.separator : '-';
    let params = {separator: separator};

    // 커스텀 formatter는 사용자에게 입력받은 함수로 fomatting
    if (type === FORMAT_TYPE.CUSTOM){
        result = formatOptions.customFormatter(cellvalue, options, rowObject);
    }
    else {

        if (type === FORMAT_TYPE.DATE){
            params.dataFormat = option.dataFormat;
            params.viewFormat = option.viewFormat;
        }

        result = UTIL.getFormatString(type, cellvalue, params);

        if (type === FORMAT_TYPE.CARD){
            result = result.value;
        }
    }

    return result;
}

/**
 * type별 unformatter 기능 설정
 *
 * @private
 * @memberOf BxGrid
 */
function _customUnFormatter(formatOptions, cellvalue, options){

    let result = '',
        type = formatOptions.type,
        option = formatOptions.options;

    // 커스텀 unformatter는 사용자에게 입력받은 함수로 unfomatting
    if (type === FORMAT_TYPE.CUSTOM){
        result = formatOptions.customUnFormatter(cellvalue, options);
    }
    else{
        result = UTIL.getUnFormatString(type, cellvalue, option);
    }

    return result;
}

/**
 * BxGrid 페이지네이션 구현을 위한 내부 메서드
 *
 * @param {int} records - 총 데이터 수
 * @param {int} currentPage -
 * @param {int} rowNum - 사용자가 입력한 페이지 당 최대 행 수.
 *
 * @memberOf BxGrid
 * @private
 */
//그리드 페이징  
function _initPage(records = 0, currentPage = 0, rowNum = 0){

    let self = this;

    // 전체 페이지 사이즈
    let totalPageSize = Math.ceil(records / rowNum); // 총 페이지

    // 페이지 그룹 사이즈
    let pageGroupSize = this.pageGroupSize;

    // 전체 페이지 그룹
    let totalPageGroupSize = Math.ceil(totalPageSize / pageGroupSize);  // 총 페이지 / 페이지 목록 갯수 --> 뿌려줄 데이터 

    // 현재 페이지 그룹
    let currentPageGroup = Math.ceil(currentPage / pageGroupSize);    // 현재 페이지가 총 페이지 리스트 중 몇번째 인지

    // 페이지 리스트가 1보다 작으면 1로 초기화    
    // 페이지 리스트가 총 페이지 리스트보다 커지면 총 페이지 리스트로 설정    
    if (currentPage < 1) currentPage = 1;
    if (currentPageGroup > totalPageGroupSize) currentPageGroup = totalPageGroupSize;

    // 시작 페이지    
    // 끝 페이지    
    let startPageGroup = ((currentPageGroup - 1) * pageGroupSize) + 1;
    let endPageGroup = (startPageGroup + pageGroupSize) - 1;

    // 시작 페이지와 끝페이지가 1보다 작으면 1로 설정    
    // 끝 페이지가 마지막 페이지보다 클 경우 마지막 페이지값으로 설정    
    if (startPageGroup < 1) startPageGroup = 1;

    if (endPageGroup > totalPageSize) endPageGroup = totalPageSize;
    if (endPageGroup < 1) endPageGroup = 1;

    // 현재 페이지 컨테이너 초기화
    this.$pageContainer.html('');

    let className = BxGrid.className;

    let $pagingContainer = $(`<div class="bx-ui-${className} bw-paging add-mg-t"></div>`);

    if (currentPageGroup > 1){

        let $prevBtnGroup = $('<div class="ctr-btn"></div>');

        let $first = _createPagingNavElement('forward', e => this.goPage(1));
        let $prev = _createPagingNavElement('pre', e => this.goPage(startPageGroup - 1));

        $prevBtnGroup.append($first);
        $prevBtnGroup.append($prev);

        $pagingContainer.append($prevBtnGroup);
    }

    let $pagingGroup = $('<div class="paging-btn"></div>');

    for (let i = startPageGroup; i <= endPageGroup; i++){

        let $a = null;

        if (i === Number(currentPage)){
            $a = $('<button type="button" class="btn-paging on" title =' + i + '>' + i + '</button>');
        }
        else{
            $a = $('<button type="button" class="btn-paging" title =' + i + '>' + i + '</button>');
            $a.on('click', (e) => this.goPage(i));
        }

        $pagingGroup.append($a);
    }

    $pagingContainer.append($pagingGroup);

    if (currentPageGroup < totalPageGroupSize){

        let $nextCtrBtn = $('<div class="ctr-btn"></div>');

        let $next = _createPagingNavElement('next', e => this.goPage(endPageGroup + 1));
        let $last = _createPagingNavElement('backward', e => this.goPage(totalPageSize));

        $nextCtrBtn.append($next);
        $nextCtrBtn.append($last);

        $pagingContainer.append($nextCtrBtn);
    }

    this.$pageContainer.append($pagingContainer);
}

/**
 * BxGrid 페이지 네비게이션 엘리먼트 생성 함수
 *
 * @param {string} directionClass - 방향에 대한 클래스
 * @param {function} clickEvent - 버튼 클릭 이벤트
 *
 * @returns {string} element
 *
 * @memberOf BxGrid
 * @private
 */
function _createPagingNavElement(directionClass = '', clickEvent = function (){
}){

    let element = `<button type="button" class="btn-paging" title="{title}"><i class="bw-bx-icon i-15 i-ctrl-${directionClass}"><span class="blind">{altText}</span></i></button>`;

    let $innerHtml = $(element.replace(/\{title}/, ALT_TEXT[directionClass])
        .replace(/\{altText}/, ALT_TEXT[directionClass]));

    $innerHtml.on('click', clickEvent);

    return $innerHtml;
}

/**
 * Ajax Method별 URL & 파라미터 객체 생성 함수
 *
 * @param {string} mType - 총 데이터 수
 * @param {string} url - 요청 보낼 URL
 * @param {object} params - 요청 보낼 파라미터
 *
 * @memberOf BxGrid
 * @private
 */
function _getUrlObjectToMethod(mType = '', url = '', params = {}){

    let result = {};

    if (!RESTFUL_METHOD[mType.toUpperCase()]){
        CONSOLE.prop.throw(BxGrid, 'Transaction Method', 'value');
    }

    switch (mType.toLowerCase()){
        case 'get':

            let tempUrl = _.cloneDeep(url);
            let invoke = "";

            if (tempUrl.indexOf('?') > -1){
                params = _mergeURLToQueryString(tempUrl, params);
                tempUrl = tempUrl.slice(0, tempUrl.indexOf('?'));
            }

            invoke = UTIL.objectToQueryString(params);
            tempUrl = invoke.length === 0 ? tempUrl : tempUrl + '?' + invoke;

            result.url = tempUrl;

            break;

        case 'put':
        case 'delete':
        case 'post':

            result.url = url;
            result.params = params;

            break;
    }

    return result;
}

/**
 * URL의 querystring과 params를 합쳐주는 함수
 *
 * @param {string} url - URL 데이터
 * @param {object} params - 파라미터
 *
 * @returns {object} - 합쳐진 parameter object
 *
 * @memberOf BxGrid
 * @private
 */
function _mergeURLToQueryString(url = '', params = {}){
    let queryObject = UTIL.queryStringToObject(url);

    _.assign(queryObject, params);

    return queryObject;
}

/**
 * 파라메터 생성 함수.
 *
 * @param {string} [optionName = ''] - optionName
 * @param {object} [extendParams = {}] - extendParams
 *
 * @returns {object} params - parameter object
 *
 * @private
 *
 * @memberOf BxGrid
 */
function _getRequestParams(optionName = '', extendParams = {}){

    const ajaxModel = _getAjaxModel.call(this, optionName);
    const reqModel = ajaxModel.reqModel;

    const ajaxOption = _getAjaxOption.call(this, optionName);
    const method = ajaxOption.method.toLowerCase();

    let params = {};

    if (method !== 'get' && !_.isEmpty(reqModel)){
        _.set(params, reqModel, {});
    }

    if (optionName === AJAX_OPTION.READ_OPTION){

        const pagingModel = DEFAULT_REQ_PARAMS_MODEL.paging;

        if (this.enablePaging){

            _.assign(_.get(params, reqModel, params), {
                [pagingModel.currentPage]: this.movePage > 0 ? this.movePage : 1,
                [pagingModel.pageSize]: this[OPTS].rowNum
            });
        }

        // 정렬 관련 parameter 확장
        $.extend(true, params, _getRequestSortParams.call(this));

        extendParams = _.size(ajaxModel.extendParams) ? ajaxModel.extendParams : extendParams;

        $.extend(true, params, extendParams);
    }
    else{
        $.extend(true, _.get(params, reqModel, params), extendParams);
    }


    return params;
}

/**
 * sort 파라메터 생성 함수.
 *
 *
 * @returns {object} params - parameter object
 *
 * @private
 *
 * @memberOf BxGrid
 */
function _getRequestSortParams(){

	const ajaxModel = this.ajaxModel;
    const reqModel = _getTransModel.call(this).reqModel;
    const method = ajaxModel[AJAX_OPTION.READ_OPTION].method.toLowerCase();

    let params = {};

    if (_.isString(reqModel) && method !== 'get'){
        _.set(params, reqModel, {});
    }

    const sortList = ajaxModel.ajaxSortList;
    const sortModel = DEFAULT_REQ_PARAMS_MODEL.sort;

    if (!_.isEmpty(sortList.columns)){

        _.assign(_.get(params, reqModel, params), {
            [sortModel.columns]: sortList.columns.join(','),
            [sortModel.orderBy]: sortList.orderBy.join(',')
        });
    }

    return params;
}

/**
 * ajaxSortList 초기 설정 함수
 * @param sortName - BxGrid 초기 sortName
 * @param sortOrder - BxGrid 초기 sortOrder
 *
 * @private
 * @memberOf BxGrid
 */
function _setAjaxSortList(sortName = '', sortOrder = ''){

    let ajaxModel = this.ajaxModel;

    ajaxModel.ajaxSortList = {
        columns: [],
        orderBy: []
    };

    if (_.isEmpty(sortName)) return;

    sortName += ' ' + sortOrder;

    let sortColArr = sortName.split(', ');

    for (let idx in sortColArr){

        let sortColDetailInfo = sortColArr[idx].split(' ');

        if (sortColDetailInfo.length !== 2) continue;

        ajaxModel.ajaxSortList.columns.push(sortColDetailInfo[0]);
        ajaxModel.ajaxSortList.orderBy.push(sortColDetailInfo[1]);
    }
}



/**
 * Ajax 통신을 이용해 반환받은 response Data 를 Grid 속성으로 변경하는 함수
 * @param {object} [data={}] 서버 response data
 * @param {boolean} [isBeforeRender=false] 그리드가 렌더링 되기 전인지 여부
 * @private
 * @memberOf BxGrid
 */
function _setResponseDataToGridParam(data = {}, isBeforeRender = false){

    let opts = this[OPTS];
    let resModel = _getTransModel.call(this).resModel;

    // 전달받은 데이터의 items 가져온다.(resModel.items 기준)
	let items = _.get(data, resModel.items);

	// 전달받은 데이터의 totalCount 가져온다.(resModel.totalCount 기준)
	let totalCount = _.get(data, resModel.totalCount);

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxGrid, ERR_MSG.notFoundItemViaResModel);
    }

    if (!TYPE.isNumber(totalCount)){
        CONSOLE.prop.throw(BxGrid, ERR_MSG.notFoundTotalCountViaResModel);
    }

    _.assign(data, {
        page: this.movePage ? this.movePage : 1,
        rows: items,
        records: totalCount,
        total: Math.ceil(totalCount / Math.abs(opts.rowNum))
    });

    let params = {
        datastr: data,
        datatype: opts.datatype
    };

    delete data[resModel.items.split('.')[0]];
    delete data[resModel.totalCount.split('.')[0]];

    if (isBeforeRender){
        _.assign(opts, params);
    }
    else{
        this.setGridParam(params);
    }
}

/**
 * 웹 접근성 적용
 * @private
 * @memberOf BxGrid
 */
function _setAccessibility(){

    _initCaption.call(this);//헤더 영역에 캡션 추가

    _theadCustomizing.call(this);// Top, middle 테이블의 Thead 커스텀
}

/**
 * caption을 생성 한다.
 * @private
 * @memberOf BxGrid
 */
function _initCaption(){

    let $tableElem = this.$root.find('table');

    _.map($tableElem, elem =>{

        if ($(elem).hasClass('ui-jqgrid-htable') && $(elem).find('.ui-jqgrid-htable > caption').length === 0){

            let columns = '';
            let flag = false;

            this[OPTS].colModel.map((data, i) =>{

                if (data.label !== undefined){

                    if (flag){
                        columns += ', ';
                    }

                    flag = true;

                    this.groupHeaders.map((groupHeader) =>{

                        if (data.name === groupHeader.startColName){
                            columns += _getSubCaption.call(this, groupHeader);
                        }

                    });

                    columns += data.label;
                }
            });

            let caption = ALT_TEXT['caption'].replace(/\{columns}/g, columns);

            $(elem).prepend('<caption class="blind">' + caption + '</caption>');
        }
    });

}

/**
 * 헤더 테이블 Thead를 바디 테이블에 동일하게 적용
 * @private
 * @memberOf BxGrid
 */
function _theadCustomizing(){

    const headerTableClass = '.ui-jqgrid-htable';
    const bodyTableBody = '.ui-jqgrid-btable tbody';

    let $root = this.$root;

    let $htable = $root.find(headerTableClass);
    let $caption = $htable.find('caption');

    let _$htable = $htable.clone();

    let $thead = _$htable.find('thead');

    let _$caption = _$htable.find('caption');

    let $th = $htable.find('thead th');
    let $tbody = $root.find(bodyTableBody);


    //리더기가 읽는거 방지
    $th.attr('aria-hidden', true);
    $caption.attr('aria-hidden', true);

    _$htable.find('button').remove();
    _$htable.find('input').remove();

    if (!UTIL.isIE9()) {

        //복사된 헤더 숨김
        $thead.addClass('blind');
    }
    else {

        // ie9일때 blind로 처리하면 body가 사라진다.
        // header에서 적용된 blind css가 body 까지 영향을 미치는듯 하다.
        _$caption.removeClass('blind');
        _$caption.css('display', 'none');

        $thead.css('display', 'none');
    }

    $tbody.before(_$caption.get(0)).before($thead.get(0));
}

/**
 * sub caption을 반환한다.
 *
 * @param {object} groupHeader 객체
 * @returns {string} subColumns - groupHeader에서 caption으로 들어갈 내용
 * @private
 */
function _getSubCaption(groupHeader = {}){

    let subColumns = '';
    let trTags = groupHeader.content;

    $.parseHTML(trTags).map((trTag) => {

        let thTags = $(trTag).html();

        $.parseHTML(thTags).map((thTag) => {
            subColumns += $(thTag).text() + ', ';
        });

    });

    return subColumns;
}

/**
 * inline Edit mode 상태 데이터 반환 함수
 *
 * @param id - rowid
 * @returns {object} o - inline Edit Mode 데이터 객체
 * @private
 */
function _getEditData(id = ''){

    let $container = this.$container,
        colModel = this[OPTS].colModel,
        $tdArr = $container.find(`tr#${id} td:not(.td_cbox):not(.jqgrid-rownum)`),
        checkElement = ['checkbox', 'radio'],
        o = {};

    if (!_.isString(id)){
        CONSOLE.prop.throw(BxGrid, 'id', 'type');
    }

    if (_.isEmpty(id)){
        CONSOLE.prop.throw(BxGrid, 'id', 'value');
    }

    for (let i = 0; i < $tdArr.length; i++){

        let $td = $($tdArr[i]),
            $editableArr = $td.children().length === 0 ? [] : $($td.find('span.editable'));

        if ($editableArr.length === 0) continue;

        let name = colModel[i].name,
            $elemArr = $($editableArr).find(`input[type=text],input[type=radio], input[type=checkbox], select, textarea`),
            data = {};

        for (let j = 0; j < $elemArr.length; j++){

            let elem = $elemArr[j],
                $elem = $(elem),
                k = $elem.attr('data-id'),
                _o = {};

            if (checkElement.indexOf(elem.type) !== -1){
                let _k = $elem.val();
                _o[k] = {};
                _.assign(_o[k], {
                    [_k]: elem.checked
                });
            }
            else{
                _o[k] = $elem.val();
            }

            _.assign(data, _o);

        }

        o[name] = data;

    }
    return o;
}

/**
 * grid param 내부에 paging 및 sort 모델 있는지 유무 반환 함수
 * @param {string}[k ='']
 * @returns {boolean}
 * @private
 */
function _isContainBasicGridParam(k = ''){

    let pagingModel = DEFAULT_REQ_PARAMS_MODEL.paging,
        sortModel = DEFAULT_REQ_PARAMS_MODEL.sort,
        result = false;

    if (_.isEqual(pagingModel.currentPage, k) || _.isEqual(pagingModel.pageSize, k)
        || _.isEqual(sortModel.columns, k) || _.isEqual(sortModel.orderBy, k)){

        result = true;
    }

    return result;
}

/**
 * 그리드의 병합 함수
 *
 * @param {array} [groupHeader = [] - 병합을 할 헤더 리스트
 * @returns {object} 생성된 인스턴스
 * @private
 */
function _setGroupHeaders(groupHeader = []){

    if (_.isEmpty(groupHeader)) return;

    let $root = this.$root,
        o = [],
        tableClass = 'bx-grid-group-header';

    // remove col resizeable
    $root.find('span.ui-jqgrid-resize').remove();

    groupHeader.map((item, i) =>{

        let html = [],
            {startColName, numberOfCol, content} = item;

        if (!_.isPlainObject(item)){
            CONSOLE.prop.throw(BxGrid, `groupOption[${i}]`, 'type');
        }

        if (!_.isString(startColName)){
            CONSOLE.prop.throw(BxGrid, `groupOption[${i}].startColName`, 'type');
        }

        if (!_.isInteger(numberOfCol)){
            CONSOLE.prop.throw(BxGrid, `groupOption[${i}].numberOfCol`, 'type');
        }

        if (!_.isString(content)){
            CONSOLE.prop.throw(BxGrid, `groupOption[${i}].content`, 'type');
        }

        html.push(`<table class="${tableClass}">`);
        html.push(`<caption class="blind">그룹 헤더 목록</caption>`);
        html.push(`<thead>`);
        html.push(content);
        html.push(`</thead>`);
        html.push('</table>');

        o.push({
            startColumnName: startColName,
            numberOfColumns: numberOfCol,
            titleText: html.join('')
        });
    });

    if (!_.isEmpty(o)){
        this.$container.setGroupHeaders({
            useColSpanStyle: true,
            groupHeaders: o
        });
    }

    let $thArr = $root.find('th:has(table)');

    for (let i = 0; i < $thArr.length; i++){
        $($thArr[i]).attr('title', '');
    }
}

/**
 * scope 속성을 삭제하는 함수
 *
 * @param {object} $container - scope속성을 삭제할 jqeury element
 * @private
 */
function _removeScopeAttr($container = $(document.documentElement)){

    if (_.isEmpty(this.groupHeaders)) return;

    if (!_.isElement($container.get(0))){
        CONSOLE.prop.throw(BxGrid, '$container', 'type');
    }

    $container.find('[scope]').removeAttr('scope');

}

/**
 * colModel에 cellattr을 설정하는 함수
 * http://www.trirand.com/jqgridwiki/doku.php?id=wiki:colmodel_options
 *
 * @param {array} colModels - colModels 객체 배열
 * @private
 */
function _setCellAttr(colModels = []){

    if (!_.isArray(colModels)){
        CONSOLE.prop.throw(BxGrid, 'colModels', 'type');
    }

    colModels.forEach((cm) =>{

        if (!cm.attr) return;

        if (!_.isFunction(cm.attr)){
            CONSOLE.prop.throw(BxGrid, 'colModel.attr', 'type');
        }

        cm.cellattr = function (rowId, val, rowObj, colModel, rData){

            let attr = cm.attr(rowId, val, colModel, rData);

            if (!_.isPlainObject(attr)){
                CONSOLE.prop.throw(BxGrid, 'colModel.attr의 리턴', 'type');
            }

            // console.log(cm.name, colModel);

            return UTIL.objectToStringAttr(attr);

        };

    });

}

/**
 * row를 선택 / 해제 하는 함수
 *
 * @param {string|array} id - row id
 * @param {type} type - 선택 타입 (select / clear)
 * @private
 */
function _setSelection(id = '', type = 'select'){

    if (!_.isString(id) && !_.isArray(id)){
        CONSOLE.prop.throw(BxGrid, 'id', 'type');
    }

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxGrid, 'type', 'type');
    }

    const $container = this.$container;
    id = _.isArray(id) ? id : [id];

    _.forEach(id, (_id) =>{

        if (!_.isString(_id)){
            CONSOLE.prop.throw(BxGrid, 'id', 'type');
        }

        const isSelected = _.indexOf(this.getSelectedId(), _id) !== -1;
        const selectable = type === 'select' ? !isSelected : isSelected;

        if (selectable) $container.setSelection(_id);

    });

}
/**
 * GroupHeader 의 마지막 th border 제거 함수
 * @param {{}} [$headerSubWrapper=null] $headerSubWrapper 'div.ui-jqgrid-hbox'
 * @private
 * @memberOf BxGrid
 */
function _removeGroupHeaderLastThBorder({ $headerSubWrapper = null } = {}){

    if (_.isNull($headerSubWrapper)){
        CONSOLE.prop.throw(BxGrid, '$headerContainer', 'value');
    }

    let { groupHeaders } = this,
        hasGroupHeader = !_.isEmpty(groupHeaders);

    const APPLY_STYLE = 'border-right: 0px !important',
        ORIGIN_HEADER_LAST_TH_CLASS = hasGroupHeader ?
            'thead > tr.jqg-second-row-header > th:visible:last' : 'thead > tr.ui-jqgrid-labels > th:visible:last',
        GROUP_HEADER_LAST_TH_CLASS = 'thead > tr.jqg-third-row-header > th:visible:last';


    let { $elem: $originLastTh, style } = _getLastTh(ORIGIN_HEADER_LAST_TH_CLASS);

    if (hasGroupHeader && $originLastTh.get(0).rowSpan === 1){
        let { $elem: $groupHeaderLastTh, style } = _getLastTh(GROUP_HEADER_LAST_TH_CLASS);

        $groupHeaderLastTh.attr('style', style + APPLY_STYLE);
    }

    $originLastTh.attr('style', style + APPLY_STYLE);

    function _getLastTh(rowClass = ''){

        let $lastTh = $headerSubWrapper.find(rowClass),
            style = $lastTh.attr('style') || '';

        return {
            $elem: $lastTh,
            style
        };
    }
}

/**
 * footers의 값을 통해 footer들을 생성 및 유효성 검사를 한다.
 * @param {array} footers - footers를 구성하는 정보
 * @private
 * @memberOf BxGrid
 */
function _initFooter(footers){

    if (_.isNull(footers))  return;

    let content = '';

    function _addRowFooter(footer, rowNum){
        footer.map(data => {

            if (!_.isString(data.type)){
                CONSOLE.prop.throw(BxGrid, 'footers.type', 'type');
            }
            if (!_.isString(data.colName)){
                CONSOLE.prop.throw(BxGrid, 'footers.colName', 'type');
            }

            if (!_.isString(data.content)){
                CONSOLE.prop.throw(BxGrid, 'footers.content', 'type');
            }
            if (!_.isInteger(data.numberOfCol)){
                CONSOLE.prop.throw(BxGrid, 'footers.numberOfCol', 'type');
            }

            let params = {
                [data.colName]: JSON.stringify(data.content)
            };

            switch (data.type){
                case 'count':
                case 'avg':
                case 'sum':
                case 'min':
                case 'max':
                    content = this.$container.jqGrid("getCol", data.colName, false, data.type);

                    params[data.colName] = content;

                    _addFooterData.call(this, 'set', params, data, rowNum);
                    break;
                case 'label':
                    _addFooterData.call(this, 'set', params, data, rowNum);
                    break;
                default:
                    CONSOLE.prop.throw(BxGrid, 'footers.type', 'type');
            }
        });
    }

    for (let i = 0; i < footers.length; i++){
        _addRowFooter.call(this, footers[i], i);
    }

}

/**
 * 맥 OS 일 경우 스크롤영역에 보더를 추가
 * @private
 * @memberOf BxGrid
 */
function _initMacScrollBorder(){

    if (!$.browser.mac) return;

    const LAST_TD_CLASS = 'tr.ui-row-ltr > td:last-child',
        MAC_SCROLL_CLASS = 'mac-scroll';

    let $container = this.$container;

    $container.find(LAST_TD_CLASS).addClass(MAC_SCROLL_CLASS);
}

/**
 * 해당 열에 footer 추가 또는 해당 열이 이미 있을 경우 업데이트를 한다
 * @param {string} action = 'set' - set or get, 값을 가져 오거나 업데이트
 * @param {string} data = '' - td 영역을 구성 하는 정보
 * @param {object} footer = {} - footer를 구성하는 전체 정보
 * @param {int} rowCount = 0 - 열 번호
 * @private
 * @memberOf BxGrid
 */
function _addFooterData(action = 'set', data = '', footer = {}, rowCount = 0){

    let success = false, res = {}, numberOfCol = footer.numberOfCol, format = true;

    function isEmpty(obj){
        var i;
        for (i in obj){
            if (obj.hasOwnProperty(i)){
                return false;
            }
        }
        return true;
    }

    if (action === undefined){
        action = "get";
    }

    if (typeof format !== "boolean"){
        format = true;
    }

    action = action.toLowerCase();

    this.$container.each(function (){ // jqgrid에서 footer customizing

        let t = this, p = t.p, vl, $td, nm, iCol;

        if (!t.grid || !p.footerrow){
            return false;
        }
        if (action === "set" && isEmpty(data)){
            return false;
        }

        success = true;

        let ftable = $(t.grid.sDiv)
            .find(".ui-jqgrid-hbox>.ui-jqgrid-ftable")[0];

        if (ftable === null || ftable.rows === null){
            return false;
        }

        if (!t.firstRow){

            t.firstRow = $(ftable.rows[0]).clone(true); // 첫 번째 row는 jqgrid가 자체 생성 해주어서 복사 해놓고 재활용

            let html = $(t.grid.bDiv).find('.jqgfirstrow').clone(true); // bodyTable에 있는 jqgfirstrow(fot auto resize) 복사

            $(ftable).find('tbody').prepend(html);
        }

        rowCount += 1;

        if (rowCount > 1){

            if (ftable.rows[rowCount] === undefined){

                $(ftable).find('tbody').append(t.firstRow.clone(true));
            }
        }

        let cells = ftable.rows[rowCount].cells,
            fcells = t.grid.fsDiv === undefined ? {} : t.grid.fsDiv.children(".ui-jqgrid-ftable")[0].rows[rowCount].cells;

        for (nm in data){
            iCol = p.iColByName[nm];
            if (data.hasOwnProperty(nm) && iCol !== undefined){
                if (action === "get"){
                    res[nm] = $(cells[iCol]).html();
                }
                else if (action === "set"){

                    if (footer.type === 'label' || footer.type === 'count'){
                        vl = data[nm];
                    }
                    else {
                        vl = t.formatter("", data[nm], iCol, data, "edit");
                    }

                    let content = footer.content.split('{0}');
                    let temp = '';

                    for (let i = 0; i < content.length; i++){
                        if (i !== 0) temp += vl;
                        temp += content[i];
                    }

                    vl = temp;

                    if (numberOfCol > 1 && $(cells[iCol]).css("display") !== 'none'){

                        for (let i = 1; i < numberOfCol; i++){

                            if (i === 1){
                                $(cells[iCol]).attr("colSpan", numberOfCol);
                            }

                            $(cells[iCol + i]).hide();
                        }

                    }

                    $td = $(cells[iCol]).add(fcells[iCol]);

                    $td.html(vl);

                    // if (p.colModel[iCol].attr) { // 추후 attr 필요할때 열어줌
                    //     $td.attr({ "title": stripHtml(vl) });
                    // }
                }
            }
        }

    });

    return action === "get" ? res : success;
}

/**
 * BxGrid 리사이즈시 footer의 td 간격을 재 조정 한다
 * @private
 * @memberOf BxGrid
 */
function _resizeFooter(){

    if (!this[OPTS].userDataOnFooter) return;

    this.$container.each(function (){

        let t = this, p = t.p;
        let colModel = p.colModel;
        let footTable = $(t.grid.sDiv)
            .find(".ui-jqgrid-hbox>.ui-jqgrid-ftable tr.jqgfirstrow").first()[0].cells;

        colModel.map((data, i) =>{
            footTable[i].style.width = data.width + "px";
        });
    });
}

/**
 * grid 상단에 아이콘을 넣을 수 있는 영역을 추가한다
 *
 * @private
 */
function _addIconWrapper(){

    let $root = this.$root;

    let $iconWarpper = $('<div class="icon-wrap ta-r"></div>');

    $root.find('.ui-widget-header.ui-jqgrid-caption').after($iconWarpper);
}

/**
 * grid에 excel export 버튼을 추가한다
 *
 * @private
 */
function _addExportButton(){

	if (!this.exportModel.enableExpButton) return;

    let $root = this.$root;
    let type = this.exportModel.type;

    let $iconWrapper = $root.find('.icon-wrap');
    let buttonText = `${EXPORT_MODEL[type].name} ${EXPORT_MODEL.downloadText}`;

    let buttonHtml = [];

    buttonHtml.push(`<button type="button" class="bx-txt-btn" title="${buttonText}">`);
    buttonHtml.push(`<span>${buttonText}</span>`);
    buttonHtml.push(`<i class="bw-icon i-25 i-ctrl-down"></i>`);
    buttonHtml.push('</button>');

    let $exportButton = $(buttonHtml.join(''));

    _addEventListenerForExportIconButton.call(this, $exportButton, type);

    $iconWrapper.append($exportButton);
}

/**
 * grid에 excel import 기능 추가
 *
 * @private
 */
function _addImportExcel() {

    if (_.isEmpty(this.importModel)) return;

    let $iconWrapper = this.$root.find('.icon-wrap');

    _addFileUploadButton.call(this, $iconWrapper);
}

/**
 * grid에 excel import 버튼을 추가한다
 *
 * @param {object} $iconWrapper - 생성 위치 타켓
 *
 * @private
 */
function _addFileUploadButton($iconWrapper = {}){

    let importModel = this.importModel;
    let type = importModel.type;
    let ajaxOption = this.importModel.ajaxOption;
    let url = ajaxOption.url;
    let enableImpButton = importModel.enableImpButton;

    let buttonText = `${IMPORT_MODEL[type].name} ${IMPORT_MODEL.uploadText}`;
    let html = [];

    let {$form: $form, $label:$importButton, $iframe:$iframe } = FORM.createFileUploadElements({
        'prefixId':'bxGrid',
        'action': url,
        'onFileChange': _sendExcelFile.bind(this)
    });

    if (UTIL.isIE9()) {
        $iconWrapper.append($form);
        $iconWrapper.append($iframe);
    }
    else {
        $iconWrapper.append($form.find('input'));
    }

    $iconWrapper.append($importButton);

    html.push(`<span>${buttonText}</span>`);
    html.push('<i class="bw-icon i-25 i-ctrl-up"></i>');

    $importButton.addClass('bx-txt-btn');
    $importButton.append($(html.join('')));

    if (!enableImpButton) $importButton.addClass('blind');
}

/**
 * 파일을 전송
 *
 * @param {object} e - 이벤트 객체
 *
 * @private
 */
function _sendExcelFile(e = {}){

    let $target = $(e.target);
    let ajaxOption = _getAjaxOption.call(this, AJAX_OPTION.IMPORT_OPTION);

    let $iconWrapper = this.$root.find('.icon-wrap');
    let $form = $iconWrapper.find('form');
    let $iframeElem = $iconWrapper.find('iframe');

    if (_.isUndefined(this.importModel)) CONSOLE.prop.throw(BxGrid, 'importModel', 'all');

    _.assign(ajaxOption.params, {'colModel': _getReverseColModel.call(this)});

    FORM.submit({
            "$formElem" : $form,
            "$iframeElem" : $iframeElem,
            "url": ajaxOption.url,
            "method": ajaxOption.method,
            "file": $target,
            "headers": ajaxOption.headers,
            "params": ajaxOption.params,
            "before": xhr => {
                ajaxOption.before.call(ajaxOption);
             },
            "success": data => {

                ajaxOption.success();

               _setDataForImportExcel.call(this,  data);
           },
            "error": ajaxOption.fail
        });

    FORM.clear($target);
}

/**
 * 서버로 부터 받은 data를 그리드에 바인딩
 *
 * @param {string} data - json 구조의 아이템
 *
 * @private
 */
function _setDataForImportExcel(data = null) {

	const importModel = this.importModel;
	const resModel = _getTransModel.call(this, AJAX_OPTION.IMPORT_OPTION).resModel;
    const enableResetData = importModel.enableResetData;
    
	if (_.isString(resModel) && resModel.length){
		data = _.get(data, resModel, data);
	}
	
    if (!_.isArray(data)){
        CONSOLE.prop.throw(BxGrid, ERR_MSG.notFoundItemViaResModel);
    }

    if (_isAjaxTransaction.call(this)){

        // 서버에 데이터를 재요청 한다.
        this.loadData();
    }
    else {

        let items = data;

        if (!enableResetData){
            items = _.concat(this.getAllRowData(), data);
        }

        this.resetGridData(items);
    }
}

/**
 * colModel 의 key 와 value 의 위치를 변경
 *
 * @private
 */
function _getReverseColModel() {

    let colModel = this[OPTS].colModel;
    let reverseCol = {};

    colModel.forEach((data, i) =>{

        if (data['hidden']) return;

        reverseCol[data['label']] = data['name'];
    });

    return JSON.stringify(reverseCol);
}

/**
 * export 아이콘에 클릭이벤트를 등록한다
 *
 * @param {object} $target = null - 타켓
 * @param {string} type = '' - 버튼 타입
 *
 * @private
 */
function _addEventListenerForExportIconButton($target = null, type = ''){

    switch(type){

        case 'excel': {

            $target.on('click', e => this.exportExcel.call(this));

            break;
        }
    }
}

/**
 * Time Stamp를 생성한다
 *
 * @returns {string} - 생성된 TimeStamp
 *
 * @private
 */
function _getTimeStamp(){

    let date = new Date();

    return UTIL.convertDateFormat(date.toString(), 'yyyyMMddhhmmss', true);
}

/**
 * export 파일을 생성 후 실행한다.
 *
 * @param {array} pageDatas - export 할 페이지 데이터 목록
 *
 * @private
 */
function _createExportFile(pageDatas = []){
	
	const exportModel = this.exportModel;
	const resModel = _getTransModel.call(this, AJAX_OPTION.EXPORT_OPTION).resModel;

	if (_.isString(resModel) && resModel.length){
		pageDatas = _.get(pageDatas, resModel, pageDatas);
	}

    if (!_.isArray(pageDatas)){
        CONSOLE.prop.throw(BxGrid, ERR_MSG.notFoundItemViaResModel);
    }

    let fileName = exportModel.fileName;
    let enableTimeStamp = exportModel.enableTimeStamp;
    let sheetName = exportModel.sheetName;

    let $exportTable = _getExportDOM.call(this, pageDatas);

    fileName = enableTimeStamp ? `${fileName}_${_getTimeStamp()}` : fileName;

    EXPORTER.exportExcel($exportTable, fileName, sheetName);
}

/**
 * export될 DOM을 가져온다
 *
 * @param {null|array|int} startIdx - 시작 페이지 번호
 * @param {int} endIdx - 마지막 페이지 번호
 * @param {object} pageDatas - 만들어질 row 데이터
 *
 * @returns {Object}
 *
 * @private
 */
function _getExportDOM(pageDatas = []){

    if (!pageDatas.length){
        CONSOLE.prop.throw(BxGrid, 'pageDatas', 'value');
    }

    let $header = $(this.$root.find('thead')[0].outerHTML);
    let footers = this.footers;

    let $wrap = $('<div></div>');
    let $table = $('<table></table>');

    let styleText = _createExportExcelStyle();
    let header = _extractHeaderDOM.call(this, $header);
    let rowDataHTML = _changePageDataToDOM.call(this, pageDatas);
    let footerHTML = _createFooterForExportDOM.call(this, footers, pageDatas);

    $table.append(header);
    $table.append(rowDataHTML);
    $table.append(footerHTML);

    $wrap.append(styleText);
    $wrap.append($table);

    return $wrap;
}

/**
 * export excel 파일의 style을 생성한다
 *
 * @returns {string}
 * @private
 */
function _createExportExcelStyle(){

    let style = [];

    style.push('<style type="text/css">');
    style.push('table th {');
    style.push('background-color: "#eee";');
    style.push('border: 1px solid;');
    style.push('}');
    style.push('</style>');

    return style.join('');
}

/**
 * export 될 header DOM으로 수정한다
 *
 * @param {object} $header - 생성된 jqgrid header 영역
 *
 * @returns {object} - 수정된 header 영역 *
 * @private
 */
function _extractHeaderDOM($header = null){

    // 첫번째 빈 tr 삭제
    $header.find('.jqg-first-row-header').remove();

    // caption 삭제
    $header.find('caption').remove();

    // checkbox 삭제
    $header.find('input').remove();

    // sort button 삭제
    $header.find('button').remove();

    // rownumber 헤더 삭제
    $header.find('.jqgh_rn').remove();

    // checkbox 헤더 삭제
    $header.find('.jqgh_cbox').remove();

    // inline 편집 헤더 삭제
    $header.find('th[id$="format"]').remove();

    // header resize 영역 삭제
    $header.find('span.ui-jqgrid-resize').remove();

    // display none 되어있는 th는 삭제
    _.map($header.find('th'), th => {

        let $th = $(th);

        if ($th.css('display') === 'none'){
            $th.remove();
        }
    });

    return $header.html();
}

/**
 * grid row 데이터를 가져온다
 *
 * @param {null|array|int} _startIdx - 시작 페이지 번호
 * @param {int|null} _endIdx - 마지막 페이지 번호
 *
 * @returns {Object[]} - grid row data
 *
 * @private
 */
function _extractPageData(_startIdx = null, _endIdx = null){

    let {startIdx, endIdx} = _validatePageIdx(_startIdx, _endIdx);

    let resultDatas = [];
    const mainDefer = $.Deferred();
    const resultPromiseFunc = (promises) => {

        $.when(...promises).done((...pageRowData) => {

            let pageDatas = _.union(...pageRowData) || [];

            return mainDefer.resolve(pageDatas);
        });
    };

    // 특정 페이지 데이터 가져오기
    if (_.isNull(startIdx)){
        resultDatas.push(this.getAllRowData());
    }
    else if (_.isArray(startIdx)){
        _.forEach(startIdx, idx => {
            resultDatas.push(this.getPageData(idx));
        });
    }
    else {
        for (let idx = startIdx; idx <= endIdx; idx++){
            resultDatas.push(this.getPageData(idx));
        }
    }

    // ajaxModel 사용 유무에 따른 처리방식 변경
    if (_isAjaxTransaction.call(this)){

        resultPromiseFunc(resultDatas);

        return mainDefer.promise();
    }
    else {
        return _.union(...resultDatas);
    }
}

/**
 * grid row 데이터를 가져온다
 *
 * @param {null|array|int} _startIdx - 시작 페이지 번호
 * @param {int|null} _endIdx - 마지막 페이지 번호
 *
 * @returns {Object[]} - grid row data
 *
 * @private
 */
function _validatePageIdx(startIdx = null, endIdx = null){

    if (!_.isNull(startIdx) && !_.isArray(startIdx)){

        startIdx = parseInt(startIdx, 10);

        if (isNaN(startIdx)){
            CONSOLE.prop.throw(BxGrid, 'startIdx', 'all');
        }
    }

    if (!_.isNull(endIdx)){

        endIdx = parseInt(endIdx, 10);

        if (isNaN(endIdx)){
            CONSOLE.prop.throw(BxGrid, 'endIdx', 'all');
        }
    }

    // 시작 페이지가 종료 페이지보다 크거나, endIdx가 null인 경우에 종료 페이지를 시작페이지와 동일하게 변경
    if (_.isNull(endIdx) || startIdx >= endIdx){
        endIdx = startIdx;
    }

    return {startIdx, endIdx};
}

/**
 * grid row 데이터를 DOM으로 변환한다
 *
 * @param {object[]} pageDatas - page row data
 *
 * @returns {string} 변환된 html
 *
 * @private
 */
function _changePageDataToDOM(pageDatas = []){

    const headerObj = _extractHeaderToColModel.call(this);
    const isFormatted = this.exportModel.dataFormat === EXPORT_MODEL.dataFormat.formatted ? true : false;

    let rowHtml = [];

    _.forEach(pageDatas, item => {

        rowHtml.push('<tr>');

        _.forEach(headerObj, headerItem => {

            let formatType = _getFormatStyle(headerItem.formatOption);
            let cellValue = item[headerItem.name];
            let id = item.id;

            if (isFormatted && !_.isUndefined(headerItem.formatOption) && headerItem.formatOption.type !== 'currency'){
                cellValue = _getFormatString.call(this, cellValue, headerItem.name, id);
            }

            rowHtml.push(`<td style='${formatType}'>${cellValue}</td>`);
        });

        rowHtml.push('</tr>');
    });

    return rowHtml.join('');
}

/**
 * 엑셀에 export 될 값의 style을 지정한다
 *
 * @param formatOption - format object
 *
 * @returns {string} excel format style
 *
 * @private
 */
function _getFormatStyle(formatOption = {}){

    let textType = 'mso-number-format:"\@"';
    let currencyType = 'text-align:right; mso-number-format:"\#\,0';

    let formatType = formatOption.type || '';
    let options = formatOption.options;

    if (!_.isPlainObject(formatOption)){
        return textType;
    }
    else if (formatType === FORMAT_TYPE.CURRENCY){

        if (!_.isUndefined(options.decimalPlaces) && options.decimalPlaces !== 0){

            currencyType += '.';

            for (let i = 0; i < options.decimalPlaces; i++){
                currencyType += '0';
            }
        }

        currencyType += '"';

        return currencyType;
    }
    else {
        return textType;
    }
}

/**
 * export 될 footer의 DOM을 생성한다
 *
 * @param {object} footers - footer 생성 option object
 * @param pageDatas - grid row datas
 *
 * @returns {string} 생성된 footer DOM
 *
 * @private
 */
function _createFooterForExportDOM(footers = {}, pageDatas = []){

    const singleCol = 1;

    let headerNameList = _extractHeaderToColModel.call(this);
    let trHTML = [];

    _.forEach(footers, tr => {

        let numberOfCol = singleCol;

        trHTML.push('<tr>');

        _.forEach(headerNameList, headerObj => {

            let tdHTML = '';

            if (numberOfCol === singleCol){

                let item = _.filter(tr, ['colName', headerObj.name]);

                if (item.length){

                    item = item[0];
                    tdHTML = _createTdHTML.call(this, item, pageDatas, headerObj.name);
                }

                if (item.numberOfCol > singleCol) numberOfCol = item.numberOfCol;
            }

            if (tdHTML.length){
                trHTML.push(tdHTML);
            }
            else if (numberOfCol > singleCol){
                numberOfCol--;
            }
            else {
                trHTML.push('<td></td>');
            }
        });

        trHTML.push('</tr>');
    });

    return trHTML.join('');
}

/**
 * footer의 td를 생성한다
 *
 * @param {object} item - 표현 될 footer object
 * @param {array} pageDatas - grid page row data
 * @param {string} cellName - header object
 *
 * @returns {string} td html
 *
 * @private
 */
function _createTdHTML(item = {}, pageDatas = [], cellName = ''){

    const REPLACE_CHAR = '{0}';

    let cellValue = item.type === 'label' ? item.content : _calcFooterValue(item.type, item.colName, pageDatas);

    cellValue = _getFormatString.call(this, cellValue, cellName);

    let replaceValue = item.content.replace(REPLACE_CHAR, cellValue);
    let colspanText = '';

    if (item.numberOfCol > 1) colspanText = `colspan='${item.numberOfCol}'`;

    return `<td ${colspanText}>${replaceValue}</td>`;
}

/**
 * 값을 포메팅 해준다
 *
 * @param {string} cellValue - 변환 할 값
 * @param {string} cellName - 변환 할 값의 헤더 이름
 * @param {string} rowId - rowId
 *
 * @returns {string} 변환된 값
 * @private
 */
function _getFormatString(cellValue = '', cellName = '', rowId = ''){

    let jqGrid = this.$container.get(0);
    let p = jqGrid.p;
    let colPos = p.iColByName[cellName];

    let resultString = cellValue;

    // jqgrid formatter를 이용해서 포매팅 관련
    // UTIL.getFormatString을 이용하여 커스텀 타입을 포매팅 하려면 customFormatter 의 arguments를 넣어줘야 포매팅이 가능하다.
    // 하지만 jqgrid에서 직접 만들어서 주는 arguments도 있어 해당 parameter는 생성해서 넣어줘야하나 매우 비효율적이라고 판단,
    // UTIL.getFormatString 대신 내장 jqgrid 함수로 포매팅하는 방법으로 변경
    resultString = jqGrid.formatter(rowId, cellValue, colPos, {[cellName]: cellValue}, "get");

    return resultString;
}

/**
 * colModel에서 Grid에 보여지는 header 정보를 순서대로 추출한다
 *
 * @returns {Array} 수집된 header 정보
 *
 * @private
 */
function _extractHeaderToColModel(){

    let headerObj = [];
    const colModel = this.getOPTS().colModel;

    _.forEach(colModel, item => {

        if ((_.isUndefined(item.hidden) || item.hidden === false) && item.name !== 'format'){
            headerObj.push({name: item.name, formatOption: item.formatOption});
        }
    });

    return headerObj;
}

/**
 * footer에 출력 될 값을 계산한다
 *
 * @param {string} type - footer 값을 계산할 타입
 * @param {string} colName - 검색 할 colName
 * @param {object[]} pageData - grid page row 데이터
 *
 * @returns {int} 계산된 값
 *
 * @private
 */
function _calcFooterValue(type = '', colName = '', pageData = []){

    let result = null;

    switch(type){

        case 'count' : {

            result = pageData.length;
            break;
        }
        case 'sum' : {

            result = 0;

            _.forEach(pageData, item => result += Number(item[colName]));

            break;
        }
        case 'avg' : {

            result = 0;

            _.forEach(pageData, item => result += Number(item[colName]));

            result /= pageData.length;

            break;
        }
        case 'min' : {

            _.forEach(pageData, item => {

                let compareValue = Number(item[colName]);

                if(_.isNull(result) || compareValue < result){
                    result = compareValue;
                }
            });

            break;
        }
        case 'max' : {

            _.forEach(pageData, item => {

                let compareValue = Number(item[colName]);

                if(_.isNull(result) || compareValue > result){
                    result = compareValue;
                }
            });

            break;
        }
    }

    return result;
}

/**
 * ajax 사용 여부를 반환한다.
 *
 * @returns {boolean} ajax 사용 여부
 *
 * @private
 */
function _isAjaxTransaction(){
    return this.ajaxModel.isAjaxTransaction;
}

/**
 * ajax option 초기화 함수
 *
 * @param {string} optionName - Ajax 통신때 사용할 옵션명
 *
 * @returns {object} - ajaxOption
 *
 * @private
 */
function _initAjaxOption(optionName = ''){

    let ajaxOption = _getAjaxOption.call(this, optionName);

	if (!_.isString(ajaxOption.url)){
        ajaxOption.url = '';
    }

	if (!_.isString(ajaxOption.method)){
        ajaxOption.method = 'post';
    }

    if (!_.isPlainObject(ajaxOption.headers)){
        ajaxOption.headers = {};
    }

    if (!_.isPlainObject(ajaxOption.params)){
        ajaxOption.params = {};
    }

    if (!_.isFunction(ajaxOption.before)){
        ajaxOption.before = function(){};
    }

    if (!_.isFunction(ajaxOption.success)){
        ajaxOption.success = function(){};
    }

    if (!_.isFunction(ajaxOption.fail)){
        ajaxOption.fail = function(){};
    }

    let ajaxModel = _getAjaxModel.call(this, optionName);

    if (optionName === AJAX_OPTION.IMPORT_OPTION){

        if (!_.isString(ajaxModel.resModel)){
            ajaxModel.resModel = DEFAULT_RES_MODEL;
        }
    }
    else if (optionName === AJAX_OPTION.EXPORT_OPTION){

        if (!_.isString(ajaxModel.resModel)){
            ajaxModel.resModel = DEFAULT_RES_MODEL;
        }
    }

    return ajaxOption;
}



module.exports = BxGrid;
