// bxuip-angular.js-ui/src/lib/bx-ui/app/components/img.js
/**
 * Created by hyunjang on 2017. 7. 3..
 */
/**
 * @class BxImg
 * @classDesc BxImg 클래스 생성자
 * @example
 *
 * <bx-img
 *      id="image1"
 *      url="http://68.media.tumblr.com/83ffbae5b815b091037d13548fd5b37e/tumblr_orwck9lRKi1slhhf0o1_1280.jpg"
 *      errUrl="http://68.media.tumblr.com/f5fdcf32198a2abb56508701ee65e41b/tumblr_osnjnznDb21slhhf0o1_1280.jpg"
 *      shape="square"
 *      overlayStyle="full"
 *      size="medium"
 *      frameType="default"
 *      data-width="400"
 *      data-height="400"
 *      data-title="테스트 타이틀 입니다"
 *      captionAlign="center"
 *      caption="caption test1"
 *      desc="align >> center"
 *      btnText="button">
 * </bx-img>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *  // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxImg().initialize();
 *
 *      var image = document.getElementById('image1');
 *
 *      // 이미지 변경
 *      image.url = "http://68.media.tumblr.com/a2aa66013283367f774b020c83ad3ace/tumblr_ol9e886Wm61slhhf0o1_1280.jpg";
 *
 *      // 이미지 설명 변경
 *      image.desc = "이미지 설명입니다";
 *
 *      // 이미지 캡션
 *      image.caption = "버튼입니다";
 *
 *      // 이미지 내부 정렬
 *      image.captionAlign = "left";
 *
 *      // 이미지 가로 사이즈 변경
 *      image['data-width'] = 100;
 *
 *      // 이미지 가로 사이즈 변경
 *      image['data-height'] = 100;
 *
 *      // 이미지 타이틀 변경
 *      image['data-title'] = '테스트 타이틀';
 *
 *      // 이미지 렌더링 callback 설정
 *      image.bxrender = function(){
 *          console.log('render complete');
 *      }
 *
 *      // 이미지 버튼 클릭 callback 설정
 *      image.bxbtnclick = function(event){
 *          console.log('button clicked', event);
 *      }
 *
 * });
 *
 * </script>
 */

/**
 * @property {function} [bx-render = function(){}] - 컴포넌트 render 시, 호출될 callback 함수
 * @callback bxrender
 *
 * @memberOf BxImg
 */

/**
 * @property {function} [bx-btnClick = function(){}] - 이미지 버튼 클릭 시, 호출될 callback 함수
 * @callback bxbtnClick
 *
 * @param {object} event - 버튼 클릭시 callback 값
 * @memberOf BxImg
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const FILE = BXUI.File;

// 이미지 모양 타입
const SHAPE = {
    "square": 'square',
    "round": 'rounded',
    "circle": 'circle'
};

// 이미지 크기
const SIZE = {
    "small": {
        "class": 'size-s',
        "value": 150
    },
    "medium": {
        "class": 'size-m',
        "value": 300
    },
    "large": {
        "class": 'size-l',
        "value": 500
    }
};

// 캡션 및 버튼 정렬
const ALIGN = {
    'center': '',
    'left': 'a-left',
    'right': 'a-right'
};

// Overlay Style
const OVERLAY_STYLE = {
    "half": 'half-type',
    "full": ''
};

const FRAME_TYPE = {
    "default": 'default',
    "horizon": 'fluid-width',
    "vertical": 'fluid-height'
};

const ATTRNAMES = {
    "errurl": "errUrl",
    "overlaystyle": "overlayStyle",
    "enablespotlight": "enableSpotlight",
    "captionalign": "captionAlign",
    "btntext": "btnText",
    "frametype": "frameType",
    "bx-btnclick": "bx-btnClick"
};

const READONLY_PROPERTY = {
    "shape": 'shape',
    "errUrl": 'errUrl',
    "size": 'size',
    "frameType": 'frameType',
    "overlayStyle": 'overlayStyle',
    "enableSpotlight": 'enableSpotlight'
};

const CLASS_NAME = {
    "wrap": 'bx-img-wrap',
    "desc": 'img-desc',
    "btn": 'img-btn'
};

// default 설정
const OPTS = {
    /**
     * @property {object} bx-img
     * @property {string} bx-img.url='' - 이미지 경로
     * @property {string} [bx-img.errUrl=''] - <span class='important'>readonly</span> - 에러시 출력할 이미지 경로
     * @property {string} [bx-img.shape='square'] - <span class='important'>readonly</span> - 이미지 모양 (square, round, circle)
     * @property {string} [bx-img.desc=''] - 이미지 하단 설명하는 텍스트 문구
     * @property {string} [bx-img.size=''] - <span class='important'>readonly</span> - 이미지 사이즈 (small, medium, large)
     * @property {object} [bx-img.data-width=null] - 이미지 가로 사이즈
     * @property {object} [bx-img.data-height=null] - 이미지 세로 사이즈
     * @property {string} [bx-img.frameType='default'] - <span class='important'>readonly</span> - 이미지 채우기 방향 설정 (default, horizon, vertical)
     * @property {string} [bx-img.overlayStyle='full'] - <span class='important'>readonly</span> - 마우스 오버시 overlay 모양 (half, full)
     * @property {string} [bx-img.captionAlign='center'] - 마우스 오버시 캡션 텍스트와 버튼의 정렬 위치 (left, center, right) <span class='important'>overlayStyle이 half 일 때 만 적용</span>
     * @property {string} [bx-img.caption=''] - 마우스 오버시 캡션 텍스트 문구
     * @property {string} [bx-img.btnText=''] - 마우스 오버시 버튼의 문구
     * @property {function} [bx-img.bx-render=function(){}] - 이미지 렌더링 완료 시, 호출될 callback 함수
     * @property {function} [bx-img.bx-btnClick=function(){}] - 이미지 내부 버튼 클릭 시, 호출될 callback 함수
     * @property {function} [bx-img.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
     *
     * @alias Attributes
     * @memberOf BxImg
     */
    // 이미지 관련 옵션
    "url": '',                              // 이미지 경로 - (get, set)
    "errUrl": '',                           // 에러시 출력할 이미지 경로 - (readonly)
    "shape": SHAPE.square,                   // 이미지 모양 - (readonly) square, round, circle
    "desc": '',                             // 이미지 외부 텍스트 - (get, set) 이미지 하단 문구
    "size": '',                             // 이미지 크기 - (readonly) small, medium, large
    "frameType": FRAME_TYPE.default,                  // 이미지 영역 - (readonly) default, horizon, vertical

    // dim 관련 옵션
    "overlayStyle": OVERLAY_STYLE.full,                     // 마우스 오버시 dim 옵션 - (readonly) half, full, clear

    // Spotlight 관련 옵션
    // "enableSpotlight": false,                // 라이트박스 표시 유무 - (readonly) true, false,

    // 캡션 관련 옵션
    "captionAlign": ALIGN.center,             // 캡션 텍스트 정렬 - (get, set) left, center
    "caption": '',                      // 캡션 텍스트 (html 적용) - (get, set)

    // 버튼 관련 옵션
    "btnText": '',                    // 버튼 텍스트 - (readonly)

    "data-width": null,
    "data-height": null,
    "data-title": null,

    // callback 관련 옵션
    "bx-render": function(){},              // 렌더링 후 callback - (get, set)
    "bx-btnClick": function(){}             // 버튼 클릭 callback - (get, set)
};

// const APIS = {
//     "show": _show,
//     "hide": _hide
// };

const DIM_STYLE = {
    color: 'black',
    enableBlur: true,
    enableIcon: true,
    iconType: 'think'
};

const HEX = 16777215;                           // ffffff 10진수값
const UPDATE_DURATION = 350;                    // 이미지 appear 대기시간
const ERROR_IMG = 'assets/images/no-img.png';   // 에러 이미지 경로

class BxImg {

    constructor(){

        this.opts = OPTS;
        // super();
    }

    /**
     * 초기화 함수
     *
     * @returns {object} 생성된 인스턴스
     *
     * @memberOf BxImg
     */
    initialize(){

        this.$img = this.$root.find('bx-img');

        this.elems = [];

        _.forEach(this.$img, (elem) => {

            const $customElem = $(elem);

            // 사용자 커스텀 엘리먼트를 파싱한다.
            const $component = _parseCustomElement.call(this, $customElem);

            $component.prop('bxrender').call($component.get(0));

            this.elems.push($component.get(0));
        });

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxImg.isCustomElement = true;

//////////////////////////////////////
//  APIS
//////////////////////////////////////
// function _show(){
//     console.log('showLightbox');
// }
//
// function _hide(){
//     console.log('hideLightbox');
// }

//////////////////////////////////////
//  private function
//////////////////////////////////////

/**
 * 사용자 `custom` 태그를 `native` 태그로 대치시킨다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - 사용자 custom 엘리먼트
 *
 * @returns {object} 생성한 $container 엘리먼트
 *
 * @private
 *
 * @memberOf BxImg
 */
function _parseCustomElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxImg, '$customElem', 'type');
    }

    // $container 엘리먼트를 생성한다.
    const $container = BASE.appendContainerElement(BxImg.$root, BxImg.className);

    // 컴포넌트 엘리먼트를 생성한다.
    const $component = _getComponentElement.call(this, $customElem);

    // 컴포넌트 이벤트 리스너 등록
    _addDocumentElementEventListener.call(this, $component);

    // 반환받은 `컴포넌트 엘리먼트`를 `컨테이너 엘리먼트`의 자식으로 추가한다.
    $container.append($component);

    // $customElem 엘리먼트를 생성된 `$container` 엘리먼트로 대치시킨다.
    $customElem.replaceWith($container);

    // 이미지 업데이트
    _update($component);

    return $component;
}

/**
 * component 엘리먼트를 반환한다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - 사용자 custom 엘리먼트
 *
 * @returns {object} 생성한 component 엘리먼트
 *
 * @private
 *
 * @memberOf BxImg
 */
function _getComponentElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxImg, '$customElem', 'type');
    }

    const opts = _createComponentOptions.call(this, $customElem, OPTS);

    const $container = _createComponentContainer.call(this, opts);

    // native event 바인드
    UTIL.addAllEventListenerWithProperty($container, opts.events);

    return $container;
}

/**
 * 각 `custom` 엘리먼트에 정의된 attr 를 opt 객체로 반환한다.
 *
 * @param {object} [targetObject=null]
 * @param {object} [defaultOpts={}]
 *
 * @returns {object} 생성된 opt 객체
 *
 * @private
 * @memberOf BxImg
 */
function _createComponentOptions(targetObject = null, defaultOpts = {}){

    if (_.isNull(targetObject)){
        CONSOLE.prop.throw(BxImg, 'targetObject', 'type');
    }

    if (!_.isFunction(targetObject.get) && !_.isPlainObject(targetObject)){
        CONSOLE.prop.throw(BxImg, 'targetObject', 'type');
    }

    if (!_.isPlainObject(defaultOpts)){
        CONSOLE.prop.throw(BxImg, 'defaultOpts', 'type');
    }

    let opts = UTIL.createComponentOptions(targetObject, defaultOpts, ATTRNAMES, function(k, v){

        // 최초 설정시 유효성 검사
        if (k === 'url' || k === 'errUrl' || k === 'caption' || k === 'desc' || k === 'btnText' || k === 'data-title'){

            if (!_.isString(v)){
                CONSOLE.prop.throw(BxImg, k, 'type');
            }
        }

        if (k === 'shape' && _.isUndefined(SHAPE[v])){
            CONSOLE.prop.throw(BxImg, k, 'type');
        }

        if (k === 'size' && _.isUndefined(SIZE[v])){
            CONSOLE.prop.throw(BxImg, k, 'type');
        }

        if (k === 'captionAlign' && _.isUndefined(ALIGN[v])){
            CONSOLE.prop.throw(BxImg, k, 'type');
        }

        if (k === 'overlayStyle' && _.isUndefined(OVERLAY_STYLE[v])){
            CONSOLE.prop.throw(BxImg, k, 'type');
        }

        if (k === 'frameType' && _.isUndefined(FRAME_TYPE[v])){
            CONSOLE.prop.throw(BxImg, k, 'type');
        }

        if (k === 'data-width' || k === 'data-height'){

            v = parseInt(v, 10);

            if (isNaN(v)){
                CONSOLE.prop.throw(BxImg, k, 'type');
            }
        }

        return v;

    }, function(k, v){

        return v;
    });

    return opts;
}

/**
 * image 컴포넌트를 생성한다
 *
 * @param {object} opts={} - 옵션 객체
 * @returns {object} 생성된 image component container
 *
 * @private
 * @memberOf BxImg
 */
function _createComponentContainer(opts = {}){

    let html = [];
    let imgAttr = UTIL.objectToStringAttr(opts.attr);

    let originSrc = opts.url;
    let size = _.isUndefined(SIZE[opts.size]) ? '' : SIZE[opts.size].class;
    let frameType = FRAME_TYPE[opts.frameType] || FRAME_TYPE.default;
    let shape = SHAPE[opts.shape] || opts.shape;
    let captionAlign = ALIGN[opts.captionAlign] || opts.captionAlign;
    let overlayStyle = OVERLAY_STYLE[opts.overlayStyle] || '';

    let fileName = FILE.getFileNameInfo(opts.url).name;
    let alt = opts.attr.alt || fileName;
    let title = opts['data-title'] || fileName;

    let desc = opts.desc;
    let caption = opts.caption;
    let btnText = opts.btnText;

    let width = opts['data-width'] || '';
    let height = opts['data-height'] || '';
    let widthStyle = width === '' ? width : `width:${width}px;`;
    let heightStyle = height === '' ? height : `height:${height}px;`;

    // overlayStyle 이 half 일 때 만 captionAlign 을 적용시킨다
    if (overlayStyle === '') captionAlign = ALIGN.center;

    // 사용자가 width 나 height 를 넣은 경우 size 를 없앤다
    if (width !== '' || height !== '') size = '';

    html.push(`<div class="${CLASS_NAME.wrap}">`);
    html.push(`<figure title="${title}" class="${shape} ${size} ${frameType} ${overlayStyle} ${captionAlign}" style="${widthStyle} ${heightStyle}">`);
    html.push(`<img ${imgAttr} alt="${alt}" title="${title}" />`);
    html.push(`<div class="dim-area"></div>`);
    html.push('<figcaption>');
    html.push(`<p>${caption}</p>`);
    html.push('</figcaption>');
    html.push(`<div class="${CLASS_NAME.btn}">`);

    if (btnText !== '') html.push(`<button type="button">${btnText}</button>`);

    html.push('</div>');
    html.push('</figure>');
    html.push(`<div class="${CLASS_NAME.desc}">`);
    html.push(`<p>${desc}</p>`);
    html.push('</div>');
    html.push('</div>');

    const $wrap = $(html.join(''));
    const $image = $wrap.find('img');
    const $button = $wrap.find('button');
    const $dimArea = $wrap.find('.dim-area');

    // 최초에 이미지는 hide처리 (ie의 아이콘 때문에)
    $image.hide();

    // 최초 컴포넌트 사이즈 설정
    _setComponentSize($dimArea, {size, width, height});

    // 컴포넌트 프로퍼티 설정
    _setElementProperties.call(this, opts, $wrap.get(0));

    // paper 설정
    // _clickOpenPaper.call(this, $image);

    // lazyload 설정
    _setLazyLoad.call(this, $image, originSrc);

    // button 설정
    $button.on('click', e => $wrap.prop('bxbtnClick').call($wrap.get(0), e));

    // dim 활성화
    _showDim.call(this, $dimArea, $image);

    return $wrap;
}


/**
 * image 엘리먼트의 프로퍼티를 설정한다.
 *
 * @param {object} [opts={}] - 옵션 객체
 * @param {object} [elem=null] - 프로퍼티가 추가될 엘리먼트
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxImg
 */
function _setElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxImg, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxImg, 'elem', 'type');
    }

    const bxPtn = /^bx/;

    _.map(opts, (v, k) => {

        // attr 속성은 무시한다.
        if (k === 'attr') return;
        // events 속성은 무시한다.
        if (k === 'events') return;

        // 읽기 전용 속성
        if (READONLY_PROPERTY[k]){
            UTIL.setReadOnlyProperty(elem, k, v);
        }
        else{

            if (k.match(bxPtn)){

                let _k = k.replace('-', '');
                delete opts[_k];

                k = _k;
            }

            let _k = UTIL.convertInternalPropertyName(k);

            elem[_k] = v;

            UTIL.setPropertyWithDescriptor(elem, k, {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: v => {

                    if (k === 'url' || k === 'desc' || k === 'caption' || k === 'btnText' || k === 'data-title'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxImg, k, 'type');
                        }
                    }

                    if (k === 'captionAlign' && _.isUndefined(ALIGN[v])){
                        CONSOLE.prop.throw(BxImg, k, 'type');
                    }

                    elem[_k] = v;

                    if (k === 'url'){
                        // $(elem).attr('src', v);
                        _update.call(this, $(elem).parents(`.${CLASS_NAME.wrap}`));
                    }

                    if (k === 'desc' || k === 'caption' || k === 'btnText' || k === 'data-title'){
                        _changeText(elem, k, v);
                    }

                    if (k === 'captionAlign'){
                        _changeTextAlign(elem, v);
                    }

                    if (k === 'data-width' || k === 'data-height'){
                        _changeComponentSize(elem, k, v);
                    }
                }
            }, v);
        }
    });

    // API 에 대한, 프로퍼티 할당
    // _.map(APIS, (v, k) => {
    //     UTIL.setReadOnlyProperty(elem, k, v);
    // });
}

/**
 * lazyload를 설정한다
 *
 * @param {object} $image - image container
 * @param {string} originImage - 사용자가 설정한 보여줄 이미지
 *
 * @private
 * @memberOf BxImg
 */
function _setLazyLoad($image = {}, originImage = ''){

    $image.one('appear', () => {

        let $figureWrap = $image.parent();
        let $dimContainer = $figureWrap.children('.dim-area');
        let $dim = $image.prop('$dim');

        let frameType = $image.prop('frameType');
        let size = $image.prop('size');
        let width = $image.prop('data-width');
        let height = $image.prop('data-height');
        let title = $image.prop('data-title');

        $image
            .on('load', e => {

                // dim container hide
                $dimContainer.hide();
                // dim 해제
                $dim.hide();

                let sizeClass = '';

                // 이미지 사이즈가 있고 width, height값이 없는 경우에만 size class 추가
                if (!_.isUndefined(SIZE[size]) && _.isNull(width) && _.isNull(height)){
                    sizeClass = SIZE[size].class;
                }

                // 이미지의 사이즈는 설정되있으니 없는 경우에 재등록 (에러 이미지에서 정상 이미지로 변경되었을 때)
                if (!$figureWrap.hasClass(sizeClass)) $figureWrap.addClass(sizeClass);

                $image.prop('isLoaded', true).show();

            })
            .on('error', e => {

                let errorImg = $image.prop('errUrl');

                if (errorImg === '') errorImg = ERROR_IMG;

                let fileName = FILE.getFileNameInfo(errorImg).name;
                title = title || fileName;

                // frameType이 horizon이나 vertical일 경우 wrapper size의 auto값이 없기 때문에 강제 설정
                if (frameType === 'horizon' || frameType === 'vertical'){

                    _setComponentSize($figureWrap, {size, width, height});

                    if (size && $figureWrap.hasClass(SIZE[size].class)){
                        $figureWrap.removeClass(SIZE[size].class);
                    }
                }

                // 에러 이미지 background로 설정
                $dimContainer.show().addClass('no-img');

                // 기본 이미지를 상대 경로로 설정했을 때 사이트의 상대경로로 보는 이슈때문에 분리
                if (errorImg !== ERROR_IMG) $dimContainer.css("background-image", `url(${errorImg})`);

                $figureWrap.attr('title', title);

                $dim.hide();
                $image.hide();
            })
            .attr('src', originImage);
    });
}

/**
 * 이미지 상태를 변경한다
 *
 * @param {object} $container - image container
 *
 * @private
 * @memberOf BxImg
 */
function _update($container = {}){

    let $image = $container.find('img');
    let $dim = $container.find('.dim-area');

    // 이미 로딩된 이미지는 진행하지 않는다
    if ($image.prop('isLoaded')) return this;

    // 이미지가 viewport 내부에 들어왔을 때 이미지 로딩을 시작한다
    if (UTIL.isInViewport($dim)){
        $image.trigger('appear');
    }
}

/**
 * element 컴포넌트 크기를 설정한다
 *
 * @param {object} $elem - element
 * @param {object} options - 사이즈 설정 오브젝트
 *
 * @private
 * @memberOf BxImg
 */

function _setComponentSize($elem = null, options = {}){

    let defaultValue = SIZE['medium'].value;
    let width = options.width === '' ? defaultValue : options.width;
    let height = options.height === '' ? defaultValue : options.height;

    if (options.size !== ''){
        width = height = defaultValue;
    }

    $elem.css({width, height});
}

/**
 * 랜덤 컬러값을 추출한다
 *
 * @param {string} val - 16진수 color 값
 *
 * @returns {string} 16진수 color 값
 *
 * @private
 * @memberOf BxImg
 */
// function _getRandomColor(val = ''){
//
//     if (val === ''){
//         val = Number((Math.random() * HEX).toFixed(0)).toString(16);
//     }
//     else if (val.length < 6){
//         val = _getRandomColor('0' + val);
//     }
//
//     return val;
// }

/**
 * 이미지 컴포넌트에 이벤트를 등록한다
 *
 * @param {object} $container - image container
 *
 * @private
 * @memberOf BxImg
 */
function _addDocumentElementEventListener($container = {}){

    // window resize event 등록
    _addResizeListenerAtWindow.call(this, $container);

    // window scroll event 등록
    _addScrollListenerAtWindow.call(this, $container);
}

/**
 * 윈도우 resize 이벤트를 등록한다
 *
 * @param $container - image container
 *
 * @private
 * @memberOf BxImg
 */
function _addResizeListenerAtWindow($container = {}){

    let $window = $(window);

    $window.on('resize', e => {
        _update.call(this, $container);
    });
}

/**
 * 윈도우 scroll 이벤트를 등록한다
 *
 * @param $container - image container
 *
 * @private
 * @memberOf BxImg
 */
function _addScrollListenerAtWindow($container = {}){

    let $window = $(window);

    $window.on('scroll', e => {

        let $image = $container.find('img');
        let timeoutObj = $image.prop('timeoutObj');

        if (!_.isUndefined(timeoutObj)) clearTimeout(timeoutObj);

        timeoutObj = setTimeout(()=>{
            _update.call(this, $container);
        }, UPDATE_DURATION);

        $image.prop('timeoutObj', timeoutObj);
    });
}

/**
 * 컴포넌트 내부 텍스트를 변경한다
 *
 * @param {object} elem - image element
 * @param {string} k - property key
 * @param {string} v - property value
 *
 * @private
 * @memberOf BxImg
 */
function _changeText(elem = null, k = '', v = ''){

    let $container = $(elem).parents(`.${CLASS_NAME.wrap}`);

    switch (k){

        case 'desc': {
            $container.find(`.${CLASS_NAME.desc} p`).html(v);
            break;
        }
        case 'caption': {
            $container.find('figcaption p').text(v);
            break;
        }
        case 'btnText': {
            $container.find('button').text(v);
            break;
        }
        case 'data-title': {
            $container.find('figure').attr('title', v);
            $container.find('img').attr('title', v);
            break;
        }
    }
}

/**
 * 캡션 및 버튼의 정렬을 변경한다
 *
 * @param {object} elem - image element
 * @param {string} v - property value
 *
 * @private
 * @memberOf BxImg
 */
function _changeTextAlign(elem = null, v = ''){

    let $container = $(elem).parent();

    // overlayStyle이 half type인 경우에만 align 적용
    if ($container.hasClass(OVERLAY_STYLE.half)){

        if ($container.hasClass(ALIGN.left)){
            $container.removeClass(ALIGN.left);
        }
        else if ($container.hasClass(ALIGN.right)){
            $container.removeClass(ALIGN.right);
        }

        $container.addClass(ALIGN[v]);
    }
}

/**
 * 컴포넌트 사이즈를 변경한다
 *
 * @param {object} elem - image element
 * @param {string} k - property key
 * @param {object} v - property value
 *
 * @private
 * @memberOf BxImg
 */
function _changeComponentSize(elem = null, k = '', v = null){

    let $container = $(elem).parent();

    k = k.split('-')[1];

    $container.css(k, `${v}px`);
}

/**
 * lazyload 전에 실행 될 dim을 그린다
 *
 * @param {object} elem - image element
 *
 * @private
 * @memberOf BxImg
 */
function _showDim($elem = null, $image = {}){

    let $dim = $elem.bxDim(DIM_STYLE);

    $dim.show();

    $image.prop('$dim', $dim);
}

module.exports = BxImg;
