// bxuip-angular.js-ui/src/lib/bx-ui/app/components/layout.js
/**
 * 렌더링 완료 시, 호출될 callback 함수
 * @callbcak onRender
 *
 * @memberOf BxLayout
 */
/**
 * panel 생성 완료 시, 호출될 callback 함수
 * @callback onTransaction
 * @param {object} $elem - 생성된 엘리먼트
 *
 * @memberOf BxLayout
 */
/**
 * @class BxLayout
 * @classDesc BxLayout 클래스 생성자
 *
 * @param {object} OPTS
 * @param {object} [OPTS.panel = {}]
 * @param {string} OPTS.panel.type - panel 타입
 * @param {string} OPTS.panel.id - panel 아이디
 * @param {string} OPTS.panel.location - 초기 위치 설정를 설정한다(top, left, bottom, right)
 * @param {string} OPTS.panel.content - panel 내용
 * @param {object} OPTS.panel.attr - panel 속성
 * @param {object[]} OPTS.panel.children -  panel 자식
 * @param {onRender} [OPTS.onRender = function(){}] - 렌더링 완료 시, 호출될 callback 함수
 * @param {onTransaction} [OPTS.onTransaction = function(){}] - panel 생성 완료 시, 호출될 callback 함수
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 * var layout = $('#layout').BxLayout({
    panel: {
        type: 'row',
        id: 'wrap',
        attr: {
            style: 'background: gray; min-width: 940px;',
            className: 'testClassName'
        },
        children: [
        {
            type: 'row',
            id: 'header',
            content: 'header',
            attr: {
                style: 'background: black; color:white'
            }
        },
        {
            type: 'row',
            id: 'container',
            children: [
                {
                    type: 'row',
                    id: 'rowMain1',
                    content: 'rowMain1',
                    attr: {
                        style: 'background:yellow;'
                    }
                },
                {
                    type: 'col',
                    id: 'side',
                    ratio: 22.34,
                    content: '',
                    attr: {
                        style: 'background:red'
                    },
                    children: [
                        {
                            type: 'row',
                            id: 'row1',
                            ratio: 20,
                            content: '<div style="color:white">row1<br><br><br><br>' +
                            '<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>' +
                            '<br><br><br><br><br><br><br><br><script type="text/javascript">console.log($("#side"));<\/script></div>',
                            attr: {
                                style: 'background:red;'
                            }
                        },
                        {
                            type: 'row',
                            id: 'row2',
                            content: 'row2',
                            attr: {
                                style: 'background:green'
                            }
                        }
                    ]

                },
                {
                    type: 'col',
                    id: 'main',
                    content: 'main'
                },
                {
                    type: 'col',
                    id: 'right',
                    content: 'right'
                }
            ]
        },
        {
            type: 'fixed',
            id: 'fixedPanel1',
            location: {
                bottom: 0
            },
            attr: {
            },
            children: [
                {
                    type: 'row',
                    id: 'footerRowMain',
                    content: 'footerRowMain',
                    attr: {
                        style: 'background:red;'
                    }
                }
            ]
        },
        {
            type: 'fixed',
            id: 'quickMenu',
            location: {
                top: 0,
                right: 0
            },
            attr: {
                style: 'background:green; width: 150px; height: 300px;'
            }
        }
    ]},
    onTransaction: function($elem){
//                console.log($elem);
    },
    onRender: function(){
        console.log('onRendered!!');
        this.$container.bxDim().show().hide();
    }
}).render();
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;

const OPTS = Symbol();

const FIRSTLOADPANELID = Symbol();

// 로드된 panel 객체를 저장한다.
const LOADEDPANELS = {};

class BxLayout {

    constructor({
        panel = {},
        onRender = function (){},
        onTransaction = function (){}
    } = {}){

        if (!_.isPlainObject(panel)){
            CONSOLE.prop.throw(BxLayout, 'panel', 'type');
        }

        if (_.isEmpty(panel)){
            CONSOLE.prop.throw(BxLayout, 'panel', 'value');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxLayout, 'onRender', 'type');
        }

        if (!_.isFunction(onTransaction)){
            CONSOLE.prop.throw(BxLayout, 'onTransaction', 'type');
        }

        this[OPTS] = {panel, onRender, onTransaction};

        this.$root = BxLayout.$root;

        // 최초 로드된 panel 아이디
        this[FIRSTLOADPANELID] = '';
    }

    /**
     * 랜더 함수
     *
     * @returns {Object} - 생성된 인스턴스
     */
    render(){

        // Render 시점에서 Container 생성 후 Root에 Append
        this.$container = BASE.appendContainerElement(this.$root, BxLayout.className);

        // _render 함수에 위임한다.
        return _render.call(this, this[OPTS].panel);
    }

    /**
     * panel 반환 함수
     *
     * @param id - panel 아이디
     * @returns {object} panel 객체
     */
    get(id = ''){

        if (!_.isString(id)){
            console.error('panel id 값이 존재하지 않습니다.');
            return this;
        }

        let stack = [this[OPTS].panel];

        let panel = null;
        let ret = null;

        // 동기 순회
        while (panel = stack.pop()){

            if (panel.id === id){
                ret = panel;
                break;
            }

            let length = panel.children ? panel.children.length : 0;

            while (length--){
                stack.push(panel.children[length]);
            }
        }

        return ret;
    }

    /**
     * panel 설정 함수
     * @param {string | object} arg - panel 객체 및 panel 아이디
     * @param {object} opt
     * @param opt.attr {object} - panel 속성
     * @param opt.content {string} - content
     * @returns {object} 생성된 인스턴스
     */
    set(arg = null, {attr = {}, content = ''} = {}){

        if (!_.isObject(arg) && !_.isString(arg)){
            console.error('panel 또는 id 값이 존재하지 않습니다.');
            return this;
        }

        let id = _.isObject(arg) ? arg.id : arg;

        let $container = this.$container;

        let $elem = $container.find('#' + id);

        if (!_.isElement($elem.get(0))){
            console.error('$elem 값이 존재하지 않습니다.');
            return this;
        }

        // 엘리먼트에 전달받은 attr 객체 속성을 추가한다.
        UTIL.attr($elem, attr);

        if (content) $elem.html(content);

        return this;
    }
}

/**
 * 내부 랜더 함수
 *
 * @param {string} type - 패널 타입
 * @param {string} id - 패널 아이디
 * @param {string} ratio - 패널 비율
 * @param {object} attr - 속성
 * @param {object[]} children - 자식 객체 배열
 * @param {string} content - HTML 및 텍스트 내용
 * @param {object} location - 패널 위치 설정
 * @param $root
 * @returns {object} 생성된 인스턴스
 * @private
 * @memberOf BxLayout
 */
function _render({
    type = '',
    id = '',
    ratio = '',
    attr = {},
    children = [],
    content = '',
    location = {}
    } = {}, $root = null){

    // 객체 해체 할당
    let panel = {type, id, ratio, attr, children, content, location};

    // 전달된 container element 가 없는 경우, 문서 조각을 할당한다.
    $root = $root || $(document.createDocumentFragment());

    // 로드된 items 에 현재 item.id 가 존재하는 경우(id 값은 unique 해야만 한다)
    if (LOADEDPANELS[panel.id]){
        throw new Error(panel.id);
    }
    // 최초 전달된 item.id 를 기억해둔다(재귀 호출의 마지막 시점을 알기위해)
    if (!this[FIRSTLOADPANELID]) this[FIRSTLOADPANELID] = panel.id;

    LOADEDPANELS[panel.id] = panel;

    let panelType = {
        'row': 'row',
        'col': 'col',
        'fixed': 'fixed'
    }[panel.type];

    // panel ratio 값을 초기화한다.
    _initRatio(children);

    let $elem = null;

    if (panelType === 'row'){

        $elem = $('<div id="' + panel.id + '" style="color:red" class="' + panelType + '">' + panel.content + '</div>');
    }
    else if (panelType === 'col'){

        $elem = $('<div id="' + panel.id + '" class="' + panelType + '">' + panel.content + '</div>');
        panel.ratio && $elem.attr('style', 'width:' + panel.ratio + '%');
    }
    else if (panelType === 'fixed'){

        $elem = $('<div id="' + panel.id + '" class="' + panelType + '">' + panel.content + '</div>');

        panel.ratio && $elem.attr('style', 'width:' + panel.ratio + '%');

        // fixed panel 의 위치를 지정한다.
        $elem.css(panel.location);
    }

    // 엘리먼트에 전달받은 attr 객체 속성을 추가한다.
    UTIL.attr($elem, panel.attr);

    // 해당 $root 엘리먼트에 현재 생성된 현재 $elem 를 추가시킨다.
    $root.append($elem);

    this[OPTS].onTransaction.call(this, $elem);

    // 자식 panel 이 존재하는 경우
    _.map(panel.children, (childpanel) =>{
        _render.call(this, childpanel, $elem);
    });

    // 재귀 호출의 마지막
    if (this[FIRSTLOADPANELID] === panel.id){

        let $container = this.$container;
        // (call)stack 자료 구조가 가지는 LIFO 에 의해, 가장 먼저 전달된 panel 객체가 가장 늣게 pop() 되는 구조를 가진다.
        $container.append($elem);
        // call render callback function
        this[OPTS].onRender.call(this);
    }

    return this;
}
/**
 * 비율 조정 함수
 *
 * @param {object[]} children - 객체 배열
 * @private
 * @memberOf BxLayout
 */
function _initRatio(children = []){

    // col type 을 가진 panel 집합
    let colTypepanels = _.filter(children, {'type': 'col'});
    let colTypepanelSize = _.size(colTypepanels);

    // 자식 panel 중 colTypepanel 이 존재하는 경우
    if (colTypepanelSize){

        let ratioSum = 0;
        // 필더된 col type panel 중, ratio 값이 존재하는 panel 집합
        let haveRatiopanels = _.filter(colTypepanels, (panel) =>{

            let ret = false;

            if (_.isNumber(panel.ratio)){
                ratioSum += panel.ratio;
                ret = true;
            }

            return ret;
        });

        // ratio property 를 가진 panel size
        let haveRatiopanelSize = _.size(haveRatiopanels);

        let noRatiopanelSize = colTypepanelSize - haveRatiopanelSize;
        // 각 panel 당 초기화될, ratio width size
        let perpanelRatioWidth = (100 - ratioSum) / noRatiopanelSize;

        _.map(colTypepanels, (panel) =>{
            if (!_.isNumber(panel.ratio)) panel.ratio = perpanelRatioWidth;
        });
    }
}

module.exports = BxLayout;
