// bxuip-angular.js-ui/src/lib/bx-ui/app/components/lnb.js
/**
 * 닫기 버튼 클릭 시, 호출될 callback 함수
 *
 * @callback onClose
 *
 * @param {object} $elem - 선택된 엘리먼트 객체
 * @memberOf BxLnb
 */

/**
 * 메뉴 선택 시, 호출될 callback 함수
 *
 * @callback onClick
 *
 * @param {object} $elem - 선택된 엘리먼트 객체
 * @memberOf BxLnb
 */

/**
 * 랜더링 완료 시, 호출될 callback 함수
 *
 * @callback onRender
 *
 * @memberOf BxLnb
 */

/**
 * @class BxLnb
 * @classDesc BxLnb 클래스 생성자
 *
 *
 * @param {object} OPTS
 * @param {object[]} [OPTS.items = []] - 메뉴 아이템
 * @param {object} OPTS.items.attr - 엘리먼트 속성
 * @param {string} OPTS.items.title - 메뉴 타이틀
 * @param {int} [OPTS.width = 150] - 메뉴 사이즈
 * @param {boolean} [OPTS.isCloseButton=false] - <s>닫기 버튼 사용 유무</s><span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @param {boolean} [OPTS.enableCloseButton=false] - 닫기 버튼 사용 유무
 * @param {boolean} [OPTS.isDynamicNavBar=false] - <s>동적 네비게이션 사용 유무</s><span class='important'> ( Added in: v3.0.0, Deprecated since: v3.0.3 )</span>
 * @param {boolean} [OPTS.enableDynamicNavBar=false] - 동적 네비게이션 사용 유무
 * @param {function} [OPTS.onClose=function(){}] - 닫기 버튼 클릭 시, 호출될 callback 함수
 * @param {function} [OPTS.onClick=function(){}] - 메뉴 선택 시, 호출될 callback 함수
 * @param {function} [OPTS.onRender=function(){}] - 랜더링 완료 시, 호출될 callback 함수
 * @returns {object} 생성된 인스턴스
 *
 * @example
 $('#BxLnb').BxLnb({
    items:[
        {attr: {id: '1', name: '1' }, title: '온라인 로그조회'},
        {attr: {id: '2'}, title: '2'},
        {attr: {id: '3'}, title: '3'}
    ],
    width: 70,
    enableCloseButton: true,
    enableDynamicNavBar: true,
    onClose : function($li){

    },
    onClick : function($li){

    },
    onRender : function(){
        this.$root.bxDim().show();
    }
}).render();
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

const ANIMATE_CONFIG = {
    duration: 300,
    marginRight: 50
};

const SELECTOR_NAME = {
    wrap: 'infinite-tab-overflow-container',
    ul: 'infinite-tab-list',
    nav: 'infinite-tab-nav',
    leftBtn: 'infinite-tab-nav-btn-left',
    rightBtn: 'infinite-tab-nav-btn-right',
    id:'bx-lnb-'
};

class BxLnb {

    constructor({
        items = [],
        width = 150,
        enableCloseButton = false,
        enableDynamicNavBar = false,
        onClose = function (){
        },
        onClick = function (){
        },
        onRender = function (){
        }
    } = {}){

        if (!_.isArray(items)){
            CONSOLE.prop.throw(BxLnb, 'items', 'type');
        }

        if (!_.isNumber(width)){
            CONSOLE.prop.throw(BxLnb, 'width', 'type');
        }

        if (!_.isBoolean(enableCloseButton)){
            CONSOLE.prop.throw(BxLnb, 'enableCloseButton', 'type');
        }

        if (!_.isBoolean(enableDynamicNavBar)){
            CONSOLE.prop.throw(BxLnb, 'enableDynamicNavBar', 'type');
        }

        if (!_.isFunction(onClose)){
            CONSOLE.prop.throw(BxLnb, 'onClose', 'type');
        }

        if (!_.isFunction(onClick)){
            CONSOLE.prop.throw(BxLnb, 'onClick', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxLnb, 'onRender', 'type');
        }

        this.$root = BxLnb.$root;

        if (!_.isArray(items)){
            throw new Error('items 값이 없습니다.');
        }

        // 옵션
        this[OPTS] = {items, width, enableCloseButton, enableDynamicNavBar, onClose, onClick, onRender};
    }

    /**
     * 렌더 함수
     * @returns {object} 생성된 인스턴스
     */
    render(){

        // Render 시점에서 Container 생성 후 Root에 Append
        this.$container = BASE.appendContainerElement(this.$root, BxLnb.className);

        // $container 엘리먼트
        let $container = this.$container;

        let html = [];

        html.push('<div class="infinite-tab-container">');
        html.push('<section class="infinite-tab-overflow-container">');

        html.push('<ul class="infinite-tab-list"></ul>');

        html.push('</section>');

        html.push('<div class="infinite-tab-nav">');
        html.push('<button type="button" class="bw-btn infinite-tab-nav-btn-left"><i class="bw-bx-icon i-15 i-ctrl-pre"></i></button>');
        html.push('<button type="button" class="bw-btn infinite-tab-nav-btn-right"><i class="bw-bx-icon i-15 i-ctrl-next"></i></button>');
        html.push('</div>');

        html.push('</div>');

        $container.append(html.join(''));

        let $ul = $container.find(`.${SELECTOR_NAME.ul}`);
        let $wrap = $container.find(`.${SELECTOR_NAME.wrap}`);
        let length = this[OPTS].items.length;

        for (let i = 0; i < length; i++){

            let item = this[OPTS].items[i];

            let $li = _createLiElem.call(this, item, i);

            // 0 번째 $li 엘리먼트를 활성화 시킨다.
            if (i === 0) $li.addClass('on');

            $ul.append($li);
        }

        // $nav 엘리먼트 및 이벤트 초기화
        _initNavigation.call(this);

        $(window).resize(_controlScrollWhenResize.bind(this));

        $container.on('focus', 'a:eq(0)', (e) =>{
            $wrap.get(0).scrollLeft = 0;
        });

        $container.on('focus', 'a, button', (e) =>{

            e.preventDefault();

            if ($(e.target).parent().index() === 0) return;

            let wrapWidth = $wrap.width();
            let ulWidth = this[OPTS].width * $ul.children().length;

            let left = wrapWidth - ulWidth;

            if (left >= $ul.position().left && wrapWidth <= ulWidth){
                $wrap.get(0).scrollLeft = ulWidth - wrapWidth;
            }

        });

        // render
        this[OPTS].onRender.call(this);

        return this;
    }

    /**
     * 아이템 삭제 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {function} [callback=function(){}] - 메뉴 삭제 후, 호출될 callback 함수
     * @returns {object} 생성된 인스턴스
     */
    remove($elem = null, callback = function (){}){

        let $ul = this.$container.find(`.${SELECTOR_NAME.ul}`);

        if (_.isNull($elem)){
            CONSOLE.prop.error(BxLnb, '$elem', 'type');
            return this;
        }

        if (!$ul.find($elem).get(0)){
            CONSOLE.prop.error(BxLnb, '$elem', 'value');
            return this;
        }

        $ul.find($elem).remove();

        if (this[OPTS].enableDynamicNavBar) _controlNav.call(this);

        _controlScrollWhenResize.call(this);

        let item = {
            id: $elem.attr('id'),
            title: $elem.text()
        };

        callback.call(this, $elem);

        return item;
    }

    /**
     * 마지막 위치에 아이템을 추가하는 함수
     *
     * @param {object} item - 아이템 객체
     * @param {object} item.attr - 아이템 객체의 속성
     * @param {string} item.title - 화면에 보여지는 라벨 값
     * @param {object} item.data - 데이터
     * @param {function} [callback=function(){}] - 메뉴 추가 후, 호출될 callback 함수
     *
     * @returns {object} 생성된 아이템
     */
    append({attr = {}, title = ''} = {}, callback = function (){}){

        let $ul = this.$container.find(`.${SELECTOR_NAME.ul}`);

        if (!_.isPlainObject(attr) || _.isUndefined(attr.id) || attr.id === ''){
            // 아이디가 없을 경우 자동 지정
            attr.id = SELECTOR_NAME.id + _.uniqueId();
        }

        // 전달된 id 값을 가진 메뉴가 존재하는 경우
        if (_.isElement($('li[id="' + attr.id + '"]').get(0))){
            console.warn('아이디값을 가진, 아이템이 이미 존재합니다.');
            return this;
        }

        let item = {attr, title};

        let $li = _createLiElem.call(this, item);

        $ul.append($li);

        if (this[OPTS].enableDynamicNavBar) _controlNav.call(this);

        callback.call(this, $li, item);

        return item;
    }

    /**
     * 맨 앞에 아이템을 추가하는 함수
     *
     * @param {object} item - 아이템 객체
     * @param {object} item.attr - 아이템 객체의 속성
     * @param {string} item.title - 화면에 보여지는 라벨 값
     * @param {object} item.data - 데이터
     * @param {function} [callback=function(){}] - 메뉴 추가 후, 호출될 callback 함수
     *
     * @returns {object} 생성된 아이템
     */
    prepend({attr = {}, title = ''} = {}, callback = function (){}){

        let $ul = this.$container.find(`.${SELECTOR_NAME.ul}`);

        if (!_.isPlainObject(attr) || _.isUndefined(attr.id) || attr.id === ''){
            // 아이디가 없을 경우 자동 지정
            attr.id = SELECTOR_NAME.id + _.uniqueId();
        }
        // 전달된 id 값을 가진 메뉴가 존재하는 경우
        if (_.isElement($('li[id="' + attr.id + '"]').get(0))){
            console.warn('아이디값을 가진, 아이템이 이미 존재합니다.');
            return this;
        }

        let item = {attr, title};

        let $li = _createLiElem.call(this, item);

        $ul.prepend($li);

        if (this[OPTS].enableDynamicNavBar) _controlNav.call(this);

        callback.call(this, $li, item);

        return item;
    }

    /**
     * 포커스 이동 함수
     *
     * @param {string} id - 아이디
     * @returns {object} 생성된 인스턴스
     */
    focus(id = ''){

        if (!_.isString(id)){
            CONSOLE.prop.error(BxLnb, 'id', 'type');
            return this;
        }

        let $container = this.$container;
        let $li = $(`.${SELECTOR_NAME.ul}>li[id="${id}"]`);

        if (!$li.get(0)){
            CONSOLE.prop.error(BxLnb, 'id', 'value');
            return this;
        }

        let $wrap = $container.find(`.${SELECTOR_NAME.wrap}`);
        let $ul = $container.find(`.${SELECTOR_NAME.ul}`);
        let liW = this[OPTS].width;
        let left = ($li.index() * liW) - $wrap.width() + liW;

        $wrap.animate({scrollLeft: left}, ANIMATE_CONFIG.duration, () =>{
            $li.find('a').focus();
        });

        let $prevElem = $ul.find('li.on');

        $prevElem.removeClass('on');
        $li.addClass('on');

        return this;
    }

    /**
     *
     * @deprecated Deprecated in: v3.4.0, Deprecated since: v3.10.0
     * lnb 재 로딩 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    resize(){
        CONSOLE.prop.throw(BxLnb, 'resize', 'deprecatedMethod');
    }
}

/**
 * 아이템을 엘리먼트로 반환하는 함수
 *
 * @param {object} attr - attribute
 * @param {string} title - 제목
 * @returns {element} li - li 엘리먼트
 * @private
 * @memberOf BxLnb
 */
function _createLiElem({attr = {}, title = ''} = {}){

    // close 버튼 노출 여부
    let enableCloseButton = this[OPTS].enableCloseButton;

    let html = [];

    // $li 엘리먼트
    let $li = $('<li style="width:' + this[OPTS].width + 'px"></li>');

    html.push('<a href="#" onclick="return false;">' + title + '</a>');

    // close 버튼이 노출될 경우
    if (enableCloseButton){

        html.push('<button type="button" class="bw-btn">');
        html.push('<i class="bw-bx-icon i-15 i-ctrl-del"></i>');
        html.push('</button>');
    }

    $li.append(html.join(''));

    // 엘리먼트에 전달받은 attr 객체 속성을 추가한다.
    UTIL.attr($li, attr);

    if (enableCloseButton){

        // close 버튼 이벤트 바인딩
        $li.children('button').on('click', (e) =>{

            this.remove($li);

            this[OPTS].onClose.call(this, $li);
        });
    }

    // $li 엘리먼트 이벤트 바인딩
    $li.on('click', e => {

        let $prevElem = this.$container.find('.infinite-tab-container li.on');

        $prevElem.removeClass('on');

        let $elem = $(e.target);

        $elem.addClass('on');

        this[OPTS].onClick.call(this, $li);
    });

    return $li;
}

/**
 * 네비게이션 초기화 함수
 *
 * @private
 * @memberOf BxLnb
 */
function _initNavigation(){

    let $container = this.$container;
    let $nav = $container.find(`.${SELECTOR_NAME.nav}`);
    let $leftBtn = $nav.find(`.${SELECTOR_NAME.leftBtn}`);
    let $rightBtn = $nav.find(`.${SELECTOR_NAME.rightBtn}`);

    // $nav 엘리먼트가 동적으로 노출되는 경우
    if (this[OPTS].enableDynamicNavBar){
        _controlNav.call(this);
        $(window).resize(_controlNav.bind(this));
    }
    else{
        $container.find(`.${SELECTOR_NAME.wrap}`).css('margin-right', ANIMATE_CONFIG.marginRight);
        $nav.css('display', 'block');
    }

    $leftBtn.click((e) =>{
        _left.call(this);
    });

    $rightBtn.click((e) =>{
        _right.call(this);
    });
}

/**
 * 동적으로 nav 버튼을 show/hide 하는 함수
 *
 * @private
 * @memberOf BxLnb
 */
function _controlNav(){

    let $container = this.$container;

    let $wrap = $container.find(`.${SELECTOR_NAME.wrap}`);
    let $ul = $container.find(`.${SELECTOR_NAME.ul}`);
    let $nav = $container.find(`.${SELECTOR_NAME.nav}`);

    let wrapW = $wrap.width();
    let ulW = $ul.children().length * this[OPTS].width;

    if (wrapW < ulW){
        $nav.css('display', 'block');
        $wrap.css('margin-right', ANIMATE_CONFIG.marginRight);
    }
    else{
        $nav.css('display', 'none');
        $wrap.css('margin-right', 0);
    }
}

/**
 * wrap의 사이즈 변경에 따른 스크롤 조정 함수
 *
 * @private
 * @memberOf BxLnb
 */
function _controlScrollWhenResize(){

    let $container = this.$container;

    let $wrap = $container.find(`.${SELECTOR_NAME.wrap}`);
    let $ul = $container.find(`.${SELECTOR_NAME.ul}`);

    let wrapW = $wrap.width();
    let ulW = $ul.children().length * this[OPTS].width;
    let scrollW = ulW - wrapW;

    if (scrollW < $wrap.get(0).scrollLeft){
        $wrap.get(0).scrollLeft = scrollW;
    }

}

/**
 * 왼쪽 이동 함수
 *
 * @private
 * @memberOf BxLnb
 */
function _left(){

    let $container = this.$container;

    let $wrap = $container.find(`.${SELECTOR_NAME.wrap}`);
    let $ul = $container.find(`.${SELECTOR_NAME.ul}`);

    // W : width, X : left point, H : hidden, I : index
    let ulX = $ul.position().left;
    let ulHW = ulX * -1;
    let liW = this[OPTS].width;

    let firstIW = Math.ceil(ulHW / liW) * liW;

    if ($ul.position().left >= 0) return;

    let left = (liW - firstIW - ulHW * -1 || liW) + ulX;

    $wrap.stop(true, true)
        .animate({scrollLeft: left * -1}, ANIMATE_CONFIG.duration);

}
/**
 * 오른쪽 이동 함수
 *
 * @private
 * @memberOf BxLnb
 */
function _right(){

    let $container = this.$container;

    let $wrap = $container.find(`.${SELECTOR_NAME.wrap}`);
    let $ul = $container.find(`.${SELECTOR_NAME.ul}`);

    // W : width, X : left point, H : hidden, I : index
    let wrapW = $wrap.width();
    let ulX = $ul.position().left;
    let ulHW = ulX * -1;
    let liW = this[OPTS].width;

    let lastX = wrapW + ulHW;
    let lastIW = Math.ceil(lastX / liW) * liW;

    if (liW * $ul.children().length <= lastX) return;

    let left = (lastIW - lastX || liW) * -1 + ulX;

    $wrap.stop(true, true)
        .animate({scrollLeft: left * -1}, ANIMATE_CONFIG.duration);

}

module.exports = BxLnb;
