// bxuip-angular.js-ui/src/lib/bx-ui/app/components/paper.js
/**
 * Created by parkbeomsoo on 2016. 9. 2..
 */

/**
 * 선택 이벤트 callback 함수
 *
 * @callback onActivate
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf BxPaper
 * */

/**
 * 드래그 이벤트 callback 함수
 *
 * @callback onDrag
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf BxPaper
 * */

/**
 * 드래그 시작 이벤트 callback 함수
 *
 * @callback onDragStart
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf BxPaper
 * */

/**
 * 드래그 종료 이벤트 callback 함수
 *
 * @callback onDragStop
 * @param {object} e - 이벤트 객체
 * @param {object} $elem -  엘리먼트 객체
 * @memberOf BxPaper
 * */

/**
 * 닫기 이벤트 callback 함수
 *
 * @callback onClosed
 * @memberOf BxPaper
 * */

/**
 * 렌더 종료 이벤트 callback 함수
 *
 * @callback onRender
 * @memberOf BxPaper
 * */

/**
 * @class BxPaper
 * @classDesc BxPaper 클래스 생성자
 * @param {object} OPTS
 * @param {int} [OPTS.width=400] BxPaper content 영역 넓이
 * @param {string|int} [OPTS.height='auto'] BxPaper content 영역 높이
 * @param {object} [OPTS.position={x:0, y:0}] 초기 렌더링 시, 화면 정 가운데를 기준으로 위치를 이동시킬 x, y 좌표
 * @param {string} [OPTS.color='white'] BxPaper 의 색상 <br/> <span class='dec'>['white', 'yellow', 'theme']</span>
 * @param {string} [OPTS.type='flat'] BxPaper 의 타입 <br/> <span class='dec'>['flat', 'round', 'line', 'stitch']</span>
 * @param {string} [OPTS.shadow='none'] BxPaper 의 그림자 <br/> <span class='dec'>['none', 'left', 'right', 'both', 'bottom']</span>
 * @param {string} [OPTS.title=''] BxPaper 의 타이틀
 * @param {string} [OPTS.titleAlign='left'] BxPaper 의 타이틀 정렬 <br/> <span class='dec'>['left', 'center']</span>
 * @param {string} [OPTS.template=''] BxPaper 의 내용
 * @param {string} [OPTS.templateUrl=''] template 로 사용할 파일의 절대 경로 <br/><span class='dec'> template 와 동시에 쓰이는 경우, templateUrl 의 우선순위가 더 높게 적용. </span>
 * @param {object} [OPTS.dimStyle={}] paper 사용 시, dim 에 대한 설정
 * @param {boolean} [OPTS.enableDrag=true] drag 사용 유무
 * @param {function} [OPTS.onActivate=function(e, $elem){}] 선택 이벤트 callback 함수
 * @param {function} [OPTS.onDrag=function(e, $elem){}] 드래그 이벤트 callback 함수
 * @param {function} [OPTS.onDragStart=function(e, $elem){}] 드래그 시작 이벤트 callback 함수
 * @param {function} [OPTS.onDragStop=function(e, $elem){}] 드래그 종료 이벤트 callback 함수
 * @param {function} [OPTS.onClosed=function(){}] 닫기 이벤트 callback 함수
 * @param {function} [OPTS.onRender=function(){}] 렌더 종료 이벤트 callback 함수
 * @example

 // 1. template 예제

 var template = [];

 template.push('<div>template 예제 입니다.</div>');
 template.push('<div> template 에 대한 내용을 입력해 주세요.</div>');
 template = template.join('');

 $().BxPaper(
 {
     width:500,
     height: 500,
     position: {
        x:0,
        y:0
     },
     color: 'theme',
     type: 'stitch',
     shadow: 'both',
     title:'template 예제 #1',
     titleAlign: 'center',
     template: template,
     dimStyle: {
        color: 'theme' // ref bxDim Document
     },
     enableDrag: true,
     onActivate: function(e, $elem) {
         console.log('onActivate', e, $elem);
     },
     onDrag: function (e, $elem){
         console.log('onDrag', e, $elem);
     },
     onDragStart: function (e, $elem){
         console.log('onDragStart', e, $elem);
     },
     onDragStop: function(e, $elem){
         console.log('onDragStop', e, $elem);
     },
     onClosed: function () {
         console.log('BxPaper - onClose');
     },
     onRender: function () {
         console.log('BxPaper - onRender')
     }
 }
 ).render();

 ----------------------------------------------------------------------

 // 2. templateUrl 예제
 $().BxPaper(
 {
     width: 500,
     height: 500,
     position: {
        x:0,
        y:0
     },
     color: 'theme',
     type: 'stitch',
     shadow: 'both',
     title:'templateUrl 예제 #1',
     titleAlign: 'center',
     templateUrl: './app/views/onlineAdmin/log/test.tpl.html',
     dimStyle: {
        color: 'theme' // ref bxDim Document
     },
     enableDrag: true,
     onActivate: function(e, $elem) {
         console.log('onActivate', e, $elem);
     },
     onDrag: function (e, $elem){
         console.log('onDrag', e, $elem);
     },
     onDragStart: function (e, $elem){
         console.log('onDragStart', e, $elem);
     },
     onDragStop: function(e, $elem){
         console.log('onDragStop', e, $elem);
     },
     onClosed: function () {
         console.log('BxPaper - onClose');
     },
     onRender: function () {
         console.log('BxPaper - onRender')
     }
 }
 ).render();


 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

const uiConfig = {
    className: {},
    styleInfo: {}
};
/*
 *   className setting
 * */
const PAPER_COLOR_CLASS_NAME = {
    white: '',
    yellow: 'paper-color1',
    theme: 'paper-color2'
};

const PAPER_TYPE_CLASS_NAME = {
    flat: '',
    round: 'paper-round',
    line: 'paper-line',
    stitch: 'paper-stitch'
};

const PAPER_SHADOW_CLASS_NAME = {
    none: '',
    left: 'paper-lift-left',
    right: 'paper-lift-right',
    both: 'paper-lift',
    bottom: 'paper-curve-btm'
};

const PAPER_STRUCTURE_CLASS_NAME = {
    wrap: 'bx-paper-wrap',
    focus: 'paper-focus',
    drag: 'paper-drag',
    header: 'paper-header',
    titleContent: 'bx-tt',
    content: 'paper-content',
    contentDesc: 'bx-desc',
    btnWrap: 'paper-btn-wrap',
    closeBtn: 'bx-btn-close',
    bwIcon: 'bw-icon',
    closeIconSize: 'i-25',
    closeIcon: 'i-close'
};
/*
 *   className assign
 */
_.assign(uiConfig.className, {
    color: PAPER_COLOR_CLASS_NAME,
    type: PAPER_TYPE_CLASS_NAME,
    shadow: PAPER_SHADOW_CLASS_NAME,
    structure: PAPER_STRUCTURE_CLASS_NAME
});

/*
 *   styleInfo setting
 */
const PAPER_HEADER_HEIGHT = {
    none: 0,
    flat: 60,
    round: 60,
    line: 61,
    stitch: 60
};

const PAPER_TITLE_ALIGN = {
    left: 'left',
    center: 'center'
};

const PAPER_CONTENT_PADDING_BOTTOM = 20;
/*
 *  styleInfo assign
 */
_.assign(uiConfig.styleInfo, {
    headerHeight: PAPER_HEADER_HEIGHT,
    titleAlign: PAPER_TITLE_ALIGN,
    contentPaddingBottom: PAPER_CONTENT_PADDING_BOTTOM
});
/*
 * 대체 택스트
 */
const ALT_TEXT = {
    close: '닫기'
};

class BxPaper {

    constructor({
            width = 300,                             // width
            height = 'auto',                         // height
            position = {},                           // 기본 center 기준 추가 x, y 값. position = { x = 1, y = 1}
            color = 'white',                         // white, yellow, theme
            type = 'flat',                           // flat, round, stitch, line
            shadow = 'none',                         // none, left, right, both, bottom
            title = '',                              // title 값의 설정유무
            titleAlign = 'left',                     // center / left
            template = '',                           // content 영역에 들어갈 template
            templateUrl = '',                        // ajax 를 이용해 가져올 template url
            dimStyle = {},                           // dimStyle 지정. default 시
            enableDrag = true,                       // enableDrag 설정 유무
            onActivate = function (e, $elem){
            },      // popup layer 선택에 대한 이벤트 callback --> dom 순서 조작해야됨.
            onDrag = function (e, $elem){
            },          // drag 되는 동안의 이벤트 callback
            onDragStart = function (e, $elem){
            },     // drag 시작 이벤트 callback
            onDragStop = function (e, $elem){
            },      // drag 종료 이벤트 callback
            onClosed = function (){
            },                // 닫기 이벤트 callback
            onRender = function (){
            }                 // render 완료시 불리는 callback
    } = {}){

        if (!_.isNumber(width)){
            CONSOLE.prop.throw(BxPaper, 'width', 'type');
        }

        if (width < 300){
            width = 300;
            console.warn('[BxPaper] width 값은 300 이상이어야 합니다. (http://bwg.co.kr/bxui/bx-ui/doc/BxPaper.html)');
        }

        if (!_.isNumber(height) && height !== 'auto'){
            CONSOLE.prop.throw(BxPaper, 'height', 'type');
        }

        if (_.isNumber(height) && height < 300){
            height = 300;
            console.warn('[BxPaper] height 값은 300 이상이어야 합니다. (http://bwg.co.kr/bxui/bx-ui/doc/BxPaper.html)');
        }

        if (_.isString(height) && height !== 'auto'){
            Console.prop.throw(BxPaper, 'height 의 문자는 auto만 입력 가능합니다.');
        }

        if (!_.isPlainObject(position)){
            CONSOLE.prop.throw(BxPaper, 'position', 'type');
        }

        if (!_.isString(color)){
            CONSOLE.prop.throw(BxPaper, 'color', 'type');
        }


        if (!_.isString(type)){
            CONSOLE.prop.throw(BxPaper, 'type', 'type');
        }


        if (!_.isString(shadow)){
            CONSOLE.prop.throw(BxPaper, 'shadow', 'type');
        }

        if (!_.isString(title)){
            CONSOLE.prop.throw(BxPaper, 'title', 'type');
        }
        else{
            titleAlign = _.isUndefined(uiConfig.styleInfo.titleAlign[titleAlign]) ? 'left' : titleAlign;
        }

        if (!_.isString(titleAlign)){
            CONSOLE.prop.throw(BxPaper, 'titleAlign', 'type');
        }

        if (!_.isString(template)){
            CONSOLE.prop.throw(BxPaper, 'template', 'type');
        }

        if (!_.isString(templateUrl)){
            CONSOLE.prop.throw(BxPaper, 'templateUrl', 'type');
        }

        if (!_.isEmpty(templateUrl)){

            template = _getTemplate(templateUrl);

            if (_.isNull(template)){
                CONSOLE.prop.throw(BxPaper, 'templateUrl', 'value');
            }
        }

        if (!_.isPlainObject(dimStyle)){
            CONSOLE.prop.throw(BxPaper, 'dimStyle', 'type');
        }

        if (!_.isBoolean(enableDrag)){
            CONSOLE.prop.throw(BxPaper, 'enableDrag', 'type');
        }

        if (!_.isFunction(onActivate)){
            CONSOLE.prop.throw(BxPaper, 'onActivate', 'type');
        }

        if (!_.isFunction(onDrag)){
            CONSOLE.prop.throw(BxPaper, 'onDrag', 'type');
        }

        if (!_.isFunction(onDragStart)){
            CONSOLE.prop.throw(BxPaper, 'onDragStart', 'type');
        }

        if (!_.isFunction(onDragStop)){
            CONSOLE.prop.throw(BxPaper, 'onDragStop', 'type');
        }

        if (!_.isFunction(onClosed)){
            CONSOLE.prop.throw(BxPaper, 'onClosed', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxPaper, 'onRender', 'type');
        }

        this[OPTS] = {
            width,
            height,
            position,
            color,
            type,
            shadow,
            title,
            titleAlign,
            template,
            templateUrl,
            dimStyle,
            enableDrag,
            onActivate,
            onDrag,
            onDragStart,
            onDragStop,
            onClosed,
            onRender
        };

        /*
         * $root: $root Element 표현. 최종 $root 는 body 를 지칭
         * $container: body 하위 -bx-ui-paper 클래스를 갖는 돔엘리먼트
         * $paperWrap: bind 함수 사용시 사용되며, 초기 bx-paper-wrap 이 포함되는 부모를 지칭
         * $dim: $dim Element
         * containerClassName: 생성된 컨테이너의 className
         */
        this.$root = null;
        this.$container = null;
        this.$paperWrap = null;

        this.$dim = null;
        this.containerClassName = null;
        /*
         * lastPosition: $paperWrap 의 마지막 위치
         * preWidth: 이전 browser width 값
         * resizer: window 에 적용할 resizer 함수
         * */
        this.lastPosition = {};
        this.preWidth = _getBrowserWidth();
        // this.resizer = _resizer.bind(this);

        return this;
    }

    /**
     * BxPaper 렌더 함수
     *
     * @param {object} [$compile = null] - template 에 Angular $scope 을 바인딩 시키기위해서는, $compile 서비스를 전달해야한다.
     * @param {object} [$scope = null] - template 에 Angular $scope 을 바인딩 시키기위해서는, $scope 을 전달해야한다.
     *
     * @returns {object} 생성된 인스턴스
     */
    render($compile = null, $scope = null){

        _initBxPaper.call(this);

        let $container = this.$container;
        let isDrag = this[OPTS].enableDrag;

        let html = _createDOMElement.call(this);

        $container.append(html);

        if (UTIL.isAngular() && $compile && $scope){
            BASE.ngCompile.apply(this, [$container, $compile, $scope]);
        }

        this.$paperWrap = $container.find(`.${uiConfig.className.structure.wrap}`);

        _setPaperWrapClass.call(this);
        _initContainerStructure.call(this);
        _setLastPosition.call(this);

        if (isDrag){
            _initDrag.call(this);
        }

        _addEvent.call(this);

        _onRender.call(this);

        return this;
    }

    /**
     * BxPaper 바인드 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    //bind(){
    //
    //    let _bind = () =>{
    //
    //        _initBxPaper.call(this, true);
    //
    //        let $paperWrap = this.$paperWrap;
    //        let opts = this[OPTS];
    //        let isDrag = opts.enableDrag;
    //        this.$container.append($paperWrap);
    //
    //        _initContainerStructure.call(this);
    //        _setLastPosition.call(this);
    //
    //        if (isDrag){
    //            _initDrag.call(this);
    //        }
    //
    //        _addEvent.call(this);
    //
    //        _onRender.call(this);
    //    };
    //
    //    window.setTimeout(() => _bind(), 1);
    //
    //    return this;
    //}

    /**
     * BxPaper 닫기 함수
     */
    close(){

        // let $root = this.$root;

        this.$container.remove();

        this.$dim.hide();
        this.$dim = null;

        // 아래 주석 부분은 multi paper 와 관련된 영역이므로 해당 부분에 대해 주석 처리함
        // let paperArr = $root.find(`.${this.containerClassName}`),
        //     paperLength = paperArr.length;
        //
        // if (paperLength === 0){
        //
        //     this.$dim.hide();
        //     this.$dim = null;
        //
        //     /*
        //      * safari 의 경우, DOM 에서 제거 되는 경우에 대해서는 blur 이벤트가 불리지 않아
        //      * 다시 setting 을 해줘야 한다.
        //      * */
        //     UTIL.setPreFocusInstanceKey('base');
        //
        //     let $target = UTIL.getPreFocusInstance();
        //
        //     if ($target){
        //         window.setTimeout(() =>{
        //             $target.focus();
        //         }, 0);
        //     }
        // }
        // else{
        //     // 창이 닫히고 포커싱이 돌아오는데 시간이 소요되 비동기로 처리한다.
        //     setTimeout(function (e){
        //         paperArr.last().find(`.${PAPER_STRUCTURE_CLASS_NAME.focus}`).focus();
        //     }, 0);
        // }

        $(window).off('resize', this.resizer);

        this[OPTS].onClosed.call(this);
    }
}

/**
 * templateUrl 반환 함수.
 *
 * @returns {string}
 * @private
 * @memberOf BxPaper
 */
function _getTemplate(url){

    let template = null;

    $.ajax({
        url: url,
        success: (req) =>{
            template = req;
        },
        contentType: 'text/plain',
        async: false
    });

    return template;
}

/**
 * BxPaper 인스턴스의 전역 속성 세팅 함수
 *
 * @private
 *
 * @memberOf BxPaper
 */
function _initBxPaper(){

    //if (isBind){

        // native 함수인 cloneNode, importNode 는 어떤 이벤트 리스너도 복사되지않는다.
        // 또한, 네이티브에서는 현재 노드에 바인딩된 이벤트 리스너들을 `탐색` 또는 `가져올 수` 있는 그 어떤 API 도 제공하지않는다.

        // JQ clone 함수를 사용하여, 노드를 복사할 경우, bx-input 엘리먼트의 inactive 속성처럼, 새롭게 추가된 프로퍼티들은
        // 복사되지 않는다(cloneNode 도 마찬가지)

        //************************************************
        //************************************************
        // 이 문제가, bind 함수를 사용할 수 없는 결정적 이유이다.
        //************************************************
        //************************************************
    //    this.$paperWrap = $root.find(`.${uiConfig.className.structure.wrap}`).clone(true, true);
    //
    //    if (_.isEmpty(this.$paperWrap)){
    //        CONSOLE.prop.throw(BxPaper, 'bind 할 자식엘리먼트의', 'value');
    //    }
    //}

    let $root = this.$root = $(document.body);

    _createDim.call(this);

    this.$container = BASE.appendContainerElement($root, BxPaper.className);
    this.containerClassName = this.$container.attr('class');

}

/**
 * dim 생성 함수
 * @private
 * @memberOf BxPaper
 */
// this.$dim 속성 설정
function _createDim(){

    let $root = this.$root;
    let dimStyle = this[OPTS].dimStyle;

    if (_hasDim($root)) {
        dimStyle.color = 'clear';
    }

    this.$dim = this.$root.bxDim(dimStyle).show();
}

/**
 * 인자로 받은 $root 에 dim 존재 유/무 반환 함수
 * @param $root - document body 의 Jquery Instance
 * @returns {boolean}
 * @private
 */
function _hasDim($root){

    const DIM_CLASS = BXUI.Base.getComponentClassName('dim');

    return $root.find(`.${DIM_CLASS}`).length > 0;
}
/**
 * browser 넓이 반환 함수
 * @returns {Number}
 * @private
 * @memberOf BxPaper
 */
function _getBrowserWidth(){

    return document.body.clientWidth || window.innerWidth;
}

/**
 * browser 높이 반환 함수
 * @returns {Number}
 * @private
 * @memberOf BxPaper
 */
function _getBrowserHeight(){

    return document.body.clientHeight || window.innerHeight;
}
/**
 * this.$paperWrap 추가 클래스 지정 함수
 * @private
 * @memberOf BxPaper
 */
function _setPaperWrapClass(){

    let opts = this[OPTS];
    let $paperWrap = this.$paperWrap;
    let contentClass = `.${uiConfig.className.structure.content}`;
    let colorClass = uiConfig.className.color[opts.color];
    let typeClass = uiConfig.className.type[opts.type];
    let shadowClass = uiConfig.className.shadow[opts.shadow];

    if (!_.isEmpty(colorClass)){
        $paperWrap.addClass(colorClass);
    }

    if (!_.isEmpty(typeClass)){
        $paperWrap.addClass(typeClass);
    }

    if (!_.isEmpty(shadowClass)){
        $paperWrap.find(contentClass).addClass(shadowClass);
    }
}

/**
 * 컨테이너 구조 설정 함수
 * @private
 * @memberOf BxPaper
 */
function _initContainerStructure(){
    // drag 영역 center 정렬 시, magin-left 에 음수 값이 들어감에 따라 container 영역 밖으로 나가는 문제를 해결하기 위해
    // $container 영역에 magin-left 를 주어 해당 영역 밖으로 나가지 않도록 조정함.
    let $container = this.$container;
    let $paperWrap = this.$paperWrap;
    let $header = $paperWrap.find(`.${uiConfig.className.structure.header}`);
    let $titleContent = $header.find(`.${uiConfig.className.structure.titleContent}`);
    let $content = $paperWrap.find(`.${uiConfig.className.structure.content}`);
    let $contentDesc = $content.find(`.${uiConfig.className.structure.contentDesc}`);
    let $closeBtn = $paperWrap.find(`.${uiConfig.className.structure.btnWrap} .${uiConfig.className.structure.closeBtn}`);
    let opts = this[OPTS];

    if (!_.isEmpty(opts.title) && _.isEqual(opts.titleAlign, 'center')){
        $titleContent.css('text-align', opts.titleAlign);
        $titleContent.css('padding-left', 0);
    }

    $contentDesc.css('width', opts.width);
    $contentDesc.css('height', opts.height);
    $paperWrap.css('margin-left', -(opts.width / 2));

    let height = _getBrowserHeight();
    let paddingLeft = opts.width / 2;
    let top = (height - $paperWrap.height()) / 2;

    let x = _.isUndefined(opts.position.x) ? ($paperWrap.offset().left + paddingLeft) : opts.width / 2 + opts.position.x;
    let y = _.isUndefined(opts.position.y) ? top : opts.position.y;

    let scrollYOffset = window.pageYOffset;

    $container.height(height);
    $container.css('top', scrollYOffset);
    $container.css('padding-left', paddingLeft);

    $paperWrap.css('left', x);
    $paperWrap.css('top', y);

}

/**
 * render 함수 사용시, html 생성 함수
 * @returns {string}
 * @private
 * @memberOf BxPaper
 */
function _createDOMElement(){

    let html = [];
    let opts = this[OPTS];
    let title = opts.title;
    let template = opts.template;
    let structureClassName = uiConfig.className.structure;

    html.push(`<div class="${structureClassName.wrap}">`);

    // html.push(`<a href="#" class="${structureClassName.focus}">`);
    html.push(`<a href="#" class="${structureClassName.focus}" onclick="return false;"><span class="blind">focus</span></a>`);
    html.push(`<div class="${structureClassName.drag}"></div>`);

    if (!_.isEmpty(title)){

        html.push(`<div class="${structureClassName.header}">`);
        html.push(`<h2 class="bx-tt">${title}</h2>`);
        html.push('</div>');
    }

    html.push(`<div class="${structureClassName.btnWrap}">`);
    html.push(`<button type="button" class="${structureClassName.closeBtn}">`);
    html.push(`<i class="${structureClassName.bwIcon} ${structureClassName.closeIconSize} ${structureClassName.closeIcon}">`);
    html.push(`<span class="blind">${ALT_TEXT.close}</span>`);
    html.push('</i>');
    html.push('</button>');
    html.push('</div>');

    html.push(`<div class="${structureClassName.content}">`);
    html.push(`<div class="${structureClassName.contentDesc}">`);
    html.push(template);
    html.push('</div>');
    html.push('</div>');
    // html.push('</a>');
    html.push('</div>');

    return html.join('');
}
/**
 * BxPaper 의 마지막 좌표 값 셋팅 함수.
 *
 * @private
 * @memberOf BxPaper
 */
function _setLastPosition(){
    let $container = this.$container;
    let $paperWrap = this.$paperWrap;

    let lastPosition = $paperWrap.offset();
    this.lastPosition = {
        top: lastPosition.top - $container.offset().top,
        left: lastPosition.left
    };
}

/**
 * 선택된 컨테이너를 body 최하위로 이동시키는 함수
 * @private
 * @memberOf BxPaper
 */
function _moveContainerBottom(){

    // this.$root.append(this.$container);

    // let a = this.$container.clone();
    // console.log(this.$root, this.$container, a);
    //
    // this.$root.replaceWith(this.$container, a);

    this.$root.append(this.$container);

    // let arr = document.getElementsByClassName('-bx-ui-paper');
    // document.getElementsByTagName('body').replaceChild()
}

/**
 * draggable 설정 함수
 * @private
 * @memberOf BxPaper
 */
function _initDrag(){

    let option = this[OPTS];
    let $paperWrap = this.$paperWrap;
    let onDrag = option.onDrag;

    let onDragStart = (e, $elem) =>{
        // dom Element 를 최하위로 옮겨 z-index 가 꼬이는 것을 방지한다.
        _moveContainerBottom.call(this);
        option.onDragStart.call(this, e, $elem);
    };

    let onDragStop = (e, $elem) =>{
        _setLastPosition.call(this);
        option.onDragStop.call(this, e, $elem);
    };

    let dragOption = {
        containment: "parent",
        snapMode: "inner",
        scroll: false,
        drag: onDrag,
        start: onDragStart,
        stop: onDragStop,
        handle: `.${uiConfig.className.structure.drag}`
        // cursor: 'crosshair',
        // cursorAt: { left: 5 }
    };

    $paperWrap.draggable(dragOption);
}

/**
 * 이벤트 설정 함수
 * @private
 * @memberOf BxPaper
 */
function _addEvent(){

    let $root = this.$root,
        $paperWrap = this.$paperWrap,
        $drag = $paperWrap.find(`.${uiConfig.className.structure.drag}`),
        closeBtnClass = `.${uiConfig.className.structure.btnWrap} .${uiConfig.className.structure.closeBtn}`,
        $focus = $paperWrap.find(`.${PAPER_STRUCTURE_CLASS_NAME.focus}`),
        focus = $focus.get(0);

    // 컨테이너 선택시, 가장 위에 오게 container 위치 조정 및 사용자 함수 call
    $drag.click((e) =>{

        if ($root.find(`.${this.containerClassName}`).length !== 1){
            _moveContainerBottom.call(this);
        }

        $focus.focus();
        this[OPTS].onActivate.call(this, e, $paperWrap);
    });

    $paperWrap.find(closeBtnClass).click(() =>{
        console.log('close btn click!!!');

        window.setTimeout(() =>{
            this.close();
        }, 0);
    });

    $paperWrap.on('focusin', function (e){

        window.setTimeout(()=>{
            UTIL.setPreFocusInstanceKey('hoverer');
        }, 0);
    });
}

/**
 * onRender 함수
 * @private
 * @memberOf BxPaper
 */
function _onRender(){

    this[OPTS].onRender.call(this);

    this.$paperWrap.find(`.${PAPER_STRUCTURE_CLASS_NAME.focus}`).focus();
}

/**
 * window resize 함수
 * @private
 * @memberOf BxPaper
 */
// function _resizer(){


//     let $container = this.$container;
//     let $paperWrap = this.$paperWrap;

//     // browser width & height value

//     let width = _getBrowserWidth();
//     let height = _getBrowserHeight();

//     let wrapMarginLeft = this[OPTS].width / 2;

//     let preWidth = this.preWidth;
//     let preHeight = $container.height();

//     let paperWrapTop = $paperWrap.offset().top - $container.offset().top;
//     let paperWrapLeft = $paperWrap.offset().left + wrapMarginLeft;

//     let top = 0;
//     let left = 0;

//     console.log('resize', this.$paperWrap);

//     // 넓이 증가
//     if (preWidth <= width){

//         if ($paperWrap.offset().left >= this.lastPosition.left){

//             left = `${this.lastPosition.left + wrapMarginLeft}`;
//         }
//         else {

//             left = `${width - this[OPTS].width + wrapMarginLeft}`;
//         }
//     }
//     // 넓이 감소
//     else{

//         if ($paperWrap.offset().left + this[OPTS].width > width || $paperWrap.offset().left < 0){

//             left = `${width - this[OPTS].width + wrapMarginLeft}`;
//         }
//         else{

//             left = `${paperWrapLeft}`;

//         }
//     }

//     // 높이 증가
//     if (preHeight < height){
//         // 컨테이너 작은경우
//         if ($paperWrap.height() <= height){
//             if (paperWrapTop >= this.lastPosition.top){

//                 top = `${this.lastPosition.top}`;

//             }
//             else{

//                 top = `${height - $paperWrap.height()}`;
//             }
//         }
//         // 컨테이너 영역이 더 큰 경우.
//         else{

//             if (paperWrapTop < this.lastPosition.top){
//                 top = `${paperWrapTop}px`;
//             }
//             else{
//                 top = `${height - $paperWrap.height()}`;
//             }

//         }
//     }
//     // 높이 감소
//     else{

//         if (paperWrapTop + $paperWrap.height() > height && paperWrapTop >= 1){

//             top = `${height - $paperWrap.height()}`;

//         }
//         else{

//             top = `${paperWrapTop}`;

//         }

//     }

//     $container.height(height);
    
//     // $paperWrap top 조정
//     if (width >= $paperWrap.width() && height >= $paperWrap.height()){
//         UTIL.setPosition($paperWrap, {top: top, left: left});
//     }

//     _setLastPosition.call(this);

//     this.preWidth = width;
// }

module.exports = BxPaper;
