// bxuip-angular.js-ui/src/lib/bx-ui/app/components/progress.js
/**
 * render 메서드 호출 시, 호출될 callback 함수
 *
 * @callback bx-render
 *
 * @memberOf BxProgress
 */

/**
 * eventName에 할당한 이벤트 발생 시, 호출될 callback 함수
 * (ex. bx-click="function(){}")
 *
 * @callback bx-eventName
 * @param {object} event - event 객체
 *
 * @memberOf BxProgress
 */

/**
 * progress의 값이 변경될 시 호출될 callback 함수
 *
 * @callback bx-change
 * @param {int} value - progress value
 *
 * @memberOf BxProgress
 */

/**
 * progress의 값이 시작값과 동일할 때 호출될 callback 함수
 *
 * @callback bx-start
 * @param {int} value - progress value
 *
 * @memberOf BxProgress
 */

/**
 * progress의 값이 종료값과 동일할 때 호출될 callback 함수
 *
 * @callback bx-end
 * @param {int} value - progress value
 *
 * @memberOf BxProgress
 */

/**
 * @class BxProgress
 * @classDesc BxProgress 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 * <script>
 *
 *  $(document).ready(function(){
 *    // DOM이 ready된 시점에서 초기화 함수를 호출한다.
 *          $().BxProgress().initialize();
 *     });
 *
 *  window.formMethod = {
 *      sliderStart: function(value){
 *           this.stopDrag();
 *           alert('start')
 *       },
 *       sliderEnd: function(value){
 *           this.stopDrag();
 *           alert('end')
 *       },
 *       sliderChange: function(value){
 *           console.log('slider 값을 ' + value + '로 변경하였습니다.');
 *       },
 *       sliderRender: function(value){
 *           console.log('render');
 *       }
 *   };
 *
 *  </script>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // bar type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *  <bx-progress id='progress1'
 *               type="bar"
 *               shape="round"
 *               label="theme"
 *               angle="horizon"
 *               effect="wave"
 *               decorator="{type:'tooltip', unit: '%'}"
 *               defaultValue="50"
 *               min="0"
 *               max="100"
 *               color="theme"
 *               colorLevel="4"
 *               size=default
 *               bx-start="formMethod.sliderStart"
 *               bx-end="formMethod.sliderEnd"
 *               bx-change="formMethod.sliderChange"
 *               bx-render ="formMethod.sliderRender">
 *  </bx-progress>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // slider type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *  <bx-progress id='progress2'
 *               type="slider"
 *               label="값을 변경하면 이벤트가 발생합니다."
 *               decorator="{type:'tooltip', unit: '%'}"
 *               defaultValue="50"
 *               angle="horizon"
 *               min="0"
 *               max="100"
 *               step="10"
 *               icon="car"
 *               color="theme"
 *               colorLevel="5"
 *               size="large"
 *               bx-start="formMethod.sliderStart"
 *               bx-end="formMethod.sliderEnd"
 *               bx-change="formMethod.sliderChange"
 *               bx-render ="formMethod.sliderRender">
 *  </bx-progress>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 * // circle type
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 *  <bx-progress id='progress3'
 *               type="circle"
 *               label="circle"
 *               shape="tooltip"
 *               decorator="{unit: '%'}"
 *               defaultValue="77"
 *               valueEffect="bounce"
 *               min="0"
 *               max="80"
 *               color="theme"
 *               colorLevel="4"
 *               size="large"
 *               bx-start="formMethod.sliderStart"
 *               bx-end="formMethod.sliderEnd"
 *               bx-change="formMethod.sliderChange">
 *</bx-progress>
 */

// 사용자에게 제공되는 API 목록
const APIS = {
    /**
     *
     * @deprecated Deprecated in: v3.2.0, Deprecated since: v3.12.0
     *
     * progress의 현재 value 반환 함수
     *
     * @returns {object} 생성된 인스턴스
     * @memberof BxProgress
     */
    getValue: () => {
        CONSOLE.prop.throw(BxProgress, 'getValue', 'deprecatedMethod');
    },
    /**
     *
     * @deprecated Deprecated in: v3.2.0, Deprecated since: v3.12.0
     *
     * progress color 설정 함수
     *
     * @returns {object} 생성된 인스턴스
     * @memberof BxProgress
     */
    setColor: () => {
        CONSOLE.prop.throw(BxProgress, 'setColor', 'deprecatedMethod');
    },
    /**
     *
     * @deprecated Deprecated in: v3.2.0, Deprecated since: v3.12.0
     *
     * progress color level 설정 함수
     *
     * @returns {object} 생성된 인스턴스
     * @memberof BxProgress
     */
    setColorLevel: () => {
        CONSOLE.prop.throw(BxProgress, 'setColorLevel', 'deprecatedMethod');
    },
    /**
     *
     * @deprecated Deprecated in: v3.2.0, Deprecated since: v3.12.0
     *
     * progress 값 설정 함수
     *
     * @returns {object} 생성된 인스턴스
     * @memberof BxProgress
     */
    setValue: () => {
        CONSOLE.prop.throw(BxProgress, 'setValue', 'deprecatedMethod');
    },
    stopDrag: _stopDrag,
    changSlider: _changeValue
};

const ATTRNAMES = {
    "defaultvalue": 'defaultValue',
    "colorlevel": 'colorLevel',
    "valueeffect": 'valueEffect'
};

const OPTS = {
    /**
     * @property {object} bx-progress - 모든 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-progress.type='bar'] - <span class='important'>readonly</span> - 타입 (bar, slider, circle)
     * @property {string} [bx-progress.label=''] - <span class='important'>readonly</span> - label의 text 값
     * @property {string} [bx-progress.effect='pingpong'] - <span class='important'>readonly</span> - 애니메이션 효과 설정 (pingpong, wave) (*IE9 미지원)
     * @property {string} [bx-progress.decorator={type : 'default', unit : ''}] - <span class='important'>readonly</span> - value 설정 객체
     * @property {int} [bx-progress.min=0] - <span class='important'>readonly</span> - 시작 값 (default : 0)
     * @property {int} [bx-progress.max=100] - <span class='important'>readonly</span> - 종료 값 (default : 0)
     * @property {string} [bx-progress.size='default'] - <span class='important'>readonly</span> - 두께 사이즈 (tiny, default, large)
     * @property {function} [bx-progress.bx-[eventName]=function(){}] - 생성된 progress 엘리먼트에 해당 이벤트 핸들러를 바인딩시킨다.(ex. bx-click -> elem.onclick = function(){})
     *
     * @property {int} [bx-progress.defaultValue=0] - 기본값 (default : 0)
     * @property {string} [bx-progress.color=''] - 색상명 (theme, red, green, navy, gray)
     * @property {int} [bx-progress.colorLevel=3] - 색상 단계 설정 (1 ~ 5)
     * @property {function} [bx-progress.bx-start=function(){}] - progress의 값이 시작값과 동일할 때 호출될 callback 함수
     * @property {function} [bx-progress.bx-end=function(){}] - progress의 값이 종료값과 동일할 때 호출될 callback 함수
     * @property {function} [bx-progress.bx-change=function(){}] - progress의 값이 변경될 시 호출될 callback 함수
     * @property {function} [bx-progress.bx-render=function(){}] - render 메서드 호출 시, 호출될 callback 함수
     *
     * @alias Attributes - All Type
     *
     * @memberof BxProgress
     */
    common: {
        "type": 'bar',
        "label": '',
        "effect": 'wave',
        "decorator":{
            type : 'default',
            unit : ''
        },
        "defaultValue": 0,
        "min": 0,
        "max": 100,
        "color": 'theme',
        "colorLevel": 3,
        "size" : 'default',
        "bx-start": function(){},
        "bx-end": function(){},
        "bx-change": function(){},
        "bx-render": function(){}
    },
    /**
     * @property {object} bx-progress - bar 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-progress.shape='round'] - <span class='important'>readonly</span> - 디자인 타입 (round, square, dotted)
     * @property {string} [bx-progress.angle='horizon'] - <span class='important'>readonly</span> - 방향 설정 (vertical, horizon)
     * @alias Attributes - Bar Type
     *
     * @memberof BxProgress
     */
    bar: {
        "angle": 'horizon',
        "shape": 'round'
    },
    /**
     * @property {object} bx-progress - slider 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @property {string} [bx-progress.shape='round'] - <span class='important'>readonly</span> -디자인 타입  (round, square)
     * @property {string} [bx-progress.angle='horizon'] - <span class='important'>readonly</span> - 방향 설정 (vertical, horizon)
     * @property {int} [bx-progress.step=1] - <span class='important'>readonly</span> - 단계 값
     * @property {string} [bx-progress.icon='default'] - <span class='important'>readonly</span> - (default, car, calbecar, horse) vertical type에서는 default icon만 지원
     * @alias Attributes - Slider Type
     *
     * @memberof BxProgress
     */
    slider: {
        "angle": 'horizon',
        "shape": 'round',
        "step": 1,
        "icon": ''
    },
    /**
     * @property {object} bx-progress - circle 타입에서 사용하는 property. <br/><span class="important">(이외 native에서 사용하는 attribute도 모두 사용 가능)</span>
     * @alias Attributes - Circle Type
     *
     * @memberof BxProgress
     */
    circle: {}
};

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;

/* color class setting */
const COLOR_CLASS_NAME = {
    theme: '',
    red: 'red',
    green: 'green',
    navy: 'navy',
    gray: 'gray'
};

/* size class setting */
const SIZE_CLASS_NAME = {
    default: 'default',
    tiny: 'tiny',
    large: 'large'
};

const SIZE_CONFIG = {
    labelSize: 20,
    defaultVerticalSize: 120
};

const COLOR_LEVEL_CLASS_NAME = ['color1', 'color2', '', 'color3', 'color4'];

/* effect class setting */
const EFFECT_CLASS_NAME = {
    pingpong: 'highlight',
    wave: 'highlight2',
    verticalpingpong: 'highlight3',
    verticalwave: 'highlight3'
};

/* animation config setting */
const ANIMATION_CONFIG = {
    duration: 2000
};

/* circle type size setting */
const CIRCLE_SIZE_CONFIG = {
    minWidth: 160,
    maxWidth: 1020,
    minFontSize: 20,
    maxFontSize: 140,
    strokeWidth: 8
};

class BxProgress {

    constructor(){
        this.$root = BxProgress.$root;
        this.opts = OPTS;
    }

    /**
     * 초기화 함수
     *
     * <p>
     *     DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
     * </p>
     *
     * @returns {object} 생성된 인스턴스
     */
    initialize(){

        this.$progress = this.$root.find('bx-progress');

        this.elems = [];

        $.each(this.$progress, (i, el) => {

            let $el = $(el),
                type = $el.attr('type') || 'bar',
                options = _createComponentOptions($el, OPTS.common);

            let $container = _parseCustomElement($el, options, type);
            let progressElem = $container.get(0);
            let bxRender = progressElem['bxrender'];

            _setValue.call($container, options.defaultValue);

            _setProgressColor($container, options.color);

            _setProgressColorLevel($container, options.colorLevel);

            if (options.type === 'slider') _setSliderMoveEvent.call($container);

            UTIL.addAllEventListenerWithProperty($container, options.events);

            bxRender.call(progressElem);

            this.elems.push($container.get(0));

        });

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxProgress.isCustomElement = true;

/**
 * progress 값 설정 함수
 * @param {int} value - progress value
 * @alias setValue
 * @private
 * @memberOf BxProgress
 */
function _setValue(value){

    if (!_.isNumber(value)) {
        CONSOLE.prop.throw(BxProgress, 'value', 'type');
    }

    let $progressElem = $(this),
        progressElem = $progressElem.get(0),
        type = progressElem.type,
        angle = progressElem.angle,
        _k = UTIL.convertInternalPropertyName('defaultValue');

    if (value >= progressElem.max) value = progressElem.max;
    if (value <= progressElem.min) value = progressElem.min;

    progressElem[_k] = value;

    _setProgressValueText.call(this, value);

    if (type === 'circle'){
        _animateCircle.call(this, value);
    }
    else {

        let _value =  _getRateToValue.call(this, value);

        if (angle === 'vertical'){
            _setSliderHeight.call(this, _value);
        }
        else {
            _setSliderWidth.call(this, _value);
        }

    }

}

/**
 * progress의 현재 value 반환 함수
 *
 * @returns {int} progress value
 * @alias getValue
 * @private
 * @memberOf BxProgress
 */
function _getValue(){

    let $progressElem = $(this);

    return $progressElem.defaultValue;
}

/**
 * progress color 설정 함수
 * @param {string} theme|red|green|navy|gray - color name (theme|red|green|navy|gray)
 * @alias setColor
 * @private
 * @memberOf BxProgress
 */
function _setColor(colorName){

    if (!_.isString(colorName)) {
        CONSOLE.prop.throw(BxProgress, 'colorName', 'type');
    }

    if (!(colorName === 'theme' || colorName === 'red' || colorName === 'green' ||
        colorName === 'navy' || colorName === 'gray' || colorName === '')) {

        CONSOLE.prop.throw(BxProgress, 'colorName', 'value');
    }

    _resetProgressColor.call(this);

    _setProgressColor($(this), colorName);
}

/**
 * progress color level 설정 함수
 * @param {int} colorLevel - color level (1~5)
 * @alias setColorLevel
 * @private
 * @memberOf BxProgress
 */
function _setColorLevel(colorLevel){

    if (colorLevel && !_.isNumber(colorLevel)) {
        CONSOLE.prop.throw(BxProgress, 'colorLevel', 'type');
    }

    colorLevel = _getColorLevel(colorLevel);

    _resetProgressColorLevel.call(this);

    _setProgressColorLevel($(this), colorLevel);
}

/**
 * slider type의 drag event를 해제하는 함수
 * @alias stopDrag
 * @memberOf BxProgress
 */
function _stopDrag(){
    _stopDragMouseInBar.call(this);
}

/**
 * custom 태그를 native 태그로 치환한다.sliderStart
 * @param {object} $customElem - progress custom 컨테이너
 * @param {object} options - component options
 * @param {string} type - progress 타입
 *
 * @returns {object} 생성된 native element
 *
 * @private
 * @memberOf BxProgress
 */
function _parseCustomElement(
    $customElem = $(document.documentElement),
    options = {},
    type = 'default'
){

    let $container = BASE.appendContainerElement(BxProgress.$root, BxProgress.className);
    let $wrapper = $('<div></div>');

    _setNativeAttribute($container, options.attr);
    _setElementProperties(options, $container.get(0));

    // let html = _getComponentHtml($container, options, type);
    let html = _getComponentHtml($wrapper, options, type);

    $wrapper.append(html);
    $container.append($wrapper);

    $customElem.replaceWith($container);

    return $container;
}
/**
 * native attribute 설정 함수
 *
 * @param {object} $container - progress 컨테이너
 * @param {object} attr - native 속성
 *
 * @private
 * @memberOf BxProgress
 */
function _setNativeAttribute(
    $container = $(document.documentElement),
    attr = {}
){

    $.each(attr, (name, value)=>{
        if (name === 'class') return;
        $container.attr(name, value);
    });

    if (attr.class) $container.addClass(attr.class);
}
/**
 * circle type의 style 정보 반환 함수
 *
 * @returns {object} $container - progress container elemnt
 *
 * @private
 * @memberOf BxProgress
 */
function _getCircleStyleInfo($container = $(document.documentElement)){

    let width = $container.width();

    width = width < 160 ? 160 : width;

    let radius = width / 2;
    let strokeWidth = CIRCLE_SIZE_CONFIG.strokeWidth;
    let circleDiameter = (radius - strokeWidth) * 2 * 3.14;
    let minWidth = CIRCLE_SIZE_CONFIG.minWidth;
    let fontDrainage = Math.round((width / minWidth));
    let fontSize = CIRCLE_SIZE_CONFIG.minFontSize * fontDrainage;
    let fontMaxSize = CIRCLE_SIZE_CONFIG.maxFontSize;

    fontSize = fontSize > fontMaxSize ? fontMaxSize : fontSize;

    return {
        svg: {
            width: width,
            height: width
        },
        circle: {
            'r': radius - strokeWidth,
            'cy': radius,
            'cx': radius,
            'stroke-dasharray': circleDiameter
        },
        value: {
            'font-size': fontSize + 'px',
            'top': radius - fontSize/2 + 'px'
        },
        label: {
            'top': radius + fontSize/2 + 10 + 'px'
        }
    };
}

/**
 * progress color level 초기화 함수
 *
 * @private
 * @memberOf BxProgress
 */
function _resetProgressColorLevel(){

    let type = this.type;
    let $target;

    if (type === 'circle') $target =  $(this).find('.bx-progress');
    else $target =  $(this).find('.bx-progress-wrap');

    COLOR_LEVEL_CLASS_NAME.forEach((className, i)=>{
        $target.removeClass(className);
    });

}

/**
 * progress color 초기화 함수
 *
 * @private
 * @memberOf BxProgress
 */
function _resetProgressColor(){

    let $target = $(this).find('.bx-progress-wrap');

    $.each(COLOR_CLASS_NAME, (colorName, className)=>{
        $target.removeClass(className);
    });

}

/**
 * circle type의 원지름 반환 함수
 *
 * @param {object} $container - progress 컨테이너
 *
 * @returns {int} 계산된 원지름
 *
 * @private
 * @memberOf BxProgress
 */
function _getCircleDiameter($container = {}){

    let width = $container.find('svg').width();

    width = width < 160 ? 160 : width;

    let radius = width / 2;
    let strokeWidth = CIRCLE_SIZE_CONFIG.strokeWidth;

    return (radius - strokeWidth) * 2 * 3.14;
}

/**
 * circle type animate 함수
 *
 * @param {int} defaultValue - progress value
 *
 * @private
 * @memberOf BxProgress
 */
function _animateCircle(defaultValue){

    let $progressElem = $(this),
        $circle = $progressElem.find('circle'),
        value =  _getCircleValuePercent($progressElem, defaultValue);

    $circle.animate({'stroke-dashoffset': value}, {
        duration: ANIMATION_CONFIG.duration,
        queue: false // animate 비동기 처리
    });
}

/**
 * value 값으로 원지름의 값을 구하는 함수
 *
 * @param {object} $container - progress 컨테이너
 * @param {object} options - progress 옵션
 *
 * @returns {int} percent value
 *
 * @private
 * @memberOf BxProgress
 */
function _getCircleValuePercent(
    $container = $(document.documentElement),
    value = 0
){
    let strokeRadius = CIRCLE_SIZE_CONFIG.strokeWidth / 2,
        diameter = _getCircleDiameter($container),
        rate =  diameter / 100;

    return (100 - _getRateToValue.call($container, value)) * rate + strokeRadius;

}

/**
 * 타입별 progress html 반환 함수
 *
 * @param {object} $container - progress 컨테이너
 * @param {object} options - progress 옵션
 * @param {string} inputType - progress 타입
 *
 * @returns {object} 화면에 그려질 progress HTML
 *
 * @private
 * @memberOf BxProgress
 */
function _getComponentHtml(
    $container = $(document.documentElement),
    options = {},
    inputType = ''
){
    inputType !== 'circle' && _setCSSProgressWrap($container, options, inputType);

    switch (inputType){
        // case 'slider'   : return _getSliderTypeProgressHTML.call(this, options);
        case 'circle'   : return _getCircleTypeProgressHTML($container, options);
        case 'bar'      :
        case 'slider'   : return _getProgressHTML($container, options, inputType);
    }
}

/**
 * BXprogress wrap의 css 설정 함수
 *
 * @param {object} $container - progress 컨테이너
 * @param {object} options - progressbar 옵션
 * @param {string} inputType - progressbar 타입
 *
 * @private
 * @memberOf BxProgress
 */
function _setCSSProgressWrap(
    $container = $(document.documentElement),
    options = {},
    inputType = 'default'
){

    let label = options.label,
        size = options.size,
        angle = options.angle,
        icon = options.icon,
        type = options.decorator.type;

    if (!_.isString(inputType) || inputType.length === 0 || inputType === 'circle') return;

    // progressbar size
    if (size !== 'default') $container.addClass(`size-${size}`);

    // progressbar 수직 or 수평 설정
    if ( angle === 'vertical') {
        $container.addClass('type-vertical');
    }
    else {
        $container.addClass('type-inline');
    }

    if (inputType === 'bar'){

        // 툴팁 위치
        if (type === 'tooltip'){

            if (angle === 'vertical'){
                $container.addClass('add-num');
            }
            else {
                $container.addClass('add-top');
            }
        }
    } else {

        // 툴팁 위치
        if (angle === 'vertical'){
            // 세로 인 경우엔 무조건 기본형 ( 아이콘이 없음 )
            $container.addClass('add-btn');
        }
        else {

            if (type === 'tooltip'){
                // 툴팁이 있는 경우
                if (!_.isUndefined(icon)){
                    // 툴팁과 아이콘이 다 있는 경우
                    $container.addClass('add-all');
                }
                else {
                    // 아이콘이 없는 경우
                    $container.addClass('add-btn-all');
                }
            }
            else {
                // 툴팁이 없는 경우
                if (!_.isUndefined(icon) && (icon === 'car' || icon === 'horse')){
                    // 아이콘이 progress 위에 있는 경우 (car, horse)
                    $container.addClass('add-top');
                }
                else if (!_.isUndefined(icon)){
                    // 아이콘이 progress 아래 있는 경우 (cablecar)
                    $container.addClass('add-btm');
                }
                else {
                    // 툴팁과 아이콘이 다 없는 경우
                    $container.addClass('add-btn');
                }
            }
        }
    }
}


/**
 * circle type progress html 반환 함수
 *
 * @param {object} $container -  progress 컨테이너
 * @param {object} options - component options
 *
 * @returns {string} 화면에 그려질 progress html
 *
 * @private
 * @memberOf BxProgress
 */
function _getCircleTypeProgressHTML(
    $container = $(document.documentElement),
    options = {}
){

    let label = options.label;
    let decorator = options.decorator;
    let styleInfo = _getCircleStyleInfo($container);

    let svgStyle = styleInfo.svg;
    let valueStyle = styleInfo.value;
    let labelStyle = styleInfo.label;
    let circleStyle = styleInfo.circle;

    let svgAttr = `style="width:${svgStyle.width}px;height:${svgStyle.width}px;"`;

    let circleAttr = [];

    circleAttr.push(`class="circle_animation"`);
    circleAttr.push(`r="${circleStyle.r}"`);
    circleAttr.push(`cx="${circleStyle.cx}"`);
    circleAttr.push(`cy="${circleStyle.cy}"`);
    circleAttr.push(`style="stroke-dasharray: ${circleStyle['stroke-dasharray']};`);
    circleAttr.push(`stroke-dashoffset:${circleStyle['stroke-dasharray']}"`);
    circleAttr.push('fill="none"');

    let html = [];

    html.push('<div class="'+ _getCircleTypeWrapClass(options) +'">');

    if (label){

        let labelAttr = 'class="bx-label"' +
            'style="top:' + labelStyle.top + ';"';
        html.push('<p '+labelAttr+'>'+ label +'</p>');
    }

    if (decorator){

        let valutStyleText = 'top:' + valueStyle.top + ';' +
            'font-size:' + valueStyle['font-size'];

        html.push(_getValueSignHTML(options, valutStyleText));
    }

    html.push('<svg '+ svgAttr +'>');
    html.push('<g>');
    html.push('<circle ' + circleAttr.join(' ') + '></circle>');
    html.push('</g>');
    html.push('</svg>');

    return html.join('');
}

/**
 * progress html 반환 함수
 *
 * @param {object} options - progress 옵션
 * @param {object} inputType - progress 타입
 *
 * @returns {string} 화면에 그려질 progress html
 *
 * @private
 * @memberOf BxProgress
 */
function _getProgressHTML(
    $container = $(document.documentElement),
    options = {},
    inputType = 'default'
){

    let label = options.label,
        icon = options.icon,
        type = options.decorator.type,
        effect = options.effect,
        angle = options.angle,
        shape = options.shape,
        rate = _getRateToValue.call($container);

    let html = [];
    let progressLength = angle === 'vertical' ? `height:${rate}%` : `width:${rate}%`;
    effect = angle === 'vertical' ? 'vertical'+effect : effect;

    // square는 명칭을 css에 맞게 변경
    // vertical dotted 타입은 전부 기본형으로 변경
    if (shape === 'square'){
        shape = 'squared';
    }
    else if (shape === 'dotted' && angle === 'vertical'){
        shape = 'round';
    }

    html.push(`<div class="bx-progress-wrap ${shape}">`);

    if (label && angle !== 'vertical') html.push(`<p class="bx-label">${label}</p>`);

    html.push(`<div class="bx-progress" ${!label ? 'style="left:0;"' : ''}>`);
    html.push(`<div class="${_getBarClass(effect)}" style="${progressLength}">`);

    inputType === 'slider' && html.push(_getIconClassHTML(angle, icon));
    type === 'tooltip' && html.push(_getValueSignHTML(options));

    html.push('</div>');
    html.push('</div>');
    html.push('</div>');

    if (label && angle === 'vertical') html.push(`<p class="bx-label">${label}</p>`);

    let height = $container.height();

    if (angle === 'vertical') {

        if (height === 0){
            // 높이를 지정 하지 않을 경우 최소 120px로 설정
            $container.css('height', SIZE_CONFIG.defaultVerticalSize);
        }
        else {
            // label 값이 있는 경우
            if (label) $container.css('height', (height - SIZE_CONFIG.labelSize));
        }

    }

    return html.join('');
}

/**
 * slider Type별 icon html 반환 함수
 *
 * @param {string} angle - progress angle
 * @param {string} icon - progress 아이콘
 *
 * @returns {string} icon html
 *
 * @private
 * @memberOf BxProgress
 */
function _getIconClassHTML(
    angle = '',
    icon = ''
){
    let html = '',
        iconClass = 'default';

    if (angle !== 'vertical'){

        switch (icon){
            case 'car'      : iconClass = 'icon-type i-car moving'; break;
            case 'cablecar' : iconClass = 'icon-type i-cablecar swing'; break;
            case 'horse'    : iconClass = 'icon-type i-horse swing'; break;
        }
    }

    html = `<span class="bx-ctrl-btn ${iconClass}"></span>`;

    return html;
}

/**
 * value decorator html 반환 함수
 *
 * @param {object} component options
 * @param {string} style - progress circle타입 style
 *
 * @returns {string} decorator html
 *
 * @private
 * @memberOf BxProgress
 */
function _getValueSignHTML(
    options = {},
    style = ''
){
    let html = [],
        elClass = 'bx-num',
        type = options.type,
        decorator = options.decorator,
        num = options.defaultValue,
        effect = options.valueEffect,
        unit = decorator.unit;

    if (type === 'circle' && effect === 'bounce') elClass += ' rubberband';

    html.push(`<span class="${elClass}" style="${style}">`);
    html.push('<b></b>');

    if (unit) html.push(`<span class="unit">${unit}</span>`);

    html.push('</span>');

    return html.join('');
}

/**
 * bar type progress class 반환 함수
 *
 * @param {string} effect - progress animation 타입
 *
 * @returns {string} 생성된 animation class
 *
 * @private
 * @memberOf BxProgress
 */
function _getBarClass(effect = ''){
    let barClass = 'bx-bar';
    if (effect) barClass += ' ' + EFFECT_CLASS_NAME[effect];
    return barClass;
}

/**
 * bar type progress의 class 반환 함수
 *
 * @param {object} component options
 *
 * @private
 * @memberOf BxProgress
 */
// function _getBarTypeWrapClass(options = {}){
//
//     let shape = options.shape,
//         type = options.type,
//         label = options.label,
//         size = options.size,
//         color = options.color,
//         decorator = options.decorator;
//
//     let elClass = 'bx-progress';
//
//     elClass += ' size-' + size;
//
//     if (color && color !== 'theme') elClass += ' ' + COLOR_CLASS_NAME[color];
//
//     if (label) elClass += ' label';
//
//     if (decorator.type === 'tooltip') elClass += ' bubble';
//
//     if (type === 'slider' && shape === 'dotted') return elClass;
//
//     if (shape && shape !== 'round') elClass += ' ' + shape;
//
//     return elClass;
// }

/**
 * circle type progress class 반환 함수
 *
 * @param {object} component options
 *
 * @returns {string} circle wrap html
 *
 * @private
 * @memberOf BxProgress
 */
function _getCircleTypeWrapClass(options = {}){

    let size = options.size,
        color = options.color,
        effect = options.effect;

    let elClass = `bx-progress circle size-${size}`;

    if (color && color !== 'theme') elClass += ' ' + COLOR_CLASS_NAME[color];

    if (effect && effect !== 'pingpong') elClass += ' ' + EFFECT_CLASS_NAME[effect];

    return elClass;
}

/**
 * value의 백분율 반환 함수
 *
 * @param {int} value - progress value
 *
 * @returns {int} 기본값을 백분율로 변환한
 *
 * @private
 * @memberOf BxProgress
 */
function _getRateToValue(value){

    let $progressElem = $(this);
    let progressElem = $progressElem.get(0);

    let min = parseInt(progressElem.min),
        max = parseInt(progressElem.max),
        num = parseInt(value);

    if (num > max) num = max;
    if (num < min) num = min;

    return parseFloat((num - min) / (max - min) * 100);
}

/**
 * colorLevel 값을 반환한다.
 *
 * @param {int} colorLevel - colorLevel 값
 *
 * @private
 *
 * @memberOf BxProgress
 */
function _getColorLevel(colorLevel){

    if (colorLevel < 1) colorLevel = 1;

    if (colorLevel > 5) colorLevel = 5;

    return colorLevel;
}

/**
 * component options 반환 함수
 *
 * @param {object} $container - Progress 컨테이너
 *
 * @returns {object} 설정된 Progress 옵션
 *
 * @private
 * @memberOf BxProgress
 */
function _createComponentOptions($targetObject = null, opt = {}){

    if (!_.isFunction($targetObject.get) && !_.isPlainObject($targetObject)){
        CONSOLE.prop.throw(BxProgress, 'targetObject', 'type');
    }

    let defaultOpts = $.extend({}, opt),
        type = $targetObject.attr('type');

    switch(type){

        case 'bar': $.extend(defaultOpts, OPTS.bar); break;

        case 'slider': $.extend(defaultOpts, OPTS.slider); break;

        case 'circle': $.extend(defaultOpts, OPTS.circle); break;

        default : $.extend(defaultOpts, OPTS.bar);

    }

    let opts = UTIL.createComponentOptions($targetObject, defaultOpts, ATTRNAMES, function(k, v){

        if (k === 'decorator') {

            return _parseObject(k, v);
        }

        return v;

    }, function(k, v){

        let _opts = this;

        //if (!_.isPlainObject(_opts.events)) _opts.events = {};

        if (k === 'class'){
            v = ' ' + v;
        }

        // 이벤트 핸들러
        //if (k.match(/^bx-/)) {
        //
        //    _opts.events[k] = v;
        //    return;
        //}

        return v;

    });

    _checkValueType(opts);// 유효성 검사

    return  $.extend(opts, _setCommonDefaultValue(opts));
}


/**
 * 프로퍼티 세팅 함수
 *
 * @param {object} 옵션 객체
 * @param {object} 프로퍼티를 세팅할 엘리먼트
 *
 * @private
 * @memberOf BxProgress
 */
function _setElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxProgress, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxProgress, 'elem', 'type');
    }

    _.map(opts, (v, k) => {

        // attr 속성은 무시한다.
        if (k === 'attr') return;
        // events 속성은 무시한다.
        if (k === 'events') return;

        if (k.match(/^bx-/)){

            let _k = k.replace('-', '');
            delete opts[k];
            k = _k;
        }

        // 읽기 전용 속성
        if (
            k === 'fill' || k === 'decorator' ||
            k === 'angle'|| k === 'min'||
            k === 'max' || k === 'icon'||
            k === 'label' || k === 'size'||
            k === 'effect' || k === 'type' ||
            k === 'shape' || k === 'step'
        ){

            if (k ==='step' || k === 'min' || k === 'max') {
                v = parseInt(v);
            }

            UTIL.setReadOnlyProperty(elem, k, v);
        }
        else {

            let _k = UTIL.convertInternalPropertyName(k);

            elem[_k] = v;

            UTIL.setPropertyWithDescriptor(elem, k, {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) => {
                    switch (k){
                        case 'color' :
                            elem[_k] = v;
                            _setColor.call(elem, v); break;

                        case 'colorLevel' :
                            elem[_k] = v;
                            _setColorLevel.call(elem, v); break;

                        case 'defaultValue' :
                            elem[_k] = v;
                            _changeValue.call(elem, v);
                            break;

                        case 'bxrender':
                        case 'bxchange':
                        case 'bxstart':
                        case 'bxend':
                            if (!_.isFunction(v) ){
                                CONSOLE.prop.throw(BxProgress, k, 'type');
                            }
                            elem[_k] = v;
                            break;
                    }
                }
            });
        }
    });

    _.map(APIS, (v, k) => {

        UTIL.setReadOnlyProperty(elem, k, v);
    });

    return this;
}

/**
 * 공통 options의 초기값 설정 함수
 *
 * @param {} options
 * @param {string} [options.label] - 라벨
 * @param {string} [options.min] - progress 최소 값
 * @param {string} [options.max] - progress 최대 값
 * @param {string} [options.size] - progress 크기
 * @param {string} [options.color] - progress 색상
 * @param {string} [options.colorLevel] - progress 색상농도
 * @param {string} [options.decorator] - value 옵션
 * @param {string} [options.defaultValue] - 초기값
 *
 * @returns {object} 초기값이 설정된 options
 *
 * @private
 * @memberOf BxProgress
 */
function _setCommonDefaultValue({
    label = '',
    min = 0,
    max = 100,
    size = 'default',
    color = 'theme',
    colorLevel = 3,
    decorator = {
        type : 'default',
        unit : ''
    },
    defaultValue = 0,
    shape = 'round'
}){

    if (_.isUndefined(decorator.type)) decorator.type = 'default';

    if (_.isUndefined(decorator.unit)) decorator.unit = '';

    if (defaultValue) defaultValue = parseInt(defaultValue);

    if (min) min = parseInt(min);

    if (max) max = parseInt(max);

    if (max <= min) max = min;

    if (_.isString(color) && _.isUndefined(COLOR_CLASS_NAME[color])) color = 'theme';

    if (!_.isUndefined(colorLevel) && isNaN(colorLevel)) colorLevel = 3;

    return {
        label: label,
        min: min,
        max: max,
        size: size,
        shape: shape,
        color: color,
        colorLevel: colorLevel,
        decorator: decorator,
        defaultValue: defaultValue
    };
}

/**
 * component option 유효성 검사 함수
 *
 * @param {object} options - progress 옵션
 *
 * @private
 * @memberOf BxProgress
 */
function _checkValueType(options = {}){

    let decorator = options.decorator,
        min = options.min,
        max = options.max,
        step = options.step,
        size = options.size,
        label = options.label,
        defaultValue = options.defaultValue;

    if (decorator && !_.isPlainObject(decorator)) {
        CONSOLE.prop.throw(BxProgress, 'decorator', 'type');
    }

    if (defaultValue && !_.isNumber(parseInt(defaultValue))){
        CONSOLE.prop.throw(BxProgress, 'defaultValue', 'type');
    }

    if (min && !_.isNumber(parseInt(min))) {
        CONSOLE.prop.throw(BxProgress, 'min', 'type');
    }

    if (max && !_.isNumber(parseInt(max))) {
        CONSOLE.prop.throw(BxProgress, 'max', 'type');
    }

    if (step && !_.isNumber(parseInt(step))) {
        CONSOLE.prop.throw(BxProgress, 'step', 'type');
    }

    if (size && (!_.isString(size) || _.isUndefined(SIZE_CLASS_NAME[size]))){
        CONSOLE.prop.throw(BxProgress, 'size', 'all');
    }

    if (label && (!_.isString(label) || label.length === 0)){
        CONSOLE.prop.throw(BxProgress, 'label', 'all');
    }

}

/**
 * 색상 레벨 설정 함수
 *
 * @param {object} $container - progress 컨테이너
 * @param {object} options - progress 옵션
 *
 * @private
 * @memberOf BxProgress
 */
function _setProgressColorLevel(
    $container = $(document.documentElement),
    colorLevel = 3
){

    let _colorLevel = _getColorLevel(colorLevel);
    let type = $container.get(0).type;
    let $progress;

    if (type === 'circle') $progress =  $container.find('.bx-progress');
    else $progress =  $container.find('.bx-progress-wrap');

    $progress.addClass(COLOR_LEVEL_CLASS_NAME[_colorLevel - 1]);
}

/**
 * 색상 설정 함수
 *
 * @param {object} $container - progress 컨테이너
 * @param {object} options - progress 옵션
 *
 * @private
 * @memberOf BxProgress
 */
function _setProgressColor(
    $container = $(document.documentElement),
    color = 'theme'
){

    let type = $container.get(0).type;

    if (type === 'circle') $container.find('.bx-progress').addClass(COLOR_CLASS_NAME[color]);
    else $container.find('.bx-progress-wrap').addClass(COLOR_CLASS_NAME[color]);

}

/**
 * slider type drag event bind 함수
 *
 * @private
 * @memberOf BxProgress
 */
function _setSliderMoveEvent(){

    this.find('.bx-progress')
        .on('mousedown', _startDragMouseInBar.bind(this));

    $(window).on('mouseup.BXprogressEnd', _stopDragMouseInBar.bind(this));
}

/**
 * slider type drag start event bind 함수
 *
 * @param {object} event 객체
 *
 * @private
 * @memberOf BxProgress
 */
function _startDragMouseInBar(e){

    let $progressElem = $(this);
    let progressElem = $progressElem.get(0);
    let value = _getSliderValueAndLength.call(this, e);

    if (progressElem.angle !== 'vertical'){
        _changeValue.call(this, Math.round(value.width));
    }
    else {
        _changeValue.call(this, Math.round(value.height));
    }

    $(window).on('mousemove.BXprogressDrag', _dragMouseInBar.bind(this));
}

/**
 * slider type drag event off 함수
 *
 * @private
 * @memberOf BxProgress
 */
function _stopDragMouseInBar(){

    let $progressElem = $(this);

    $progressElem.css('cursor', 'pointer');

    $(window).off('mousemove.BXprogressDrag');
}

/**
 * slider type drag event
 *
 * @param {object} event 객체
 *
 * @private
 * @memberOf BxProgress
 */
function _dragMouseInBar(e){

    let $progressElem = $(this);
    let progressElem = $progressElem.get(0);
    let value = _getSliderValueAndLength.call(this, e);

    if (progressElem.defaultValue !== value){

        if (progressElem.angle !== 'vertical'){
            _changeValue.call(this, Math.round(value.width));
        }
        else {
            _changeValue.call(this, Math.round(value.height));
        }
    }
}

/**
 * slider change handler
 *
 * @param {int} value - 변경될 값
 *
 * @private
 * @memberOf BxProgress
 */
function _changeValue(value){
    _setValue.call(this, value);
    _callDirectiveBXEvent.call(this);
}

/**
 * value 세팅 함수
 *
 * @param {int} value - 설정 할 값
 *
 * @private
 * @memberOf BxProgress
 */
function _setProgressValueText(value){

    if (value >= this.max) value = this.max;
    if (value <= this.min) value = this.min;

    $(this).find('.bx-num b').text(value);
}

/**
 * slider bar width 설정 함수
 * @param {int} value - 설정 할 값
 * @private
 * @memberOf BxProgress
 */
function _setSliderWidth(width){

    $(this).find('.bx-bar').css('width', width + '%');
}

/**
 * slider bar height 설정 함수
 * @param {int} value - 설정 할 값
 * @private
 * @memberOf BxProgress
 */
function _setSliderHeight(height){

    $(this).find('.bx-bar').css('height', height + '%');
}

/**
 * slider value, width 반환 함수
 * @param {object} e - drag event obejct
 * @private
 * @memberOf BxProgress
 */
function _getSliderValueAndLength(e){

    let $progressElem = $(this);
    let progressElem = $progressElem.get(0);

    let max = parseInt(progressElem.max),
        min = parseInt(progressElem.min),
        value = parseInt(progressElem.defaultValue);

    let step = parseInt(progressElem.step) || 1;

    let $target = this.find('.bx-progress');
    let prevWidth = $target.find('.bx-bar').width();
    let prevHeight = $target.find('.bx-bar').height();

    if (step < min) step = min;
    if (step > max) step = max;

    let totalWidth = $target.width();
    let totalHeight = $target.height();
    let widthPointer = e.clientX - $target.offset().left;
    let heightPointer = ($target.offset().top + totalHeight) - e.pageY;

    let widthPerStep = totalWidth * (step / max);
    let heightPerStep = totalHeight * (step / max);

    let widthPosition = widthPointer - prevWidth;
    let heightPosition = heightPointer - prevHeight;

    let widthCell = widthPosition % step;
    let heightCell = heightPosition % step;
    let modifiedWidthPosition = (widthCell > (step /2))? widthPosition + step - widthCell : widthPosition - widthCell;
    let modifiedHeightPosition = (heightCell > (step /2))? heightPosition + step - heightCell : heightPosition - heightCell;

    let offsetWidthStep = Math.round(modifiedWidthPosition / widthPerStep);
    let offsetHeightStep = Math.round(modifiedHeightPosition / heightPerStep);
    let offsetWidth = offsetWidthStep * widthPerStep;
    let offsetHeight = offsetHeightStep * heightPerStep;

    let _widthPosition = value + (offsetWidthStep * step);
    let _heightPosition = value + (offsetHeightStep * step);
    let widthRate = (prevWidth + offsetWidth) / totalWidth * 100;
    let heightRate = (prevHeight + offsetHeight) / totalHeight * 100;

    if (widthRate > 100) widthRate = 100;
    if (heightRate > 100) heightRate = 100;

    return {
        widthPosition: _widthPosition,
        width: widthRate,
        heightPosition: _heightPosition,
        height: heightRate
    };
}


/**
 * directive event bind 함수
 * @private
 * @memberOf BxProgress
 */
function _callDirectiveBXEvent(){

    let $progressElem = $(this);
    let progressElem = $progressElem.get(0);

    let bxChange = progressElem['bxchange'];
    let bxStart = progressElem['bxstart'];
    let bxEnd = progressElem['bxend'];
    let value = progressElem.defaultValue;

    if (bxChange) bxChange.call(progressElem, value);
    if (bxStart && value === parseInt(progressElem.min)) bxStart.call(progressElem, value);
    if (bxEnd && value === parseInt(progressElem.max)) bxEnd.call(progressElem, value);
}

/**
 * object parsing 함수
 * @param {string} name - 함수 명
 * @param {string} string - JSON Object로 변환 할 텍스트
 * @returns {object} 변환된 JSON Object
 * @private
 * @memberOf BxProgress
 */
function _parseObject(name = '', string = ''){

    try {
        return UTIL.JSParser(string);
    }
    catch (e){
        CONSOLE.prop.throw(BxProgress, name, 'type');
    }
}

module.exports = BxProgress;
