// bxuip-angular.js-ui/src/lib/bx-ui/app/components/select.js
/**
 * Created by seonggyun on 2017. 01. 03.
 */
/**
 *
 *  @class BxSelect
 *  @desc BxSelect 클래스 생성자
 *  @example
 *
 * <bx-select
 *
 *  id="test1"
 *  type="default"
 *  shape="default"
 *  multiple="false"
 *  disabled="false"
 *  icon=""
 *  label="-- 전체 --"
 *  maxHeight="200"

 *  bx-click="function(){ console.log('bx-click'); }"

 *  bx-toggle="function(state){ console.log('bx-toggle'); console.log(state); }"
 *  bx-show="function(){ console.log('show'); }"
 *  bx-hide="function(){ console.log('hide'); }"

 *  bx-prepend="function(type, item){ console.log('bx-prepend'); console.log(type, item);}"
 *  bx-add="function(type, item){ console.log('bx-add'); console.log(type, item); }"
 *  bx-insertBefore="function(type, item){ console.log('bx-insertBefore'); console.log(type, item); }"
 *  bx-insertAfter="function(type, item){ console.log('bx-insertAfter'); console.log(type, item); }"

 *  bx-remove="function(type, item){ console.log('bx-remove'); console.log(type, item); }"

 *  bx-selected="function(type, state){ console.log('bx-selected'); console.log(type, state);}"
 *  bx-cleared="function(type, state){ console.log('bx-cleared'); console.log(type, state);}"
 *  bx-disabled="function(type, state){ console.log('bx-disabled'); console.log(type, state);}"
 *  bx-render="function(){ console.log('bx-render');}"
 *  >
 *      <bx-optgroup id="group1" name="group1" label="group1" disabled="false" selected="false">
 *      <bx-option id="option11" name="option11" label="option11" value="option11" selected="false" disabled="false" icon=""></bx-option>
 *      <bx-option id="option12" name="option12" label="option12" value="option12" selected="false" disabled="false" icon=""></bx-option>
 *      <bx-option id="option13" name="option13" label="option13" value="option13" selected="false" disabled="false" icon=""></bx-option>
 *      </bx-optgroup>
 *      <bx-optgroup id="group2" name="group2" label="group2" disabled="false" selected="false">
 *      <bx-option id="option21" name="option21" label="option21" value="option21" selected="false" disabled="false" icon=""></bx-option>
 *      <bx-option id="option22" name="option22" label="option22" value="option22" selected="false" disabled="false" icon=""></bx-option>
 *      <bx-option id="option23" name="option23" label="option23" value="option23" selected="false" disabled="false" icon=""></bx-option>
 *      </bx-optgroup>
 *      <bx-option id="option1" name="option1" label="option1" value="option1" selected="false" disabled="false" icon=""></bx-option>
 *      <bx-option id="option2" name="option2" label="option2" value="option2" selected="false" disabled="false" icon=""></bx-option>
 *
 *  </bx-select>
 *
 * <script type="text/javascript">
 *
 * $().ready(function(){
 *
 *    // BxSelect initialized
 *     $().BxSelect().initialize();
 *
 *     var elem = document.getElementById('test1');
 *
 *     console.log(elem.multiple);
 *     console.log(elem.bxOptGroups[0].selected);
 *     console.log(elem.bxOptGroups[0].bxOptions[0].label);
 *     console.log(elem.bxOptGroups[0].bxOptions[0].value);
 *     console.log(elem.bxOptions[0].selected);
 *
 *     elem.bxOptGroups[0].bxOptions[2].label = 'test';
 *     elem.bxOptGroups[0].bxOptions[2].value = 'test';
 *     elem.bxOptGroups[0].bxOptions[2].selected = true;
 *
 *     console.log(elem.bxOptGroups[0].bxOptions[2].label);
 *     console.log(elem.bxOptGroups[0].bxOptions[2].value);
 *
 *     elem.bxshow = function(){ alert('bx-show');};
 *
 *     elem.bxhide = function(){ alert('bx-hide');};
 * });
 *
 * </script>
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');
const Mousetrap = require('mousetrap');


const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;

// 전역 메세지 객체
const MESSAGE = {
    "notOptGroupType": '엘리먼트 타입이 `optGroup` 이 아닙니다',
    "notOptionType": '엘리먼트 타입이 `option` 이 아닙니다',
    "notOptGroupElement": '`optGroup` 엘리먼트가 아닙니다',
    "notMultiple": '해당 select 엘리먼트의 multiple 속성이 true 가 아닙니다',
    "notSupportDisabledProperty": '해당 select 엘리먼트는 `disabled` 속성이 제공되지 않습니다.',
    "notSupportValueProperty": '해당 select 엘리먼트는 `value` 속성이 제공되지 않습니다.',
    "noArguments": '필수 인자값이 전달되지 않았습니다.'
};

// 전역 클래스 객체
const CLASS_NAME = {
    "bxUiSelect": '-bx-ui-select',
    "bxUiInput": '-bx-ui-input',
    "bxSelectWrap": 'bx-select-wrap',
    "overflow": 'overflow',
    "bxSelectHeader": 'bx-select-header',
    "bxSelectList": 'bx-select-list',
    "bxSelectHeaderClick": 'bx-select-header-click',
    "option": 'option',
    "optionGroup": 'option-group',
    "optionValue": 'option-value',
    "icon": 'icon',
    "disabled": 'disabled',
    "overflowSelector": 'overflow-selector',
    "focus": 'focus',
    "selected": 'selected',
    "bxCheckbox": 'bx-checkbox',
    "text": 'text'
};

const ELEMENT_TYPE = {
    "select": 'select',
    "optGroup": 'optGroup',
    "option": 'option',
    "optGroupList": 'optGroupList'
};

// 엘리먼트 옵션
const OPTS = {

    select: {
        /**
         * @property {object} bx-select
         * @property {string} [bx-select.type=default] - <span class='important'>readonly</span> - 엘리먼트 타입(default, text)
         * @property {string} [bx-select.shape=default] - <span class='important'>readonly</span> - list 박스 타입(default, bubble)
         * @property {boolean} [bx-select.multiple=false] - <span class='important'>readonly</span> - 엘리먼트의 multiple 유/무
         * @property {string} [bx-select.icon=''] - 아이콘(<b class="dec">multiple 속성이 false 인 경우에만 동작한다</b>)
         * @property {string} [bx-select.label=''] - 선택된 option 엘리먼트의 label
         * @property {string} [bx-select.value=''] - 선택된 option 엘리먼트의 value(multiple 속성이 true 인 경우에는 값이 갱신되지 않는다)
         * @property {number} [bx-select.maxHeight=200] - <span class='important'>readonly</span> - 리스트 박스의 세로 높이
         * @property {boolean} [bx-select.disabled=false] - 엘리먼트의 disabled 유/무(<b class="dec">multiple 속성이 false 인 경우에만 동작한다</b>)
         * @property {function} [bx-select.bx-toggle=function(){}] - toggle 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-show=function(){}] - show 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-hide=function(){}] - hide 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-prepend=function(){}] - prepend 추가 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-add=function(){}] - add 추가 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-insertBefore=function(){}] - insertBefore 추가 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-insertAfter=function(){}] - insertAfter 추가 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-remove=function(){}] - remove 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-selected=function(){}] - selected 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-cleared=function(){}] - cleared 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-disabled=function(){}] - disabled 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-render=function(){}] - render 시, 호출될 callback 함수
         * @property {function} [bx-select.bx-[eventName]=function(){}] - 생성된 (div)엘리먼트에 해당 이벤트 핸들러를 바인딩시킨다.(ex. bx-click -> elem.onclick = function(){})
         *
         * @property {object} bx-optgroup
         * @property {string} [bx-optgroup.label=''] - label
         * @property {boolean} [bx-optgroup.selected=false] - 엘리먼트의 selected 유/무
         * @property {boolean} [bx-optgroup.disabled=false] - 엘리먼트의 disabled 유/무
         *
         * @property {object} bx-option
         * @property {string} [bx-option.label=''] - label
         * @property {string} [bx-option.value=''] - value
         * @property {string} [bx-option.icon=''] - 아이콘
         * @property {boolean} [bx-option.selected=false] - 엘리먼트의 selected 유/무
         * @property {boolean} [bx-option.disabled=false] - 엘리먼트의 disabled 유/무
         *
         *
         * @alias Attributes
         *
         * @memberof BxSelect
         */
        "type": 'default',
        "shape": 'default',
        "multiple": false,
        "disabled": false,
        "icon": '',
        "label": '',
        "value": '',
        "maxHeight": 200,

        /* jshint ignore:start */
        /**
         * @property {function} [bx-toggle = function(){}] - toggle 시, 호출될 callback 함수
         * @callback bxtoggle
         *
         * @param {string} state - 상태값
         *
         * @memberof BxSelect
         */
        "bx-toggle": function (){
        },
        /**
         * @property {function} [bx-show = function(){}] - show 시, 호출될 callback 함수
         * @callback bxshow
         * @memberof BxSelect
         */
        "bx-show": function (){
        },
        /**
         * @property {function} [bx-hide = function(){}] - hide 시, 호출될 callback 함수
         * @callback bxhide
         * @memberof BxSelect
         */
        "bx-hide": function (){
        },
        /**
         * @property {function} [bx-prepend = function(){}] - prepend 추가 시, 호출될 callback 함수
         * @callback bxprepend
         *
         * @param {string} type - 아이템 타입
         * @param {object|object[]} item - 아이템 또는 아이템 집합
         *
         * @memberof BxSelect
         */
        "bx-prepend": function (){
        },
        /**
         * @property {function} [bx-add = function(){}] - add 추가 시, 호출될 callback 함수
         * @callback bxadd
         *
         * @param {string} type - 아이템 타입
         * @param {object|object[]} item - 아이템 또는 아이템 집합
         *
         * @memberof BxSelect
         */
        "bx-add": function (){
        },
        /**
         * @property {function} [bx-insertBefore = function(){}] - insertBefore 추가 시, 호출될 callback 함수
         * @callback bxinsertBefore
         *
         * @param {string} type - 아이템 타입
         * @param {object|object[]} item - 아이템 또는 아이템 집합
         *
         * @memberof BxSelect
         */
        "bx-insertBefore": function (){
        },
        /**
         * @property {function} [bx-insertAfter = function(){}] - insertAfter 추가 시, 호출될 callback 함수
         * @callback bxinsertAfter
         *
         * @param {string} type - 아이템 타입
         * @param {object|object[]} item - 아이템 또는 아이템 집합
         *
         * @memberof BxSelect
         */
        "bx-insertAfter": function (){
        },
        /**
         * @property {function} [bx-remove = function(){}] - remove 시, 호출될 callback 함수
         * @callback bxremove
         *
         * @param {string} type - 추가된 아이템 타입
         * @param {object|object[]} item - 아이템 집합
         *
         * @memberof BxSelect
         */
        "bx-remove": function (){
        },
        /**
         * @property {function} [bx-selected = function(){}] - selected 시, 호출될 callback 함수
         * @callback bxselected
         *
         * @param {string} type - 선택된 아이템 타입
         * @param {boolean} state - 선택 유/무
         *
         * @memberof BxSelect
         */
        "bx-selected": function (){
        },
        /**
         * @property {function} [bx-cleared = function(){}] - cleared 시, 호출될 callback 함수
         * @callback bxcleared
         *
         * @param {string} type - 선택된 아이템 타입
         * @param {boolean} state - 선택 유/무
         *
         * @memberof BxSelect
         */
        "bx-cleared": function (){
        },
        /**
         * @property {function} [bx-disabled = function(){}] - disabled 시, 호출될 callback 함수
         * @callback bxdisabled
         *
         * @param {string} type - 선택된 아이템 타입
         * @param {boolean} state - disabled 유/무
         *
         * @memberof BxSelect
         */
        "bx-disabled": function (){
        },

        /**
         * @property {function} [bx-render = function(){}] - render 시, 호출될 callback 함수
         * @callback bxrender
         *
         * @memberof BxSelect
         */
        "bx-render": function (){
        }
        /* jshint ignore:end */
    },
    optGroup: {
        "label": '',
        "disabled": false,
        "value": '',
        /*
         ------------------------------------------------
         ------------------------------------------------
         옵션 그룹의 `selected` 는 총 3 가지의 상태값을 가진다.
         ------------------------------------------------
         ------------------------------------------------
         true: 선택되었을 경우
         false: 선택되지 않았을 경우
         undetermined: 그룹 옵션 중 하나라도 선택이 해제 되었을 경우
         ------------------------------------------------
         */
        "selected": false
    },
    option: {
        "label": '',
        "value": '',
        "icon": '',
        "disabled": false,
        "selected": false
    }
};

// 사용자에게 제공되는 API 목록
const APIS = {

    addGroup: _addGroup,
    prependGroup: _prependGroup,
    insertBeforeGroup: _insertBeforeGroup,
    insertAfterGroup: _insertAfterGroup,

    addOptions: _addOptions,
    prependOptions: _prependOptions,
    insertBeforeOptions: _insertBeforeOptions,
    insertAfterOptions: _insertAfterOptions,

    getSelectedGroups: _getSelectedGroups,
    selectGroups: _selectGroups,
    clearSelectedGroups: _clearSelectedGroups,

    getSelectedOptions: _getSelectedOptions,
    selectOptions: _selectOptions,
    clearSelectedOptions: _clearSelectedOptions,

    removeGroups: _removeGroups,
    removeOptions: _removeOptions,

    toggle: _toggle,
    show: _show,
    hide: _hide
};

// `IE9` 대응을 위한, attr 이름 목록
const ATTRNAMES = {
    "bx-insertafter": "bx-insertAfter",
    "bx-insertbefore": "bx-insertBefore",
    "maxheight": "maxHeight"
};

// 내부 아이디 생성을 위한 prefixe 문자열
const prefixId = 'bxUiSelect';

// 마지막으로 생성된 인스턴스 객체(mutiple 속성이 false 인 경우에만 마지막 엘리먼트를 저장한다)
let lastShowInstance = null;

// 위 / 아래 키를통해, 옵션 이동 시, browser scroll 막은 후, 다시 이전값을 돌려주기위해 사용하는 캐시값
const LAGACYS = {};

class BxSelect {

    constructor(){
        this.$root = BxSelect.$root;
        this.opts = OPTS.select;
    }

    /**
     * 초기화 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    initialize(){

        this.$selects = this.$root.find('bx-select');
        this.elems = [];

        $.each(this.$selects, (i, elem) =>{

            const $customElem = $(elem);

            // 사용자 커스텀 엘리먼트를 파싱한다.
            const $component = _parseCustomElement($customElem);
            const _this = $component.get(0);

            $component.prop('bxrender').call(_this);
            this.elems.push($component.get(0));
        });

        _addDocumentElementEventListener();

        LAGACYS.overflow = $(document.body).css('overflow');

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxSelect.isCustomElement = true;

//TODO PUBLIC 메서드 영역
/*
 ==================================================================================================================================
 ==================================================================================================================================

 PUBLIC 메서드 영역

 ==================================================================================================================================
 ==================================================================================================================================
 */
/**
 *
 * 새로운 그룹을 가장 뒤에 추가한다.
 *
 * @param {object} groupItem - 그룹 아이템
 * @param {string} groupItem.label - label
 * @param {boolean} groupItem.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [groupItem.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} groupItem.attr - attr 객체
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias addGroup
 *
 * @memberOf BxSelect
 */
function _addGroup(groupItem = {}, optionItems = []){

    if (!_.isPlainObject(groupItem)){
        CONSOLE.prop.throw(BxSelect, 'groupItem', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = _createOptGroupElement(groupItem, $selectElem);
    let optGroupElem = $optGroupElem.get(0);
    let $optGroupUlElem = $optGroupElem.children('div');

    // 엘리먼트의 disabled 를 설정한다.
    _setDisabled($optGroupElem);

    // 옵션 그룹을 추가한다.
    _addElementItem(this.bxOptGroups, optGroupElem);

    $overflowElem.append($optGroupElem);

    UTIL.setReadOnlyProperty(optGroupElem, 'bxOptions', []);

    _addOptionElementToItems($optGroupUlElem, optGroupElem.bxOptions, optionItems);

    // list 컨테이너 엘리먼트에 `group` class 를 추가한다.
    _toggleGroupClassName($selectElem);

    // overflow 컨테이너 엘리먼트에 class 를 추가 또는 삭제한다.
    _toggleOverFlowClassNameWithChildrenSize($selectElem);

    // 엘리먼트의 selected 를 설정한다.
    _setSelected($optGroupElem);

    this['bxadd']('group', optGroupElem);

    return this;
}

/**
 *
 * 새로운 그룹을 가장 앞에 추가한다.
 *
 * @param {object} groupItem - 그룹 아이템
 * @param {string} groupItem.label - label
 * @param {boolean} groupItem.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [groupItem.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} groupItem.attr - attr 객체
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias prependGroup
 *
 * @memberOf BxSelect
 */
function _prependGroup(groupItem = {}, optionItems = []){

    if (!_.isPlainObject(groupItem)){
        CONSOLE.prop.throw(BxSelect, 'groupItem', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = _createOptGroupElement(groupItem, $selectElem);
    let optGroupElem = $optGroupElem.get(0);
    let $optGroupUlElem = $optGroupElem.children('div');

    // 엘리먼트의 disabled 를 설정한다.
    _setDisabled($optGroupElem);

    // 옵션 그룹을 추가한다.
    prependElementItem(this.bxOptGroups, optGroupElem);

    $overflowElem.prepend($optGroupElem);

    UTIL.setReadOnlyProperty(optGroupElem, 'bxOptions', []);

    _addOptionElementToItems($optGroupUlElem, optGroupElem.bxOptions, optionItems);

    // list 컨테이너 엘리먼트에 `group` class 를 추가한다.
    _toggleGroupClassName($selectElem);

    // overflow 컨테이너 엘리먼트에 class 를 추가 또는 삭제한다.
    _toggleOverFlowClassNameWithChildrenSize($selectElem);

    // 엘리먼트의 selected 를 설정한다.
    _setSelected($optGroupElem);

    this['bxprepend']('group', optGroupElem);

    return this;
}

/**
 *
 * 새로운 그룹을 정해진 그룹 이전에 추가한다.
 *
 * @param {string} groupId - 그룹 아이디
 *
 * @param {object} groupItem - 그룹 아이템
 * @param {string} groupItem.label - label
 * @param {boolean} groupItem.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [groupItem.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} groupItem.attr - attr 객체
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias insertBeforeGroup
 *
 * @memberOf BxSelect
 */
function _insertBeforeGroup(groupId = '', groupItem = {}, optionItems = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isPlainObject(groupItem)){
        CONSOLE.prop.throw(BxSelect, 'groupItem', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $selectedOptGroupElem = $overflowElem.find(`#${groupId}`);
    let selectedGroupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($selectedOptGroupElem);

    if (selectedGroupIndex === -1) return this;

    let $optGroupElem = _createOptGroupElement(groupItem, $selectElem);
    let optGroupElem = $optGroupElem.get(0);
    let $optGroupUlElem = $optGroupElem.children('div');

    // 엘리먼트의 disabled 를 설정한다.
    _setDisabled($optGroupElem);

    // 옵션 그룹을 추가한다.
    _insertBeforeElementItem(this.bxOptGroups, optGroupElem, (selectedGroupIndex));

    $optGroupElem.insertBefore($selectedOptGroupElem);

    UTIL.setReadOnlyProperty(optGroupElem, 'bxOptions', []);

    _addOptionElementToItems($optGroupUlElem, optGroupElem.bxOptions, optionItems);

    // 엘리먼트의 selected 를 설정한다.
    _setSelected($optGroupElem);

    this['bxinsertBefore']('group', optGroupElem);

    return this;
}

/**
 *
 * 새로운 그룹을 정해진 그룹 이후에 추가한다.
 *
 * @param {string} groupId - 그룹 아이디
 *
 * @param {object} groupItem - 그룹 아이템
 * @param {string} groupItem.label - label
 * @param {boolean} groupItem.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [groupItem.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} groupItem.attr - attr 객체
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias insertAfterGroup
 *
 * @memberOf BxSelect
 */
function _insertAfterGroup(groupId = '', groupItem = {}, optionItems = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isPlainObject(groupItem)){
        CONSOLE.prop.throw(BxSelect, 'groupItem', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'options', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $selectedOptGroupElem = $overflowElem.find(`#${groupId}`);
    let selectedGroupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($selectedOptGroupElem);

    if (selectedGroupIndex === -1) return this;

    let $optGroupElem = _createOptGroupElement(groupItem, $selectElem);
    let optGroupElem = $optGroupElem.get(0);
    let $optGroupUlElem = $optGroupElem.children('div');

    // 엘리먼트의 disabled 를 설정한다.
    _setDisabled($optGroupElem);

    // 옵션 그룹을 추가한다.
    _insertAfterElementItem(this.bxOptGroups, optGroupElem, selectedGroupIndex);

    $optGroupElem.insertAfter($selectedOptGroupElem);

    UTIL.setReadOnlyProperty(optGroupElem, 'bxOptions', []);

    _addOptionElementToItems($optGroupUlElem, optGroupElem.bxOptions, optionItems);

    // 엘리먼트의 selected 를 설정한다.
    _setSelected($optGroupElem);

    this['bxinsertAfter']('group', optGroupElem);

    return this;
}

/**
 *
 * 새로운 옵션을 (그룹의)가장 뒤에 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디(만약 그룹 아이디가 있는 경우에는 해당 그룹의 옵션으로 추가된다)
 *
 * @param {object|object[]} optionItems - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias addOptions
 *
 * @memberOf BxSelect
 */
function _addOptions(...args){

    let optionItems = [];

    if (args.length < 1){
        CONSOLE.prop.throw(BxSelect, MESSAGE.noArguments);
    }

    if (args.length === 1){

        optionItems = args[0];

        if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
            CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
        }
    }
    else if (args.length === 2){
        return _addGroupOptions.call(this, args[0], args[1]);
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptions;

    _addOptionElementToItems($overflowElem, bxOptions, optionItems, false);

    // overflow 컨테이너 엘리먼트에 class 를 추가 또는 삭제한다.
    _toggleOverFlowClassNameWithChildrenSize($selectElem);

    this['bxadd'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 새로운 옵션을 (그룹의)가장 앞에 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디(만약 그룹 아이디가 있는 경우에는 해당 그룹의 옵션으로 추가된다)
 *
 * @param {object|object[]} optionItems - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias prependOptions
 *
 * @memberOf BxSelect
 */
function _prependOptions(...args){

    let optionItems = [];

    if (args.length < 1){
        CONSOLE.prop.throw(BxSelect, MESSAGE.noArguments);
    }

    if (args.length === 1){

        optionItems = args[0];

        if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
            CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
        }
    }
    else if (args.length === 2){
        return _prependGroupOptions.call(this, args[0], args[1]);
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptions;

    _prependOptionElementToItems($overflowElem, bxOptions, optionItems, false);

    // overflow 컨테이너 엘리먼트에 class 를 추가 또는 삭제한다.
    _toggleOverFlowClassNameWithChildrenSize($selectElem);

    this['bxprepend'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 새로운 옵션을 정해진 (그룹)옵션 이전에 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디(만약 그룹 아이디가 있는 경우에는 해당 그룹의 옵션으로 추가된다)
 *
 * @param {string} optionId - 옵션 아이디
 *
 * @param {object|object[]} optionItems - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias insertBeforeOptions
 *
 * @memberOf BxSelect
 */
function _insertBeforeOptions(...args){

    let optionId = '';
    let optionItems = [];

    if (args.length < 2){
        CONSOLE.prop.throw(BxSelect, MESSAGE.noArguments);
    }

    if (args.length === 2){

        optionId = args[0];
        optionItems = args[1];

        if (!_.isString(optionId) || _.trim(optionId).length === 0){
            CONSOLE.prop.throw(BxSelect, 'optionId', 'type');
        }

        if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
            CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
        }
    }
    else if (args.length === 3){
        return _insertBeforeGroupOptions.call(this, args[0], args[1], args[2]);
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optionElem = $overflowElem.find(`#${optionId}`);
    let optionIndex = $overflowElem.find('div[data-element-type="option"]').index($optionElem);

    if (optionIndex === -1) return this;

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptions;

    _insertBeforeOptionElementToItems($optionElem, bxOptions, optionItems, optionIndex, false);

    this['bxinsertBefore'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 새로운 옵션을 정해진 (그룹)옵션 이후에 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디(만약 그룹 아이디가 있는 경우에는 해당 그룹의 옵션으로 추가된다)
 *
 * @param {string} optionId - 옵션 아이디
 *
 * @param {object|object[]} optionItems - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias insertAfterOptions
 *
 * @memberOf BxSelect
 */
function _insertAfterOptions(...args){

    let optionId = '';
    let optionItems = [];

    if (args.length < 2){
        CONSOLE.prop.throw(BxSelect, MESSAGE.noArguments);
    }

    if (args.length === 2){

        optionId = args[0];
        optionItems = args[1];

        if (!_.isString(optionId) || _.trim(optionId).length === 0){
            CONSOLE.prop.throw(BxSelect, 'optionId', 'type');
        }

        if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
            CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
        }
    }
    else if (args.length === 3){
        return _insertAfterGroupOptions.call(this, args[0], args[1], args[2]);
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optionElem = $overflowElem.find(`#${optionId}`);
    let optionIndex = $overflowElem.find('div[data-element-type="option"]').index($optionElem);

    if (optionIndex === -1) return this;

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptions;

    _insertAfterOptionElementToItems($optionElem, bxOptions, optionItems, optionIndex, false);

    this['bxinsertAfter'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 선택된 그룹을 가져온다.
 *
 * @returns {array} 선택된 그룹
 *
 * @alias getSelectedGroups
 *
 * @memberOf BxSelect
 */
function _getSelectedGroups(){
    return _.filter(this.bxOptGroups, {"selected": true});
}

/**
 *
 * 그룹을 선택한다.
 *
 * @param {empty|string|string[]} [groupIds=[]] - 그룹 아이디 (전달값이 없을 경우 모든 그룹이 선택된다)
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias selectGroups
 *
 * @memberOf BxSelect
 */
function _selectGroups(groupIds = []){

    if (!_.isString(groupIds) && !_.isArray(groupIds)){
        CONSOLE.prop.throw(BxSelect, 'groupIds', 'type');
    }

    if (_.isString(groupIds) && _.trim(groupIds).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupIds', 'type');
    }

    groupIds = _.isString(groupIds) ? [groupIds] : groupIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    if (!groupIds.length){
        groupIds = _getElementIds($selectElem.prop('bxOptGroups'));
    }

    _updateElementToItems($overflowElem, groupIds, $optGroupElem =>{

        let elementType = $optGroupElem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.optGroup) return;

        $optGroupElem.prop('selected', true);

        let $children = $optGroupElem.children();

        _.map($children, optionElem =>{

            let elementType = $(optionElem).attr('data-element-type');

            if (elementType !== ELEMENT_TYPE.option) return;

            optionElem['selected'] = true;
        });
    });

    return this;
}

/**
 *
 * 선택된 그룹을 해제한다.
 *
 * @param {empty|string|string[]} [groupIds=[]] - 그룹 아이디 (전달값이 없을 경우 모든 그룹이 선택 해제된다)
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias clearSelectedGroups
 *
 * @memberOf BxSelect
 */
function _clearSelectedGroups(groupIds = []){

    if (!_.isString(groupIds) && !_.isArray(groupIds)){
        CONSOLE.prop.throw(BxSelect, 'groupIds', 'type');
    }

    if (_.isString(groupIds) && _.trim(groupIds).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupIds', 'type');
    }

    groupIds = _.isString(groupIds) ? [groupIds] : groupIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    if (!groupIds.length){
        groupIds = _getElementIds($selectElem.prop('bxOptGroups'));
    }

    _updateElementToItems($overflowElem, groupIds, $optGroupElem =>{

        let elementType = $optGroupElem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.optGroup) return;

        $optGroupElem.prop('selected', false);

        let $children = $optGroupElem.children();

        _.map($children, optionElem =>{

            let elementType = $(optionElem).attr('data-element-type');

            if (elementType !== ELEMENT_TYPE.option) return;

            optionElem['selected'] = false;
        });
    });

    return this;
}

/**
 *
 * 선택된 옵션을 가져온다.
 *
 * @param {string} [groupId=''] - 그룹 아이디(만약 그룹 아이디가 있는 경우에는 해당 그룹의 선택된 옵션이 반환된다)
 *
 * @returns {array} 선택된 옵션
 *
 * @alias getSelectedOptions
 *
 * @memberOf BxSelect
 */
function _getSelectedOptions(...args){

    if (args.length === 0){
        return _getSelectedAllNodes.call(this);
    }
    else if (args.length === 1){
        return _getSelectedGroupOptions.call(this, args[0]);
    }
}

/**
 *
 * 옵션을 선택한다.
 *
 * @param {empty|string|string[]} [optionIds=[]] - 옵션 아이디 (전달값이 없을 경우 모든 옵션이 선택된다)
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias selectOptions
 *
 * @memberOf BxSelect
 */
function _selectOptions(optionIds = []){

    if (!_.isString(optionIds) && !_.isArray(optionIds)){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    if (_.isString(optionIds) && _.trim(optionIds).length === 0){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    optionIds = _.isString(optionIds) ? [optionIds] : optionIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let bxOptGroups = $selectElem.prop('bxOptGroups');
    let bxOptions = $selectElem.prop('bxOptions');

    if (!optionIds.length){
        return _selectAllNodes.call(this);
    }

    _.map(bxOptGroups, v => _selectGroupOptions.call(this, $(v).attr('id'), optionIds));

    _updateElementToItems($overflowElem, optionIds, $optionElem =>{

        let elementType = $optionElem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.option) return;

        $optionElem.prop('selected', true);
    });

    return this;
}

/**
 *
 * 선택된 옵션을 해제한다.
 *
 * @param {empty|string|string[]} [optionIds=[]] - 옵션 아이디 (전달값이 없을 경우 모든 옵션이 선택 해제된다)
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias clearSelectedOptions
 *
 * @memberOf BxSelect
 */
function _clearSelectedOptions(optionIds = []){

    if (!_.isString(optionIds) && !_.isArray(optionIds)){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    if (_.isString(optionIds) && _.trim(optionIds).length === 0){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    optionIds = _.isString(optionIds) ? [optionIds] : optionIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let bxOptGroups = $selectElem.prop('bxOptGroups');
    let bxOptions = $selectElem.prop('bxOptions');

    if (!optionIds.length){
        return _clearSelectedAllNodes.call(this);
    }

    _.map(bxOptGroups, v => _clearSelectedGroupOptions.call(this, $(v).attr('id'), optionIds));

    _updateElementToItems($overflowElem, optionIds, $optionElem =>{

        let elementType = $optionElem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.option) return;

        $optionElem.prop('selected', false);
    });

    return this;
}

/**
 *
 * 선택된 그룹을 삭제한다.
 *
 * @param {empty|string|string[]} [groupIds=[]] - 그룹 아이디 (전달값이 없을 경우 모든 그룹이 삭제된다)
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias removeGroups
 *
 * @memberOf BxSelect
 */
function _removeGroups(groupIds = []){

    if (!_.isString(groupIds) && !_.isArray(groupIds)){
        CONSOLE.prop.throw(BxSelect, 'groupIds', 'type');
    }

    if (_.isString(groupIds) && _.trim(groupIds).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupIds', 'type');
    }

    groupIds = _.isString(groupIds) ? [groupIds] : groupIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let bxOptGroups = this.bxOptGroups;

    if (!groupIds.length){
        groupIds = _getElementIds(bxOptGroups);
    }

    // 삭제될 옵션이 현재 선택된 옵션일 경우, select 엘리먼트의 `label` 및 `value` 값을 기본값으로 변경 시킨다.
    _.map(groupIds, v =>{

        let $optGroupElem = $overflowElem.find(`#${v}`);

        _.map($optGroupElem.prop('bxOptions'), vv =>{

            if (!$selectElem.prop('multiple') && $(vv).prop('selected')){
                _initSelectElement($selectElem);
            }
        });
    });

    _removeOptGroupElementToItems($overflowElem, bxOptGroups, groupIds);

    _toggleGroupClassName($selectElem);
    _toggleOverFlowClassNameWithChildrenSize($selectElem);

    this['bxremove']('group', bxOptGroups);

    return this;
}

/**
 *
 * 옵션을 삭제한다.
 *
 * @param {empty|string|string[]} [optionIds=[]] - 옵션 아이디 (전달값이 없을 경우 모든 그룹 및 옵션이 삭제된다)
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias removeOptions
 *
 * @memberOf BxSelect
 */
function _removeOptions(optionIds = []){

    if (!_.isString(optionIds) && !_.isArray(optionIds)){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    if (_.isString(optionIds) && _.trim(optionIds).length === 0){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    optionIds = _.isString(optionIds) ? [optionIds] : optionIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let bxOptGroups = this.bxOptGroups;
    let bxOptions = this.bxOptions;

    if (!optionIds.length){

        optionIds = _getElementIds(bxOptions);

        _.map(bxOptGroups, v =>{
            _.map(v['bxOptions'], vv => optionIds.push($(vv).attr('id')));
        });
    }

    _.map(optionIds, v =>{

        let $optionElem = $overflowElem.find(`#${v}`);

        let $optGroupElem = $($optionElem.parents(`.${CLASS_NAME.optionGroup}`).get(0));
        let id = $optGroupElem.attr('id');

        // option 엘리먼트의 부모가 optGroup 엘리먼트일 경우
        if ($optGroupElem.get(0)) _removeGroupOptions.call(this, id, [v]);
        else _removeOptionElementToItems($overflowElem, bxOptions, [v]);

        // 삭제될 옵션이 현재 선택된 옵션일 경우, select 엘리먼트의 `label` 및 `value` 값을 기본값으로 변경 시킨다.
        if (!$selectElem.prop('multiple') && $optionElem.prop('selected')){
            _initSelectElement($selectElem);
        }
    });

    _toggleOverFlowClassNameWithChildrenSize($selectElem);

    this['bxremove'](ELEMENT_TYPE.option, _getAllNodes.call(this));

    return this;
}

/**
 *
 * show 함수
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias show
 *
 * @memberOf BxSelect
 */
function _show(){

    let $selectElem = $(this);
    let $bxSelectListElem = $selectElem.find(`.${CLASS_NAME.bxSelectList}`);
    let $bxSelectHeaderElem = $selectElem.find(`.${CLASS_NAME.bxSelectHeader}`);

    let multiple = $selectElem.prop('multiple');
    let disabled = $selectElem.prop('disabled');

    if (!multiple && disabled) return this;

    window.setTimeout(() =>{

        $bxSelectListElem.css('display', 'block');
        $bxSelectHeaderElem.addClass(CLASS_NAME.bxSelectHeaderClick);

        if (!multiple) lastShowInstance = this;

        this['bxshow']();

    }, 30);

    return this;
}

/**
 *
 * hide 함수
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias hide
 *
 * @memberOf BxSelect
 */
function _hide(){

    let $selectElem = $(this);
    let $bxSelectListElem = $selectElem.find(`.${CLASS_NAME.bxSelectList}`);
    let $bxSelectHeaderElem = $selectElem.find(`.${CLASS_NAME.bxSelectHeader}`);
    let multiple = $selectElem.prop('multiple');

    window.setTimeout(() =>{

        $bxSelectListElem.css('display', 'none');
        $bxSelectHeaderElem.removeClass(CLASS_NAME.bxSelectHeaderClick);

        if (!multiple) lastShowInstance = null;

        this['bxhide']();

    }, 1);

    return this;
}

/**
 *
 * toggle 함수
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias toggle
 *
 * @memberOf BxSelect
 */
function _toggle(){

    let handler = this['bxtoggle'];

    if (_isShow.call(this)){
        _hide.call(this);
        handler('hide');
    }
    else{
        _show.call(this);
        handler('show');
    }

    return this;
}

//TODO PRIVATE 메서드 영역
/*
 ==================================================================================================================================
 ==================================================================================================================================

 PRIVATE 메서드 영역

 ==================================================================================================================================
 ==================================================================================================================================
 */


/**
 * 사용자 `custom` 태그를 `native` 태그로 대치시킨다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - 사용자 custom 엘리먼트
 *
 * @returns {object} 생성된 $container 엘리먼트
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _parseCustomElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$customElem', 'type');
    }

    // $container 엘리먼트를 생성한다.
    let $container = BASE.appendContainerElement(BxSelect.$root, BxSelect.className);

    // 컴포넌트 엘리먼트를 생성한다.
    let $component = _getComponentElement($customElem);

    // 반환받은 `컴포넌트 엘리먼트`를 `컨테이너 엘리먼트`의 자식으로 추가한다.
    $container.append($component);

    // $customElem 엘리먼트를 생성된 `$container` 엘리먼트로 대치시킨다.
    $customElem.replaceWith($container);

    return $component;
}


/**
 * component 엘리먼트를 반환한다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - 사용자 custom 엘리먼트
 *
 * @returns {object} 생성된 component 엘리먼트
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _getComponentElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$customElem', 'type');
    }

    let selectType = $customElem.attr('type');

    switch (selectType){
        case 'default' :
            return _createSelectElement($customElem);
        default :
            return _createSelectElement($customElem);
    }
}

/**
 * select 엘리먼트를 생성한다.
 *
 * @param {object} [$customElem=$(document.documentElement)] - `custom` 엘리먼트
 *
 * @returns {object} select 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _createSelectElement($customElem = $(document.documentElement)){

    if (!_.isElement($customElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$customElem', 'type');
    }

    // `bx-select` 엘리먼트에 할당된 옵션을 반환한다.
    let opts = _createComponentOptions($customElem, OPTS.select);

    let _attr = opts.attr;

    let selectClass = _attr.class ? ' ' + _attr.class : '';

    if (selectClass) delete _attr.class;

    let attr = UTIL.objectToStringAttr(_attr);

    let type = opts.type;
    let shape = opts.shape;
    let icon = opts.icon;
    let label = opts.label;
    let disabled = opts.disabled;
    let multiple = opts.multiple;
    let maxHeight = opts.maxHeight;

    let multipleClassName = multiple ? ' multiple' : '';

    if (type === 'text') type = 'text-type';

    let $selectElem = $(`<div ${attr} data-element-type="${ELEMENT_TYPE.select}" class="${CLASS_NAME.bxSelectWrap}${selectClass}${multipleClassName}"></div>`);

    let selectElem = $selectElem.get(0);

    // 할당된 옵션값을 통해 엘리먼트 프로퍼티를 생성한다.
    _setSelectElementProperties(opts, selectElem);

    // 최초 설정한 label 값을 새로운 프로퍼티로 남겨둔다.
    UTIL.setReadOnlyProperty(selectElem, 'defaultLabel', label);

    let topHtml = [];

    if (!multiple){

        topHtml.push(`<div class="${CLASS_NAME.bxSelectHeader} ${type}">`);
        topHtml.push('<span class="wrap">');

        topHtml.push(`<span class="icon"></span>`);
        topHtml.push(`<a href="#" class="text" onclick="return false;" title="${label}">${label}</a>`);
        topHtml.push('</span>');

        topHtml.push('<i class="bx-ctrl-icon i-15 i-ctrl-down"></i>');
        topHtml.push('</div>');
    }

    topHtml.push(`<div class="${CLASS_NAME.bxSelectList} ${shape}">`);
    // firefox focus 이슈 대응으로 tabindex 사용
    topHtml.push(`<div class="${CLASS_NAME.overflow} ${CLASS_NAME.overflowSelector}" tabindex="-1" style="max-height:${maxHeight}px"></div>`);
    topHtml.push('</div>');

    $selectElem.append(topHtml.join(''));

    _addSelectElementEventListener($selectElem, opts);
    // select 엘리먼트의 아이콘을 설정한다.
    _setIconWithSelectElement($selectElem);

    // 엘리먼트의 disabled 를 설정한다.
    _setDisabled($selectElem);

    UTIL.setReadOnlyProperty(selectElem, 'bxOptGroups', []);
    UTIL.setReadOnlyProperty(selectElem, 'bxOptions', []);

    // 커스텀 엘리먼트의 자식 엘리먼트들
    let $children = $customElem.children();

    let groupNodeName = 'BX-OPTGROUP';
    let optionNodeName = 'BX-OPTION';

    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    _.map($children, v =>{

        // 자식 엘리먼트
        let $child = $(v);

        // 그룹 엘리먼트일 경우
        if (v.nodeName === groupNodeName){

            let $optGroupElem = _createOptGroupElement($child, $selectElem);

            let $optGroupUlElem = $optGroupElem.children('div');

            let $children = $child.children();

            // group 엘리먼트를 추가한다.
            _addElementItem($selectElem.prop('bxOptGroups'), $optGroupElem.get(0));

            // 생성된 group 컬렉션에 option 컬렉션을 추가한다.
            UTIL.setReadOnlyProperty($optGroupElem.get(0), 'bxOptions', []);

            _.map($children, v =>{

                let $child = $(v);

                // 자식 엘리먼트의 nodeName 이 bx-option 일 경우
                if (v.nodeName === optionNodeName){

                    let $elem = _createOptionElementIntoGroupElement($child, $selectElem);
                    let $optionElem = _addElement($optGroupUlElem, $elem);

                    _addElementItem($optGroupElem.prop('bxOptions'), $optionElem.get(0));
                }
            });

            // group 엘리먼트를 추가한다.
            $overflowElem.append($optGroupElem);

            // 엘리먼트의 selected 를 설정한다.
            _setSelectedWithOptGroupElement($optGroupElem);

            // 엘리먼트의 disabled 를 설정한다.
            _setDisabledWithOptGroupElement($optGroupElem);

            _toggleGroupClassName($selectElem);
        }

        if (v.nodeName === optionNodeName){

            let $elem = _createOptionElementIntoOverFlowElement($child, $selectElem);

            let $optionElem = _addElement($overflowElem, $elem);

            _addElementItem($selectElem.prop('bxOptions'), $optionElem.get(0));

            // 엘리먼트의 disabled 를 설정한다.
            _setSelected($optionElem);

            // 엘리먼트의 disabled 를 설정한다.
            _setDisabled($optionElem);
        }
    });

    // bx-* 패턴으로 할당된 모든 네이티브 이벤트를 바인딩 시킨다.
    UTIL.addAllEventListenerWithProperty($selectElem, opts.events);

    return $selectElem;
}

/**
 * optGroup 엘리먼트를 생성한다.
 *
 * @param {object} [targetObject=null]
 * @param {object} [$selectElem=null] - select 엘리먼트 객체
 *
 * @returns {object} optGroup 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _createOptGroupElement(targetObject = null, $selectElem = null){

    if (!_.isObject(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let opts = _createComponentOptions(targetObject, OPTS.optGroup);

    let label = opts.label;

    let _attr = opts.attr;

    if (!_attr.id) _attr.id = _.uniqueId(prefixId + '_' + 'optGroup_');

    let optGroupClass = _attr.class ? ' ' + _attr.class : '';

    if (optGroupClass) delete _attr.class;

    let attr = UTIL.objectToStringAttr(_attr);

    let multiple = $selectElem.prop('multiple');

    let extendOpts = _.extend(opts, {multiple});

    let html = [];

    html.push(`<div ${attr} data-element-type="${ELEMENT_TYPE.optGroup}" class="${CLASS_NAME.optionGroup}${optGroupClass}">`);
    html.push(`<h3>`);

    if (multiple){
        html.push(`<bx-input type="checkbox" title="${label}"></bx-input>`);
    }

    html.push(`<span class="${CLASS_NAME.optionValue}"></span>`);
    html.push(`</h3>`);
    html.push(`<div></div>`);
    html.push(`</div>`);

    let $elem = $(html.join(''));

    multiple && $elem.bxInput().initialize();

    _setOptGroupElementProperties(opts, $elem.get(0));
    _addOptGroupElementEventListener($elem, extendOpts);

    $elem.prop('selected') && setTimeout(()=>{_setSelected($elem);});
    // 엘리먼트의 label 을 설정한다.
    _setLabel($elem);

    return $elem;
}

/**
 * overFlow 엘리먼트 아래에 option 엘리먼트를 생성한다.
 *
 * @param {object} [targetObject=null]
 * @param {object} [$selectElem=null] - select 엘리먼트 객체
 *
 * @returns {object} option 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _createOptionElementIntoOverFlowElement(targetObject = null, $selectElem = null){

    if (!_.isObject(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let opts = _createComponentOptions(targetObject, OPTS.option);

    let label = opts.label;

    let _attr = opts.attr;

    if (!_attr.id) _attr.id = _.uniqueId(prefixId + '_' + 'option_');

    let optionClass = _attr.class ? ' ' + _attr.class : '';

    if (optionClass) delete _attr.class;

    let attr = UTIL.objectToStringAttr(_attr);

    let multiple = $selectElem.prop('multiple');

    let extendOpts = _.extend(opts, {multiple});

    let html = [];

    html.push(`<div ${attr} data-element-type="${ELEMENT_TYPE.option}" class="${CLASS_NAME.option}${optionClass}">`);

    if (multiple){
        html.push(`<bx-input type="checkbox" title="${label}"></bx-input>`);
    }

    html.push(`<span class="icon"></span>`);

    html.push(`<span>`);

    if (multiple){
        html.push(`<p class="${CLASS_NAME.optionValue}"></p>`);
    }
    else{
        html.push(`<a href="#" onclick="return false;" class="${CLASS_NAME.optionValue}" title="${label}"></a>`);
    }

    html.push(`</span>`);

    html.push(`</div>`);

    let $elem = $(html.join(''));

    multiple && $elem.bxInput().initialize();

    _setOptionElementProperties(opts, $elem.get(0));
    _addOptionElementEventListener($elem, extendOpts, $selectElem);

    // 엘리먼트의 아이콘을 설정한다.
    _setIconWithOptionElement($elem);

    // 엘리먼트의 label 을 설정한다.
    _setLabel($elem);

    return $elem;
}

/**
 * optGroup 엘리먼트 아래에 option 엘리먼트를 생성한다.
 *
 * @param {object} [targetObject=null]
 * @param {object} [$selectElem=null] - select 엘리먼트 객체
 *
 * @returns {object} option 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _createOptionElementIntoGroupElement(targetObject = null, $selectElem = null){

    if (!_.isObject(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let opts = _createComponentOptions(targetObject, OPTS.option);

    let label = opts.label;

    let _attr = opts.attr;

    if (!_attr.id) _attr.id = _.uniqueId(prefixId + '_' + 'option_');

    let optionClass = _attr.class ? ' ' + _attr.class : '';

    if (optionClass) delete _attr.class;

    let attr = UTIL.objectToStringAttr(_attr);

    let multiple = $selectElem.prop('multiple');

    let extendOpts = _.extend(opts, {multiple});

    let html = [];

    html.push(`<div ${attr} data-element-type="${ELEMENT_TYPE.option}" class="${CLASS_NAME.option}${optionClass}">`);

    if (multiple){
        html.push(`<bx-input type="checkbox" title="${label}"></bx-input>`);
    }

    html.push(`<span class="icon"></span>`);

    html.push(`<span>`);

    if (multiple){
        html.push(`<p class="${CLASS_NAME.optionValue}"></p>`);
    }
    else{
        html.push(`<a href="#" onclick="return false;" class="${CLASS_NAME.optionValue}" title="${label}"></a>`);
    }

    html.push(`</span>`);

    html.push(`</div>`);

    let $elem = $(html.join(''));

    multiple && $elem.bxInput().initialize();

    _setOptionElementProperties(opts, $elem.get(0));
    _addOptionElementEventListener($elem, extendOpts, $selectElem);

    // 엘리먼트의 아이콘을 설정한다.
    _setIconWithOptionElement($elem);

    // 엘리먼트의 label 을 설정한다.
    _setLabel($elem);

    return $elem;
}

/**
 * select list 엘리먼트에 그룹 클래스를 추가한다.
 *
 * @param {object} [$selectElem=null] - select 엘리먼트 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _toggleGroupClassName($selectElem = null){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);
    let $optGroupElems = $overflowElem.find(`.${CLASS_NAME.optionGroup}`);

    if ($optGroupElems.length){
        $selectElem.find(`.${CLASS_NAME.bxSelectList}`).addClass('group');
    }
    else{
        $selectElem.find(`.${CLASS_NAME.bxSelectList}`).removeClass('group');
    }

    return this;
}

/**
 * overflow 엘리먼트의 자식 엘리먼트 Size 를 가지고, overflow class 를 toggle 시킨다.
 *
 * @param {object} [$selectElem=null] - select 엘리먼트 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _toggleOverFlowClassNameWithChildrenSize($selectElem = null){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    if (!$overflowElem.children().length){
        $overflowElem.removeClass(`${CLASS_NAME.overflow}`);
    }
    else{
        $overflowElem.addClass(`${CLASS_NAME.overflow}`);
    }

    return this;
}

/**
 * document 엘리먼트의 이벤트를 바인딩한다.
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _addDocumentElementEventListener(){

    // document 클릭 시, 해당 컴포넌트(마지막으로 열린)를 hide 시킨다.
    $(document).click(e =>{

        let $elem = $(e.target);
        let $selectElem = $elem.parents(`.${CLASS_NAME.bxUiSelect}`);

        if ($selectElem.get(0)) return false;

        // 마지막으로 열린 컴포넌트를 hide 시킨다.
        if (lastShowInstance){
            _hide.call(lastShowInstance);
        }
    });

    return this;
}
/**
 * select 엘리먼트의 이벤트를 바인딩한다.
 *
 * @param {object} [$selectElem=null]
 * @param {object} [opts={}]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _addSelectElementEventListener($selectElem = null, opts = {}){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxSelect, 'opts', 'type');
    }

    let selectElem = $selectElem.get(0);
    let multiple = $selectElem.prop('multiple');

    // document 클릭 시, 해당 컴포넌트(마지막으로 열린)를 hide 시킨다.
    $selectElem.click(e =>{

        let disabled = $selectElem.prop('disabled');

        if (multiple || disabled) return false;

        // 마지막 인스턴스가 존재할 경우
        if (lastShowInstance){

            // 마지막으로 열린 컴포넌트를 hide 시킨다.
            _hide.call(lastShowInstance);
        }

        if (!_.isEqual(selectElem, lastShowInstance)){

            _show.call(selectElem);

            _selectFirstOptionElem($selectElem);
        }
    });

    return this;
}

/**
 * selectElement 의 첫 번째 요소를 선택한다.
 *
 * @param {object} [$selectElem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _selectFirstOptionElem($selectElem = null){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $firstElem = $($overflowElem.find('div:first-child').get(0));
    let isOptGroupElem = $firstElem.is(`.${CLASS_NAME.optionGroup}`);

    $firstElem = isOptGroupElem ? $firstElem.find(`.${CLASS_NAME.option}:first-child`) : $firstElem;

    let $focusElem = $firstElem.find(`.${CLASS_NAME.optionValue}`);

    window.setTimeout(() => $focusElem.focus(), 30);

    return this;
}

/**
 * optGroup 엘리먼트의 이벤트들을 바인딩한다.
 *
 * @param {object} [$optGroupElem=null]
 * @param {object} [extendOpts={}]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _addOptGroupElementEventListener($optGroupElem = null, extendOpts = {}){

    if (!_.isElement($optGroupElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optGroupElem', 'type');
    }

    if (!_.isObject(extendOpts)){
        CONSOLE.prop.throw(BxSelect, 'extendOpts', 'type');
    }

    let multiple = extendOpts.multiple;

    let $h3Elem = $optGroupElem.children('h3');

    // 이벤트 바인딩
    $h3Elem.click(_click);
    $h3Elem.find(`.${CLASS_NAME.bxCheckbox}`).focus(_focus).blur(_blur);

    // 그룹 엘리먼트에대한, 키 이벤트는 `multiple` 이 true 인 경우에만 바인딩시킨다.
    if (multiple){

        let optGroupElem = $optGroupElem.get(0);

        Mousetrap(optGroupElem).bind('up', _up);
        Mousetrap(optGroupElem).bind('down', _down);
    }

    // 이벤트 함수
    function _click(){

        let $optGroupElem = $(this).parents(`.${CLASS_NAME.optionGroup}`);

        let disabled = $optGroupElem.prop('disabled');

        if (disabled) return false;

        if (multiple){
            $optGroupElem.prop('selected', ($optGroupElem.prop('selected') === true ? false : true));
            $h3Elem.find(`.${CLASS_NAME.bxCheckbox}`).focus();
        }
    }

    function _focus(){
        $h3Elem.addClass(CLASS_NAME.focus);
    }

    function _blur(){
        $h3Elem.removeClass(CLASS_NAME.focus);
    }

    function _up(e){

        let $elem = $(e.target);
        let groupElem = $elem.parents('h3').parents(`.${CLASS_NAME.optionGroup}`).get(0);

        if (groupElem){
            _moveOptGroupElem($optGroupElem, 'up');
        }

        // 키 이벤트를 통해, 발생되는 `기본 이벤트`를 막는다(browser scroll 이벤트 등)
        e.preventDefault();
    }

    function _down(e){

        let $elem = $(e.target);
        let groupElem = $elem.parents('h3').parents(`.${CLASS_NAME.optionGroup}`).get(0);

        if (groupElem){
            _moveOptGroupElem($optGroupElem, 'down');
        }

        // 키 이벤트를 통해, 발생되는 `기본 이벤트`를 막는다(browser scroll 이벤트 등)
        e.preventDefault();
    }

    return this;
}

/**
 * option 엘리먼트의 이벤트들을 바인딩한다.
 *
 * @param {object} [$optGroupElem=null]
 * @param {object} [extendOpts={}]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _addOptionElementEventListener($optionElem = null, extendOpts = {}, $selectElem = null){

    if (!_.isElement($optionElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    if (!_.isObject(extendOpts)){
        CONSOLE.prop.throw(BxSelect, 'extendOpts', 'type');
    }

    let multiple = extendOpts.multiple;

    let _click = function (){
    };

    let $focusElem = null;

    if (!multiple){

        _click = function (e){

            let disabled = $optionElem.prop('disabled');

            if (disabled) return false;

            $optionElem.prop('selected', true);

            $optionElem.parents(`.${CLASS_NAME.bxSelectWrap}`).find(`.${CLASS_NAME.text}`).focus();

            e.stopPropagation();
        };

        $focusElem = $optionElem.find(`.${CLASS_NAME.optionValue}`);
    }
    else{

        _click = function (e){

            let disabled = $optionElem.prop('disabled');

            if (disabled) return false;

            $optionElem.prop('selected', ($optionElem.prop('selected') ? false : true));

            $optionElem.find(`.${CLASS_NAME.bxCheckbox}`).focus();
        };


        $focusElem = $optionElem.find(`.${CLASS_NAME.bxCheckbox}`);

    }

    // 각종 이벤트 바인딩

    $optionElem.click(_click);
    $focusElem.focus(_focus).blur(_blur);

    let optionElem = $optionElem.get(0);

    Mousetrap(optionElem).bind('up', _up);
    Mousetrap(optionElem).bind('down', _down);


    // 이벤트 함수 정의
    function _focus(){
        $optionElem.addClass(CLASS_NAME.focus);
    }

    function _blur(){
        $optionElem.removeClass(CLASS_NAME.focus);
    }

    function _up(e){

        if (multiple) _moveOptionElemWithMultiple($optionElem, 'up');
        else _moveOptionElemNotMultiple($optionElem, 'up');

        // 키 이벤트를 통해, 발생되는 `기본 이벤트`를 막는다(browser scroll 이벤트 등)
        e.preventDefault();
    }

    function _down(e){

        if (multiple) _moveOptionElemWithMultiple($optionElem, 'down');
        else _moveOptionElemNotMultiple($optionElem, 'down');

        // 키 이벤트를 통해, 발생되는 `기본 이벤트`를 막는다(browser scroll 이벤트 등)
        e.preventDefault();
    }

    return this;
}

/**
 * 그룹 엘리먼트에 대한, 키 이동 함수
 *
 * @param {object} [$groupElem=null] - 그룹 엘리먼트
 * @param {string} [keyType={}] - 이벤트 key 값
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _moveOptGroupElem($groupElem = null, keyType = ''){

    if (!_.isElement($groupElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$groupElem', 'type');
    }

    if (!_.isString(keyType) || !_.trim(keyType).length){
        CONSOLE.prop.throw(BxSelect, 'keyType', 'all');
    }

    // 형제 엘리먼트
    let $siblingElem = null;

    if (keyType === 'up'){

        // 이전 형제 엘리먼트
        $siblingElem = $groupElem.prev();

        let isOptGroupElem = $siblingElem.is(`div.${CLASS_NAME.optionGroup}`);

        $siblingElem = isOptGroupElem ? $siblingElem.find(`div.${CLASS_NAME.option}:last-child`) : $siblingElem;
    }
    else{
        $siblingElem = $groupElem.find(`div.${CLASS_NAME.option}:first-child`);
    }

    if ($siblingElem.get(0)){

        $siblingElem.find(`.${CLASS_NAME.bxCheckbox}`).focus();
        _setScrollTopWithKeyEvent($siblingElem, keyType);
    }

    return this;
}

/**
 * option 엘리먼트에 대한, 키 이동 함수(multiple 이 true 일 경우)
 *
 * @param {object} [$optionElem=null] - 옵션 엘리먼트
 * @param {string} [keyType={}] - 이벤트 key 값
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _moveOptionElemWithMultiple($optionElem = null, keyType = ''){

    if (!_.isElement($optionElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    if (!_.isString(keyType) || !_.trim(keyType).length){
        CONSOLE.prop.throw(BxSelect, 'keyType', 'all');
    }

    // 해당 옵션 엘리먼트가 그룹에 포함되어있는지에대한 여부
    let isOptGroupOptionElem = $optionElem.parents(`div.${CLASS_NAME.optionGroup}`).length ? true : false;

    // 형제 엘리먼트
    let $siblingElem = keyType === 'up' ? $optionElem.prev() : $optionElem.next();

    if (isOptGroupOptionElem){

        if (!$siblingElem.get(0)){

            $siblingElem = $optionElem.parents(`div.${CLASS_NAME.optionGroup}`);

            if (keyType === 'down') $siblingElem = $siblingElem.next();
        }
    }
    else{

        let isOptGroupElem = $siblingElem.is(`div.${CLASS_NAME.optionGroup}`);

        if (isOptGroupElem){

            if (keyType === 'up'){
                $siblingElem = $siblingElem.find(`div.${CLASS_NAME.option}:last-child`);
            }
        }
    }

    if ($siblingElem.get(0)){

        $($siblingElem.find(`.${CLASS_NAME.bxCheckbox}`).get(0)).focus();
        _setScrollTopWithKeyEvent($optionElem, keyType);
    }

    return this;
}

/**
 * option 엘리먼트에 대한, 키 이동 함수(multiple 이 false 일 경우)
 *
 * @param {object} [$optionElem=null] - 옵션 엘리먼트
 * @param {string} [keyType={}] - 이벤트 key 값
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _moveOptionElemNotMultiple($optionElem = null, keyType = ''){


    if (!_.isElement($optionElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    if (!_.isString(keyType) || !_.trim(keyType).length){
        CONSOLE.prop.throw(BxSelect, 'keyType', 'all');
    }

    let isGroupOptionElem = $optionElem.parents(`div.${CLASS_NAME.optionGroup}`).length ? true : false;

    // (이전 / 이후)에 대한 형제 엘리먼트
    let $siblingElem = keyType === 'up' ? $optionElem.prev() : $optionElem.next();
    let selector = keyType === 'up' ? 'last-child' : 'first-child';

    // 그룹 내부 옵션인 경우
    if (isGroupOptionElem){

        // 형제 엘리먼트가 더이상 존재하지 않는 경우
        if (!$siblingElem.get(0)){

            let $parentElem = $optionElem.parents(`div.${CLASS_NAME.optionGroup}`);

            // (이전 / 이후)에 대한 형제 엘리먼트
            $siblingElem = keyType === 'up' ? $parentElem.prev() : $parentElem.next();

            let isGroupElem = $siblingElem.is(`div.${CLASS_NAME.optionGroup}`);
            $siblingElem = isGroupElem ? $siblingElem.find(`div.${CLASS_NAME.option}:${selector}`) : $siblingElem;
        }
    }
    else{
        // 현재 엘리먼트가 그룹 엘리먼트인지에 대한 여부
        let isGroupElem = $siblingElem.is(`div.${CLASS_NAME.optionGroup}`);
        $siblingElem = isGroupElem ? $siblingElem.find(`div.${CLASS_NAME.option}:${selector}`) : $siblingElem;
    }

    if ($siblingElem.get(0)){

        $siblingElem.find(`.${CLASS_NAME.optionValue}`).focus();

        _setScrollTopWithKeyEvent($optionElem, keyType);
    }

    return this;
}

/**
 * up / down 키 이벤트에 대한 `scrollTop` 값 설정 함수
 *
 * @param {object} [$targetElem=null] - 대상 엘리먼트
 * @param {string} [keyType={}] - 이벤트 key 값
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _setScrollTopWithKeyEvent($targetElem = null, keyType = ''){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    let $selectElem = $targetElem.parents(`.${CLASS_NAME.bxSelectWrap}`);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let currentScrollTop = $overflowElem.scrollTop();
    let height = $targetElem.height();
    let scrollTop = keyType === 'up' ? (currentScrollTop - height) : (currentScrollTop + height);

    // 메뉴 이동에 대한 `duration` 값은 `0.03` 초로 초기화시킨다.
    let duration = 30;

    $overflowElem.stop().animate({scrollTop: scrollTop}, duration);

    return this;
}

/**
 * 각 `custom` 엘리먼트에 정의된 attr 를 opt 객체로 반환한다.
 *
 * @param {object} [targetObject=null]
 * @param {object} [defaultOpts={}]
 *
 * @returns {object} 생성된 opt 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _createComponentOptions(targetObject = null, defaultOpts = {}){

    if (_.isNull(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isFunction(targetObject.get) && !_.isPlainObject(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isPlainObject(defaultOpts)){
        CONSOLE.prop.throw(BxSelect, 'defaultOpts', 'type');
    }


    let opts = UTIL.createComponentOptions(targetObject, defaultOpts, ATTRNAMES, function (k, v){

        if (
            k === 'multiple' ||
            k === 'disabled' ||
            k === 'selected'
        ){

            if (
                v !== 'true' &&
                v !== 'false' &&
                v !== true &&
                v !== false){

                CONSOLE.prop.throw(BxSelect, k, 'type');
            }

            v = v === 'true' || v === true ? true : false;
        }

        if (k === 'maxheight'){

            v = parseInt(v);

            if (isNaN(v)){
                CONSOLE.prop.throw(BxSelect, k, 'type');
            }
        }

        return v;

    }, function (k, v){
        return v;
    });

    return opts;
}

/**
 * select 엘리먼트의 프로퍼티를 설정한다.
 *
 * @param {object} [opts={}] - 옵션 객체
 * @param {object} [elem=null] - 프로퍼티가 추가될 엘리먼트
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setSelectElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxSelect, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxSelect, 'elem', 'type');
    }

    const $elem = $(elem);
    const _this = $elem.get(0);
    const elementType = $elem.attr('data-element-type');
    const bxPtn = /^bx/;

    if (opts.multiple) {
        opts.value = [];
    }

    _.map(opts, (v, k) =>{

        // attr 속성은 무시한다.
        if (k === 'attr') return;
        // events 속성은 무시한다.
        if (k === 'events') return;

        // 읽기 전용 속성
        if (
            k === 'type' ||
            k === 'shape' ||
            k === 'multiple' ||
            k === 'maxHeight'
        ){
            UTIL.setReadOnlyProperty(elem, k, v);
        }
        else{

            if (k.match(bxPtn)){

                let _k = k.replace('-', '');
                delete opts[_k];

                k = _k;
            }

            let _k = UTIL.convertInternalPropertyName(k);

            elem[_k] = v;

            UTIL.setPropertyWithDescriptor(elem, k, {
                configurable: true,
                enumerable: true,
                get: () =>{
                    return elem[_k];
                },
                set: v =>{

                    let multiple = $elem.prop('multiple');

                    if (k === 'label' && !_.isString(v)){
                        CONSOLE.prop.throw(BxSelect, k, 'type');
                    }

                    if (k === 'value'){
                        if (multiple && !_.isArray(v)){
                            CONSOLE.prop.throw(BxSelect, k, 'type');
                        }

                        if (!multiple && !_.isString(v)){
                            CONSOLE.prop.throw(BxSelect, k, 'type');
                        }
                    }

                    if (k === 'icon' && !_.isString(v)){
                        CONSOLE.prop.throw(BxSelect, k, 'type');
                    }

                    if (k === 'disabled'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxSelect, k, 'type');
                        }

                        if (multiple){
                            CONSOLE.prop.throw(BxSelect, MESSAGE.notSupportDisabledProperty);
                        }
                    }

                    if (k.match(bxPtn)){

                        if (!_.isFunction(v)){
                            CONSOLE.prop.throw(BxSelect, k, 'type');
                        }
                    }

                    elem[_k] = v;

                    if (k === 'label'){

                        let $textElem = $elem.find(`.${CLASS_NAME.bxSelectHeader} .${CLASS_NAME.text}`);

                        $textElem.text(v);
                    }

                    if (k === 'value'){

                        _.map($elem.prop('bxOptGroups'), groupElem =>{
                            _.map($(groupElem).prop('bxOptions'), elem =>{
                                if (elem.value === v && !elem.selected) elem.selected = true;
                            });
                        });

                        _.map($elem.prop('bxOptions'), elem =>{
                            if (elem.value === v && !elem.selected) elem.selected = true;
                        });
                    }

                    if (k === 'icon'){
                        _setIconWithSelectElement($elem);
                    }

                    if (k === 'disabled'){

                        _setDisabled($elem);

                        if (_isShow.call(elem)) _hide.call(elem);

                        $elem.prop('bxdisabled').call(_this, elementType, v);
                    }
                }
            });
        }
    });

    // API 에 대한, 프로퍼티 할당
    _.map(APIS, (v, k) =>{
        UTIL.setReadOnlyProperty(elem, k, v);
    });

    return this;
}

/**
 * optGroup 엘리먼트의 프로퍼티를 설정한다.
 *
 * @param {object} [opts={}] - 옵션 객체
 * @param {object} [elem=null] - 프로퍼티가 추가될 엘리먼트
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setOptGroupElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxSelect, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxSelect, 'elem', 'type');
    }

    let $elem = $(elem);
    let elementType = $elem.attr('data-element-type');

    _.map(opts, (v, k) =>{

        // attr 속성은 무시한다.
        if (k === 'attr') return;

        let _k = UTIL.convertInternalPropertyName(k);

        elem[_k] = v;

        UTIL.setPropertyWithDescriptor(elem, k, {
            configurable: true,
            enumerable: true,
            get: () =>{
                return elem[_k];
            },
            set: v =>{

                const $selectElem = $elem.parents(`.${CLASS_NAME.bxSelectWrap}`);
                const _this = $selectElem.get(0);

                if (k === 'label' && !_.isString(v)){
                    CONSOLE.prop.throw(BxSelect, k, 'type');
                }

                if (k === 'selected'){

                    if (v !== 'undetermined' && !_.isBoolean(v)){
                        CONSOLE.prop.throw(BxSelect, k, 'type');
                    }
                }

                if (k === 'disabled' && !_.isBoolean(v)){
                    CONSOLE.prop.throw(BxSelect, k, 'type');
                }

                elem[_k] = v;

                if (k === 'label'){
                    _setLabel($elem);
                }

                if (k === 'selected'){
                    window.setTimeout(() =>{
                        _setSelected($elem);
                    });

                    if (v) $selectElem.prop('bxselected').call(_this, elementType, v);
                    else $selectElem.prop('bxcleared').call(_this, elementType, v);
                }

                if (k === 'disabled'){

                    _setDisabled($elem);

                    $selectElem.prop('bxdisabled').call(_this, elementType, v);
                }
            }
        });
    });

    return this;
}

/**
 * option 엘리먼트의 프로퍼티를 설정한다.
 *
 * @param {object} [opts={}] - 옵션 객체
 * @param {object} [elem=null] - 프로퍼티가 추가될 엘리먼트
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setOptionElementProperties(opts = {}, elem = null){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxSelect, 'opts', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxSelect, 'elem', 'type');
    }

    let $elem = $(elem);
    let elementType = $elem.attr('data-element-type');

    _.map(opts, (v, k) =>{

        // attr 속성은 무시한다.
        if (k === 'attr') return;

        let _k = UTIL.convertInternalPropertyName(k);

        if (k === 'disabled'){

            window.setTimeout(function (){

                let $optGroupElem = $elem.parents(`.${CLASS_NAME.optionGroup}`);

                // 옵션 그룹의 disabled 속성이 true 인 경우
                if ($optGroupElem.prop('disabled')){

                    elem[_k] = true;

                    _setDisabled($elem);
                }

            }, 1);
        }

        elem[_k] = v;

        UTIL.setPropertyWithDescriptor(elem, k, {
            configurable: true,
            enumerable: true,
            get: () =>{
                return elem[_k];
            },
            set: v =>{

                const $selectElem = $elem.parents(`.${CLASS_NAME.bxSelectWrap}`);
                const _this = $selectElem.get(0);
                const $optGroupElem = $elem.parents(`.${CLASS_NAME.optionGroup}`);

                let multiple = $selectElem.prop('multiple');

                let _tmpV = $elem.prop(k);

                if (k === 'label' && !_.isString(v)){
                    CONSOLE.prop.throw(BxSelect, k, 'type');
                }

                if (k === 'value' && !_.isString(v)){
                    CONSOLE.prop.throw(BxSelect, k, 'type');
                }

                if (k === 'icon' && !_.isString(v)){
                    CONSOLE.prop.throw(BxSelect, k, 'type');
                }

                if (k === 'selected' && !_.isBoolean(v)){

                    CONSOLE.prop.throw(BxSelect, k, 'type');
                }

                if (k === 'disabled'){

                    if (!_.isBoolean(v)){
                        CONSOLE.prop.throw(BxSelect, k, 'type');
                    }

                    // 옵션 그룹의 disabled 속성이 true 인 경우
                    if ($optGroupElem.prop('disabled') && !v){
                        v = true;
                    }
                }

                elem[_k] = v;

                if (k === 'label'){
                    _setLabel($elem);
                }

                if (k === 'icon'){
                    _setIconWithOptionElement($elem);
                }

                if (k === 'selected'){

                    let text = $elem.text();

                    if (multiple){

                        window.setTimeout(() =>{
                            _setSelected($elem);

                            if ($optGroupElem.get(0)){
                                _selectOptGroupElement($optGroupElem, $elem);
                            }

                        }, 1);
                    }
                    else{

                        if (v){

                            _unSelectAllOptionElement($selectElem, $elem);

                            let selectElem = $selectElem.get(0);
                            if (_isShow.call(selectElem)) _hide.call(selectElem);

                            let selected = CLASS_NAME.selected;

                            // $selectElem label 변경
                            $selectElem.prop('label', text);

                            // 기존 selectElem 에서는 value 에 프로퍼티로 접근 하는 기능이 없었지만 추가되면서 문제가 발생 하여 아래와 같이 수정하였다.
                            selectElem[UTIL.convertInternalPropertyName('value')] = $elem.prop('value');

                            $selectElem.find('.' + selected).removeClass(selected);
                            $elem.addClass(selected);
                        }

                        if (_tmpV && !v){
                            _initSelectElement($selectElem);
                        }
                    }


                    if (v) $selectElem.prop('bxselected').call(_this, elementType, v);
                    else $selectElem.prop('bxcleared').call(_this, elementType, v);
                }

                if (k === 'disabled'){

                    _setDisabled($elem);

                    $selectElem.prop('bxdisabled').call(_this, elementType, v);
                }
            }
        });
    });

    return this;
}

/**
 * targetElem 의 가장 뒤에 새로운 엘리먼트를 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _addElement($targetElem = null, $elem = null){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    $targetElem.append($elem);

    return $elem;
}

/**
 * targetElem 의 가장 앞에 새로운 엘리먼트를 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _prependElement($targetElem = null, $elem = null){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    $targetElem.prepend($elem);

    return $elem;
}

/**
 * targetElem 의 이전에 새로운 엘리먼트를 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _insertBeforeElement($targetElem = null, $elem = null){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    $elem.insertBefore($targetElem);

    return $elem;
}

/**
 * targetElem 의 이후에 새로운 엘리먼트를 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 엘리먼트 객체
 *
 * @private
 * @memberOf BxSelect
 */
function _insertAfterElement($targetElem = null, $elem = null){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    $elem.insertAfter($targetElem);

    return $elem;
}

/**
 * 엘리먼트를 삭제한다.
 *
 * @param {object} [$targetElem=null]
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _removeElement($targetElem = null, $elem = null){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isElement($elem.get(0))) return this;


    $targetElem.children('#' + $elem.attr('id')).remove();

    return this;
}

/**
 * targetObject 의 가장 뒤에 새로운 아이템을 추가한다.
 *
 * @param {array} [targetObject=[]]
 * @param {object} [item=null]
 *
 * @returns {object} 생성된 item
 *
 * @private
 * @memberOf BxSelect
 */
function _addElementItem(targetObject = [], item = null){

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement(item)){
        CONSOLE.prop.throw(BxSelect, 'item', 'type');
    }

    targetObject.push(item);

    return targetObject[targetObject.length - 1];
}

/**
 * targetObject 의 가장 앞에 새로운 아이템을 추가한다.
 *
 * @param {array} [targetObject=[]]
 * @param {object} [item=null]
 *
 * @returns {object} 생성된 item
 *
 * @private
 * @memberOf BxSelect
 */
function prependElementItem(targetObject = [], item = null){

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement(item)){
        CONSOLE.prop.throw(BxSelect, 'item', 'type');
    }

    targetObject.unshift(item);

    return targetObject[0];
}

/**
 * targetObject 의 targetIndex 이전에 새로운 아이템을 추가한다.
 *
 * @param {array} [targetObject=[]]
 * @param {object} [item=null]
 * @param {number} [targetIndex=0]
 *
 * @returns {object} 생성된 item
 *
 * @private
 * @memberOf BxSelect
 */
function _insertBeforeElementItem(targetObject = [], item = null, targetIndex = 0){

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement(item)){
        CONSOLE.prop.throw(BxSelect, 'item', 'type');
    }

    if (!_.isNumber(targetIndex)){
        CONSOLE.prop.throw(BxSelect, 'targetIndex', 'type');
    }

    let length = targetObject.length;

    if (targetIndex < 0 || targetIndex > (length - 1)){
        CONSOLE.prop.throw(BxSelect, 'targetIndex', 'type');
    }

    targetObject.splice(targetIndex, 0, item);

    return targetObject[targetIndex];
}

/**
 * targetObject 의 targetIndex 이후에 새로운 아이템을 추가한다.
 *
 * @param {array} [targetObject=[]]
 * @param {object} [item=null]
 * @param {number} [targetIndex=0]
 *
 * @returns {object} 생성된 item
 *
 * @private
 * @memberOf BxSelect
 */
function _insertAfterElementItem(targetObject = [], item = null, targetIndex = 0){

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement(item)){
        CONSOLE.prop.throw(BxSelect, 'item', 'type');
    }

    if (!_.isNumber(targetIndex)){
        CONSOLE.prop.throw(BxSelect, 'targetIndex', 'type');
    }

    let length = targetObject.length;

    if (targetIndex < 0 || targetIndex > (length - 1)){
        CONSOLE.prop.throw(BxSelect, 'targetIndex', 'type');
    }

    targetObject.splice(++targetIndex, 0, item);

    return targetObject[targetIndex];
}

/**
 * 아이템을 삭제한다.
 *
 * @param {array} [targetObject=[]]
 * @param {object} [elem=null]
 *
 * @returns {object} 마지막 item
 *
 * @private
 * @memberOf BxSelect
 */
function _removeElementItem(targetObject = [], elem = null){

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isElement(elem)){
        CONSOLE.prop.throw(BxSelect, 'elem', 'type');
    }

    let idx = targetObject.indexOf(elem);

    targetObject.splice(idx, 1);

    return targetObject[targetObject.length - 1];
}

/**
 * option 엘리먼트 및 아이템들을 $targetElem 및 targetObject 의 가장 뒤에 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {array} [targetObject=[]]
 * @param {array} [optionItems=[]]
 * @param {boolean} [isIntoGroupElem=true] - optGroup 엘리먼트 아래 포함시킬지 여/부
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _addOptionElementToItems($targetElem = null, targetObject = [], optionItems = [], isIntoGroupElem = true){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    if (!_.isBoolean(isIntoGroupElem)){
        CONSOLE.prop.throw(BxSelect, 'isIntoGroupElem', 'type');
    }

    let $selectElem = $targetElem.parents(`.${CLASS_NAME.bxSelectWrap}`);

    _.map(optionItems, v =>{

        let $elem = isIntoGroupElem ? _createOptionElementIntoGroupElement(v, $selectElem) : _createOptionElementIntoOverFlowElement(v, $selectElem);

        _addElement($targetElem, $elem);

        _setSelected($elem);

        _setDisabled($elem);

        _addElementItem(targetObject, $elem.get(0));
    });

    return this;
}

/**
 * option 엘리먼트 및 아이템들을 $targetElem 및 targetObject 의 가장 앞에 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {array} [targetObject=[]]
 * @param {array} [optionItems=[]]
 * @param {boolean} [isIntoGroupElem=true] - optGroup 엘리먼트 아래 포함시킬지 여/부
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _prependOptionElementToItems($targetElem = null, targetObject = [], optionItems = [], isIntoGroupElem = true){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    if (!_.isBoolean(isIntoGroupElem)){
        CONSOLE.prop.throw(BxSelect, 'isIntoGroupElem', 'type');
    }

    let $selectElem = $targetElem.parents(`.${CLASS_NAME.bxSelectWrap}`);

    _.map(optionItems, v =>{

        let $elem = isIntoGroupElem ? _createOptionElementIntoGroupElement(v, $selectElem) : _createOptionElementIntoOverFlowElement(v, $selectElem);

        _prependElement($targetElem, $elem);

        _setSelected($elem);

        _setDisabled($elem);

        prependElementItem(targetObject, $elem.get(0));
    });

    return this;
}

/**
 * option 엘리먼트 및 아이템들을 $targetElem 및 targetObject 이전에 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {array} [targetObject=[]]
 * @param {array} [optionItems=[]]
 * @param {number} [targetIndex=0]
 * @param {boolean} [isIntoGroupElem=true] - optGroup 엘리먼트 아래 포함시킬지 여/부
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _insertBeforeOptionElementToItems($targetElem = null, targetObject = [], optionItems = [], targetIndex = 0, isIntoGroupElem = true){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    if (!_.isNumber(targetIndex)){
        CONSOLE.prop.throw(BxSelect, 'targetIndex', 'type');
    }

    if (!_.isBoolean(isIntoGroupElem)){
        CONSOLE.prop.throw(BxSelect, 'isIntoGroupElem', 'type');
    }

    let $selectElem = $targetElem.parents(`.${CLASS_NAME.bxSelectWrap}`);

    _.map(optionItems, v =>{

        let $elem = isIntoGroupElem ? _createOptionElementIntoGroupElement(v, $selectElem) : _createOptionElementIntoOverFlowElement(v, $selectElem);

        _insertBeforeElement($targetElem, $elem);

        _setSelected($elem);

        _setDisabled($elem);

        _insertBeforeElementItem(targetObject, $elem.get(0), targetIndex);
    });

    return this;
}

/**
 * option 엘리먼트 및 아이템들을 $targetElem 및 targetObject 이후에 추가한다.
 *
 * @param {object} [$targetElem=null]
 * @param {array} [targetObject=[]]
 * @param {array} [optionItems=[]]
 * @param {number} [targetIndex=0]
 * @param {boolean} [isIntoGroupElem=true] - optGroup 엘리먼트 아래 포함시킬지 여/부
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _insertAfterOptionElementToItems($targetElem = null, targetObject = [], optionItems = [], targetIndex = 0, isIntoGroupElem = true){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    if (!_.isNumber(targetIndex)){
        CONSOLE.prop.throw(BxSelect, 'targetIndex', 'type');
    }

    if (!_.isBoolean(isIntoGroupElem)){
        CONSOLE.prop.throw(BxSelect, 'isIntoGroupElem', 'type');
    }

    let $selectElem = $targetElem.parents(`.${CLASS_NAME.bxSelectWrap}`);

    _.map(optionItems, v =>{

        let $elem = isIntoGroupElem ? _createOptionElementIntoGroupElement(v, $selectElem) : _createOptionElementIntoOverFlowElement(v, $selectElem);

        _insertAfterElement($targetElem, $elem);

        _setSelected($elem);

        _setDisabled($elem);

        _insertAfterElementItem(targetObject, $elem.get(0), targetIndex);
    });

    return this;
}

/**
 * optGroup 엘리먼트 및 아이템들을 삭제한다.
 *
 * @param {object} [$targetElem=null]
 * @param {array} [targetObject=[]]
 * @param {array} [items=[]]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _removeOptGroupElementToItems($targetElem = null, targetObject = [], items = []){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxSelect, 'items', 'type');
    }

    _.map(items, v =>{

        let $elem = $targetElem.find(`#${v}`);

        let elementType = $elem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.optGroup){
            CONSOLE.prop.warn(BxSelect, MESSAGE.notOptGroupType);
            return;
        }

        _removeElement($targetElem, $elem);

        _removeElementItem(targetObject, $elem.get(0));
    });

    return this;
}

/**
 * option 엘리먼트 및 아이템들을 삭제한다.
 *
 * @param {object} [$targetElem=null]
 * @param {array} [targetObject=[]]
 * @param {array} [items=[]]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _removeOptionElementToItems($targetElem = null, targetObject = [], items = []){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(targetObject)){
        CONSOLE.prop.throw(BxSelect, 'targetObject', 'type');
    }

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxSelect, 'items', 'type');
    }

    _.map(items, v =>{

        let $elem = $targetElem.find(`#${v}`);

        let elementType = $elem.attr('data-element-type');

        // 엘리먼트 타입 검사
        if (elementType !== ELEMENT_TYPE.option){
            CONSOLE.prop.warn(BxSelect, MESSAGE.notOptionType);
            return;
        }

        _removeElement($targetElem, $elem);

        _removeElementItem(targetObject, $elem.get(0));
    });

    return this;
}

/**
 * 엘리먼트 및 아이템 업데이트를 위한 함수
 *
 * @param {object} [$targetElem=null]
 * @param {array} [items=[]]
 * @param {function} [callback=[]]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _updateElementToItems($targetElem = null, items = [], callback = function (){
}){

    if (!_.isElement($targetElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$targetElem', 'type');
    }

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxSelect, 'items', 'type');
    }

    if (!_.isFunction(callback)){
        CONSOLE.prop.throw(BxSelect, 'callback', 'type');
    }

    _.map(items, v =>{

        let $elem = $targetElem.find(`#${v}`);

        callback($targetElem.children('#' + $elem.attr('id')));
    });

    return this;
}

/**
 * show 여부를 반환한다.
 *
 * @returns {object} show 유/무
 *
 * @private
 * @memberOf BxSelect
 */
function _isShow(){

    let $componentElem = $(this);
    let $bxSelectListElem = $componentElem.find(`.${CLASS_NAME.bxSelectList}`);

    return $bxSelectListElem.css('display') === 'block' ? true : false;
}

/**
 * optGroup 엘리먼트를 선택한다.
 *
 * @param {object} [$optGroupElem=null]
 * @param {object} [$optionElem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _selectOptGroupElement($optGroupElem = null, $optionElem = null){

    if (!_.isElement($optGroupElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optGroupElem', 'type');
    }

    if (!_.isElement($optionElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    let bxOptions = $optGroupElem.prop('bxOptions');
    let selectedLength = _.filter(bxOptions, {selected: true}).length;
    let unSelectedLength = _.filter(bxOptions, {selected: false}).length;
    let optGroupElemSelected = $optGroupElem.prop('selected');

    if (!$optionElem.prop('selected') && bxOptions.length !== unSelectedLength){
        if (optGroupElemSelected === true) $optGroupElem.prop('selected', 'undetermined');
    }

    if (bxOptions.length === selectedLength){
        if (optGroupElemSelected === 'undetermined' || optGroupElemSelected === false) $optGroupElem.prop('selected', true);
    }

    if (bxOptions.length === unSelectedLength){
        if (optGroupElemSelected) $optGroupElem.prop('selected', false);
    }

    return this;
}

/**
 * select 엘리먼트의 아이콘을 설정한다.
 *
 * @param {object} [$selectElem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setIconWithSelectElement($selectElem = null){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let icon = $selectElem.prop('icon');
    let $headerElem = $selectElem.find(`.${CLASS_NAME.bxSelectHeader}`);
    let $iconElem = $headerElem.find(`.${CLASS_NAME.icon}`);

    $headerElem.removeClass('add-icon');
    $iconElem.find('i').remove();

    if (_.trim(icon).length > 0){
        $headerElem.addClass('add-icon');
        $iconElem.append(`<i class="${icon}"></i>`);
    }

    return this;
}

/**
 * option 엘리먼트의 아이콘을 설정한다.
 *
 * @param {object} [$optionElem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setIconWithOptionElement($optionElem = null){

    if (!_.isElement($optionElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    let icon = $optionElem.prop('icon');
    let $iconElem = $optionElem.find(`.${CLASS_NAME.icon}`);

    $iconElem.find('i').remove();

    if (_.trim(icon).length > 0){
        $iconElem.append(`<i class="${icon}"></i>`);
    }

    return this;
}

/**
 * 엘리먼트 label 을 설정한다.
 *
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setLabel($elem = null){

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    let label = $elem.prop('label');
    let $textElem = $($elem.find(`.${CLASS_NAME.optionValue}`).get(0));

    $textElem.text(label);

    return this;
}

/**
 * select 엘리먼트의 label 및 value 에 기본값을 설정한다.
 *
 * @param {object} [$selectElem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _initSelectElement($selectElem = null){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    let defaultLabel = $selectElem.prop('defaultLabel');
    let _value = UTIL.convertInternalPropertyName('value');

    // value는 이미 set에 등록 되어 있고 이 함수를 자기 자신이 호출 하고 있다
    // 프로퍼티로 접근시 해당 함수로 무한 반복 하기 때문에 convert 된
    // 이름으로 접근하면 set을 타는것을 방지 할 수 있다.

    $selectElem.prop('label', defaultLabel);
    $selectElem.prop(_value, '');

    return this;
}

/**
 * 커스텀 엘리먼트 아이디를 반환한다.
 *
 * @param {array} [targetArray=[]]
 *
 * @returns {array} 아이디 배열
 *
 * @private
 * @memberOf BxSelect
 */
function _getElementIds(targetArray = []){

    if (!_.isArray(targetArray)){
        CONSOLE.prop.throw(BxSelect, 'targetArray', 'type');
    }

    let ret = [];

    _.map(targetArray, v => ret.push($(v).attr('id')));

    return ret;
}

/**
 * 엘리먼트를 선택한다.
 *
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setSelected($elem = null){

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    let $selectElem = $elem.parents(`.${CLASS_NAME.bxSelectWrap}`);
    let multiple = $selectElem.prop('multiple');

    let elementType = $elem.attr('data-element-type');
    let selected = $elem.prop('selected');

    // select 엘리먼트의 multiple 속성이 true 인 경우
    if (multiple){

        let _selected = elementType === ELEMENT_TYPE.optGroup && selected === 'undetermined' ? false : selected;
        let $checkboxElem = $($elem.find('input[type="checkbox"]').get(0));

        const value = $elem.prop('value');

        $checkboxElem.prop('checked', _selected);

        if (_.isUndefined(value)) return;

        if (_selected) {
            $selectElem.prop('value').push(value);
        }
        else {
            const _value = $selectElem.prop('value');
            const idx = _value.indexOf(value);
            if (idx === -1) return;
            _value.splice(idx, 1);
            $selectElem.prop('value', _value);
        }
    }
    else{

        if (selected){

            let text = $elem.text();
            let value = $elem.prop('value');

            $selectElem.prop('label', text);
            $selectElem.prop('value', value);

        }

    }

    if (elementType === ELEMENT_TYPE.optGroup){

        _.map($elem.prop('bxOptions'), elem =>{

            if (selected === true){
                if (!elem.selected) elem.selected = true;
            }
            else if (selected === false){
                if (elem.selected) elem.selected = false;
            }
        });
    }

    return this;
}

/**
 * 모든 옵션 엘리먼트의 선택을 해제한다.
 *
 * @param {object} [$selectElem=null]
 * @param {object} [$optionElem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _unSelectAllOptionElement($selectElem = null, $optionElem = null){

    if (!_.isElement($selectElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$selectElem', 'type');
    }

    if (!_.isElement($optionElem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$optionElem', 'type');
    }

    let selectElem = $selectElem.get(0);
    let optionElem = $optionElem.get(0);

    let selected = $optionElem.prop('selected');

    let bxOptGroups = selectElem['bxOptGroups'];
    let bxOptions = selectElem['bxOptions'];

    _.map(bxOptGroups, (v, k) =>{
        _.map(v['bxOptions'], (vv, kk) =>{
            if (vv !== optionElem){
                if (vv.selected) vv.selected = false;
            }
        });
    });

    _.map(bxOptions, (v, k) =>{
        if (v !== optionElem){
            if (v.selected) v.selected = false;
        }
    });

    return this;
}

/**
 * optGroup 엘리먼트를 선택한다.
 *
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setSelectedWithOptGroupElement($elem = null){

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    let elementType = $elem.attr('data-element-type');
    let selected = $elem.prop('selected');

    let _selected = elementType === ELEMENT_TYPE.optGroup && selected === 'undetermined' ? false : selected;

    let $checkboxElem = $($elem.find('input[type="checkbox"]').get(0));

    $checkboxElem.prop('checked', _selected);

    if (elementType === ELEMENT_TYPE.optGroup){

        let bxOptions = $elem.prop('bxOptions');

        _.map(bxOptions, elem =>{

            if (selected === true){
                elem.selected = true;
            }
            else if (selected === false){
                let selected = elem.selected;
                elem.selected = selected;
            }
        });
    }

    return this;
}

/**
 * 엘리먼트를 disabled 한다.
 *
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setDisabled($elem = null){

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    let elementType = $elem.attr('data-element-type');
    let disabled = false;

    if (elementType === ELEMENT_TYPE.select){

        disabled = $elem.prop('disabled');

        let label = $elem.prop('label');

        $elem = $elem.find(`.${CLASS_NAME.bxSelectHeader}`);

        let $wrapElem = $elem.find('.wrap');
        $wrapElem.find('a').remove();
        $wrapElem.find('.text').remove();

        if (disabled){
            $wrapElem.append(`<span class="text">${label}</span>`);
        }
        else{
            $wrapElem.append(`<a href="#" class="text" onclick="return false;" title="${label}">${label}</a>`);
        }
    }
    else{
        disabled = $elem.prop('disabled');
    }

    if (disabled) $elem.addClass('disabled');
    else $elem.removeClass('disabled');

    if (elementType === ELEMENT_TYPE.optGroup){

        let $checkedElems = $elem.find('input[type="checkbox"]');
        let $optionElems = $elem.find('div div');

        if ($checkedElems.length) _.map($checkedElems, (v, k) => $(v).attr('disabled', disabled));
        _.map($optionElems, (v, k) => $(v).prop('disabled', disabled));
    }
    else if (elementType === ELEMENT_TYPE.option){

        let $checkedElem = $elem.find('input[type="checkbox"]');
        if ($checkedElem.get(0)) $checkedElem.attr('disabled', disabled);
    }

    return this;
}

/**
 * optGroup 엘리먼트를 disabled 한다.
 *
 * @param {object} [$elem=null]
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxSelect
 */
function _setDisabledWithOptGroupElement($elem = null){

    if (!_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxSelect, '$elem', 'type');
    }

    let elementType = $elem.attr('data-element-type');
    let disabled = false;

    if (elementType === ELEMENT_TYPE.select){
        disabled = $elem.prop('disabled');
        $elem = $elem.find(`.${CLASS_NAME.bxSelectHeader}`);
    }
    else{
        disabled = $elem.prop('disabled');
    }

    if (disabled) $elem.addClass('disabled');
    else $elem.removeClass('disabled');

    if (elementType === ELEMENT_TYPE.optGroup){

        let $checkedElems = $elem.find('input[type="checkbox"]');
        let $optionElems = $elem.find('div div');

        if (disabled){

            if ($checkedElems.length) _.map($checkedElems, (v, k) => $(v).attr('disabled', disabled));
            _.map($optionElems, (v, k) => $(v).prop('disabled', disabled));
        }
        else{

            if ($checkedElems.length){

                _.map($checkedElems, (v, k) =>{

                    let $v = $(v);
                    $v.attr('disabled', $v.prop('disabled'));
                });
            }

            _.map($optionElems, (v, k) =>{

                let $v = $(v);
                $v.prop('disabled', $v.prop('disabled'));
            });
        }
    }
    else if (elementType === ELEMENT_TYPE.option){

        let $checkedElem = $elem.find('input[type="checkbox"]');
        if ($checkedElem.get(0)) $checkedElem.attr('disabled', disabled);
    }

    return this;
}

/**
 *
 * 정해진 그룹의 가장 뒤에 새로운 옵션을 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias addGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _addGroupOptions(groupId = '', optionItems = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let groupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($optGroupElem);
    let $optGroupUlElem = $($optGroupElem.children('div').get(0));

    if (groupIndex === -1) return this;

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptGroups[groupIndex].bxOptions;

    _addOptionElementToItems($optGroupUlElem, bxOptions, optionItems);

    this['bxadd'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 정해진 그룹의 가장 앞에 새로운 옵션을 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias prependGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _prependGroupOptions(groupId = '', optionItems = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let groupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($optGroupElem);
    let $optGroupUlElem = $optGroupElem.children('div');

    if (groupIndex === -1) return this;

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptGroups[groupIndex].bxOptions;

    _prependOptionElementToItems($optGroupUlElem, bxOptions, optionItems);

    this['bxprepend'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 정해진 그룹의 정해진 옵션 이전에 새로운 옵션을 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 * @param {string} [optionId=''] - 옵션 아이디
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias insertBeforeGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _insertBeforeGroupOptions(groupId = '', optionId = '', optionItems = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isString(optionId) || _.trim(optionId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'optionId', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let groupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($optGroupElem);
    let $optGroupUlElem = $optGroupElem.children('div');

    if (groupIndex === -1) return this;

    let $optionElem = $optGroupUlElem.find(`#${optionId}`);
    let optionIndex = $optionElem.index();

    if (optionIndex === -1) return this;

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptGroups[groupIndex].bxOptions;

    _insertBeforeOptionElementToItems($optionElem, bxOptions, optionItems, optionIndex);

    this['bxinsertBefore'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 정해진 그룹의 정해진 옵션 이후에 새로운 옵션을 추가한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 * @param {string} [optionId=''] - 옵션 아이디
 *
 * @param {object|object[]} [optionItems=[]] - 옵션 아이템
 * @param {string} optionItems.label - label
 * @param {string} optionItems.value - value
 * @param {string} optionItems.icon - icon
 * @param {string} optionItems.selected - 엘리먼트의 selected 유/무
 * @param {boolean} [optionItems.disabled=false] - 엘리먼트의 disabled 유/무
 * @param {object} optionItems.attr - attr 객체
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias insertAfterGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _insertAfterGroupOptions(groupId = '', optionId = '', optionItems = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isString(optionId) || _.trim(optionId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'optionId', 'type');
    }

    if (!_.isPlainObject(optionItems) && !_.isArray(optionItems)){
        CONSOLE.prop.throw(BxSelect, 'optionItems', 'type');
    }

    optionItems = _.isPlainObject(optionItems) ? [optionItems] : optionItems;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let groupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($optGroupElem);
    let $optGroupUlElem = $optGroupElem.children('div');

    if (groupIndex === -1) return this;

    let $optionElem = $optGroupUlElem.find(`#${optionId}`);
    let optionIndex = $optionElem.index();

    if (optionIndex === -1) return this;

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptGroups[groupIndex].bxOptions;

    _insertAfterOptionElementToItems($optionElem, bxOptions, optionItems, optionIndex);

    this['bxinsertAfter'](ELEMENT_TYPE.option, bxOptions);

    return this;
}

/**
 *
 * 정해진 그룹의 옵션을 삭제한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 * @param {empty|string|string[]} [optionIds=[]] - 옵션 아이디
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias removeGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _removeGroupOptions(groupId = '', optionIds = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isString(optionIds) && !_.isArray(optionIds)){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    optionIds = _.isString(optionIds) ? [optionIds] : optionIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let groupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($optGroupElem);
    let $optGroupUlElem = $optGroupElem.children('div');

    if (groupIndex === -1) return this;

    let bxOptions = this.bxOptGroups[groupIndex].bxOptions;

    if (!optionIds.length){
        optionIds = _getElementIds(bxOptions);
    }

    _removeOptionElementToItems($optGroupUlElem, bxOptions, optionIds);

    return this;
}

/**
 *
 * 전체 노드를 선택한다.
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias selectAllNodes
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _selectAllNodes(){

    let bxOptions = this['bxOptions'];
    let bxOptGroups = this['bxOptGroups'];

    _.map(bxOptions, (v, k) => v.selected = true);

    _.map(bxOptGroups, (v, k) =>{

        v.selected = true;
        _.map(v['bxOptions'], (vv, kk) => vv.selected = true);
    });

    return this;
}

/**
 *
 * 전체 노드의 선택을 해제한다.
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias clearSelectedAllNodes
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _clearSelectedAllNodes(){

    let selectedBxOptions = _getSelectedOptions.call(this);
    let bxOptGroups = this['bxOptGroups'];

    _.map(selectedBxOptions, (v, k) => v.selected = false);

    _.map(bxOptGroups, (v, k) =>{

        v.selected = false;
        _.map(v['bxOptions'], (vv, kk) => vv.selected = false);
    });

    return this;
}

/**
 *
 * 선택된 모든 노드를 가져온다.
 *
 * @returns {array} 선택된 노드
 *
 * @alias getSelectedAllNodes
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _getSelectedAllNodes(){

    let bxOptGroups = this['bxOptGroups'];
    let bxOptions = this['bxOptions'];

    let ret = _.filter(bxOptions, {selected: true});

    _.map(bxOptGroups, (v, k) =>{

        _.map(v['bxOptions'], (vv, kk) =>{
            if (vv.selected) ret.push(vv);
        });
    });

    return ret;
}

/**
 *
 * 모든 노드를 가져온다.
 *
 * @returns {array} 선택된 노드
 *
 * @alias getAllNodes
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _getAllNodes(){

    let bxOptGroups = this['bxOptGroups'];
    let bxOptions = this['bxOptions'];

    let ret = bxOptions;

    _.map(bxOptGroups, (v, k) =>{
        _.map(v['bxOptions'], (vv, kk) => ret.push(vv));
    });

    return ret;
}

/**
 *
 * 선택된 그룹 옵션을 가져온다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 *
 * @returns {array} 선택된 그룹 옵션
 *
 * @alias getSelectedGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _getSelectedGroupOptions(groupId = ''){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'all');
    }

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let groupIndex = $overflowElem.find('div[data-element-type="optGroup"]').index($optGroupElem);

    // 옵션 그룹을 추가한다.
    let bxOptions = this.bxOptGroups[groupIndex].bxOptions;

    let ret = _.filter(bxOptions, {"selected": true});

    return ret;
}

/**
 *
 * 그룹 옵션을 선택한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 * @param {empty|string|string[]} [optionIds=[]] - 옵션 아이디
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias selectGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _selectGroupOptions(groupId = '', optionIds = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isString(optionIds) && !_.isArray(optionIds)){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    optionIds = _.isString(optionIds) ? [optionIds] : optionIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let $optGroupUlElem = $optGroupElem.children('div');

    if (!optionIds.length){
        optionIds = _getElementIds($optGroupElem.prop('bxOptions'));
    }

    _updateElementToItems($optGroupUlElem, optionIds, $optionElem =>{

        let elementType = $optionElem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.option) return;

        $optionElem.prop('selected', true);
    });

    return this;
}

/**
 *
 * 선택된 그룹 옵션을 해제한다.
 *
 * @param {string} [groupId=''] - 그룹 아이디
 * @param {empty|string|string[]} [optionIds=[]] - 옵션 아이디
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias clearSelectedGroupOptions
 *
 * @private
 *
 * @memberOf BxSelect
 */
function _clearSelectedGroupOptions(groupId = '', optionIds = []){

    if (!_.isString(groupId) || _.trim(groupId).length === 0){
        CONSOLE.prop.throw(BxSelect, 'groupId', 'type');
    }

    if (!_.isString(optionIds) && !_.isArray(optionIds)){
        CONSOLE.prop.throw(BxSelect, 'optionIds', 'type');
    }

    optionIds = _.isString(optionIds) ? [optionIds] : optionIds;

    let $selectElem = $(this);
    let $overflowElem = $selectElem.find(`.${CLASS_NAME.overflowSelector}`);

    let $optGroupElem = $overflowElem.find(`#${groupId}`);
    let $optGroupUlElem = $optGroupElem.children('div');

    if (!optionIds.length){
        optionIds = _getElementIds($optGroupElem.prop('bxOptions'));
    }

    _updateElementToItems($optGroupUlElem, optionIds, $optionElem =>{

        let elementType = $optionElem.attr('data-element-type');

        if (elementType !== ELEMENT_TYPE.option) return;

        $optionElem.prop('selected', false);
    });

    return this;
}

///**
// * object parsing 함수
// * @param {string} attribute name
// * @param {string} attribute value
// * @private
// * @memberOf BxSelect
// */
//function _parseObject(name = '', string = ''){
//
//    try{
//        return UTIL.JSParser(string);
//    }
//    catch(e){
//        CONSOLE.prop.throw(BxSelect, name, 'type');
//    }
//}

/**
 * 브라우저 스크롤 설정 함수
 *
 * @param hidden
 *
 * @private
 * @memberOf BxSelect
 */
function _setBrowserScroll(hidden = false){

    if (!_.isBoolean(hidden)){
        CONSOLE.prop.throw(BxSelect, 'hidden', 'type');
    }

    let $body = $(document.body);

    if (hidden){
        $body.css('overflow', 'hidden');
    }
    else{
        $body.css('overflow', LAGACYS.overflow);
    }

    return this;
}

module.exports = BxSelect;
