// bxuip-angular.js-ui/src/lib/bx-ui/app/components/sidebar.js
/**
 * show 메서드 호출 시, 호출될 callback 함수
 *
 * @callback onShow
 *
 * @memberOf BxSidebar
 */

/**
 * hide 메서드 호출 시, 호출될 callback 함수
 *
 * @callback onHide
 *
 * @memberOf BxSidebar
 */

/**
 * @class BxSidebar
 * @classDesc BxSidebar 클래스 생성자
 * @param {Object} OPTS
 * @param {String} OPTS.type='default' - sidebar 타입(default, full, icon)
 * @param {String} [OPTS.transitions='overlay'] - sidebar 의 기본 Body영역 (overlay, push)
 * @param {String} [OPTS.position='left'] - sidebar 출력 위치(left, right, top, bottom)
 * @param {Number} [OPTS.size=200] - contents sidebar 너비

 * @param {String} OPTS.items(templateUrl) - templateUrl 방식으로 sidebar를 그리는 데이터
 * @param {String} OPTS.items(selector) - jquery selector 방식으로 sidebar를 그리는 데이터
 * @param {Object[]} OPTS.items - Array.Object 형태의 데이터로 sidebar를 그린다

 * @param {String} [OPTS.items[].label=''] - 표시될 문구 지정
 * @param {String} [OPTS.items[].image=''] - type 이 <b class="dec">icon</b> 인 경우 사용이 가능하며, 무조건 Array 방식으로 시작해야 한다.<br/> 아이콘 클래스 지정
 * @param {Object} [OPTS.items[].attr={}] - 해당 label의 엘리먼트의 속성값을 지정
 * @param {Object[]} [OPTS.items[].items=''] - 하위 아이템을 지정<br/>
 * <b class="dec">templateUrl, selector, Array</b> 방식 전부 가능하다.

 * @param {String} [OPTS.color='black'] - sidebar의 background 색상 (black, white)
 * @param {Boolean} [OPTS.enableCloseButton=false] - sidebar 내부 닫기버튼 유무
 * @param {Boolean} [OPTS.enableResizeBody=false] - (push 타입 한정) 밀어내는 영역 사이즈를 자동 조절 유무
 * @param {Object} [OPTS.dimStyle={}] - sidebar를 제외한 영역의 dim처리 옵션
 * @param {Function} [OPTS.onHide=function(){}] - sidebar가 숨겨졌을 때 callback
 * @param {Function} [OPTS.onShow=function(){}] - sidebar가 보여졌을 때 callback
 * @param {Function} [OPTS.onRender=function(){}] - sidebar가 전부 그려졌을 때 callback
 *
 * @returns {Object} 생성된 인스턴스
 *
 * @example

 // 1. item 예제

 var sidebarDefaultItemInHeader = [
 {
     label: 'icon1 header1',
     attr: {
         class: 'class11',
         id: 'id11',
         name: 'name11',
         title: '타이틀11'
     },
     items: [
         {
             label: 'header1 item1',
             attr: {
                 class: 'class111',
                 id: 'id111',
                 name: 'name111',
                 title: '타이틀111'
             },
         },
         {
             label: 'header1 item2',
             attr: {
                 class: 'class112',
                 id: 'id112',
                 name: 'name112',
                 title: '타이틀112'
             },
         },
         {
             label: 'header1 item3',
             attr: {
                 class: 'class113',
                 id: 'id113',
                 name: 'name113',
                 title: '타이틀113'
             },
         }
     ]
 },
 {
     label: 'icon1 header2',
     attr: {
         class: 'class12',
         id: 'id12',
         name: 'name12',
         title: '타이틀12'
     },
     items: [
         {
             label: 'header2 item1',
             attr: {
                 class: 'class121',
                 id: 'id121',
                 name: 'name121',
                 title: '타이틀121'
             },
         },
         {
             label: 'header2 item2',
             attr: {
                 class: 'class122',
                 id: 'id122',
                 name: 'name122',
                 title: '타이틀122'
             },
         },
         {
             label: 'header2 item3',
             attr: {
                 class: 'class123',
                 id: 'id123',
                 name: 'name123',
                 title: '타이틀123'
             },
         }
     ]
 },
 {
     label: 'icon1 header3',
     attr: {
         class: 'class13',
         id: 'id13',
         name: 'name13',
         title: '타이틀13'
     },
     items: [
         {
             label: 'header3 item1',
             attr: {
                 class: 'class131',
                 id: 'id131',
                 name: 'name131',
                 title: '타이틀131'
             },
         },
         {
             label: 'header3 item2',
             attr: {
                 class: 'class132',
                 id: 'id132',
                 name: 'name132',
                 title: '타이틀132'
             },
         },
         {
             label: 'header3 item3',
             attr: {
                 class: 'class133',
                 id: 'id133',
                 name: 'name133',
                 title: '타이틀133'
             },
         }
     ]
 }
 ];

 var sidebar = $().BxSidebar({
    type: 'default',
    transitions: 'overlay',
    position: 'left',
    size: 220,
    items: sidebarDefaultItemInHeader,
    color: 'black',
    enableCloseButton: true,
    enableResizeBody: true,
    dimStyle: {
        color: 'theme',
    },
    onHide: function(){
        console.log('sidebar onHide');
    },
    onShow: function(){
        console.log('sidebar onShow');
    },
    onRender: function(){
        console.log('sidebar onRender');
    }
 }).render();

 sidebar.show();

 ----------------------------------------------------------------------

 // 2. templateUrl 예제

 // template.tpl.html
<div>
     <div>
         <ul>
             <li>test1111</li>
             <li>test2222</li>
             <li>test3333</li>
         </ul>
     <div>
        <span>test4444</span>
     </div>
     <div>
        <span>test5555</span>
     </div>
     <div>
         <span>test6666</span>
     </div>
 </div>

 // js

 var sidebar = $().BxSidebar({
    type: 'default',
    transitions: 'overlay',
    position: 'left',
    size: 220,
    items: 'this/is/your/template/url/template.tpl.html',
    color: 'black',
    enableCloseButton: true,
    enableResizeBody: true,
    dimStyle: {
        color: 'theme',
    },
    onHide: function(){
        console.log('sidebar onHide');
    },
    onShow: function(){
        console.log('sidebar onShow');
    },
    onRender: function(){
        console.log('sidebar onRender');
    }
 }).render();

 sidebar.show();

 ----------------------------------------------------------------------

 // 3. selector 예제

 // template

 <div id="nav">
     <ul>
         <li>test1111</li>
         <li>test2222</li>
         <li>test3333</li>
     </ul>
     <div>
        <span>test4444</span>
     </div>
     <div>
        <span>test5555</span>
     </div>
     <div>
        <span>test6666</span>
     </div>
 </div>


 // js

 var sidebar = $().BxSidebar({
    type: 'default',
    transitions: 'overlay',
    position: 'left',
    size: 220,
    items: '#nav',
    color: 'black',
    enableCloseButton: true,
    enableResizeBody: true,
    dimStyle: {
        color: 'theme',
    },
    onHide: function(){
        console.log('sidebar onHide');
    },
    onShow: function(){
        console.log('sidebar onShow');
    },
    onRender: function(){
        console.log('sidebar onRender');
    }
 }).render();

 sidebar.show();

 ----------------------------------------------------------------------

 // 4. templateUrl + icon Type 예제

 // template.tpl.html
<div>
     <div>
         <ul>
             <li>test1111</li>
             <li>test2222</li>
             <li>test3333</li>
         </ul>
     <div>
        <span>test4444</span>
     </div>
     <div>
        <span>test5555</span>
     </div>
     <div>
        <span>test6666</span>
     </div>
 </div>

 // js

 var sidebar = $().BxSidebar({
    type: 'default',
    transitions: 'overlay',
    position: 'left',
    size: 220,
    items: [
        {
            label: 'icon', // icon, header
            image: 'i-ctrl-minus',
            attr: {
                class: 'icon1',
                id: 'icon1',
                name: 'icon1',
                title: '아이콘1'
            },
            items: 'this/is/your/template/url/template.tpl.html'
        },
        {
            label: 'icon', // icon, header
            image: 'i-ctrl-next',
            attr: {
                class: 'icon2',
                id: 'icon2',
                name: 'icon2',
                title: '아이콘2'
            },
            items: 'this/is/your/template/url/template.tpl.html'
        },
        {
            label: 'icon', // icon, header
            image: 'i-ctrl-pre',
            attr: {
                class: 'icon3',
                id: 'icon3',
                name: 'icon3',
                title: '아이콘3'
            },
            items: 'this/is/your/template/url/template.tpl.html'
        }
    ],
    color: 'black',
    enableCloseButton: true,
    enableResizeBody: true,
    dimStyle: {
        color: 'theme',
    },
    onHide: function(){
        console.log('sidebar onHide');
    },
    onShow: function(){
        console.log('sidebar onShow');
    },
    onRender: function(){
        console.log('sidebar onRender');
    }
 }).render();

 sidebar.show();

 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const OPTS = Symbol();

// Sidebar 타입
const SIDEBAR_TYPE = {
    DEFAULT: 'default',
    FULL: 'full',
    ICON: 'icon'
};

// Sidebar Animation 방법
const SIDEBAR_TRANSITIONS = {
    PUSH: 'push',
    OVERLAY: 'overlay'
};

// Sidebar 출력 위치
const SIDEBAR_POSITION = {
    TOP: 'top',
    BOTTOM: 'bottom',
    LEFT: 'left',
    RIGHT: 'right'
};

const SIDEBAR_MARGIN = {
    RIGHT : 'margin-right',
    LEFT : 'margin-left',
    BOTTOM : 'margin-bottom',
    TOP : 'margin-top'
};

const SIDEBAR_ORIENTATION = {
    WIDTH: 'width',
    HEIGHT: 'height'
};

const SIDEBAR_COLOR = {
    BLACK: 'black',
    WHITE: 'white'
};

// Sidebar 클래스명
const SIDEBAR_CLASS_NAME = {
    WRAP: 'bx-sidebar-wrap',
    WRAP_CONTENT: '-bx-ui-sidebar-contents',
    ITEM_GROUP_WRAP: 'item-group-wrap',
    ITEM_GROUP: 'item-group',
    ITEM: 'item',
    HEADER: 'header'
};

// Sidebar Error 메세지
const SIDEBAR_ERR_MSG = {
    CONTENTS_TYPE_NOT_ICON: 'default type이나 full type인 경우에는 item은 templateUrl string이나 구성하는 array만 가능합니다',
    CONTENTS_TYPE_ICON: 'icon type인 경우 item은 array만 가능합니다'
};

// Sidebar Contents Container 기본 사이즈
const SIDEBAR_DEFAULT_CONTAINER_SIZE = 200;
// Sidebar Icon Container 기본 사이즈
const SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE = 40;
// Sidebar Animation 기본 딜레이
const SIDEBAR_DEFAULT_ANIMATE_DURATION = 500;

const SIDEBAR_DEFAULT_ERASING = 'easeInOutExpo';

class BxSidebar {

    constructor({
        type = 'default',
        transitions = 'overlay',
        position = 'left',
        size = 0,
        items = null,
        color = 'black',
        enableCloseButton = false,
        enableResizeBody = false,
        dimStyle = null,
        onHide = function(){},
        onShow = function(){},
        onRender = function(){}
    } = {}){

        if (!_.isString(type) || _.isUndefined(SIDEBAR_TYPE[type.toUpperCase()])){
            CONSOLE.prop.throw(BxSidebar, 'type', 'type');
        }

        if (!_.isString(transitions) || _.isUndefined(SIDEBAR_TRANSITIONS[transitions.toUpperCase()])){
            CONSOLE.prop.throw(BxSidebar, 'transitions', 'type');
        }

        if (transitions === SIDEBAR_TRANSITIONS.PUSH && !_.isBoolean(enableResizeBody)){
            CONSOLE.prop.throw(BxSidebar, 'enableResizeBody', 'type');
        }

        if (!_.isString(position) || _.isUndefined(SIDEBAR_POSITION[position.toUpperCase()])){
            CONSOLE.prop.throw(BxSidebar, 'type', 'type');
        }

        let maxSize = position === SIDEBAR_POSITION.LEFT || position === SIDEBAR_POSITION.RIGHT ? $(document).width() - SIDEBAR_DEFAULT_CONTAINER_SIZE : $(document).height() - SIDEBAR_DEFAULT_CONTAINER_SIZE;

        if (!_.isInteger(size)){
            CONSOLE.prop.throw(BxSidebar, 'size', 'type');
        }
        else if (size < SIDEBAR_DEFAULT_CONTAINER_SIZE){
            size = SIDEBAR_DEFAULT_CONTAINER_SIZE;
        }
        else if (size > maxSize){
            size = maxSize;
        }

        if (_.isNull(items)){
            CONSOLE.prop.throw(BxSidebar, 'items', 'type');
        }
        else if((type === SIDEBAR_TYPE.DEFAULT || type === SIDEBAR_TYPE.FULL) && (!_.isString(items) && !_.isArray(items))){
            CONSOLE.prop.throw(BxSidebar, SIDEBAR_ERR_MSG.CONTENTS_TYPE_NOT_ICON, 'empty');
        }
        else if (type === SIDEBAR_TYPE.ICON && (!_.isArray(items) && !_.isString(items))){
            CONSOLE.prop.throw(BxSidebar, SIDEBAR_ERR_MSG.CONTENTS_TYPE_ICON, 'empty');
        }

        if (!_.isString(color) || _.isUndefined(SIDEBAR_COLOR[color.toUpperCase()])){
            CONSOLE.prop.throw(BxSidebar, 'color', 'type');
        }

        if (!_.isBoolean(enableCloseButton)){
            CONSOLE.prop.throw(BxSidebar, 'enableCloseButton', 'type');
        }

        if (!_.isFunction(onHide)){
            CONSOLE.prop.throw(BxSidebar, 'onHide', 'type');
        }

        if (!_.isFunction(onShow)){
            CONSOLE.prop.throw(BxSidebar, 'onHide', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxSidebar, 'onHide', 'type');
        }

        this.onShow = onShow;
        this.onHide = onHide;
        this.onRender = onRender;
        this.orientation = position === SIDEBAR_POSITION.LEFT || position === SIDEBAR_POSITION.RIGHT ? SIDEBAR_ORIENTATION.WIDTH : SIDEBAR_ORIENTATION.HEIGHT;
        this.reversePosition = _getReversePosition(position);
        this.rootSize = this.orientation === 'width' ? BxSidebar.$root.width() : BxSidebar.$root.height();

        this.$dim = null;
        this.isOnlyIcon = _checkOnlyIcon.call(this, items, type);

        this[OPTS] = {
            type,
            transitions,
            position,
            size,
            items,
            color,
            enableCloseButton,
            enableResizeBody,
            dimStyle
        };
    }

    /**
     * BxSidebar 생성하는 함수
     *
     * @memberOf BxSidebar
     */
    render(){

        let $root = BxSidebar.$root;
        let $bodyContainer = null;
        let $container = null;
        let style = 'position:absolute;';

        // $root 가 body인지 체크
        let isRootBody = $root.get(0).tagName === 'BODY';

        let tempContainer = $(`<div class="${SIDEBAR_CLASS_NAME.WRAP_CONTENT}" style="${style}"></div>`);

        // 포지션별 컨테이너 시작 px
        let positionStartPx = Number($root.css(this[OPTS].position).replace(/px|auto/g,''));
        positionStartPx = positionStartPx <= 0 ? 0 : positionStartPx;

        if (this[OPTS].type === SIDEBAR_TYPE.ICON){

            let position = this[OPTS].position;

            if (position === 'left' || position === 'top'){
                tempContainer.css({
                    [position]: SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE
                });
            }
            else{
                tempContainer.css({
                    ['margin-' + position]: SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE
                });
            }
        }

        $root.children().wrapAll(tempContainer);

        $bodyContainer = $(`.${SIDEBAR_CLASS_NAME.WRAP_CONTENT}`);

        // $root가 body인 경우엔 body에 생성, 아닐경우 $root와 같은 level에 컨테이너 생성
        $container = BASE.appendContainerElement(BxSidebar.$root, BxSidebar.className);

        $($bodyContainer).after($container);

        this.$root = $root;
        this.$container = $container;
        this.$bodyContainer = $bodyContainer;
        this.positionStartPx = positionStartPx;

        _createDOMElement.call(this, $container);

        if ($bodyContainer.length){
            _createDim.call(this, $bodyContainer);
        }
        else {
            _createDim.call(this, $root);
        }

        if (this[OPTS].enableResizeBody){
            _setResizeEventListener.call(this);
        }

        this.onRender.call(this, this, $container);

        return this;
    }

    /**
     * Sidebar Contents Container를 보여주는 함수
     *
     * @param {Number} [index = 1] - Icon type인 경우 불러올 아이콘의 Index
     *
     * @memberOf BxSidebar
     */
    show(index = 1){
        _showContentsContainer.call(this, index - 1);
    }

    /**
     * Sidebar Contents Container를 숨겨주는 함수
     *
     * @memberOf BxSidebar
     */
    hide(){
        _hideContentsContainer.call(this);
    }
}

/**
 * Sidebar의 DOM 생성 함수
 *
 * @param {Object} [$container = {}] - 생성된 $container
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _createDOMElement($container = {}){

    let options = this[OPTS];
    let type = options.type;
    let $contentsContainer = null;
    let $iconContainer = null;

    if (type === SIDEBAR_TYPE.ICON){
        $iconContainer = _createIconContainer.call(this, options);
    }

    // icon만 있는경우 contentsContainer 제외
    if (!this.isOnlyIcon){
        $contentsContainer = _createContentsContainer.call(this, options);
    }

    // animation 때문에 Contents Container를 먼저 추가
    $container.append($iconContainer);
    $container.append($contentsContainer);

    this.$iconContainer = $iconContainer;
    this.$contentsContainer = $contentsContainer;
}

/**
 * Sidebar의 Icon Container 생성 함수
 *
 * @param {Object} [options = {}] - 초기 설정 options
 *
 * @returns {Object} 생성된 Icon Container Element
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _createIconContainer(options = {}){

    if (!_.isPlainObject(options)){
        CONSOLE.prop.throw(BxSidebar, 'options', 'all');
    }

    let items = options.items;
    let position = options.position;
    let type = options.type;
    let iconType = type === SIDEBAR_TYPE.ICON ? 'icon-type' : 'icon-type combiner';
    let color = options.color === SIDEBAR_COLOR.WHITE ? 'invert' : '';

    let $html = $(`<div class="${SIDEBAR_CLASS_NAME.WRAP} ${iconType} ${position} ${color}"></div>`);
    let $itemGroup = $(`<div class="${SIDEBAR_CLASS_NAME.ITEM_GROUP}"></div>`);

    _.map(items, (item)=>{
        $itemGroup.append(_createIconElement.call(this, item));
    });

    $html.append($itemGroup);

    return $html;
}

/**
 * Icon Container 내부 Element를 생성하여 반환하는 함수
 *
 * @param {Object} [iconObj = {}] - 아이콘 관련 options
 *
 * @returns {Object} 생성된 Icon 내부 Element
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _createIconElement(iconObj = {}){

    let imgReg = /\.(png|jpg|gif|jpeg)/g;
    let faReg = /^fa\s/;
    let attr = iconObj.attr || {};
    let title = _.isUndefined(attr.title) ? '' : `title=${attr.title}`;

    let $html = $(`<a class="${SIDEBAR_CLASS_NAME.ITEM}" href="javascript:;"></a>`);

    // title을 제외한 속성값 a Tag에 추가
    $html = _setDOMAttribute.call(this, $html, attr);

    // 아이콘 선택
    if (imgReg.test(iconObj.image)){
        $html.append(`<img class="i-25" src="${iconObj.image}" ${title}>`);
    }
    else {

        let imageClass = faReg.test(iconObj.image) ? iconObj.image : `bw-icon i-25 ${iconObj.image}`;

        $html.append(`<i class="${imageClass}" ${title}></i>`);
    }

    // 이벤트 바인딩
    $html.on('click', _clickIcon.bind(this));

    return $html;
}

/**
 * Sidebar icon 클릭시 callback 함수
 *
 * @param {Object} - icon click event callback
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _clickIcon(e){

    let target = $(e.currentTarget);

    // on Class 추가
    _addOnClass.call(this, target, 'icon');

    // contentsContainer 보이기 animation
    _showContentsContainer.call(this, target.index());
}

/**
 * Sidebar의 Contents Container DOM을 생성하여 반환한다
 *
 * @param {Object} [options = {}] - 초기 설정 options
 *
 * @returns {Object} 생성된 Contents Container Elements
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _createContentsContainer(options = {}){

    if (!_.isPlainObject(options)){
        CONSOLE.prop.throw(BxSidebar, 'contents container ptions', 'all');
    }

    let items = options.items;
    let type = options.type;
    let enableCloseButton = options.enableCloseButton;
    let $html = null;
    let itemType = _checkItemType(items);

    $html = _initContentsContainer.call(this, options);

    // 최초 item 값이 templateUrl인 경우 (default type이나 full type만)
    switch (itemType){
        case 'url':
        case 'selector':

            let template = _getTemplate(items, itemType);

            $html.append(template);

            break;

        case 'item':

            _.map(items, (item)=>{

                let childItemType = _checkItemType(item.items);

                // Mix 인 경우에 클래스 추가
                if (type === SIDEBAR_TYPE.ICON && !this.isOnlyIcon){
                    $html.addClass('combiner');
                }

                // icon type인데 templateUrl이 있는 경우
                if (type === SIDEBAR_TYPE.ICON && (childItemType === 'url' || childItemType === 'selector' )){

                    let $groupWrap = $(`<div class=${SIDEBAR_CLASS_NAME.ITEM_GROUP_WRAP}>`);

                    $groupWrap.append(_getTemplate(item.items, childItemType));

                    $groupWrap.hide();

                    $html.append($groupWrap);
                }
                else {
                    // default type, full type, icon type인데 templateUrl이 없는 경우 여기서 그리기 시작
                    let tempItems = type === SIDEBAR_TYPE.ICON ? item.items : items;

                    $html.append(_createContentsHeader.call(this, tempItems));
                }
            });

            break;
    }

    if (enableCloseButton){

        let $closeButton = $('<button type="button" class="close-btn"></button>');

        // x 버튼 이벤트 등록
        $closeButton.on('click', _hideContentsContainer.bind(this));

        $html.append($closeButton);
    }

    return $html;
}

/**
 * Type별 Contents Container 초기값을 설정한다
 *
 * @param {Object} [options = {}] - Contents 관련 options
 *
 * @returns {Object} Contents Container Wrapper Element
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _initContentsContainer(options = {}){

    if (!_.isPlainObject(options)){
        CONSOLE.prop.throw(BxSidebar, 'init options', 'all');
    }

    let $html = null;
    let position = options.position;
    let type = options.type;
    let size = options.size;

    let orientation = this.orientation;
    let positionStartPx = this.positionStartPx;
    let color = options.color === SIDEBAR_COLOR.WHITE ? 'invert' : '';

    let contentSize = '';
    let hiddenSize = '';
    let totalSize = 0;

    // type별 contentsContainer 길이 구하기
    totalSize = _calcContainerWidth.call(this, type, size, position);

    // 처음 contents 영역은 숨김
    hiddenSize = `${position}: -${totalSize - positionStartPx}px`;

    if (type === SIDEBAR_TYPE.FULL){
        $html = $(`<div class="${SIDEBAR_CLASS_NAME.WRAP} ${type} ${color}" style="${position}: -${totalSize}px;"></div>`);
    }
    else {

        let iconType = 'default-type';

        // 가로 or 세로 길이 설정
        contentSize = `${orientation}:${size}px`;

        $html = $(`<div class="${SIDEBAR_CLASS_NAME.WRAP} ${iconType} ${position} ${color} hide" style="${contentSize}; ${hiddenSize};"></div>`);
    }

    // 최초에는 hide 상태로 시작
    $html.hide();

    this.hiddenSize = totalSize - positionStartPx;

    return $html;
}

/**
 * Contents Container Header Elements를 생성하여 반환 한다
 *
 * @param {Array} [items = []] - header 영역을 그릴 items
 *
 * @returns {Object} 생성된 Header Elements
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _createContentsHeader(items = []){

    if (!_.isArray(items)){
        CONSOLE.prop.throw(BxSidebar, 'header items', 'all');
    }

    let $groupWrap = $(`<div class=${SIDEBAR_CLASS_NAME.ITEM_GROUP_WRAP}>`);

    _.map(items, (item)=>{

        // header 하위가 있는경우
        if (_.isArray(item.items) && item.items.length){

            let $header = $(`<div class="header">${item.label}</div>`);
            let $group = $(`<div class="${SIDEBAR_CLASS_NAME.ITEM_GROUP}"></div>`);

            // header 영역 등록
            $group.append($header);

            _.map(item.items, (childItem)=>{
                $group.append(_createContentsItems.call(this, childItem));
            });

            $groupWrap.append($group);
        }
        else {
            $groupWrap.append(_createContentsItems.call(this, item));
        }
    });

    // 최초에는 모든 groupWrap을 전부 hidden 처리
    $groupWrap.hide();

    return $groupWrap;
}

/**
 * Contents Container Item Elements를 생성하여 반환 한다
 *
 * @param {Object} [items = {}] - item Elements를 생성하기 위한 데이터
 *
 * @returns {Object} 생성된 Item Elements
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _createContentsItems(items = {}){

    if (!_.isPlainObject(items)){
        CONSOLE.prop.throw(BxSidebar, 'Item Elements items', 'all');
    }

    let $childItem = $(`<a class="${SIDEBAR_CLASS_NAME.ITEM}" href="javascript:;"></a>`);
    let $label = $(`<b>${items.label}</b>`);

    $childItem = _setDOMAttribute.call(this, $childItem, items.attr);

    $childItem.append($label);

    // 이벤트 등록
    $childItem.on('click', _clickChildItem.bind(this));

    return $childItem;
}

/**
 * Contents Item 클릭 이벤트 callback 함수
 *
 * @param {Object} - Contents Item click event callback
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _clickChildItem(e){

    let target = $(e.currentTarget);

    _addOnClass.call(this, target, 'contents');
}

/**
 * Class 'on' 을 붙이는 함수
 *
 * @param {Object} [target = {}] - on을 붙일 Elements
 * @param {String} [type = ''] - on을 붙이는 곳의 Container Type
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _addOnClass(target = {}, type = ''){

    if (!_.isObject(target)){
        CONSOLE.prop.throw(BxSidebar, `${type} on class target`, 'type');
    }

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxSidebar, `target class type`, 'type');
    }

    if (type === 'icon'){

        // 형제의 on 클래스 삭제
        target.siblings().removeClass('on');
    }
    else {

        // 모든 컨텐츠 영역에서 on Class 삭제
        target.parents('.bx-sidebar-wrap').find('.on').removeClass('on');
    }

    // 클릭한 target on 클래스 추가
    target.addClass('on');
}

/**
 * template HTML 반환 함수.
 *
 * @param {String} [item = ''] - 요청할 item 값
 * @param {String} [type = ''] - 요청 할 아이템의 형식
 *
 * @returns {String} - 가져온 html text
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _getTemplate(item = '', type = ''){

    let template = null;

    if (type === 'url'){
        $.ajax({
            url: item,
            success: (req) => {
                template = req;
            },
            contentType: 'text/plain',
            async: false
        });
    }
    else if(type === 'selector') {
        template = $(item);
    }

    return template;
}

/**
 * dim 생성 함수
 *
 * @private
 *
 * @memberOf BxSidebar
 */
// this.$dim 속성 설정
function _createDim(body){

    let dimStyle = this[OPTS].dimStyle;

    if (!_.isNull(dimStyle)){

        let $dim = body.bxDim(dimStyle);

        this.$dim = $dim;
    }
}

/**
 * Elements에 attr 옵션의 값을 설정 하는 함수
 *
 * @param {object} [$elem = {}] - attr을 설정할 element
 * @param {object} [attrObj = {}] - 설정할 attr 값
 *
 * @returns {object} attr 옵션이 설정된 Elements
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _setDOMAttribute($elem = {}, attrObj = {}){

    let $html = _.cloneDeep($elem);

    _.map(attrObj, (attrValue, attrKey)=>{

        if (attrKey === 'title'){
            return;
        }
        else if (attrKey.toLowerCase() === 'class'){
            $html.addClass(attrValue);
        }
        else {
            $html.attr(attrKey, attrValue);
        }
    });

    return $html;
}

/**
 * Contents Container를 보이기 위한 설정 및 animation 함수
 *
 * @param {Number} [index = 0] - icon Type인 경우 선택한 icon index
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _showContentsContainer(index = 0){

    if (!_.isInteger(index) || index < 0){
        CONSOLE.prop.throw(BxSidebar, 'icon index', 'all');
    }

    let $contentsContainer = this.$contentsContainer;
    let $bodyContainer = this.$bodyContainer;
    let opts = this[OPTS];
    let $dim = this.$dim;
    let onlyIcon = this.isOnlyIcon;

    let type = opts.type;
    let transitions = opts.transitions;
    let position = opts.position;
    let enableResizeBody = opts.enableResizeBody;

    let transitionObj = _calcContainerTransition.call(this, 'show', 'sidebar');
    let transitionBodyObj = _calcContainerTransition.call(this, 'show', 'body');

    if (!onlyIcon){

        // container animation
        if (!$contentsContainer.hasClass('show') && !$contentsContainer.hasClass('velocity-animating')){

            $contentsContainer.show();

            $contentsContainer.velocity( transitionObj, {
                duration: SIDEBAR_DEFAULT_ANIMATE_DURATION,
                complete: _callbackAnimation.bind(this, 'show', enableResizeBody, this.onShow),
                easing: SIDEBAR_DEFAULT_ERASING
            });

            // push 타입인 경우 바디도 animation 처리
            if (transitions === SIDEBAR_TRANSITIONS.PUSH){

                //push이며 enableResizeBody일 경우 body width 조절
                if (this[OPTS].enableResizeBody){

                    let fullSize = $(window).width();
                    // let calcSize = `calc(${fullSize} - ${opts.size}px)`;
                    let calcSize = `${fullSize - opts.size}`;

                    if (type === 'icon'){
                        // calcSize = `calc(${fullSize} - ${opts.size + SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE}px)`;
                        calcSize = `${fullSize - opts.size + SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE}`;
                    }

                    if (position === 'left' || position === 'right'){
                        $bodyContainer.css({
                            [SIDEBAR_ORIENTATION.HEIGHT]: '100%',
                            [SIDEBAR_ORIENTATION.WIDTH]: calcSize,
                        });
                    }
                    else{
                        $bodyContainer.css({
                            [SIDEBAR_ORIENTATION.WIDTH]: '100%',
                            [SIDEBAR_ORIENTATION.HEIGHT]: calcSize,
                        });
                    }
                }
                else{

                    let size = this[OPTS].size;

                    if (type === 'icon'){
                        size += SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE;
                    }

                    if (position === 'right') $bodyContainer.css({ [SIDEBAR_MARGIN.RIGHT] : size });
                    if (position === 'bottom') $bodyContainer.css({ [SIDEBAR_MARGIN.BOTTOM] : size });
                }

                $bodyContainer.velocity( transitionBodyObj, {
                    duration: SIDEBAR_DEFAULT_ANIMATE_DURATION,
                    complete: _callbackAnimation.bind(this, 'show', enableResizeBody),
                    easing: SIDEBAR_DEFAULT_ERASING
                });
            }

            if (!_.isNull(this.$dim) && type !== SIDEBAR_TYPE.FULL){
                $dim.show();
            }
        }

        // 선택한 contents 영역 이외엔 감추기
        $contentsContainer.children('.item-group-wrap').hide();

        // contents 영역만 보여주기
        $contentsContainer.children().eq(index).show();
    }
}

/**
 * Contents Container를 숨기기 위한 설정 및 animation 함수
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _hideContentsContainer(){

    let $contentsContainer = this.$contentsContainer;
    let $bodyContainer = this.$bodyContainer;
    let transitions = this[OPTS].transitions;
    let onlyIcon = this.isOnlyIcon;
    let position = this[OPTS].position;
    let enableResizeBody = this[OPTS].enableResizeBody;
    let type = this[OPTS].type;

    let transitionObj = _calcContainerTransition.call(this, 'hide', 'sidebar');
    let transitionBodyObj = _calcContainerTransition.call(this, 'hide', 'body');

    if (!onlyIcon){

        if (!$contentsContainer.hasClass('hide') && !$contentsContainer.hasClass('velocity-animating')) {
            // container animation
            $contentsContainer.velocity( transitionObj, {
                duration: SIDEBAR_DEFAULT_ANIMATE_DURATION,
                complete: _callbackAnimation.bind(this, 'hide', enableResizeBody, this.onHide),
                easing: SIDEBAR_DEFAULT_ERASING
            });

            // push 타입인 경우 바디도 animation 처리
            if (transitions === SIDEBAR_TRANSITIONS.PUSH){

                //push이며 enableResizeBody일 경우 body width 초기화
                if (this[OPTS].enableResizeBody){

                    $bodyContainer.css({
                        [SIDEBAR_ORIENTATION.HEIGHT]: ``,
                        [SIDEBAR_ORIENTATION.WIDTH]: ``,
                    });

                }
                else{

                    let size = 0;

                    if (type === 'icon'){
                        size += SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE;
                    }

                    if (position === 'right') $bodyContainer.css({ [SIDEBAR_MARGIN.RIGHT] : size });
                    if (position === 'bottom') $bodyContainer.css({ [SIDEBAR_MARGIN.BOTTOM] : size });
                }

                $bodyContainer.velocity( transitionBodyObj, {
                    duration: SIDEBAR_DEFAULT_ANIMATE_DURATION,
                    begin: ()=>{
                        if (type === SIDEBAR_TYPE.FULL) $bodyContainer.show();
                    },
                    complete: _callbackAnimation.bind(this, 'hide', enableResizeBody),
                    easing: SIDEBAR_DEFAULT_ERASING
                });
            }
        }

        // dim이 있는 경우 안보이게 하기
        if (!_.isNull(this.$dim) && this[OPTS].type !== SIDEBAR_TYPE.FULL) this.$dim.hide();
    }
}

/**
 * Jquery Animate에서 사용 될 움직이는 값을 계산하는 함수
 *
 * @param {String} [action = ''] - animate 상태 (show, hide)
 * @param {String} [target = ''] - animate 대상 (sidebar, body)
 *
 * @returns {Object} jquery animate에 사용될 움직이는 값들
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _calcContainerTransition(action = '', target = ''){

    let result = {};
    let options = this[OPTS];
    let orientation = this.orientation;
    let position = options.position;
    let reversePosition = this.reversePosition;
    let type = options.type;
    let enableResizeBody = options.enableResizeBody;

    let totalSize = _calcContainerWidth.call(this, options.type, options.size, position, target);

    if (target === 'sidebar' && action === 'show'){

        result[position] = 0;
    }
    else if (target === 'sidebar' && action === 'hide'){

        if (options.type === SIDEBAR_TYPE.ICON){
            totalSize -= SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE;
        }

        result[position] = '-=' + totalSize;
    }
    else if (target === 'body' && action === 'show'){

        if (position === 'right' || position === 'bottom'){

            if (enableResizeBody) result[reversePosition] = 0;
            else result[reversePosition] = '-=' + totalSize;

        }
        else{
            result[position] = '+=' + totalSize;
        }

    }
    else if (target === 'body' && action === 'hide'){

        if (position === 'right' || position === 'bottom'){

            result[reversePosition] = 0;

        }
        else{
            result[position] = options.type === SIDEBAR_TYPE.ICON ? SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE : 0;
        }


    }

    return result;
}

/**
 * Contents Container 너비값을 계산하는 함수
 *
 * @param {String} [type = ''] - Sidebar Type
 * @param {Number} [width = 0] - Sidebar width
 * @param {String} [position = ''] - Sidebar Position
 * @param {String} [target = ''] - Sidebar Target
 *
 * @returns {Number} - 계산된 Container 너비값
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _calcContainerWidth(type = '', width = 0, position = '', target = ''){

    let totalWidth = width;
    let orientation = this.orientation;

    if (type === SIDEBAR_TYPE.FULL){
        if (orientation === 'width'){
            totalWidth = $(document).width();
        }
        else {
            totalWidth = $(document).height();
        }
    }

    return totalWidth;
}

/**
 * Contents Container Animation 후 실행 되는 함수
 *
 * @param {String} [type = ''] - animation 상태
 * @param {Function} [callback = function(){}] - animation 후 실행할 callback 함수
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _callbackAnimation(type = '', enableResizeBody = false, callback = function(){}){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxSidebar, 'animation move type', 'all');
    }

    if (!_.isFunction(callback)){
        callback = function(){};
    }

    let $contentsContainer = this.$contentsContainer;
    let $bodyContainer = this.$bodyContainer;

    if (type === 'show'){

        $contentsContainer.removeClass('hide').addClass('show');
    }
    else {
        let totalSize = _calcContainerWidth.call(this, this[OPTS].type, this[OPTS].size, this[OPTS].position);
        let positionStartPx = this.positionStartPx;
        let hiddenSize = totalSize - positionStartPx;

        $contentsContainer.css(this[OPTS].position, `-${hiddenSize}px`);
        $contentsContainer.removeClass('show').addClass('hide');
        $contentsContainer.hide();
    }

    callback.call(this);
}

/**
 * Icon type일 때 Contents Container를 그릴 것이 있는지 유무 확인 함수
 *
 * @param {} [items] - Sidebar Items
 * @param {String} [type = ''] - Sidebar type
 *
 * @returns {Boolean} - 확인된 Container값
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _checkOnlyIcon(items, type = ''){

    let templateUrl = items;
    let child = items[0].items;

    try{

        if (type === SIDEBAR_TYPE.ICON){

            if (_.isString(templateUrl) && templateUrl.length) return false;

            if (_.isArray(child) && child.length) return false;

            if ($(child).length) return false;

            return true;
        }
        else {
            return false;
        }
    }
    catch (e){
        return true;
    }

}

/**
 * item Type을 구별하는 함수
 *
 * @param {String} [item = null] - check할 string값
 *
 * @return {String} - templateUrl 확인된 값
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _checkItemType(item = null){

    let htmlReg = /\.(html)/;

    try{

        if (htmlReg.test(item)) return 'url';

        if (_.isArray(item)) return 'item';

        if ($(item).length) return 'selector';

        return '';
    }
    catch (e){
        return '';
    }
}

/**
 * Position 반대방향을 구별하는 함수
 *
 * @param {String} [position = ''] - check할 position 값
 *
 * @return {String} - position 반대방향
 *
 * @private
 *
 * @memberOf BxSidebar
 */
function _getReversePosition(position = ''){

    switch (position){
        case SIDEBAR_POSITION.LEFT : return SIDEBAR_POSITION.RIGHT;
        case SIDEBAR_POSITION.RIGHT : return SIDEBAR_POSITION.LEFT;
        case SIDEBAR_POSITION.TOP : return SIDEBAR_POSITION.BOTTOM;
        case SIDEBAR_POSITION.BOTTOM : return SIDEBAR_POSITION.TOP;
    }

    return position;
}

/**
 * 브라우저 resize 이벤트 등록
 *
 * @private
 */
function _setResizeEventListener(){

    let $window = $(window);
    let $bodyContainer = this.$bodyContainer;
    let $contentsContainer = this.$contentsContainer;
    let opts = this[OPTS];

    let position = opts.position;
    let type = opts.type;
    let containerSize = 0;

    // if ($contentsContainer.hasClass('show')){
    //     containerSize += opts.size;
    // }

    if (type === SIDEBAR_TYPE.ICON){
        containerSize += SIDEBAR_DEFAULT_ICON_CONTAINER_SIZE;
    }

    $window.resize(()=>{

        let calcSize = 0;

        if (position === SIDEBAR_POSITION.LEFT || position === SIDEBAR_POSITION.RIGHT){

            calcSize = $(window).width() - containerSize;
            $bodyContainer.css(SIDEBAR_ORIENTATION.WIDTH, calcSize);
        }
        else {

            calcSize = $(window).height() - containerSize;
            $bodyContainer.css(SIDEBAR_ORIENTATION.HEIGHT, calcSize);
        }
    });
}

module.exports = BxSidebar;
