// bxuip-angular.js-ui/src/lib/bx-ui/app/components/tab.js
/**
 * 탭 선택 시, 호출될 callback 함수
 *
 * @callback onActive
 * @param {object} event - 이벤트 객체
 * @param {object} $newTab - 현재 선택된 tab 엘리먼트
 * @param {object} $newPanel - 현재 선택된 panel 엘리먼트
 * @param {object} $oldTab - 이전 선택된 tab 엘리먼트
 * @param {object} $oldPanel - 이전 선택된 panel 엘리먼트
 * @param {object} newIndex - 현재 선택된 index
 * @param {object} oldIndex - 이전 선택된 index
 *
 * @memberOf BxTab
 */

/**
 * 랜더링 완료 시, 호출될 callback 함수
 *
 * @callback onRender
 *
 * @memberOf BxTab
 */

// @param {number} maxTabCounter - 최대 탭의 수를 설정 할 수 있다.

//* @param {string} [OPTS.heightStyle='content'] - 'auto','fill','content'로 값을 설정할 수 있으며,
// * 'auto'는 가장 큰 컨텐츠의 높이에 맞춰서 모든 탭의 높이를 일정하게 반환하며,
// * 'fill'은 컨테이너의 크기에 맞춰서 반환하며, 'content'는 해당 패널에 적합한 크기를 반환한다.
/**
 * @class BxTab
 * @classDesc BxTab 클래스 생성자
 *
 * @param {object} OPTS
 * @param {object[]} [OPTS.items = [] ] 
 * @param {string} OPTS.items.id - 탭 아이디
 * @param {string} OPTS.items.title - 탭 제목
 * @param {string} OPTS.items.content - <s>탭 내용</s><span class='important'> ( Added in: v3.0.0, Deprecated since: v3.22.0 )</span>
 * @param {string} OPTS.items.template - 탭 내용
 * @param {string} OPTS.items.templateUrl - 탭 내용이 포함된 Url
 * @param {number} [OPTS.height = 0] - 컨텐츠 높이
 * @param {string} [OPTS.event = 'click'] - 사용될 이벤트명 ('click', 'mouseover')
 *
 * @param {Function} [OPTS.onActivate = function(){}] - 탭 선택 시, 호출될 callback 함수
 * @param {Function} [OPTS.onRender = function(){}] - 랜더링 완료 시, 호출될 callback 함수
 *
 * @returns {object} 생성된 인스턴스
 * @example
 *

// 1. template 예제

 $('#tab').BxTab({
    items :[
        {title:'탭 목록1', template:'<h1>탭 내용 1</h1><br/><br/>', id:1},
        {title:'탭 목록2', template:'<h1>탭 내용 2</h1><br/><br/>', id:2},
        {title:'탭 목록3', template:'<h1>탭 내용 3</h1><br/><br/>', id:3}
    ],
    event : 'click',
    onActivate : function(e, $newTab, $newPanel, $oldTab, $oldPanel, newIndex, oldIndex){
        console.log('active');
    },
    onRender : function(){
        console.log('Render complete');
    },
    height : 300
}).render();

----------------------------------------------------------------------

// 2. templateUrl 예제

$('#tab').BxTab({
    items :[
        {title:'탭 목록1', templateUrl:'tab1.html', id:1},
        {title:'탭 목록2', templateUrl:'tab2.html', id:2},
        {title:'탭 목록3', templateUrl:'tab3.html', id:3}
    ],
    event : 'click',
    onActivate : function(e, $newTab, $newPanel, $oldTab, $oldPanel, newIndex, oldIndex){
        console.log('active');
    },
    onRender : function(){
        console.log('Render complete');
    },
    height : 300
}).render();

 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

class BxTab {

    constructor({
        items = [],
        height = 0,
        event = 'click',
        // content / fill 로 지정
        heightStyle = 'fill',
        maxTabCounter = 10,
        onActivate = function(){},
        onRender = function(){}
    } = {}){

        if (!_.isArray(items)){
            CONSOLE.prop.throw(BxTab, 'items', 'type');
        }

        if (!_.isNumber(height)){
            CONSOLE.prop.throw(BxTab, 'height', 'type');
        }

        if (!_.isString(event)){
            CONSOLE.prop.throw(BxTab, 'event', 'type');
        }

        if (!_.isString(heightStyle)){
            CONSOLE.prop.throw(BxTab, 'heightStyle', 'type');
        }

        if (!_.isNumber(maxTabCounter)){
            CONSOLE.prop.throw(BxTab, 'maxTabCounter', 'type');
        }

        if (!_.isFunction(onActivate)){
            CONSOLE.prop.throw(BxTab, 'onActivate', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxTab, 'onRender', 'type');
        }

        this.$root = BxTab.$root;

        this[OPTS] = {event, heightStyle};

        // 추후, add tab 기능 추가시에 현재 tabCounter와 maxTabCounter를 이용해
        // 생성되는 텝을 제한.
        // this.tabCounter = items.length;
        // this.maxTabCounter = maxTabCounter;

        this.height = height;
        this.items = items;

        this.onActivate = onActivate;
        this.onRender = onRender;
    }

    /**
     * 렌더 함수
     *
     * @param {object} [$compile = null] - template 에 Angular $scope 을 바인딩 시키기위해서는, $compile 서비스를 전달해야한다.
     * @param {object} [$scope = null] - template 에 Angular $scope 을 바인딩 시키기위해서는, $scope 을 전달해야한다.
     *
     * @returns {object} 생성된 인스턴스
     */
    render($compile = null, $scope = null){

        let $container = this.$container = BASE.appendContainerElement(this.$root, BxTab.className);

        let html  = _parseToHTML(this.items);

        $container.append(html);

        // 이 부분은 디렉티브로 제거할 수 없다.
        // 이유는 차후(각 입력 컴포넌트들에게 state 기능이 생겼을때)에는 DOM 포퍼먼스 문제(여러 탭 생성 시, 초기 로드 문제와, 다수의 DOM 이 누적되어 느려지는 현상)
        // 를 해결하기위해, 동적 페이징 로드를 적용해야하기때문이다.
        // 결국 그때는 (각 템플릿 파일에서 개별 컨트롤러를 사용하기 위해)페이퍼와 같이 render 함수를 통해, $compile 및 $scope 객체를 전달받아 사용해야할것이다.
        if (UTIL.isAngular() && $compile && $scope){
            BASE.ngCompile.apply(this, [$container, $compile, $scope]);
        }

        _addEvent.call(this);

        _initTab.call(this);

        return this;
    }

    /**
     * 템플릿 바인딩 함수
     *
     * @return {object} 생성된 인스턴스
     */
    bind(){

        /*
         내부 바인드 함수
         */
        let _bind = () => {

            let html = this.$root.children();

            this.$container = BASE.appendContainerElement(this.$root, BxTab.className);

            this.$container.html(html);

            _addEvent.call(this);

            _initTab.call(this);
        };

        window.setTimeout(() => _bind(), 1);

        return this;
    }

    /**
     * 새로고침 함수
     *
     * @returns {object} 생성된 인스턴스
     * @private
     */
    refresh(){

        this.$container.tabs('refresh');
        return this;
    }
}

/**
 * 탭 생성함수
 * @private
 */
function _initTab(){

    if (this.height !== 0) {

        this.$container.height(this.height);
    }

    this.$container.tabs(this[OPTS]);

    // this[OPTS].load가 작동하지 않아, custom 함수를 사용하도록 한다.
    _onRender.call(this, () => {
        this.onRender.apply(this);
    });
}

/**
 * 이벤트 추가 함수
 *
 * @private
 */
function _addEvent(){
    // onActive 설정.
    this[OPTS].activate = (e, ui) => {

        let $newTab = ui.newTab;
        let $oldTab = ui.oldTab;

        let $newPanel = ui.newPanel;
        let $oldPanel = ui.oldPanel;

        let newIndex = $newPanel.index();
        let oldIndex = $oldPanel.index();

        this.onActivate.apply(this, [e, $newTab, $newPanel, $oldTab, $oldPanel, newIndex, oldIndex]);

    };
    
}

/**
 * 랜더링 완료 callback 함수
 *
 * @param {function} success - 성공시 실행할 function
 * @param {function} fail - 실패시 실행할 function
 * @private
 * @memberOf BxTab
 */
function _onRender(success = function(){}, fail = function(){}){
    _checkElementSize(this.$container.find('.ui-tabs-panel'), this.items, success, fail);
}

/**
 * 입력받은 items를 HTML tag로 변환해서 반환한다.
 *
 * @param {array} [items=[]] = items
 * @returns {string}
 * @private
 * @memberOf BxTab
 */
function _parseToHTML(items = []){

    let ul = [];
    let divs = [];

    ul.push('<ul>');

    _.map(items, function(item, idx){

        let title = item.title || '';
        let content = item.content || item.template || '';
        let templateUrl = item.templateUrl || '';

        if (!_.isString(title)){
            CONSOLE.prop.throw(BxTab, 'item.title', 'type');
        }

        if (!_.isString(content)){
            CONSOLE.prop.throw(BxTab, 'item.content(or template)', 'type');
        }

        if (!_.isString(templateUrl)){
            CONSOLE.prop.throw(BxTab, 'item.templateUrl', 'type');
        }

        if (!_.isEmpty(templateUrl)){
            content = _getTemplate(templateUrl);
        }

        ul.push(`<li><a href="#${idx}">${title}<span class="triangle"></span></a></li>`);
        divs.push(`<div id="${idx}">${content}</div>`);
    });

    ul.push('</ul>');


    return ul.concat(divs).join('');
}

/**
 * 렌더링 이벤트를 명시적으로 제공하는 않는 라이브러리를 위해, 전달받은 엘리먼트 size 와 item length 를 비교 후, onRender 시점을 찾아낸다.
 *
 * @param {object} [$elem=null] - 엘리먼트 집합
 * @param {array} [items=[]] - 아이템 집합
 * @param {function} [success=function(){}] - 렌더 완료 시, 호출될 callback 함수
 * @param {function} [fail=function(){}] - 제한 시간안에 렌더되지않았을때, 호출될 callback 함수
 * @param {int} [mTime=3000] - 렌더 제한 시간
 *
 * @returns {string} - 생성된 컴포넌트 클래스 이름
 *
 * @private
 *
 * @memberof BxTab
 */
function _checkElementSize($elem = null, items = [], success = function (){}, fail = function (){}, mTime = 3000){

    if (_.isNull($elem)){
        console.error('$elem 값이 존재하지 않습니다.');
        return this;
    }

    if (!_.isArray(items)){
        console.error('items 값이 존재하지 않습니다.');
        return this;
    }

    let sTime = new Date().getTime();

    let timerId = window.setInterval(() =>{

        let eTime = new Date().getTime() - sTime;

        let $elemSize = $elem.length;
        let itemsLength = items.length;

        if ($elemSize >= itemsLength){
            console.log(`$elemSize:${$elemSize},itemsLength:${itemsLength}`);
            window.clearInterval(timerId);
            success();
        }

        if (eTime >= mTime){
            console.log(`eTime${eTime},mTime:${mTime}`);
            window.clearInterval(timerId);
            fail();
        }

    }, 300);


    return this;
}

/**
 * templateUrl 반환 함수.
 *
 * @returns {string} - template
 *
 * @private
 *
 * @memberOf BxTab
 */
function _getTemplate(url){

    let template = '';

    $.ajax({
        url: url,
        success: (data) =>{
            template = data;
        },
        contentType: 'text/plain',
        // 동기식 처리
        async: false
    });

    return template;
}


module.exports = BxTab;
