// bxuip-angular.js-ui/src/lib/bx-ui/app/components/ticTac.js
/**
 * @class BxTicTac
 * @classDesc BxTicTac 클래스 생성자
 *
 * @example
 *
 * $().BxTicTac({
 *     position : 'BC', //'TL', 'TC', 'TR','BL', 'BC', 'BR'
 *     duration : 0,
 *     zIndex : null,
 *     offsetX : 0,
 *     offsetY : 0,
 *     spacing:10
 * }).success({
 *     title: 'Success Message',
 *     content: 'success',
 *     onHide :function(){console.log('hide');},
 *     onClick:function(){console.log('click');}
 * });
 *
 * @param {object} OPTS
 * @param {string} [OPTS.position='BC'] - 위치 (TL, TC, TR, BL, BC, BR)
 * @param {int} [OPTS.duration=0] - ticatac이 보여질 시간 설정. 0일 경우 자동으로 닫히지 않음
 * @param {int} [OPTS.zIndex=null] - z-index
 * @param {int} [OPTS.offsetX=0] - x축 offset
 * @param {int} [OPTS.offsetY=0] - y축 offset
 * @param {int} [OPTS.spacing=10] - tictac들 사이의 높이
 *
 * @returns {object} 생성된 인스턴스
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const OPTS = Symbol();

const TICTAC_SELECTOR = {
    custom: 'custom',
    ticTac: 'bx-tictac',
    message: 'bx-tictac-message',
    icon: 'bx-tictac-icon',
    addIcon: 'add-icon',
    ticTacClass: '.bx-tictac',
};

const MESSAGE_TYPE = {
    error: 'error',
    success: 'success',
    warning: 'warning',
    info: 'info',
    custom: 'custom'
};

const ANIMATE_TYPE = {
    open: 'addItem',
    close: 'removeItem',
    top: 'top',
    bottom: 'bottom'
};

const POSITION = {
    tl: 'TL',
    tc: 'TC',
    tr: 'TR',
    bl: 'BL',
    bc: 'BC',
    br: 'BR'
};

class BxTicTac{

    constructor({
        position = POSITION.bc, //'TL', 'TC', 'TR','BL', 'BC', 'BR'
        duration = 3000,
        zIndex = null,
        offsetX = 0,
        offsetY = 0,
        spacing=10
    } = {}
    ){

        if (!_.isString(position)) {
            CONSOLE.prop.throw(BxTicTac, 'position', 'type');
        }

        if (_.indexOf(Object.values(POSITION), position) === -1) {
            CONSOLE.prop.throw(BxTicTac, 'position', 'value');
        }

        if (!_.isNumber(duration)) {
            CONSOLE.prop.throw(BxTicTac, 'duration', 'type');
        }

        if (!_.isNull(zIndex) && !_.isNumber(zIndex)) {
            CONSOLE.prop.throw(BxTicTac, 'zIndex', 'type');
        }

        if (!_.isNumber(offsetX)) {
            CONSOLE.prop.throw(BxTicTac, 'offsetX', 'type');
        }

        if (!_.isNumber(offsetY)) {
            CONSOLE.prop.throw(BxTicTac, 'offsetY', 'type');
        }

        if (!_.isNumber(spacing)) {
            CONSOLE.prop.throw(BxTicTac, 'spacing', 'type');
        }

        this[OPTS] = _.assign({}, {position, duration, zIndex, offsetX, offsetY, spacing});
    }

    /**
     * success message
     *
     * @param {object} opts
     * @param {string} [opts.title = 'success'] - 제목
     * @param {string} [opts.content = ''] - 내용
     * @param {function} [opts.onHide = function(){}] - 닫힐 시, 호출될 callback 함수
     * @param {function} [opts.onClick = function(){}] - 클릭 시, 호출될 callback 함수
     *
     * @returns {object} 생성된 엘리먼트
     *
     * @memberOf BxTicTac
     */
    success(...args) {

        return _renderDOM.call(this[OPTS], MESSAGE_TYPE.success, ...args);
    }
    /**
     * error message
     *
     * @param {object} opts
     * @param {string} [opts.title = 'error'] - 제목
     * @param {string} [opts.content = ''] - 내용
     * @param {function} [opts.onHide =function(){}] - 닫힐 시, 호출될 callback 함수
     * @param {function} [opts.onClick = function(){}] - 클릭 시, 호출될 callback 함수
     * @returns {object} 생성된 엘리먼트
     *
     * @memberOf BxTicTac
     */
    error(...args) {

        return _renderDOM.call(this[OPTS], MESSAGE_TYPE.error, ...args);
    }
    /**
     * warning message
     *
     * @param {object} opts
     * @param {string} [opts.title = 'waring'] - 제목
     * @param {string} [opts.content = ''] - 내용
     * @param {function} [opts.onHide =function(){}] - 닫힐 시, 호출될 callback 함수
     * @param {function} [opts.onClick = function(){}] - 클릭 시, 호출될 callback 함수
     * @returns {object} 생성된 엘리먼트
     *
     * @memberOf BxTicTac
     */
    warning(...args) {

        return _renderDOM.call(this[OPTS], MESSAGE_TYPE.warning, ...args);
    }
    /**
     * info message
     *
     * @param {object} opts
     * @param {string} [opts.title = 'info'] - 제목
     * @param {string} [opts.content = ''] - 내용
     * @param {function} [opts.onHide =function(){}] - 닫힐 시, 호출될 callback 함수
     * @param {function} [opts.onClick = function(){}] - 클릭 시, 호출될 callback 함수
     * @returns {object} 생성된 엘리먼트
     *
     * @memberOf BxTicTac
     */
    info(...args) {

        return _renderDOM.call(this[OPTS], MESSAGE_TYPE.info, ...args);
    }
    /**
     * custom message
     *
     * @param {object} opts
     * @param {string} [opts.title = 'custom'] - 제목
     * @param {string} [opts.content = ''] - 내용
     * @param {string} [opts.iconSrc = ''] - url 입력시 아이콘 변경
     * @param {function} [opts.onHide =function(){}] - 닫힐 시, 호출될 callback 함수
     * @param {function} [opts.onClick = function(){}] - 클릭 시, 호출될 callback 함수
     * @returns {object} 생성된 엘리먼트
     *
     * @memberOf BxTicTac
     */
    custom(...args) {

        return _renderDOM.call(this[OPTS], MESSAGE_TYPE.custom, ...args);
    }
}

/**
 * type에 맞게 Dom을 생성
 *
 * @param {string} type - 메시지 타입
 * @param {object} opts - 옵션 정보
 *
 * @private
 * @memberOf BxTicTac
 */
function _renderDOM(
    type = '', {
        title = type, // 기본 값은 타입 이름
        content = '',
        iconSrc = null,
        onHide = function() {}, // 토스트 창이 닫혔을 때
        onClick = function() {} // 컨테이너를 클릭 했을때
    } = {}
    ){

    if (!_.isString(title)) {
        CONSOLE.prop.throw(BxTicTac, 'title', 'type');
    }

    if (!_.isString(content)) {
        CONSOLE.prop.throw(BxTicTac, 'content', 'type');
    }

    if (!_.isFunction(onHide)) {
        CONSOLE.prop.throw(BxTicTac, 'onHide', 'type');
    }

    if (!_.isFunction(onClick)) {
        CONSOLE.prop.throw(BxTicTac, 'onClick', 'type');
    }

    let position = this.position;
    let spacing = this.spacing;
    let offsetX = this.offsetX;
    let $container = BASE.appendContainerElement(BxTicTac.$root, BxTicTac.className);
    let titleHtml = `<strong class="title">${title}</strong>`;
    let contentHtml = `<p class="desc">${content}</p>`;
    let customHtml = '';
    let img = null;

    if (type === MESSAGE_TYPE.custom) {

        if (!_.isString(iconSrc) && !_.isNull(iconSrc)) {
            CONSOLE.prop.throw(BxTicTac, 'iconSrc', 'type');
        }

        if (!_.isNull(iconSrc)) {
            img = `<div class="custom-img" style="background-image:url(${iconSrc})"></div>`;
        }
        else {
            img = `<div class="custom-img"></div>`;
        }

        customHtml = `<div class=${TICTAC_SELECTOR.icon}>${img}</div>`;
    }

    let html = `<div class="${TICTAC_SELECTOR.ticTac} ${type} ${TICTAC_SELECTOR.addIcon}" data-position=${position} data-offsetX=${offsetX} data-spacing=${spacing}> ${customHtml} <div class="${TICTAC_SELECTOR.message}">${titleHtml}${contentHtml}</div></div>`;
    let style = _getStyleOffset.call(this, position);
    let $component = $(html).css(style);

    $container.append($component);

    _setRePosition.call(this, ANIMATE_TYPE.open);
    _setEventHandler.call(this, $component, {onHide , onClick});

    return $component.get(0);
}

/**
 * 위치에 맞게 offset 을 계산
 *
 * @param {string} position - 위치
 *
 * @private
 * @memberOf BxTicTac
 */
function _getStyleOffset(position = POSITION.bc) {

    let offsetX = this.offsetX;
    let zIndex = this.zIndex;
    let isLeftDirection = true;
    let isTopDirection = true;
    let isCenterDirection = false;

    let style = {
        'left' : null,
        'right' : null,
        'margin' : null,
        'width' : 300,
        'opacity' : 0,
        'z-index' : zIndex
    };

    switch (position) {
        case POSITION.bl :
        case POSITION.bc :
        case POSITION.br :
            isTopDirection = false;
            break;
    }

    switch (position) {
        case POSITION.tl :
        case POSITION.bl :
            style.left = offsetX;
            break;
        case POSITION.tc :
        case POSITION.bc :
            style.left = offsetX;
            style.right = 0;
            style.margin = '0px auto';
            isCenterDirection = true;
            break;
        case POSITION.tr :
        case POSITION.br :
            isLeftDirection = false;
            style.right = offsetX;
            break;
    }

    this.isTopDirection = isTopDirection;
    this.isLeftDirection = isLeftDirection;
    this.isCenterDirection = isCenterDirection;

    return style;
}

/**
 * message 위치를 재 정렬
 *
 * @param {string} type - 애니메이션 종류
 *
 * @private
 * @memberOf BxTicTac
 */
function _setRePosition(type = ANIMATE_TYPE.open) {

    let _this = this;
    let spacing = this.spacing;
    let offsetX = this.offsetX;
    let offsetY = this.offsetY;
    let position = this.position;
    let isTopDirection = this.isTopDirection;
    let selector = `${TICTAC_SELECTOR.ticTacClass}[data-position=${position}]`;
    // 신규 메시지는 항상 위에서 쌓인다.
    let $elements = isTopDirection ? $($(selector).get().reverse()) : $(selector);
    let isFirstElemAni = false;
    let lastLength = $elements.length - 1;

    $elements.each(function(i) {

        let $component = $(this);

        if (isTopDirection) {

            isFirstElemAni = i === 0 && type === ANIMATE_TYPE.open;

            _animateComponent.call(_this, $component, ANIMATE_TYPE.top, offsetX, offsetY, isFirstElemAni);
        }
        else {

            isFirstElemAni = i === lastLength && type === ANIMATE_TYPE.open;

            _animateComponent.call(_this, $component, ANIMATE_TYPE.bottom, offsetX, offsetY, isFirstElemAni);
        }

        offsetY = (parseInt(offsetY) + parseInt(spacing)) + $(this).outerHeight();
    });
}

/**
 * ticTac event 바인드
 *
 * @param {object} $target jquery element
 * @param {function}[onClick=function(){}] click callback 함수
 * @param {function}[onHide=function(){}] hide callback 함수
 *
 * @private
 * @memberOf BxTicTac
 */
function _setEventHandler($target = null, {onClick = function(){}, onHide = function (){}}={}){

    _.assign(this, {onClick, onHide});

    _setClickEvent.call(this, $target);
    _setDelayEvent.call(this, $target);
}

/**
 * ticTac message 에 클릭시 메시지가 닫히는 이벤트를 등록
 *
 * @param {object} $target jquery element
 *
 * @private
 * @memberOf BxTicTac
 */
function _setClickEvent($target = null) {

    let _this = this;

    $target.click((e) => {

        _this.onClick.call(_this, e, $target);

        _hideMessage.call(_this, $target);
    });
}

/**
 * ticTac message 에 delay 이후 닫히는 이벤트 등록
 *
 * @param {object} $target jquery element
 *
 * @private
 * @memberOf BxTicTac
 */
function _setDelayEvent($target = null) {

    let duration = this.duration;
    let delayEvent = null;
    let _this = this;

    let hideMessage = function() {
        _hideMessage.call(_this, $target);
    };

    if (duration !== 0) {

        delayEvent = setTimeout(hideMessage , duration);

        $target.hover((e) => {
            switch (e.type) {
                case 'mouseenter' :
                    clearTimeout(delayEvent);
                    break;
                case 'mouseleave' :
                    delayEvent = setTimeout(hideMessage , duration);
                    break;
            }
        });
    }
}

/**
 * ticTac message 를 닫는다.
 *
 * @param {object} $target jquery element
 *
 * @private
 * @memberOf BxTicTac
 */
function _hideMessage($target = null) {

    this.onHide.call(this, $target);

    // 닫을떄 가장 마지막에 추가된 정보(offset, spacing)으로 재설정
    _setStateLastComponent.call(this);

    _animateComponent.call(this, $target, ANIMATE_TYPE.close);
}

/**
 * animation 을 컴포넌트에 적용
 *
 * @param {object} $component 애니메이션 대상
 * @param {string} type 애니메이션 타입
 * @param {int} offsetX offsetX 값
 * @param {int} offsetY offsetY 값
 * @param {boolean} isFirstComponent 첫번째 컴포넌트
 * @private
 * @memberOf BxTicTac
 */
function _animateComponent($component = {}, type = ANIMATE_TYPE.close, offsetX = 0, offsetY = 0, isFirstComponent = false) {

    const DELAY = 500;

    let baseX = 600;
    let aniOffsetX = {};
    let aniOffsetY = {};
    let aniBaseX = {};
    let isLeftDirection = this.isLeftDirection;
    let isCenterDirection = this.isCenterDirection;
    let _this = this;

    if (isLeftDirection) {
        _.assign(aniBaseX, {left: -baseX});
        _.assign(aniOffsetX, {left: offsetX});
    }
    else {
        _.assign(aniBaseX, {right: -baseX});
        _.assign( aniOffsetX, {right: offsetX});
    }

    type === ANIMATE_TYPE.top ? _.assign( aniOffsetY, {top: offsetY}) :  _.assign( aniOffsetY, {bottom: offsetY});

    let aniOffsetXY = _.assign(aniOffsetX, aniOffsetY);

    if (type === ANIMATE_TYPE.close) {

        if(isCenterDirection) {

            $component.parent().remove();
            _setRePosition.call(_this, type);
        }
        else {
            $component.velocity(
                aniBaseX,
                {
                    complete : function (){
                        $component.parent().remove();
                        _setRePosition.call(_this, type);
                    }
                }

            );
        }
    }
    else {

        if (isFirstComponent) {

            $component.velocity(
                {opacity:1},
                {
                    duration: DELAY,
                    begin: function () {

                        if (type === ANIMATE_TYPE.bottom) {
                            $component.css({bottom: offsetY});
                        }
                        else {
                            $component.css({top: offsetY});
                        }
                    },
                    progress: function (elements, complete) {

                        $component.css({opacity: complete});

                        if (!isCenterDirection) { // 가운대 일때는 애니메이션 없음

                            if (isLeftDirection) {
                                $component.css({left: -baseX + ((baseX + aniOffsetX.left) * complete)});
                            }
                            else {
                                $component.css({right: -baseX + ((baseX + aniOffsetX.right) * complete)});
                            }
                        }
                    }
                });
        }
        else {
            $component.css(aniOffsetXY);
        }
    }
}

/**
 * 가장 마지막에 추가된 아이템의 상태 값으로 재설정
 *
 * @private
 * @memberOf BxTicTac
 */
function _setStateLastComponent() {

    let $elem = null;
    let position = this.position;
    let selector = `${TICTAC_SELECTOR.ticTacClass}[data-position=${position}]`;
    let isTopDirection = this.isTopDirection;

    if (isTopDirection) {

        $elem = $(selector).last();
        this.offsetY = $elem.css('top');
    }
    else {

        $elem = $(selector).first();
        this.offsetY = $elem.css('bottom');
    }

    this.spacing = $elem.attr('data-spacing');
    this.offsetX = $elem.attr('data-offsetX');
}

module.exports = BxTicTac;
