// bxuip-angular.js-ui/src/lib/bx-ui/app/components/timePicker.js
/**
 * 시간 또는 meridian 변경 시, 호출될 callback 함수
 *
 * @callback onChanged
 * @param {object} time - 시간 데이터
 *
 * @memberOf BxTimePicker
 */

/**
 * show 메서드 호출 시, 호출될 callback 함수
 *
 * @callback onShow
 *
 * @memberOf BxTimePicker
 */

/**
 * hide 메서드 호출 시, 호출될 callback 함수
 *
 * @callback onHide
 *
 * @memberOf BxTimePicker
 */

/**
 * 컴포넌트 초기화 시, 호출될 callback 함수
 *
 * @callback onInitialize
 *
 * @memberOf BxTimePicker
 */


/**
 * @class BxTimePicker
 * @classDesc BxTimePicker 클래스 생성자
 * @param {object} OPTS
 * @param {string} [OPTS.type = 'default'] - timePicker 타입(default, bubble, round, line)
 * @param {boolean} [OPTS.show24Hours = true] - 24 시간제 사용 유/무(`false`: 12 시간제)
 * @param {int} [OPTS.minHours = 0] - 최소 시간(`default` type 은 적용되지 않는다)
 * @param {int} [OPTS.maxHours = 23]- 최대 시간(`default` type 은 적용되지 않는다)
 * @param {string} [OPTS.formatter = 각 type 에 따른, 기본 값] - `시간 표기`에 대한, formatter(h:m:s or hh:mm:ss)
 * @param {string} [OPTS.defaultTime = 각 시간제에 따른, 기본 값] - 기본 시간(12:00:00)
 * @param {object} [OPTS.steps]
 * @param {int} [OPTS.steps.hour = 1] - 시간 간격(`default` 타입은 적용되지 않는다)
 * @param {int} [OPTS.steps.min = 1] - 분 간격
 * @param {int} [OPTS.steps.sec = 1] - 초 간격(`default` 타입은 적용되지 않는다)

 * @param {function} [OPTS.onChanged = function(){}] - 시간 또는 meridian 변경 시, 호출될 callback 함수
 * @param {function} [OPTS.onShow = function(){}] - show 메서드 호출 시, 호출될 callback 함수
 * @param {function} [OPTS.onHide = function(){}] - hide 메서드 호출 시, 호출될 callback 함수
 * @param {function} [OPTS.onInitialize = function(){}] - <s>컴포넌트 초기화 시, 호출될 callback 함수</s> <span class='important'> ( Added in: v3.3.0, Deprecated since: v3.4.0 )</span>
 * @param {function} [OPTS.onRender = function(){}] - 컴포넌트 초기화 시, 호출될 callback 함수
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 $('#selector').BxTimePicker({
	type: 'bubble',
	show24Hours: true,
	minHours: 0,
	maxHours: 13,
	formatter: 'hh:mm:ss',
	defaultTime: '00:00:00',
	steps: {
	    hour: 3,
	    min: 30,
	    sec: 30
	},
	onChanged: function(time){
	    console.log(this);
	    console.log(time);
	},
	onShow: function(){
	    console.log(this);
	},
	onHide: function(){
	    console.log(this);
	},
	onRender: function(){
	    console.log(this);
	}
}).render();
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const OPTS = Symbol();

// 전역 메세지 객체
const MESSAGE = {
    "am": 'AM',
    "pm": 'PM',
    "emptyComponentError": '해당 컴포넌트가 존재하지 않습니다.',
    "notInputElementError": '$root 엘리먼트는 반드시, INPUT 엘리먼트여야 합니다.',
    "meridianValueError": '12 시간제가 아니거나, 값은 반드시 `AM` 또는 `PM` 이여야합니다.'
};

// 전역 클래스 객체
const CLASS_NAME = {
    "bxTimePickerWrap": 'bx-timepicker-wrap',
    "bxUiInput": '-bx-ui-input',
    "bxInput": 'bx-input',
    "bxTimePickerInput": 'bx-timepicker-input',
    "pickerInput": 'picker-input',
    "hour": 'hour',
    "min": 'min',
    "sec": 'sec',
    "btnUp": 'btn-up',
    "btnDown": 'btn-down',
    "bxCtrlBtn": 'bx-ctrl-btn',
    "meridian": 'meridian',
    "col1": 'col-1',
    "col2": 'col-2',
    "col3": 'col-3',
    "col4": 'col-4'
};

// 마지막으로 생성된 인스턴스 객체
let lastShowInstance = null;

class BxTimePicker {

    constructor({
        type = 'default',
        show24Hours = true,
        minHours = 0,
        maxHours = 23,
        formatter = '',
        defaultTime = '',
        steps = {
            hour: 1,
            min: 1,
            sec: 1
        },
        onChanged = function (){
        },
        onShow = function (){
        },
        onHide = function (){
        },
        // @deprecated.
        onInitialize = function (){
        },
        onRender = function (){
        }
    } = {}){

        if (!_.isString(type)){
            CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
        }
        else if (_.trim(type).length === 0){
            type = 'default';
        }

        if (!_.isNumber(minHours) || _.isNaN(minHours)){
            CONSOLE.prop.throw(BxTimePicker, 'minHours', 'type');
        }

        if (!_.isNumber(maxHours) || _.isNaN(maxHours)){
            CONSOLE.prop.throw(BxTimePicker, 'maxHours', 'type');
        }

        // 각 시간제(12 / 24)에 따라, 해당 formatter 를 초기화 시킨다.
        if (!_.isString(formatter)){
            CONSOLE.prop.throw(BxTimePicker, 'formatter', 'type');
        }
        else if (_.trim(formatter).length === 0){
            formatter = this.initFormatter(type);
        }

        if (!_.isString(defaultTime)){
            CONSOLE.prop.throw(BxTimePicker, 'defaultTime', 'type');
        }

        if (!_.isPlainObject(steps)){
            CONSOLE.prop.throw(BxTimePicker, 'steps', 'type');
        }

        if (!_.isFunction(onChanged)){
            CONSOLE.prop.throw(BxTimePicker, 'onChanged', 'type');
        }


        if (!_.isFunction(onShow)){
            CONSOLE.prop.throw(BxTimePicker, 'onShow', 'type');
        }

        if (!_.isFunction(onHide)){
            CONSOLE.prop.throw(BxTimePicker, 'onHide', 'type');
        }

        if (!_.isFunction(onInitialize)){
            CONSOLE.prop.throw(BxTimePicker, 'onInitialize', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxTimePicker, 'onRender', 'type');
        }

        let $root = BxTimePicker.$root;

        this[OPTS] = {
            type,
            show24Hours,
            minHours,
            maxHours,
            formatter,
            defaultTime,
            steps,
            onChanged,
            onShow,
            onHide,
            // @deprecated.
            onInitialize,
            onRender
        };

        // formatter 를 초기화 한다.
        this[OPTS].formatters = _initFormatters.call(this, formatter);

        this.$root = $root;
        this.$container = null;

        // `defaultTime` 이 한번이라도 설정됐는지 여부
        this.isSetDefaultTime = false;

        // 사용중인 `meridian` 정보
        this.useMeridian = {'AM': false, 'PM': false};

        // 시간에 대한, `min` / `max` 와 `useMeridian` 정보를 초기화 한다.
        this.hourLimit = _initMinMaxHourToUseMeridian.call(this);

        // `meridian` 상태(AM / PM)
        this.meridian = _initMeridian.call(this);

        // 시간 데이터 집합
        this.stores = {

            'value': {
                items: []
            },
            'hour': {
                // 24 시간제 아이템
                '24Items': [],
                // 12 시간제 AM 아이템
                '12AMItems': [],
                // 12 시간제 PM 아이템
                '12PMItems': []
            },
            'min': {
                items: []
            },
            'sec': {
                items: []
            },
            'meridian': {
                items: []
            }
        };
    }

    /**
     * 초기화 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    render(){

        this.initialize();

        return this;
    }

    /**
     * 초기화 함수
     *
     * @private
     * @returns {object} 생성된 인스턴스
     */
    initialize(){

        let opts = this[OPTS];

        let $root = this.$root;
        let className = BxTimePicker.className;

        // $root 엘리먼트가 `bx-input` 인지 아닌지를 반환한다.
        let isBxInput = _isBXInput($root.attr('class'));

        // $root 엘리먼트가 `bx-input` 이 아닌경우
        if (!isBxInput){
            if ($root.get(0).nodeName !== 'INPUT'){
                CONSOLE.prop.throw(BxTimePicker, '$root 엘리먼트는 반드시 `INPUT` 엘리먼트여야합니다');
            }
        }

        // $root 엘리먼트가 `bx-input` 인 아닐때, 래핑될, `component $root` 엘리먼트
        let $cRoot = $(`<div class="${CLASS_NAME.bxUiInput}"></div>`);

        // $root 엘리먼트가 `bx-input` 이 아닌 경우
        if (!isBxInput){

            // clone 메서드의 첫 번째 인수가 `true` 일 경우, 현재 바인딩된 `event handler` 들이 전부 복사된다.

            // 기존 $root 엘리먼트를 복사한 후, $cRoot 엘리먼트의 하위로 `append` 시킨다.
            let _$root = $root.clone(true);
            _$root.addClass(CLASS_NAME.bxTimePickerInput);

            // 기존 $root 엘리먼트를 대치시킨다.
            $root.replaceWith($cRoot.append(_$root));

            this.$root = $cRoot;
        }
        else{
            $root.find(`.${CLASS_NAME.bxInput}`).addClass(CLASS_NAME.bxTimePickerInput);
        }

        // 초기화 이벤트 바인딩
        _addInitEventListener.call(this);

        // $container 엘리먼트 생성
        this.$container = BASE.appendContainerElement(this.$root, className);
        this.$container.append(_getComponentHtml.call(this));

        // 컴포넌트 이벤트 바인딩
        _addEventListener.call(this);

        // @deprecated.(차후 제거될 이벤트 핸들러)
        opts.onInitialize.call(this);
        opts.onRender.call(this);

        return this;
    }

    /**
     *
     * 시간 데이터를 반환한다.
     *
     * @returns {object} 시간 데이터
     */
    getTime(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return null;
        }

        let time = _getTime.call(this);

        let hour = _zeroToTwelveHour.call(this, time.hour);

        time.value = _mergeTime(hour, time.min, time.sec);
        time.hour = hour;

        return time;
    }

    /**
     *
     * 24 시간제 유/무를 반환한다.
     *
     * @returns {boolean} 24 시간제 유/무
     */
    getShow24Hours(){
        return this[OPTS].show24Hours;
    }

    /**
     *
     * `AM` 유/무를 반환한다.
     *
     * @returns {boolean} AM 유/무
     */
    getAM(){

        let ret = false;

        // 12 시간제일 경우
        if (!this.getShow24Hours()){
            ret = (_getMeridian.call(this) === MESSAGE.am) ? true : false;
        }

        return ret;
    }

    /**
     *
     * `PM` 유/무를 반환한다.
     *
     * @returns {boolean} PM 유/무
     */
    getPM(){

        let ret = false;

        // 12 시간제일 경우
        if (!this.getShow24Hours()){
            ret = (_getMeridian.call(this) === MESSAGE.pm) ? true : false;
        }

        return ret;
    }

    /**
     *
     * 시간 값을 반환한다.
     *
     * @returns {string} 시간 값
     */
    getHour(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return '';
        }

        let v = _getTime.call(this);

        return v.hour;
    }

    /**
     *
     * 분 값을 반환한다.
     *
     * @returns {string} 분 값
     */
    getMin(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return '';
        }

        let v = _getTime.call(this);

        return v.min;
    }

    /**
     *
     * 초 값을 반환한다.
     *
     * @returns {string} 초 값
     */
    getSec(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return '';
        }

        let v = _getTime.call(this);

        return v.sec;
    }

    /**
     *
     * formatter 값을 반환한다.
     *
     * @returns {string} formatter 값
     */
    getFormatter(){
        return this[OPTS].formatter;
    }

    /**
     *
     * 시간 값을 설정한다.
     *
     * @param {string} [value = ''] - 시간 값(12:00:00 (`AM` / `PM`))
     *
     * @returns {object} 생성된 인스턴스
     */
    setTime(value = ''){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        if (!_.isString(value)){
            CONSOLE.prop.throw(BxTimePicker, 'value', 'all');
        }

        let values = _splitTime(value);

        let time = values.time;
        let meridian = _setMeridian.call(this, values.meridian);

        let pickerType = this[OPTS].type;

        // 정해진 `formatter` 를 통해, 시간값을 재 설정한다.
        let times = _convertTimeWithFormatter.call(this, time);

        let hour = times.hour;
        let min = times.min;
        let sec = times.sec;

        // 최종 할당될 meridian 값
        let _meridian = meridian || _getMeridian.call(this);

        if (
            pickerType === 'bubble' ||
            pickerType === 'round' ||
            pickerType === 'line'
        ){
            // 각 input 엘리먼트들에게 해당 시간 값을 할당한다.
            _setInputValues.call(this, hour, min, sec, _meridian);
        }

        // root input 엘리먼트에 값을 할당한다.
        _setRootInputValue.call(this, hour, min, sec, _meridian);

        this[OPTS].onChanged.call(this, this.getTime());

        return this;
    }

    /**
     *
     * `AM` 으로 설정한다.
     *
     * @returns {object} 생성된 인스턴스
     */
    setAM(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        let meridian = MESSAGE.am;

        if (_isMultiMeridian.call(this)){

            let values = _getTime.call(this);

            let time = _mergeTime(values.hour, values.min, values.sec, meridian);

            this.setTime(time);

            _setMeridian.call(this, meridian);
        }

        return this;
    }

    /**
     *
     * `PM` 으로 설정한다.
     *
     * @returns {object} 생성된 인스턴스
     */
    setPM(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        let meridian = MESSAGE.pm;

        if (_isMultiMeridian.call(this)){

            let values = _getTime.call(this);

            let time = _mergeTime(values.hour, values.min, values.sec, meridian);

            this.setTime(time);

            _setMeridian.call(this, meridian);
        }

        return this;
    }

    /**
     *
     * 시간 값을 설정한다.
     *
     * @param {int} value - 시간 값
     *
     * @returns {object} 생성된 인스턴스
     */
    setHour(value){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        if (!_.isNumber(value)){
            CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
        }

        let values = _getTime.call(this);

        let time = _mergeTime(_.toString(value), values.min, values.sec, values.meridian);

        this.setTime(time);

        return this;
    }

    /**
     *
     * 분 값을 설정한다.
     *
     * @param {int} value - 분 값
     *
     * @returns {object} 생성된 인스턴스
     */
    setMin(value){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        if (!_.isNumber(value)){
            CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
        }

        let values = _getTime.call(this);

        let time = _mergeTime(values.hour, _.toString(value), values.sec, values.meridian);

        this.setTime(time);

        return this;
    }

    /**
     *
     * 초 값을 설정한다.
     *
     * @param {int} value - 초 값
     *
     * @returns {object} 생성된 인스턴스
     */
    setSec(value){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        if (!_.isNumber(value)){
            CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
        }

        let values = _getTime.call(this);

        let time = _mergeTime(values.hour, values.min, _.toString(value), values.meridian);

        this.setTime(time);

        return this;
    }

    /**
     *
     * toggle 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    toggle(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        _toggle.call(this);

        return this;
    }

    /**
     *
     * show 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    show(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        _show.call(this);

        return this;
    }

    /**
     *
     * hide 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    hide(){

        if (_.isNull(this.$container)){
            console.warn(MESSAGE.emptyComponentError);
            return this;
        }

        _hide.call(this);

        return this;
    }

    /**
     * formatter 초기 값을 반환한다.
     *
     * @param {string} [type = ''] - picker 타입
     *
     * @returns {string} formatter
     *
     * @private
     *
     * @memberOf BxTimePicker
     */
    initFormatter(type = ''){

        if (!_.isString(type) || _.trim(type).length === 0){
            CONSOLE.prop.throw(BxTimePicker, 'type', 'all');
        }

        return _initFormatter(type);
    }
}

/**
 * 시간에 대한, min/max 및 useMeridian 정보를 초기화 한다.
 *
 * @returns {object} min/max 데이터
 *
 * @private
 *
 * @memberOf BxTimePicker
 */

function _initMinMaxHourToUseMeridian(){

    let ret = {};

    let opts = this[OPTS];

    let pickerType = opts.type;

    let show24Hours = opts.show24Hours;
    let minHours = opts.minHours;
    let maxHours = opts.maxHours;

    // `default` 타입인 경우
    if (pickerType === 'default'){

        // `default` 타입은 min / max 가 적용되지 않는다(즉 0 ~ 23 범위로 fixed 된다)
        ret['__24MinMax__'] = {
            min: 0,
            max: 23
        };

        // 12 시간제인 경우
        if (!show24Hours){

            ret['__12AmMinMax__'] = {
                min: 0,
                max: 12
            };

            ret['__12PmMinMax__'] = {
                min: 0,
                max: 12
            };

            this.useMeridian[MESSAGE.am] = true;
            this.useMeridian[MESSAGE.pm] = true;
        }
    }
    else{

        // `default` 타입을 제외한 모든 타

        // minHours 가 최소 시간 보다 작을 경우
        if (minHours < 0) minHours = 0;
        // minHours 가 최대 시간 보다 클 경우
        if (minHours > 23) minHours = 23;


        // maxHours 가 최소 시간 보다 작을 경우
        if (maxHours < 0) maxHours = 0;
        // maxHours 가 최대 시간 보다 클 경우
        if (maxHours > 23) maxHours = 23;

        // maxHours 보다 minHours 가 클 경우
        if (minHours > maxHours) minHours = maxHours;

        // 24 시간제 데이터는 무조건 초기화된다.
        ret['__24MinMax__'] = {
            min: minHours,
            max: maxHours
        };

        // 12 시간제
        if (!show24Hours){

            // AM 만 있는 경우
            if (minHours < 12 && maxHours < 12){

                ret['__12AmMinMax__'] = {
                    min: minHours,
                    max: maxHours
                };

                this.useMeridian[MESSAGE.am] = true;
            }

            // PM 만 있는 경우
            if (minHours > 11 && maxHours > 11){

                ret['__12PmMinMax__'] = {
                    min: _to12Hour.call(this, minHours),
                    max: _to12Hour.call(this, maxHours)
                };

                this.useMeridian[MESSAGE.pm] = true;
            }

            // AM / PM 이 있는 경우
            if (minHours < 12 && maxHours > 11){

                ret['__12AmMinMax__'] = {
                    min: minHours,
                    max: 11
                };

                ret['__12PmMinMax__'] = {
                    min: 0,
                    max: _to12Hour.call(this, maxHours)
                };

                this.useMeridian[MESSAGE.am] = true;
                this.useMeridian[MESSAGE.pm] = true;
            }
        }
    }

    return ret;
}

/**
 * formatter 초기 값을 반환한다.
 *
 * @param {string} [type = ''] - picker 타입
 *
 * @returns {string} formatter
 *
 * @private
 *
 * @memberOf BxTimePicker
 */

function _initFormatter(type = ''){
    return type === 'default' ? 'hh:mm' : 'hh:mm:ss';
}


/**
 * formatters 객체를 초기화 시킨다.
 *
 * @param {string} [formatter = ''] - formatter
 *
 * @returns {object} formatters 객체
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _initFormatters(formatter = ''){

    if (!_.isString(formatter)){
        CONSOLE.prop.throw(BxTimePicker, 'formatter', 'type');
    }

    let inputType = this[OPTS].type;

    // 초기화 객체
    let ret = {

        hour: {
            enableZero: false,
            isShow: false
        },
        min: {
            enableZero: false,
            isShow: false
        },
        sec: {
            enableZero: false,
            isShow: false
        }
    };

    let formatters = formatter.split(':');

    if (formatters.length > 3){
        CONSOLE.prop.throw(BxTimePicker, 'formatter', 'type');
    }

    // 시간에 대한, format 이 있을 경우
    if (formatters.length >= 1){

        let hourFormat = formatters[0];

        if (hourFormat !== 'h' && hourFormat !== 'hh'){
            CONSOLE.prop.throw(BxTimePicker, 'formatter.(h or hh)', 'type');
        }

        if (hourFormat === 'hh') ret.hour.enableZero = true;

        ret.hour.isShow = true;
    }

    // 분에 대한, format 이 있을 경우
    if (formatters.length >= 2){

        let minFormat = formatters[1];

        if (minFormat !== 'm' && minFormat !== 'mm'){
            CONSOLE.prop.throw(BxTimePicker, 'formatter.(m or mm)', 'type');
        }

        if (minFormat === 'mm') ret.min.enableZero = true;

        ret.min.isShow = true;
    }

    // 초에 대한, format 이 있을 경우
    // defalut type 은 초 단위가 존재하지 않는다.
    if (inputType !== 'default' && formatters.length === 3){

        let secFormat = formatters[2];

        if (secFormat !== 's' && secFormat !== 'ss'){
            CONSOLE.prop.throw(BxTimePicker, 'formatter.(s or ss)', 'type');
        }

        if (secFormat === 'ss') ret.sec.enableZero = true;

        ret.sec.isShow = true;
    }

    return ret;
}

/**
 * 컴포넌트 할당할 `html` 을 반환한다.
 *
 * @returns {string} html 문자열
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getComponentHtml(){

    let type = this[OPTS].type;

    switch (type){

        case 'default':
            return _getDefaultHTML.call(this);
        case 'bubble':
            return _getBubbleHTML.call(this, type);
        case 'round':
            return _getBubbleHTML.call(this, type);
        case 'line':
            return _getBubbleHTML.call(this, type);
        default :
            return _getDefaultHTML.call(this);
    }
}

/**
 * `default` type 의 `html` 을 반환한다.
 *
 * @returns {string} html 문자열
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getDefaultHTML(){

    // default type 에 대한 template 을 반환한다.
    let html = [];

    let steps = _getDefaultTimesWithMinStep.call(this);

    let formatters = this[OPTS].formatters;
    let isHourShow = formatters.hour.isShow;
    let isMinShow = formatters.min.isShow;

    html.push(`<div class="${CLASS_NAME.bxTimePickerWrap} type-default">`);
    html.push('<ul class="bx-timepicker">');

    let length = steps.length;

    for (let i = 0; i < length; i++){

        let hour = isHourShow ? steps[i].hour + '' : '';
        let min = isMinShow ? steps[i].min : '';

        hour = isMinShow ? hour + ':' : hour;

        // 12 시간제일 경우에만 해당 값을 적용시킨다.
        let meridian = !this[OPTS].show24Hours ? ' ' + steps[i].meridian : '';

        html.push(`<li>${hour}${min}${meridian}</li>`);
    }

    html.push('</ul>');
    html.push('</div>');


    return html.join('');
}

/**
 * `bubble`/`round`/`line` 타입의 html 을 반환한다.
 *
 * @param {string} [formType = 'bubble'] - formType
 *
 * @returns {string} html 문자열
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getBubbleHTML(formType = 'bubble'){

    // bubble type 에 대한 template 을 반환한다.
    let html = [];
    let classNameLists = ['hour', 'min', 'sec'];

    // 각 시간 리스트들의 목록화를 진행한다.
    _getTimesWithSteps.call(this, 'hour');
    _getTimesWithSteps.call(this, 'min');
    _getTimesWithSteps.call(this, 'sec');
    _getMeridians.call(this);

    let colClassName = _getColumnClassName.call(this);
    let isShow24Hours = this.getShow24Hours();

    html.push(`<div class="${CLASS_NAME.bxTimePickerWrap} type-${formType}">`);
    html.push(`<ul class="bx-timepicker ${colClassName}">`);

    let length = classNameLists.length;

    for (let i = 0; i < length; i++){

        let className = classNameLists[i];

        if (this[OPTS].formatters[className].isShow){

            html.push(`<li class="${className}">`);
            html.push('<div class="bx-ctrl-btn btn-up"><i></i></div>');
            html.push('<div>');
            html.push(`<input type="text" class="${CLASS_NAME.pickerInput}" maxlength="2" />`);
            html.push('</div>');
            html.push('<div class="bx-ctrl-btn btn-down"><i></i></div>');
            html.push('</li>');
        }
    }

    // 12 시간제일 경우
    if (!isShow24Hours){

        html.push('<li class="meridian">');
        html.push('<div class="bx-ctrl-btn btn-up"><i></i></div>');
        html.push('<div>');
        html.push(`<input type="text" class="${CLASS_NAME.pickerInput}" disabled="disabled" />`);
        html.push('</div>');
        html.push('<div class="bx-ctrl-btn btn-down"><i></i></div>');
        html.push('</li>');
    }

    html.push('</ul>');
    html.push('</div>');

    return html.join('');
}

/**
 *
 * 컴포넌트 초기화 이벤트 바인딩 함수
 *
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _addInitEventListener(){

    // document 클릭 시, 해당 컴포넌트(마지막으로 열린)를 hide 시킨다.
    $(document).on('focus click', (e) =>{

        let $elem = $(e.target);

        let $input = $elem.parents(`.${CLASS_NAME.bxUiInput}`);

        if ($input.get(0)) return;

        // target 인스턴스가, 마지막으로 열린 인스턴스와 동일한 경우
        if (_.isEqual(this, lastShowInstance)){
            _hide.call(this);
        }
    });

    // $root 엘리먼트 클릭 시, 해당 컴포넌트를 노출 시킨다.
    this.$root.on('click', '.' + CLASS_NAME.bxTimePickerInput, (e) =>{

        if (_.isEqual(this, lastShowInstance)) return false;

        // 마지막 인스턴스가 존재할 경우
        if (lastShowInstance){

            // 마지막으로 열린 컴포넌트를 hide 시킨다.
            _hide.call(lastShowInstance);
        }

        // 해당 bx-input 컴포넌트에, timePicker 컴포넌트가 존재하지 않는 경우
        _show.call(this);

        // `defaultTime` 을 설정한다.
        _setDefaultTime.call(this);
    });

    return this;
}

/**
 *
 * 컴포넌트 이벤트 바인딩 함수
 *
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _addEventListener(){

    let $container = this.$container;
    let pickerType = this[OPTS].type;

    if (pickerType === 'default'){

        $container.find('ul').click((e) =>{

            let $elem = $(e.target);

            this.setTime($elem.text());

            _hide.call(this);
        });
    }
    else{

        // 각 input 엘리먼트(hour, min, sec)에 대한, 유효성 검사 처리(값이 숫자가 아닌 경우)
        $container.find(`.${CLASS_NAME.bxTimePickerWrap}`).on('keyup keydown', 'input', (e) =>{

            let $elem = $(e.target);

            let v = $elem.val();
            let ptn = /\D/g;

            if (v.match(ptn)) $elem.val(v.replace(ptn, ''));
        });

        $container.find(`.${CLASS_NAME.hour}`).on('click', `.${CLASS_NAME.bxCtrlBtn}`, (e) =>{

            let $elem = $(e.target);

            if (!_.isElement($elem.get(0))) return false;

            if ($elem.get(0).nodeName === 'I'){
                $elem = $elem.parent();
            }

            let className = $elem.attr('class');
            let value = '';

            if (className.indexOf(`${CLASS_NAME.btnUp}`) > -1){
                value = _getHourByUPAndDown.call(this, 'up');
            }
            else if (className.indexOf(`${CLASS_NAME.btnDown}`) > -1){
                value = _getHourByUPAndDown.call(this, 'down');
            }

            this.setHour(_.toNumber(value));
        });

        $container.find(`.${CLASS_NAME.min}`).on('click', `.${CLASS_NAME.bxCtrlBtn}`, (e) =>{

            let $elem = $(e.target);

            if (!_.isElement($elem.get(0))) return false;

            if ($elem.get(0).nodeName === 'I'){
                $elem = $elem.parent();
            }

            let className = $elem.attr('class');
            let value = '';

            if (className.indexOf(`${CLASS_NAME.btnUp}`) > -1){
                value = _getMinByUPAndDown.call(this, 'up');
            }
            else if (className.indexOf(`${CLASS_NAME.btnDown}`) > -1){
                value = _getMinByUPAndDown.call(this, 'down');
            }

            this.setMin(_.toNumber(value));
        });

        $container.find(`.${CLASS_NAME.sec}`).on('click', `.${CLASS_NAME.bxCtrlBtn}`, (e) =>{

            let $elem = $(e.target);

            if (!_.isElement($elem.get(0))) return false;

            if ($elem.get(0).nodeName === 'I'){
                $elem = $elem.parent();
            }

            let className = $elem.attr('class');
            let value = '';

            if (className.indexOf(`${CLASS_NAME.btnUp}`) > -1){
                value = _getSecByUPAndDown.call(this, 'up');
            }
            else if (className.indexOf(`${CLASS_NAME.btnDown}`) > -1){
                value = _getSecByUPAndDown.call(this, 'down');
            }

            this.setSec(_.toNumber(value));
        });

        // meridian UI 가 노출될 경우
        if (_isMultiMeridian.call(this)){

            $container.find(`.${CLASS_NAME.meridian}`).on('click', `.${CLASS_NAME.bxCtrlBtn}`, (e) =>{

                let $elem = $(e.target);

                if (!_.isElement($elem.get(0))) return false;

                if ($elem.get(0).nodeName === 'I'){
                    $elem = $elem.parent();
                }

                let className = $elem.attr('class');

                if (className.indexOf(`${CLASS_NAME.btnUp}`) > -1 ||
                    className.indexOf(`${CLASS_NAME.btnDown}`) > -1){
                    _toggleMeridian.call(this);
                }

                let values = _getInputValues.call(this);
                let time = _mergeTime(values.hour, values.min, values.sec, values.meridian);

                this.setTime(time);
            });
        }
    }

    return this;
}

/**
 *
 * defaultTime 을 설정한다.
 *
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setDefaultTime(){

    if (this.isSetDefaultTime) return this;

    let tmpTime = _getInputValues.call(this);

    let defaultTimes = _splitTime(this[OPTS].defaultTime);

    let defaultTime = _convertTimeWithFormatter.call(this, defaultTimes.time);

    let defaultMeridian = defaultTimes.meridian;

    let hour = tmpTime.hour || defaultTime.hour;
    let min = tmpTime.min || defaultTime.min;
    let sec = tmpTime.sec || defaultTime.sec;
    let meridian = tmpTime.meridian || defaultMeridian;

    let time = _mergeTime(hour, min, sec, meridian);

    // defaultTime 을 설정한다.
    this.setTime(time);

    // defaultTime 설정 여부에 true 를 할당한다.
    this.isSetDefaultTime = true;

    return this;
}

/**
 * 현재 시간을 반환한다.
 *
 * @returns {object} 시간 데이터
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getTime(){

    let pickerType = this[OPTS].type;
    let value = '';

    if (pickerType === 'default'){
        value = _getRootInputValue.call(this).split(/\s/)[0];
    }
    else{
        let values = _getInputValues.call(this);
        value = [values.hour, values.min, values.sec].join(':');
    }

    return _convertTimeWithFormatter.call(this, value);
}

/**
 * 정해진 formatter 를 통해, 시간 데이터를 변환한다.
 *
 * @param {string} [value = ''] - 시간 값
 *
 * @returns {object} 시간 데이터
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _convertTimeWithFormatter(value = ''){

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
    }

    let values = value ? value.split(':') : [];

    // 정해진 `formatter` 를 통한, 최대 크기보다 클 경우
    if (values.length > 3){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
    }

    // 시간
    let hour = _getSafeTime.call(this, 'hour', values[0]);
    // 분
    let min = _getSafeTime.call(this, 'min', values[1]);
    // 초
    let sec = _getSafeTime.call(this, 'sec', values[2]);

    return _getTimeObject.call(this, hour, min, sec);
}

/**
 * 시간 데이터를 반환한다.
 *
 * @param {string} [hour = ''] - 시간 값
 * @param {string} [min = ''] - 분 값
 * @param {string} [sec = ''] - 초 값
 *
 * @returns {object} 시간 데이터
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getTimeObject(hour = '', min = '', sec = ''){

    if (!_.isString(hour)){
        CONSOLE.prop.throw(BxTimePicker, 'hour', 'type');
    }

    if (!_.isString(min)){
        CONSOLE.prop.throw(BxTimePicker, 'min', 'type');
    }

    if (!_.isString(sec)){
        CONSOLE.prop.throw(BxTimePicker, 'sec', 'type');
    }

    let values = [];
    let ms = 0;

    hour && values.push(hour);
    min && values.push(min);
    sec && values.push(sec);

    ms += hour && _.toNumber(hour) * 3600;
    ms += min && _.toNumber(min) * 60;
    ms += sec && _.toNumber(sec);

    return {
        value: _.trim(values.join(':')),
        hour: hour,
        min: min,
        sec: sec,
        meridian: _getMeridian.call(this),
        ms: (ms * 1000)
    };
}

/**
 * 정해진 formatters 객체를 통해, 시간 값을 변환한다.
 *
 * @param {int} [value = 0] - 시간 값
 *
 * @returns {string} 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _convertHourWithFormatter(value = 0){

    if (!_.isNumber(value)){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'all');
    }

    let formatters = this[OPTS].formatters;

    if (!formatters.hour.isShow) return '';

    // 상황에 따라 문자열(0)이 추가된다.
    return (formatters.hour.enableZero && value < 10) ? `0${value}` : String(value);
}

/**
 * 정해진 formatters 객체를 통해, 분 값을 변환한다.
 *
 * @param {int} [value = 0] - 분 값
 *
 * @returns {string} 분 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _convertMinWithFormatter(value = 0){

    if (!_.isNumber(value)){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'all');
    }

    let formatters = this[OPTS].formatters;

    if (!formatters.min.isShow) return '';

    // 상황에 따라 문자열(0)이 추가된다.
    return (formatters.min.enableZero && value < 10) ? `0${value}` : String(value);
}

/**
 * 정해진 formatters 객체를 통해, 초 값을 변환한다.
 *
 * @param {int} [value = 0] - 초 값
 *
 * @returns {string} 초 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _convertSecWithFormatter(value = 0){

    if (!_.isNumber(value)){
        CONSOLE.prop.throw(BxTimePicker, 'v', 'all');
    }

    let formatters = this[OPTS].formatters;

    if (!formatters.sec.isShow) return '';

    // 상황에 따라 문자열(0)이 추가된다.
    return (formatters.sec.enableZero && value < 10) ? `0${value}` : String(value);
}

/**
 * 시간 간격을 반환한다.
 *
 * @returns {int} 시간 간격
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getHourStep(){

    let steps = this[OPTS].steps;

    // 해당 step 이 없을 경우 `1` 로 초기화 한다.
    steps.hour = _.isNumber(steps.hour) ? steps.hour : 1;

    // 해당 step 이 0 과 같거나 작을경우
    if (steps.hour <= 0){
        steps.hour = 1;
    }

    return steps.hour;
}

/**
 * 분 간격을 반환한다.
 *
 * @returns {int} 분 간격
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getMinStep(){

    let steps = this[OPTS].steps;

    // 해당 step 이 없을 경우 `1` 로 초기화 한다.
    steps.min = _.isNumber(steps.min) ? steps.min : 1;

    // 해당 step 이 0 과 같거나 작을경우
    if (steps.min <= 0){
        steps.min = 1;
    }

    return steps.min;
}

/**
 * 초 간격을 반환한다.
 *
 * @returns {int} 초 간격
 *
 * @private
 * @memberOf BxTimePicker
 */

function _getSecStep(){

    let steps = this[OPTS].steps;

    // 해당 step 이 없을 경우 `1` 로 초기화 한다.
    steps.sec = _.isNumber(steps.sec) ? steps.sec : 1;

    // 해당 step 이 0 과 같거나 작을경우
    if (steps.sec <= 0){
        steps.sec = 1;
    }

    return steps.sec;
}

/**
 * 분 step 을 통해, `default` 타입의 아이템 리스트를 설정 및 반환한다.
 *
 * @returns {array} 아이템 리스트
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getDefaultTimesWithMinStep(){

    let ret = [];
    let isShow24Hours = this.getShow24Hours();

    let minLimitHour = 0;
    let maxLimitHour = 24 * 60;

    let middleTime = maxLimitHour / 2;
    let minStep = _getMinStep.call(this);

    let expression = function (){
        return minLimitHour < maxLimitHour;
    };

    while (expression()){

        let hour = minLimitHour > 0 ? parseInt(minLimitHour / 60, 10) : 0;
        let min = minLimitHour > 0 ? minLimitHour - (hour * 60) : 0;

        if (!isShow24Hours && hour === 0) hour = 12;
        else if (!isShow24Hours && hour > 12) hour = hour - 12;

        hour = _convertHourWithFormatter.call(this, hour);
        min = _convertMinWithFormatter.call(this, min);

        ret.push({
            hour: hour,
            min: min,
            meridian: minLimitHour < middleTime ? MESSAGE.am : MESSAGE.pm
        });

        minLimitHour += minStep;
    }

    this.stores.value.items = ret;

    return ret;
}

/**
 * 각 시간(hour, min, sec)에 대한, 아이템 리스트를 설정 및 반환한다.
 *
 * @param {string} [type = ''] - 시간 타입
 *
 * @returns {array} 아이템 리스트
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getTimesWithSteps(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    let ret = {};

    let step, converter;

    let i = 0;
    let length = 0;

    let limit = type === 'hour' ? this.hourLimit['__24MinMax__'] : _getMinMaxTime.call(this, type);

    if (type === 'hour'){

        // type 이 시간일 경우
        step = _getHourStep.call(this);

        converter = _convertHourWithFormatter;

        // 최소 시간
        i = limit.min;

        // 최대 시간
        length = limit.max;
    }
    else if (type === 'min'){

        step = _getMinStep.call(this);

        converter = _convertMinWithFormatter;

        // 최소 시간
        i = limit.min;

        // 최대 시간
        length = limit.max;
    }
    else if (type === 'sec'){

        step = _getSecStep.call(this);

        converter = _convertSecWithFormatter;

        // 최소 시간
        i = limit.min;

        // 최대 시간
        length = limit.max;
    }

    if (type === 'hour'){

        let _24Items = [];
        let _12AMItems = [];
        let _12PMItems = [];

        while (i <= length){

            // 24 시간용 데이터
            _24Items.push(converter.call(this, i));

            // 12 AM 시간용 데이터
            if (i < 12){
                _12AMItems.push(converter.call(this, i));
            }

            // 12 PM 시간용 데이터
            if (i >= 12){
                _12PMItems.push(converter.call(this, function (){

                    let ii = i;

                    if (i === 12) ii = 0;

                    if (i > 12){
                        return i - 12;
                    }

                    return ii;

                }()));
            }


            i += step;
        }

        this.stores[type]['24Items'] = _24Items;
        this.stores[type]['12AMItems'] = _12AMItems;
        this.stores[type]['12PMItems'] = _12PMItems;
    }
    else{

        let items = [];

        while (i <= length){

            items.push(converter.call(this, i));

            i += step;
        }

        this.stores[type].items = items;
    }

    return ret;
}

/**
 * 각 시간(hour, min, sec)에 대한, min/max 객체를 반환한다.
 *
 * @param {string} [type = ''] - 시간 타입
 *
 * @returns {object} min/max 객체
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getMinMaxTime(type = ''){

    let min = 0;
    let max = 0;

    if (type === 'hour'){

        let isShow24Hours = this.getShow24Hours();
        let hourLimit = this.hourLimit;

        let limit = null;

        // 24 시간제일경우
        if (isShow24Hours){
            limit = hourLimit['__24MinMax__'];
        }
        else{

            // 12 시간제이면서, AM 일 경우
            if (this.getAM()){
                limit = hourLimit['__12AmMinMax__'];
            }
            // 12 시간제이면서, PM 일 경우
            else if (this.getPM()){
                limit = hourLimit['__12PmMinMax__'];
            }
        }

        if (_.isEmpty(limit)){
            CONSOLE.prop.throw(BxTimePicker, '해당 `meridian` 타입의 아이템이 존재하지 않습니다.');
        }

        min = limit.min;
        max = limit.max;
    }
    else if (type === 'min' || type === 'sec'){
        min = 0;
        max = 59;
    }

    // 분 또는 초에 대한 최소, 최대 시간
    return {
        min: min,
        max: max
    };
}

/**
 * 초기 `meridian` 값을 반환한다.
 *
 * @returns {string} 초기 meridian 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _initMeridian(){

    let isShow24Hours = this.getShow24Hours();

    // 24 시간제일 경우
    if (isShow24Hours) return '';


    if (this.useMeridian['AM'] && this.useMeridian['PM']){
        this.meridian = MESSAGE.am;
    }
    else if (this.useMeridian['AM'] && !this.useMeridian['PM']){
        this.meridian = MESSAGE.am;
    }
    else if (!this.useMeridian['AM'] && this.useMeridian['PM']){
        this.meridian = MESSAGE.pm;
    }

    return this.meridian;
}

/**
 * meridian 값을 반환한다.
 *
 * @returns {string} meridian 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getMeridian(){

    return this.meridian;
}

/**
 * meridian 값을 설정한다.
 *
 * @returns {string} meridian 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setMeridian(meridian = ''){

    if (!_.isString(meridian)){
        CONSOLE.prop.throw(BxTimePicker, 'meridian', 'type');
    }

    let isShow24Hours = this.getShow24Hours();

    // 24 시간제인 경우
    if (isShow24Hours) return '';

    if (_.trim(meridian).length > 0){

        // 값이 `AM` 또는 `PM` 이 아닌 경우
        if (meridian !== MESSAGE.am && meridian !== MESSAGE.pm){
            CONSOLE.prop.throw(BxTimePicker, MESSAGE.meridianValueError);
        }

        this.meridian = meridian;
    }

    return meridian;
}

/**
 * meridian 아이템 리스트를 설정 및 반환한다.
 *
 * @returns {array} 아이템 리스트
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getMeridians(){

    let ret = [];

    // 24 시간제가 아닐 경우
    if (this.getShow24Hours()) return ret;

    ret.push(MESSAGE.am);
    ret.push(MESSAGE.pm);

    this.stores.meridian.items = ret;

    return ret;
}

/**
 * 시간 아이템 리스트에 대한, `상위` or `하위` 값을 가져온다.
 *
 * @param {string} [type = ''] - 상위 or 하위(up or down) 타입
 *
 * @returns {string} 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getHourByUPAndDown(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    let ret = '';

    if (type === 'up'){
        // 현재 시간보다 한단계 큰 값을 가져온다.
        ret = _getTimeWithUP.call(this, 'hour');
    }
    else{
        ret = _getTimeWithDown.call(this, 'hour');
    }

    return ret;
}

/**
 * 분 아이템 리스트에 대한, `상위` or `하위` 값을 가져온다.
 *
 * @param {string} [type = ''] - 상위 or 하위(up or down) 타입
 *
 * @returns {string} 분 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getMinByUPAndDown(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    let ret = '';

    if (type === 'up'){
        // 현재 시간보다 한단계 큰 값을 가져온다.
        ret = _getTimeWithUP.call(this, 'min');
    }
    else{
        ret = _getTimeWithDown.call(this, 'min');
    }

    return ret;
}

/**
 *
 * 초 아이템 리스트에 대한, `상위` or `하위` 값을 가져온다.
 *
 * @param {string} [type = ''] - 상위 or 하위(up or down) 타입
 *
 * @returns {string} 초 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getSecByUPAndDown(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    let ret = '';

    if (type === 'up'){
        // 현재 시간보다 한단계 큰 값을 가져온다.
        ret = _getTimeWithUP.call(this, 'sec');
    }
    else{
        ret = _getTimeWithDown.call(this, 'sec');
    }

    return ret;
}

/**
 * 각 시간(hour, min, sec)에 대한, `상위` 값을 가져온다.
 *
 * @param {string} [type = ''] - 시간 타입
 *
 * @returns {string} 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getTimeWithUP(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'all');
    }

    let ret = '';
    let compareNum = '';

    let store = this.stores[type];
    let items = [];

    let isShow24Hours = this.getShow24Hours();

    switch (type){

        case 'hour':
            compareNum = this.getHour();
            break;
        case 'min':
            compareNum = this.getMin();
            break;
        case 'sec':
            compareNum = this.getSec();
            break;
    }

    if (compareNum){

        let _compareNum = _.toNumber(compareNum);

        // 시간 타입일 경우..
        if (type === 'hour'){

            // 12 기간제일 경우
            if (!isShow24Hours){
                if (_compareNum >= 12) _compareNum = 0;
            }

            items = _getHourItemsWithMeridian.call(this, store);
        }
        else{
            items = store.items;
        }

        ret = _.minBy(items, (n) =>{

            n = _.toNumber(n);

            if (_compareNum < n){
                return n;
            }

        }) || compareNum;
    }

    // 현재 time 값
    return ret;
}

/**
 * 각 시간(hour, min, sec)에 대한, `하위` 값을 가져온다.
 *
 *
 * @param {string} [type = ''] - 시간 타입
 *
 * @returns {string} 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getTimeWithDown(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    let ret = '';
    let compareNum = '';

    let store = this.stores[type];
    let items = [];

    let isShow24Hours = this.getShow24Hours();

    switch (type){

        case 'hour':
            compareNum = this.getHour();
            break;
        case 'min':
            compareNum = this.getMin();
            break;
        case 'sec':
            compareNum = this.getSec();
            break;
    }

    if (compareNum){

        let _compareNum = _.toNumber(compareNum);

        // 시간 타입이면서, 12 시간제일 경우..
        if (type === 'hour'){

            // 12 기간제일 경우
            if (!isShow24Hours){
                if (_compareNum >= 12) _compareNum = 0;
            }

            items = _getHourItemsWithMeridian.call(this, store);
        }
        else{
            items = store.items;
        }

        ret = _.maxBy(items, (n) =>{

            n = _.toNumber(n);

            if (_compareNum > n){
                return n;
            }

        }) || compareNum;
    }

    // 현재 time 값
    return ret;
}

/**
 *
 * `bubble`/`round`/`line` 타입에 대한, `columnClassName` 을 반환한다.
 *
 **
 * @returns {string} columnClassName
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getColumnClassName(){

    let isShow24Hours = this.getShow24Hours();

    let length = _.filter(this[OPTS].formatters, (o, i) =>{
        return o.isShow;
    }).length;

    // 12 시간제일 경우
    if (!isShow24Hours) ++length;

    let ret = '';

    if (length === 1){
        ret = CLASS_NAME.col1;
    }
    else if (length === 2){
        ret = CLASS_NAME.col2;
    }
    else if (length === 3){
        ret = CLASS_NAME.col3;
    }
    else if (length === 4){
        ret = CLASS_NAME.col4;
    }

    return ret;
}

/**
 *
 * $rootInput 엘리먼트의 value 값을 반환한다.
 *
 **
 * @returns {string} $rootInput 엘리먼트의 value 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getRootInputValue(){

    let $root = this.$root;
    let $input = $root.find(`.${CLASS_NAME.bxTimePickerInput}`);

    // 현재 input value 를 반환한다.
    return $input.val();
}

/**
 * $rootInput 엘리먼트의 value 값을 설정한다.
 *
 * @param {string} [hour = ''] - 시간 값
 * @param {string} [min = ''] - 분 값
 * @param {string} [sec = ''] - 초 값
 * @param {string} [meridian = ''] - meridian 값
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setRootInputValue(hour = '', min = '', sec = '', meridian = ''){

    if (!_.isString(hour)){
        CONSOLE.prop.throw(BxTimePicker, 'hour', 'type');
    }

    if (!_.isString(min)){
        CONSOLE.prop.throw(BxTimePicker, 'min', 'type');
    }

    if (!_.isString(sec)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    if (!_.isString(meridian)){
        CONSOLE.prop.throw(BxTimePicker, 'meridian', 'type');
    }

    let $root = this.$root;
    let $input = $root.find(`.${CLASS_NAME.bxTimePickerInput}`);

    hour = _zeroToTwelveHour.call(this, hour);

    let time = _mergeTime(hour, min, sec, meridian);

    $input.val(time);

    return this;
}

/**
 * `bubble`/`round`/`line` 타입의, 각 input 엘리먼트의 value 값을 반환한다.
 *
 * @returns {object} 시간 데이터
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getInputValues(){

    let $container = this.$container;

    let $pickerWrap = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);

    let $hourInput = $pickerWrap.find(`.${CLASS_NAME.hour} .${CLASS_NAME.pickerInput}`);
    let $minInput = $pickerWrap.find(`.${CLASS_NAME.min} .${CLASS_NAME.pickerInput}`);
    let $secInput = $pickerWrap.find(`.${CLASS_NAME.sec} .${CLASS_NAME.pickerInput}`);
    let $meridianInput = $pickerWrap.find(`.${CLASS_NAME.meridian} .${CLASS_NAME.pickerInput}`);

    let hour = $hourInput.val();
    hour = _twelveToZeroHour.call(this, hour);

    return {
        hour: hour,
        min: $minInput.val(),
        sec: $secInput.val(),
        meridian: $meridianInput.val()
    };
}

/**
 *
 * `bubble`/`round`/`line` 타입의, 각 input 엘리먼트의 value 값을 설정한다.
 *
 * @param {string} [hour = ''] - 시간 값
 * @param {string} [min = ''] - 분 값
 * @param {string} [sec = ''] - 초 값
 * @param {string} [meridian = ''] - meridian 값
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setInputValues(hour = '', min = '', sec = '', meridian = ''){

    _setHourInputValue.call(this, hour);
    _setMinInputValue.call(this, min);
    _setSecInputValue.call(this, sec);
    _setMeridianInputValue.call(this, meridian);

    return this;
}

/**
 * 시간 input 엘리먼트의 value 값을 설정한다.
 *
 * @param {string} [hour = ''] - 시간 값
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setHourInputValue(hour = ''){

    if (!_.isString(hour)){
        CONSOLE.prop.throw(BxTimePicker, 'hour', 'type');
    }

    let $container = this.$container;
    let $pickerWrap = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);

    let $input = $pickerWrap.find(`.${CLASS_NAME.hour} .${CLASS_NAME.pickerInput}`);

    let _hour = _zeroToTwelveHour.call(this, hour);

    $input.val(_hour);

    return this;
}

/**
 * 분 input 엘리먼트의 value 값을 설정한다.
 *
 * @param {string} [min = ''] - 분 값
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setMinInputValue(min = ''){

    if (!_.isString(min)){
        CONSOLE.prop.throw(BxTimePicker, 'min', 'type');
    }

    let $container = this.$container;
    let $pickerWrap = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);

    let $input = $pickerWrap.find(`.${CLASS_NAME.min} .${CLASS_NAME.pickerInput}`);

    $input.val(min);

    return this;
}

/**
 * 초 input 엘리먼트의 value 값을 설정한다.
 *
 * @param {string} [sec = ''] - 초 값
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxTimePicker
 */
function _setSecInputValue(sec = ''){

    if (!_.isString(sec)){
        CONSOLE.prop.throw(BxTimePicker, 'sec', 'type');
    }

    let $container = this.$container;
    let $pickerWrap = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);

    let $input = $pickerWrap.find(`.${CLASS_NAME.sec} .${CLASS_NAME.pickerInput}`);

    $input.val(sec);

    return this;
}

/**
 * meridian input 엘리먼트의 value 값을 설정한다.
 *
 * @param {string} [meridian = ''] - meridian 값
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _setMeridianInputValue(meridian = ''){

    if (!_.isString(meridian)){
        CONSOLE.prop.throw(BxTimePicker, 'meridian', 'type');
    }

    let $container = this.$container;
    let $pickerWrap = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);

    let $input = $pickerWrap.find(`.${CLASS_NAME.meridian} .${CLASS_NAME.pickerInput}`);

    $input.val(meridian);

    return this;
}


/**
 * `default` type 이 아닌, 체크된 시간(hour, min, sec) 값을 반환한다.
 *
 * @param {string} [type = ''] - 시간 타입
 * @param {string} [value = ''] - 시간 값
 *
 * @returns {string} 시간 값
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _getSafeTime(type = '', value = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxTimePicker, 'type', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
    }

    let isShow24Hours = this.getShow24Hours();

    let pickerType = this[OPTS].type;
    // 각 타입에 따른, 제한 시간을 가져온다.
    let limit = _getMinMaxTime.call(this, type);

    let _value = _.toNumber(value);

    let min = limit.min;
    let max = limit.max;

    // min / max 값을 우회하기위한 로직!!(중요)
    // hour 타입일 경우
    // 12 시간제일 경우
    // value 값이 12 인 경우
    if (
        type === 'hour' && !isShow24Hours &&
        pickerType !== 'default'
    ){

        if (_value === 12) _value = 0;

        min = _to12Hour(min);
        max = _to12Hour(max);
    }

    // 24 시간제일 경우
    if (_value < min) _value = min;
    if (_value > max) _value = max;

    if (type === 'hour'){
        value = _convertHourWithFormatter.call(this, _value);
    }
    else if (type === 'min'){
        value = _convertMinWithFormatter.call(this, _value);
    }
    else if (type === 'sec'){
        value = _convertSecWithFormatter.call(this, _value);
    }

    return value;
}

/**
 * meridian 값을 toggle 한다.
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 * @memberOf BxTimePicker
 */
function _toggleMeridian(){

    // AM / PM 설정을 토글한다.
    this.getAM() ? this.setPM() : this.setAM();

    return this;
}

/**
 * component toggle
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxTimePicker
 */
function _toggle(){

    if (_isShow.call(this)) _hide.call(this);
    else _show.call(this);

    return this;
}

/**
 * component show 유/무를 반환한다.
 *
 * @returns {boolean} show 유/무
 *
 * @private
 * @memberOf BxTimePicker
 */
function _isShow(){

    let $container = this.$container;

    let $elem = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);
    let display = $elem.css('display');

    return display === 'block' ? true : false;
}

/**
 * component 를 show 한다.
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxTimePicker
 */
function _show(){

    let $root = this.$root;
    let $container = this.$container;

    let $elem = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);
    let $inputElem = $root.find(`.${CLASS_NAME.bxTimePickerInput}`);

    // `defaultTime` 을 설정한다.
    _setDefaultTime.call(this);

    window.setTimeout(() =>{

        $inputElem.focus();

        $elem.css('display', 'block');

        this[OPTS].onShow.call(this);

        lastShowInstance = this;

    }, 30);

    return this;
}

/**
 * component 를 hide 한다.
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxTimePicker
 */
function _hide(){

    let $root = this.$root;
    let $container = this.$container;

    let $elem = $container.find(`.${CLASS_NAME.bxTimePickerWrap}`);
    let $inputElem = $root.find(`.${CLASS_NAME.bxTimePickerInput}`);

    let pickerType = this[OPTS].type;

    if (pickerType !== 'default'){

        // 현재 input 엘리먼트들에 있는 값을 설정한다.
        let values = _getInputValues.call(this);
        let time = _mergeTime(values.hour, values.min, values.sec, values.meridian);

        this.setTime(time);
    }

    window.setTimeout(() =>{

        $inputElem.blur();

        $elem.css('display', 'none');

        this[OPTS].onHide.call(this);

        lastShowInstance = null;

    }, 1);

    return this;
}


/**
 * bx-input 유/무 를 반환한다.
 *
 * @returns {boolean} bx-input 유/무
 *
 * @private
 * @memberOf BxTimePicker
 */
function _isBXInput(className = ''){

    if (!_.isString(className)){
        CONSOLE.prop.throw(BxTimePicker, 'className', 'type');
    }

    return (className && className.indexOf(CLASS_NAME.bxUiInput) > -1) ? true : false;
}

/**
 * meridian 값을 통해, 12 시간 기준값 으로 변환한다.
 *
 * @returns {int} value - 시간값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _to12Hour(value = 0){

    if (!_.isNumber(value)){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
    }

    let ret = value;

    // `PM` 상태이면서, value 값이 12 이상인 경우
    if (value > 11){
        ret = value - 12;
    }

    return ret;
}

/**
 * 전달된 item 값을 통해, 시간을 반환한다.
 *
 * @returns {object} item - 시간 객체
 *
 * @private
 * @memberOf BxTimePicker
 */
function _getHourItemsWithMeridian(store = {}){

    if (!_.isPlainObject(store)){
        CONSOLE.prop.throw(BxTimePicker, 'store', 'type');
    }

    let isShow24Hours = this.getShow24Hours();

    let items = [];

    // 24 시간제인 경우
    if (isShow24Hours){
        items = store['24Items'];
    }
    else{

        let meridian = _getMeridian.call(this);

        // `AM` 또는 `PM` 에 제한될 경우..
        if (!_isMultiMeridian.call(this)){
            items = store['12' + meridian + 'Items'];
        }
        else{

            if (this.getAM()){
                items = store['12AMItems'];
            }
            else if (this.getPM()){
                items = store['12PMItems'];
            }
        }
    }

    return items;
}

/**
 * multi meridian 노출 여부
 *
 * @returns {boolean} - multi meridian 노출 여부
 *
 * @private
 * @memberOf BxTimePicker
 */
function _isMultiMeridian(){

    let ret = false;

    let useMeridian = this.useMeridian;

    // 24 시간제일 경우
    if (this.getShow24Hours()) return ret;

    // `AM` / `PM` 모두 사용할경우
    if (useMeridian['AM'] && useMeridian['PM']) ret = true;

    return ret;
}

/**
 * 나눠진 시간들을 merge 시킨다.
 *
 * @param {string} [hour = ''] - 시간 값
 * @param {string} [min = ''] - 분 값
 * @param {string} [sec = ''] - 초 값
 * @param {string} [meridian = ''] - meridian 값
 *
 * @returns {string} - 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _mergeTime(hour = '', min = '', sec = '', meridian = ''){

    if (!_.isString(hour)){
        CONSOLE.prop.throw(BxTimePicker, 'hour', 'type');
    }

    if (!_.isString(min)){
        CONSOLE.prop.throw(BxTimePicker, 'min', 'type');
    }

    if (!_.isString(sec)){
        CONSOLE.prop.throw(BxTimePicker, 'sec', 'type');
    }

    if (!_.isString(meridian)){
        CONSOLE.prop.throw(BxTimePicker, 'meridian', 'type');
    }

    let times = [];

    hour && times.push(hour);
    min && times.push(min);
    sec && times.push(sec);

    let ret = times.join(':');

    if (_.trim(meridian).length > 0){
        ret += ' ' + meridian;
    }

    return ret;
}

/**
 * 시간을 정해진 패턴으로 분리 시킨다.
 *
 * @param {string} [value = ''] - 시간 값
 *
 * @returns {object} - 분리된 시간 데이터
 *
 * @private
 * @memberOf BxTimePicker
 */
function _splitTime(value = ''){

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxTimePicker, 'value', 'type');
    }

    let values = value.split(' ');

    return {
        time: values[0] || '',
        meridian: values[1] || ''
    };
}

/**
 * `0` 값을 `12` 로 변경한다.
 *
 * @param {string} [hour = ''] - 시간 값
 *
 * @returns {string} - 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _zeroToTwelveHour(hour = ''){

    if (!_.isString(hour)){
        CONSOLE.prop.throw(BxTimePicker, 'hour', 'type');
    }

    let isShow24Hours = this.getShow24Hours();

    if (!isShow24Hours && _.toNumber(hour) === 0) hour = '12';

    return hour;
}

/**
 * `12` 값을 `0` 으로 변경한다.
 *
 * @param {string} [hour = ''] - 시간 값
 *
 * @returns {string} - 시간 값
 *
 * @private
 * @memberOf BxTimePicker
 */
function _twelveToZeroHour(hour = ''){

    if (!_.isString(hour)){
        CONSOLE.prop.throw(BxTimePicker, 'hour', 'type');
    }

    let isShow24Hours = this.getShow24Hours();

    if (!isShow24Hours && _.toNumber(hour) === 12) hour = '0';

    return hour;
}

module.exports = BxTimePicker;
