// bxuip-angular.js-ui/src/lib/bx-ui/app/components/toggle.js
/**
 * 랜더링 완료 이벤트의 callback 함수
 *
 * @callback onRender
 *
 * @memberOf BxToggle
 */

/**
 *  메뉴 선택 이벤트의 callback 함수
 *
 *  @callback onSelected
 *  @param {object} e - 이벤트 객체
 *  @param {Object} $elem - 선택된 엘리먼트
 *  @memberOf BxToggle
 */

/**
 * @class BxToggle
 * @classDesc BxToggle 클래스 생성자
 *
 * @param {object} OPTS
 * @param {object[]} [OPTS.items = []]
 * @param {object} OPTS.items.attr - 엘리먼트 속성
 * @param {string} OPTS.items.title - 타이틀
 * @param {string} OPTS.items.icon - 아이콘
 * @param {string[]} OPTS.items.items - 자식 아이템 집합
 * @param {boolean} OPTS.items.active - 메뉴 활성화 유무
 * @param {function} [OPTS.onSelected = function(){}] - 메뉴 선택 시, 호출될 callback 함수
 * @param {function} [OPTS.onRendered = function(){}] - 랜더링 완료 시, 호출될 callback 함수
 *
 * @returns {Object} - 생성된 인스턴스
 * @example
 $('#toggle').BxToggle({
    items : [
        {
            attr: {id: '1', name: '1'},
            title: '목록1',
            items: [
                    '<a href="#" style="padding-left:55px">아이템1</a>',
                    '<a href="#" style="padding-left:55px">아이템2</a>',
                    '<a href="#" style="padding-left:55px">아이템3</a>',
            ],
            icon: '',
            active: true
        },
        {
            attr: {id: '2', name: '2'},
            title: '목록2',
            items: [
                '아이템1',
                '아이템2',
                '아이템3'
            ],
            icon: 'bw-icon i-30 i-mng-online',
            active: false
        },
    ],
    onSelected : function(e, $elem){
        console.log('select');
    },
    onRender : function(){
        console.log('render Complate')
    }
 }).render();
 //---------------------------------------------------------------

 // bind 사용 예제 angularJS
 <div id="toggle">
     <ul>
         <li id="1" name="1">
            <h3 class="toggle-header active">목록1</h3>
            <div class="toggle-content" style="display: none;">
                <ul class="bw-sub">
                    <li ng-click="vm.go('deal')">아이템1</li>
                    <li ng-click="vm.go('dashBoard')">아이템2</li>
                </ul>
            </div>
         </li>
         <li id="2" name="2">
            <h3 class="toggle-header active">
                <i class="bw-icon i-30 i-mng-online"></i>목록2
            </h3>
            <div class="toggle-content" style="display: none;">
                <ul class="bw-sub">
                    <li>아이템1</li>
                    <li>아이템2</li>
                </ul>
            </div>
         </li>
     </ul>
 </div>

 $('#toggle').BxToggle({
    onSelected : function(e, $elem){
        console.log('select');
    },
    onRender : function(){
        console.log('render Complate')
    }
 }).bind();
 //---------------------------------------------------------------

 // bind 사용 예제 react
 <div id="toggle">
     <ul>
         <li id="1" name="1">
            <h3 className="toggle-header active">목록1</h3>
            <div className="toggle-content">
                <ul className="bw-sub">
                    <li><Link to="uriPath">아이템1</Link></li>
                    <li><Link to="uriPath">아이템2</Link></li>
                </ul>
            </div>
         </li>
         <li id="1" name="1">
            <h3 className="toggle-header active">
                <i className="bw-icon i-30 i-mng-online"></i>목록2
            </h3>
            <div className="toggle-content">
                <ul className="bw-sub">
                    <li><Link to="uriPath">아이템1</Link></li>
                    <li><Link to="uriPath">아이템2</Link></li>
                </ul>
            </div>
         </li>
     </ul>
 </div>

 $('#toggle').BxToggle({
    onSelected : function(e, $elem){
        console.log('select');
    },
    onRender : function(){
        console.log('render Complate')
    }
 }).bind();

 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const OPTS = Symbol();

class BxToggle {

    constructor({
        items = [],
        onSelected = function (){
        },
        onRender = function (){
        }
    } = {}){

        this.$root = BxToggle.$root;

        if (!_.isArray(items)){
            CONSOLE.prop.throw(BxToggle, 'items', 'type');
        }

        if (!_.isFunction(onSelected)){
            CONSOLE.prop.throw(BxToggle, 'onSelected', 'type');
        }

        if (!_.isFunction(onRender)){
            CONSOLE.prop.throw(BxToggle, 'onRender', 'type');
        }

        this[OPTS] = {items, onRender, onSelected};
    }

    /**
     * 랜더 함수
     *
     * @returns {Object} - 생성된 인스턴스
     */
    render(){

        this.$container = BASE.appendContainerElement(this.$root, BxToggle.className);

        const $container = this.$container;

        let html = [];

        html.push('<ul></ul>');

        $container.append(html.join(''));

        let $ul = $container.find('ul');

        let length = this[OPTS].items.length;

        for (let i = 0; i < length; i++){

            let item = this[OPTS].items[i];

            let $li = _createElement.call(this, item);

            $ul.append($li);
        }

        this[OPTS].onRender.apply(this);

        return this;
    }

    /**
     * 템플릿 바인딩 함수
     *
     * @returns {object} 생성된 인스턴스
     */
    bind(){

        // var self = this;

        // angular의 경우, linking 과정 전에 호출하는 것을 방지
        window.setTimeout(() => _bind.call(this), 1);

        return this;
    }

    /**
     * 컨텐츠 열기 함수
     *
     * @param {object} $elem - 선택된 엘리먼트
     * @private
     */
    show($elem = null){

        if (!_.isObject($elem)){
            CONSOLE.prop.error(BxToggle, '$elem', 'type');
        }

        if (_.isNull($elem)){
            CONSOLE.prop.error(BxToggle, '$elem', 'value');
        }

        // target 엘리먼트의 title 영역
        let $h3 = $elem.find('h3');

        // target 엘리먼트의 content 엘리먼트 영역
        let $content = $elem.find('div:first');

        // target 엘리먼트를 오픈한다.
        $content.show();

        // target 엘리먼트의 class 속성에 active 값을 할당한다.
        $h3.addClass('active');
    }
}

/*
 내부 바인드 함수
 */
function _bind(){

    let $children = this.$root.children();

    this.$container = BASE.appendContainerElement(this.$root, BxToggle.className);
    /*
     * $children DOM Tree 이동
     * Jq.append 함수 이용시 존재하는 DOM을 인자로 사용하는 경우, 해당 돔의 위치가 이동하게 된다.
     * $children은 Jq Dom 인스턴스이며, 내부에 context를 통해 해당 돔을 삭제 및 지정 한돔에 추가 한다.
     * http://api.jquery.com/append/ 참고.
     */
    this.$container.append($children);

    /*
     this.$container.find('ul>li') 로 find 검색시,
     세부 목록의 li까지 전부 가져오게 되므로, className으로 접근해서 검색하도록 한다.
     토글 메뉴에 id, name이라는 속성이 없을 수도 있기 때문이다.
     아래와 같이 nth-child로 접근하는 이유는 nth-child로 접근하지 아니할시에 하위 컴포넌트들을 모두 가져오는 문제를 해결하기 위해서이며,
     이벤트 바인딩을 할시 해당 DOM이아닌 jquery의 인스턴스로 이용하기 위해서 이다.
     */

    let liFinder = '.' + this.$container.attr('class') + ' >ul >li';
    let liLength = this.$root.find(liFinder).length;

    for (let i = 0; i < liLength; i++){

        let $li = this.$root.find(liFinder + ':nth-child(' + ( i + 1 ) + ')');
        let $h3 = $li.find('h3');
        let $div = $li.find('div.toggle-content');

        $h3.click((e) =>{
            _toggle.call(this, $h3, $div);
            this[OPTS].onSelected(e, $li);
        });

        /*
         toggle header가 active true 값을 갖는 경우
         이에 대해 활성화를 시킬 수 있음..
         */
        let active = $h3.hasClass('active');
        active && this.show($li);
    }

    /*
     render 종료시, onRender callback 함수를 call.
     */
    this[OPTS].onRender.apply(this);
}

/**
 * 토글 함수
 * @param {object} $header - 선택된 엘리먼트의 해더부분의 돔 객체
 * @param {object} $content - 선택된 엘리먼트의 내용부분의 돔 객
 * @private
 */
function _toggle($header = null, $content = null){

    if ($header.hasClass('active')){
        // target 엘리먼트가 열려있는 경우, 해당 엘리먼트를 닫는다.
        $content.slideUp('fast', 'swing');

        // target 엘리먼트의 class 를 지운다.
        $header.removeClass('active');
    }
    else{
        // target 엘리먼트를 오픈한다.
        $content.slideDown();

        // target 엘리먼트의 class 속성에 active 값을 할당한다.
        $header.addClass('active');
    }
}

/**
 * 엘리먼트 생성 함수
 *
 * @param ITEM
 * @param {object} [ITEM.attr = {}] - 토글 메뉴에 추가할 attribute
 * @param {string} [ITEM.title = ''] - 토글 메뉴 title
 * @param {object[]} [ITEM.items = []] - 토글 메뉴 아이템
 * @param {string} [ITEM.icon = ''] - 토글 아이콘
 * @param {boolean} [ITEM.active = false] - 메뉴 선택 시, 발생될 callback 함수
 * @returns {object} $li - $('li') 을 리턴.
 * @private
 * @memberOf BxToggle
 */

function _createElement({attr = {}, title = '', items = [], icon = '', active = false} = {}){

    let $li = $('<li></li>');

    // 엘리먼트에 전달받은 attr 객체 속성을 추가한다.
    UTIL.attr($li, attr);

    let $h3 = $('<h3 class="toggle-header"><a href="#" onclick="return false;"><i class="' + icon + '"></i>' + title + '</a></h3>');
    let $div = $('<div class="toggle-content"><ul class="bw-sub"></ul></div>');

    $h3.click((e) =>{

        _toggle.call(this, $h3, $div);

        this[OPTS].onSelected(e, $li);

        return false;
    });

    let $ul = $div.find('ul');
    let length = items.length;

    for (let i = 0; i < length; i++){

        let content = items[i];
        let $contentElem = $(content);
        let contentElem = $contentElem.get(0);

        if (_.isElement(contentElem) && contentElem.tagName === 'A'){
            $ul.append($('<li>' + content + '</li>'));
        }
        else{
            $ul.append($('<li><a href="#" onclick="return false;">' + content + '</a></li>'));
        }

    }

    $li.append($h3);
    $li.append($div);

    active && this.show($li);

    return $li;
}

module.exports = BxToggle;
