// bxuip-angular.js-ui/src/lib/bx-ui/app/components/tooltip.js
/**
 * @class BxTooltip
 * @classDesc BxTooltip 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 * <script>
 *
 *      $(document).ready(function(){
 *
 *          // DOM이 ready된 시점에서 초기화 함수를 호출한다.
 *          $().BxTooltip().initialize();
 *
 *          // 엘리먼트로 접근하여 property를 가져올 수 있다.
 *          var tooltipElem = document.getElementById('tooltip1');
 *
 *          console.log(tooltipElem.for);
 *          console.log(tooltipElem.type);
 *          console.log(tooltipElem.fill);
 *          console.log(tooltipElem.fixed);
 *          console.log(tooltipElem.content);
 *          console.log(tooltipElem.width);
 *          console.log(tooltipElem.position);
 *
 *          // readonly가 아닌 property는 초기 설정 값을 변경할 수 있다.
 *          // event callback property로 접근할 때는 property명의 '-' 를 생략한다. ('bx-show' -> 'bxshow')
 *          tooltipElem.bxshow = function(){ console.log('bx-show'); };
 *      });
 *
 * </script>
 *
 * //---------------------------------------------------------------------------------------------------------
 * //---------------------------------------------------------------------------------------------------------
 *
 * <button type="button" id="tooltip">툴팁</button>
 * <bx-tooltip for="tooltip"
 *             type="default"
 *             fill="true"
 *             fixed="false"
 *             content="툴팁입니다."
 *             width="200"
 *             position="TC"
 *             bx-show="function(){ console.log('show')}"
 *             bx-hide="function(){ console.log('hide')}"
 *             bx-render="function(){ console.log('render')}">
 * </bx-tooltip>
 *
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

//const PREFIX = '-bx-ui-';

const ERR_MSG = {
    notTargetElement: '연결할 TargetElement Id를 입력하여야 합니다.',
    notSupportPositionType: '지원하지 않는 Position입니다.'
};

const SIZES = {
    "animationSpeed": 500,
    "maxWidth": 300
};

const POSITIONS = {
    "TR": 'TR',
    "RT": 'RT',
    "RM": 'RM',
    "RB": 'RB',
    "BR": 'BR',
    "BC": 'BC',
    "TL": 'TL',
    "BL": 'BL',
    "LB": 'LB',
    "LM": 'LM',
    "LT": 'LT',
    "TC": 'TC'
};


const TOOLTIP_SELECTOR = {
    //className: 'tooltip',
    "bxTooltip": 'bx-tooltip',
    bxTooltipText: 'bx-tooltip-text',
    squareShape: 'square-shape',
    bubbleShape: 'bubble-shape',
    lineType: 'line-type',
    fillType: 'fill-type'
};

const POSITION_CHARS = {
    "top": 'T',
    "bottom": 'B',
    "left": 'L',
    "right": 'R',
    "center": 'C',
    "middle": 'M'
};

const OPTS = {

    /**
     * @property {object} bx-tooltip - tooltip property.
     * @property {string} bx-tooltip.for - <span class='important'>readonly</span> - 연결시킬 element의 ID
     * @property {string} [bx-tooltip.type = 'default'] - <span class='important'>readonly</span> - tooltip의 타입 (default, square)
     * @property {boolean} [bx-tooltip.fill = true] - <span class='important'>readonly</span> - 색상 채움 여부
     * @property {boolean} [bx-tooltip.fixed = false] - <span class='important'>readonly</span> - 위치 고정 여부
     * @property {string} [bx-tooltip.content = ''] - content 값
     * @property {int} [bx-tooltip.width] - <span class='important'>readonly</span> - tooltip의 넓이
     * @property {string} [bx-tooltip.position = 'TC'] - <span class='important'>readonly</span> - tooltip의 위치 및 방향
     *                                                   <br/><span class="important"> position: 'TL', 'TC', 'TR','RT', 'RM', 'RB','LT, 'LM', 'LB','BL', 'BC', 'BR'</span>
     * @property {function} [bx-tooltip.bx-show=function(){}] - show 메서드 호출 시, 호출 될 callback 함수
     * @property {function} [bx-tooltip.bx-hide=function(){}] - hide 메서드 호출 시, 호출 될 callback 함수
     * @property {function} [bx-tooltip.bx-render=function(){}] - render 시, 호출될 callback 함수
     *
     * @alias Attributes
     *
     * @memberof BxTooltip
     */

    "for": '',
    //@private
    //"event": 'mouseover',
    "type": 'default',
    "fill": true,
    "fixed": false,
    "content": '',
    "width": 0,
    "position": 'TC',

    /* jshint ignore:start */
    /**
     * @property {function} [bx-show = function(){}] - show 시, 호출될 callback 함수
     * @callback bxshow
     * @memberof BxTooltip
     */
    "bx-show": function (){
    },
    /**
     * @property {function} [bx-hide = function(){}] - hide 시, 호출될 callback 함수
     * @callback bxhide
     * @memberof BxTooltip
     */
    "bx-hide": function (){
    },
    /**
     * @property {function} [bx-render = function(){}] - render 시, 호출될 callback 함수
     * @callback bxrender
     * @memberof BxTooltip
     */
    "bx-render": function (){
    }
    /* jshint ignore:end */
};

const READONLY_ATTRS = ['for', 'type', 'fill', 'fixed', 'width', 'position'];

const APIS = {
    show: _show,
    hide: _hide
};

class BxTooltip {

    constructor(){
        this.opts = OPTS;
    }

    /**
     * 초기화 함수
     *
     * <p>
     *     DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
     * </p>
     *
     * @returns {object} 생성된 인스턴스
     */
    initialize(){

        let $tooltips = this.$root.find(TOOLTIP_SELECTOR.bxTooltip);

        this.elems = [];

        $.each($tooltips, (i, elem) =>{

            let $customElem = $(elem);

            let $container = _parseCustomElement($customElem);

            this.elems.push($container.get(0));
        });

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxTooltip.isCustomElement = true;

/*
 ==================================================================================================================================
 ==================================================================================================================================

 PRIVATE 메서드 영역

 ==================================================================================================================================
 ==================================================================================================================================
 */
/**
 * custom 태그를 native 태그로 치환한다.
 *
 * @param {object} custom jquery element
 *
 * @returns {object} 생성된 $container 엘리먼트
 *
 * @private
 * @memberOf BxTooltip
 */
function _parseCustomElement($customElem = null){

    if (_.isNull($customElem)){
        CONSOLE.prop.throw(BxTooltip, '$customElem', 'type');
    }

    if (!_.isFunction($customElem.get) && !_.isPlainObject($customElem)){
        CONSOLE.prop.throw(BxTooltip, '$customElem', 'type');
    }

    let targetId = $customElem.attr('for');
    let $target = $(`#${targetId}`);

    //'for' validation
    if (!_.isElement($target.get(0))){
        CONSOLE.prop.throw(BxTooltip, ERR_MSG.notTargetElement);
    }

    let $container = BASE.appendContainerElement(BxTooltip.$root, BxTooltip.className);

    let opts = _createComponentOptions($customElem, OPTS);

    let $component = _createComponentElement($customElem, opts);

    $container.append($component);
    $customElem.remove();

    $target.after($container);

    _setElementProperties($component, opts);

    _setContainerForm($component, opts);

    _init($target, $component);

    _setAPI($component);

    return $container;
}

function _createComponentElement($elem = null, opts = {}){

    let html = [];

    let content = opts.content;

    let targetId = $elem.attr('for');

    if (opts.attr.class){
        opts.attr.class = `${TOOLTIP_SELECTOR.bxTooltip} ${opts.attr.class}`;
    }
    else{
        opts.attr.class = TOOLTIP_SELECTOR.bxTooltip;
    }

    let attr = UTIL.objectToStringAttr(opts.attr);

    html.push(`<div for="${targetId}" ${attr}>`);
    html.push(`<span class="${TOOLTIP_SELECTOR.bxTooltipText}">${content}</span>`);
    html.push(`</div>`);

    return $(html.join(''));
}
/**
 * component option 반환 함수
 *
 * @param {object} custom jquery element
 * @param {object} 기본값 옵션 객체
 *
 * @return {object} component opts
 *
 * @private
 * @memberOf BxTooltip
 */
function _createComponentOptions($elem = null, opts = {}){

    opts = UTIL.createComponentOptions($elem, opts, {}, function (k, v){

        if (k === 'fill' || k === 'fixed'){

            if (!_.isBoolean(v)){
                CONSOLE.prop.throw(BxTooltip, k, 'type');
            }

            //v = v === 'true' || v === true ? true : false;
        }

        if (k === 'width'){

            //v = parseInt(v);

            if (!TYPE.isNumber(v)){
                CONSOLE.prop.throw(BxTooltip, k, 'type');
            }
        }

        if (k === 'position'){

            if (_.isUndefined(POSITIONS[v])){
                CONSOLE.prop.throw(BxTooltip, ERR_MSG.notSupportPositionType);
            }
        }

        return v;

    });

    return _setMandatoryOPTS(opts);
}

/**
 * tooltip property 설정 함수
 *
 * @param {object} component opts
 * @param {object} property 셋팅 할 element
 *
 * @return {object} 생성된 인스턴스
 *
 * @private
 * @memberOf BxTooltip
 */
function _setElementProperties($elem = null, opts = {}){

    let elem = $elem.get(0);

    _.map(opts, (v, k) => {

        // attr 속성은 무시한다.
        if (k === 'attr') return;
        // events 속성은 무시한다.
        if (k === 'events') return;

        if (k.match(/^bx-/)){

            let _k = k.replace('-', '');

            delete opts[k];

            k = _k;
        }

        // 읽기 전용 속성
        if (_.indexOf(READONLY_ATTRS, k) > -1){
            UTIL.setReadOnlyProperty(elem, k, v);
            return;
        }

        let _k = UTIL.convertInternalPropertyName(k);

        elem[_k] = v;

        UTIL.setPropertyWithDescriptor(elem, k, {
            configurable: true,
            enumerable: true,
            get: function () {
                return this[_k];
            },
            set: function (v) {

                if (k === 'content'){

                    if (!_.isString(v) || !_.trim(v).length){
                        CONSOLE.prop.throw(BxTooltip, k, 'all');
                    }

                    let $bxTooltipText = $elem.find(`.${TOOLTIP_SELECTOR.bxTooltipText}`);
                    $bxTooltipText.html(v);
                }

                elem[_k] = v;
            }
        });
    });

    return this;
}

/**
 * tooltip event 바인드
 *
 * @param {object} target jquery element
 * @param {object} container jquery element
 * @param {object} component opts
 *
 * @private
 * @memberOf BxTooltip
 */
function _init($target = null, $component = null){

    let enablePosition = null;
    let titleTemp = null;

    $target.hover(() =>{

            titleTemp = $target.prop('title');
            if (titleTemp) $target.removeAttr('title');

            $component.stop(true, true).fadeIn(SIZES.animationSpeed);

            enablePosition = _checkEnablePosition($target, $component);

            _setPosition($target, $component, enablePosition);

            //callback
            $component.prop('bxshow')();

        }, () =>{
            $component.stop(true, true).fadeOut();

            if (titleTemp){
                $target.prop('title', titleTemp);
            }

            //callback
            $component.prop('bxhide')();
        }
    );
}


/**
 * component option 필수 값 세팅 후 반환하는 함수
 *
 * @param {object} 기본값 옵션 객체
 *
 * @return {object} component opts
 *
 * @private
 * @memberOf BxInput
 */
function _setMandatoryOPTS(opts = {}){

    let attr = opts.attr;

    let id = attr.id;

    if (!id) {
        attr.id = UTIL.uniqueId(TOOLTIP_SELECTOR.bxTooltip);
    }

    return opts;
}

/**
 * tooltip 속성에 따른 모양 설정
 *
 * @param {object} jquery element
 * @param {object} component opts
 *
 * @private
 * @memberOf BxTooltip
 */
function _setContainerForm($component = null, opts = {}){

    if (opts.type === 'square'){
        $component.addClass(TOOLTIP_SELECTOR.squareShape);
    }
    else{
        $component.addClass(TOOLTIP_SELECTOR.bubbleShape);
    }

    if (opts.fill){
        $component.addClass(TOOLTIP_SELECTOR.fillType);
    }
    else{
        $component.addClass(TOOLTIP_SELECTOR.lineType);
    }

    if (opts.width){
        $component.find('.' + TOOLTIP_SELECTOR.bxTooltipText).css('width', opts.width);
    }
    else{
        $component.find('.' + TOOLTIP_SELECTOR.bxTooltipText).css('max-Width', SIZES.maxWidth);
    }
}
/**
 * tooltip position 가능 여부 확인
 *
 * @param {object} target jquery element
 * @param {object} container jquery element
 *
 * @return {string} enablePosition
 *
 * @private
 * @memberOf BxTooltip
 */
function _checkEnablePosition($target = null, $container = null){

    let ret = '';

    let position = $container.prop('position');

    if ($container.prop('fixed')){

        ret = position;
    }
    else{

        //현재 포지션에서 tooltip이 벗어나는지 체크
        let currentPosition = position;

        _setPosition($target, $container, currentPosition);

        if (UTIL.checkVisibleArea($container).isFullVisible){
            ret = currentPosition;
        }
        else{

            let areaLists = [];

            //12방향에 대한 넓이를 areaLists에 각각 저장
            _.map(POSITIONS, (position) =>{
                _setPosition($target, $container, position);
                areaLists[position] = UTIL.checkVisibleArea($container).visibleArea;
            });

            //areaLists를 정렬하여 bestPositionLists에 저장 후, 0번째 값을 return
            let bestPositionLists = Object.keys(areaLists).sort((a, b) =>{
                return areaLists[b] - areaLists[a];
            });

            ret = bestPositionLists[0];
        }
    }

    return ret;
}

/**
 * tooltip position 설정 함수
 *
 * @param {object} target jquery element
 * @param {object} container jquery element
 * @param {string} 설정 할 position
 *
 * @private
 * @memberOf BxTooltip
 */
function _setPosition($target = null, $container = null, position = ''){

    let $bxTooltipText = $container.find(`.${TOOLTIP_SELECTOR.bxTooltipText}`);

    _.forEach(POSITIONS, (position) =>{
        $bxTooltipText.removeClass('position-' + position.toLowerCase());
    });

    $bxTooltipText.addClass('position-' + position.toLowerCase());
    $container.css({'top': '', 'bottom': '', 'left': '', 'right': ''});

    //type option에 따른 설정 값 적용
    let calc = 0;
    let leftCalc = 0;

    if ($container.prop('type') === 'default'){
        calc = 10;
    }
    else{
        calc = 3;
    }

    if (position.split('')[1] === POSITION_CHARS.left || position.split('')[1] === POSITION_CHARS.right){
        leftCalc = 7;
    }

    //위치 설정에 필요한 값 정의
    let {left: targetLeft, top: targetTop} = $target.position();

    let targetWidth = $target.outerWidth(true);
    let targetHeight = $target.outerHeight(true);

    let containerWidth = $container.width();
    let containerHeight = $container.height();

    let center = (targetWidth - containerWidth) / 2;
    let middle = (targetHeight - containerHeight) / 2;

    //top, bottom 위치 설정
    if (position.charAt(0) === POSITION_CHARS.top || position.charAt(0) === POSITION_CHARS.bottom){

        switch (position.charAt(1)){

            case POSITION_CHARS.left :
                $container.css('left', targetLeft - containerWidth - leftCalc);
                break;
            case POSITION_CHARS.center :
                $container.css('left', targetLeft + center);
                break;
            case POSITION_CHARS.right :
                $container.css('left', targetLeft + targetWidth + leftCalc);
                break;
        }

        if (position.charAt(0) === POSITION_CHARS.top){

            $container.css('top', targetTop - containerHeight - calc);
        }
        else{
            $container.css('top', targetTop + targetHeight + calc);
        }
    }
    //right, left 위치 설정
    if (position.charAt(0) === POSITION_CHARS.right || position.charAt(0) === POSITION_CHARS.left){

        switch (position.charAt(1)){

            case POSITION_CHARS.top :
                $container.css('top', targetTop);
                break;
            case POSITION_CHARS.middle :
                $container.css('top', targetTop + middle);
                break;
            case POSITION_CHARS.bottom :
                $container.css('top', targetTop + targetHeight - containerHeight);
                break;
        }

        if (position.charAt(0) === POSITION_CHARS.left){

            $container.css('left', targetLeft - containerWidth - calc);
        }
        else{
            $container.css('left', targetLeft + targetWidth + calc);
        }
    }
}

/*
 * ========================================================================================================
 * below function is API function.
 * ========================================================================================================
 */

 /**
 * API 설정 함수
 *
 * @param {object} [elem =null]- BxTextarea DOM
 * @private
 * @memberOf BxTextarea
 */
function _setAPI($elem = null){

    let elem = $elem.get(0);

    _.map(APIS, (v, k) => {
        UTIL.setReadOnlyProperty(elem, k, v);
    });
}

/**
 * show 함수
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias show
 *
 * @memberOf BxTooltip
 */
function _show(){

    let $elem = $(this);

    let targetId = $elem.attr('for');
    let $target = $(`#${targetId}`);

    let title = $target.prop('title');

    let $bxTooltipText = $elem.find(`.${TOOLTIP_SELECTOR.bxTooltipText}`);

    //동적 title 처리
    if (title){
        $bxTooltipText.attr('data-title', title);
        $target.removeAttr('title');
    }

    $elem.stop(true, true).fadeIn(SIZES.animationSpeed);

    let enablePosition = _checkEnablePosition($target, $elem);

    _setPosition($target, $elem, enablePosition);

    //callback
    $elem.prop('bxshow')();

    return this;
}
/**
 * hide 함수
 *
 * @returns {object} 생성된 인스턴스
 *
 * @alias hide
 *
 * @memberOf BxTooltip
 */
function _hide(){

    let $elem = $(this);

    let targetId = $elem.attr('for');
    let $target = $(`#${targetId}`);

    let $bxTooltipText = $elem.find(`.${TOOLTIP_SELECTOR.bxTooltipText}`);

    let title = $bxTooltipText.attr('data-title');

    //동적 title 처리
    if (title){
        $target.attr('title', title);
        $bxTooltipText.removeAttr('data-title');
    }

    $elem.stop(true, true).fadeOut();

    //callback
    $elem.prop('bxhide')();

    return this;
}

module.exports = BxTooltip;
