// bxuip-angular.js-ui/src/lib/bx-ui/app/components/tree.js
/**
 *  랜더링 완료 시, 호출될 callback 함수
 *
 *  @callback onLoaded
 *  @param {object} e - 이벤트 객체
 *  @param {Object} data - 트리 인스턴스
 *  @memberOf BxTree
 */

/**
 *  노드 선택 시, 호출될 callback 함수
 *
 *  @callback onSelected
 *  @param {object} e - 이벤트 객체
 *  @param {Object} data - 노드 객체
 *  @memberOf BxTree
 */

/**
 *  노드 더블클릭 시, 호출될 callback 함수
 *
 *  @callback onDblClick
 *  @param {object} e - 이벤트 객체
 *  @param {Object} data - 노드 객체
 *  @memberOf BxTree
 */

/**
 *  드래그 시작 이벤트 callback 함수
 *
 *  @callback onDragStart
 *  @param {object} e - 이벤트 객체
 *  @param {Object} data - 노드 객체
 *  @param {Object} data.current - 선택 된 node 객체
 *  @param {Object} data.current.node - 선택 된 node id
 *  @param {Object} data.current.element - 선택 된 DOM
 *  @param {Object} data.parent - 부모 node 객체
 *  @param {Object} data.parent.node - 부모 node id
 *  @param {Object} data.parent.element - 부모 DOM
 *  @memberOf BxTree
 */

/**
 *  드래그 이벤트 callback 함수
 *
 *  @callback onDrag
 *  @param {object} e - 이벤트 객체
 *  @param {Object} data - 노드 객체
 *  @param {Object} data.current - 선택 된 node 객체
 *  @param {Object} data.current.node - 선택 된 node id
 *  @param {Object} data.current.element - 선택 된 DOM
 *  @param {Object} data.parent - 부모 node 객체
 *  @param {Object} data.parent.node - 부모 node id
 *  @param {Object} data.parent.element - 부모 DOM
 *  @memberOf BxTree
 */

/**
 *  드래그 종료 이벤트 callback 함수
 *
 *  @callback onDragStop
 *  @param {object} e - 이벤트 객체
 *  @param {Object} data - 노드 객체
 *  @param {Object} data.current - 선택 된 node 객체
 *  @param {Object} data.current.node - 선택 된 node id
 *  @param {Object} data.current.element - 선택 된 DOM
 *  @param {Object} data.parent - 부모 node 객체
 *  @param {Object} data.parent.node - 부모 node id
 *  @param {Object} data.parent.element - 부모 DOM
 *  @param {function} reject - 트리 반영 거부 함수
 *  @memberOf BxTree
 */

/**
 * @class BxTree
 * @classDesc BxTree 클래스 생성자
 *
 * @param {object} OPTS
 * @param {object[]} [OPTS.colModel = []] colModel 을 통해, 각 item columnName(id, text) 을 유동적으로 할당할 수 있다.
 * @param {string} OPTS.colModel.name - item columnName 과 매칭될 name
 * @param {string} OPTS.colModel.type - item columnType 과 매칭될 type
 * @param {object} [OPTS.data = {}] - 트리 데이터
 * @param {string} OPTS.data.id - 아이디
 * @param {string} OPTS.data.text - 노드 텍스트
 * @param {string} OPTS.data.icon - 해당 노드 개별 아이콘 설정
 * @param {object[]} OPTS.data.children - 자식 아이템 집합
 * @param {boolean} [OPTS.enableOpenAll = true] - 초기 모든 노드에 대한 열기 여부 <br/><b class="dec">(stateOptions.fold = true 일 경우, fold 상태값이 우선적으로 적용)</b>
 * @param {object} [OPTS.stateOptions = {}] - 상태 저장 옵션
 * @param {string} [OPTS.stateOptions.key  = '']- 트리 고유의 state key <br/><b class="dec">(필수 값)</b>
 * @param {boolean} [OPTS.stateOptions.expire = true] - 브라우저 종료 시 state 초기화
 * @param {boolean} [OPTS.stateOptions.fold = false] - 열림 닫힘 저장
 * @param {boolean} [OPTS.stateOptions.checkbox = false] - 체크박스 저장
 * @param {boolean} [OPTS.stateOptions.selected = false] - 선택 노드 저장
 * @param {boolean} [OPTS.stateOptions.disabled = false] - 비활성 노드 저장
 * @param {string} [OPTS.animation = 0] - duration
 * @param {object} [OPTS.icons = {}] - 아이콘 집합
 * @param {boolean} [OPTS.enableDeleteRootNode = false] - root 노드 삭제 가능 여부
 * @param {boolean} [OPTS.enableCheckbox = true] - 체크 박스 사용 여부
 * @param {object} [OPTS.checkBoxOptions = {}] - 체크 박스 옵션 기능들을 정할 수 있다
 * @param {boolean} [OPTS.checkBoxOptions.enableCheckRelated = false] - 상위 노드까지 체크의 유무
 * @param {boolean} [OPTS.checkBoxOptions.enableCheckLabel = false] - 라벨을 클릭 했을때 체크의 유무
 * @param {boolean} [OPTS.enableDnd = false] - drag and drop 사용 여부
 * @param {object} [OPTS.dndOptions = {}] - drag and drop 옵션 기능들을 정할 수 있다
 * @param {string} [OPTS.dndOptions.insertPosition = 'before'] - 하위노드가 있는 노드로의 이동 시, 위치를 상단 또는 하단에 지정할 수 있다 ('before', 'after')
 * @param {boolean} [OPTS.enableLine = true] - 관계선 여부
 * @param {string} [OPTS.lineType = 'default'] - 관계선 타입(default, dot)
 * @param {function} [OPTS.onSelected = function(){}] - 노드 선택 시, 호출될 callback 함수
 * @param {function} [OPTS.onDragStart = function(){}] - 드래그 시작 이벤트 callback 함수
 * @param {function} [OPTS.onDrag = function(){}] - 드래그 이벤트 callback 함수
 * @param {function} [OPTS.onDragStop = function(){}] - 드래그 종료 이벤트 callback 함수
 * @param {function} [OPTS.onDblClick = function(){}] - 노드 더블 클릭 시, 호출될 callback 함수
 * @param {function} [OPTS.onLoaded = function(){}] - 랜더링 완료 시, 호출될 callback 함수
 *
 * @returns {Object} - 생성된 인스턴스
 *
 * @example

 var idx = 0;

 var tree = $('#tree').BxTree({
    colModel: [
        {
            name: 'colId',
            type: 'id'
        },
        {
            name: 'colText',
            type: 'text'
        }
    ],
    data: {
        colId : 'root',
        colText : '<span style="color:red">ROOT</span>',
        icon : 'root-icon-class-name',
        children : [
            {
                colId : 'child1',
                colText : 'child1',
                icon : 'child-icon-class-name'
            },
            {
                colId : 'child2',
                colText : 'child2',
                children : [
                    {
                        colId: 'child2-1',
                        colText: 'child2-1'
                    },
                    {
                        colId: 'child2-2',
                        colText: 'child2-2'
                    }
                ]
            }
        ]
    },
    animation : 500,
    icons: {
        open: '',
        close: '',
        none: ''
    },
    //전체 노드 열림 or 닫힘 여부
    enableOpenAll: true,
    //상태 저장 옵션
    stateOptions: {
        key : 'tree',
        expire : true,
        fold : false,
        checkbox : false,
        selected : false,
        disabled : false
    },
    // 루트 노드를 삭제할 수 있는지 여부
    enableDeleteRootNode: false,
    // 체크 박스 사용 여부
    enableCheckbox: true,
    // 체크 박스가 True 일 경우만 작동
    checkBoxOptions: {
        enableCheckRelated : true, // 상위 노드까지 체크의 유무
        enableCheckLabel : false // 라벨을 클릭 했을때 체크의 유무
    },
    // drag and drop 사용 여부
    enableDnd : false,
    // enableDnd가 true일 경우만 작동
    dndOptions: {
        insertPosition : 'before' // 하위노드가 있는 노드로의 이동 시, 상단 또는 하단에 위치 지정
    },
    //관계선 여부
    enableLine : true,
    lineType : 'default',
    onSelected: function (e, data) {

        var node = data.node;

        this.addNode({
            id: 'newId' + (idx++), text: 'testName', icon: 'bw-bx-icon i-15 i-ctrl-null', callback: function(){
           console.log(this);
        }});

        this.renameNode('reNameNode');

    },
    onDragStart: function (e, data) {

        console.log('onDragStart!!');
        console.log(data);

    },
    onDrag: function (e, data) {

        console.log('onDrag!!');
        console.log(data);

    },
    onDragStop: function (e, data, reject) {

        console.log('onDragStop!!');
        console.log(id);

    },
    onDblClick: function (e, data) {
        console.log('onDblClick!!');
        console.log(data);
    },
    onLoaded: function (e, data) {
        console.log('onLoaded!!');
        console.log(data);
    }
 }).render();
 *
 *
 */

const BXUI = require('../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../base');
const CONSOLE = BXUI.Console;
const UTIL = BXUI.Util;
const STORAGE = BXUI.Storage;

// declare Tree Class
const TREE_CLASS = {
    bxUiTree: '-bx-ui-tree',
    lineType: 'line-type',
    dotType: 'dot-type',
    noneType: 'none-type',
    jstreeContainerUl: 'jstree-container-ul'
};

//default state options
const DEFAULT_STATE_OPTIONS = {
    key : '',
    expire : true,
    fold : false,
    checkbox : false,
    selected : false,
    disabled : false
};

//dnd options
const DND_OPTIONS = {
    before : 'first',
    after : 'last',
    openTime : 100
};

const OPTS = Symbol();
const THIS = Symbol();

class BxTree {

    constructor({
        colModel = [],
        data = {},
        enableOpenAll = true,
        stateOptions,
        enableCheckbox = false,
        checkBoxOptions = {},
        enableDnd = false,
        dndOptions = {},
        // icon 크기 조정.gru
        // small, medium, large
        variant = 'large',
        // 여러개 하위 노드 선택
        //multiple = true, // default : true
        // anmation 시간
        animation = 0,
        //// 더블 클릭시, 토글
        //dblclick_toggle = false,

        //visible = true,
        //keep_selected_style = false,
        //whole_node = false,
        //three_state = false,
        icons = {},
        enableDeleteRootNode = false,
        enableLine = false,
        lineType = 'default',
        onSelected = function (){
        },
        onDragStart = function (){
        },
        onDrag = function (){
        },
        onDragStop = function (){
        },
        onLoaded = function (){
        },
        onDblClick = function (){
        },

        // plugin 설정
        // *   wholerow : row 전체 선택, 해당 플러그인 없을 시 이름에 해당부분에만 selected css적용
        // *  checkbox : 체크박스 사용
        // *  dnd : drag & drop plguin
        // *  changed : drag & drop 플러그인 사용 시, tree 구조 변화 적용
        // *  contextmenu : 우클릭시 contextmenu 사용가능.
        // *  unique : 같은 이름의 노드를 형제 노드로 갖지 않음.
        plugins = []
    } = {}){
        if (!_.isArray(colModel)){
            CONSOLE.prop.throw(BxTree, 'colModel', 'all');
        }

        if (!_.isPlainObject(icons)){
            CONSOLE.prop.throw(BxTree, 'icons', 'type');
        }

        if (!_.isPlainObject(data)){
            CONSOLE.prop.throw(BxTree, 'data', 'all');
        }

        if (!_.isBoolean(enableOpenAll)){
            CONSOLE.prop.throw(BxTree, 'enableOpenAll', 'type');
        }

        if (!_.isUndefined(stateOptions)) {

            if (_.isPlainObject(stateOptions)){
                initStateOptions(stateOptions);
            }
            else {
                CONSOLE.prop.throw(BxTree, 'stateOptions', 'type');
            }
        }

        function initStateOptions(stateOptions = {}){

            const { key, expire, fold, checkbox, selected, disabled } = stateOptions;

            if (_.isUndefined(key)) stateOptions.key = DEFAULT_STATE_OPTIONS.key;

            if (_.isUndefined(expire)) stateOptions.expire = DEFAULT_STATE_OPTIONS.expire;

            if (_.isUndefined(fold)) stateOptions.fold = DEFAULT_STATE_OPTIONS.fold;

            if (_.isUndefined(checkbox)) stateOptions.checkbox = DEFAULT_STATE_OPTIONS.checkbox;

            if (_.isUndefined(selected)) stateOptions.selected = DEFAULT_STATE_OPTIONS.selected;

            if (_.isUndefined(disabled))  stateOptions.disabled = DEFAULT_STATE_OPTIONS.disabled;


            if (_.isString(stateOptions.key)) {

                if (_.isEmpty(stateOptions.key.trim())) {
                    CONSOLE.prop.throw(BxTree, 'stateOptions.key', 'value');
                }
            }
            else {
                CONSOLE.prop.throw(BxTree, 'stateOptions.key', 'type');
            }

            if (!_.isBoolean(stateOptions.expire)){
                CONSOLE.prop.throw(BxTree, 'stateOptions.expire', 'type');
            }

            if (!_.isBoolean(stateOptions.fold)){
                CONSOLE.prop.throw(BxTree, 'stateOptions.fold', 'type');
            }

            if (!_.isBoolean(stateOptions.checkbox)){
                CONSOLE.prop.throw(BxTree, 'stateOptions.checkbox', 'type');
            }

            if (!_.isBoolean(stateOptions.selected)){
                CONSOLE.prop.throw(BxTree, 'stateOptions.selected', 'type');
            }

            if (!_.isBoolean(stateOptions.disabled)){
                CONSOLE.prop.throw(BxTree, 'stateOptions.disabled', 'type');
            }
        }

        if (!_.isBoolean(enableDeleteRootNode)){
            CONSOLE.prop.throw(BxTree, 'enableDeleteRootNode', 'type');
        }

        if (!_.isBoolean(enableDeleteRootNode)){
            CONSOLE.prop.throw(BxTree, 'enableDeleteRootNode', 'type');
        }

        if (!_.isBoolean(enableLine)){
            CONSOLE.prop.throw(BxTree, 'enableLine', 'type');
        }

        if (!_.isString(lineType)){
            CONSOLE.prop.throw(BxTree, 'lineType', 'type');
        }

        if (!_.isBoolean(enableCheckbox)){
            CONSOLE.prop.throw(BxTree, 'enableCheckbox', 'type');
        }

        if (!_.isPlainObject(checkBoxOptions)){
            CONSOLE.prop.throw(BxTree, 'checkBoxOptions', 'type');
        }

        if (!_.isBoolean(enableDnd)){
            CONSOLE.prop.throw(BxTree, 'enableDnd', 'type');
        }

        if (!_.isPlainObject(dndOptions)){
            CONSOLE.prop.throw(BxTree, 'dndOptions', 'type');
        }

        if (!_.isFunction(onSelected)){
            CONSOLE.prop.throw(BxTree, 'onSelected', 'type');
        }

        if (!_.isFunction(onDragStart)){
            CONSOLE.prop.throw(BxTree, 'onDragStart', 'type');
        }

        if (!_.isFunction(onDrag)){
            CONSOLE.prop.throw(BxTree, 'onDrag', 'type');
        }

        if (!_.isFunction(onDragStop)){
            CONSOLE.prop.throw(BxTree, 'onDragStop', 'type');
        }

        if (!_.isFunction(onLoaded)){
            CONSOLE.prop.throw(BxTree, 'onLoaded', 'type');
        }

        if (!_.isFunction(onDblClick)){
            CONSOLE.prop.throw(BxTree, 'onDblClick', 'type');
        }

        icons.open = icons.open || 'bw-tree-icon i-tree-open';
        icons.close = icons.close || 'bw-tree-icon i-tree-close';
        icons.none = icons.none || 'bw-tree-icon i-tree-transparent';

        //let core = {themes : {variant}, multiple, animation, check_callback,  dblclick_toggle, data};



        let core = {
            multiple: false, //multiple 지원하지 않음
            themes: {variant},
            animation,
            data,
            /* jshint camelcase: false */
            dblclick_toggle: false,
            /* jshint camelcase: false */
            check_callback: _getCheckCallbackFunction(this)
            // 초기 로드 시 loading... 문구
            // 만약 추가 사항 또는 변경 사항이 발생한다면, 아래 key 값을 통해, 변경할 수 있음
            //strings: {
            //    'Loading ...' : ' '
            //}
        };

        //let checkbox = {visible, keep_selected_style, whole_node, three_state};
        //this[OPTS] = {core, checkbox, plugins};

        this[OPTS] = {core, plugins};

        this.$root = BxTree.$root;

        // 유동 컬럼(이름)을 위한, colModel
        this.colModel = colModel;
        // 기본 colModelName
        this.defaultColNames = ['id', 'text', 'icon'];

        // 모든 node 에 할당되는 icon
        this.icons = icons;
        // root node 삭제 여부
        this.enableDeleteRootNode = enableDeleteRootNode;

        //초기 모든 노드에 대한 열기 여부
        this.enableOpenAll = enableOpenAll;
        //현재 상태 저장 옵션
        this.stateOptions = stateOptions;

        // 선택된 아이템
        this.selectedNode = null;

        // Check Box 초기화
        this.enableCheckbox = enableCheckbox;
        this.checkBoxOptions = checkBoxOptions;

        //Drag and Drop 초기화
        this.enableDnd = enableDnd;
        this.dndOptions = dndOptions;

        // 변화에 대한 트리 반영 여부
        this.apply = true;

        //관계선 여부
        this.enableLine = enableLine;
        this.lineType = lineType;

        // Event
        this.onSelected = onSelected;
        this.onDragStart = onDragStart;
        this.onDrag = onDrag;
        this.onDragStop = onDragStop;
        this.onLoaded = onLoaded;
        this.onDblClick = onDblClick;
    }

    /**
     * 랜더 함수
     *
     * @returns {object} - 생성된 인스턴스
     */
    render(){
        // Render 시점에서 Container 생성 후 Root에 Append
        this.$container = BASE.appendContainerElement(this.$root, BxTree.className);
        this.$container.get(0)[THIS] = this;

        let checkBoxOptions = this.checkBoxOptions;
        let dndOptions = this.dndOptions;
        let stateOptions = this.stateOptions;
        let opts = this[OPTS];

        //state session 설정
        if (stateOptions) _setSession.call(this);

        // node 데이터 초기화
        if (!_.isEmpty(opts.core.data)){
            _initNodeData.call(this, opts.core.data);
            this.enableCheckbox && _initCheckBox.call(this, checkBoxOptions.enableCheckRelated, checkBoxOptions.enableCheckLabel);
            this.enableDnd && _initDnd.call(this, dndOptions.insertPosition);
        }

        // 트리 초기화
        this.$container.jstree(opts);

        // 각 이벤트 바인딩
        this.$container.on('ready.jstree', (e, data) =>{

            if (!_.isEmpty(opts.core.data)){

                //state option 설정
                if (stateOptions && _getItems(stateOptions.key + '_state')){

                    _getOptionState.call(this);

                    if (!stateOptions.fold){
                        _initFirstOpen.call(this, data);
                    }

                }
                else{
                    _initFirstOpen.call(this, data);
                }
                _initSettingLine.call(this);
            }
            _setCustomIconHandler.call(this);

            this.onLoaded.apply(this, [e, data]);
        });

        this.$container.on('refresh.jstree', (e, data) =>{

            if (!_.isEmpty(opts.core.data)){
                _initFirstOpen.call(this, data);
                _initSettingLine.call(this);
            }

            this.onLoaded.apply(this, [e, data]);
        });

        if (stateOptions){

            if (stateOptions.fold){

                this.$container.on('open_node.jstree', _setOptionState.bind(this));

                this.$container.on('close_node.jstree', _setOptionState.bind(this));

            }

            if (stateOptions.checkbox){

                this.$container.on('check_node.jstree', _setOptionState.bind(this));

                this.$container.on('uncheck_node.jstree', _setOptionState.bind(this));

            }

            if (stateOptions.disabled){

                this.$container.on('disable_node.jstree', _setOptionState.bind(this));

                this.$container.on('enable_node.jstree', _setOptionState.bind(this));

            }
        }

        this.$container.on('move_node.jstree', (e, data) =>{

            //parent 노드의 child 유무에 따른 아이콘 셋팅
            let newParent = this.getNode(data['parent']);
            let oldParent = this.getNode(data['old_parent']);

            if (newParent.children.length === 1) this.setCloseIcon(newParent);
            else this.setOpenIcon(newParent);

            if (oldParent.children.length === 0) this.setNoneIcon(oldParent);

        });
        // click과 double click event는 함께 바인딩이 안되므로, select_node에서 처리한다.
        this.$container.on('select_node.jstree', (e, data) =>{

            let event = data.event;

            // 사용자의 action이 아닌 select_node API에 의해 trigger된 경우에는
            // 이벤트 핸들러를 실행하지 않는다.
            if (!event) return;

            let node = data.node;

            this.selectedNode = node;

            let $this = $(event.currentTarget).find('.text-node');

            let ret = {
                'node' : data.selected,
                'element' : $(`#${data.selected}`)[0]
            };

            // double click
            if ($this.hasClass('clicked')){

                $this.removeClass('clicked');
                this.onDblClick.apply(this, [e, ret]);

            }
            // click
            else{

                $this.addClass('clicked');
                setTimeout(() =>{
                    if ($this.hasClass('clicked')){

                        $this.removeClass('clicked');
                        this.onSelected.apply(this, [e, ret]);
                    }
                }, 200);

            }

            if (stateOptions && stateOptions.selected) _setOptionState.call(this);

        });

        return this;
    }

    /**
     * 노드 추가 함수
     *
     * @param {object} opts
     * @param {string} [opts.id=''] - 노드 아이디
     * @param {string} [opts.text=''] - 노드 타이틀
     * @param {string} [opts.icon=''] - 아이콘 클래스
     * @param {Function} [opts.callback = function(){}] - 노드 추가 후, 호출될 callback 함수
     *
     * @returns {object} - 생성된 아이템
     */
    addNode({
        id = '',
        text = '',
        icon = '',
        callback = function (){
        },
    } = {}){

        // 선택된 node 가 없을 경우(즉 부모 노드가 존재하지 않을 경우)
        if (!this.selectedNode){
            CONSOLE.prop.throw(BxTree, '선택한 node 데이터', 'value');
            return this;
        }

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxTree, 'id', 'type');
            return this;
        }

        if (!_.isString(text) || !_.trim(text).length){
            CONSOLE.prop.throw(BxTree, 'text', 'all');
            return this;
        }

        if (!_.isString(icon)){
            CONSOLE.prop.throw(BxTree, 'icon', 'type');
            return this;
        }

        if (!_.isFunction(callback)){
            CONSOLE.prop.throw(BxTree, 'callback', 'type');
            return this;
        }

        // id 값이 없는 경우에는 jsTree 가 자동으로 id 값을 생성한다.
        let node = {id, text: _getNodeHTML(text, icon), icon: this.icons.none};

        // jstree는 생성된 노드의 아이디를 반환
        let _id = this.$container.jstree('create_node', this.selectedNode, node, 'last', callback);

        // node 추가 후 오픈
        this.open(this.selectedNode);

        return _.assign(node, {id:_id});
    }

    /**
     * 노드 반환 함수
     *
     * @param {string} id - 노드 아이디
     *
     * @returns {object} - 노드 객체
     */
    getNode(id = ''){

        if (!_.isString(id) || !_.trim(id).length){
            CONSOLE.prop.throw(BxTree, 'id', 'all');
            return this;
        }

        let node = this.$container.jstree('get_node', id); // {id:id}} 이나 id 는 jstree 안에서는 같은 방식으로 처리 하고 있음
                                                           // 단, '[]' 형태 일 경우 내부에서 jQuery.closest 을 이용해 검색을 하기 때문에 트리가 닫혀 있을 경우 Dom에서 찾지 못하여 Null값을 리턴함
        return !node ? null : _parsing.call(this, node);
    }

    /**
     * 전체 노드 반환 함수
     *
     * @returns {object} - 전체 tree 객체
     */
    getAllNodes(){

        return _getAllNode.call(this);

    }

    /**
     * 선택된 노드를 반환 한다
     *
     * @returns {object} - 노드 객체
     */
    getSelectedNode(){

        if (!this.selectedNode) return null;

        let node = this.$container.jstree('get_node', '[id="' + this.selectedNode.id + '"]');

        return !node ? null : _parsing.call(this, this.selectedNode);
    }

    /**
     * 하나 또는 여러개의 노드를 체크 한다
     *
     * @param {string | string[]} [ids=null]- 노드 아이디
     * @returns {object} - 생성된 인스턴스
     */
    checkNodes(ids = null){

        _validationOfString(ids);

        if (_.isArray(ids)){

            _.map(ids, id =>{

                this.$container.jstree('check_node', id);
            });
        }
        else{
            this.$container.jstree('check_node', ids);
        }

        return this;
    }

    /**
     * 모든 노드를 체크 한다.
     *
     * @returns {object} - 생성된 인스턴스
     */
    checkNodeAll(){

        this.$container.jstree('check_all');

        if (this.stateOptions && this.stateOptions.checkbox) _setOptionState.call(this);

        return this;
    }

    /**
     * 하나 또는 여러개의 노드를 체크 해제 한다.
     *
     * @param {string | string[]} [ids=null]- 노드 아이디
     * @returns {object} - 생성된 인스턴스
     */
    clearCheckedNodes(ids = null){

        _validationOfString(ids);

        if (_.isArray(ids)){

            _.map(ids, id =>{

                this.$container.jstree('uncheck_node', id);
            });
        }
        else{
            this.$container.jstree('uncheck_node', ids);
        }

        return this;
    }

    /**
     * 모든 노드를 체크 해제 한다.
     *
     * @returns {object} - 생성된 인스턴스
     */
    clearCheckedNodeAll(){

        this.$container.jstree('uncheck_all');

        if (this.stateOptions && this.stateOptions.checkbox) _setOptionState.call(this);

        return this;
    }

    /**
     * 하나 또는 여러개의 노드를 비활성화 시킨다.
     *
     * @param {string | string[]} [ids=null]- 노드 아이디
     * @returns {object} - 생성된 인스턴스
     */
    setDisableNodes(ids = null){

        _validationOfString(ids);

        if (_.isArray(ids)){

            _.map(ids, id =>{

                _.forEach(_getChildrensNode.call(this, id, true), (value) =>{ // 여러개의 자식 노드들을 비 활성화 한다

                    this.$container.jstree('disable_node', value);
                    if (!this.checkBoxOptions.enableCheckRelated) return false;
                });
            });
        }
        else{

            _.map(_getChildrensNode.call(this, ids, true), node =>{

                this.$container.jstree('disable_node', node);
            });
        }

        return this;
    }

    /**
     * 하나 또는 여러개의 노드를 활성화 시킨다.
     *
     * @param {string | string[]} [ids=null]- 노드 아이디
     * @returns {object} - 생성된 인스턴스
     */
    setEnableNodes(ids = null){

        _validationOfString(ids);

        if (_.isArray(ids)){

            _.map(ids, id =>{

                _.map(_getChildrensNode.call(this, id, true), node =>{

                    this.$container.jstree('enable_node', node);
                    if (!this.checkBoxOptions.enableCheckRelated) return false;
                });
            });
        }
        else{

            _.map(_getChildrensNode.call(this, ids, true), node =>{

                this.$container.jstree('enable_node', node);
            });
        }

        return this;
    }

    /**
     * 비 활성화된 노드를 전부 반환 시킨다.
     *
     * @returns {array} - 노드 객체들을 배열에 담아 반환 한다.
     */
    getDisabledNodes(){

        let nodeList = _getChildrensNode.call(this, '#'); // jstree 에서는 #.children_d 를 이용하여
        // 각 노드의 ID를 가지고 온다.
        let disabledList = [];

        if (_.isArray(nodeList)){

            _.map(nodeList, id =>{
                if (this.$container.jstree('is_disabled', id)){

                    let node = this.$container.jstree('get_node', id);
                    // map 메서드의 return 문이 존재하지않기때문에, 아래 식은 사용하되않은것으로 간주됨.
                    !node ? null : disabledList.push(_parsing.call(this, node));
                }
            });
        }

        return disabledList;
    }

    /**
     * 체크 되어진 노드를 전부 반환 시킨다.
     *
     * @returns {array} - 노드 객체들을 배열에 담아 반환 한다.
     */
    getCheckedNodes(){

        let nodeList = _getChildrensNode.call(this, '#');
        let checkedList = [];

        if (_.isArray(nodeList)){

            _.map(nodeList, id =>{
                if (this.$container.jstree('is_checked', id) === true){
                    let node = this.$container.jstree('get_node', id);

                    // map 메서드의 return 문이 존재하지않기때문에, 아래 식은 사용하되않은것으로 간주됨.
                    !node ? null : checkedList.push(_parsing.call(this, node));
                }
            });
        }

        return checkedList;
    }

    /**
     * 노드가 체크 되어 있는지 반환 한다.
     *
     * @param {string} id - 노드 아이디
     * @returns {boolean} - true or false
     */
    isChecked(id = null){

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxTree, 'id', 'type');
            return this;
        }

        return this.$container.jstree('is_checked', id) === true;
    }

    /**
     * 노드 삭제 함수
     *
     * @param {object} [node = selectedNode] - 노드 객체
     * @returns {object} - 생성된 인스턴스
     */
    deleteNode(node = {}){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node 데이터', 'type');
            return this;
        }

        node = _.isEmpty(node) ? this.selectedNode : node;

        if (!this.enableDeleteRootNode && this.isRootNode(node)) return;


        // 노드 삭제
        this.$container.jstree('delete_node', node);

        // 삭제한 노드의 부모 노드가 더이상 자식을 갖고 있지 않는 경우 아이콘 변경
        let parentNode = this.getNode(node.parent);

        !this.$container.jstree('is_parent', parentNode) && this.setNoneIcon(parentNode);

        this.selectedNode = null;

        return node;
    }

    /**
     * 노드 이름 변경 함수
     *
     * @param {string} text - 노드 제목
     * @param {object} [node = selectedNode] - 노드 객체
     * @returns {object} - 생성된 인스턴스
     */
    renameNode(text = '', node = {}){

        if (_.isEmpty(this[OPTS].core.data)){
            return this;
        }

        if (!_.isString(text) || !_.trim(text).length){
            CONSOLE.prop.throw(BxTree, 'text', 'all');
            return this;
        }

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        node = _.isEmpty(node) ? this.selectedNode : node;

        text = _getNodeHTML(text, _getIconClassFromNodeHTML(node.text));

        this.$container.jstree('rename_node', node, text);

        return node;
    }

    /**
     * rootNode 구분 함수
     *
     * @param {object} node - 노드 객체
     * @returns {boolean} - Root 노드 여부
     * @private
     */
    isRootNode(node = {}){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        return node.parent === '#';
    }

    /**
     * 노드 열기 함수
     *
     * @param {object} node - 노드 객체
     * @param {boolean} [isOpenRelated=false] - 상위 노드에 대한 열기 여부
     * @returns {object} - 생성된 인스턴스
     */
    open(node = {}, isOpenRelated = false){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        if (!_.isBoolean(isOpenRelated)){
            CONSOLE.prop.throw(BxTree, 'isOpenRelated', 'type');
            return this;
        }

        if (isOpenRelated){
            _.map(_getParentsNode.call(this, node.id, true), (nodeId) =>{ // 선택된 ID의 상위 노드 ID를 가져온다

                if (node.id !== '#' && !this.$container.jstree('is_open', nodeId)){
                    let node = this.getNode(nodeId);

                    this.$container.jstree('open_node', node);
                    !_.isEmpty(node.children) && this.setOpenIcon(node);
                }
            });
        }
        else{
            this.$container.jstree('open_node', node);
            !_.isEmpty(node.children) && this.setOpenIcon(node);
        }
        return this;
    }

    /**
     * 노드 닫기 함수
     *
     * @param {object} node - 노드 객체
     * @param {boolean} [isCloseRelated=false] - 하위 노드에 대한 닫기 여부
     * @returns {object} - 생성된 인스턴스
     */
    close(node = {}, isCloseRelated = false){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        if (!_.isBoolean(isCloseRelated)){
            CONSOLE.prop.throw(BxTree, 'isCloseRelated', 'type');
            return this;
        }

        if (isCloseRelated){
            _.map(_getChildrensNode.call(this, node.id, true), (nodeId) =>{ // 선택된 ID의 상위 노드 ID를 가져온다

                if (node.id !== '#' && this.$container.jstree('is_open', nodeId)){
                    let node = this.getNode(nodeId);

                    this.$container.jstree('close_node', node);
                    !_.isEmpty(node.children) && this.setCloseIcon(node);
                }
            });
        }
        else{
            this.$container.jstree('close_node', node);
            !_.isEmpty(node.children) && this.setCloseIcon(node);
        }

        return this;
    }

    /**
     * 노드 열기 / 닫기 토글 함수
     *
     * @param {object} node - 노드 객체
     * @returns {object} - 생성된 인스턴스
     */
    toggle(node = {}){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        // 자식 노드가 없는 경우
        if (_.isEmpty(node.children)) return this;

        // 노드의 이전 상태(열림, 닫힘)에 따라 아이콘을 세팅한다.
        if (!this.$container.jstree('is_open', node)){
            this.open(node);
        }
        else{
            this.close(node);
        }

        return this;
    }

    /**
     * open 아이콘 설정 함수
     *
     * @param {object} node - 노드 객체
     * @returns {object} - 생성된 인스턴스
     */
    setOpenIcon(node = {}){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        this.$container.jstree('set_icon', node, this.icons.open);

        return this;
    }

    /**
     * close 아이콘 설정 함수
     *
     * @param {object} node - 노드 객체
     * @returns {object} - 생성된 인스턴스
     */
    setCloseIcon(node = {}){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        this.$container.jstree('set_icon', node, this.icons.close);

        return this;
    }

    /**
     * none 아이콘 설정 함수
     *
     * @param {object} node - 노드 객체
     * @returns {object} - 생성된 인스턴스
     */
    setNoneIcon(node = {}){

        if (!_.isPlainObject(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'type');
            return this;
        }

        if (_.isEmpty(node)){
            CONSOLE.prop.throw(BxTree, 'node', 'value');
            return this;
        }

        this.$container.jstree('set_icon', node, this.icons.none);

        return this;
    }

    /**
     * node 가 가진 데이터를 반환한다.
     *
     * @param {string} [id=rootId] - 노드 아이디
     * @param {boolean} [isFlat=false] - flat 데이터 유무
     * @returns {object} - 노드 데이터
     */
    getJson(id = '', isFlat = false){

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxTree, 'id', 'type');
            return this;
        }

        if (!_.isBoolean(isFlat)){
            CONSOLE.prop.throw(BxTree, 'isFlat', 'type');
            return this;
        }

        if (_.isEmpty(this[OPTS].core.data)){
            return [];
        }
        else{
            let node = this.$container.jstree('get_node', '[id="' + id + '"]');

            return _parsing.call(this, node, isFlat);
        }
    }

    /**
     * BxTree 데이터 리셋 함수
     *
     * @param data - 데이터 객체
     * @returns {object} - 생성된 인스턴스
     */
    reset(data = {}) {

        if (!_.isPlainObject(data)){
            CONSOLE.prop.throw(BxTree, 'data', 'all');
        }

        let stateOptions = this.stateOptions;

        if (stateOptions) STORAGE.localStorage.remove(stateOptions.key + '_state');

        let opts = this[OPTS];

        // enableOpenAll 속성 및 데이터가 있는 경우에만 모두 닫기 후 데이터 변경.
        if (!this.enableOpenAll){
            /*
             * close_all
             * param #1 : tree data
             * param #2 : animation duration
             * */
            this.$container.jstree('close_all', opts.core.data, 0);
        }

        opts.core.data = data;

        if (!_.isEmpty(data)){
            _initNodeData.call(this, opts.core.data);
        }

        this.$container.jstree(true).settings.core.data = opts.core.data;
        this.selectedNode = {};

        _refresh.call(this);

        return this;
    }

    /**
     * BxTree 관계선 설정 함수
     *
     * @param {boolean} [isLine = true]- 관계선 셋팅 유무
     * @returns {object} - 생성된 인스턴스
     */
    setLine(isLine = true){

        if (!_.isBoolean(isLine)){
            CONSOLE.prop.throw(BxTree, 'isLine', 'type');
        }
        _initSettingLine.call(this, isLine);

        return this;
    }
}

//1. bx-tree render쪽에 이벤트가 위치할 시, 트리 갯수만큼 동일 이벤트가 생기는 이슈로 인해, document 영역에서 접근하도록 함
//2. 각 tree instance를 제어하기 위해 document의 event를 부여
//3. document의 event에서 이벤트 실행 시, 각 tree instance의 callback 함수를 호출
(function addDNDEventListeners() {

    const event = {
        start: 'dnd_start.vakata.jstree',
        move: 'dnd_move.vakata.jstree',
        stop: 'dnd_stop.vakata.jstree',
    };

    $(document).on(event.start, (e, data) => {

        const $container = $(data.element).parents(`.${TREE_CLASS.bxUiTree}`);
        const self = $container.get(0)[THIS];

        const current = _getCurrentDataModel(data);
        const isRoot = _isEqualRootAndDataModel(self, current);

        if (isRoot) return;

        const parent = _getParentDataModel(self, current.node);
        const dataModel = {
            current, parent
        };

        e.target = current.element;

        self.onDragStart.apply(self, [e, dataModel]);
    });

    $(document).on(event.move, (e, data) => {

        const $container = $(data.element).parents(`.${TREE_CLASS.bxUiTree}`);
        const self = $container.get(0)[THIS];

        const current = _getCurrentDataModel(data);
        const isRoot = _isEqualRootAndDataModel(self, current);

        if (isRoot) {
            _hideDndArea();
            return;
        }
        const parent = _getParentDataModel(self, current.node);

        const dataModel = {
            current, parent
        };

        e.target = current.element;

        self.onDrag.apply(self, [e, dataModel]);
    });

    $(document).on(event.stop, (e, data) => {

        const $container = $(data.element).parents(`.${TREE_CLASS.bxUiTree}`);
        const self = $container.get(0)[THIS];

        const current = _getCurrentDataModel(data);
        const isRoot = _isEqualRootAndDataModel(self, current);

        if (isRoot) return;
        const event = data.event;
        const { $refElement, isSiblingWithRef } = event;
        const $targetLi = $refElement.parent();
        const targetId = $targetLi.attr('id');
        const targetNode = self.getNode(targetId);
        const isSiblingRoot = self.isRootNode(targetNode);
        let parent;

        const initApply = () => {
            setTimeout(()=>{
                self.apply = true;
            });
        };

        const reject = () => {
            self.apply = false;
            initApply();
        };

        if (isSiblingRoot && isSiblingWithRef) {
            reject();
            return;
        }

        if (isSiblingWithRef) {
            parent = _getParentDataModel(self, targetId);
        }
        else {
            parent = _getDefaultDataModel();
            parent.node = targetId;
            parent.element = $targetLi.get(0);

            _setOpenState(self, parent);
        }

        const dataModel = {
            current, parent
        };

        e.target = current.element;

        self.onDragStop.apply(self, [e, dataModel, reject]);
    });
})();

/**
 * 현재 데이터 모델 반환 객체
 * @param data - jsTree vendor 가공 후 dnd 이벤트에 할당한 data객체
 * @private
 */
function _getCurrentDataModel(data) {
    const currentDataModel = _getDefaultDataModel();
    currentDataModel.node = data.data.nodes[0];
    currentDataModel.element = data.data.obj[0];

    return currentDataModel;
}

/**
 * 데이터 모델의 초기 값 전달 함수
 * @returns {{node: null, element: null}}
 * @private
 */
function _getDefaultDataModel() {
    const dataModel = {
        node: null,
        element: null
    };

    return dataModel;
}

/**
 * 데이터 모델이 root 와 같은지 여부 반환 함수
 * @param self - jsTree instance
 * @param dataModel - 데이터 모델
 * @returns {boolean}
 * @private
 */
function _isEqualRootAndDataModel(self, dataModel) {
    return self.isRootNode(self.getNode(dataModel.node));
}

/**
 * 부모의 데이터 모델 반환 함수
 * @param self - jsTree Instance
 * @param childId - 자식 ID
 * @returns {{}} dataModel - 데이터 모델
 * @private
 */
function _getParentDataModel(self, childId){

    const parentDataModel = _getDefaultDataModel();

    parentDataModel.node = self.getNode(childId).parent;
    parentDataModel.element = self.$container.find(`#${childId}`).get(0);

    return parentDataModel;
}

/**
 * dnd 기능 활성화로 인해 동적으로 생성되는 dnd 및 화살표 영역에 대해 제거하는 함수
 * @private
 */
function _hideDndArea(){
    const dndArea = 'vakata-dnd';
    const markerArea = 'jstree-marker';

    $(`#${dndArea}`).remove();
    $(`#${markerArea}`).remove();
}

/**
 * dnd 이후 기존 부모의 state 값 및 아이콘 적용함수
 * @param self - jsTree instance
 * @param dataModel - dataModel
 * @private
 */
function _setOpenState(self, dataModel){

    const data = self.$container.jstree()._model.data[dataModel.node];
    const children = data.children;
    const opened = data.state.opened;

    if (children.length === 0 || !opened) {
        data.state.opened = false;
        setTimeout(()=>{
            self.setCloseIcon(self.getNode(dataModel.node));
        });
    }
}

/**
 * CheckCallback 함수 반환 함수
 * @param self - bxTree context
 * @returns {function} - check_callback 함수
 * @private
 */
function _getCheckCallbackFunction(self) {

    // Operator about check_callback
    const OPERATORS = {
        "create": "create_node",
        "rename": "rename_node",
        "delete": "delete_node",
        "move": "move_node",
        "copy": "copy_node",
        "edit": "edit"
    };

    return (operation, node, nodeParent, nodePosition, more) => {
        // console.log(operation);
        let apply;
        switch (operation) {
            case OPERATORS.move:
                apply = self.apply;
                break;
            default:
                apply = true;
                break;
        }

        return apply;
    };
}

/**
 * 트리 노드 초기화 함수
 *
 * @param {object} [data = {}] 아이템
 * @private
 * @memberOf BxTree
 */
function _initNodeData(data = {}){

    _.cloneDeepWith(data, (nodeData = {}) =>{

        if (_.isPlainObject(nodeData)){

            let colModel = this.colModel;
            let length = colModel.length;
            let isMatchColName = false;

            // colModel 이 존재할 경우(colModel 추가후 조건)
            if (length){

                for (let i = 0; i < length; i++){

                    let model = colModel[i];

                    for (let n in nodeData){

                        if (model.name === n){

                            if (model.type === 'text'){

                                nodeData[n] = _getNodeHTML(nodeData[n], nodeData.icon);
                                nodeData.icon = _getExtendIcon.call(this, nodeData);
                            }

                            nodeData[model.type] = nodeData[n];

                            // colModel 에 정의한 이름이 기본 colName(id, text) 가 아닐 경우에만 기존 key 값을 삭제한다.
                            if (this.defaultColNames.indexOf(n) === -1) delete nodeData[n];

                            isMatchColName = true;
                        }
                    }

                    // 각 item key 가 colModel 에 정의한 columnName 과 일치하지 않을 경우
                    if (!isMatchColName) CONSOLE.prop.throw(BxTree, 'colModel 에 대한, item.key', 'type');
                }
            }
            else{

                // 기존과 같이, colModel 이 존재하지 않으며 {id, text} key 를 통해, 아이템이 할당된 경우
                nodeData.text = _getNodeHTML(nodeData.text, nodeData.icon);
                nodeData.icon = _getExtendIcon.call(this, nodeData);
            }
        }
    });
}

/**
 * 노드 내부 HTML 반환 함수
 *
 * @param {string} text
 * @param {string} icon class name
 * @returns {string}
 * @private
 * @memberOf BxTree
 */
function _getNodeHTML(text = '', iconClass = ''){

    if (!_.isString(text) || !_.trim(text).length) return '';

    let html = [];

    if (iconClass) html.push(_getNodeIcon(iconClass));

    html.push(_getTextNode(text));

    return html.join('');

}

/**
 * 노드 내부 텍스트 반환 함수
 *
 * @param {string} text
 * @returns {string}
 * @private
 * @memberOf BxTree
 */
function _getTextNode(text = ''){

    let html = [];

    html.push('<span class="text-node">');
    html.push(text);
    html.push('</span>');

    return html.join('');
}

/**
 * 노드 내부 아이콘 반환 함수
 *
 * @param {string} icon class name
 * @returns {string}
 * @private
 * @memberOf BxTree
 */
function _getNodeIcon(iconCls){
    return '<i class="' + iconCls + '"></i>';
}

/**
 * 노드 확장 아이콘 반환 함수
 * @param {object} node 객체
 * @returns {object} icon 객체
 * @private
 * @memberOf BxTree
 */
function _getExtendIcon(node = {}){

    let icon = '';
    let colModelNames = _getColModelNames.call(this);
    let stateOptions = this.stateOptions;

    // node.text 가 없는 경우에는, 해당 아이콘을 노출하지 않는다.
    // 정의된 colModel 을 통해, text key 에 대한, columnName 을 가져온다.
    if (!_.isEmpty(colModelNames)){
        if (!_.trim(node[colModelNames.text]).length) return '';
    }
    else{
        // columnName 이 text 인 경우
        if (!_.trim(node.text).length) return '';
    }

    if (_.isEmpty(node.children)){
        icon = this.icons.none;
        return icon;
    }

    //state 있는 경우에는 기본 아이콘을 close로 할당한다.

    if (stateOptions && stateOptions.fold && _getItems(stateOptions.key + '_state')){
        icon = this.icons.close;
    }
    else{
        if (this.enableOpenAll){

            icon = this.icons.open;

        }
        else{
            // enableOpenAll 프로퍼티가 false 인 경우에는 아이콘 open/close 여부를 가려 할당한다.
            icon = this.icons.close;
        }
    }

    return icon;

}

/**
 * 노드 html에서 사용자 아이콘의 class 반환
 * @param {string} node html
 * @returns {string} icon class
 * @private
 * @memberOf BxTree
 */
function _getIconClassFromNodeHTML(nodeHtml = ''){

    let target = $(nodeHtml).get(0);

    if (target.tagName !== 'I') return;

    return $(target).attr('class');
}

function _getColModelNames(){

    let o = {};

    _.map(this.colModel, function (model){
        o[model.type] = model.name;
    });

    return o;
}

/**
 * refresh 함수
 * @returns {BxTree.refresh}
 * @private
 * @memberOf BxTree
 */
function _refresh(){
    this.$container.jstree(true).refresh(true);

    return this;
}

// /**
//  * node 전체 삭제 함수
//  * @returns {BxTree.clearData}
//  * @private
//  */
// function _clearData(){
//     this[OPTS].core.data = {};
//     this.$container.jstree(true).settings.core.data = this[OPTS].core.data;
//     _refresh.call(this);
//
//     return this;
// }
/**
 * state session set 함수
 *
 * @private
 * @memberOf BxTree
 */
function _setSession(){

    let stateOptions = this.stateOptions;

    if (stateOptions.expire){
        let key = stateOptions.key;
        let session = STORAGE.sessionStorage.get(key);

        if (!session){
            STORAGE.localStorage.remove(key + '_state');
            STORAGE.sessionStorage.set(stateOptions.key, 'sessionOn');
        }
    }
}
/**
 * 현재 core data get 함수
 *
 * @returns {object} all data state
 * @private
 * @memberOf BxTree
 */
function _getAllNode(){

    let node = this.$container.jstree('get_json', '#')[0];

    let data  = _createCoreData.call(this, node);

    return data;

}
/**
 * state option set 함수
 *
 * @private
 * @memberOf BxTree
 */
function _setOptionState(){

    let state = _getState.call(this);

    _setItems(state, this.stateOptions.key + '_state');

}
/**
 * state option get 함수
 *
 * @returns {object} option state
 * @private
 * @memberOf BxTree
 */
function _getOptionState(){

    let stateOptions = this.stateOptions;
    let state = _getItems(stateOptions.key + '_state');

    if (stateOptions.fold){
        _initFoldState.call(this, state);
    }
    if (stateOptions.checkbox){
        _initCheckboxState.call(this, state);
    }
    if (stateOptions.selected){
        _initSelectedState.call(this, state);
    }
    if (stateOptions.disabled){
        _initDisabledState.call(this, state);
    }

    return state;
}
/**
 * 노드의 fold state 설정 함수
 *
 * @param {object} [state = {}] - 상태 저장 내역
 * @private
 * @memberOf BxTree
 */
function _initFoldState(state = {}){

    if (!_.isPlainObject(state)) CONSOLE.prop.throw(BxTree, 'state', 'type');

    let openNodes = state.open;
    if (openNodes.length){
        for (let i in openNodes){

            let node = this.$container.jstree('get_node', openNodes[i]);

            if (node) this.open(node);
        }
    }
}
/**
 * 노드의 checkbox state 설정 함수
 *
 * @param {object} [state = {}] - 상태 저장 내역
 * @private
 * @memberOf BxTree
 */
function _initCheckboxState(state = {}){

    if (!_.isPlainObject(state)) CONSOLE.prop.throw(BxTree, 'state', 'type');

    this.checkNodes(state.checked);
}
/**
 * 노드의 selected state 설정 함수
 *
 * @param {object} [state = {}] - 상태 저장 내역
 * @private
 * @memberOf BxTree
 */
function _initSelectedState(state = {}){

    if (!_.isPlainObject(state)) CONSOLE.prop.throw(BxTree, 'state', 'type');

    let node = this.$container.jstree('get_node', state.selected);

    this.selectedNode = node;
    this.$container.jstree('select_node', node, false, true);

}
/**
 * 노드의 disabled state 설정 함수
 *
 * @param {object} [state = {}] - 상태 저장 내역
 * @private
 * @memberOf BxTree
 */
function _initDisabledState(state = {}){

    if (!_.isPlainObject(state)) CONSOLE.prop.throw(BxTree, 'state', 'type');

    this.setDisableNodes(state.disabled);
}
/**
 * state get 함수
 *
 * @returns {object} state 값
 * @private
 * @memberOf BxTree
 */
function _getState(){

    let state = {
        'open': [],
        'checked': [],
        'selected': [],
        'disabled': []
    };

    let data = this.$container.jstree()._model.data;

    for (let i in data){
        if (data.hasOwnProperty(i)){
            if (i !== $.jstree.root){
                if (data[i].state.opened){
                    state.open.push(i);
                }
                if (data[i].state.checked){
                    state.checked.push(i);
                }
                if (data[i].state.selected){
                    state.selected.push(i);
                }
                if (data[i].state.disabled){
                    state.disabled.push(i);
                }
            }
        }
    }

    return state;

}
/**
 * localStorage에 items set 함수
 *
 * @param {object} data - jstree instance
 * @param {string} items key
 * @private
 * @memberOf BxTree
 */
function _setItems(data = {}, key = null){

    if (!_.isPlainObject(data)) CONSOLE.prop.throw(BxTree, 'data', 'type');

    if (!_.isString(key)) CONSOLE.prop.throw(BxTree, 'key', 'type');

    STORAGE.localStorage.set(key, JSON.stringify(data));
}

/**
 * localStorage에서 items get 함수
 *
 * @param {string} items key
 * @returns {object} data - store instance
 * @private
 * @memberOf BxTree
 */
function _getItems(key){

    if (!_.isString(key)) CONSOLE.prop.throw(BxTree, 'key', 'type');

    let storedItem = JSON.parse(STORAGE.localStorage.get(key));

    return storedItem;
}

/**
 * 노드의 초기 Open, Close 설정 함수
 *
 * @param {object} data - jsTree instance
 * @private
 * @memberOf BxTree
 */
function _initFirstOpen(data = {}){

    if (!_.isPlainObject(data)) CONSOLE.prop.throw(BxTree, 'data', 'type');

    let node = null;

    _.map(data.instance._model.data, (data, key) =>{
        if (this.isRootNode(data)){
            node = data;
        }
    });

    // 전체 node 를 open 시킨다.
    if (this.enableOpenAll){
        // 아이콘을 변경 시킨다.
        //this.setOpenIcon(node);
        this.$container.jstree('open_all', node);
    }
    else{
        this.$container.jstree('close_all', node);
    }

}

/**
 * 관계선 초기화 함수
 *
 * @private
 * @memberOf BxTree
 */
function _initSettingLine(isLine = null){

    let setLine;
    if (!_.isBoolean(isLine))  setLine = this.enableLine;
    else setLine = isLine;

    let type = this.lineType;
    let $node = this.$container.find(`.${TREE_CLASS.jstreeContainerUl}`);

    $node.removeClass(`${TREE_CLASS.lineType}`)
        .removeClass(`${TREE_CLASS.dotType}`)
        .removeClass(`${TREE_CLASS.noneType}`);

    if (!setLine){
        $node.addClass(`${TREE_CLASS.noneType}`);
    }
    else{
        if (type === 'dot'){
            $node.addClass(`${TREE_CLASS.dotType}`);
        }
        else{
            $node.addClass(`${TREE_CLASS.lineType}`);
        }
    }
}

/**
 * Node의 text의 태그를 삭제 한다.
 * @param {object} [node = {}] 아이템
 * @param {boolean} [isFlat=false] - flat 데이터 유무
 * @returns {object} - 생성된 인스턴스
 * @private
 * @memberOf BxTree
 */
function _parsing(node = {}, isFlat = false){

    if (!_.isPlainObject(node)) CONSOLE.prop.throw(BxTree, 'node', 'type');

    let obj = [];
    let nodeList = !isFlat ? [UTIL.toJSON(JSON.stringify(node))] : this.$container.jstree('get_json', node, {flat: isFlat});

    _.map(nodeList, data =>{

        obj.push({
            'id': data.id,
            'text': $(data.text).filter('span.text-node').html(),
            'icon': data.icon,
            'parent': data.parent,
            'children': data.children
        });
    });

    return !isFlat ? obj[0] : obj;
}

/**
 * CheckBox의 기능을 초기화 한다
 *
 * @param {boolean} [enableCheckRelated=false] - 상위 노드까지 체크의 유무
 * @param {boolean} [enableCheckLabel=false] - 라벨을 클릭 했을때 체크의 유무
 * @private
 * @memberOf BxTree
 */
function _initCheckBox(enableCheckRelated = false, enableCheckLabel = false){

    if (!_.isBoolean(enableCheckRelated) || !_.isBoolean(enableCheckLabel)){
        CONSOLE.prop.throw(BxTree, 'checkBoxOptions', 'type');
    }

    let opts = this[OPTS];

    opts.plugins.push('checkbox');

    opts.checkbox = {
        'three_state': enableCheckRelated,
        'whole_node': enableCheckLabel,
        'tie_selection': false // 체크, 선택 기능을 동등하게 볼지 여부
    };

    // whole_node 가 true일떄 Select 기능이 작동 안하도록 jstree는 되어 있어 core 코드와 동일하게 재구성
    if (opts.checkbox['whole_node']){

        this.$container.on('click', ".jstree-anchor", (e)=>{

            if (this.$container.jstree('is_disabled', e.currentTarget)) return;

            this.$container.jstree('deselect_all', true);
            this.$container.jstree('select_node', e.currentTarget, false, true, e);
            this.$container.jstree(true)['_data'].core['last_clicked'] = this.$container.jstree('get_node', (e.currentTarget));
        });
    }
}

/**
 * Drag and Drop의 기능을 초기화 한다
 *
 * @param {boolean} [insertPosition='before'] - 하위노드가 있는 노드로의 이동 시, 상단 또는 하단에 위치 지정
 * @private
 * @memberOf BxTree
 */
function _initDnd(insertPosition = 'before'){

    if (!_.isString(insertPosition)){
        CONSOLE.prop.throw(BxTree, 'insertPosition', 'type');
    }

    // 마우스 우측 현재노드(vakata-dnd)위치 수정
    $.vakata.dnd.settings = {
        "scroll_speed" : 10,
        "scroll_proximity" : 20,
        "helper_left" : 5,
        "helper_top" : -5, //10에서 -5로 수정
        "threshold" : 20,
        "threshold_touch" : 20
    };

    let opts = this[OPTS];

    opts.plugins.push('dnd');

    if (insertPosition === 'after') insertPosition = DND_OPTIONS.after;
    else insertPosition = DND_OPTIONS.before;

    opts.dnd = {
        'copy': false,
        'open_timeout': DND_OPTIONS.openTime,
        'inside_pos': insertPosition
    };
}

/**
 * Custom icon에 이벤트를 생성한다.
 *
 * @private
 * @memberOf BxTree
 */
function _setCustomIconHandler(){

    this.$container.off('click', '.jstree-ocl'); // jsTree 안에 있는 이벤트를 제거

    this.$container.on('click', '.jstree-themeicon-custom', (e) =>{

        this.toggle(this.$container.jstree('get_node', (e.currentTarget)));

        e.preventDefault();
        e.stopPropagation();
    });
}

/**
 * String or String[] 인지 유효성 검사를 한다
 * @param {string| string[] } [ids = null] 노드의 ID
 * @private
 * @memberOf BxTree
 */

function _validationOfString(ids = null){

    if (!(_.isArray(ids) || _.isString(ids))){
        CONSOLE.prop.throw(BxTree, 'ids', 'type');
        return this;
    }

    _.map(ids, id =>{

        if (!_.isString(id)){
            CONSOLE.prop.throw(BxTree, 'ids', 'type');
            return this;
        }
    });

    return this;
}

/**
 * 선택된 노드의 자식들을 반환한다
 *
 * @param {string} [id = null] 노드의 ID
 * @param {boolean} [isSelfNode = false] 선택된 노드가 함께 반환 여부
 * @returns {array} - 자식 노드들
 * @private
 * @memberOf BxTree
 */
function _getChildrensNode(id = null, isSelfNode = false){

    let nodeList = isSelfNode ? [id] : [];

    if (this.$container.jstree('get_node', id)){
        return nodeList.concat(this.$container.jstree()._model.data[id]['children_d']);
    }
    else{
        return nodeList;
    }
}
/**
 * 선택된 노드의 부모를 반환한다
 *
 * @param {string} [id = null] 노드의 ID
 * @param {boolean} [isSelfNode = false] 선택된 노드가 함께 반환 여부
 * @returns {array} - 부모 노드들
 * @private
 * @memberOf BxTree
 */
function _getParentsNode(id = null, isSelfNode = false){

    let nodeList = isSelfNode ? [id] : [];

    if (this.$container.jstree('get_node', id)){
        return nodeList.concat(this.$container.jstree()._model.data[id]['parents']);
    }
    else{
        return nodeList;
    }
}
/**
 * 업데이트 된 core.data 를 생성해준다.
 *
 * @param {object} [data = {}] jstree의 최신 트리정보
 * @returns {object} - 변화된 data
 * @private
 * @memberOf BxTree
 */
function _createCoreData(data = {}){

    _.cloneDeepWith(data, (nodeData = {}) =>{

        if (_.isPlainObject(nodeData)){

            delete nodeData['a_attr'];
            delete nodeData['data'];
            delete nodeData['li_attr'];
            delete nodeData['state'];
            nodeData['icon'] = $(nodeData['text']).filter('i').attr('class');
            nodeData['text'] = $(nodeData['text']).filter('span.text-node').html();
        }
    });

    return data;
}
module.exports = BxTree;
