// bxuip-angular.js-ui/src/app/util.js

const $ = require('jquery');
const _ = require('lodash');
const BXUI = require('../lib/bx-ui-core/entry');

const UTIL = {
    createCustomElement: _createCustomElement,
    separateEventsAndAttrs: _separateEventsAndAttrs,
    applyAttrs: _applyAttrs,
    addEvent: _addEvent,
    isNgAttr: _isNgAttr,
    callBxRender: _callBxRender,
    hasNgRepeat: _hasNgRepeat,
    deleteNgRepeat: _deleteNgRepeat
};

/**
 * Element 반환 함수
 * @param tagName - tagName
 * @returns {string}
 * @private
 */
function _createCustomElement(tagName = '') {

    if (!_.isString(tagName)) {
        BXUI.Console.prop.throw('BxNg.Util', 'tagName', 'type');
    }

    if (_.isEmpty(tagName.trim())) {
        BXUI.Console.prop.throw('BxNg.Util', 'tagName', 'value');
    }

    return `<${tagName}></${tagName}>`;
}

/**
 * event 와 attribute 분리 함수
 * @param attributes
 * @private
 */
function _separateEventsAndAttrs($scope = null , attributes = null) {

    if(_.isNull($scope)){
        BXUI.Console.prop.throw('BxNg.Util', '$scope', 'value');
    }

    if(_.isNull(attributes)){
        BXUI.Console.prop.throw('BxNg.Util', 'attributes', 'value');
    }

    const events = {};
    const attrs = {};

    const $attr = attributes.$attr;

    _.map($attr, (v, k) => {
        const isBxFunction = k.indexOf('bx') === 0;
        let _v = attributes[k];

        if (isBxFunction) {
            const isFunctionCall = _v.search(/\(|\)/g) === -1;
            const errMsg = `${_v} 를 ${_v}() 로 변경해주세요.`;

            if (isFunctionCall) {
                BXUI.Console.prop.throw('BxNg.Util', errMsg);
            }

            _v = _v.replace(/\(\s*\S*\)$/g, '');

            events[v] = _.get($scope, _v);
        }
        else {
            attrs[v] = _v;
        }
    });
    return { events, attrs };
}

/**
 * directive 의 attribute 를 customElement 의 attribute 로 할당하는 함수
 * @param $scope
 * @param $elem
 * @param attrs
 * @private
 */
function _applyAttrs($elem = null, attrs = {}) {

    if (_.isNull($elem)) {
        BXUI.Console.prop.throw('BxNg.Util', '$elem', 'value');
    }

    if (!_.isObject($elem)) {
        BXUI.Console.prop.throw('BxNg.Util', 'attrs', 'value');
    }

    const elem = $elem.get(0);

    _.forEach(attrs, (v, k) => {
        elem.setAttribute(k, v);
    });
}

/**
 * 이벤트 할당 함수
 * @param $elem
 * @param events
 * @param customEvents
 * @private
 */
function _addEvent($elem = null, events = null, customEvents = null) {

    if(_.isNull($elem)){
        BXUI.Console.prop.throw('BxNg.Util', '$elem', 'value');
    }

    if(_.isNull(events)){
        BXUI.Console.prop.throw('BxNg.Util', 'events', 'value');
    }

    if(_.isNull(customEvents)){
        BXUI.Console.prop.throw('BxNg.Util', 'customEvents', 'value');
    }

    const customEventList = Object.getOwnPropertyNames(customEvents);
    const lowerCustomEventList = customEventList.map((v) => v.toLowerCase());

    _.forEach(events, (v, k) => {
        const idx = lowerCustomEventList.indexOf(k);
        const isProperty = idx !== -1;

        if (isProperty) {
            const prop = customEventList[idx].replace(/-/,'');
            $elem.get(0)[prop] = v;
        }
        else {
            const eventName = k.replace('bx-','').toLowerCase();
            $elem.on(eventName, function (e){
                v && v.call(this, e);
            });
        }
    });
}

/**
 * ng 지시자 여부 반환 함수
 * @param {string} attrName - attribute 명
 * @returns {boolean}
 * @private
 */
function _isNgAttr(attrName = ''){

    if(!_.isString(attrName)) {
        BXUI.Console.prop.throw('BxNg.Util', 'attrName', 'value');
    }

    return attrName.split('-')[0] === 'ng';
}

/**
 * bxRender 호출 함수
 * @param elem - target Element
 * @param events - event 객체
 * @private
 */
function _callBxRender(elem = null, events = {}){
    if (!_.isElement(elem)) {
        BXUI.Console.prop.throw('BxNg.Util', 'elem', 'type');
    }

    if (!_.isPlainObject(events)) {
        BXUI.Console.prop.throw('BxNg.Util', 'events', 'type');
    }

    if (!_hasBxRender(events)) return;

    elem.bxrender = events[_getBxRenderName()];
    elem.bxrender();
}

/**
 * bxRender 함수 존재 유무 반환 함수
 * @param events
 * @returns {boolean}
 * @private
 */
function _hasBxRender(events = {}) {

    if (!_.isPlainObject(events)) {
        BXUI.Console.prop.throw('BxNg.Util', 'events', 'type');
    }

    const renderFunc = _getBxRenderName();
    return events.hasOwnProperty(renderFunc) && events[renderFunc];
}

/**
 * bx-render 함수 key 값 반환 함수
 * @returns {string}
 * @private
 */
function _getBxRenderName() {
    return 'bx-render';
}

/**
 * bx-render 함수 반환 함수
 * @param events
 * @private
 */
function _getBxRender(events = {}) {

    if (!_.isPlainObject(events)) {
        BXUI.Console.prop.throw('BxNg.Util', 'events', 'type');
    }

    const renderFunc = _getBxRenderName();
    return events[renderFunc];
}

/**
 * ngRepeat 유무 반환 함수
 * @param attrs
 * @returns {boolean|*}
 * @private
 */
function _hasNgRepeat(attrs = {}) {

    const ngRepeat = _getNgRepeatName();
    return attrs.hasOwnProperty(ngRepeat) && attrs[ngRepeat];
}

function _getNgRepeatName(){

    return 'ng-repeat';
}

/**
 * ngRepeat 삭제 함수
 * @param attrs
 * @private
 */
function _deleteNgRepeat(attrs = {}) {

    delete attrs[_getNgRepeatName()];
}

module.exports = UTIL;
