// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/base.js
/**
 * @class Base
 * @classDesc input Base 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');
const Mousetrap = require('mousetrap');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

class Base {

    constructor(){

        this.keyCode = {
            "37": 'LEFT',
            "38": 'TOP',
            "39": 'RIGHT',
            "40": 'BOTTOM',
            "8": 'BACKSPACE'
        };

        this.altText = {
            "increase": '증가',
            "decrease": '감소',
            "select": '선택'
        };

        this.inputSelector = {
            "blind": 'blind',
            "bxActionIcon": 'bx-action-icon',
            "bxHinting": 'bx-hinting',
            "label": 'bx-label',
            "bxTimepicker": '-bx-ui-time-picker',
            "bxInput": 'bx-input',
            "bxInputWrap": 'bx-input-wrap',
            "bxSelectWrap": 'bx-select-wrap',
            "placeholder": 'placeholder'
        };

        // 모든 input 컴포넌트가 가지는 기본 옵션
        this.opts = {
            //"attr": {},
            "label": '',
            "inactive": false,
            "bx-render": function (){}
        };
    }

    /**
     * input wrapper html 반환 함수
     *
     * @param {object} component opts
     *
     * @return {string} input wrap html string
     *
     * @private
     * @memberOf BxInput
     */
    getWrapperElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(Base, 'opts', 'type');
        }

        let classNames = [];

        let type = opts.attr.type;
        let shape = opts.shape;

        classNames.push(`bx-${_getInputClassTypeName(type)}-wrap`);

        if (shape) classNames.push(shape);

        let $wrapper = $('<div>');

        $wrapper.attr('class', classNames.join(' '));

        return $wrapper;
    }
    /**
     * icon 클래스명을 반환한다.
     *
     * @param {object} component opts
     *
     * @return {string} classNames
     *
     * @private
     *
     * @memberOf BxInput
     */
    getIconClassNames(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(Base, 'opts', 'type');
        }

        let label = opts.label;
        let icons = opts.icons;

        let left = icons.left;
        let right = icons.right;

        let classNames = [];

        if (!label && left && !right) classNames.push('add-left');

        if (right || left) classNames.push('add-icon');

        if (right && ((left && label || !left))) classNames.push('add-right');

        if (!label && left && right) classNames.push('add-all');

        return classNames.join(' ');
    }
    /**
     * input 아이콘을 생성하는 함수
     *
     * @param {string} 아이콘 클래스 명
     * @param {string} 대체 텍스트
     *
     * @return {string} 생성된 icon html
     *
     * @private
     * @memberOf BxInput
     */
    createIconElement(iconClass = '', altText = ''){

        if (!_.isString(iconClass)){
            CONSOLE.prop.throw(Base, 'iconClass', 'type');
        }

        if (!_.isString(altText)){
            CONSOLE.prop.throw(Base, 'altText', 'type');
        }

        let html = [];

        html.push(`<i class="${iconClass || ''}">`);
        html.push(`<span class="${this.inputSelector.blind}">${altText || ''}</span>`);
        html.push('</i>');

        return html.join('');
    }

    /**
     * validator 세팅 함수
     *
     * @param {string} 유효성 검사 옵션 객체
     * @param {object} 대상 jquery 엘리먼트 객체
     *
     * @private
     * @memberOf BxInput
     */
    setInputValidator($elem = null, validateOption = null){

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isPlainObject(validateOption)){
            CONSOLE.prop.throw(Base, 'validateOption', 'type');
        }

        let events = validateOption.events;
        let validator = validateOption.validator;

        if (!_.isArray(events)){
            CONSOLE.prop.throw(Base, 'validateOption.events', 'type');
        }

        if (!_.isFunction(validator)){
            CONSOLE.prop.throw(Base, 'validateOption.validator', 'type');
        }

        // 바인딩되어있는 이벤트를 삭제한다.
        $elem.off('.bxValidator');

        events.forEach(eventName => {

            // event namespace 를 통해, 이벤트를 바인딩한다.
            $elem.on(eventName + '.bxValidator', 'input', e => {

                let $target = $(e.currentTarget);

                let result = validator($target.val());
                let $wrap = $target.parent();

                $wrap.removeClass('confirm error')
                .addClass(result ? 'confirm' : 'error');
            });
        });

        return this;
    }
    /**
     * inactive 설정 함수
     *
     * @param {object} element
     * @param {boolean} disabled
     *
     * @return {object} component opts
     *
     * @private
     * @memberOf BxInput
     */
    setInactive($elem = null, inactive = null){

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isBoolean(inactive)){
            CONSOLE.prop.throw(Base, 'inactive', 'type');
        }
        
        let $input = $elem.find('input');
        let $select = $elem.find(`.${this.inputSelector.bxSelectWrap}`);

        let $inputWrapper = null;

        let isMultipleInput = $elem.get(0).tagName.toLowerCase() !== 'input';

        if (isMultipleInput){

            $inputWrapper = $elem;

            _.forEach($input, v => {
                $(v).prop('disabled', inactive);
            });

            _.forEach($select, v => {
                $(v).prop('disabled', inactive);
            });
        }
        else{

            $inputWrapper = $elem.parents(`.${this.inputSelector.bxInputWrap}`);

            $elem.prop('disabled', inactive);
        }

        if (inactive) $inputWrapper.addClass('disabled');
        else $inputWrapper.removeClass('disabled');

        $inputWrapper.removeClass('confirm error');

        return this;
    }
    /**
     * Container에 hinting 영역만큼의 padding을 설정하는 함수
     *
     * @param {object} container jquery element
     * @param {object} component opts
     *
     * @private
     * @memberOf BxInput
     */
    setContainerPadding($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(Base, 'opts', 'type');
        }

        let padding = 15;

        if (!_.isEmpty(opts.hinting)){

            if (opts.shape === 'line'){
                $elem.css('padding-bottom', padding);
            }
            else{
                $elem.css('padding-top', padding);
            }
        }

        return this;
    }
    /**
     * component option 반환 함수
     *
     * @param {object} custom jquery element
     * @param {object} 기본값 옵션 객체
     *
     * @return {object} component opts
     *
     * @private
     * @memberOf BxInput
     */
    createComponentOptions(
        $elem = null,
        opts = {},
        attrNames = {},
        withOptionKeyCallBack = function (){},
        withoutOptionKeyCallBack = function (){}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) && !_.isPlainObject($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(Base, 'opts', 'type');
        }

        if (!_.isPlainObject(attrNames)){
            CONSOLE.prop.throw(Base, 'attrNames', 'type');
        }

        if (!_.isFunction(withOptionKeyCallBack)){
            CONSOLE.prop.throw(Base, 'withOptionKeyCallBack', 'type');
        }

        if (!_.isFunction(withoutOptionKeyCallBack)){
            CONSOLE.prop.throw(Base, 'withoutOptionKeyCallBack', 'type');
        }

        let _opts = UTIL.createComponentOptions($elem, opts, attrNames, withOptionKeyCallBack, withoutOptionKeyCallBack);

        return _setMandatoryOPTS.call(this, _opts);
    }
    setElementProperties($elem = null, opts = {}, readOnlyAttrs = [], callback = () => {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) && !_.isPlainObject($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(Base, 'opts', 'type');
        }

        if (!_.isArray(readOnlyAttrs)){
            CONSOLE.prop.throw(Base, 'readOnlyAttrs', 'type');
        }

        if (!_.isFunction(callback)){
            CONSOLE.prop.throw(Base, 'callback', 'type');
        }

        let elem = $elem.get(0);

        _.map(opts, (v, k) =>{

            // attr skip
            if (k === 'attr') return;
            // events skip
            if (k === 'events') return;

            if (k.match(/^bx-/)){

                let _k = k.replace('-', '');

                delete opts[k];

                k = _k;
            }

            // 읽기 전용 속성
            if (_.indexOf(readOnlyAttrs, k) > -1){
                UTIL.setReadOnlyProperty(elem, k, v);
                return;
            }

            let _k = UTIL.convertInternalPropertyName(k);

            elem[_k] = v;

            let descriptor = callback(elem, k, v);

            UTIL.setPropertyWithDescriptor(elem, k, descriptor, v);
        });

        return this;
    }
    /**
     * value 배열 초기화
     *
     * @param {string} 초기화할 문자열
     * @param {string} dataType
     * @param {string} detailType
     *
     * @returns {array} 초기화된 배열
     *
     * @private
     * @memberOf bxInput
     */
    splitValue(value = '', inputInfos = [], separateValue = ''){

        if (!_.isString(value)){
            CONSOLE.prop.throw(Base, 'value', 'type');
        }

        if (!_.isArray(inputInfos)){
            CONSOLE.prop.throw(Base, 'inputInfos', 'type');
        }

        if (!_.isString(separateValue)){
            CONSOLE.prop.throw(Base, 'separateValue', 'type');
        }

        let ret = [];

        let values = value.split(separateValue);

        _.forEach(inputInfos, (v, k) => {

            let value = values[k] || '';
            let digit = v.digit;
            let deletePtn = v.deletePtn;

            if (!_.isUndefined(deletePtn)){
                value = value.replace(deletePtn, '');
            }

            if (!_.isUndefined(digit)){
                value = value.substr(0, digit);
            }

            ret.push(value);
        });


        return ret;
    }
    /**
     * formatter 세팅 함수
     *
     * @param {string} 포맷 옵션 객체
     * @param {object} 대상 jquery 엘리먼트 객체
     *
     * @private
     * @memberOf BxInput
     */
    setInputFormatter($elem = null, formatter = null){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }
            
        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isPlainObject(formatter) && !_.isString(formatter)){
            CONSOLE.prop.throw(Base, 'formatter', 'type');
        }

        let type = '';
        let options = null;

        let $input = $elem.find('input');

        let ptns = {
            "date": /[^yMdhms.]/g,
            "notNumeric": /\D/g,
            "notNumericDot": /[^0-9\.]/g
        };
        
        // FORMATTER TYPE
        let formatType = {
            "currency": 'currency',
            "date": 'date',
            "pid": 'pid',
            "bzno": 'bzno',
            "tel": 'tel',
            "card": 'card',
            "exp": 'exp',
            "pct": 'pct',
            "zip": 'zip',
            "custom": 'custom'
        };        

        //object, string 모두 지원, 초기화
        if (_.isPlainObject(formatter)){

            type = formatter.type;
            options = formatter.options || {};
        }
        else if (_.isString(formatter)){

            type = formatter;
            options = {};

            //string 시, 필수 default 지정
            if (type === 'date'){
                options = {'viewFormat': 'yyyy-MM-dd'};
            }
            else if (type === 'currency'){
                options = {'decimalPlaces': 0};
            }
        }

        // 전달받은 옵션에 대한, 유효성 검사를 진행한다.
        validateOptions();

        // 각 타입에 바인딩된 이벤트 핸들러
        let focusInEventHandler = function(){};
        let focusOutEventHandler = function(){};
        let keyUpAndDownEventHandler = function(){};
        let keyUpAndDownFocusOutEventHandler = function(){};

        //event 설정
        if (type === formatType.custom){

            focusInEventHandler = e => {

                $input.val(UTIL.getUnFormatString(type, $input.val(), options));
            };

            focusOutEventHandler = e => {

                let originValue = $input.val();
                let ret = originValue.replace(options.reg, options.pattern);

                $input.val(ret);
                $input.prop('originValue', originValue);
                $input.prop('value', ret);
            };
        }
        else if (type === formatType.date){

            options.dataFormat = options.viewFormat.replace(ptns.date, '');

            focusInEventHandler = e =>{
                $input.val(UTIL.getUnFormatString(type, $input.val(), options));
            };

            focusOutEventHandler = e =>{

                let originValue = $input.val().replace(ptns.notNumeric, '');
                let ret = UTIL.getFormatString(type, originValue, options);

                $input.val(ret);
                $input.prop('originValue', originValue);
                $input.prop('value', ret);

            };

            keyUpAndDownEventHandler = e =>{

                if (!_.isUndefined(this.keyCode[e.keyCode])) return;

                $input.val($input.val().replace(ptns.notNumeric, ''));

            };
        }
        else if (type === formatType.currency){

            keyUpAndDownFocusOutEventHandler = e =>{

                let originValue = $input.val().replace(ptns.notNumericDot, '');

                let {integer, decimal} = _getValidDecimal(originValue);

                originValue = integer + decimal;

                let ret = UTIL.getFormatString(formatType.currency, originValue, options);

                if (!_.isUndefined(this.keyCode[e.keyCode])){

                    if (this.keyCode[e.keyCode] === 'BACKSPACE') {
                        $input.prop('originValue', originValue);
                    }

                }

                $input.val(ret);
                $input.prop('originValue', originValue);
                $input.prop('value', ret);

            };
        }
        else{

            // 나머지 타입들(pid, bzno, tel, card, exp, pct, zip)
            keyUpAndDownFocusOutEventHandler = e => {

                let value = $input.val();
                let originValue = value.replace(ptns.notNumeric, '');

                if (type === 'pct'){

                    originValue = value.replace(ptns.notNumericDot, '');

                    let {integer, decimal} = _getValidDecimal(originValue);

                    originValue = integer + decimal;
                }

                let ret = UTIL.getFormatString(type, originValue, options);

                if (type === 'card') ret = ret.value;

                if (!_.isUndefined(this.keyCode[e.keyCode])){

                    if (this.keyCode[e.keyCode] === 'BACKSPACE') {
                        $input.prop('originValue', originValue);
                    }

                    return;
                }

                $input.val(ret);
                $input.prop('originValue', originValue);
                $input.prop('value', ret);
            };
        }

        $elem.on('focusin', focusInEventHandler);
        $elem.on('focusout', focusOutEventHandler);
        $elem.on('keyup keydown', keyUpAndDownEventHandler);
        $elem.on('keyup keydown focusout', keyUpAndDownFocusOutEventHandler);



        // 전달받은 옵션에 대한 유효성 검사 함수
        function validateOptions(){

            if (!_.isString(type) || _.isUndefined(formatType[type])){
                CONSOLE.prop.throw(Base, 'formatter type', 'type');
            }

            if (type === formatType.currency && !_.isUndefined(options.decimalPlaces)){

                if (!_.isNumber(options.decimalPlaces)){
                    CONSOLE.prop.throw(Base, 'decimalPlaces', 'type');
                }

                if (options.decimalPlaces < 0){
                    CONSOLE.prop.throw(Base, 'decimalPlaces 값은 양수만 입력 가능합니다');
                }

                if (_.toString(options.decimalPlaces).indexOf('.') > -1){
                    CONSOLE.prop.throw(Base, 'decimalPlaces 값은 정수만 입력 가능합니다');
                }
            }
            else if (type === formatType.date &&
                (!_.isString(options.viewFormat) || options.viewFormat.length === 0)){

                CONSOLE.prop.throw(Base, 'viewFormat 값을 입력해주세요.');
            }
            else if (type === formatType.custom &&
                (_.isUndefined(options.reg) || _.isUndefined(options.pattern))){

                CONSOLE.prop.throw(Base, 'custom 타입은 object 형식으로 입력해야하며, reg 프로퍼티와 pattern 프로퍼티가 반드시 필요합니다.');
            }
        }

        return this;
    }

    /**
     * input focus 될 시 border 효과 handler
     *
     * @private
     * @memberOf BxInput
     */
    setInputFocusHandler($elem = null){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        $elem.on('focus', `input, .${this.inputSelector.bxSelectWrap}`, e =>{
            $(e.delegateTarget).children('div').addClass('focus');
        });

        return this;
    }

    /**
     * input blur 될 시 focus(border) 효과 삭제 handler
     *
     * @private
     * @memberOf BxInput
     */
    setInputBlurHandler($elem = null){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        $elem.on('blur', `input, .${this.inputSelector.bxSelectWrap}`, e =>{
            $(e.delegateTarget).children('div').removeClass('focus');
        });

        return this;
    }

    /**
     * input type 의 right icon handler 설정 함수
     *
     * @private
     * @memberOf BxInput
     */
    setInputIconHandler($elem = null, callback = function(){}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction(callback)){
            CONSOLE.prop.throw(Base, 'callback', 'type');
        }

        let input = $elem.find('input').get(0);

        $elem.on('click', 'i.' + this.inputSelector.bxActionIcon, e => {
            callback.call(input, e);
        });

    }
    getValues($elem = null, separateValue = ''){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isString(separateValue)){
            CONSOLE.prop.throw(Base, 'separateValue', 'type');
        }

        let ret = [];

        // 값이 채워졌는지에대한 유/무
        let isFull = false;

        let $inputs = $elem.find('input');

        _.forEach($inputs, (v, k) => {

            let value = $(v).val();

            if (value) isFull = true;

            ret.push(value);
        });

        ret = isFull ? ret.join(separateValue) : '';

        return ret;

    }
    setValues($elem = null, values = null){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isArray(values)){
            CONSOLE.prop.throw(Base, 'values', 'type');
        }

        let $inputs = $elem.find('input');

        _.forEach($inputs, (v, k) => {

            let $input = $(v);
            let value = values[k] || '';

            $input.val(value);
        });

        if (UTIL.isIE9()){
            $inputs.removeClass(this.inputSelector.placeholder);
        }

        return this;
    }
    getPlaceHolders($elem = null, separateValue = ''){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isString(separateValue)){
            CONSOLE.prop.throw(Base, 'separateValue', 'type');
        }

        let ret = [];

        // 값이 채워졌는지에대한 유/무
        let isFull = false;

        let $inputs = $elem.find('input');

        _.forEach($inputs, (v, k) => {

            let placeholder = $(v).attr('placeholder');

            if (placeholder) isFull = true;

            ret.push(placeholder);
        });

        ret = isFull ? ret.join(separateValue) : '';

        return ret;
    }
    setPlaceHolders($elem = null, placeholders = ''){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isArray(placeholders)){
            CONSOLE.prop.throw(Base, 'placeholders', 'type');
        }

        let $inputs = $elem.find('input');

        _.forEach($inputs, (v, k) => {

            let $input = $(v);
            let placeholder = placeholders[k] || '';

            $input.attr('placeholder', placeholder);

            UTIL.setIE9PlaceHolder($input);
        });

        return this;
    }
    /**
     * 다음 input으로 이동하는 mousetrap
     *
     * @param {object} 기준 input
     * @param {string} 이동시 연결할 keyValue
     *
     * @private
     * @memberOf bxInput
     */
    moveToNextInput($elem = null, keyValue = ''){

        if (_.isNull($elem)) {
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(Base, '$elem', 'type');
        }

        if (!_.isString(keyValue)){
            CONSOLE.prop.throw(Base, 'keyValue', 'type');
        }

        Mousetrap($elem.get(0)).bind(keyValue, e => {

            let $next = $elem.next();

            if (
                !$next.get(0) ||
                $next.prop('nodeName') !== 'INPUT' ||
                $next.prop('disabled')){

                return false;
            }

            $next.focus();

        }, 'keyup');

        return this;
    }
    concatInputDefaultClassName(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(Base, 'opts', 'type');
        }

        let type = opts.attr.type;

        let classNames = `bx-${_getInputClassTypeName(type)}`;

        if (opts.attr.class){
            classNames += ` ${opts.attr.class}`;
        }

        opts.attr.class = classNames;

        return opts;
    }
    //removeNativeEventHandler(opts = {}){
    //
    //    if (!_.isPlainObject(opts)){
    //        CONSOLE.prop.throw(Base, 'opts', 'type');
    //    }
    //
    //    let attr = opts.attr;
    //    let ptn = /^bx-/;
    //
    //    _.map(attr, (v, k) => {
    //
    //        if (k.match(ptn)) delete opts[k];
    //    });
    //
    //    return opts;
    //}
}


/**
 * input type 반환 함수
 *
 * @param {string} attribute type value
 *
 * @return {string} input type
 *
 * @private
 * @memberOf BxInput
 */
function _getInputClassTypeName(type = ''){

    if (!_.isString(type)){
        CONSOLE.prop.throw(Base, 'type', 'type');
    }

    let ret = 'input';

    if (type === 'checkbox' || type === 'radio') ret = type;

    return ret;
}

/**
 * component option 필수 값 세팅 후 반환하는 함수
 *
 * @param {object} 기본값 옵션 객체
 *
 * @return {object} component opts
 *
 * @private
 * @memberOf BxInput
 */
function _setMandatoryOPTS(opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(Base, 'opts', 'type');
    }

    let attr = opts.attr;
    let type = attr.type;

    let id = _.trim(attr.id);

    if (!id) {
        attr.id = UTIL.uniqueId(this.inputSelector.bxInput);
    }

    return opts;
}


/**
 * 유효한 소수점 반환 함수
 *
 * @param {value} 체크할 값
 *
 * @return {array} 체크된 정수 및 소수 배열
 *
 * @private
 * @memberOf BxInput
 */
function _getValidDecimal(value = ''){

    if (!_.isString(value)){
        CONSOLE.prop.throw(Base, 'value', 'type');
    }

    let splitValue = [];
    let integer = '';
    let decimal = '';

    let dotPtn = /\./gi;
    let notNumericPtn = /\D/g;

    if (value.match(dotPtn) && value.match(dotPtn).length > 1){
        value = value.replace(notNumericPtn, '');
    }

    if (value.match(dotPtn) && value.match(dotPtn).length === 1){
        splitValue = value.split('.');
    }
    else{
        splitValue[0] = value;
    }

    //integer
    integer = splitValue[0].replace(/(^0+)/, '');

    if (integer.length === 0) integer = '0';

    //decimal
    if (!_.isUndefined(splitValue[1])) decimal = '.' + splitValue[1];

    return {
        "integer": integer,
        "decimal": decimal
    };
}

module.exports = Base;
