// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/bzno.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputBzno
 */

/**
 * @class BxInputBzno
 * @classDesc BxInputBzno 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="bzno"
 *            id="bznoInput"
 *            value="123-45-67890"
 *            placeholder="123-45-67890"
 *            label="bzno"
 *            shape="line"
 *            hinting="사업자번호를 입력하세요."
 *            inactive="false"
 *            bx-render="function(){console.log('bzno render')}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().bxInput().initialize();
 *
 *      var bznoElem = document.getElementById('bznoInput');
 *
 *      // readonly가 아닌 property는 초기 설정 값을 변경할 수 있다.
 *      // event callback property로 접근할 때는 property명의 '-' 를 생략한다. ('bx-changed' -> 'bxchanged')
 *      bznoInput.bxrender = function(){ alert('bx-render'); };
 *na
 *  });
 *
 * </script>
 */

/**
 * @property {object} bx-input - bzno 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.value=''] - value 값
 * @property {string} [bx-input.placeholder=''] - placeholder 값
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 *
 * @memberOf BxInputBzno
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');
const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const INPUT_SELECTOR = {
    "bxTypeFormatter": 'bx-type-formatter',
    "bznoTypeWrap": 'biz'
};

const INPUT_INFOS = [
    {
        "digit": 3,
        "title": '사업자번호 첫번째 자리',
        "deletePtn": /\D+/
    },
    {
        "digit": 2,
        "title": '사업자번호 두번째 자리',
        "deletePtn": /\D+/
    },
    {
        "digit": 5,
        "title": '사업자번호 세번째 자리',
        "deletePtn": /\D+/
    }
];

const SEPARATE_VALUE = '-';

const OPTS = {
    "shape": 'default',
    "hinting": '',
    "value": '',
    "placeholder": ''
};

const READONLY_ATTRS = ['shape', 'label', 'hinting'];

class BxInputBzno extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        _.assign(INPUT_SELECTOR, this.inputSelector);
    }

    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputBzno, 'opts', 'type');
        }

        let html = [];

        let type = opts.attr.type;
        let label = opts.label;
        let id = opts.attr.id;
        let hinting = opts.hinting;

        let values = super.splitValue(opts.value, INPUT_INFOS, SEPARATE_VALUE);
        //let placeholders = super.splitValue(opts.placeholder, INPUT_INFOS, SEPARATE_VALUE);

        let labelId = UTIL.uniqueId(type);

        let $wrapper = super.getWrapperElement(opts);
        let tmpClassNames = opts.attr.class;

        delete opts.attr.class;

        $wrapper.attr(opts.attr);
        $wrapper.addClass(tmpClassNames);

        if (!label) $wrapper.addClass('input-single');


        if (label){
            html.push(`<label class="bx-label" for="${labelId}">${label}</label>`);
        }

        html.push(`<div class="${INPUT_SELECTOR.bxTypeFormatter} ${INPUT_SELECTOR.bxInput} ${INPUT_SELECTOR.bznoTypeWrap}">`);

        _.forEach(INPUT_INFOS, (v, k) => {

            let digit = v.digit;
            let title = v.title;

            let value = values[k];

            let idAttr = k === 0 ? `id="${labelId}"`: '';

            html.push(`<input type="text" ${idAttr} class="bx-input digit-${digit}" value="${value}" title="${title}" />`);
        });

        html.push(`</div>`);

        if (hinting){
            html.push(`<span class="${INPUT_SELECTOR.bxHinting}">${hinting}</span>`);
        }

        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        return super.createComponentOptions($elem, opts, {}, (k, v) => {

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputBzno, k, 'type');
                }
            }

            return v;
        });
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {

                    let ret = elem[_k];

                    if (k === 'value'){
                        ret = super.getValues($elem, SEPARATE_VALUE);
                    }

                    if (k === 'placeholder'){
                        ret = super.getPlaceHolders($elem, SEPARATE_VALUE);
                    }

                    return ret;
                },
                set: (v) =>{


                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputBzno, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    if (k === 'value'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputBzno, k, 'type');
                        }

                        let values = super.splitValue(v, INPUT_INFOS, SEPARATE_VALUE);

                        super.setValues($elem, values);
                    }

                    if (k === 'placeholder'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputBzno, k, 'type');
                        }

                        let placeholders = super.splitValue(v, INPUT_INFOS, SEPARATE_VALUE);

                        super.setPlaceHolders($elem, placeholders);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputBzno, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputBzno, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputBzno, 'opts', 'type');
        }

        super.setContainerPadding($elem, opts);

        let placeholders = super.splitValue(opts.placeholder, INPUT_INFOS, SEPARATE_VALUE);

        super.setPlaceHolders($elem, placeholders);


        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        _setBznoTypeHandler.call(this, $elem);

        super.setInactive($elem.find(`.${INPUT_SELECTOR.bxInputWrap}`), opts.inactive);

        return this;
    }

}

/**
 * bzno type의 입력에 따른 이벤트 세팅 함수
 *
 * @param {object} $container - $container jquery 객체
 *
 * @private
 */
function _setBznoTypeHandler($elem = null){

    let $inputs = $elem.find('input');

    _.forEach($inputs, (v, k) => {

        let $elem = $(v);
        let digit = INPUT_INFOS[k].digit;
        let ptn = INPUT_INFOS[k].deletePtn;

        $elem.on('keyup keydown focusout', e => {

            if (!_.isUndefined(this.keyCode[e.keyCode])) return;

            let value = $elem.val();

            $elem.val(value.replace(ptn, '').substr(0, digit));
        });
    });
}

module.exports = BxInputBzno;
