// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/email.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputEmail
 */

/**
 * @class BxInputEmail
 * @classDesc BxInputEmail 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="email"
 *            id="emailInput"
 *            value="aaa@bankwareglobal.com"
 *            placeholder="bbb@bankwareglobal.com"
 *            domainList="naver.com, gmail.com"
 *            label="email"
 *            shape="line"
 *            hinting="이메일을 입력하세요."
 *            inactive="false"
 *            bx-render="function(){console.log('email render')}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().bxInput().initialize();
 *
 *      var emailElem = document.getElementById('emailInput');
 *
 *      // readonly가 아닌 property는 초기 설정 값을 변경할 수 있다.
 *      // event callback property로 접근할 때는 property명의 '-' 를 생략한다. ('bx-changed' -> 'bxchanged')
 *      emailInput.bxrender = function(){ alert('bx-render'); };
 *
 *  });
 *
 * </script>
 */

/**
 * @property {object} bx-input - email 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.domainList='naver.com, gmail.com, daum.net, hanmail.net, nate.com'] - <span class='important'>readonly</span> - 도메인 설정
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.value=''] - value 값
 * @property {string} [bx-input.placeholder=''] - placeholder 값
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 *
 * @memberOf BxInputEmail
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');
const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const INPUT_SELECTOR = {
    "bxTypeFormatter": 'bx-type-formatter',
    "emailTypeWrap": 'email',
    "domain": 'domain'
};

const DOMAIN_LIST = ['naver.com', 'gmail.com', 'daum.net', 'hanmail.net', 'nate.com'];

const INPUT_INFOS = [
    {
        "digit": 1000,
        "title": '이메일 앞자리'
    },
    {
        "digit": 1000,
        "title": '이메일 도메인 주소'
    }
];

const SEPARATE_VALUE = '@';

const OPTS = {
    "shape": 'default',
    "hinting": '',
    "value": '',
    "placeholder": '',
    "domainList": DOMAIN_LIST.join(',')
};

const READONLY_ATTRS = ['shape', 'label', 'hinting', 'domainList'];

const ATTR_NAMES = {
    "domainlist": 'domainList'
};

class BxInputEmail extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        _.assign(INPUT_SELECTOR, this.inputSelector);
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputEmail, 'opts', 'type');
        }

        let html = [];

        let type = opts.attr.type;
        let label = opts.label;
        let id = opts.attr.id;
        let hinting = opts.hinting;
        let domainList = opts.domainList.split(',');

        let values = super.splitValue(opts.value, INPUT_INFOS, SEPARATE_VALUE);
        //let placeholders = super.splitValue(opts.placeholder, INPUT_INFOS, SEPARATE_VALUE);

        let labelId = UTIL.uniqueId(type);

        let $wrapper = super.getWrapperElement(opts);
        let tmpClassNames = opts.attr.class;

        delete opts.attr.class;

        $wrapper.attr(opts.attr);
        $wrapper.addClass(tmpClassNames);

        if (!label) $wrapper.addClass('input-single');

        if (label){
            html.push(`<label class="bx-label" for="${labelId}">${label}</label>`);
        }

        html.push(`<div class="${INPUT_SELECTOR.bxTypeFormatter} ${INPUT_SELECTOR.bxInput} ${INPUT_SELECTOR.emailTypeWrap}">`);

        _.forEach(INPUT_INFOS, (v, k) => {

            let title = v.title;

            let value = values[k];

            let className = '';

            let idAttr = '';

            if (k === 0){
                className = 'bx-input digit-8';
                idAttr = `id="${labelId}"`;
            }
            else if (k === 1){
                className = 'bx-input digit-8 domain';
            }

            html.push(`<input type="text" class="${className}" ${idAttr} value="${value}" title="${title}" />`);
        });

        html.push(_getSelectHTML.call(this, domainList));

        html.push(`</div>`);

        if (hinting){
            html.push(`<span class="${INPUT_SELECTOR.bxHinting}">${hinting}</span>`);
        }

        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        return super.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputEmail, k, 'type');
                }
            }

            return v;
        });
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {

                    let ret = elem[_k];

                    if (k === 'value'){
                        ret = _getValues($elem);
                    }

                    if (k === 'placeholder'){
                        ret = _getPlaceHolders($elem);
                    }

                    return ret;
                },
                set: (v) =>{


                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputEmail, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    if (k === 'value'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputEmail, k, 'type');
                        }

                        let $selects = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);

                        // 콤보박스는 직접입력으로 변경한다.
                        $selects.prop('value', 'directDomain');

                        _setValues.call(this, $elem, v);
                    }

                    if (k === 'placeholder'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputEmail, k, 'type');
                        }

                        _setPlaceHolders.call(this, $elem, v);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputEmail, 'opts', 'type');
        }

        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        _initBxSelect.call(this, $elem);

        super.setContainerPadding($elem, opts);

        _setPlaceHolders.call(this, $elem, opts.placeholder);

        _setEmailTypeHandler.call(this, $elem);

        super.setInactive($elem.find(`.${INPUT_SELECTOR.bxInputWrap}`), opts.inactive);

        return this;
    }
}

/**
 * email type의 입력에 따른 이벤트 세팅 함수 (@입력 시, select 이동)
 *
 * @param {object} $container - $container jquery 객체
 *
 * @private
 */
function _setEmailTypeHandler($elem = null){

    let $input = $elem.find('input');

    _.forEach($input, (v) => {

        let $elem = $(v);

        $elem.on('keyup keydown focusout', e => {

            if (!_.isUndefined(this.keyCode[e.keyCode])) return;

            let value = $elem.val();

            value = value.replace(RegExp(SEPARATE_VALUE), '');

            $elem.val(value);
        });

        this.moveToNextInput($elem, SEPARATE_VALUE);
    });
}
/**
 * select html 반환 함수
 *
 * @param {string} select id
 * @param {array} optionList
 *
 * @return {string} select HTML
 *
 * @private
 */
function _getSelectHTML(optionValues = []){

    if (!_.isArray(optionValues)){
        CONSOLE.prop.throw(BxInputEmail, 'optionValues', 'type');
    }

    let html = [];

    html.push(`<bx-select style="width:140px" label="${this.altText.select}">`);

    _.forEach(optionValues, v =>{
        html.push(`<bx-option label="${v}" value="${v}"></bx-option>`);
    });

    html.push(`<bx-option label="직접입력" value="directDomain" selected="true"></bx-option>`);
    html.push(`</bx-select>`);

    return html.join('');
}

function _initBxSelect($elem = null){

    // 별도의 초기화된 인스턴스가 필요한 경우 해당 인스턴스를 생성한다.
    $elem.bxSelect().initialize();

    let $domain = $($elem.find(`.${INPUT_SELECTOR.bxInput}.${INPUT_SELECTOR.domain}`).get(0));
    let $select = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);

    $select.get(0).bxselected = function(){

        let value = $select.val();

        if (value === 'directDomain'){

            $domain.attr('disabled', false);
            $domain.val('');
        }
        else{
            $domain.attr('disabled', true);
            $domain.val(value);
        }
    };
}

function _getValues($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    let ret = [];
    let isFull = false;

    let $inputs = $elem.find('input');

    ret.push($($inputs.get(0)).val());
    ret.push($($inputs.get(1)).val());

    _.forEach(ret, (v, k) => {
        if (v) isFull = true;
    });

    ret = isFull ? ret.join(SEPARATE_VALUE) : '';

    return ret;
}

function _setValues($elem = null, value = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxInputEmail, 'value', 'type');
    }

    let $inputs = $elem.find('input');

    let values = this.splitValue(value, INPUT_INFOS, SEPARATE_VALUE);

    let $input1 = $($inputs.get(0));
    let $input2 = $($inputs.get(1));

    $($input1).val(values[0]);
    $($input2).val(values[1]);

    if (UTIL.isIE9()){
        $inputs.removeClass(INPUT_SELECTOR.placeholder);
    }

    return this;
}

function _getPlaceHolders($elem = null){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    let ret = [];
    let isFull = false;

    let placeHolderAttrName = 'placeholder';

    let $inputs = $elem.find('input');

    let $input1 = $($inputs.get(0));
    let $input2 = $($inputs.get(1));

    ret.push($input1.attr(placeHolderAttrName));
    ret.push($input2.attr(placeHolderAttrName));

    _.forEach(ret, (v, k) => {
        if (v) isFull = true;
    });

    ret = isFull ? ret.join(SEPARATE_VALUE) : '';

    return ret;
}


function _setPlaceHolders($elem = null, placeholder = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputEmail, '$elem', 'type');
    }

    if (!_.isString(placeholder)){
        CONSOLE.prop.throw(BxInputEmail, 'placeholder', 'type');
    }

    let placeHolderAttrName = 'placeholder';

    let $inputs = $elem.find('input');

    let placeholders = this.splitValue(placeholder, INPUT_INFOS, SEPARATE_VALUE);

    let $input1 = $($inputs.get(0));
    let $input2 = $($inputs.get(1));

    $input1.attr(placeHolderAttrName, placeholders[0]);
    $input2.attr(placeHolderAttrName, placeholders[1]);

    UTIL.setIE9PlaceHolder($input1);
    UTIL.setIE9PlaceHolder($input2);

    return this;
}

module.exports = BxInputEmail;
