// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/file.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputFile
 */

/**
 *
 * @class BxInputFile
 * @classDesc BxInputFile 클래스 생성자
 * @example
 *
 *  <bx-input type="file"
 *            id="inputFile"
 *            shape="round"
 *            multiple="true"
 *            maxFileCount= "2"
 *            maxFileSize= "1"
 *            accept="image/png"
 *            allowFileType="png"
 *            class="test-form"
 *            label="number"
 *            bx-render="function(){console.log('render')}"
 *            bx-selected="function(file){console.log('file',file)}"
 *            bx-error="function(errorObj){console.log('errorObj',errorObj)}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxInput().initialize();
 *
 *      var inputFile = document.getElementById('inputFile');
 *
 *      inputFile.bxselected = function(files){
 *          console.log('files', files);
 *      }
 *
 *      inputFile.bxerror = function(errorObj){
 *          console.log('errorObj', errorObj);
 *      }
 *
 *  });
 *
 * </script>
 */

/**
 * @property {object} bx-input - file 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {int} [bx-input.maxFileCount=0] - <span class='important'>readonly</span> - 업로드 할 최대 파일 개수
 * @property {int} [bx-input.maxFileSize=0] - <span class='important'>readonly</span> - 개당 파일 최대 용량 설정, 기본은 MB,<br/> <span class='important'>ie9는 지원되지 않음</span>
 * @property {int} [bx-input.maxTotalSize=0] - <span class='important'>readonly</span> - 업로드 파일 총 용량 설정, 기본은 MB,<br/> <span class='important'>ie9는 지원되지 않음</span>
 * @property {string} [bx-input.allowFileType=''] - <span class='important'>readonly</span> - 허용할 파일 확장자 설정
 * @property {string} [bx-input.accept=''] - 파일 선택 시 제한 설정,<br/> <span class='important'>ie9는 지원되지 않음</span>
 * @property {boolean} [bx-input.multiple=false] - 파일 다중선택 유무 설정,<br/> <span class='important'>ie9, 사파리 6.0 하위 버전에서는 multiple 속성이 지원되지 않음</span>
 * @property {function} [bx-input.bx-selected=function(){}] - 파일 선택 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-error=function(){}] - 파일에서 에러가 발생 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 * @alias Attributes
 * @memberOf BxInputFile
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');

const {Console: CONSOLE, Util: UTIL, Type: TYPE, File: FILE} = BXUI;

const INPUT_SELECTOR = {
    "fileText": 'file-name bx-input',
    "fileError": 'file-error',
    "fileCapacity": 'file-capa',
    "fileToolTip": '.bx-tooltip',
    "prefixTooltip": 'tooltip-'
};

const OPTS = {
    "shape": 'default',
    "hinting": '',
    "maxFileCount": 0,
    "maxFileSize": 0,
    "maxTotalSize": 0,
    "allowFileType": '',
    "bx-selected": function (){},
    "bx-error": function (){}
};

const READONLY_ATTRS = [
    'shape',
    'label',
    'hinting',
    'maxFileCount',
    'maxFileSize',
    'maxTotalSize',
    'allowFileType'
];

const ATTR_NAMES = {
    "maxfilecount": 'maxFileCount',
    "maxfilesize": 'maxFileSize',
    "maxtotalsize": 'maxTotalSize',
    "allowfiletype": 'allowFileType'
};


class BxInputFile extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        _.assign(INPUT_SELECTOR, this.inputSelector);
    }
    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputFile, 'opts', 'type');
        }

        let html = [];

        let id = opts.attr.id;
        let label = opts.label;
        let hinting = opts.hinting;
        let maxTotalSize = opts.maxTotalSize;

        let attr = UTIL.objectToStringAttr(opts.attr);

        let $wrapper = super.getWrapperElement(opts);

        $wrapper.attr('id', INPUT_SELECTOR.prefixTooltip + id);
        $wrapper.addClass('file-type');

        if (!label) $wrapper.addClass('input-single');

        if (label){
            html.push(`<label class="bx-label">${label}</label>`);
        }

        html.push(`<label for=${opts.attr.id}></label>`);

        html.push(`<input ${attr} />`);

        html.push(`<input type="text" class="${INPUT_SELECTOR.fileText}" value="" disabled />`);

        if (hinting){
            html.push(`<span class="${INPUT_SELECTOR.bxHinting}">${hinting}</span>`);
        }

        if (!UTIL.isIE9()){

            if (maxTotalSize !== 0){

                html.push(`<div style="display:none" class="${INPUT_SELECTOR.fileCapacity}"><span class="bar" style="width: 0%"></span></div>`);
            }

            html.push(`<bx-tooltip class='blind' for=${INPUT_SELECTOR.prefixTooltip + opts.attr.id} fill="false"></bx-tooltip>`);
        }

        html.push(`<div class="file-error" style="display:none"><p>error</p></div>`);


        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        opts = super.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputFile, k, 'type');
                }
            }

            if (k === 'maxTotalSize' || k === 'maxFileSize' || k === 'maxFileCount'){

                if (!TYPE.isNumber(v)){
                    CONSOLE.prop.throw(BxInputFile, k, 'type');
                }
            }

            return v;

        }, (k, v) => {

            if (k === 'multiple'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputFile, k, 'type');
                }
            }
        });

        // 사파리 `6` 하위 버전이거나, multiple 옵션이 false 인 경우
        if (UTIL.isBelowSafariVersion(6) || !opts.attr.multiple){
            delete opts.attr.multiple;
        }

        return opts;
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) =>{

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputFile, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputFile, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputFile, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputFile, 'opts', 'type');
        }

        let $file = $elem.find('input[type="file"]');
        let $input = $elem.find('input[type="text"]');

        let _this = {
            $container: $elem,
            $file: $file,
            $input: $input,
            opts: opts,
            errorList:[]
        };

        if (!UTIL.isIE9() && opts.multiple) $file.prop('multiple', opts.multiple);

        $().bxTooltip().initialize();

        super.setContainerPadding($elem, opts);

        $file.on('change', _onSelectedFile.bind(this, _this));

        _setInputFocusHandler.call(_this);

        super.setInactive($elem.find('input'), opts.inactive);

        return this;
    }
}

/**
 * 파일을 선택 했을때
 *
 * @param {object} _this = {} - context
 * @param {object} e={} - file 선택 시 이벤트 객체
 *
 * @private
 */
function _onSelectedFile(_this = {}, e = {}){

    let $target = $(e.target);
    let filesInfo = FILE.getFilesInfo($target);

    _validateFiles.call(_this, filesInfo);
}

/**
 * file validation 체크
 *
 * @param {object} fileInfo - 선택된 파일들 정보
 *
 * @private
 *
 */
function _validateFiles(filesInfo = []){

    let $container = this.$container;
    let $file = this.$file;
    let opts = this.opts;

    let totalFileSize = 0;
    let maxTotalSize = UTIL.convertStringToByte(opts.maxTotalSize);
    let maxFileCount = opts.maxFileCount;
    let maxFileSize = UTIL.convertStringToByte(opts.maxFileSize);

    let $fileProgressbar = $container.find('span.bar');
    let $fileCapacity = $container.find('.file-capa');

    let allowedList = opts.allowFileType.replace(/\s/g, '').split(',');

    let fileLength = filesInfo.length;

    _.forEach(filesInfo, file => totalFileSize += file.size);

    let remainFilePercent = parseInt(totalFileSize / maxTotalSize * 100);

    _.forEach(filesInfo, file => {

        let extensions = file.ext;
        let fileSize = file.size;
        let fileName = file.name;

        // 확장자 체크
        if (opts.allowFileType.length && _.intersection(allowedList, [extensions]).length === 0){

            _addErrorMessage.call(this, 'notAllowExtensions', filesInfo, fileName);
        }
        // 개별 파일의 크기 체크
        if ((maxFileSize !== 0) && maxFileSize < fileSize){

            _addErrorMessage.call(this, 'overFileSize', filesInfo, fileName);
        }

    });

    if (remainFilePercent >= 100){

        _addErrorMessage.call(this, 'overTotalSize');
    }

    if (maxFileCount !== 0 && maxFileCount < fileLength) {

        _addErrorMessage.call(this, 'overFileCount');
    }

    if (this.errorList.length > 0) {

        _animateErrorContainer.call(this);
        _setInputText.call(this);
        _setTooltipText.call(this);

        $fileCapacity.hide();

        $fileProgressbar.css('width', '0%');
    }
    else if(fileLength > 0){

        _setTooltipText.call(this, filesInfo);
        _setInputText.call(this, filesInfo[0].name);

        $fileCapacity.show();

        $fileProgressbar.css('width', remainFilePercent + '%');

        $file.prop('bxselected')(filesInfo);
    }
}

/**
 * 에러 메시지를 저장
 *
 * @param {string} errorType='' - 에러 타입
 * @param {array} filesInfo=[] - 선택된 파일들 정보
 * @param {array} fileName='' - 파일 이름
 * @private
 *
 */
function _addErrorMessage(errorType = '', filesInfo = [], fileName = '') {

    let errorList = this.errorList;
    let msg = FILE.MESSAGES[errorType];

    // 파일 정보에 따라 반환되는 값이 달라짐
    if (filesInfo.length > 0) {

        errorList.push({
            type: errorType,
            fileName: fileName,
            file: filesInfo,
            msg: msg
        });
    }
    else {

        errorList.push({
            type: errorType,
            msg: msg
        });
    }
}

/**
 * 에러 메시지 팝업 애니 메이션
 *
 * @private
 */
function _animateErrorContainer(){

    const duration = 500;
    const delay = 1000;

    let $file = this.$file;
    let $errorContainer = this.$container.find('.file-error');

    let errorList = this.errorList;
    let errorObj = errorList[0];

    let type = errorObj.type;
    let msg = errorObj.msg;
    let changeText = $file.prop('bxerror')(errorObj);

    if (!_.isString(changeText) || changeText.length === 0) changeText = msg;

    if (type === 'overFileSize') {

        changeText = _getCountError(errorList, type) + msg;
    }

    $errorContainer.children('p').text(changeText);

    $errorContainer.show()
        .velocity("slideDown", {duration: duration})
        .velocity("slideUp", {
            delay: delay,
            duration: duration
        }).hide();

    this.errorList = [];
}

/**
 * 해당 에러 타입으로 발생된 횟수 반환
 *
 * @param {array} errorList - 에러 리스트
 * @param {string} type - 에러 타입
 * @returns {int} count - 발생된 횟수
 *
 * @private
 */
function _getCountError(errorList = [], type = '') {

    let count = 0;

    _.map(errorList, item => {

        if (item.type === type){
            count++;
        }
    });

    return count;
}

/**
 * input의 text를 변경
 *
 * @param {string} text - 텍스트
 *
 * @private
 */
function _setInputText(text = '') {

    let $input = this.$input;

    $input.prop('value', text);
}

/**
 * tooltip의 text를 변경
 *
 * @param {array} filesInfo - 선택된 파일들 정보
 *
 * @private
 */
function _setTooltipText(filesInfo = []) {

    let $toolTip = this.$container.find(INPUT_SELECTOR.fileToolTip);
    let fullFileName = '';

    _.forEach(filesInfo, file => fullFileName += file.name + '</br>');

    let html = `<div style="text-align:left">${fullFileName}</div>`;

    if (filesInfo.length >= 2) {

        $toolTip.removeClass('blind');
        $toolTip.prop('content', html);
    }
    else {

        $toolTip.addClass('blind');
    }
}

/**
 * input focus 될 시 border 효과 handler
 *
 * @private
 */
function _setInputFocusHandler(){

    let $container = this.$container;

    $container.on('focus', 'input', e =>{

        $(e.delegateTarget).children('div').addClass('focus');
    });

    $container.on('blur change', 'input', e =>{

        $(e.delegateTarget).children('div').removeClass('focus');
    });
}

module.exports = BxInputFile;
