// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/input.js
/**
 * @class BxInput
 * @classDesc BxInput 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = require('../../base');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const TEXT_CLASS = require('./text');
const BZNO_CLASS = require('./bzno');
const IP_CLASS = require('./ip');
const PID_CLASS = require('./pid');
const EMAIL_CLASS = require('./email');
const TEL_CLASS = require('./tel');
const PASSWORD_CLASS = require('./password');
const CHECKBOX_CLASS = require('./checkbox');
const RADIO_CLASS = require('./radio');
const NUMBER_CLASS = require('./number');
const TIMEPICKER_CLASS = require('./timepicker');
const FILE_CLASS = require('./file');

// Input Factory 클래스
class BxInput {

    constructor(){

        this.text = new TEXT_CLASS();
        this.bzno = new BZNO_CLASS();
        this.ip = new IP_CLASS();
        this.pid = new PID_CLASS();
        this.email = new EMAIL_CLASS();
        this.tel = new TEL_CLASS();
        this.password = new PASSWORD_CLASS();
        this.checkbox = new CHECKBOX_CLASS();
        this.radio = new RADIO_CLASS();
        this.number = new NUMBER_CLASS();
        this.timepicker = new TIMEPICKER_CLASS();
        this.file = new FILE_CLASS();
    }

    /**
     * 초기화 함수
     *
     * <p>
     *     DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
     * </p>
     *
     *
     * @returns {object} 생성된 인스턴스
     *
     */
    initialize($compile = null, $scope = null){

        // @deprecated
        if (!_.isNull($compile)){
            CONSOLE.prop.throw(BxInput, '$compile', 'deprecatedParameter');
        }

        // @deprecated
        if (!_.isNull($scope)){
            CONSOLE.prop.throw(BxInput, '$scope', 'deprecatedParameter');
        }

        let $inputs = this.$root.find('bx-input');

        this.elems = [];

        $.each($inputs, (i, elem) => {

            let $customElem = $(elem);

            let type = $(elem).attr('type') || 'text';

            type = type.toLowerCase();

            const $container = _parseCustomElement($customElem, this[type] || this.text);

            this.elems.push($container.get(0));

        });

        return this;
    }
}

/**
 *
 * Custom Element 를 구별하기위한 플래그
 *
 * `커스텀 엘리먼트`의 경우에는 반드시 아래 프로퍼티값을 정의해야한다.
 *
 * @private
 * @type {boolean}
 */
BxInput.isCustomElement = true;

/**
 * custom 태그를 native 태그로 치환한다.
 *
 * @param {object} custom jquery element
 * @returns {object} 생성된 $container 엘리먼트
 *
 * @private
 * @memberOf BxInput
 */
function _parseCustomElement($customElem = null, instance = {}){

    if (_.isNull($customElem)){
        CONSOLE.prop.throw(BxInput, '$customElem', 'type');
    }

    if (!_.isFunction($customElem.get) && !_.isPlainObject($customElem)){
        CONSOLE.prop.throw(BxInput, '$customElem', 'type');
    }

    if (!_.isObject(instance)){
        CONSOLE.prop.throw(BxInput, 'instance', 'type');
    }

    let $container = BASE.appendContainerElement(BxInput.$root, BxInput.className);

    let opts = instance.createComponentOptions($customElem, instance.opts);

    let $component = instance.createComponentElement(opts);

    let $elem = _getElement($component, opts.attr.type);
    let elem = $elem.get(0);

    $customElem.replaceWith($container);
    $container.append($component);

    // native event 바인드
    UTIL.addAllEventListenerWithProperty($elem, opts.events);

    instance.setElementProperties($elem, opts);

    // 컴포넌트 설정을 초기화한다.
    instance.init($container, opts);

    elem.bxrender.call(elem);

    return $container;
}

function _getElement($elem = null, type = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInput, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInput, '$elem', 'type');
    }

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxInput, 'type', 'type');
    }

    let _$elem = $elem;

    if (type === 'file'){
        $elem = _$elem.find('input[type=file]');
    }
    else if (
        type === 'bzno' ||
        type === 'email' ||
        type === 'ip' ||
        type === 'pid' ||
        type === 'tel'
    ){
        $elem = _$elem;
    }
    else{
        $elem = _$elem.find('input');
    }

    return $elem;
}


module.exports = BxInput;
