// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/number.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputNumber
 */

/**
 * 값 변경 시, 호출될 callback 함수
 *
 * @callback bx-changed
 * @param {object} data - 변경 전/후 데이터
 *
 * @memberOf BxInputNumber
 */

/**
 * @class BxInputNumber
 * @classDesc BxInputNumber 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="number"
 *            shape="line"
 *            class="test-form"
 *            id="numberInput"
 *            label="number"
 *            min="0"
 *            max="100"
 *            step="5"
 *            buttonType="side"
 *            buttonIcon="arrow"
 *            bx-render="function(){console.log('render')}"
 *            bx-changed="function(e, data){console.log(data)}"
 *            bx-click="function(){console.log('click')}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxInput().initialize();
 *
 *      var numberElem = document.getElementById('numberInput');
 *
 *      console.log(numberInput.max);
 *      console.log(numberInput.min);
 *      console.log(numberInput.step);
 *
 *      // readonly가 아닌 property는 초기 설정 값을 변경할 수 있다.
 *      // event callback property로 접근할 때는 property명의 '-' 를 생략한다. ('bx-changed' -> 'bxchanged')
 *      numberElem.bxchanged = function(){ alert('bx-changed'); };
 *
 *  });
 *
 * </script>
 */

/**
 * @property {object} bx-input - number 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {object} [bx-input.icons=null] - <span class='important'>readonly</span> - icon의 class 설정 및 click 이벤트에 대한 handler 설정 객체
 * @property {object} [bx-input.icons.left=null] - <span class='important'>readonly</span> - input 왼쪽의 icon 설정 객체<br/><span class='important'>* label과 함께 적용 시, 왼쪽 아이콘은 노출되지 않는다.</span>
 * @property {string} [bx-input.icons.left.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.left.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {object} [bx-input.icons.right=null] - <span class='important'>readonly</span> - input 오른쪽의 icon 설정 객체
 * @property {string} [bx-input.icons.right.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.right.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {function} [bx-input.icons.right.onClick=null] - <span class='important'>readonly</span> - icon click callback 함수
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.placeholder=''] - placeholder 값(<span class='important'>ie9는 지원하지않음</span>)
 * @property {string} [bx-input.min] - 최소값
 * @property {string} [bx-input.max] - 최대값
 * @property {string} [bx-input.step=1] - 증감 단위값
 * @property {string} [bx-input.buttonType=default] - <span class='important'>readonly</span> - 증감 버튼 디자인 타입 (default, side, inline)
 * @property {string} [bx-input.buttonIcon=default] - <span class='important'>readonly</span> - 증감 버튼 기호 (default, arrow)
 * @property {function} [bx-input.bx-changed=function(){}] - 값 변경 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 *
 * @memberOf BxInputNumber
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');
const Mousetrap = require('mousetrap');

const BASE_CLASS = require('./base');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const INPUT_SELECTOR = {
    "numberTypeWrap": 'bx-type-number',
    "numberSideType": 'side-btn-style',
    "numberInlineType": 'separation-style',
    "numberDefaultType": 'border-style',
    "numberBtnNav": 'quantity-nav',
    "numberBtn": 'btn-quantity',
    "increase": 'increase',
    "decrease": 'decrease',
    "incBtn": 'inc-btn',
    "decBtn": 'dec-btn',
    "ctrlIcon": 'bx-ctrl-icon i-10',
    "upIcon": 'i-ctrl-up inc-btn',
    "downIcon": 'i-ctrl-down dec-btn',
    "plusIcon": 'i-ctrl-num-plus inc-btn',
    "minusIcon": 'i-ctrl-num-minus dec-btn',
    "disabled": 'disabled'
};

const OPTS = {
    "shape": 'default',
    "hinting": '',
    "icons": {left: null, right: null},
    "buttonType": 'default',
    "buttonIcon": 'default',
    "bx-changed": function (){}
};

const READONLY_ATTRS = ['shape', 'label', 'icons', 'hinting', 'buttonType', 'buttonIcon'];

const ATTR_NAMES = {
    "buttontype": 'buttonType',
    "buttonicon": 'buttonIcon'
};

const DEFAULT_VALUE = {
    "min": '-999999999999999999999999999999',
    "max": '999999999999999999999999999999',
    "step": '1'
};

class BxInputNumber extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        _.assign(INPUT_SELECTOR, this.inputSelector);
    }

    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
        }

        let html = [];

        let label = opts.label;
        let id = opts.attr.id;
        let icons = opts.icons;
        let hinting = opts.hinting;

        // 전달받은 class 값과 bx-input class 를 병합시킨다.
        opts = super.concatInputDefaultClassName(opts);

        let attr = opts.attr;

        let $wrapper = super.getWrapperElement(opts);

        if (icons){
            // icon 관련 class name 을 가져온다.
            $wrapper.addClass(super.getIconClassNames(opts));
        }

        if (!label) $wrapper.addClass('input-single');

        const valueInfo = _getQualifiedValueInfo(attr.min, attr.max, attr.value);

        attr.value = valueInfo.value;

        if (opts.shape === 'line' && opts.buttonType === 'side'){
            opts.buttonType = 'default';
        }

        if (label) html.push(`<label class="${INPUT_SELECTOR.label}" for="${id}">${label}</label>`);

        if (!label && icons && icons.left){

            let icon = icons.left;

            html.push(super.createIconElement(icon.className, icon.title));
        }

        html.push(`<div class="${_getNumberWrapClass(opts)}">`);

        switch (opts.buttonType){

            case 'side':
                html.push(_getNumberSideTypeHTML.call(this, opts));
                break;
            case 'inline':
                html.push(_getNumberDefaultTypeHTML.call(this, opts));
                break;
            default :
                html.push(_getNumberDefaultTypeHTML.call(this, opts));
                break;
        }

        html.push('</div>');

        if (icons && icons.right){

            let icon = icons.right;
            let className = icon.className + ' ' + INPUT_SELECTOR.bxActionIcon;

            html.push(super.createIconElement(className, icon.title));
        }

        if (hinting){
            html.push(`<span class="${INPUT_SELECTOR.bxHinting}">${hinting}</span>`);
        }

        $wrapper = $wrapper.append(html.join(''));

        _setDisabledIcon($wrapper, valueInfo.state);

        return $wrapper;
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        opts = super.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'icons'){

                v = UTIL.parseObject(k, v);

                if (!_.isPlainObject(v)){
                    CONSOLE.prop.throw(BxInputNumber, k, 'type');
                }
            }

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputNumber, k, 'type');
                }
            }

            return v;

        }, (k, v) => {

            if (k === 'type') v = 'text';

            if (k === 'min' || k === 'max' || k === 'step' || k === 'value'){

                if (!TYPE.isNumber(v)){
                    CONSOLE.prop.throw(BxInputNumber, k, 'type');
                }

                // TODO - x.0 에 대한 계산을 하기 위해서 string 으로 존재하는게 나을 듯 하다.
                // if (k !== 'value'){
                //     v = parseFloat(v);
                // }
            }

            return v;

        });

        const attr = opts.attr;
        attr.min = _.isUndefined(attr.min) ? DEFAULT_VALUE.min : attr.min;
        attr.max = _.isUndefined(attr.max) ? DEFAULT_VALUE.max : attr.max;
        attr.step = _.isUndefined(attr.step) ? DEFAULT_VALUE.step : attr.step;
        // IE9 의 경우, 옵션을 추가해준다.
        if (UTIL.isIE9()){
            opts.min = attr.min;
            opts.max = attr.max;
            opts.step = attr.step;
        }

        return opts;
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) => {

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputNumber, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    // TODO - min / max / step 에 대해서 string 화 시켜서 추가
                    if (k === 'min' || k === 'max' || k === 'step' || k === 'value') {
                        if (!TYPE.isNumber(v)){
                            CONSOLE.prop.throw(BxInputNumber, k, 'type');
                        }

                        v = v.toString();
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputNumber, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputNumber, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
        }

        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        _setIconPosition($elem, opts);
        super.setContainerPadding($elem, opts);

        if (opts.icons && opts.icons.right && opts.icons.right.onClick){
            super.setInputIconHandler($elem, opts.icons.right.onClick);
        }

        // 이후 추가 기능으로 들어갈 부분이다.
        UTIL.setIE9PlaceHolder($elem.find('input'));

        _setHandler($elem, opts);
        _setNumberTypeIncBtnHandler($elem, opts);
        _setNumberTypeDecBtnHandler($elem, opts);

        super.setInactive($elem.find('input'), opts.inactive);

        return this;
    }
}

/**
 * number input의 side type HTML 반환 함수
 * - side : 버튼이 좌우에 위치한 타입
 *
 * @param {object} component opts
 *
 * @return {string} number side type HTML
 *
 * @private
 * @memberOf bxInput
 */
function _getNumberSideTypeHTML(opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
    }

    let html = [];
    let buttonIcon = _getNumberButtonClass(opts.buttonIcon);

    let attr = UTIL.objectToStringAttr(opts.attr);

    let numberBtnClass = INPUT_SELECTOR.numberBtn;

    html.push(`<div class="${numberBtnClass}">`);
    html.push(this.createIconElement(buttonIcon.decIcon, this.altText.decrease));
    html.push('</div>');

    html.push(`<input ${attr} />`);

    html.push(`<div class="${numberBtnClass}">`);
    html.push(this.createIconElement(buttonIcon.incIcon, this.altText.increase));
    html.push('</div>');

    return html.join('');
}

/**
 * number input의 default type HTML 반환 함수
 *
 * @param {object} component opts
 *
 * @return {string} number default type HTML
 *
 * @private
 * @memberOf bxInput
 */
function _getNumberDefaultTypeHTML(opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
    }

    let html = [];
    let buttonIcon = _getNumberButtonClass(opts.buttonIcon);
    let attr = UTIL.objectToStringAttr(opts.attr);
    let numberBtnClass = INPUT_SELECTOR.numberBtn + ' ';

    html.push(`<input ${attr} />`);

    html.push(`<div class="${INPUT_SELECTOR.numberBtnNav}">`);

    html.push(`<div class="${numberBtnClass + INPUT_SELECTOR.increase}">`);
    html.push(this.createIconElement(buttonIcon.incIcon, this.altText.increase));
    html.push('</div>');

    html.push(`<div class="${numberBtnClass + INPUT_SELECTOR.decrease}">`);
    html.push(this.createIconElement(buttonIcon.decIcon, this.altText.decrease));
    html.push('</div>');

    html.push('</div>');

    return html.join('');
}

/**
 * number type의 wrap class 선택자 반환 함수
 *
 * @param {object} component opts
 *
 * @return {string} wrap class
 *
 * @private
 * @memberOf bxInput
 */
function _getNumberWrapClass(opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
    }

    let wrapClass = INPUT_SELECTOR.numberTypeWrap + ' ';

    switch (opts.buttonType){

        case 'side'     :
            wrapClass += INPUT_SELECTOR.numberSideType;
            break;
        case 'inline'   :
            wrapClass += INPUT_SELECTOR.numberInlineType;
            break;
        default         :
            if (opts.shape === 'default') wrapClass += INPUT_SELECTOR.numberDefaultType;
            break;
    }

    return wrapClass;
}

/**
 * number type input의 버튼 class 선택자 반환 함수
 *
 * @param {string} component opts
 *
 * @return {string} number button class
 *
 * @private
 * @memberOf bxInput
 */
function _getNumberButtonClass(buttonIcon = 'default'){

    if (!_.isString(buttonIcon)){
        CONSOLE.prop.throw(BxInputNumber, 'buttonIcon', 'type');
    }

    let commonClass = INPUT_SELECTOR.ctrlIcon + ' ';
    let incIcon = commonClass;
    let decIcon = commonClass;

    if (buttonIcon === 'arrow'){
        incIcon += INPUT_SELECTOR.upIcon;
        decIcon += INPUT_SELECTOR.downIcon;
    }
    else{
        incIcon += INPUT_SELECTOR.plusIcon;
        decIcon += INPUT_SELECTOR.minusIcon;
    }

    return {
        incIcon: incIcon,
        decIcon: decIcon
    };
}

/**
 * number type input의 event 세팅 함수 (키 입력 제한, 값 체크 등)
 *
 * @private
 *
 * @memberOf bxInput
 */
function _setHandler($elem = null, opts = {}){

    let input = $elem.find(`.${INPUT_SELECTOR.bxInput}`).get(0);

    $elem.on('keydown', `.${INPUT_SELECTOR.bxInput}`, _disableCharacterKey);

    $elem.on('keyup', `.${INPUT_SELECTOR.bxInput}`, e => { _numberTypeChangedHandler.call($(e.currentTarget)); });

    Mousetrap(input).bind('up', _handleUpKeyDown.bind(opts));

    Mousetrap(input).bind('down', _handleDownKeyDown.bind(opts));

    Mousetrap(input).bind('.', _handleDotKeyDown);

    Mousetrap(input).bind('-', _handleMinusKeyDown);

    $elem.on('blur', 'input', _numberTypeBlurHandler);

    // TODO - 위윔되는게 안불림
    // $elem.on('change', 'input', _numberTypeChangeHandler);
}

/**
 * 문자 키 입력을 비활성화 하는 이벤트 핸들러 (숫자 관련 키보드 이벤트)
 *
 * @param {object} event 객체
 *
 * @private
 * @memberOf bxInput
 */
function _disableCharacterKey(e = {}){

    let keyCode = e.keyCode;

    // key code 표 참조
    // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

    let isNumberKey = (keyCode >= 48 && keyCode <= 57) ||   // 숫자열 0 ~ 9 : 48 ~ 57
        (keyCode >= 96 && keyCode <= 105);    //키패드 0 ~ 9 : 96 ~ 105

    let isDotKey = keyCode === 110 ||    //소수점(.) : 문자키배열
        keyCode === 190;      //소수점(.) : 키패드

    let isMinusKey = keyCode === 189 ||    //- 키 : 문자키 배열
        keyCode === 173 ||    //- 키 : 문자키 배열 (Gecko 계열 브라우저)
        keyCode === 109;      //- 키 : 키패드

    let isUpKey = keyCode === 38;

    let isDownKey = keyCode === 40;

    let isControlKey = keyCode === 8 ||    //BackSpace
        keyCode === 46 ||    //Delete
        keyCode === 37 ||    //좌 화살표
        keyCode === 39 ||    //우 화살표
        keyCode === 35 ||    //End 키
        keyCode === 36 ||    //Home 키
        keyCode === 9  ||    //Tab 키
        keyCode === 13;      //Enter 키

    if (!isNumberKey && !isControlKey && !isDotKey && !isMinusKey && !isUpKey && !isDownKey){
        return false;
    }
}

/**
 * up 방향키 key down 이벤트 핸들러 (number type)
 *
 * @param {object} event 객체
 *
 * @private
 * @memberOf bxInput
 */
function _handleUpKeyDown(e = {}){

    let opts = this;

    _increaseNumberTypeValue.call(e.currentTarget, e, opts);
}

/**
 * down 방향키 key down 이벤트 핸들러 (number type)
 *
 * @param {object} event 객체
 *
 * @private
 * @memberOf bxInput
 */
function _handleDownKeyDown(e = {}){

    let opts = this;

    _decreaseNumberTypeValue.call(e.currentTarget, e, opts);
}

// /**
//  * number type input의 change 전 값을 저장하는 함수
//  *
//  * @param {object} event 객체
//  *
//  * @private
//  * @memberOf bxInput
//  */
// function _numberTypeFocusHandler(e = {}){
//
//     let $input = $(e.currentTarget);
//
//     let value = _.trim($input.val());
//
//     if (!TYPE.isNumber(value)){
//
//         $input.val('');
//         $input.attr('data-prev-value', '');
//     }
// }

/**
 * number type changed event handler
 *
 *
 * @private
 * @memberOf bxInput
 */
function _numberTypeChangedHandler(){

    const $input = this;
    const $wrapper = $input.parents(`.${INPUT_SELECTOR.numberTypeWrap}`);

    const value = $input.val();

    const min = $input.prop('min');
    const max = $input.prop('max');

    let prevValue = $input.attr('data-prev-value') || '';
    let currentValue = _.trim(value);

    const currentValueInfo = _getQualifiedValueInfo(min, max, currentValue);

    currentValue = currentValueInfo.value;

    const item = {
        prevValue: prevValue,
        currentValue: currentValue
    };

    _setDisabledIcon($wrapper, currentValueInfo.state);

    if (prevValue !== currentValue){
        $input.prop('bxchanged').call($input, item);
    }

    $input.val(currentValue);
}

/**
 * . 키 down 이벤트 핸들러 (number type)
 *
 * @param {object} event 객체
 * @return {boolean} keydown 허용 여부
 *
 * @private
 * @memberOf bxInput
 */
function _handleDotKeyDown(e = {}){

    let input = e.currentTarget;
    let $input = $(input);
    let value = _.trim($input.val());

    if (input.selectionStart === 0){

        $input.val('0.');
        return false;
    }

    let cursorIndex = value.indexOf('.');

    if (cursorIndex !== -1){

        input.selectionStart = cursorIndex;
        input.selectionEnd = cursorIndex;

        return false;
    }

    if (value.substr(0, input.selectionStart) === '-'){

        $input.val('-0.' + value.substr(1, value.length - 1));

        return false;
    }
}

/**
 * - 키 down 이벤트 핸들러 (number type)
 *
 * @param {object} event 객체
 * @return {boolean} keydown 허용 여부
 *
 * @private
 * @memberOf bxInput
 */
function _handleMinusKeyDown(e = {}){

    let input = e.currentTarget;
    let $input = $(input);
    let value = _.trim($input.val());

    if (input.selectionStart === 0) return true;

    if (value.indexOf('-') === -1) $input.val('-' + value);
    else $input.val(value.replace('-', ''));

    return false;
}

/**
 * number type blur event handler (숫자 유효성 check)
 *
 * @param {object} event 객체
 *
 * @private
 * @memberOf bxInput
 */
function _numberTypeBlurHandler(e = {}){

    const $elem = $(e.target);

    const max = $elem.prop('max');
    const min = $elem.prop('min');
    const step = $elem.prop('step').replace(/\d/g, '0');

    const $input = $(e.currentTarget);
    let value = _.trim($input.val());

    if (value.charAt(value.length - 1) === '.'){
        $input.val(value.slice(0, -1));
    }

    if (value === '-'){
        $input.val('');
        return;
    }

    value = _getQualifiedValueInfo(min, max, value).value;

    if (value.length === 2 && value.indexOf('.') === 1) {
        value = _convertFloatingPointToDecimalPoint(value, step);
    }

    $elem.attr('data-prev-value', value);
    $input.val(value);
}

/**
 * number type input의 값 증가 함수
 *
 * @param {object} 옵션 객체
 *
 * @private
 * @memberOf bxInput
 */
function _increaseNumberTypeValue(e = {}, opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
    }

    const $this = $(this);

    // TODO - 아래 코드 변경되어야 한다. [변경완]
    if ($this.hasClass('placeholder')){
        $this.val('0');
        $this.removeClass('placeholder');
    }
    if ($this.prop('disabled')) return;

    const min = $this.prop('min');
    const max = $this.prop('max');
    const step = $this.prop('step');

    let currentValue = _.trim($this.val()) || '0';

    $this.attr('data-prev-value', currentValue);
    currentValue = _convertFloatingPointToDecimalPoint(currentValue, step);
    currentValue = _getQualifiedValueInfo(min, max, _.toString(currentValue)).value;

    $this.val(currentValue);
}

/**
 * number type input의 값 감소 함수
 *
 * @param {object} 옵션 객체
 *
 * @private
 * @memberOf bxInput
 */
function _decreaseNumberTypeValue(e = {}, opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputNumber, 'opts', 'type');
    }

    let $this = $(this);

    if ($this.hasClass('placeholder')){
        $this.val('0');
        $this.removeClass('placeholder');
    }
    if ($this.prop('disabled')) return;

    const min = $this.prop('min');
    const max = $this.prop('max');
    const step = `-${$this.prop('step')}`;

    let currentValue = _.trim($this.val()) || '0';

    $this.attr('data-prev-value', currentValue);

    currentValue = _convertFloatingPointToDecimalPoint(currentValue, step);
    currentValue = _getQualifiedValueInfo(min, max, currentValue).value;

    $this.val(currentValue);
}

/**
 * 부동소수점 연산 함수
 * http://stackoverflow.com/questions/588004/is-floating-point-math-broken 참고
 *
 * @param {string} value
 * @param {string} step
 *
 * @private
 * @memberOf bxInput
 */
function _convertFloatingPointToDecimalPoint(value = '', step = '1'){

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxInputNumber, 'value', 'type');
    }

    if (!TYPE.isNumber(step)){
        CONSOLE.prop.throw(BxInputNumber, 'step', 'type');
    }

    let ret = '';

    if (value){

        const nValue = parseFloat(value || 0);
        const sStep = step;

        if (value.indexOf('.') === -1 && sStep.indexOf('.') === -1){
            ret = nValue + parseFloat(step);
        }
        else{
            // 실수를 정수로 만들기 위한 소수자리 값을 구한다.
            const vpl = (value.split('.')[1] || '').length;
            const spl = (sStep.split('.')[1] || '').length;

            const point = vpl > spl ? vpl : spl;

            const T = Number('1e' + point);
            // 실수를 정수로 만들어 연산한 후, 다시 실수로 변환하여 그 값을 반환한다.
            ret = (Math.round((nValue + parseFloat(step)) * T) / T).toFixed(point);
        }
    }
    return ret.toString();
}

/**
 * number type input의 + 버튼 이벤트 세팅 함수
 *
 * @private
 * @memberOf bxInput
 */
function _setNumberTypeIncBtnHandler($elem = null, opts = {}){

    $elem.on('click', `.${INPUT_SELECTOR.incBtn}`, e => {

        let $component = $(e.delegateTarget).find(`.${INPUT_SELECTOR.bxInputWrap}`);
        let $input = $component.find(`.${INPUT_SELECTOR.bxInput}`);

        _increaseNumberTypeValue.call($input, e, opts);
        _numberTypeChangedHandler.call($input);
    });
}

/**
 * number type input의 - 버튼 이벤트 세팅 함수
 *
 * @private
 * @memberOf bxInput
 */
function _setNumberTypeDecBtnHandler($elem = null, opts = {}){

    $elem.on('click', `.${INPUT_SELECTOR.decBtn}`, (e) => {

        let $component = $(e.delegateTarget).find(`.${INPUT_SELECTOR.bxInputWrap}`);
        let $input = $component.find(`.${INPUT_SELECTOR.bxInput}`);

        _decreaseNumberTypeValue.call($input, e, opts);
        _numberTypeChangedHandler.call($input);
    });
}

/**
 * number type의 검증된 숫자 값을 반환하는 함수
 * @param {string} min
 * @param {string} max
 * @param {string} value
 * @returns {{value: string, state: string}}
 * @private
 */
function _getQualifiedValueInfo(min = '0', max = '0', value = ''){

    if (!TYPE.isNumber(min)){
        CONSOLE.prop.throw(BxInputNumber, 'min', 'type');
    }

    if (!TYPE.isNumber(max)){
        CONSOLE.prop.throw(BxInputNumber, 'max', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxInputNumber, 'value', 'type');
    }

    let state = '';

    if (TYPE.isNumber(value)){

        state = 'between';

        const _value = parseFloat(value);
        const _min = parseFloat(min);
        const _max = parseFloat(max);

        if (_value <= _min){
            value = min;
            state = 'min';
        }

        if (_value >= _max){
            value = max;
            state = 'max';
        }

        if (value.indexOf('.') === 0) {
            value = _value;
        }
    }

    return {
        value: value,
        state: state
    };
}

/**
 * bxInput Number Type 의 Icon 및 structure css 재조정 함수
 * - safari 5.1.7 에서 float 로 인해 structure 구조 변경. calc 에 대한 연산 적용.
 * @param {{}} $container - $container instance
 * @param {{}} options - component option 객체
 * @private
 * @memberOf bxInput
 */
function _setIconPosition($elem = null, options = {}){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputNumber, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputNumber, '$elem', 'type');
    }

    if (!_.isPlainObject(options)){
        CONSOLE.prop.throw(BxInputNumber, 'options', 'type');
    }

    // 하위 작업은 window browser 에서 아이콘의 위치가 calc 로 인해 밀려나는 문제로 인해 pixel 연산 후 적용함.
    let {buttonType} = options;

    let selectorObj = UTIL.getSelectorInstance($elem, {
        quantity: 'btn-quantity',
        ctrlIcon: 'bx-ctrl-icon'
    });

    selectorObj.wrap = $elem;
    let $ctrlIcon = selectorObj.ctrlIcon;

    $ctrlIcon.length > 1 && Object.assign(selectorObj, {
        $ctrlIcons: $ctrlIcon
    });

    _.isEqual(buttonType, 'default') && _setDefaultIconPosition(selectorObj);

    _.isEqual(buttonType, 'inline') && _setInlineIconPosition(selectorObj);

    _.isEqual(buttonType, 'side') && _setSideIconPosition(selectorObj);
}

/**
 * default buttonType 아이콘 포지션 설정 함수
 * @param {{}} [$wrap=null] wrap 인스턴스
 * @param {{}}[$quantity=null] quantity 인스턴스
 * @param {{}}[$ctrlIcon=null] icon 인스턴스
 * @private
 */
function _setDefaultIconPosition({wrap: $wrap = null, quantity: $quantity = null, ctrlIcon: $ctrlIcon = null} = {}){

    let mid = Math.floor(($wrap.height() - 1) / 2);

    $($quantity[0]).height(mid);

    let iTop = ($quantity.height() - $ctrlIcon.height()) / 2,
        iLeft = ($quantity.width() - $ctrlIcon.width()) / 2;

    UTIL.setPosition($ctrlIcon, {top: iTop, left: iLeft});
}

/**
 * inline buttonType 아이콘 포지션 설정 함수
 * @param {{}} [$wrap=null] wrap 인스턴스
 * @param {{}}[$quantity=null] quantity 인스턴스
 * @param {{}}[$ctrlIcon=null] icon 인스턴스
 * @private
 */
function _setInlineIconPosition({wrap: $wrap = null, quantity: $quantity = null, ctrlIcon: $ctrlIcon = null} = {}){

    let top = ($wrap.height() - 15) / 2;

    UTIL.setPosition($quantity, {top: top});

    let iTop = ($quantity.height() - $ctrlIcon.height()) / 2 + 1,
        iLeft = ($quantity.width() - $ctrlIcon.width()) / 2;

    UTIL.setPosition($ctrlIcon, {top: iTop, left: iLeft});
}

/**
 * side buttonType 아이콘 포지션 설정 함수
 * @param {{}} [$wrap=null] wrap 인스턴스
 * @param {{}}[$quantity=null] quantity 인스턴스
 * @param {{}}[$ctrlIcons=null] icon 인스턴스 배열
 *
 * @private
 */
function _setSideIconPosition({quantity: $quantity = null, $ctrlIcons = null} = {}){

    let top = ($quantity.height() / 2 - 5);
    let left = top;

    for (let i = 0; i < $ctrlIcons.length; i++){

        let $icon = $($ctrlIcons[i]);

        UTIL.setPosition($icon, {top: top, left: left});
    }
}

/**
 * 최대, 최소값 인 경우 이미지 disabled 설정 함수
 * @param {object} [$wrapper={}] input wrapper
 * @param {string} [state=''] 변경 할 icon state
 *
 * @private
 */
function _setDisabledIcon($elem = null, state = ''){

    let $i = $elem.find('i');

    let disabled = INPUT_SELECTOR.disabled;

    if (state === 'between'){
        $i.removeClass(disabled);
    }
    else if (state === 'max'){
        $i.removeClass(disabled).filter(`.${INPUT_SELECTOR.incBtn}`).addClass(disabled);
    }
    else if (state === 'min'){
        $i.removeClass(disabled).filter(`.${INPUT_SELECTOR.decBtn}`).addClass(disabled);
    }
}

module.exports = BxInputNumber;
