// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/password.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputPassword
 */

/**
 * @class BxInputPassword
 * @classDesc BxInputPassword 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="password"
 *            shape="round"
 *            class="text-form"
 *            id="password"
 *            name="password"
 *            label="password"
 *            hinting="6~20자리 영문자 혹은 숫자로 입력해주세요."
 *            validateOption="{events: ['blur'], validator: formMethod.validator}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxInput().initialize();
 *
 *      // 엘리먼트로 접근하여 property를 가져올 수 있다.
 *      var inputElem = document.getElementById('shape');
 *
 *      console.log(inputElem.label);
 *      console.log(inputElem.shape);
 *      console.log(inputElem.icons);
 *      console.log(inputElem.inactive);
 *      console.log(inputElem.bxRender);
 *
 *  });
 *
 *  window.formMethod = {
 *
 *      validator: function(value){
 *
 *          var reg = /[a-zA-Z0-9]{6,20}/g;
 *          var result = reg.test(value);
 *
 *          if (!result){
 *              $().BxDialog().alert({
 *                  title: 'error',
 *                  content: '비밀번호는 6~20자리의 영문 또는 숫자만 가능합니다.'
 *              });
 *          }
 *
 *          return result;
 *      }
 *  };
 * </script>
 */

/**
 * @property {object} bx-input - password 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {object} [bx-input.icons=null] - <span class='important'>readonly</span> - icon의 class 설정 및 click 이벤트에 대한 handler 설정 객체
 * @property {object} [bx-input.icons.left=null] - <span class='important'>readonly</span> - input 왼쪽의 icon 설정 객체<br/><span class='important'>* label과 함께 적용 시, 왼쪽 아이콘은 노출되지 않는다.</span>
 * @property {string} [bx-input.icons.left.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.left.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {object} [bx-input.icons.right=null] - <span class='important'>readonly</span> - input 오른쪽의 icon 설정 객체
 * @property {string} [bx-input.icons.right.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.right.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {function} [bx-input.icons.right.onClick=null] - <span class='important'>readonly</span> - icon click 이벤트 callback 함수
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {object} [bx-input.validateOption=null] - 입력값에 대한 유효성 검사 설정 객체
 * @property {object[]} [bx-input.validateOption.events=null] - 유효성 검사를 발생할 이벤트 명 배열
 * @property {function} [bx-input.validateOption.validator=null] - 유효성 결과 값을 리턴하는 handler
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.placeholder=''] - <span class='important'>readonly</span> - placeholder 값
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 * @memberOf BxInputPassword
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');
const TEXT_CLASS = require('./text');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

let readOnlyAttrs = [];

const ATTR_NAMES = {};

class BxInputPassword extends TEXT_CLASS {

    constructor(){

        super();

        // text type 의 `formatter` 옵션을 제거한다.
        this.opts = UTIL.removePropertys(this.opts, 'formatter');

        this.base = new BASE_CLASS();

        readOnlyAttrs = _.union(UTIL.removeItems(this.readOnlyAttrs, 'formatter'), readOnlyAttrs);
        _.assign(ATTR_NAMES, UTIL.removePropertys(this.attrNames, 'originvalue'));

    }
    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputPassword, 'opts', 'type');
        }

        return super.createComponentElement(opts);
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        opts = this.base.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'icons' || k === 'validateOption'){

                v = UTIL.parseObject(k, v);

                if (!_.isPlainObject(v)){
                    CONSOLE.prop.throw(BxInputPassword, k, 'type');
                }

            }

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputPassword, k, 'type');
                }
            }

            return v;
        });

        return opts;
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return this.base.setElementProperties($elem, opts, readOnlyAttrs, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) => {

                    if (k === 'validateOption'){

                        v = UTIL.parseObject(k, v);

                        if (!_.isPlainObject(v)){
                            CONSOLE.prop.throw(BxInputPassword, k, 'type');
                        }

                        super.setInputValidator($elem, v);
                    }

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputPassword, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputPassword, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputPassword, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputPassword, 'opts', 'type');
        }

        if (opts.validateOption) this.base.setInputValidator($elem, opts.validateOption);

        UTIL.setIE9PlaceHolder($elem.find('input'));

        this.base.setContainerPadding($elem, opts);

        this.base.setInputFocusHandler($elem);
        this.base.setInputBlurHandler($elem);

        if (opts.icons.right && opts.icons.right.onClick){
            this.base.setInputIconHandler($elem, opts.icons.right.onClick);
        }

        this.base.setInactive($elem.find('input'), opts.inactive);

        return this;
    }
}

module.exports = BxInputPassword;
