// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/pid.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputPid
 */

/**
 * @class BxInputPid
 * @classDesc BxInputPid 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="pid"
 *            id="pidInput"
 *            value="123456-7777777"
 *            placeholder="666666-777777"
 *            enableMask="true"
 *            label="pid"
 *            shape="line"
 *            hinting="주민번호를 입력하세요."
 *            inactive="false"
 *            bx-render="function(){console.log('pid render')}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().bxInput().initialize();
 *
 *      var pidElem = document.getElementById('pidInput');
 *
 *      // readonly가 아닌 property는 초기 설정 값을 변경할 수 있다.
 *      // event callback property로 접근할 때는 property명의 '-' 를 생략한다. ('bx-changed' -> 'bxchanged')
 *      pidInput.bxrender = function(){ alert('bx-render'); };
 *
 *  });
 *
 * </script>
 */

/**
 * @property {object} bx-input - pid 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {boolean} [bx-input.enableMask=false] - <span class='important'>readonly</span> - 주민번호 뒷자리 마스킹 여부
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.value=''] - value 값
 * @property {string} [bx-input.placeholder=''] - placeholder 값
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 *
 * @memberOf BxInputPid
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');
const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const INPUT_SELECTOR = {
    "bxTypeFormatter": 'bx-type-formatter',
    "pidTypeWrap": 'pid',
    "pidTypeHiddenWrap": 'pid-hidden'
};

const INPUT_INFOS = {
    "default": [
        {
            "digit": 6,
            "title": '주민등록번호 첫번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 7,
            "title": '주민등록번호 두번째 자리',
            "deletePtn": /\D+/
        }
    ],
    "enableMask": [
        {
            "digit": 6,
            "title": '주민등록번호 첫번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 1,
            "title": '주민등록번호 두번째 자리',
            "deletePtn": /\D+/
        }
    ]
};

const SEPARATE_VALUE = '-';

const OPTS = {
    "shape": 'default',
    "hinting": '',
    "value": '',
    "placeholder": '',
    "enableMask": false
};

const READONLY_ATTRS = ['shape', 'label', 'hinting', 'enableMask'];

const ATTR_NAMES = {
    "enablemask": 'enableMask'
};

class BxInputPid extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);
        
        _.assign(INPUT_SELECTOR, this.inputSelector);
    }

    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputPid, 'opts', 'type');
        }

        let html = [];

        let type = opts.attr.type;
        let label = opts.label;
        let id = opts.attr.id;
        let hinting = opts.hinting;
        let enableMask = opts.enableMask;

        let inputInfos = _getInputInfo(enableMask);

        let values = super.splitValue(opts.value, inputInfos, SEPARATE_VALUE);
        //let placeholders = super.splitValue(opts.placeholder, inputInfos, SEPARATE_VALUE);

        let labelId = UTIL.uniqueId(type);

        let $wrapper = super.getWrapperElement(opts);
        let tmpClassNames = opts.attr.class;

        delete opts.attr.class;

        $wrapper.attr(opts.attr);
        $wrapper.addClass(tmpClassNames);

        if (!label) $wrapper.addClass('input-single');

        if (label){
            html.push(`<label class="bx-label" for="${labelId}">${label}</label>`);
        }

        html.push(`<div class="${_getPidWrapClassName(opts)}">`);


        _.forEach(inputInfos, (v, k) => {

            let type = '';
            let className = '';
            let idAttr = '';

            let digit = v.digit;
            let title = v.title;

            let value = values[k];
            //let placeholder = placeholders[k];

            if (k === 0){

                type = 'text';
                className = 'bx-input digit-6';
                idAttr = `id="${labelId}"`;

            }
            else if (k === 1){

                type = 'password';
                className = 'bx-input digit-7';

                if (enableMask) value = values[1].substr(0, 1);
            }

            html.push(`<input type="${type}" ${idAttr} class="${className}" value="${value}" title="${title}" maxlength="${digit}" />`);
        });

        html.push(`<span class="bx-btn i-view i-25"></span>`);
        html.push(`</div>`);

        if (hinting){
            html.push(`<span class="${INPUT_SELECTOR.bxHinting}">${hinting}</span>`);
        }

        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        return super.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'inactive' || k === 'enableMask'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputPid, k, 'type');
                }
            }

            return v;
        });
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        let enableMask = opts.enableMask;
        let inputInfos = _getInputInfo(enableMask);

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {

                    let ret = elem[_k];

                    if (k === 'value'){
                        ret = super.getValues($elem, SEPARATE_VALUE);
                    }

                    if (k === 'placeholder'){
                        ret = super.getPlaceHolders($elem, SEPARATE_VALUE);
                    }

                    return ret;
                },
                set: (v) =>{


                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputPid, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    if (k === 'value'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputPid, k, 'type');
                        }

                        let values = super.splitValue(v, inputInfos, SEPARATE_VALUE);

                        super.setValues($elem, values);
                    }

                    if (k === 'placeholder'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputPid, k, 'type');
                        }

                        let placeholders = super.splitValue(v, inputInfos, SEPARATE_VALUE);

                        super.setPlaceHolders($elem, placeholders);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputPid, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputPid, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputPid, 'opts', 'type');
        }

        super.setContainerPadding($elem, opts);

        let inputInfos = _getInputInfo(opts.enableMask);
        let placeholders = super.splitValue(opts.placeholder, inputInfos, SEPARATE_VALUE);

        super.setPlaceHolders($elem, placeholders);

        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        _setPidTypeHandler.call(this, $elem, opts);

        super.setInactive($elem.find(`.${INPUT_SELECTOR.bxInputWrap}`), opts.inactive);

        return this;
    }

}


/**
 * email type의 wrap class 선택자 반환 함수
 *
 * @param {object} component opts
 *
 * @return {string} wrap class
 *
 * @private
 */
function _getPidWrapClassName(opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputPid, 'opts', 'type');
    }

    let className = `${INPUT_SELECTOR.bxTypeFormatter} ${INPUT_SELECTOR.bxInput}`;

    if (opts.enableMask){
        className += ` ${INPUT_SELECTOR.pidTypeHiddenWrap}`;
    }
    else{
        className += ` ${INPUT_SELECTOR.pidTypeWrap}`;
    }

    //if (UTIL.isIE9()) className += ' ie9';

    return className;
}
/**
 * pid type의 입력에 따른 이벤트 세팅 함수 (숫자 입력, 다음 input 이동, 남여 아이콘 삽입, 눈 아이콘에 따른 이벤트)
 *
 * @param {object} $container - $container jquery 객체
 * @param {object} opts - component opts
 *
 * @private
 */
function _setPidTypeHandler($elem = null, opts = {}){

    let $input = $elem.find('input');

    let inputInfos = _getInputInfo(opts.enableMask);
    let ptn = /\D+/g;

    _.forEach($input, (v, k) => {

        let $elem = $(v);
        let digit = inputInfos[k] ? inputInfos[k].digit : 0;

        $elem.on('keyup keydown focusout', e => {

            if (!_.isUndefined(this.keyCode[e.keyCode])) return;

            let value = $elem.val();

            $elem.val(value.replace(ptn, '').substr(0, digit));
        });

        if (k === ($input.length - 1)){

            $elem.parent().find(".i-view").on('click',  e => {

                let $eyeBtn = $(e.currentTarget);

                if ($elem.attr('type') === 'text'){

                    $elem.attr('type', 'password');

                    $eyeBtn.addClass('i-view');
                    $eyeBtn.removeClass('i-blind');
                }
                else {

                    $elem.attr('type', 'text');

                    $eyeBtn.addClass('i-blind');
                    $eyeBtn.removeClass('i-view');
                }
            });
        }
    });
}

function _getInputInfo(enableMask = false){

    if (!_.isBoolean(enableMask)){
        CONSOLE.prop.throw(BxInputPid, 'enableMask', 'type');
    }

    return !enableMask ? INPUT_INFOS['default'] : INPUT_INFOS['enableMask'];
}

module.exports = BxInputPid;
