// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/radio.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputRadio
 */

/**
 * @class BxInputRadio
 * @classDesc BxInputRadio 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="radio"
 *            class="test-form"
 *            id="female"
 *            value="female"
 *            label="남자"
 *            inactive="false"
 *            name="sex"
 *            bx-click="handleInputClick">
 *  </bx-input>
 *
 *  <bx-input type="radio"
 *            class="test-form"
 *            id="male"
 *            value="male"
 *            label="여자"
 *            inactive="false"
 *            name="sex"
 *            bx-click="handleInputClick">
 *  </bx-input>
 *
 * <script type="text/javascript">
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxInput().initialize();
 *
 *      // 엘리먼트로 접근하여 property를 가져올 수 있다.
 *      var inputElem = document.getElementById('shape');
 *
 *      console.log(inputElem.label);
 *      console.log(inputElem.inactive);
 *      console.log(inputElem.bxRender);
 *
 *  });
 *
 *  function handleInputClick(e){
 *
 *      var value = $(this).val();
 *
 *      console.log(value);
 *  }
 *
 * </script>
 */

/**
 * @property {object} bx-input - radio 타입에서 사용하는 attributes
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 *
 * @memberOf BxInputRadio
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const READONLY_ATTRS = ['label'];

class BxInputRadio extends BASE_CLASS {

    constructor(){

        super();
    }
    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputRadio, 'opts', 'type');
        }

        let html = [];
        let label = opts.label;
        let id = opts.attr.id;
        let title = opts.attr.title || '';
        let altText = title || this.altText.select;

        // 전달받은 class 값과 bx-input class 를 병합시킨다.
        opts = super.concatInputDefaultClassName(opts);

        // input 엘리먼트의 title 속성은 제거한다(동일한 기능을 label 엘리먼트가 대신한다)
        delete opts.attr.title;

        let attr = UTIL.objectToStringAttr(opts.attr);

        let $wrapper = super.getWrapperElement(opts);

        html.push(`<input ${attr}>`);

        if (title){
            html.push(`<label class="bx-label" for="${id}" title="${title}">`);
        }
        else{
            html.push(`<label class="bx-label" for="${id}">`);
        }

        if (label){
            html.push(`<span>${label}</span>`);
        }
        else {
            html.push(`<span class="blind">${altText}</span>`);
        }

        html.push('</label>');

        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        return super.createComponentOptions($elem, opts, {}, (k, v) => {

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputRadio, k, 'type');
                }
            }

            return v;

        });
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) =>{

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputRadio, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputRadio, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputRadio, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputRadio, 'opts', 'type');
        }

        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        super.setInactive($elem.find('input'), opts.inactive);

        return this;
    }
}

module.exports = BxInputRadio;
