// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/tel.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputTel
 */

/**
 * @class BxInputTel
 * @classDesc BxInputTel 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="tel"
 *            id="telInput"
 *            value="010-123-4567"
 *            placeholder="010-123-4567"
 *            detailType="mobile"
 *            firstNumList="010, 011, 019"
 *            label="tel"
 *            shape="line"
 *            hinting="주민번호를 입력하세요."
 *            inactive="false"
 *            bx-render="function(){console.log('tel render')}">
 *  </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().bxInput().initialize();
 *
 *      var telElem = document.getElementById('telInput');
 *
 *      // readonly가 아닌 property는 초기 설정 값을 변경할 수 있다.
 *      // event callback property로 접근할 때는 property명의 '-' 를 생략한다. ('bx-changed' -> 'bxchanged')
 *      telInput.bxrender = function(){ alert('bx-render'); };
 *
 *  });
 *
 * </script>
 */

/**
 * @property {object} bx-input - tel 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.detailType=default] - <span class='important'>readonly</span> - tel의 상세 타입(default, mobile, telecom, area, nation)
 * @property {string} [bx-input.firstNumList='010, 011, 016, 017, 018, 019'] - <span class='important'>readonly</span> - 휴대폰 번호 앞자리 설정 <span class='important'>* mobile, telecom에서만 사용 가능</span>
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.value=''] - value 값
 * @property {string} [bx-input.placeholder=''] - placeholder 값
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 *
 * @memberOf BxInputTel
 *
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');
const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

// 통신사
const TELECOM = [
    {"label": 'SKT', "value": 'SKT'},
    {"label": 'KT', "value": 'KT'},
    {"label": 'LGT', "value": 'LGT'}
];

// 통신사 기본 번호
const FIRST_NUM_LIST = [
    {"label": '010', "value": '010'},
    {"label": '011', "value": '011'},
    {"label": '016', "value": '016'},
    {"label": '017', "value": '017'},
    {"label": '018', "value": '018'},
    {"label": '019', "value": '019'}
];

// 지역번호
const AREA = [
    {'label':'서울','value':'02'},
    {'label':'부산','value':'051'},
    {'label':'대구','value':'053'},
    {'label':'인천','value':'032'},
    {'label':'광주','value':'062'},
    {'label':'대전','value':'042'},
    {'label':'울산','value':'052'},
    {'label':'세종','value':'044'},
    {'label':'경기','value':'031'},
    {'label':'강원','value':'033'},
    {'label':'충북','value':'043'},
    {'label':'충남','value':'041'},
    {'label':'전북','value':'063'},
    {'label':'전남','value':'061'},
    {'label':'경북','value':'054'},
    {'label':'경남','value':'055'},
    {'label':'제주','value':'064'}
];

// 국가별 번호
const NATION =  [
    {'label':'Alaska','KOR':'알라스카','value':'1907'},
    {'label':'Albania','KOR':'알바니아','value':'335'},
    {'label':'Algeria','KOR':'알제리아','value':'213'},
    {'label':'American samoa','KOR':'미국령 사모아','value':'684'},
    {'label':'Andorra','KOR':'안도라','value':'376'},
    {'label':'Angola','KOR':'앙골라','value':'244'},
    {'label':'Anguilla','KOR':'안퀼라','value':'1264'},
    {'label':'Antarctica','KOR':'안타티카','value':'6721'},
    {'label':'Antigua','KOR':'안티과','value':'1809'},
    {'label':'Argentian','KOR':'아르헨티나','value':'54'},
    {'label':'Armenia','KOR':'아르메니아','value':'374'},
    {'label':'Ascension','KOR':'아세손','value':'247'},
    {'label':'Austria','KOR':'오스트리아','value':'43'},
    {'label':'Azerbaidjan','KOR':'아제르바이젠','value':'994'},
    {'label':'Azores Is','KOR':'아조레스섬','value':'351'},
    {'label':'Austalia','KOR':'호주','value':'61'},
    {'label':'Bahamas','KOR':'바하마','value':'1242'},
    {'label':'Bahrain','KOR':'바레인','value':'973'},
    {'label':'Bangladesh','KOR':'방글라데쉬','value':'880'},
    {'label':'Barbados','KOR':'바베이도즈','value':'1246'},
    {'label':'Belarus','KOR':'벨라루스','value':'375'},
    {'label':'Beigium','KOR':'벨지움','value':'32'},
    {'label':'Belize','KOR':'벨리제','value':'501'},
    {'label':'Benin','KOR':'베넹','value':'229'},
    {'label':'Bermuda','KOR':'버뮤다','value':'1441'},
    {'label':'Bhutan','KOR':'부탄','value':'975'},
    {'label':'Bolivia','KOR':'볼리비아','value':'591'},
    {'label':'Bosnia-Herzegovina','KOR':'보스니아헤르코비나','value':'387'},
    {'label':'Botswana','KOR':'보츠와나','value':'267'},
    {'label':'Brazil','KOR':'브라질','value':'55'},
    {'label':'Brunei','KOR':'브루나이','value':'673'},
    {'label':'Bulgaria','KOR':'불가리아','value':'359'},
    {'label':'Burkina Faso','KOR':'브르키나파소','value':'226'},
    {'label':'Burundi','KOR':'브룬디','value':'257'},
    {'label':'Canbodia','KOR':'캄보디아','value':'855'},
    {'label':'Cameroon','KOR':'카메룬','value':'237'},
    {'label':'Canada','KOR':'캐나다','value':'1'},
    {'label':'Canary Is.','KOR':'카나리제도','value':'3428'},
    {'label':'Cape Verde','KOR':'카푸버데','value':'238'},
    {'label':'Cayman Is.','KOR':'케이멘제도','value':'1345'},
    {'label':'Central African Rep.','KOR':'중앙아프리카','value':'236'},
    {'label':'Chad','KOR':'차드','value':'235'},
    {'label':'Chile','KOR':'칠레','value':'56'},
    {'label':'China(P.R.C)','KOR':'중국','value':'86'},
    {'label':'Christmas Is.','KOR':'크리스마스제도','value':'6191'},
    {'label':'Cocos Is','KOR':'코코스섬','value':'6191'},
    {'label':'Colombia','KOR':'콜롬비아','value':'57'},
    {'label':'Comoros','KOR':'코모로','value':'269'},
    {'label':'Congo','KOR':'콩코','value':'242'},
    {'label':'Cook Is.','KOR':'쿡스제도','value':'682'},
    {'label':'Costa Rica','KOR':'코스타리카','value':'506'},
    {'label':'Croatia','KOR':'크로아티아','value':'385'},
    {'label':'Cuba','KOR':'쿠바','value':'53'},
    {'label':'Cyprus','KOR':'싸이프러스','value':'357'},
    {'label':'Czech','KOR':'체코','value':'420'},
    {'label':'Denmark','KOR':'덴마크','value':'45'},
    {'label':'Deutschland','KOR':'Germany ,독일, 도이칠란트','value':'49'},
    {'label':'Diego Garcia','KOR':'디에고가르시아','value':'246'},
    {'label':'Djibouti','KOR':'지브타','value':'253'},
    {'label':'Dominica Is.','KOR':'도미니카','value':'1767'},
    {'label':'Dominican Rep.','KOR':'도미니카공화국','value':'1767'},
    {'label':'Ecuador','KOR':'에콰도르','value':'593'},
    {'label':'Egypt','KOR':'이집트','value':'20'},
    {'label':'El Salvador','KOR':'엘살바도로','value':'503'},
    {'label':'Equatorial Guinea','KOR':'적도기니아','value':'240'},
    {'label':'Eritrea','KOR':'에리트리아','value':'291'},
    {'label':'Estonia','KOR':'에스토니아','value':'372'},
    {'label':'Ethiopia','KOR':'에티오피아','value':'251'},
    {'label':'Falkland Is.','KOR':'포클랜드섬','value':'500'},
    {'label':'Foroe Is.(DEN)','KOR':'DEN','value':'298'},
    {'label':'Fiji','KOR':'피지','value':'679'},
    {'label':'Finland','KOR':'핀란드','value':'358'},
    {'label':'France','KOR':'프랑스','value':'33'},
    {'label':'Gabon','KOR':'가봉','value':'241'},
    {'label':'Gambia','KOR':'잠비아','value':'220'},
    {'label':'Georgia','KOR':'그루지아','value':'995'},
    {'label':'Ghana','KOR':'가나','value':'233'},
    {'label':'Gibrltar','KOR':'지브랄타','value':'350'},
    {'label':'Greece','KOR':'그리스','value':'30'},
    {'label':'Greenland(DEN)','KOR':'그린란드','value':'299'},
    {'label':'Grenada','KOR':'그레나다','value':'1809'},
    {'label':'Guadeloupe Is.','KOR':'과델로프','value':'590'},
    {'label':'Guatemala','KOR':'과테말라','value':'502'},
    {'label':'Guiana(French)','KOR':'가아나','value':'594'},
    {'label':'Guinea Bissau','KOR':'가나아비쏘우','value':'245'},
    {'label':'Guinea(R.P)','KOR':'기니아','value':'224'},
    {'label':'Guyana','KOR':'기이니아','value':'592'},
    {'label':'Haiti','KOR':'아이티','value':'509'},
    {'label':'Hawaii','KOR':'하와이','value':'1808'},
    {'label':'Honduras','KOR':'온듀라스','value':'504'},
    {'label':'Hong Kong','KOR':'홍콩','value':'852'},
    {'label':'Hungaruy','KOR':'헝가리','value':'36'},
    {'label':'Iceland','KOR':'아이슬랜드','value':'354'},
    {'label':'India','KOR':'인디아','value':'91'},
    {'label':'Indonesia','KOR':'인도네시아','value':'62'},
    {'label':'Iran','KOR':'이란','value':'98'},
    {'label':'Iraq','KOR':'이라크','value':'964'},
    {'label':'Ireland','KOR':'아일랜드','value':'353'},
    {'label':'Israel','KOR':'이스라엘','value':'972'},
    {'label':'Italy','KOR':'이태리','value':'39'},
    {'label':'Ivory Coast','KOR':'아이보리코스트','value':'225'},
    {'label':'Jamaica','KOR':'자마이카','value':'1876'},
    {'label':'Japan','KOR':'일본','value':'81'},
    {'label':'Jordan','KOR':'요르단','value':'962'},
    {'label':'Kazakhstan','KOR':'카자흐스탄','value':'7'},
    {'label':'Kenya','KOR':'케냐','value':'254'},
    {'label':'Kiribati','KOR':'키리바티','value':'686'},
    {'label':'South Korea','KOR':'대한민국, 한국, 남한','value':'82'},
    {'label':'Kuwait','KOR':'쿠웨이트','value':'965'},
    {'label':'Kyrayzstan','KOR':'키르기즈스탄공화국','value':'7'},
    {'label':'Laos','KOR':'라오스','value':'856'},
    {'label':'Latvia','KOR':'라트비아','value':'371'},
    {'label':'Lithuania','KOR':'리투아니아','value':'370'},
    {'label':'Luxembourg','KOR':'룩셈부르크','value':'362'},
    {'label':'Lebanon','KOR':'레바논','value':'961'},
    {'label':'Lesotho','KOR':'레스토','value':'266'},
    {'label':'Liberia','KOR':'라이베리아','value':'231'},
    {'label':'Libya','KOR':'리비아','value':'218'},
    {'label':'Liechtenstein','KOR':'리히텐슈타인','value':'4175'},
    {'label':'Mozambique','KOR':'모잠비크','value':'258'},
    {'label':'Myanmar(Burma)','KOR':'미얀마','value':'95'},
    {'label':'Macao','KOR':'마카오','value':'853'},
    {'label':'Macedonia','KOR':'마케도니아','value':'389'},
    {'label':'Madagascar','KOR':'마다가스칼','value':'261'},
    {'label':'Madeira Is.','KOR':'마데이라섬','value':'35191'},
    {'label':'Malawi','KOR':'말라위','value':'265'},
    {'label':'Malaysia','KOR':'말레이지아','value':'60'},
    {'label':'Maldives','KOR':'몰디브','value':'960'},
    {'label':'Mali','KOR':'말리','value':'223'},
    {'label':'Malta','KOR':'몰타','value':'356'},
    {'label':'Marshall Is.','KOR':'마샬제도','value':'692'},
    {'label':'Martinique','KOR':'마르티니크','value':'596'},
    {'label':'Mouritania','KOR':'모리타니아','value':'222'},
    {'label':'Mauritius','KOR':'모리셔스','value':'230'},
    {'label':'Mayotte','KOR':'마요트','value':'2696'},
    {'label':'Mexico','KOR':'멕시코','value':'52'},
    {'label':'Moldova','KOR':'몰도바','value':'373'},
    {'label':'Monaco','KOR':'모나코','value':'377'},
    {'label':'Mongolla','KOR':'몽골','value':'976'},
    {'label':'Montserrat','KOR':'몬데세라토','value':'1664'},
    {'label':'Morocco','KOR':'모로코','value':'212'},
    {'label':'Norfolk Is.','KOR':'노르폭섬','value':'6723'},
    {'label':'Norway','KOR':'노르웨이','value':'47'},
    {'label':'Namibia','KOR':'나미비아','value':'264'},
    {'label':'Nauru','KOR':'나우르','value':'674'},
    {'label':'Nepal','KOR':'네팔','value':'977'},
    {'label':'Netherlands','KOR':'네델란드','value':'31'},
    {'label':'Netherlands antilles','KOR':'안틸레스','value':'599'},
    {'label':'New Caledonia','KOR':'뉴칼레도니아','value':'687'},
    {'label':'New Zealand','KOR':'뉴질랜드','value':'64'},
    {'label':'Nicaragua','KOR':'니카라과','value':'505'},
    {'label':'Niger','KOR':'니제르','value':'227'},
    {'label':'Nigeria','KOR':'나이제리아','value':'234'},
    {'label':'Niue Is.','KOR':'나우에섬','value':'683'},
    {'label':'North Korea','KOR':'북조선, 북한','value':'85'},
    {'label':'Oman','KOR':'오만','value':'968'},
    {'label':'Portugal','KOR':'포르투칼','value':'351'},
    {'label':'Puerto Rico','KOR':'푸에르토리코','value':'1787'},
    {'label':'Pakistan','KOR':'파키스탄','value':'92'},
    {'label':'Palau','KOR':'팔라우','value':'680'},
    {'label':'Panama','KOR':'파나마','value':'507'},
    {'label':'Papua New Guines','KOR':'파푸뉴기니아','value':'675'},
    {'label':'Paraguay','KOR':'파라과이아','value':'595'},
    {'label':'Peru','KOR':'페루','value':'51'},
    {'label':'Philippines','KOR':'필리핀','value':'63'},
    {'label':'Poland','KOR':'폴란드','value':'48'},
    {'label':'Qatar','KOR':'카타르','value':'974'},
    {'label':'Reunion Is.','KOR':'레위니옹','value':'262'},
    {'label':'Romania','KOR':'루마니아','value':'40'},
    {'label':'Russia','KOR':'러시아','value':'7'},
    {'label':'Rwanda','KOR':'르완다','value':'250'},
    {'label':'Saipan','KOR':'사이판','value':'670'},
    {'label':'San Marino','KOR':'산마리노','value':'378'},
    {'label':'Saudi Arabia','KOR':'사우디아라비아','value':'966'},
    {'label':'Senegal','KOR':'세네칼','value':'221'},
    {'label':'Seychelles','KOR':'세이셀','value':'248'},
    {'label':'Sierra Leone','KOR':'씨에라네온','value':'232'},
    {'label':'Singapore','KOR':'싱가폴','value':'65'},
    {'label':'Slovak','KOR':'슬로바키아','value':'421'},
    {'label':'Slovenia','KOR':'슬로베니아','value':'386'},
    {'label':'Solomon Is.','KOR':'솔로몬제도','value':'677'},
    {'label':'Somalia','KOR':'소말리아','value':'252'},
    {'label':'South Africa','KOR':'남아프리카공화국, 남아공','value':'27'},
    {'label':'España, Spain','KOR':'에스파냐, 스페인','value':'34'},
    {'label':'Sri Lanca','KOR':'스리랑카','value':'94'},
    {'label':'St. christopher','KOR':'세인트크리스토퍼','value':'1809'},
    {'label':'St. Helena','KOR':'세인트헬레나','value':'290'},
    {'label':'St. Kitts','KOR':'세인드키츠','value':'1809'},
    {'label':'St. Lucia','KOR':'세인트루시아','value':'1758'},
    {'label':'St. Vincent','KOR':'세인트빈센트','value':'1809'},
    {'label':'Sudan','KOR':'수단','value':'249'},
    {'label':'Suriname','KOR':'수리남','value':'597'},
    {'label':'Swaziland','KOR':'스와질랜드','value':'268'},
    {'label':'Sweden','KOR':'스웨덴','value':'46'},
    {'label':'Switzerland','KOR':'스위스','value':'41'},
    {'label':'Syria','KOR':'시리아','value':'963'},
    {'label':'Tadzhikistan','KOR':'타자키스탄공화국','value':'7'},
    {'label':'Tahiti','KOR':'타이티','value':'689'},
    {'label':'Taiwan','KOR':'타이완, 대만','value':'886'},
    {'label':'Tanzania','KOR':'탄자니아','value':'255'},
    {'label':'Thailand','KOR':'태국','value':'66'},
    {'label':'Togo','KOR':'토고','value':'228'},
    {'label':'Tonga','KOR':'통가','value':'676'},
    {'label':'Trinidad &Tobogo','KOR':'트리니다드토바고','value':'1868'},
    {'label':'Tunisia','KOR':'튀니지','value':'216'},
    {'label':'Turkiye, Turkey','KOR':'투르키예, 터키','value':'90'},
    {'label':'Turkmenistan','KOR':'투르크메니스탄','value':'993'},
    {'label':'Turks &Caicos Is.','KOR':'턱스케이코스제도','value':'1649'},
    {'label':'Tuvalu','KOR':'투발루','value':'688'},
    {'label':'Uganda','KOR':'우간다','value':'256'},
    {'label':'Ukraine','KOR':'우크라이나','value':'380'},
    {'label':'United Arab Emirates','KOR':'아랍에미레이터','value':'971'},
    {'label':'United Kingdom','KOR':'영국','value':'44'},
    {'label':'Uruguay','KOR':'우르과이','value':'598'},
    {'label':'U.S.A','KOR':'미국 본토','value':'1'},
    {'label':'Uzbekistan','KOR':'우주베키스탄','value':'7'},
    {'label':'Vanuatu','KOR':'바누아투','value':'678'},
    {'label':'Vatican City','KOR':'바티칸시티','value':'396'},
    {'label':'Venezuela','KOR':'베네쥬엘라','value':'58'},
    {'label':'Viet Nam','KOR':'베트남','value':'84'},
    {'label':'Westem Samoa','KOR':'서사모아','value':'685'},
    {'label':'Yemen Arab Rep.','KOR':'북예멘','value':'967'},
    {'label':'Yemen(PDR)','KOR':'예멘','value':'969'},
    {'label':'Yogoslavia','KOR':'유고슬라비아','value':'381'},
    {'label':'Zaire','KOR':'자이레','value':'243'},
    {'label':'Zambia','KOR':'잠비아','value':'260'},
    {'label':'Zimbabwe','KOR':'짐바브웨','value':'263'}
];

const INPUT_SELECTOR = {
    "bxTypeFormatter": 'bx-type-formatter',
    "telTypeWrap": 'tel',
    "mobileTypeWrap": 'tel-mobile',
    "nationTypeWrap": 'tel-int',
    "bx-select-wrap": 'bx-select-wrap'
};

const INPUT_INFOS = {

    "default": [
        {
            "digit": 4,
            "title": '전화번호 첫번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 4,
            "title": '전화번호 두번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 4,
            "title": '전화번호 세번째 자리',
            "deletePtn": /\D+/
        }
    ],
    "mobile": [
        {
            "digit": 3,
            "title": '휴대폰 전화번호 첫번째 자리'
        },
        {
            "digit": 4,
            "title": '휴대폰 전화번호 두번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 4,
            "title": '휴대폰 전화번호 세번째 자리',
            "deletePtn": /\D+/
        }
    ],
    "area": [
        {
            "digit": 3,
            "title": '지역번호'
        },
        {
            "digit": 4,
            "title": '전화번호 첫번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 4,
            "title": '전화번호 두번째 자리',
            "deletePtn": /\D+/
        }
    ],
    "telecom": [
        {
            "digit": 3,
            "title": '통신사명'
        },
        {
            "digit": 4,
            "title": '전화번호 두번째 자리'
        },
        {
            "digit": 4,
            "title": '전화번호 세번째 자리',
            "deletePtn": /\D+/
        },
        {
            "digit": 4,
            "title": '전화번호 세번째 자리',
            "deletePtn": /\D+/
        }
    ],
    "nation": [
        {
            "digit": 4,
            "title": '국가 전화번호'
        },
        {
            "digit": 12,
            "title": '전화번호',
            "deletePtn": /\D+/
        }
    ]
};

const SEPARATE_VALUE = '-';

const OPTS = {
    "shape": 'default',
    "hinting": '',
    "value": '',
    "placeholder": '',
    "detailType": 'default',
    "firstNumList": ''
};

const READONLY_ATTRS = ['shape', 'label', 'hinting', 'firstNumList', 'detailType'];

const ATTR_NAMES = {
    "firstnumlist": 'firstNumList',
    "detailtype": 'detailType'
};


class BxInputTel extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        _.assign(INPUT_SELECTOR, this.inputSelector);
    }

    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputTel, 'opts', 'type');
        }

        let html = [];

        let type = opts.attr.type;
        let label = opts.label;
        let id = opts.attr.id;
        let hinting = opts.hinting;
        let detailType = opts.detailType;
        let firstNumList = opts.firstNumList ? _convertArrayToArrayObject(opts.firstNumList.split(',')) : FIRST_NUM_LIST;

        let labelId = UTIL.uniqueId(type);
        let inputInfos = _getInputInfo(detailType);

        let values = super.splitValue(opts.value, inputInfos, SEPARATE_VALUE);
        //let placeholders = super.splitValue(opts.placeholder, inputInfos, SEPARATE_VALUE);

        let $wrapper = super.getWrapperElement(opts);
        let tmpClassNames = opts.attr.class;

        delete opts.attr.class;

        $wrapper.attr(opts.attr);
        $wrapper.addClass(tmpClassNames);

        if (!label) $wrapper.addClass('input-single');

        let getHTML = function(){};

        if (label){
            html.unshift(`<label class="bx-label" for="${labelId}">${label}</label>`);
        }

        html.push(`<div class="${_getTelWrapClassName(opts)}">`);


        if (detailType === 'default'){

            getHTML = (v, k) => {

                let title = v.title;

                let value = values[k];
                //let placeholder = placeholders[k];

                let idAttr = k === 0 ? `id="${labelId}"`: '';

                html.push(`<input type="text" class="bx-input digit-4" ${idAttr} value="${value}" title="${title}" />`);
            };
        }
        else if (detailType === 'mobile' || detailType === 'area' || detailType === 'nation'){

            let optionValues = [];
            let className = 'bx-input digit-4';

            if (detailType === 'mobile'){
                optionValues = firstNumList;
            }
            else if (detailType === 'area'){
                optionValues = AREA;
            }
            else if (detailType === 'nation'){
                optionValues = NATION;
                className = 'bx-input digit-8';
            }

            getHTML = (v, k) => {

                let title = v.title;

                let value = values[k];
                //let placeholder = placeholders[k];

                let idAttr = k === 1 ? `id="${labelId}"`: '';

                if (k === 0){
                    html.push(_getSelectHTML.call(this, detailType, value, optionValues));
                }
                else{
                    html.push(`<input type="text" class="${className}" ${idAttr} value="${value}" title="${title}" />`);
                }
            };
        }
        else if (detailType === 'telecom'){

            getHTML = (v, k) => {

                let title = v.title;

                let value = values[k];
                //let placeholder = placeholders[k];

                let idAttr = k === 2 ? `id="${labelId}"`: '';

                if (k === 0){
                    html.push(_getSelectHTML.call(this, detailType, value, TELECOM));
                }
                else if (k === 1){
                    html.push(_getSelectHTML.call(this, detailType, value, firstNumList));
                }
                else{
                    html.push(`<input type="text" class="bx-input digit-4" ${idAttr} value="${value}" title="${title}" />`);
                }
            };
        }

        // wrapper 엘리먼트에 바인딩할 HTML 을 생성한다.
        _.forEach(inputInfos, getHTML);


        html.push(`</div>`);

        if (hinting){
            html.push(`<span class="${INPUT_SELECTOR.bxHinting}">${hinting}</span>`);
        }

        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        return super.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputTel, k, 'type');
                }
            }

            return v;
        });
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        let detailType = opts.detailType;

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {

                    let ret = elem[_k];

                    if (k === 'value'){
                        ret = _getValues($elem, detailType);
                    }

                    if (k === 'placeholder'){
                        ret = _getPlaceHolders($elem, detailType);
                    }

                    return ret;
                },
                set: (v) =>{

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputTel, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    if (k === 'value'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputTel, k, 'type');
                        }

                        _setValues.call(this, $elem, detailType, v);
                    }

                    if (k === 'placeholder'){

                        if (!_.isString(v)){
                            CONSOLE.prop.throw(BxInputTel, k, 'type');
                        }

                        _setPlaceHolders.call(this, $elem, detailType, v);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputTel, 'opts', 'type');
        }

        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        _initBxSelect($elem, opts);

        super.setContainerPadding($elem, opts);

        _setPlaceHolders.call(this, $elem, opts.detailType, opts.placeholder);

        _setTelTypeHandler.call(this, $elem, opts);

        super.setInactive($elem.find(`.${INPUT_SELECTOR.bxInputWrap}`), opts.inactive);

        return this;
    }
}


/**
 * email type의 wrap class 선택자 반환 함수
 *
 * @param {object} component opts
 *
 * @return {string} wrap class
 *
 * @private
 */
function _getTelWrapClassName(opts = {}){

    if (!_.isPlainObject(opts)){
        CONSOLE.prop.throw(BxInputTel, 'opts', 'type');
    }

    let className = `${INPUT_SELECTOR.bxTypeFormatter} ${INPUT_SELECTOR.bxInput}`;

    if (opts.detailType === 'nation') {
        className += ` ${INPUT_SELECTOR.nationTypeWrap}`;
    }
    else if (opts.detailType === 'mobile' || opts.detailType === 'telecom') {
        className += ` ${INPUT_SELECTOR.mobileTypeWrap}`;
    }
    else{
        className += ` ${INPUT_SELECTOR.telTypeWrap}`;
    }

    return className;
}
/**
 * tel type의 입력에 따른 이벤트 세팅 함수 (숫자 입력, 다음 input 이동)
 *
 * @param {object} $container - $container jquery 객체
 * @param {object} opts - component opts
 *
 * @private
 */
function _setTelTypeHandler($elem = null, opts = {}){

    let $input = $elem.find('input');
    let digit = opts.detailType === 'nation' ? 12 : 4;

    let defaultTypePtn = /\D+/;
    let nationTypePtn = /[^0-9\-]+/;

    _.forEach($input, (v) => {

        let $elem = $(v);

        $elem.on('keyup keydown focusout', e => {

            if (!_.isUndefined(this.keyCode[e.keyCode])) return;

            let value = $elem.val();

            if (opts.detailType === 'nation'){
                value = value.replace(nationTypePtn, '').substr(0, digit);
            }
            else{
                value = value.replace(defaultTypePtn, '').substr(0, digit);
            }

            $elem.val(value);
        });
    });
}
/**
 * select html 반환 함수
 *
 * @param {string} opts html
 * @param {string} id id
 * @param {array} optionArr
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 */
function _getSelectHTML(type = '', value = '', optionValues = null){

    if (!_.isString(type)){
        CONSOLE.prop.throw(BxInputTel, 'type', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxInputTel, 'value', 'type');
    }

    if (!_.isArray(optionValues)){
        CONSOLE.prop.throw(BxInputTel, 'optionValues', 'type');
    }

    let html = [];
    let width = type === 'nation' ? 140 : 80;

    html.push(`<bx-select style="width:${width}px" label="${this.altText.select}" value="${value}">`);

    _.forEach(optionValues, (v, k) => {

        let label = v.label;
        let value = v.value;

        if (type === 'area' || type === 'nation'){
            label = `${label} (${value})`;
        }

        html.push(`<bx-option label="${label}" value="${value}"></bx-option>`);

    });

    html.push(`</bx-select>`);

    return html.join('');
}

function _convertArrayToArrayObject(v = []){

    if (!_.isArray(v)){
        CONSOLE.prop.throw(BxInputTel, 'v', 'type');
    }

    let ret = [];

    _.forEach(v, vv => {
        ret.push({
            label: vv,
            value: vv
        });
    });

    return ret;
}


function _initBxSelect($elem = null, opts = {}){

    // 별도의 초기화된 인스턴스가 필요한 경우 해당 인스턴스를 생성한다.
    $elem.bxSelect().initialize();

    let detailType = opts.detailType;

    let $bxSelectWrap = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);

    let $firstNumber = null;
    let firstNumberValue = '';

    if (detailType === 'telecom'){

        let $telecom = $($bxSelectWrap.get(0));
        let telecomValue = $telecom.val();

        $firstNumber = $($bxSelectWrap.get(1));
        firstNumberValue = $firstNumber.val();

        $telecom.val(telecomValue);
        $firstNumber.val(firstNumberValue);
    }
    else{

        $firstNumber = $($bxSelectWrap.get(0));
        firstNumberValue = $firstNumber.val();

        $firstNumber.val(firstNumberValue);
    }

    return this;
}

function _getInputInfo(detailType = ''){

    if (!_.isString(detailType)){
        CONSOLE.prop.throw(BxInputTel, 'detailType', 'type');
    }

    if (_.isUndefined(INPUT_INFOS[detailType])){
        CONSOLE.prop.throw(BxInputTel, 'detailType', 'type');
    }

    return INPUT_INFOS[detailType];
}

function _getValues($elem = null, detailType = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isString(detailType)){
        CONSOLE.prop.throw(BxInputTel, 'detailType', 'type');
    }

    let ret = [];
    let isFull = false;

    let $selects = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);
    let $inputs = $elem.find('input');

    if (detailType === 'default'){

        _.forEach($inputs, (v, k) => {
            ret.push($(v).val());
        });
    }
    else if (detailType === 'mobile' || detailType === 'area'){

        ret.push($selects.prop('value'));

        ret.push($($inputs.get(0)).val());
        ret.push($($inputs.get(1)).val());
    }
    else if (detailType === 'nation'){

        ret.push($selects.prop('value'));

        ret.push($($inputs.get(0)).val());
    }
    else if (detailType === 'telecom'){

        ret.push($($selects.get(0)).prop('value'));
        ret.push($($selects.get(1)).prop('value'));

        ret.push($($inputs.get(0)).val());
        ret.push($($inputs.get(1)).val());
    }

    _.forEach(ret, (v, k) => {
        if (v) isFull = true;
    });

    ret = isFull ? ret.join(SEPARATE_VALUE) : '';

    return ret;
}


function _setValues($elem = null, detailType = '', value = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isString(detailType)){
        CONSOLE.prop.throw(BxInputTel, 'detailType', 'type');
    }

    if (!_.isString(value)){
        CONSOLE.prop.throw(BxInputTel, 'value', 'type');
    }

    let inputInfos = _getInputInfo(detailType);

    let $selects = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);
    let $inputs = $elem.find('input');

    let values = this.splitValue(value, inputInfos, SEPARATE_VALUE);

    if (detailType === 'default'){

        _.forEach($inputs, (v, k) => {
            $(v).val(values[k]);
        });
    }
    else if (detailType === 'mobile' || detailType === 'area'){

        $selects.prop('value', values[0]);

        $($inputs.get(0)).val(values[1]);
        $($inputs.get(1)).val(values[2]);
    }
    else if (detailType === 'nation'){

        $selects.prop('value', values[0]);

        $($inputs.get(0)).val(values[1]);
    }
    else if (detailType === 'telecom'){

        $($selects.get(0)).prop('value', values[0]);
        $($selects.get(1)).prop('value', values[1]);

        $($inputs.get(0)).val(values[2]);
        $($inputs.get(1)).val(values[3]);
    }

    if (UTIL.isIE9()){
        $inputs.removeClass(INPUT_SELECTOR.placeholder);
    }

    return this;
}

function _getPlaceHolders($elem = null, detailType = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isString(detailType)){
        CONSOLE.prop.throw(BxInputTel, 'detailType', 'type');
    }

    let ret = [];
    let isFull = false;

    let $selects = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);
    let $inputs = $elem.find('input');

    let placeHolderAttrName = 'placeholder';

    if (detailType === 'default'){

        _.forEach($inputs, (v, k) => {
            ret.push($(v).attr(placeHolderAttrName));
        });
    }
    else if (detailType === 'mobile' || detailType === 'area'){

        let $select1 = $($selects.get(0));

        let $input1 = $($inputs.get(0));
        let $input2 = $($inputs.get(1));

        ret.push($select1.attr(placeHolderAttrName));

        ret.push($input1.attr(placeHolderAttrName));
        ret.push($input2.attr(placeHolderAttrName));
    }
    else if (detailType === 'nation'){

        let $select1 = $($selects.get(0));
        let $input1 = $($inputs.get(0));

        ret.push($select1.attr(placeHolderAttrName));

        ret.push($input1.attr(placeHolderAttrName));
    }
    else if (detailType === 'telecom'){

        let $select1 = $($selects.get(0));
        let $select2 = $($selects.get(1));

        let $input1 = $($inputs.get(0));
        let $input2 = $($inputs.get(1));

        ret.push($select1.attr(placeHolderAttrName));
        ret.push($select2.attr(placeHolderAttrName));

        ret.push($input1.attr(placeHolderAttrName));
        ret.push($input2.attr(placeHolderAttrName));
    }

    _.forEach(ret, (v, k) => {
        if (v) isFull = true;
    });

    ret = isFull ? ret.join(SEPARATE_VALUE) : '';

    return ret;
}

function _setPlaceHolders($elem = null, detailType = '', placeholder = ''){

    if (_.isNull($elem)){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
        CONSOLE.prop.throw(BxInputTel, '$elem', 'type');
    }

    if (!_.isString(detailType)){
        CONSOLE.prop.throw(BxInputTel, 'detailType', 'type');
    }

    if (!_.isString(placeholder)){
        CONSOLE.prop.throw(BxInputTel, 'placeholder', 'type');
    }

    let inputInfos = _getInputInfo(detailType);

    let $selects = $elem.find(`.${INPUT_SELECTOR.bxSelectWrap}`);
    let $inputs = $elem.find('input');

    let placeholders = this.splitValue(placeholder, inputInfos, SEPARATE_VALUE);
    let placeHolderAttrName = 'placeholder';

    if (detailType === 'default'){

        _.forEach($inputs, (v, k) => {

            let $input = $(v);

            $input.attr(placeHolderAttrName, placeholders[k]);

            UTIL.setIE9PlaceHolder($input);
        });
    }
    else if (detailType === 'mobile' || detailType === 'area'){

        let $select1 = $($selects.get(0));

        let $input1 = $($inputs.get(0));
        let $input2 = $($inputs.get(1));

        $select1.attr(placeHolderAttrName, placeholders[0]);

        $input1.attr(placeHolderAttrName, placeholders[1]);
        $input2.attr(placeHolderAttrName, placeholders[2]);

        UTIL.setIE9PlaceHolder($input1);
        UTIL.setIE9PlaceHolder($input2);
    }
    else if (detailType === 'nation'){

        let $select1 = $($selects.get(0));
        let $input1 = $($inputs.get(0));

        $select1.attr(placeHolderAttrName, placeholders[0]);
        $input1.attr(placeHolderAttrName, placeholders[1]);

        UTIL.setIE9PlaceHolder($input1);
    }
    else if (detailType === 'telecom'){

        let $select1 = $($selects.get(0));
        let $select2 = $($selects.get(1));

        let $input1 = $($inputs.get(0));
        let $input2 = $($inputs.get(1));

        $select1.attr(placeHolderAttrName, placeholders[0]);
        $select2.attr(placeHolderAttrName, placeholders[1]);

        $input1.attr(placeHolderAttrName, placeholders[2]);
        $input2.attr(placeHolderAttrName, placeholders[3]);

        UTIL.setIE9PlaceHolder($input1);
        UTIL.setIE9PlaceHolder($input2);
    }

    return this;
}

module.exports = BxInputTel;
