// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/text.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputText
 */

/**
 * @class BxInputText
 * @classDesc BxInputText 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 *  <bx-input type="text"
 *            shape="line"
 *            class="test-form"
 *            id="shape"
 *            label="shape"
 *            bx-render="function(){console.log('render')}"
 *            bx-click="function(){console.log('click')}">
 *  </bx-input>
 *
 *  <bx-input type="password"
 *            shape="round"
 *            class="text-form"
 *            id="password"
 *            name="password"
 *            label="password"
 *            hinting="6~20자리 영문자 혹은 숫자로 입력해주세요."
 *            validateOption="{events: ['blur'], validator: formMethod.validator}">
 *  </bx-input>
 *
 *  <bx-input type="text"
 *            shape="round"
 *            class="text-form"
 *            id="custom"
 *            name="custom"
 *            label="custom"
 *            hinting="주민번호 custom 입니다."
 *            formatter="{type:'custom', options:{reg:/(\d{6})([1-4]\d{0})/, pattern: '$1-$2'}}">
 *            <!--formatter="{type:'currency', options:{separator:',', prefix: '$', decimalPlaces: 4}}">-->
 *            <!--formatter="pid">-->
 *  </bx-input>
 *
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxInput().initialize();
 *
 *      // 엘리먼트로 접근하여 property를 가져올 수 있다.
 *      var inputElem = document.getElementById('shape');
 *
 *      console.log(inputElem.label);
 *      console.log(inputElem.shape);
 *      console.log(inputElem.icons);
 *      console.log(inputElem.inactive);
 *      console.log(inputElem.bxRender);
 *
 *  });
 *
 *  window.formMethod = {
 *
 *      validator: function(value){
 *
 *          var reg = /[a-zA-Z0-9]{6,20}/g;
 *          var result = reg.test(value);
 *
 *          if (!result){
 *              $().BxDialog().alert({
 *                  title: 'error',
 *                  content: '비밀번호는 6~20자리의 영문 또는 숫자만 가능합니다.'
 *              });
 *          }
 *
 *          return result;
 *      }
 *  };
 *
 * </script>
 */

/**
 * @property {object} bx-input - text 타입에서 사용하는 attributes
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {object} [bx-input.icons=null] - <span class='important'>readonly</span> - icon의 class 설정 및 click 이벤트에 대한 handler 설정 객체
 * @property {object} [bx-input.icons.left=null] - <span class='important'>readonly</span> - input 왼쪽의 icon 설정 객체<br/><span class='important'>* label과 함께 적용 시, 왼쪽 아이콘은 노출되지 않는다.</span>
 * @property {string} [bx-input.icons.left.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.left.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {object} [bx-input.icons.right=null] - <span class='important'>readonly</span> - input 오른쪽의 icon 설정 객체
 * @property {string} [bx-input.icons.right.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.right.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {function} [bx-input.icons.right.onClick=null] - <span class='important'>readonly</span> - icon click 이벤트 callback 함수
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {object} [bx-input.validateOption=null] - 입력값에 대한 유효성 검사 설정 객체
 * @property {object|string} [bx-input.formatter=null] - <span class='important'>readonly</span> - 입력값에 대한 formatting 설정
 * @property {string} [bx-input.formatter.type - 해당 셀의 formatting 형식을 지정<br/>
 * currency - 통화<br/>
 * date - 날짜<br/>
 * pid - 주민등록번호<br/>
 * bzno - 사업자등록번호<br/>
 * tel - 전화번호(핸드폰번호)<br/>
 * card - 카드번호<br/>
 * exp - 카드유효기간<br/>
 * pct - %<br/>
 * zip - zipCode<br/>
 * custom - 사용자 설정 <span class="dec">* custom 타입은 string 형식을 지원하지 않습니다.</span><br/>
 * @property {object} [bx-input.formatter.options = {}]
 * @property {string} [bx-input.formatter.options.separator = '-'] - 구분자로 사용할 값을 설정한다.<br/>
 *      * <span class="dec">exp(/), currency(,), pct(%)</span> 는 default값을 예외로 둔다.<br/>
 *      * <span class="dec">custom, date, zipcode</span> 는 지원하지 않는다.
 * @property {string} bx-input.formatter.options.reg - <span class="dec">formatter.type</span> 이 <span class="dec">'custom'</span> 인 경우 사용이 가능하며, 정규식에 사용할 reg를 설정한다.
 * @property {string} bx-input.formatter.options.pattern - <span class="dec">formatter.type</span> 이 <span class="dec">'custom'</span> 인 경우 사용이 가능하며, 정규식에 사용할 pattern을 설정한다.
 * @property {string} [bx-input.formatter.options.prefix = ''] - <span class="dec">formatter.type</span> 이 <span class="dec">'currency'</span> 인 경우 사용이 가능하며, 가장 앞에 붙일 문자를 설정한다.
 * @property {string} [bx-input.formatter.options.suffix = ''] - <span class="dec">formatter.type</span> 이 <span class="dec">'currency'</span> 인 경우 사용이 가능하며, 가장 뒤에 붙일 문자를 설정한다.
 * @property {int} [bx-input.formatter.options.decimalPlaces = 0] - <span class="dec">formatter.type</b>이 <span class="dec">'currency'</span> 인 경우 사용이 가능하며, 소수점 출력 자릿수를 설정한다.
 * @property {string} bx-input.formatter.options.viewFormat - <span class="dec">formatter.type</span> 이 <span class="dec">'date'</span> 인 경우 사용이 가능하며, 날짜 데이터의 형식을 설정한다.
 *     yyyy - 년도 (4자리)<br/>
 *     MM - 월<br/>
 *     dd - 일<br/>
 *     hh - 시간<br/>
 *     mm - 분<br/>
 *     ss - 초<br/>
 * @property {string} [bx-input.originValue = ''] - <b class="important">formatter 사용 시,</b> formatter 가 적용되지않은 원본 값을 반환한다.
 * @property {object[]} [bx-input.validateOption.events=null] - 유효성 검사를 발생할 이벤트 명 배열
 * @property {function} [bx-input.validateOption.validator=null] - 유효성 결과 값을 리턴하는 handler
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.placeholder=''] - <span class='important'>readonly</span> - placeholder 값
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 * @memberOf BxInputText
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const OPTS = {
    "shape": 'default',
    "icons": {left: null, right: null},
    "hinting": '',
    "validateOption": null,
    "formatter": null
};

const READONLY_ATTRS = ['shape', 'label', 'icons', 'hinting', 'formatter'];

const ATTR_NAMES = {
    "validateoption": 'validateOption',
    "originvalue": 'originValue'
};

class BxInputText extends BASE_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        this.readOnlyAttrs = READONLY_ATTRS;
        this.attrNames = ATTR_NAMES;
    }
    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputText, 'opts', 'type');
        }

        let html = [];

        let label = opts.label;
        let icons = opts.icons;
        let hinting = opts.hinting;

        // 전달받은 class 값과 bx-input class 를 병합시킨다.
        opts = super.concatInputDefaultClassName(opts);

        let attr = UTIL.objectToStringAttr(opts.attr);

        // wrapper 엘리먼트를 반환한다.
        let $wrapper = super.getWrapperElement(opts);

        if (icons){
            // icon 관련 class 를 반환한다.
            $wrapper.addClass(super.getIconClassNames(opts));
        }

        if (!label) $wrapper.addClass('input-single');

        if (label){
            html.push(`<label class="bx-label" for="${opts.attr.id}">${label}</label>`);
        }

        if (!label && icons && icons.left){

            let icon = icons.left;

            html.push(super.createIconElement(icon.className, icon.title));
        }

        html.push(`<input ${attr} />`);

        if (icons && icons.right){

            let icon = icons.right;
            let className = icon.className + ' ' + this.inputSelector.bxActionIcon;

            html.push(super.createIconElement(className, icon.title));
        }

        if (hinting){
            html.push(`<span class="${this.inputSelector.bxHinting}">${hinting}</span>`);
        }

        return $wrapper.append(html.join(''));
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        opts = super.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'icons' || k === 'validateOption'){

                v = UTIL.parseObject(k, v);

                if (!_.isPlainObject(v)){
                    CONSOLE.prop.throw(BxInputText, k, 'type');
                }

            }

            if (k === 'inactive'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputText, k, 'type');
                }
            }

            if (k === 'formatter'){

                if (TYPE.isPlainObject(v)){

                    v = UTIL.parseObject(k, v);

                    if (!_.isPlainObject(v)){
                        CONSOLE.prop.throw(BxInputText, k, 'type');
                    }
                }
            }

            return v;
        });

        return opts;
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){

        return super.setElementProperties($elem, opts, READONLY_ATTRS, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) => {

                    if (k === 'validateOption'){

                        v = UTIL.parseObject(k, v);

                        if (!_.isPlainObject(v)){
                            CONSOLE.prop.throw(BxInputText, k, 'type');
                        }

                        super.setInputValidator($elem, v);
                    }

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputText, k, 'type');
                        }

                        super.setInactive($elem, v);
                    }

                    elem[_k] = v;
                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     *
     * @returns 생성된 인스턴스 객체
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputText, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputText, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputText, 'opts', 'type');
        }

        if (opts.validateOption) super.setInputValidator($elem, opts.validateOption);
        if (opts.formatter) super.setInputFormatter($elem, opts.formatter);

        UTIL.setIE9PlaceHolder($elem.find('input'));

        super.setContainerPadding($elem, opts);
        super.setInputFocusHandler($elem);
        super.setInputBlurHandler($elem);

        if (opts.icons.right && opts.icons.right.onClick){
            super.setInputIconHandler($elem, opts.icons.right.onClick);
        }

        super.setInactive($elem.find('input'), opts.inactive);

        return this;
    }
}


module.exports = BxInputText;
