// bxuip-angular.js-ui/src/lib/bx-ui/app/components/input/timepicker.js
/**
 * DOM이 생성된 이후 (DOM Ready) 초기화 함수를 호출한다.
 *
 * @method initialize
 * @memberOf BxInputTimePicker
 */

/**
 * @class BxInputTimePicker
 * @classDesc BxInputTimePicker 클래스 생성자
 *
 * @returns {object} 생성된 인스턴스
 *
 * @example
 *
 * <bx-input
 *       id="picker1"
 *       label="timePicker"
 *       type="timePicker"
 *       pickerType="bubble"
 *       show24Hours="true"
 *       minHours="0"
 *       maxHours="13"
 *       formatter="hh:mm:ss"
 *       defaultTime="00:00:00 (AM or PM)"
 *       steps = "{hour: 3, min: 30, sec: 30}"
 *       bx-changed="function(time){ console.log(time); }"
 *       bx-show="function(){ console.log(this); }"
 *       bx-hide="function(){ console.log(this); }"
 *       bx-render="function(){ console.log(this); }">
 * </bx-input>
 *
 * <script>
 *
 *  $(document).ready(function(){
 *
 *      // DOM 이 ready 된 시점에서 초기화 함수를 호출한다.
 *      $().BxInput().initialize();
 *
 *      var pickerElem = document.getElementById('picker1');
 *
 *      console.log(pickerElem.pickerType);
 *      console.log(pickerElem.show24Hours);
 *      console.log(pickerElem.formatter);
 *
 *  });
 *
 * </script>
 */
/**
 * @property {object} bx-input - timePicker 타입에서 사용하는 attributes <br/><span class="important">관련 API 내용은 http://bxuip.bwg.co.kr/#/uiComponents/timepicker 문서를 참조하시기 바랍니다.</span>
 * @property {string} [bx-input.shape=default] - <span class='important'>readonly</span> - input 디자인 타입(default, round, line, clear)
 * @property {string} [bx-input.label=''] - <span class='important'>readonly</span> - label의 text 값
 * @property {object} [bx-input.icons=null] - <span class='important'>readonly</span> - icon의 class 설정 및 click 이벤트에 대한 handler 설정 객체
 * @property {object} [bx-input.icons.left=null] - <span class='important'>readonly</span> - input 왼쪽의 icon 설정 객체<br/><span class='important'>* label과 함께 적용 시, 왼쪽 아이콘은 노출되지 않는다.</span>
 * @property {string} [bx-input.icons.left.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.left.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {object} [bx-input.icons.right=null] - <span class='important'>readonly</span> - input 오른쪽의 icon 설정 객체
 * @property {string} [bx-input.icons.right.className=null] - <span class='important'>readonly</span> - icon class 선택자 명
 * @property {string} [bx-input.icons.right.title=null] - <span class='important'>readonly</span> - icon 대체 텍스트
 * @property {function} [bx-input.icons.right.onClick=null] - <span class='important'>readonly</span> - icon click callback 함수
 * @property {string} [bx-input.hinting=''] - <span class='important'>readonly</span> - hinting 메세지 값
 * @property {boolean} [bx-input.inactive=false] - 비활성화 여부
 * @property {string} [bx-input.placeholder=''] - <span class='important'>readonly</span> - placeholder 값
 * @property {string} [bx-input.pickerType=null] - <span class='important'>readonly</span> - timePicker 타입(default, bubble, round, line)
 * @property {boolean} [bx-input.show24Hours=false] - <span class='important'>readonly</span> - 24 시간제 사용 유/무(`false`: 12 시간제)
 * @property {int} [bx-input.minHours=0] - <span class='important'>readonly</span> - 최소 시간(`default` type 은 적용되지 않는다)
 * @property {int} [bx-input.maxHours=23] - <span class='important'>readonly</span> - 최대 시간(`default` type 은 적용되지 않는다)
 * @property {string} [bx-input.formatter=각 type 에 따른, 기본 값] - <span class='important'>readonly</span> - `시간 표기`에 대한, formatter(h:m:s or hh:mm:ss)
 * @property {string} [bx-input.defaultTime=각 시간제에 따른, 기본 값] - <span class='important'>readonly</span> - 기본 시간(12:00:00)
 * @property {object} [bx-input.steps]
 * @property {int} [bx-input.steps.hour = 1] - <span class='important'>readonly</span> - 시간 간격(`default` 타입은 적용되지 않는다)
 * @property {int} [bx-input.steps.min = 1] - <span class='important'>readonly</span> - 분 간격
 * @property {int} [bx-input.steps.sec = 1] - <span class='important'>readonly</span> - 초 간격(`default` 타입은 적용되지 않는다)
 * @property {function} [bx-input.bx-render=function(){}] - render 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-changed = function(){}] - 시간 또는 meridian 변경 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-show = function(){}] - show 메서드 호출 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-hide = function(){}] - hide 메서드 호출 시, 호출될 callback 함수
 * @property {function} [bx-input.bx-[eventName]=function(){}] - eventName(click, focus 등) 에 해당되는, 네이티브 이벤트를 바인딩시킨다.
 *
 * @alias Attributes
 * @memberOf BxInputTimePicker
 */

const BXUI = require('../../../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE_CLASS = require('./base');
const TEXT_CLASS = require('./text');

const {Console: CONSOLE, Util: UTIL, Type: TYPE} = BXUI;

const OPTS = {
    "pickerType": 'default',
    "show24Hours": true,
    "minHours": 0,
    "maxHours": 23,
    "formatter": '',
    "defaultTime": '',
    "steps": {
        "hour": 1,
        "min": 1,
        "sec": 1
    },
    "bx-changed": function (){},
    "bx-show": function (){},
    "bx-hide": function (){}
};

let readOnlyAttrs = [
    'pickerType',
    'show24Hours',
    'minHours',
    'maxHours',
    'defaultTime',
    'steps'
];

const ATTR_NAMES = {
    "pickertype": 'pickerType',
    "show24hours": 'show24Hours',
    "minhours": 'minHours',
    "maxhours": 'maxHours',
    "defaulttime": 'defaultTime'
};

class BxInputTimePicker extends TEXT_CLASS {

    constructor(){

        super();

        this.opts = _.assign(this.opts, OPTS);

        this.base = new BASE_CLASS();

        readOnlyAttrs = _.union(this.readOnlyAttrs, readOnlyAttrs);
    }
    /**
     * 컴포넌트 엘리먼트를 생성한다.
     *
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 컴포넌트 엘리먼트
     *
     * @private
     */
    createComponentElement(opts = {}){

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputTimePicker, 'opts', 'type');
        }

        let $elem = super.createComponentElement(opts);

        return $elem;
    }
    /**
     * 컴포넌트 옵션을 생성한다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 옵션 객체
     *
     * @private
     */
    createComponentOptions($elem = null, opts = {}){

        opts = this.base.createComponentOptions($elem, opts, ATTR_NAMES, (k, v) => {

            if (k === 'icons' || k === 'steps'){

                v = UTIL.parseObject(k, v);

                if (!_.isPlainObject(v)){
                    CONSOLE.prop.throw(BxInputTimePicker, k, 'type');
                }
            }

            if (k === 'inactive' || k === 'show24Hours'){

                if (!_.isBoolean(v)){
                    CONSOLE.prop.throw(BxInputTimePicker, k, 'type');
                }
            }

            if (k === 'minHours' || k === 'maxHours'){

                if (!TYPE.isNumber(v)){
                    CONSOLE.prop.throw(BxInputTimePicker, k, 'type');
                }
            }

            return v;
        });

        return opts;
    }
    /**
     * 컴포넌트 프로퍼티를 바인딩 시킨다.
     *
     * @param {object} $elem - 바인딩될 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} Base 객체
     *
     * @private
     */
    setElementProperties($elem = null, opts = {}){


        return this.base.setElementProperties($elem, opts, readOnlyAttrs, (elem, k) => {

            let _k = UTIL.convertInternalPropertyName(k);

            return {
                configurable: true,
                enumerable: true,
                get: () => {
                    return elem[_k];
                },
                set: (v) =>{

                    if (k === 'inactive'){

                        if (!_.isBoolean(v)){
                            CONSOLE.prop.throw(BxInputTimePicker, k, 'type');
                        }

                        this.base.setInactive($elem, v);
                    }

                    elem[_k] = v;

                    if (k.indexOf('bx') !== -1 && OPTS[_.replace(k, 'bx', 'bx-')]){

                        let $parent = $elem.parent().parent();

                        $parent.find(`.${this.inputSelector.bxTimepicker}`).detach();
                        opts.defaultTime = $elem.val();
                        _initTimePicker.apply(this, [$parent, opts]);

                    }

                }
            };
        });
    }
    /**
     *
     * 이벤트 핸들러 바인딩 함수
     *
     * @param {object} $elem - 엘리먼트 객체
     * @param {object} opts - 컴포넌트 옵션
     *
     * @returns {object} 생성된 인스턴스
     *
     * @private
     */
    init($elem = null, opts = {}){

        if (_.isNull($elem)){
            CONSOLE.prop.throw(BxInputTimePicker, '$elem', 'type');
        }

        if (!_.isFunction($elem.get) || !_.isElement($elem.get(0))){
            CONSOLE.prop.throw(BxInputTimePicker, '$elem', 'type');
        }

        if (!_.isPlainObject(opts)){
            CONSOLE.prop.throw(BxInputTimePicker, 'opts', 'type');
        }

        _initComponent.apply(this, [$elem, opts]);

        // 이후 기능 추가로 들어갈 기능
        //UTIL.setIE9PlaceHolder($elem.find('input'));

        this.base.setContainerPadding($elem, opts);

        this.base.setInputFocusHandler($elem);
        this.base.setInputBlurHandler($elem);

        if (opts.icons.right && opts.icons.right.onClick){

            this.base.setInputIconHandler($elem, opts.icons.right.onClick);
        }

        this.base.setInactive($elem.find('input'), opts.inactive);

        return this;
    }
}


/**
 * timePicker 컴포넌트 초기화 함수
 *
 * @param {object} $elem - 엘리먼트 객체
 * @param {object} opts - 컴포넌트 옵션
 *
 *
 * @returns {object} 생성된 인스턴스
 *
 * @private
 */
function _initComponent($elem = null, opts = {}){


    let $input = $elem.find(`.${this.inputSelector.bxInput}`);
    let input = $input.get(0);

    // timePicker 컴포넌트를 초기화한다.
    let instance = _initTimePicker.call(this, $elem, opts);

    let APIS = {
        getTime: instance.getTime.bind(instance),
        getShow24Hours: instance.getShow24Hours.bind(instance),
        getAM: instance.getAM.bind(instance),
        getPM: instance.getPM.bind(instance),
        getHour: instance.getHour.bind(instance),
        getMin: instance.getMin.bind(instance),
        getSec: instance.getSec.bind(instance),
        getFormatter: instance.getFormatter.bind(instance),
        setTime: instance.setTime.bind(instance),
        setAM: instance.setAM.bind(instance),
        setPM: instance.setPM.bind(instance),
        setHour: instance.setHour.bind(instance),
        setMin: instance.setMin.bind(instance),
        setSec: instance.setSec.bind(instance),
        toggle: instance.toggle.bind(instance),
        show: instance.show.bind(instance),
        hide: instance.hide.bind(instance)
    };

    // 전달받은 opts 객체의 formatter 값을 다시 초기화시킨다.
    if (!_.trim(opts.formatter).length){
        opts.formatter = instance.initFormatter(opts.pickerType);
    }

    _.map(APIS, (v, k) =>{
        UTIL.setReadOnlyProperty(input, k, v);
    });
}

function _initTimePicker($elem = null, opts = {}){
    let $input = $elem.find(`.${this.inputSelector.bxInput}`);
    let input = $input.get(0);

    let pickerOpts = {

        type: opts.pickerType,
        show24Hours: opts.show24Hours,
        minHours: opts.minHours ? _.toNumber(opts.minHours) : opts.minHours,
        maxHours: opts.maxHours ? _.toNumber(opts.maxHours) : opts.maxHours,
        formatter: opts.formatter,
        defaultTime: opts.defaultTime,
        steps: opts.steps,
        onChanged: $input.prop('bxchanged').bind(input),
        onShow: $input.prop('bxshow').bind(input),
        onHide: $input.prop('bxhide').bind(input)
        // @deprecated.(차후 제거될 이벤트 핸들러)
        //onInitialize: opts['bxinitialize']
    };

    // timePicker 컴포넌트를 초기화한다.
    return  $elem.bxTimePicker(pickerOpts).render();
}


module.exports = BxInputTimePicker;
