// bxuip-angular.js-ui/src/lib/bx-ui/app/base.js
/**
 * Created by sgjeon on 16. 8. 19..
 */

/**
 *
 * @namespace BXUI
 *
 */

const BXUI = require('../../bx-ui-core/entry');
const $ = require('jquery');
const _ = require('lodash');

const BASE = _.assign({}, {

    register: _register,
    appendContainerElement: _appendContainerElement,
    ngCompile: _ngCompile,
    mixin: _mixin
});


/**
 * 컴포넌트 등록 함수
 *
 * @param {String} fnName - JQuery.fn 객체에 할당될 함수명
 * @param {Function}[constructor=function(){}] constructor - 생성자 함수 객체
 *
 * @returns {object} 생성된 컴포넌트 객체
 *
 * @private
 *
 * @memberof BXUI
 */
function _register(fnName = '', constructor = function (){
}){

    if (!_.isString(fnName) || fnName === ''){
        throw new Error('fnName 값이 존재하지 않습니다.');
    }

    // JQuery 프로토타입 객체에 할당될 메서드 이름
    if ($.fn[fnName]){
        throw new Error('이미 존재하는 메서드입니다.');
    }

    if (!_.isFunction(constructor)){
        throw new Error('constructor 값이 존재하지 않습니다.');
    }

    // constructor 를 변경한다.
    Object.defineProperty(constructor.prototype, 'constructor', {
        enumerable: false,
        configurable: true,
        writable: false,
        value: constructor
    });

    // 커스텀 엘리먼트를 구별하기 위한 프로퍼티값
    let isCustomElement = constructor.isCustomElement;
    let fnNames = [fnName];

    if (/^bx[a-zA-Z]+/.test(fnName)){
        fnNames.push(fnName.replace('bx', 'Bx'));
    }

    // bx or Bx 로 시작하는 함수를 등록한다.(기존 사용자를 위해, 생성자 함수를 두 가지로 등록시킨다)
    _.forEach(fnNames, (v, k) => {
        registerComponentMethod(v);
    });

    // 컴포넌트 생성자 함수를 등록한다.
    function registerComponentMethod(fnName = ''){

        $.fn[fnName] = componentMethod;

        if (_.isBoolean(isCustomElement)){
            // `커스텀 엘리먼트`의 경우, jquery 생성자 함수안에 동일한 프로퍼티값을 할당시켜놓는다.
            $.fn[fnName]['isCustomElement'] = isCustomElement;
        }
    }

    function componentMethod(){

        // 생성자 함수로 전달할 arguments 집합을 생성한다.
        let args = [].slice.call(arguments, 0);

        let root = this.get(0) || document.body;

        // 컴포넌트 $root 엘리먼트
        let $root = $(root);

        // 컴포넌트 이름
        let componentName = fnNames.length === 2 ? fnNames[1] : fnNames[0];

        // 컴포넌트 className
        let componentClassName = fnName.replace(/^bx|^Bx/, '');
        componentClassName = componentClassName.replace(/([A-Z])/g, '-$1').toLowerCase();
        componentClassName = componentClassName[0] === '-' ? componentClassName.substr(1) : componentClassName;

        // 사용자로부터 전달받은 $root 엘리먼트
        constructor.$root = $root;
        constructor.componentName = componentName;
        constructor.className = componentClassName;

        // 각 컴포넌트에 대한 인스턴스 생성
        // null 을 Object.create(null) === 빈 객체로 대치시키자
        let o = new (Function.prototype.bind.apply(constructor, [null].concat(args)));

        _.assign(o, {
            "cid": _.uniqueId(fnName + '_'),
            "$root": $root,
            "componentName": componentName,
            "componentClassName": componentClassName
        });

        return o;
    }
}


/**
 * $container 엘리먼트 생성 함수
 *
 * @param {Object} $root - root 엘리먼트
 * @param {String} componentName - class 에 할당될 컴포넌트 이름
 * @param {Object}[$elem=null] - 대치시킬 엘리먼트
 *
 * @returns {object} - 생성된 $container 엘리먼트
 *
 * @private
 *
 * @memberof BXUI
 */
function _appendContainerElement($root = null, className = '', $elem = null){

    if (_.isNull($root)){
        throw new Error('$root 엘리먼트가 존재하지 않습니다.');
    }

    if (!_.isString(className) || className === ''){
        throw new Error('className 값이 존재하지 않습니다.');
    }

    let componentClassName = BXUI.Base.getComponentClassName(className);

    // 라이브러에서 지정된 tagName 이 있는 경우(grid 라이브러리의 경우, table 엘리먼트만이 root 엘리먼트로 가능하다)
    if ($elem){

        if (!_.isElement($elem.get(0))){
            throw new Error('$elem 엘리먼트가 존재하지 않습니다.');
        }

        let _class = $elem.attr('class') || '';

        $elem.attr('class', _class + ' ' + componentClassName);

        $root.append($elem);
    }
    else{
        $root.append('<div class="' + componentClassName + '"></div>');
    }

    // $root 엘리먼트의 자식으로 해당 컴포넌트가 여러개 할당되었을경우, 그중 가장 마지막 자식 (컴포넌트)엘리먼트를 반환해준다.
    let $containers = $root.children('.' + componentClassName);
    let size = $containers.length;
    size = size > 0 ? size - 1 : 0;

    return $($containers.get(size));
}


/**
 * angular 컴파일 함수
 *
 * @param {string | Object} arg - 컴파일시킬 html 또는 엘리먼트
 * @param {Object} $compile - 컴파일 서비스 객체
 * @param {Object} $scope - linking 시킬 $scope 객체
 *
 * @returns {string} - 컴파일된 html 문자열
 *
 * @private
 *
 * @memberof BXUI
 */
function _ngCompile(arg = null, $compile = null, $scope = null){

    if (_.isNull(arg)){
        throw new Error('arg 값이 존재하지 않습니다.');
    }

    if (_.isString(arg) && _.trim(arg).length === 0){
        throw new Error('arg 값이 존재하지 않습니다.');
    }

    if (_.isNull($compile)){
        throw new Error('$compile 값이 존재하지 않습니다.');
    }

    if (_.isNull($scope)){
        throw new Error('$scope 값이 존재하지 않습니다.');
    }

    return $compile(arg)($scope);
}


/**
 * mixin 함수
 *
 * @param {Class} superClass
 *
 * @private
 *
 * @memberOf BXUI
 */

function _mixin(superClass){

    class Mix {

        constructor(superClass){
            this.superclass = superClass;
        }

        with(...mixins){
            return mixins.reduce((c, mixin) => mixin(c), this.superclass);
        }
    }

    return new Mix(superClass);
}

//function _mixin(...args) {
//
//    class Mix {}
//
//	_.map(args, v => {
//
//		if (_.isPlainObject(v)){
//			_copy(Mix.prototype, v);
//		}
//		else{
//	        _copy(Mix, v);
//	        _copy(Mix.prototype, v.prototype);
//		}
//    });
//
//
//
//    function _copy(target, source) {
//
//        _.map(Object.getOwnPropertyNames(source), v => {
//
//			// constructor 를 통해, 해당 static method 들도 전부 복사된다.
//			// Mix 클래스의 constructor, prototype, name 프로퍼티는 복사하지 않는다.
//            if (
//	                v !== 'constructor' &&
//	                v !== 'prototype' &&
//	                v !== 'name'
//				){
//
//                let _v = Object.getOwnPropertyDescriptor(source, v);
//
//                // property 를 할당한다.
//                Object.defineProperty(target, v, _v);
//            }
//        });
//    }
//
//    return Mix;
//}


module.exports = BASE;
