// bxuip-builder-core/src/server.js
const path = require('path');

const _ = require('lodash');
const open = require('open');

const hound = require('hound');
const chalk = require('chalk');

const webpack = require('webpack');
const WebpackDevServer = require('webpack-dev-server');

const SingleEntryPlugin = require('webpack/lib/SingleEntryPlugin');

const Config = require('./lib/config');
const Common = new (require('./lib/common'));

const envArgs = Common.getEnvArgs();
/**
 * 파일 추가가 가능한 개발 서버
 */
class Server extends Config{

    constructor({
        webpackConfig = {}
    } = {}){

        super();

        this.isBrowserOpen = false;
        this.serverInfo = _getServerInfo.call(this);

        this.webpackConfig = webpackConfig;

        this.compiler = null;
    }
    /**
     * 서버를 시작한다.
     */
    start(){

        const serverInfo = this.serverInfo;

        // 파일 수정 시, 브라우저 새로고침을 위한 코드.
        const webpackConfig = this.webpackConfig;
        _.map(webpackConfig.entry, v => { v.push(`webpack-dev-server/client?${serverInfo.uri}`); });

        this.compiler = webpack(webpackConfig);

        const devConfig = {
            stats: { colors: true },
            watchOptions: {
              aggregateTimeout: 300,
              poll: 1000
            },
            proxy: serverInfo.proxy
        };

        if (envArgs.verbose) devConfig.stats = 'verbose';

        const server = new WebpackDevServer(this.compiler, devConfig);

        server.listen(serverInfo.port, serverInfo.host, () => {
            console.log(chalk.blueBright(`Project is running at ${serverInfo.uri}/`));
        });

        this.compiler.plugin('done', () => {

            if (!this.isBrowserOpen){

                // 브라우저 오픈
                open(serverInfo.uri);

                this.isBrowserOpen = true;
            }
        });

        this.addWatchEvent();
    }
    /**
     * (파일 추가를 위한)왓치 이벤트를 등록한다
     */
    addWatchEvent(){

        const builderConfig = this.builder;
        const builderPath = builderConfig.path;

        hound.watch(builderPath.src).on('create', (f, stats) => {
        	
            const parse = path.parse(f);
            const dir = parse.dir;
            const name = parse.name;
            const ext = parse.ext;
            console.log(dir, name, ext);
            let entryKey = path.normalize(`${dir}/${name}`);
            entryKey = entryKey.replace(path.normalize(`${builderPath.src}/`), '');

            if (ext === '.js'){
                // 동적으로 파일을 추가하기위해, webpack 내부 플러그인을 사용했다.
                this.compiler.apply(new SingleEntryPlugin(path.normalize(`${dir}`), f, entryKey));
            }

            // 아래 소스 살리면 무한루프 돌게 됩니다..
//            this.compiler.run(() => {});
        });

        //hound.watch(builderPath.src).on('delete', (f, stats) => {
        //
        //    console.log('deletedeletedeletedeletedeletedeletedeletedeletedeletedeletedelete');
        //
        //    const parse = path.parse(f);
        //    const dir = parse.dir;
        //    const name = parse.name;
        //
        //    let entryKey = path.normalize(`${dir}/${name}`);
        //    entryKey = entryKey.replace(path.normalize(`${builderPath.src}/`), '');
        //
        //    this.compiler.apply(new RemoveChunkEntryPlugin({chunks: [entryKey]}));
        //
        //    this.compiler.run(() => { console.log('compilercompilercompilercompilercompiler'); });
        //});
    }
}

/**
 * 서버 정보를 반환한다.
 * @returns {{host: string, port: number, proxy: {}, uri: *}}
 * @private
 */
function _getServerInfo(){

    // dev server 설정
    let host = 'localhost';
    let port = 8088;
    let proxy = {};

    const appConfig = this.app;

    if (_.isPlainObject(appConfig.devServer)){

        const devServer = appConfig.devServer;

        host = devServer.host || host;
        port = devServer.port || port;

        proxy = devServer.proxy || proxy;
    }

    const uri = `http://${host}:${port}`;

    return {host, port, proxy, uri};
}

module.exports = Server;
