// bxuip-builder-core/src/config/webpack.base.config.js
/**
 * Created by UI/UX Team on 2018. 2. 14..
 */

const path = require('path');

const _ = require('lodash');

const webpack = require('webpack');

// builder, app, env config
const {builder: builderConfig, app: appConfig, env: envConfig} = new (require('../lib/config'));
const Common = new (require('../lib/common'));
const Entry = new (require('../lib/entry'));

const {Config: WebpackConfig, environment} = require('webpack-config');

const FixModuleIdAndChunkIdPlugin = require('fix-moduleid-and-chunkid-plugin');

const CleanWebpackPlugin = require('clean-webpack-plugin');
const CopyWebpackPlugin = require('copy-webpack-plugin');

const HtmlWebpackPlugin = require('html-webpack-plugin');
const InjectAssetsVersionPlugin = require('./plugins/injectAssetVersionPlugin');
const InjectTargetFilePlugin = require('./plugins/injectTargetFilePlugin');

const builderPath = builderConfig.path;

const appContext = appConfig.context;

const envArgs = Common.getEnvArgs();
// -----------------------------------------------------------------------------
// -----------------------------------------------------------------------------
// config
// -----------------------------------------------------------------------------
// -----------------------------------------------------------------------------
const config = {
    // 설정된 context 으로, watch 영역이 할당된다.
    context: builderPath.root,
    entry: Entry.create(),
    output: {
        filename: '[name].js',
        path: builderPath.build
    },
    module: {
        rules: [
        {
            test: /\.js$/,
            include: builderPath.app,
            use: {
                loader: 'babel-loader',
                options: {
                    cacheDirectory: true,
                    presets: ['es2015'],
                    plugins: ['transform-runtime']
                }
            }
        },
        {
            test: /\.html$/,
            exclude: path.join(builderPath.src, 'index.html'),
            use: {
              loader: 'html-loader'
            }
        }]
    },
    stats: (() => {

        let ret = 'normal';

        if (envArgs.verbose) ret = 'verbose';

        return ret;
    })(),
    cache: true,
    resolve: {
        modules: ['node_modules'],
        /*
         모듈명 뒤에 여기 명시된 확장자명들을 붙여보며 탐색을 수행한다. 즉 아래처럼 extensions: ['.js', '.css']
         으로 설정되어 있으면 require('abc')를 resolve 하기 위해 abc, abc.js, abc.css 로 탐색한다.
         */
        extensions: ['.js']
    },
    plugins: [
        new CleanWebpackPlugin(builderPath.build, {root: builderPath.root, verbose: false, watch: false}),
        // chunkId 를 가변 index(숫자)값에서, 고정된값인 파일 경로로 고정 시킨다.(bundle 활성화시 반드시 필요한 기능이다)
        new FixModuleIdAndChunkIdPlugin(),
        new CopyWebpackPlugin((() => {

            const ret = [];

            const copy = appConfig.copy;

            if (_.isArray(copy)){

                _.forEach(copy, (v) => {

                    const context = v.context || appContext;

                    const form = path.join(context, v.from);
                    const to = v.to;
                    const ignore = v.ignore;

                    ret.push({context: context, from: form, to: to, ignore: ignore});
                });
            }

            return ret;

        })(),
        {
            // watch or webpack-dev-server build 시, 수정된 파일들만 복사한다.
            copyUnmodified: false
        }),
        new HtmlWebpackPlugin((() => {
            return _.assign({
                chunksSortMode: 'manual',
                template: 'src/index.html'
            }, envConfig);
        })()),
        new webpack.optimize.CommonsChunkPlugin((() => {

            return {
                name: Entry.chunk.name,
                filename: Entry.chunk.filename,
                minChunks: Entry.chunk.minChunks
            };
        })()),
        new InjectAssetsVersionPlugin(),
        new InjectTargetFilePlugin()
    ]
};


module.exports = new WebpackConfig().merge(config);
