// bxuip-angular.js-ui/config/webpack.dev.config.js
/**
 * Created by UI/UX Team on 2018. 2. 14..
 */

const path = require('path');
const fs = require('fs');

const _ = require('lodash');
const glob = require('glob');

const {Config, environment} = require('webpack-config');

const envConfig = environment.get('config');

const HtmlWebpackPlugin = require('html-webpack-plugin');
const CopyWebpackPlugin = require('copy-webpack-plugin');

const SassBuildPlugin = require('./plugins/sassBuildPlugin');
const AddEntryPlugin = require('./plugins/addEntryPlugin');
const JSDocPlugin = require('./plugins/jsDocPlugin');

const rootPath = path.join(__dirname, '..');

const srcPath = path.join(rootPath, 'src');
const buildPath = path.join(rootPath, 'dist');

const testEntry = glob.sync(path.join(srcPath, 'test/**/*.js'));

const BXUICoreEntry = path.join(srcPath, 'lib/bx-ui-core/entry.js');
const BXUIEntry = envConfig.vendors.concat(glob.sync(path.join(srcPath, 'lib/bx-ui/app/**/*.js')));
const libraryEntry = [BXUICoreEntry].concat(BXUIEntry);

const entry = libraryEntry.concat(glob.sync(path.join(srcPath, 'app/**/*.js')));

const config = {
    entry: {
        test: testEntry,
        "bx-ui-angular": entry
    },
    devServer: {
        contentBase: buildPath,
        host: 'localhost',
        port: '8088',
        proxy: {},
        watchOptions: {
            aggregateTimeout: 300,
            poll: 1000
        },
        open: true
    },
    plugins: [
        new SassBuildPlugin({
            outputFileName: 'bx-ui.css'
        }),
        new HtmlWebpackPlugin({
            hash: true,
            template: path.join(srcPath, 'index.html')
        }),
        new CopyWebpackPlugin([
            {context: srcPath, from: `${path.join(srcPath, 'test/**/*')}`, ignore: ['*.js']}
        ]),
        new AddEntryPlugin({
            entry: glob.sync(path.join(srcPath, 'assets/sass/**/*'), {nodir: true})
        }),
        new JSDocPlugin()
    ]
};

module.exports = new Config().extend(path.join(__dirname, 'webpack.base.config.js')).merge(config);
