// bxuip-angular.js-ui/config/webpack.prod.config.js
/**
 * Created by UI/UX Team on 2018. 2. 14..
 */

const path = require('path');
const fs = require('fs');

const _ = require('lodash');
const glob = require('glob');

const webpack = require('webpack');

const {Config, environment} = require('webpack-config');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
const SassBuildPlugin = require('./plugins/sassBuildPlugin');
const JSDocPlugin = require('./plugins/jsDocPlugin');

const envConfig = environment.get('config');

const rootPath = path.join(__dirname, '..');
const srcPath = path.join(rootPath, 'src');

const BXUICoreEntry = path.join(srcPath, 'lib/bx-ui-core/entry.js');
const BXUIEntry = envConfig.vendors.concat(glob.sync(path.join(srcPath, 'lib/bx-ui/app/**/*.js')));
const libraryEntry = [BXUICoreEntry].concat(BXUIEntry);

const entry = libraryEntry.concat(glob.sync(path.join(srcPath, 'app/**/*.js')));

const config = {
    entry: {
        "bx-ui-angular": entry,
        "bx-ui-angular.min": entry
    },
    plugins: [
        new UglifyJsPlugin({
            include: /\.min\.js$/,
            uglifyOptions: {mangle: false}
        }),
        new SassBuildPlugin({
            outputFileName: 'bx-ui.min.css',
            outputStyle: 'compressed'
        }),
        new JSDocPlugin()
    ]
};

module.exports = new Config().extend(path.join(__dirname, 'webpack.base.config.js')).merge(config);
